diff --git a/sift4g/src/sift_scores.cpp b/sift4g/src/sift_scores.cpp
index d9b8345..f1700de 100644
--- a/sift4g/src/sift_scores.cpp
+++ b/sift4g/src/sift_scores.cpp
@@ -8,10 +8,12 @@
 
 #include <stdio.h>
 #include <math.h>
+#include <string>
 #include <cstring>
 #include <ostream>
 #include <iostream>
 #include <fstream>
+#include <sstream>
 #include <regex>
 #include <iomanip>
 #include <algorithm>
