--- src/pyOpenMS/CMakeLists.txt	2022-02-01 14:39:51.000000000 +0100
+++ src/pyOpenMS/CMakeLists.txt	2022-02-01 15:09:51.000000000 +0100
@@ -660,7 +660,6 @@
 ELSE()
     IF(APPLE)
         add_custom_target(pyopenms
-            COMMAND ./mac_fix_dependencies.rb -l ${CMAKE_BINARY_DIR}/pyOpenMS/pyopenms -e "@rpath/" -f -v
             #COMMAND ${Python_EXECUTABLE} setup.py bdist_egg
             COMMAND ${Python_EXECUTABLE} -m pip wheel -w dist --no-deps --no-binary=pyopenms .
             #COMMAND ${Python_EXECUTABLE} setup.py bdist --format=zip
