--- src/pyOpenMS/CMakeLists.txt	2020-01-27 16:37:16.000000000 +0100
+++ src/pyOpenMS/CMakeLists.txt	2020-01-28 12:49:52.000000000 +0100
@@ -454,7 +454,7 @@
 endforeach()
 	
 # fixup these dependencies (i.e. get other dynamic dependencies recursively)
-if (APPLE) ## On APPLE use our script because the executables need to be relinked
+if (FALSE) ## On APPLE use our script because the executables need to be relinked
 ## this is done before setup.py so that we fixup the resulting pyopenms.so's too
 else()
     ## Assemble common required non-system libraries
