diff --git a/cmake/compiler_flags.cmake b/cmake/compiler_flags.cmake
index 48c2a042ea..6f1b1b0f0a 100644
--- a/cmake/compiler_flags.cmake
+++ b/cmake/compiler_flags.cmake
@@ -103,7 +103,7 @@ elseif (MSVC)
 elseif ("${CMAKE_C_COMPILER_ID}" MATCHES "Clang")
   set(CMAKE_COMPILER_IS_CLANG true CACHE INTERNAL "Is CLang compiler (clang++)")
   # add clang specific warning levels
-  add_compile_options(-Weverything)
+  add_compile_options(-Wall -Wextra -Wno-register)
   # .. and disable some of the harmless ones
   add_compile_options(
                   -Wno-sign-conversion
