--- src/pyOpenMS/MANIFEST.in	2020-01-26 17:08:42.000000000 +0100
+++ src/pyOpenMS/MANIFEST.in	2020-01-28 14:18:49.000000000 +0100
@@ -1,11 +1,5 @@
-graft pyopenms/share
-prune pyopenms/share/OpenMS/examples
-include pyopenms/*.dll
-include pyopenms/*.dylib
-include pyopenms/*.so
+include pyopenms/pyopenms*.dll
+include pyopenms/pyopenms*.dylib
+include pyopenms/pyopenms*.so
 include pyopenms/*.txt
-graft pyopenms/*.framework
-include pyopenms/Qt*
 exclude pyopenms/*.cpp
-global-exclude pyopenms.cpp
-global-exclude HMDB2StructMapping.tsv
