--- ./src/pyOpenMS/setup.py	2018-10-29 20:07:55.000000000 +0100
+++ ./src/pyOpenMS/setup.py	2019-12-18 12:21:21.000000000 +0100
@@ -153,6 +153,7 @@
 elif sys.platform == "darwin":
     # we need to manually link to the Qt Frameworks
     extra_compile_args = ["-Qunused-arguments"]
+    extra_compile_args.append("-isysroot"+os.environ['CONDA_BUILD_SYSROOT'])
 
 if IS_DEBUG:
     extra_compile_args.append("-g2")
