diff -ruN OpenMS-2.8.0-org/src/openms/source/ANALYSIS/ID/SiriusAdapterAlgorithm.cpp OpenMS-2.8.0/src/openms/source/ANALYSIS/ID/SiriusAdapterAlgorithm.cpp
--- OpenMS-2.8.0-org/src/openms/source/ANALYSIS/ID/SiriusAdapterAlgorithm.cpp	2022-05-17 10:03:33.295974032 +0200
+++ OpenMS-2.8.0/src/openms/source/ANALYSIS/ID/SiriusAdapterAlgorithm.cpp	2022-05-17 10:05:40.491967708 +0200
@@ -564,10 +564,6 @@
       // the actual process
       QProcess qp;
       QString executable_qstring = SiriusAdapterAlgorithm::determineSiriusExecutable(executable).toQString();
-      QString wd = File::path(executable).toQString();
-      qp.setWorkingDirectory(wd); //since library paths are relative to sirius executable path
-      //since library paths are relative to sirius executable path
-      qp.setWorkingDirectory(File::path(executable).toQString());
       qp.start(executable_qstring, command_line); // does automatic escaping etc... start
 
       std::stringstream ss;
