diff --git a/src/openms/thirdparty/evergreen/src/FFT/FFT.hpp b/src/openms/thirdparty/evergreen/src/FFT/FFT.hpp
index ba9b34955f..0d644e59ff 100644
--- a/src/openms/thirdparty/evergreen/src/FFT/FFT.hpp
+++ b/src/openms/thirdparty/evergreen/src/FFT/FFT.hpp
@@ -4,7 +4,7 @@
 // #include this file to import all FFT utilities from this
 // subdirectory.
 
-const unsigned char FFT1D_MAX_LOG_N = 31u;
+const unsigned char FFT1D_MAX_LOG_N = 16u;
 
 #include "../BitReversedShuffle/RecursiveShuffle.hpp"
 
diff --git a/src/openms/thirdparty/evergreen/src/Tensor/TensorUtils.hpp b/src/openms/thirdparty/evergreen/src/Tensor/TensorUtils.hpp
index 243bf56996..ff603c09be 100644
--- a/src/openms/thirdparty/evergreen/src/Tensor/TensorUtils.hpp
+++ b/src/openms/thirdparty/evergreen/src/Tensor/TensorUtils.hpp
@@ -8,7 +8,7 @@
 #include "Vector.hpp"
 
 #ifndef MAX_TENSOR_DIMENSION
-#define MAX_TENSOR_DIMENSION 24
+#define MAX_TENSOR_DIMENSION 12
 #endif
 
 // Tuple types:
