--- a/gridss_extract_overlapping_fragments
+++ b/gridss_extract_overlapping_fragments
@@ -206,7 +206,7 @@ for tool in gridsstools samtools ; do
 	fi
 	write_status "Found $(which $tool)"
 done
-samtools_version=$(samtools --version | grep samtools | cut -b 10-)
+samtools_version=$(samtools --version | head -n1 | grep samtools | cut -b 10-)
 write_status "samtools version: $samtools_version"
 samtools_major_version=$(echo $samtools_version | cut -f 1 -d ".")
 samtools_minor_version=$(echo $samtools_version | cut -f 2 -d ".")
