diff --git a/Makefile b/Makefile
index 013275e..ad9330e 100644
--- a/Makefile
+++ b/Makefile
@@ -1,5 +1,5 @@
-CC=g++
-CFLAGS=-O3
+CC?=${CXX}
+CFLAGS=-O3 -std=c++17 -D_LIBCPP_DISABLE_AVAILABILITY -march=x86-64-v3
 LDFLAGS=-static
 
 prog=neff converter
@@ -7,13 +7,13 @@ prog=neff converter
 all: ${prog}
 
 neff: code/neff.cpp
-	${CC} ${CFLAGS} -std=c++17 code/flagHandler.cpp code/common.cpp code/msaReader.cpp code/msaWriter.cpp code/multimerHandler.cpp code/neff.cpp -o neff
+	${CC} ${CFLAGS} code/flagHandler.cpp code/common.cpp code/msaReader.cpp code/msaWriter.cpp code/multimerHandler.cpp code/neff.cpp -o neff
 
 converter: code/converter.cpp
-	${CC} ${CFLAGS} -std=c++17 code/flagHandler.cpp code/common.cpp code/msaReader.cpp code/msaWriter.cpp code/converter.cpp -o converter
+	${CC} ${CFLAGS} code/flagHandler.cpp code/common.cpp code/msaReader.cpp code/msaWriter.cpp code/converter.cpp -o converter
 
 install: ${prog}
 	cp ${prog} ./bin
-	
+
 clean:
 	rm -f ${prog}
diff --git a/setup.py b/setup.py
index c8639a2..ab6fac7 100644
--- a/setup.py
+++ b/setup.py
@@ -1,7 +1,7 @@
 # Downloads C++ code from GitHub, compiles it using a Makefile,
 # and includes the resulting executable files in the `bin` directory of the package to be used by python methods.
 
-from setuptools import setup, find_packages
+from setuptools import setup, find_namespace_packages
 from setuptools.command.build_py import build_py
 import subprocess
 import os
@@ -93,7 +93,7 @@ setup(
     long_description=long_description,
     long_description_content_type="text/markdown",
     url='https://github.com/Maryam-Haghani/NEFFy',
-    packages=find_packages(),
+    packages=find_namespace_packages(),
     include_package_data=True,
     classifiers=[
         "Programming Language :: Python :: 3",
@@ -115,4 +115,4 @@ setup(
         'build_py': CustomBuildPyCommand,
         'bdist_wheel': bdist_wheel,
     },
-)
\ No newline at end of file
+)
