Description: Do not enforce -mavx2
Author: Nilesh Patra <nilesh@debian.org>
Bug-Debian: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=992270
Forwarded: not-needed
Last-Update: 2021-08-18
--- a/source/Makefile
+++ b/source/Makefile
@@ -40,7 +40,6 @@
 # Defaults, can be overridden by make arguments or environment
 CXXFLAGS ?= -pipe -Wall -Wextra
 CFLAGS ?= -pipe -Wall -Wextra -O3
-CXXFLAGS_SIMD ?= -mavx2
 
 # Unconditionally set essential flags and optimization options
 CXXFLAGS_common := -std=c++11 -fopenmp $(COMPTIMEPLACE) $(CCFLAGS_common_add)
@@ -104,7 +103,7 @@
 
 opal/opal.o : opal/opal.cpp opal/opal.h
 	cd opal && \
-	$(CXX) -c -I./ -std=c++11 $(CPPFLAGS) $(CXXFLAGS) $(CXXFLAGSextra) $(CXXFLAGS_SIMD) opal.cpp
+	$(CXX) -c -I./ -std=c++11 $(CPPFLAGS) $(CXXFLAGS) $(CXXFLAGSextra) opal.cpp
 
 .PHONY: clean
 clean:
