new file mode 100755
--- /dev/null
+++ simd-dispatch.sh
@@ -0,0 +1,31 @@
+#!/bin/bash
+
+SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
+DIR=$SCRIPT_DIR
+BASE=${DIR}/$(basename "$0")
+CMDARGS="$@"
+
+# Normalize CPU flags to ensure consistent detection
+cpu_flags=$(grep -oP 'flags\s+:\s+\K.*' /proc/cpuinfo | sort | uniq | tr '\n' ' ')
+cpu_flags=${cpu_flags//_/}
+
+function test_and_run () {
+    if [[ $cpu_flags == *"$1"* ]] && [ -x "${BASE}-$1" ]; then
+        cmd="${BASE}-$1"
+        shift
+        # echo "${cmd}" "$@"
+        "${cmd}" "$@"
+        exit
+    fi
+}
+
+# Test for SIMD levels in order of preference
+for SIMD in avx2 avx sse4.1 sse4_1 ssse3 sse3; do
+    test_and_run ${SIMD} "$@"
+done
+
+# Fallback to plain option if no SIMD level matches
+if [ -x "$BASE-plain" ]; then
+    "$BASE-plain" "$@"
+else
+    echo "Error: No suitable STAR binary found for SIMD level." >&2
+    exit 1
+fi
+
