diff -ru a/source/Makefile b/source/Makefile
--- a/source/Makefile	2024-01-25 21:55:29.000000000 +0000
+++ b/source/Makefile	2025-04-26 21:41:45.041518209 +0100
@@ -134,7 +134,7 @@
 Depend.list: $(SOURCES) parametersDefault.xxd htslib
 	echo $(SOURCES)
 	'rm' -f ./Depend.list
-	$(CXX) $(CXXFLAGS_common) -MM $^ >> Depend.list
+	$(CXX) ${CXXFLAGS} $(CXXFLAGS_common) -MM $^ >> Depend.list
 include Depend.list
 endif
 endif
Only in b/source: Makefile~
