diff --git a/Makefile b/Makefile
index 7028a9a..730aa9a 100644
--- a/source/Makefile
+++ b/source/Makefile
@@ -42,6 +42,11 @@ GIT_BRANCH_COMMIT_DIFF := -D'GIT_BRANCH_COMMIT_DIFF="$(GIT_BRANCH_COMMIT_DIFF)"'
 # Defaults, can be overridden by make arguments or environment
 CXXFLAGS ?= -pipe -Wall -Wextra
 CFLAGS ?= -pipe -Wall -Wextra -O3
+ifeq ($(shell uname -m),x86_64)
+avxCXXFLAGS_SIMD ?=
+else
+avxCXXFLAGS_SIMD ?= -mavx2
+endif
 
 # Unconditionally set essential flags and optimization options
 CXXFLAGS_common := -std=c++11 -fopenmp $(COMPTIMEPLACE) $(CCFLAGS_common_add)
