diff --git a/Makefile b/Makefile2
index 9072034..08e928d 100644
--- a/Makefile
+++ b/Makefile2
@@ -1,14 +1,14 @@
-CC=gcc
-CFLAGS=-g -Wall -O2 -Wno-unused-function -Wno-format
-BINDIR=/usr/local/bin
+CC=$(CC)
+CFLAGS=-g -Wall -O3 -Wno-unused-function -Wno-format
+BINDIR=$(PREFIX)/bin
 
 all:seqtk
 
 seqtk:seqtk.c khash.h kseq.h
-		$(CC) $(CFLAGS) seqtk.c -o $@ -lz -lm
+		$(CC) $(CFLAGS) seqtk.c -o $@ $(LDFLAGS) -lz -lm
 
 install:all
-		install seqtk $(BINDIR)
+		install -v -m 0755 seqtk $(BINDIR)
 
 clean:
 		rm -fr gmon.out *.o ext/*.o a.out seqtk trimadap *~ *.a *.dSYM session*
