diff --git a/Makefile b/Makefile
index 12bd575..1e5309e 100644
--- ./Makefile
+++ ./Makefile
@@ -1,10 +1,10 @@
-CC=gcc
-CFLAGS=-g -Wall -O2 -Wno-unused-function
+CC ?= gcc
+CFLAGS := $(CFLAGS) -g -Wall -O2 -Wno-unused-function
 
 all:seqtk
 
 seqtk:seqtk.c khash.h kseq.h
-		$(CC) $(CFLAGS) seqtk.c -o $@ -lz -lm
+		$(CC) $(CFLAGS) $(LDFLAGS) seqtk.c -o $@ -lz -lm
 
 clean:
 		rm -fr gmon.out *.o ext/*.o a.out seqtk trimadap *~ *.a *.dSYM session*
