--- src/CMakeLists.txt	2018-01-30 09:34:20.302944803 +0100
+++ src/CMakeLists.txt	2018-01-30 09:41:41.688401196 +0100
@@ -3,15 +3,13 @@
 
 list(REMOVE_ITEM sources main.cpp)
 
-include_directories(../ext/htslib)
-
 add_library(kallisto_core ${sources} ${headers})
 target_include_directories(kallisto_core PUBLIC ${CMAKE_CURRENT_SOURCE_DIR})
 
 add_executable(kallisto main.cpp)
 
 find_package( Threads REQUIRED )
-target_link_libraries(kallisto kallisto_core pthread ${CMAKE_CURRENT_SOURCE_DIR}/../ext/htslib/libhts.a)
+target_link_libraries(kallisto kallisto_core pthread libhts)
 
 if(LINK MATCHES static)
     set(BUILD_SHARED_LIBS OFF)
@@ -56,4 +54,4 @@
 endif(LINK MATCHES static)
 
 
-install(TARGETS kallisto DESTINATION "${CMAKE_INSTALL_BINDIR}")
\ No newline at end of file
+install(TARGETS kallisto DESTINATION "${CMAKE_INSTALL_BINDIR}")
