--- CMakeLists.txt	2018-01-30 09:17:45.092860072 +0100
+++ CMakeLists.txt	2018-01-30 09:20:18.641290204 +0100
@@ -35,25 +35,9 @@
 ENDIF(LINK MATCHES static)
 
 
-include(ExternalProject)
-ExternalProject_Add(htslib
-    PREFIX ${PROJECT_SOURCE_DIR}/ext/htslib
-    SOURCE_DIR ${PROJECT_SOURCE_DIR}/ext/htslib
-    BUILD_IN_SOURCE 1
-    CONFIGURE_COMMAND ${PROJECT_SOURCE_DIR}/ext/htslib/configure 
-        --prefix=${PREFIX} --disable-bz2 --disable-lzma --disable-libcurl
-    BUILD_COMMAND make lib-static
-    INSTALL_COMMAND ""
-)
-
-include_directories(${htslib_PREFIX}/src/htslib)
-
-
-
 # add_compile_options(-Wdeprecated-register)
 
 add_subdirectory(src)
-include_directories(${EXT_PROJECTS_DIR})
 
 option(BUILD_TESTING "Build unit tests." OFF)
 include(CTest)
