diff --git a/Makefile b/Makefile
index f82a060..749cd4d 100644
--- a/Makefile
+++ b/Makefile
@@ -11,9 +11,9 @@ OBJ = $(MUSTANG)/obj
 BIN = $(MUSTANG)/bin
 
 #compiler options
-CPP = g++ 
-CPPFLAGS = -traditional -Wall -O3 
-LDFLAGS =
+CPP = $(CXX)
+CPPFLAGS = -Wall -O3 -I$(PREFIX)/include -Wno-deprecated-declarations -Wno-header-guard -Wno-array-parameter
+LDFLAGS = -L$(PREFIX)/lib
 .cpp.o:
 	$(CPP) $(CPPFLAGS) -c -o $@ $<
 
