diff --git a/utils/helper.cpp b/utils/helper.cpp
index c1474b6..4de32ff 100644
--- a/utils/helper.cpp
+++ b/utils/helper.cpp
@@ -1,6 +1,5 @@
 #include "helper.h"
 
-#include "byteswap.h"
 #include <sys/stat.h>
 #include <regex>
 
@@ -544,6 +543,18 @@ int PgHelpers::readsSufPreCmp(const char* suffixPart, const char* prefixRead) {
     return 0;
 }
 
+static __inline unsigned short
+bswap_16 (unsigned short __x)
+{
+    return (__x >> 8) | (__x << 8);
+}
+
+static __inline unsigned int
+bswap_32 (unsigned int __x)
+{
+    return (bswap_16 (__x & 0xffff) << 16) | (bswap_16 (__x >> 16));
+}
+
 int PgHelpers::strcmplcp(const char* lStrPtr, const char* rStrPtr, int length) {
     
     int i = 0;
