diff --git a/pymummer/nucmer.py b/pymummer/nucmer.py
index 5c7d0b9..88c31c6 100644
--- a/pymummer/nucmer.py
+++ b/pymummer/nucmer.py
@@ -136,7 +136,7 @@ def run(self):
         qry = os.path.abspath(self.qry)
         ref = os.path.abspath(self.ref)
         outfile = os.path.abspath(self.outfile)
-        tmpdir = tempfile.mkdtemp(prefix='tmp.run_nucmer.', dir=os.getcwd())
+        tmpdir = tempfile.mkdtemp(prefix='tmp.run_nucmer.')
         original_dir = os.getcwd()
         os.chdir(tmpdir)
         script = 'run_nucmer.sh'
