--- configure	2019-01-25 09:45:33.192414169 +0100
+++ configure.patched	2019-01-26 14:47:04.555112587 +0100
@@ -2881,10 +2881,10 @@
 cd ../;
 ## we add the '-w' flag to suppress all the warnings hdf5 prints
 ./configure --with-pic --enable-shared=no --enable-cxx \
-    --with-szlib \
+    --with-szlib --with-zlib=${PREFIX} --prefix=${PREFIX} \
     CXX="${CXX}" CXXFLAGS="${CXXFLAGS} ${CXXPICFLAGS} -w" \
     CC="${CC}" CFLAGS="${CFLAGS} ${CPICFLAGS} -w" \
-    LDFLAGS="-L`pwd -P`/szip/src/.libs" \
+    LDFLAGS="-L${PREFIX}/szip/src/.libs" \
     CPPFLAGS="-I`pwd -P`/szip/src/"
 $MAKE lib
 cd ../../
