--- configure	2018-12-20 17:09:38.001376321 +0100
+++ configure.patched	2018-12-20 17:25:42.834031679 +0100
@@ -2881,7 +2881,7 @@
 cd ../;
 ## we add the '-w' flag to suppress all the warnings hdf5 prints
 ./configure --with-pic --enable-shared=no --enable-cxx \
-    --with-szlib \
+    --with-szlib --with-zlib=${PREFIX} --prefix=${PREFIX} \
     CXX="${CXX}" CXXFLAGS="${CXXFLAGS} ${CXXPICFLAGS} -w" \
     CC="${CC}" CFLAGS="${CFLAGS} ${CPICFLAGS} -w" \
     LDFLAGS="-L`pwd -P`/szip/src/.libs" \
