diff --git a/config.sub b/config.sub
index 1d8e98b..384040c 100755
--- a/config.sub
+++ b/config.sub
@@ -450,6 +450,9 @@ case $basic_machine in
 	| ymp-* \
 	| z8k-* | z80-*)
 		;;
+	arm64-apple)
+		basic_machine=aarch64-apple
+		;;
 	# Recognize the basic CPU types without company name, with glob match.
 	xtensa*)
 		basic_machine=$basic_machine-unknown
diff --git a/configure.ac b/configure.ac 
index 20a3458..1912b7e 100644
--- a/configure.ac
+++ b/configure.ac 
@@ -6,7 +6,6 @@ AM_SILENT_RULES([yes])
 AC_CONFIG_SRCDIR([jellyfish])
 AC_CONFIG_HEADERS([config.h])
 LT_INIT
-AC_LIB_RPATH
 PKG_PROG_PKG_CONFIG

 # Change default compilation flags
