diff --git a/Makefile b/Makefile
index ad29043..4dff5ca 100644
--- a/Makefile
+++ b/Makefile
@@ -15,7 +15,7 @@ $(INSTALL_PREFIX)/bin/% : exec/%
 	install -C $< $(INSTALL_PREFIX)/bin
 
 exec/hetmers: src_ploidyplot/PloidyPlot.c src_ploidyplot/libfastk.c src_ploidyplot/libfastk.h src_ploidyplot/matrix.c src_ploidyplot/matrix.h
-	gcc $(CFLAGS) -o $@ src_ploidyplot/PloidyPlot.c src_ploidyplot/libfastk.c src_ploidyplot/matrix.c -lpthread -lm
+	$(CC) $(CFLAGS) -o $@ src_ploidyplot/PloidyPlot.c src_ploidyplot/libfastk.c src_ploidyplot/matrix.c -lpthread -lm
 
 
 .PHONY : clean
