diff --git a/Makefile b/Makefile
index bb4da2b..60d679e 100644
--- a/Makefile
+++ b/Makefile
@@ -11,16 +11,16 @@ HET_KMERS_INST = $(INSTALL_PREFIX)/bin/hetmers $(INSTALL_PREFIX)/bin/extract_kme
 default: exec/hetmers exec/extract_kmer_pairs
 
 .PHONY : install
-install : $(HET_KMERS_INST) 
+install : $(HET_KMERS_INST)
 
 $(INSTALL_PREFIX)/bin/% : exec/%
 	install -C $< $(INSTALL_PREFIX)/bin
 
 exec/hetmers: src/lib/PloidyPlot.c src/lib/libfastk.c src/lib/libfastk.h src/lib/matrix.c src/lib/matrix.h
-	gcc $(CFLAGS) -o $@ src/lib/PloidyPlot.c src/lib/libfastk.c src/lib/matrix.c -lpthread -lm
+	$(CC) $(CFLAGS) -o $@ src/lib/PloidyPlot.c src/lib/libfastk.c src/lib/matrix.c -pthread -lm
 
 exec/extract_kmer_pairs: src/lib/PloidyList.c src/lib/libfastk.c src/lib/libfastk.h src/lib/matrix.c src/lib/matrix.h
-	gcc $(CFLAGS) -o $@ src/lib/PloidyList.c src/lib/libfastk.c src/lib/matrix.c -lpthread -lm
+	$(CC) $(CFLAGS) -o $@ src/lib/PloidyList.c src/lib/libfastk.c src/lib/matrix.c -pthread -lm
 
 .PHONY : clean
 clean :
