--- Makefile
+++ Makefile
@@ -12,33 +12,33 @@
 
 gerpcol.o:	gerpcol.cc etree.o emodel.o MIter.o Mseq.o Seq.o
-	g++ $(COMPILE_OPTS) -c gerpcol.cc
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) $(LDFLAGS) $(COMPILE_OPTS) -c gerpcol.cc
 
 gerpcol:	gerpcol.o $(GERPCOL_LIBFILE)
-	g++ $(COMPILE_OPTS) -o gerpcol MIter.o Mseq.o Seq.o gerpcol.o etree.o emodel.o $(GERPCOL_LIBFILE)
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) $(LDFLAGS) $(COMPILE_OPTS) -o gerpcol MIter.o Mseq.o Seq.o gerpcol.o etree.o emodel.o $(GERPCOL_LIBFILE)
 
 gerpelem.o:	gerpelem.cc
-	g++ $(COMPILE_OPTS) -c gerpelem.cc
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) $(LDFLAGS) $(COMPILE_OPTS) -c gerpelem.cc
 
 gerpelem:	gerpelem.o
-	g++ $(COMPILE_OPTS) -o gerpelem gerpelem.o
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) $(LDFLAGS) $(COMPILE_OPTS) -o gerpelem gerpelem.o
 
 $(GERPCOL_LIBFILE):	$(OBJS)
-	ar -rc $(GERPCOL_LIBFILE) $(OBJS)
-	ranlib $(GERPCOL_LIBFILE)
+	$(AR) -rc $(GERPCOL_LIBFILE) $(OBJS)
+	$(RANLIB) $(GERPCOL_LIBFILE)
 
 Seq.o:	Seq.cc Seq.h Vec.h
-	g++ $(COMPILE_OPTS) -c Seq.cc
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) $(LDFLAGS) $(COMPILE_OPTS) -c Seq.cc
 
 Mseq.o:	Mseq.cc $(MSEQ_H)
-	g++ $(COMPILE_OPTS) -c Mseq.cc
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) $(LDFLAGS) $(COMPILE_OPTS) -c Mseq.cc
 
 MIter.o:	MIter.cc MIter.h Seq.h Vec.h
-	g++ $(COMPILE_OPTS) -c MIter.cc
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) $(LDFLAGS) $(COMPILE_OPTS) -c MIter.cc
 
 etree.o:	etree.cc etree.h MIter.o Mseq.o Seq.o
-	g++ $(COMPILE_OPTS) -c etree.cc
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) $(LDFLAGS) $(COMPILE_OPTS) -c etree.cc
 
 emodel.o:	emodel.cc emodel.h MIter.o Mseq.o Seq.o
-	g++ $(COMPILE_OPTS) -c emodel.cc
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) $(LDFLAGS) $(COMPILE_OPTS) -c emodel.cc
 
 clean:
