--- gerpcol.cc	2009-09-29 11:50:02.000000000 -0500
+++ gerpcol.cc.new	2019-01-29 00:14:34.255545765 -0600
@@ -26,6 +26,7 @@
 #include <map>
 #include <cmath>
 #include <sstream>
+#include <algorithm>
 
 #include "etree.h"
 #include "emodel.h"
--- gerpelem.cc	2009-09-29 12:05:28.000000000 -0500
+++ gerpelem.cc.new	2019-01-29 00:15:05.141542758 -0600
@@ -27,6 +27,7 @@
 #include <ctime>
 #include <algorithm>
 #include <cmath>
+#include <cstring>
 
 #include "Mseq.h"
 #include "MIter.h"
--- etree.cc	2009-09-29 11:50:02.000000000 -0500
+++ etree.cc.new	2019-01-29 00:54:16.298091489 -0600
@@ -25,6 +25,7 @@
 #include <iostream>
 #include <fstream>
 #include <string>
+#include <cstdlib>
 
 ETree::ETree(string filename) {
   ifstream ifs(filename.c_str());
