--- snpit/core.py.orig	2019-02-25 20:48:57.479475516 +0200
+++ snpit/core.py	2019-02-25 20:49:54.493443355 +0200
@@ -28,7 +28,7 @@
         self.threshold=threshold
 
         # construct the relative path in the package to the library file which contains a list of all the lineages and sub-lineages
-        resource_path = '/'.join(('..','lib', 'library.csv'))
+        resource_path = '/'.join(('..', '..', '..', '..', 'share', 'snpit', 'library.csv'))
 
         utf8_reader = codecs.getreader("utf-8")
 
@@ -51,7 +51,7 @@
             self.lineages[lineage_name]={'species':record['species'],'lineage':record['lineage'],'sublineage':record['sublineage']}
 
             # now we know the name construct the relative path to this lineage file
-            lineage_path='/'.join(('..','lib',lineage_name))
+            lineage_path='/'.join(('..', '..', '..', '..', 'share', 'snpit',lineage_name))
 
             # open a stream object to that file ready for reading
             lineage_file = pkg_resources.resource_stream("snpit", lineage_path)
@@ -176,7 +176,7 @@
         """
 
         # make the relative path to the H37Rv TB reference GenBank file
-        genbank_path = '/'.join(('..','lib', "H37Rv.gbk"))
+        genbank_path = '/'.join(('..', '..', '..', '..', 'share', 'snpit', "H37Rv.gbk"))
 
         # open a stream object ready for reading
         genbank_file = pkg_resources.resource_filename("snpit", genbank_path)
