diff --git a/AvxWindowFmIndex/CMakeLists.txt b/AvxWindowFmIndex/CMakeLists.txt
index 21bde02..95ee590 100644
--- a/AvxWindowFmIndex/CMakeLists.txt
+++ b/AvxWindowFmIndex/CMakeLists.txt
@@ -11,8 +11,8 @@ set(CMAKE_POSITION_INDEPENDENT_CODE ON)
 #add support for OpenMP
 find_package(OpenMP REQUIRED)
 if(OpenMP_FOUND)
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
-    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS}")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS} -I$(PREFIX)/include")
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS} -I$(PREFIX)/include")
 endif()
 
 set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY build)
@@ -238,4 +238,4 @@ add_custom_command(
 )
 
 # Add a dependency on build_submodule
-add_dependencies(awfmindex_static build_divsufsort)
\ No newline at end of file
+add_dependencies(awfmindex_static build_divsufsort)
diff --git a/AvxWindowFmIndex/lib/libdivsufsort/CMakeLists.txt b/AvxWindowFmIndex/lib/libdivsufsort/CMakeLists.txt
index 7859943..b87671b 100644
--- a/AvxWindowFmIndex/lib/libdivsufsort/CMakeLists.txt
+++ b/AvxWindowFmIndex/lib/libdivsufsort/CMakeLists.txt
@@ -17,10 +17,10 @@ set(CPACK_SOURCE_GENERATOR "TGZ;TBZ2;ZIP")
 include(ProjectCPack)

 ## Project options ##
-option(BUILD_SHARED_LIBS "Set to OFF to build static libraries" ON)
+option(BUILD_SHARED_LIBS "Set to OFF to build static libraries" OFF)
 option(BUILD_EXAMPLES "Build examples" ON)
-option(BUILD_DIVSUFSORT64 "Build libdivsufsort64" OFF)
-option(USE_OPENMP "Use OpenMP for parallelization" OFF)
+option(BUILD_DIVSUFSORT64 "Build libdivsufsort64" ON)
+option(USE_OPENMP "Use OpenMP for parallelization" ON)
 option(WITH_LFS "Enable Large File Support" ON)

 ## Installation directories ##
