From c47d59a3b8e0680e97cecbe0ad6aeb06ff7f3e73 Mon Sep 17 00:00:00 2001
From: Anders Goncalves da Silva <andersgs@gmail.com>
Date: Sat, 27 Jul 2019 18:10:03 +0000
Subject: [PATCH] Fix Makefile to work with conda

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 6490932..631db18 100644
--- a/Makefile
+++ b/Makefile
@@ -3,7 +3,7 @@ CC=			gcc
 CFLAGS=		-g -Wall -O2 -Wno-unused-function
 WRAP_MALLOC=-DUSE_MALLOC_WRAPPERS
 AR=			ar
-DFLAGS=		-DHAVE_PTHREAD $(WRAP_MALLOC)
+DFLAGS=		-DHAVE_PTHREAD $(WRAP_MALLOC) $(LDFLAGS)
 LOBJS=		utils.o kthread.o kstring.o ksw.o bwt.o bntseq.o bwa.o bwamem.o bwamem_pair.o malloc_wrap.o
 AOBJS=		QSufSort.o bwt_gen.o bwase.o bwaseqio.o bwtgap.o bwtaln.o bamlite.o \
 			is.o bwtindex.o bwape.o kopen.o pemerge.o \
-- 
2.14.2

