From 6c463143f714a99a2d26ef7e9f68669fb5fe2e2e Mon Sep 17 00:00:00 2001
From: Anders Goncalves da Silva <andersgs@gmail.com>
Date: Sun, 28 Jul 2019 22:43:34 -0700
Subject: [PATCH] Update Makefile

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 6490932..631db18 100644
--- a/Makefile
+++ b/Makefile
@@ -3,7 +3,7 @@ CC=			gcc
 CFLAGS=		-g -Wall -O2 -Wno-unused-function
 WRAP_MALLOC=-DUSE_MALLOC_WRAPPERS
 AR=			ar
-DFLAGS=		-DHAVE_PTHREAD $(WRAP_MALLOC)
+DFLAGS=		-DHAVE_PTHREAD $(WRAP_MALLOC) $(LDFLAGS)
 LOBJS=		utils.o kthread.o kstring.o ksw.o bwt.o bntseq.o bwa.o bwamem.o bwamem_pair.o malloc_wrap.o
 AOBJS=		QSufSort.o bwt_gen.o bwase.o bwaseqio.o bwtgap.o bwtaln.o bamlite.o \
 			is.o bwtindex.o bwape.o kopen.o pemerge.o \
-- 
2.20.1 (Apple Git-117)

