diff --git a/t/002_icpp.t b/t/002_icpp.t
index 2d9b0d0..f3a452d 100644
--- a/t/002_icpp.t
+++ b/t/002_icpp.t
@@ -16,7 +16,7 @@ my @DATA = (
     },
   ],
   [
-    { os => 'MSWin32', cc => 'gcc', config => {ccflags => '', ldflags => ''} },
+    { os => 'MSWin32', cc => $ENV{"CC"}, config => {ccflags => '', ldflags => ''} },
     {
       is_sunstudio => 0,
       is_msvc => undef, is_gcc => 1, is_clang => 0,
@@ -25,7 +25,7 @@ my @DATA = (
     },
   ],
   [
-    { os => 'MSWin32', cc => 'gcc', config => {ccflags => '', ldflags => 'static-libstdc++'} },
+    { os => 'MSWin32', cc => $ENV{"CC"}, config => {ccflags => '', ldflags => 'static-libstdc++'} },
     {
       is_sunstudio => 0,
       is_msvc => undef, is_gcc => 1, is_clang => 0,
@@ -34,7 +34,7 @@ my @DATA = (
     },
   ],
   [
-    { os => 'freebsd', cc => 'gcc', config => {ccflags => ''}, osvers => 9 },
+    { os => 'freebsd', cc => $ENV{"CC"}, config => {ccflags => ''}, osvers => 9 },
     {
       is_sunstudio => 0,
       is_msvc => undef, is_gcc => 1, is_clang => 0,
@@ -43,7 +43,7 @@ my @DATA = (
     },
   ],
   [
-    { os => 'freebsd', cc => 'gcc', config => {gccversion => 'Clang', ccflags => ''}, osvers => 10 },
+    { os => 'freebsd', cc => $ENV{"CC"}, config => {gccversion => 'Clang', ccflags => ''}, osvers => 10 },
     {
       is_sunstudio => undef,
       is_msvc => undef, is_gcc => undef, is_clang => 1,
@@ -52,7 +52,7 @@ my @DATA = (
     },
   ],
   [
-    { os => 'netbsd', cc => 'gcc', config => {ccflags => ''} },
+    { os => 'netbsd', cc => $ENV{"CC"}, config => {ccflags => ''} },
     {
       is_sunstudio => 0,
       is_msvc => undef, is_gcc => 1, is_clang => 0,
@@ -61,7 +61,7 @@ my @DATA = (
     },
   ],
   [
-    { os => 'linux', cc => 'clang', config => {gccversion => 'Clang', ccflags => ''} },
+    { os => 'linux', cc => $ENV{"CC"}, config => {gccversion => 'Clang', ccflags => ''} },
     {
       is_sunstudio => 0,
       is_msvc => undef, is_gcc => undef, is_clang => 1,
@@ -70,7 +70,7 @@ my @DATA = (
     },
   ],
   [
-    { os => 'linux', cc => 'gcc', config => {ccflags => ''} },
+    { os => 'linux', cc => $ENV{"CC"}, config => {ccflags => ''} },
     {
       is_sunstudio => 0,
       is_msvc => undef, is_gcc => 1, is_clang => 0,
