--- src/ReScoreDB.cpp
+++ src/ReScoreDB.cpp
@@ -379,7 +379,7 @@
 	ofstream new_res(new_res_file);
 	if (! new_res.is_open() || ! new_res.good())
 	{
-			cout << "Error: couldn't open new inspect results file for writing:" << new_res << endl;
+			cout << "Error: couldn't open new inspect results file for writing:" << new_res_file << endl;
 		exit(1);
 	}
 
@@ -583,7 +583,7 @@
 	ofstream new_res(new_res_file);
 	if (! new_res.is_open() || ! new_res.good())
 	{
-			cout << "Error: couldn't open new inspect results file for writing:" << new_res << endl;
+			cout << "Error: couldn't open new inspect results file for writing:" << new_res_file << endl;
 		exit(1);
 	}
 
