diff --git a/configure.sh b/configure.sh
index 95a0bf0..d7713b5 100755
--- a/configure.sh
+++ b/configure.sh
@@ -87,7 +87,7 @@ if [ ! -x $PYTHON ] ; then
   WITHOUT_ATAC="atac-driver/ seatac/"
 else
   echo "Python executable found in '$PYTHON'"
-  CFLAGS_PYTHON=`$PYTHON -c "from distutils import sysconfig; print sysconfig.get_python_inc()"`
+  CFLAGS_PYTHON=`$PYTHON -c "from distutils import sysconfig; print(sysconfig.get_python_inc())"`
   
   if [ -z "$CFLAGS_PYTHON" -o ! -d "$CFLAGS_PYTHON" ] ; then
     echo "WARNING:  Python development environment not found."
@@ -360,7 +360,7 @@ esac
 
 
 cat <<EOF >> Make.compilers
-PERL              := /usr/bin/perl
+PERL              := /usr/bin/env perl
 .EXE              := 
 .SO               := .so
 .A                := .a
