--- setup.py	2018-03-14 14:58:28.000000000 +0100
+++ setup_new.py	2018-08-03 17:31:43.925379385 +0200
@@ -57,18 +57,7 @@
         install_requires = [
             'cython',
             'numpy',
-            'scipy',
-            'matplotlib',
-            'pandas',
-            'docopt',
-            'tqdm',
-            'joblib',
-            'xlrd',
-            'openpyxl',
-            'graphviz',
-            'biopython',
-            'mygene',
-            'misc==0.2.8'
+            'scipy'
         ],
         include_package_data=True,
         test_suite='nose.collector',
