--- source/makefile
+++ source/makefile
@@ -73 +73 @@
-			$(CC) main.cpp -o SolexaQA++ $(LIBBOOST_FS) $(LIBBOOST_SYS) $(LIBBOOST_REGEX) $(LIBBOOST_IO) -I. -I/usr/local/include/boost -I/usr/local/include/ -L/usr/local/lib $(LDFLAGS) $(CXXFLAGS) $(LIBS)
+			$(CC) main.cpp -o SolexaQA++ $(LIBBOOST_FS) $(LIBBOOST_SYS) $(LIBBOOST_REGEX) $(LIBBOOST_IO) -I. -I$(BOOST_PATH)/include/boost -I$(BOOST_PATH)/include/ -L$(BOOST_PATH)/lib $(LDFLAGS) $(CXXFLAGS) $(LIBS)
