--- scripts/test_install	2016-07-09 13:44:07.643819412 +0400
+++ scripts/test_install	2016-07-09 13:41:18.239457732 +0400
@@ -1,6 +1,6 @@
 #!/usr/bin/env python
 from subprocess import call, Popen, PIPE
-import sys, os 
+import sys, os
 
 
 if sys.version_info[0] == 2:
@@ -40,11 +40,11 @@
     args = ['Rscript','--no-init-file','-e',rcmd]
     #output = check_output(args) #only in python2.7
     try:
-        output = Popen(args, stdout=PIPE).communicate()[0]
+        output = Popen(args, stdout=PIPE).communicate()[0].decode("utf-8")
         if output.split('\n')[-1].strip() != "42":
             raise "bad value"
-    except Exception,e: 
-            print("Exception: " + str(e) )
+    except Exception as e:
+            print("Exception: ")
             print(output)
             sys.exit("Simple test failed, please copy the whole output to kmergenie@cse.psu.edu")
     print("OK")
