diff --git a/meson.build b/meson.build
index 7391022..9393491 100644
--- a/meson.build
+++ b/meson.build
@@ -14,8 +14,8 @@ foreach ext, flag : extensions
     executable('PretextMap' + ext, 'PretextMap.cpp', dependencies : thread_dep, install : true, cpp_args : flags, link_with : 
         [   static_library('deflate' + ext, 'libdeflate/lib/deflate_compress.c', 'libdeflate/lib/deflate_decompress.c', 'libdeflate/lib/utils.c', c_args : flags, link_with : 
             [   static_library('deflate_x86' + ext, 'libdeflate/lib/x86/cpu_features.c', c_args : flags),
-                static_library('deflate_arm' + ext, 'libdeflate/lib/arm/cpu_features.c', c_args : flags)]),
-            static_library('mpc' + ext, 'mpc/mpc.c', c_args : flags)])
+                static_library('deflate_arm' + ext, 'libdeflate/lib/arm/cpu_features.c', c_args : flags)], include_directories : 'libdeflate/'),
+            static_library('mpc' + ext, 'mpc/mpc.c', c_args : flags)], include_directories : 'libdeflate/')
 endforeach
 
 e = executable('PretextMap', 'wrapper/SIMDExtensionWrapper.cpp', install : true, cpp_args : base_flags + ['-DProgramName=PretextMap'])
