From 644f0db61395be9283bc7dfe3b6a77a90a50c40b Mon Sep 17 00:00:00 2001
From: Matus Kosut <matus.kosut@ntnu.no>
Date: Sun, 1 Nov 2020 15:38:16 +0100
Subject: [PATCH 2/3] import shared Eigen3

---
 src/Regenie.hpp | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/Regenie.hpp b/src/Regenie.hpp
index a51db25..b6f5506 100644
--- a/src/Regenie.hpp
+++ b/src/Regenie.hpp
@@ -58,8 +58,8 @@
 #include <boost/math/distributions.hpp>
 
 #include "bgen_to_vcf.hpp"
-#include "eigen3.3/Dense"
-#include "eigen3.3/StdVector"
+#include "Eigen/Dense"
+#include "Eigen/StdVector"
 
 #ifdef __linux__
 #include <omp.h>
-- 
2.23.0

