From 28910770bb6b31df80f2c9eff25f517e6bc86432 Mon Sep 17 00:00:00 2001
From: Matus Kosut <matus.kosut@ntnu.no>
Date: Sun, 1 Nov 2020 18:11:11 +0100
Subject: [PATCH] fix fail exit on help

---
 src/Regenie.cpp | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/Regenie.cpp b/src/Regenie.cpp
index 759f95a..bf8f020 100755
--- a/src/Regenie.cpp
+++ b/src/Regenie.cpp
@@ -190,11 +190,11 @@ void read_params_and_check(int argc, char *argv[], struct param* params, struct
     if (vm.count("help")){
       print_header(std::cout);
       std::cout << AllOptions.help({"", "Main"}) << '\n' << webinfo << "\n\n";
-      exit(-1);
+      exit(0);
     } else if (vm.count("helpFull")) {
       print_header(std::cout);
       std::cout << AllOptions.help({"", "Main", "Additional"}) << '\n' << webinfo << "\n\n";
-      exit(-1);
+      exit(0);
     } 
     
     if (!vm.count("out")){
-- 
2.23.0

