--- post_isoseq_cluster/filter_by_count.py	2018-04-12 17:08:58.000000000 +0000
+++ post_isoseq_cluster/filter_by_count.py.new	2018-05-06 01:59:17.528883017 +0000
@@ -1,7 +1,3 @@
-__author__ = 'etseng@pacb.com'
-
-__version__ = '2.0'
-
 #!/usr/bin/env python
 
 from pbtools.pbtranscript.io import GFF
@@ -94,4 +90,4 @@
     parser.add_argument("--min_count", type=int, default=2, help="Minimum FL count (default: 2)")
 
     args = parser.parse_args()
-    filter_by_count(args.input_prefix, args.output_prefix, args.min_count)
\ No newline at end of file
+    filter_by_count(args.input_prefix, args.output_prefix, args.min_count)
--- targeted/count_hits_hg19.R	2018-04-12 17:08:58.000000000 +0000
+++ targeted/count_hits_hg19.R.new	2018-05-07 22:53:10.011665113 +0000
@@ -1,4 +1,10 @@
-x<-read.table('isoseq_flnc.fasta.hg19.sam.probe_hit.txt',sep='\t',header=T)
+#!/opt/anaconda1anaconda2anaconda3/bin/Rscript
+
+args = commandArgs(trailingOnly=TRUE)
+if (length(args)==0) {
+  stop("Input file must be specified.")
+}
+x<-read.table(args[1],sep='\t',header=T)
 print("# hit align hg19: ")
 print(length(unique(x$read_id)))
 
