--- post_isoseq_cluster/filter_by_count.py	2018-04-12 17:08:58.000000000 +0000
+++ post_isoseq_cluster/filter_by_count.py.new	2018-05-06 01:59:17.528883017 +0000
@@ -1,7 +1,3 @@
-__author__ = 'etseng@pacb.com'
-
-__version__ = '2.0'
-
 #!/usr/bin/env python
 
 from pbtools.pbtranscript.io import GFF
@@ -94,4 +90,4 @@
     parser.add_argument("--min_count", type=int, default=2, help="Minimum FL count (default: 2)")
 
     args = parser.parse_args()
-    filter_by_count(args.input_prefix, args.output_prefix, args.min_count)
\ No newline at end of file
+    filter_by_count(args.input_prefix, args.output_prefix, args.min_count)
