--- setup.py	2020-10-17 13:12:55.613817546 +0100
+++ setup.py	2020-10-17 13:13:09.385985958 +0100
@@ -46,7 +46,6 @@
                'cupcake/tofu/filter_by_count.py',
                'cupcake/tofu/filter_away_subset.py',
                'cupcake/tofu/fusion_finder.py',
-               'cupcake/tofu/fusion_collate_info.py',
                'cupcake/tofu/counting/chain_samples.py',
                'cupcake/tofu/counting/chain_fusion_samples.py',
                'cupcake/tofu/counting/summarize_sample_GFF_junctions.py',
