--- setup.py.bak	2017-02-12 23:39:14.000000000 +0100
+++ setup.py	2017-02-12 23:39:35.000000000 +0100
@@ -36,7 +36,7 @@
                         'pandas>=0.11.0',
                         'biopython>=1.62b',
                         'scikit-learn>=0.13.1',
-                        'nose==1.3.0',
+                        'nose',
                         'sphinx-rtd-theme>=0.1.6',
                         'Sphinx>=1.2.2'],
       entry_points="""
