diff --git i/concoct/transform.py w/concoct/transform.py
index 20e90d3..aff6e7f 100644
--- i/concoct/transform.py
+++ w/concoct/transform.py
@@ -2,6 +2,7 @@ from sklearn.decomposition import PCA
 import pandas as p
 
 def perform_pca(d, nc, seed):
+    d.columns = d.columns.astype(str)  # add this exact line
     pca_object = PCA(n_components=nc, random_state=seed).fit(d)
     return pca_object.transform(d), pca_object
 
