diff --git a/Cargo.toml b/Cargo.toml
index 007efdd..66cc96d 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -12,29 +12,32 @@ exclude = [ # Max upload is 10MB, as of writing test data was 15MB
   "tests/*",
 ]
 default-run = "coverm"
+edition = "2015"
 
 [dependencies]
-bio = "1.1.0"
-rust-htslib = { version="0.44.*", default-features = false }
+bio = "2.0.*"
+rust-htslib = { version="0.47.*", default-features = false } # On some systems, https://github.com/rust-bio/rust-htslib/issues/434 bites, so restrict to 0.46
 clap = { version="4.*", features = ["cargo"] } # cargo feature required for crate_version!
 clap_complete = "4.*"
 log = "0.4.*"
-env_logger = "0.10.*"
-nix = "0.26.*"
+env_logger = "0.11.*"
+nix = { version="0.29.*", features = ["fs"] }
 tempdir = "0.3.*"
 tempfile = "3.*"
 ansi_term = "0.12"
 lazy_static = "1.4.0"
 rand = "0.8.*"
 serde = "1.0"
-version-compare = "0.1.0"
-bird_tool_utils = "0.5.0"
-# bird_tool_utils = { git = "https://github.com/wwood/bird_tool_utils" }
+version-compare = "0.2.0"
+# bird_tool_utils = "0.5.0"
+bird_tool_utils = { git = "https://github.com/wwood/bird_tool_utils" }
+# bird_tool_utils = { path = "../bird_tool_utils" }
 galah = "0.4.0"
 # galah = { git = "https://github.com/wwood/galah" }
 bird_tool_utils-man = "0.4.0"
 roff = "0.2.*"
 needletail = "0.5.*"
+csv = "1.*"
 
 [dev-dependencies]
 assert_cli = "0.6.*"
