--- configure.py    2016-08-04 20:35:56.000000000 +0300
+++ configure.py        2016-08-04 20:40:38.000000000 +0300
@@ -3,7 +3,7 @@

 - Create defines.mk
 """
-import commands
+import subprocess
 import contextlib
 import optparse
 import os
@@ -23,7 +23,7 @@

 def shell(cmd):
     log('`%s`'%cmd)
-    status, output = commands.getstatusoutput(cmd)
+    status, output = subprocess.getstatusoutput(cmd)
     if status:
         raise Exception('%d <- %r' %(status, cmd))
     log(output)

--- libcpp/configure.py     2016-08-04 20:46:18.000000000 +0300
+++ libcpp/configure.py 2016-08-04 20:42:45.000000000 +0300
@@ -9,7 +9,7 @@

 This is not used by './unittest/'.
 """
-import commands
+import subprocess
 import contextlib
 import os
 import sys
@@ -22,7 +22,7 @@

 def shell(cmd):
     log(cmd)
-    status, output = commands.getstatusoutput(cmd)
+    status, output = subprocess.getstatusoutput(cmd)
     if status:
         raise Exception('%d <- %r' %(status, cmd))
     return output
