--- erds1.1/em.pl        2018-05-07 15:21:03.587125000 +0200
+++ erds1.1/em.pl    2018-05-07 15:21:11.855736000 +0200
@@ -164,7 +164,7 @@
        my $total_line=$_[1];
        open(W,">$weight_file") or die "Can't open $weight_file";
        for(my $k=0; $k<=$total_line; $k++){
-               printf W ("1\t1\n",1);
+               printf W ("1\t1\n");
        }
        close(W) or die "Can't close $weight_file";
 }

