--- erds1.1/erds_pipeline.pl	2018-05-07 12:52:32.978255013 +0200
+++ erds1.1/erds_pipeline.pl	2018-05-07 12:52:43.121833391 +0200
@@ -88,7 +88,7 @@
 $sd_dir="$script_dir/database/SD_list_"."$sd";
 
 if(!defined($samtools)){
-	$samtools="$script_dir/software/samtools";
+	$samtools="samtools";
 }
 if(!defined($samplename)){
 	$samplename=get_samplename();
@@ -467,4 +467,4 @@
 	
 	_EOT_
 	exit(1);
-}
\ No newline at end of file
+}
