diff --git a/cmake/BuildBoost.cmake b/cmake/BuildBoost.cmake
index d796ad5..7522c29 100644
--- a/cmake/BuildBoost.cmake
+++ b/cmake/BuildBoost.cmake
@@ -29,11 +29,11 @@ ExternalDependency_Add(
         URL ${BOOST_URL}
         SOURCE_DIR ${BOOST_SRC}
         BINARY_DIR ${BOOST_SRC}
-        CONFIGURE_COMMAND "./bootstrap.sh"
+        CONFIGURE_COMMAND ./bootstrap.sh --with-toolset=gcc
         BUILD_COMMAND
             echo "Building boost, build log is ${BOOST_LOG}" &&
             ./b2 --prefix=${BOOST_ROOT} --layout=system link=static
-                    threading=multi install ${BOOST_BUILD_LIBS} > ${BOOST_LOG} 2>&1
+                    threading=multi strip=on variant=release cflags="-std=c++14" install toolset=gcc ${BOOST_BUILD_LIBS} > ${BOOST_LOG} 2>&1
         INSTALL_COMMAND "true"
 )
 
