diff --git a/cmake/BuildSamtools.cmake b/cmake/BuildSamtools.cmake
index 7ba6103..35a2aa6 100644
--- a/cmake/BuildSamtools.cmake
+++ b/cmake/BuildSamtools.cmake
@@ -42,7 +42,7 @@ ExternalProject_Add(
     URL ${CMAKE_SOURCE_DIR}/vendor/xz-5.2.4.tar.gz
     SOURCE_DIR ${XZ_SRC}
     BINARY_DIR ${XZ_SRC}
-    CONFIGURE_COMMAND ./configure --prefix=${XZ_ROOT}
+    CONFIGURE_COMMAND autoreconf -if && ./configure --prefix=${XZ_ROOT} CC=clang
     BUILD_COMMAND make
     INSTALL_COMMAND make install
 )
@@ -94,7 +94,7 @@ ExternalProject_Add(
     # which were causing problems on my OS X machine
     # RTSP remains enabled in the minimal build container,
     # so we leave out --disable-rtsp 
-    CONFIGURE_COMMAND ./configure --prefix=${CURL_ROOT} --with-mbedtls=${MBEDTLS_ROOT} --without-zlib --without-brotli --without-winssl --without-schannel --without-darwinssl --without-secure-transport --without-amissl --without-ssl --without-gnutls --without-wolfssl --without-mesalink --without-nss --without-libpsl --without-libmetalink --without-librtmp --without-winidn --without-libidn2 --without-nghttp2 --without-ngtcp2 --without-nghttp3 --without-quiche --without-zsh-functions-dir --without-fish-functions-dir --disable-ldap --disable-ldaps
+    CONFIGURE_COMMAND autoreconf -if && sed -i.bak -e "s|10.8|10.13|" configure && ./configure --prefix=${CURL_ROOT} --with-mbedtls=${MBEDTLS_ROOT} --without-zlib --without-brotli --without-winssl --without-schannel --without-darwinssl --without-secure-transport --without-amissl --without-ssl --without-gnutls --without-wolfssl --without-mesalink --without-nss --without-libpsl --without-libmetalink --without-librtmp --without-winidn --without-libidn2 --without-nghttp2 --without-ngtcp2 --without-nghttp3 --without-quiche --without-zsh-functions-dir --without-fish-functions-dir --disable-ldap --disable-ldaps CC=clang
     BUILD_COMMAND make
     INSTALL_COMMAND make install
     DEPENDS mbedtls
