--- boost-1.55-bamrc/bootstrap.sh	2024-06-06 08:22:28.944000000 +0000
+++ boost-1.55-bamrc/bootstrap.sh	2024-06-06 08:21:55.140000000 +0000
@@ -277,7 +277,7 @@
 
     if test "x$PYTHON_ROOT" = x; then
         echo -n "Detecting Python root... "
-        PYTHON_ROOT=`$PYTHON -c "import sys; print sys.prefix"`
+        PYTHON_ROOT=`$PYTHON -c "import sys; print(sys.prefix)"`
         echo $PYTHON_ROOT
     fi    
 fi
