diff --git a/configure b/configure
index 9fa777a..6a4c04e 100755
--- a/configure
+++ b/configure
@@ -20779,7 +20779,7 @@ printf "%s\n" "$tst_cflags" >&6; };
 printf %s "checking for *version-min in CFLAGS... " >&6; }
     min=""
     if test -z "$(echo $CFLAGS | grep m.*os.*-version-min)"; then
-      min="-mmacosx-version-min=10.8"
+      min="-mmacosx-version-min=10.13"
       CFLAGS="$CFLAGS $min"
     fi
     if test -z "$min"; then
