--- a/get_pvalue.R.orig	2018-09-03 08:14:34.850659791 -0400
+++ b/get_pvalue.R	2018-09-03 09:54:11.911808821 -0400
@@ -1,4 +1,4 @@
-#!/usr/bin/Rscript
+#!/opt/anaconda1/anaconda2anaconda3/bin/Rscript
 # vim: tabstop=8 expandtab shiftwidth=2 softtabstop=2
 # integrate prior info of bkg error distribution (based on duplex data) and data specific error distribution
 # modified the prior distribution using more stringint consensus criteria. 
@@ -8,7 +8,6 @@
 
 rm(list=ls())
 library(plyr)
-bkgErrorDistSimulation <- '/srv/qgen/data/annotation/bkg.error.v2.RData'
 
 ##############################
 ##       Parameters         ##
@@ -25,6 +24,7 @@
 rpb <- as.numeric(args[9])
 minAltUMI <- as.numeric(args[10])
 inputVCF <- args[11]
+bkgErrorDistSimulation <- args[12]
 min.mtDepth <- 1000
 
 # set working directory
