diff -rupN idba-1.1.3/src/basic/kmer.h idba-1.1.3-new/src/basic/kmer.h
--- idba-1.1.3/src/basic/kmer.h	2016-04-01 07:52:56.000000000 +0200
+++ idba-1.1.3-new/src/basic/kmer.h	2018-05-02 14:17:59.607555356 +0200
@@ -197,7 +197,7 @@ public:
     { return kMaxSize; }
 
 
-    static const uint32_t kNumUint64 = 4;
+    static const uint32_t kNumUint64 = 10;
     static const uint32_t kBitsForSize = ((kNumUint64 <= 2) ? 6 : ((kNumUint64 <= 8) ? 8 : 16));
     static const uint32_t kBitsForKmer = (kNumUint64 * 64 - kBitsForSize);
     static const uint32_t kMaxSize = kBitsForKmer / 2;
diff -rupN idba-1.1.3/src/sequence/short_sequence.h idba-1.1.3-new/src/sequence/short_sequence.h
--- idba-1.1.3/src/sequence/short_sequence.h	2016-04-01 07:52:56.000000000 +0200
+++ idba-1.1.3-new/src/sequence/short_sequence.h	2018-05-02 14:18:01.671539022 +0200
@@ -99,7 +99,7 @@ public:
         return size() < seq.size();
     }
 
-    static const uint32_t kMaxShortSequence = 128;
+    static const uint32_t kMaxShortSequence = 600;
     static const uint32_t kNumBytes = (kMaxShortSequence + 3) / 4 + 2;
 
 private:
