diff --git a/src/assembly/assembly_utility.h b/src/assembly/assembly_utility.h
index 02d4128..c4daae9 100644
--- a/src/assembly/assembly_utility.h
+++ b/src/assembly/assembly_utility.h
@@ -24,7 +24,7 @@
 class ContigGraph;
 class HashGraph;
 class Sequence;
-class ShortSequence;
+#include "sequence/short_sequence.h"
 class Sequence;
 class HashAlignerRecord;
 class HashAligner;
diff --git a/src/basic/kmer.h b/src/basic/kmer.h
index 70146a1..37247c9 100644
--- a/src/basic/kmer.h
+++ b/src/basic/kmer.h
@@ -197,7 +197,7 @@ public:
     { return kMaxSize; }
 
 
-    static const uint32_t kNumUint64 = 4;
+    static const uint32_t kNumUint64 = 10;
     static const uint32_t kBitsForSize = ((kNumUint64 <= 2) ? 6 : ((kNumUint64 <= 8) ? 8 : 16));
     static const uint32_t kBitsForKmer = (kNumUint64 * 64 - kBitsForSize);
     static const uint32_t kMaxSize = kBitsForKmer / 2;
diff --git a/src/sequence/short_sequence.h b/src/sequence/short_sequence.h
index 53c1fe9..f4f1c02 100644
--- a/src/sequence/short_sequence.h
+++ b/src/sequence/short_sequence.h
@@ -99,7 +99,7 @@ public:
         return size() < seq.size();
     }
 
-    static const uint32_t kMaxShortSequence = 512;
+    static const uint32_t kMaxShortSequence = 600;
     static const uint32_t kNumBytes = (kMaxShortSequence + 3) / 4 + 2;
 
 private:
