--- flu-amd/LABEL	2021-03-11 13:45:48.000000000 -0600
+++ LABEL	2021-06-04 16:40:06.961358066 -0500
@@ -253,11 +253,11 @@
 bin="$bpath"/$resources/scripts
 hmmscore="$bin"/hmmscore_$OS
 align2model="$bin"/align2model_$OS
-muscle="$bin"/muscle_$OS
-FastTreeMP="$bin"/FastTreeMP_$OS
+muscle="muscle"
+FastTreeMP="FastTreeMP"
 modelfromalign="$bin"/modelfromalign_$OS
 shogun="$bin"/shogun_$OS
-PARALLEL="$bin/parallel"
+PARALLEL="parallel"
 
 
 if [ "$DO_WEB" -eq "0" ];then
