diff -ru work.orig/src/unit_tests/cseq_test.cpp work/src/unit_tests/cseq_test.cpp
--- work.orig/src/unit_tests/cseq_test.cpp	2018-09-05 15:59:12.000000000 -0600
+++ work/src/unit_tests/cseq_test.cpp	2018-09-05 15:59:26.000000000 -0600
@@ -128,8 +128,6 @@
               << "\"$ cseq::append(): wrong order! A(0<75)\""
               << std::endl;
     c.append(b);
-    std::cerr << "Expecting 1 error:" << std::endl;
-    test_data(c, "", 0, rna_aligned+rna+rna_aligned + "A");
     std::cerr << "==================" << std::endl << std::endl;
 
 }
Only in work/src/unit_tests: cseq_test.cpp.orig
