--- core/Makefile	2019-12-08 23:32:07.647136701 +0900
+++ core/linux-makefile	2019-12-09 09:24:47.196427934 +0900
@@ -47,7 +47,7 @@
 STDF = -std=c99
 endif
 
-MYCFLAGS = $(MNO_CYGWIN) $(ENABLE_MULTITHREAD) $(ENABLE_ATOMIC) $(STDF) $(CFLAGS)
+MYCFLAGS = $(MNO_CYGWIN) $(ENABLE_MULTITHREAD) $(ENABLE_ATOMIC) $(STDF) $(CFLAGS) -static
 
 INSTALL = install
 
@@ -161,7 +161,7 @@
 dlls : $(DLLS)
 
 $(DASH_CLIENT): dash_client.go
-	go build dash_client.go
+	go build --ldflags '-extldflags "-static"' dash_client.go
 
 univscript: univscript.tmpl Makefile
 	sed "s:_PROGS:$(PROGS):" univscript.tmpl  > univscript
