diff --git a/build.rs b/build.rs
index 9a5b4c1..7aa001f 100644
--- a/build.rs
+++ b/build.rs
@@ -67,6 +67,16 @@ fn main() {
         dst_piscem_cpp.join("lib").display()
     );

+    // Arghhh conda!
+    println!(
+        "cargo:rustc-link-search=native={}",
+        dst_cf.join("lib64").display()
+    );
+    println!(
+        "cargo:rustc-link-search=native={}",
+        dst_piscem_cpp.join("lib64").display()
+    );
+
     println!("cargo:rustc-link-lib=static=kmc_core");
     //println!("cargo:rustc-link-lib=static=pesc_static");
     //println!("cargo:rustc-link-lib=static=build_static");
