From 9852a11e4c38f92725f72eefc041cfe5fadc2ea6 Mon Sep 17 00:00:00 2001
From: Travis Wrightsman <tw493@cornell.edu>
Date: Tue, 23 Jan 2024 22:18:44 -0500
Subject: [PATCH] Remove need for CuTest during build

---
 C/impl/sonLibGlobalsInternal.h | 1 -
 1 file changed, 1 deletion(-)

diff --git a/C/impl/sonLibGlobalsInternal.h b/C/impl/sonLibGlobalsInternal.h
index f8542a0..9a2e51d 100644
--- a/C/impl/sonLibGlobalsInternal.h
+++ b/C/impl/sonLibGlobalsInternal.h
@@ -29,7 +29,6 @@
 #include <math.h>
 #include <inttypes.h>
 
-#include "CuTest.h"
 #include "sonLib.h"
 #include "sonLibListPrivate.h"
 
-- 
2.39.2

