diff --git a/src/densities.cpp b/src/densities.cpp
index ec5f347..039e063 100644
--- a/src/densities.cpp
+++ b/src/densities.cpp
@@ -505,7 +505,7 @@ void matrixMultiply(int transposeA, int transposeB,
   if (k != nrowB) {
     cout<<"dimension do not match "<<k<<" != "<<nrowB<<"\n";
   }
-  F77_CALL(dgemm)(&transa, &transb, &m, &n, &k, &one, a, &nrowA, b, &nrowB, &zero, c, &m);
+  F77_CALL(dgemm)(&transa, &transb, &m, &n, &k, &one, a, &nrowA, b, &nrowB, &zero, c, &m, 1, 1);
 }
   
 
