--- ssake/SSAKE	2018-01-21 04:39:05.000000000 +0000
+++ ssake/SSAKE.new	2018-07-10 06:42:03.002749762 +0000
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 
 #AUTHOR
--- ssake/tools/analyzePositionSSAKE.pl	2018-01-21 04:39:05.000000000 +0000
+++ ssake/tools/analyzePositionSSAKE.pl.new	2018-07-10 06:42:47.153809903 +0000
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # Rene Warren, October 2010
 
 use strict;
--- ssake/tools/getStats.pl	2018-01-21 04:39:05.000000000 +0000
+++ ssake/tools/getStats.pl.new	2018-07-10 06:43:03.580832276 +0000
@@ -1,4 +1,4 @@
-#!/home/martink/bin/perl
+#!/usr/bin/env perl
 ##Rene Warren 2002-2014
 
 use Statistics::Descriptive;
--- ssake/tools/makeFastaFileFromScaffolds.pl	2018-01-21 04:39:05.000000000 +0000
+++ ssake/tools/makeFastaFileFromScaffolds.pl.new	2018-07-10 06:43:13.359845599 +0000
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # Rene Warren, November 2007
 # Uses SSAKE Scaffold instructions (.scaffolds) to organize contigs (fasta sequence) within a scaffold.
 
--- ssake/tools/makePairedOutput2EQUALfiles.pl	2018-01-21 04:39:05.000000000 +0000
+++ ssake/tools/makePairedOutput2EQUALfiles.pl.new	2018-07-10 06:43:26.264863178 +0000
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 #RLW 2010
 
 use strict;
--- ssake/tools/makePairedOutput2UNEQUALfiles.pl	2018-01-21 04:39:05.000000000 +0000
+++ ssake/tools/makePairedOutput2UNEQUALfiles.pl.new	2018-07-10 06:43:36.058876517 +0000
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 #RLW 2010
 
 use strict;
--- ssake/tools/makePairedOutput.pl	2018-01-21 04:39:05.000000000 +0000
+++ ssake/tools/makePairedOutput.pl.new	2018-07-10 06:43:48.128892961 +0000
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 #Rene Warren -April 2008
 #rwarren at gmail dot com
 
--- ssake/tools/nLength.pl	2018-01-21 04:39:05.000000000 +0000
+++ ssake/tools/nLength.pl.new	2018-07-10 06:43:56.560904445 +0000
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 #Rene Warren 2002-2008
 
 my $min = 200;
--- ssake/tools/qseq2fasta.pl	2018-01-21 04:39:05.000000000 +0000
+++ ssake/tools/qseq2fasta.pl.new	2018-07-10 06:44:08.770921077 +0000
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use strict;
 
--- ssake/tools/qseq2fastq.pl	2018-01-21 04:39:05.000000000 +0000
+++ ssake/tools/qseq2fastq.pl.new	2018-07-10 06:44:19.470935652 +0000
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
--- ssake/tools/splitInput.pl	2018-01-21 04:39:05.000000000 +0000
+++ ssake/tools/splitInput.pl.new	2018-07-10 06:44:33.339954543 +0000
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 #Rene Warren 2008
 #rwarren at bcgsc dot ca
 
--- ssake/tools/TQSfastq.pl	2018-01-21 04:39:05.000000000 +0000
+++ ssake/tools/TQSfastq.pl.new	2018-07-10 06:44:47.158973367 +0000
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 
 #AUTHOR
--- ssake/tools/TQSfastq.py	2018-01-21 04:39:05.000000000 +0000
+++ ssake/tools/TQSfastq.py.new	2018-07-10 06:45:28.611029831 +0000
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/usr/bin/env python
 
 __doc__ = """
 TQS
