diff --git a/add.c b/add.c
index 752dd5d..7f75e47 100644
--- a/add.c
+++ b/add.c
@@ -240,7 +240,10 @@ void op_add_apply
 			v = NULL;
 			chromSpec = find_chromosome_spec (chrom);
 			if (chromSpec != NULL) v = chromSpec->valVector;
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
 			strncpy (prevChrom, chrom, sizeof(prevChrom));
+#pragma GCC diagnostic pop
 			}
 
 		if (chromSpec == NULL) continue;
@@ -533,7 +536,10 @@ void op_subtract_apply
 			v = NULL;
 			chromSpec = find_chromosome_spec (chrom);
 			if (chromSpec != NULL) v = chromSpec->valVector;
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
 			strncpy (prevChrom, chrom, sizeof(prevChrom));
+#pragma GCC diagnostic pop
 			}
 
 		if (chromSpec == NULL) continue;
diff --git a/genodsp.c b/genodsp.c
index ac1704e..00377b1 100644
--- a/genodsp.c
+++ b/genodsp.c
@@ -1229,7 +1229,10 @@ void read_intervals
 			v = NULL;
 			chromSpec = find_chromosome_spec (chrom);
 			if (chromSpec != NULL) v = chromSpec->valVector;
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
 			strncpy (prevChrom, chrom, sizeof(prevChrom));
+#pragma GCC diagnostic pop
 			}
 
 		if (chromSpec == NULL) continue;
diff --git a/logical.c b/logical.c
index fb295e3..6689e51 100644
--- a/logical.c
+++ b/logical.c
@@ -497,7 +497,10 @@ void op_or_apply
 			v = NULL;
 			chromSpec = find_chromosome_spec (chrom);
 			if (chromSpec != NULL) v = chromSpec->valVector;
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
 			strncpy (prevChrom, chrom, sizeof(prevChrom));
+#pragma GCC diagnostic pop
 			}
 
 		if (chromSpec == NULL) continue;
@@ -810,7 +813,10 @@ void op_and_apply
 				if (chromSpec->flag) goto chrom_not_together;
 				}
 
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
 			strncpy (prevChrom, chrom, sizeof(prevChrom));
+#pragma GCC diagnostic pop
 			}
 
 		// if the current chromosome is not of any interest, ignore this
diff --git a/mask.c b/mask.c
index 4fd5ebd..283a5eb 100644
--- a/mask.c
+++ b/mask.c
@@ -253,7 +253,10 @@ void op_mask_apply
 			v = NULL;
 			chromSpec = find_chromosome_spec (chrom);
 			if (chromSpec != NULL) v = chromSpec->valVector;
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
 			strncpy (prevChrom, chrom, sizeof(prevChrom));
+#pragma GCC diagnostic pop
 			}
 
 		if (chromSpec == NULL) continue;
@@ -548,7 +551,10 @@ void op_mask_not_apply
 				if (chromSpec->flag) goto chrom_not_together;
 				}
 
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
 			strncpy (prevChrom, chrom, sizeof(prevChrom));
+#pragma GCC diagnostic pop
 			}
 
 		// if the current chromosome is not of any interest, ignore this
diff --git a/minmax.c b/minmax.c
index c222bd0..4c9684e 100644
--- a/minmax.c
+++ b/minmax.c
@@ -261,7 +261,10 @@ void op_min_in_interval_apply
 				if (chromSpec->flag) goto chrom_not_together;
 				}
 
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
 			strncpy (prevChrom, chrom, sizeof(prevChrom));
+#pragma GCC diagnostic pop
 			}
 
 		// if the current chromosome is not of any interest, ignore this
@@ -664,7 +667,10 @@ void op_max_in_interval_apply
 				if (chromSpec->flag) goto chrom_not_together;
 				}
 
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
 			strncpy (prevChrom, chrom, sizeof(prevChrom));
+#pragma GCC diagnostic pop
 			}
 
 		// if the current chromosome is not of any interest, ignore this
@@ -1941,7 +1947,10 @@ void op_min_with_apply
 			v = NULL;
 			chromSpec = find_chromosome_spec (chrom);
 			if (chromSpec != NULL) v = chromSpec->valVector;
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
 			strncpy (prevChrom, chrom, sizeof(prevChrom));
+#pragma GCC diagnostic pop
 			}
 
 		if (chromSpec == NULL) continue;
@@ -2227,7 +2236,10 @@ void op_max_with_apply
 			v = NULL;
 			chromSpec = find_chromosome_spec (chrom);
 			if (chromSpec != NULL) v = chromSpec->valVector;
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
 			strncpy (prevChrom, chrom, sizeof(prevChrom));
+#pragma GCC diagnostic pop
 			}
 
 		if (chromSpec == NULL) continue;
diff --git a/multiply.c b/multiply.c
index 4c1b550..8e2fa56 100644
--- a/multiply.c
+++ b/multiply.c
@@ -260,7 +260,10 @@ void op_multiply_apply
 				if (chromSpec->flag) goto chrom_not_together;
 				}
 
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
 			strncpy (prevChrom, chrom, sizeof(prevChrom));
+#pragma GCC diagnostic pop
 			}
 
 		// if the current chromosome is not of any interest, ignore this
@@ -654,7 +657,10 @@ void op_divide_apply
 				if (chromSpec->flag) goto chrom_not_together;
 				}
 
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
 			strncpy (prevChrom, chrom, sizeof(prevChrom));
+#pragma GCC diagnostic pop
 			}
 
 		// if the current chromosome is not of any interest, ignore this
