--- a/EukHighConfidenceFilter.in
+++ b/EukHighConfidenceFilter.in
@@ -13,7 +13,7 @@
 # --------------------------------------------------------------------
  
 use strict;
-use lib "@libdir@/tRNAscan-SE";
+use lib '@libdir@/tRNAscan-SE';
 use Getopt::Long;
 
 use tRNAscanSE::tRNA;
diff --git a/tRNAscan-SE.src b/tRNAscan-SE.src
index 771cc2d..796b952 100755
--- a/tRNAscan-SE.src
+++ b/tRNAscan-SE.src
@@ -18,7 +18,7 @@
 #                           
 
 use strict;
-use lib "@libdir@/tRNAscan-SE";
+use lib '@libdir@/tRNAscan-SE';
 use Getopt::Long;
 use tRNAscanSE::Configuration;
 use tRNAscanSE::tRNA;
@@ -44,7 +44,7 @@ our $release_date = "July 2021";
 our $program_id = "tRNAscan-SE-".$version;
 
 # modified by 'make'
-our $default_conf = "@bindir@/tRNAscan-SE.conf";
+our $default_conf = '@bindir@/tRNAscan-SE.conf';
 
 # Signal handling
 $SIG{'TERM'} = 'error_handler';
