--- src/genome_sites.c	2024-11-20 21:03:59.000000000 +0000
+++ src/genome_sites_modified.c	2025-03-08 19:30:57.429345500 +0000
@@ -413,7 +413,7 @@
 #else      
       debug2(printf("found is %08X, -found & found is %08X\n",found,-found & found));
       pos = offset + mod_37_bit_position[(lowbit = -found & found) % 37];
-      sites = Intlistpool_push(sites,pos
+      sites = Intlistpool_push(sites,intlistpool,pos
 			       intlistpool_trace(__FILE__,__LINE__));
       found -= lowbit;
 #endif
