--- source/Makefile
+++ source/Makefile
@@ -1,31 +1,31 @@
 # BayeScan makefile
 
 bayescan2: start.o beta.o dirichlet.o RJupdates.o MHupdates.o likelihood.o read_write.o anyoption.o 
-	g++ -o bayescan2 -static start.o beta.o dirichlet.o RJupdates.o MHupdates.o likelihood.o read_write.o anyoption.o 
+	$(CXX) $(LDFLAGS) -o bayescan2 start.o beta.o dirichlet.o RJupdates.o MHupdates.o likelihood.o read_write.o anyoption.o 
 
 start.o: start.cpp errors.cpp anyoption.h global_defs.h
-	g++ -c start.cpp errors.cpp 
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) -c start.cpp errors.cpp 
 
 beta.o: beta.cpp global_defs.h
-	g++ -c beta.cpp 
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) -c beta.cpp 
       
 dirichlet.o: dirichlet.cpp global_defs.h
-	g++ -c dirichlet.cpp 
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) -c dirichlet.cpp 
 
 RJupdates.o: RJupdates.cpp global_defs.h
-	g++ -c RJupdates.cpp 
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) -c RJupdates.cpp 
 
 MHupdates.o: MHupdates.cpp global_defs.h
-	g++ -c MHupdates.cpp 
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) -c MHupdates.cpp 
 
 likelihood.o: likelihood.cpp global_defs.h
-	g++ -c likelihood.cpp 
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) -c likelihood.cpp 
 
 read_write.o: read_write.cpp errors.cpp global_defs.h
-	g++ -c read_write.cpp errors.cpp 
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) -c read_write.cpp errors.cpp 
 
 anyoption.o: anyoption.cpp anyoption.h 
-	g++ -c anyoption.cpp 
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) -c anyoption.cpp 
 
 clean: 
 	rm *.o bayescan2
