diff --git a/source/Makefile b/source/Makefile
index fde05fc..c5b38c9 100644
--- a/source/Makefile
+++ b/source/Makefile
@@ -1,31 +1,31 @@
 # BayeScan makefile
 
-bayescan_2.1: start.o beta.o dirichlet.o RJupdates.o MHupdates.o likelihood.o read_write.o anyoption.o 
-	g++ -fopenmp -static -lpthread -o bayescan_2.1 start.o beta.o dirichlet.o RJupdates.o MHupdates.o likelihood.o read_write.o anyoption.o 
+bayescan_2.1: start.o beta.o dirichlet.o RJupdates.o MHupdates.o likelihood.o read_write.o anyoption.o
+	$(CXX) $(CXXFLAGS) -std=c++14 -fopenmp -static -pthread -o bayescan_2.1 start.o beta.o dirichlet.o RJupdates.o MHupdates.o likelihood.o read_write.o anyoption.o
 
 start.o: start.cpp errors.cpp anyoption.h global_defs.h
-	g++ -fopenmp -c start.cpp errors.cpp 
+	$(CXX) $(CXXFLAGS) -std=c++14 -fopenmp $(CPPFLAGS) -c start.cpp errors.cpp
 
 beta.o: beta.cpp global_defs.h
-	g++ -fopenmp -c beta.cpp 
-      
+	$(CXX) $(CXXFLAGS) -std=c++14 -fopenmp $(CPPFLAGS) -c beta.cpp
+
 dirichlet.o: dirichlet.cpp global_defs.h
-	g++ -fopenmp -c dirichlet.cpp 
+	$(CXX) $(CXXFLAGS) -std=c++14 -fopenmp $(CPPFLAGS) -c dirichlet.cpp
 
 RJupdates.o: RJupdates.cpp global_defs.h
-	g++ -fopenmp -c RJupdates.cpp 
+	$(CXX) $(CXXFLAGS) -std=c++14 -fopenmp $(CPPFLAGS) -c RJupdates.cpp
 
 MHupdates.o: MHupdates.cpp global_defs.h
-	g++ -fopenmp -c MHupdates.cpp 
+	$(CXX) $(CXXFLAGS) -std=c++14 -fopenmp $(CPPFLAGS) -c MHupdates.cpp
 
 likelihood.o: likelihood.cpp global_defs.h
-	g++ -fopenmp -c likelihood.cpp 
+	$(CXX) $(CXXFLAGS) -std=c++14 -fopenmp $(CPPFLAGS) -c likelihood.cpp
 
 read_write.o: read_write.cpp errors.cpp global_defs.h
-	g++ -fopenmp -c read_write.cpp errors.cpp 
+	$(CXX) $(CXXFLAGS) -std=c++14 -fopenmp $(CPPFLAGS) -c read_write.cpp errors.cpp
 
-anyoption.o: anyoption.cpp anyoption.h 
-	g++ -fopenmp -c anyoption.cpp 
+anyoption.o: anyoption.cpp anyoption.h
+	$(CXX) $(CXXFLAGS) -std=c++14 -fopenmp $(CPPFLAGS) -c anyoption.cpp
 
-clean: 
+clean:
 	rm *.o bayescan_2.1
