diff --git a/CMakeLists.txt b/CMakeLists.txt
index fce4a26..b10e474 100755
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -35,19 +35,19 @@ project(mpboot)
 add_definitions(-DIQ_TREE)
 # The version number.
 set (iqtree_VERSION_MAJOR 1)
-set (iqtree_VERSION_MINOR 1)
-set (iqtree_VERSION_PATCH 1)
+set (iqtree_VERSION_MINOR 2)
+set (iqtree_VERSION_PATCH 0)
 
 set(BUILD_SHARED_LIBS OFF)
 
 message("IQ-TREE flags : ${IQTREE_FLAGS}")
 
-if (NOT CMAKE_BUILD_TYPE) 
+if (NOT CMAKE_BUILD_TYPE)
 	set(CMAKE_BUILD_TYPE "Release")
 endif()
 
-if (CMAKE_BUILD_TYPE STREQUAL "Release") 
-	message("Builde mode   : Release")
+if (CMAKE_BUILD_TYPE STREQUAL "Release")
+	message("Build mode   : Release")
 endif()
 
 if (CMAKE_GENERATOR MATCHES "Xcode")
@@ -104,11 +104,11 @@ set(CLANG "FALSE") # Clang compiler
 set(ICC "FALSE")   # Intel compiler
 set(VCC "FALSE")   # MS Visual C Compiler, note that it is different from MSVC variable
 
-if (CMAKE_COMPILER_IS_GNUCXX) 	
+if (CMAKE_COMPILER_IS_GNUCXX)
 	message("Compiler      : GNU Compiler (gcc)")
 	set(GCC "TRUE")
 #	set(COMBINED_FLAGS "-Wall -Wno-unused-function -Wno-sign-compare -pedantic -D_GNU_SOURCE -fms-extensions -Wno-deprecated")
-#	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++98")
+	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14")
 	set(CMAKE_CXX_FLAGS_RELEASE "-O3 -g0")
 	set(CMAKE_C_FLAGS_RELEASE "-O3 -g0")
 elseif (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
@@ -122,7 +122,7 @@ elseif (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
 		set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -stdlib=libc++")
         set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
 		SET(CMAKE_XCODE_ATTRIBUTE_CLANG_CXX_LIBRARY "libc++")
-    endif()	
+    endif()
 elseif (CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
 	set(VCC "TRUE")
 	message("Compiler      : MS Visual C++ Compiler")
@@ -462,7 +462,6 @@ endif()
 # add the install targets
 ##############################################################
 install (TARGETS mpboot DESTINATION bin)
-install (FILES "${PROJECT_SOURCE_DIR}/examples/example.phy" DESTINATION .)
 install (FILES "${PROJECT_SOURCE_DIR}/Documents/iqtree-manual-1.0.pdf" DESTINATION .)
 
 if (WIN32)
