diff -rNu bbmap.orig/bbduk.sh bbmap/bbduk.sh
--- bbmap.orig/bbduk.sh	2019-03-28 14:39:45.000000000 -0600
+++ bbmap/bbduk.sh	2019-04-22 15:15:41.795088854 -0600
@@ -373,8 +373,8 @@
 		module load samtools/1.4
 		module load pigz
 	fi
-	#local CMD="java -Djava.library.path=$NATIVELIBDIR $EA $z $z2 -cp $CP jgi.BBDuk $@"
-	local CMD="java $EA $z $z2 -cp $CP jgi.BBDuk $@"
+	local CMD="java -Djava.library.path=$NATIVELIBDIR $EA $z $z2 -cp $CP jgi.BBDuk $@"
+	#local CMD="java $EA $z $z2 -cp $CP jgi.BBDuk $@"
 	if [[ $silent == 0 ]] && [[ $json == 0 ]]; then
 		echo $CMD >&2
 	fi
diff -rNu bbmap.orig/bbmap.sh bbmap/bbmap.sh
--- bbmap.orig/bbmap.sh	2019-02-11 18:13:12.000000000 -0700
+++ bbmap/bbmap.sh	2019-04-22 15:12:54.583258066 -0600
@@ -371,8 +371,8 @@
 		module load samtools/1.4
 		module load pigz
 	fi
-	#local CMD="java -Djava.library.path=$NATIVELIBDIR $EA $z -cp $CP align2.BBMap build=1 overwrite=true fastareadlen=500 $@"
-	local CMD="java $EA $z -cp $CP align2.BBMap build=1 overwrite=true fastareadlen=500 $@"
+	local CMD="java -Djava.library.path=$NATIVELIBDIR $EA $z -cp $CP align2.BBMap build=1 overwrite=true fastareadlen=500 $@"
+	#local CMD="java $EA $z -cp $CP align2.BBMap build=1 overwrite=true fastareadlen=500 $@"
 	echo $CMD >&2
 	eval $CMD
 }
diff -rNu bbmap.orig/bbmerge-auto.sh bbmap/bbmerge-auto.sh
--- bbmap.orig/bbmerge-auto.sh	2019-02-07 12:45:54.000000000 -0700
+++ bbmap/bbmerge-auto.sh	2019-04-22 15:16:05.711069577 -0600
@@ -69,8 +69,8 @@
 		module load java/1.8.0_144
 		module load pigz
 	fi
-	#local CMD="java -Djava.library.path=$NATIVELIBDIR $EA $z $z2 -cp $CP jgi.BBMerge $@"
-	local CMD="java $EA $z $z2 -cp $CP jgi.BBMerge $@"
+	local CMD="java -Djava.library.path=$NATIVELIBDIR $EA $z $z2 -cp $CP jgi.BBMerge $@"
+	#local CMD="java $EA $z $z2 -cp $CP jgi.BBMerge $@"
 	echo $CMD >&2
 	eval $CMD
 }
diff -rNu bbmap.orig/bbmerge.sh bbmap/bbmerge.sh
--- bbmap.orig/bbmerge.sh	2019-02-21 15:50:04.000000000 -0700
+++ bbmap/bbmerge.sh	2019-04-22 15:14:05.515178226 -0600
@@ -271,8 +271,8 @@
 		module load java/1.8.0_144
 		module load pigz
 	fi
-	#local CMD="java -Djava.library.path=$NATIVELIBDIR $EA $z $z2 -cp $CP jgi.BBMerge $@"
-	local CMD="java $EA $EOOM $z $z2 -cp $CP jgi.BBMerge $@"
+	local CMD="java -Djava.library.path=$NATIVELIBDIR $EA $EOOM $z $z2 -cp $CP jgi.BBMerge $@"
+	#local CMD="java $EA $EOOM $z $z2 -cp $CP jgi.BBMerge $@"
 	echo $CMD >&2
 	eval $CMD
 }
diff -rNu bbmap.orig/mapPacBio.sh bbmap/mapPacBio.sh
--- bbmap.orig/mapPacBio.sh	2019-02-07 12:45:58.000000000 -0700
+++ bbmap/mapPacBio.sh	2019-04-22 15:16:29.103051800 -0600
@@ -68,8 +68,8 @@
 		module load samtools/1.4
 		module load pigz
 	fi
-	#local CMD="java -Djava.library.path=$NATIVELIBDIR $EA $z -cp $CP align2.BBMapPacBio build=1 overwrite=true minratio=0.40 fastareadlen=6000 ambiguous=best minscaf=100 startpad=10000 stoppad=10000 midpad=6000 $@"
-	local CMD="java $EA $z -cp $CP align2.BBMapPacBio build=1 overwrite=true minratio=0.40 fastareadlen=6000 ambiguous=best minscaf=100 startpad=10000 stoppad=10000 midpad=6000 $@"
+	local CMD="java -Djava.library.path=$NATIVELIBDIR $EA $z -cp $CP align2.BBMapPacBio build=1 overwrite=true minratio=0.40 fastareadlen=6000 ambiguous=best minscaf=100 startpad=10000 stoppad=10000 midpad=6000 $@"
+	#local CMD="java $EA $z -cp $CP align2.BBMapPacBio build=1 overwrite=true minratio=0.40 fastareadlen=6000 ambiguous=best minscaf=100 startpad=10000 stoppad=10000 midpad=6000 $@"
 	echo $CMD >&2
 	eval $CMD
 }
diff -rNu bbmap.orig/rqcfilter2.sh bbmap/rqcfilter2.sh
--- bbmap.orig/rqcfilter2.sh	2019-02-21 15:50:05.000000000 -0700
+++ bbmap/rqcfilter2.sh	2019-04-22 15:15:11.807114541 -0600
@@ -254,8 +254,8 @@
 		module load pigz
 		export TZ="America/Los_Angeles" 
 	fi
-	#local CMD="java -Djava.library.path=$NATIVELIBDIR $EA $z $z2 -cp $CP jgi.RQCFilter2 $@"
-	local CMD="java $EA $z $z2 -cp $CP jgi.RQCFilter2 $@"
+	local CMD="java -Djava.library.path=$NATIVELIBDIR $EA $z $z2 -cp $CP jgi.RQCFilter2 $@"
+	#local CMD="java $EA $z $z2 -cp $CP jgi.RQCFilter2 $@"
 	echo $CMD >&2
 	eval $CMD
 }
