diff --git a/setup.py b/setup.py
index 89afbb4..3e1ef30 100644
--- a/setup.py
+++ b/setup.py
@@ -43,7 +43,7 @@ packages = [this for this in packages if this not in ['test']]
 
 on_rtd = os.environ.get('READTHEDOCS', None) == 'True'
 if on_rtd:
-    extra_packages = []
+    extra_packages = ["pillow"]
 else:
     extra_packages = []
 
@@ -71,8 +71,8 @@ setup(
     # mock is for the test only
     install_requires = ["easydev>=0.9.31", "reports>=0.3.0", "matplotlib",
         "pyVCF", "pandas", "cutadapt>=1.9.1", "bioservices>=1.4.14",
-        "biokit>=0.3.4", "pysam", "docutils", "sphinx-gallery", "mock",
-        "numpydoc", "pillow", "blist", "sphinx", "ruamel.yaml>=0.13.2",
+        "biokit>=0.3.4", "pysam", "docutils", "mock",
+        "numpydoc",  "blist", "sphinx", "ruamel.yaml>=0.13.2",
         "colorlog"] + extra_packages,
 
     # here below '': pattern means include that pattern in all packages

