diff --git a/setup.py b/setup.py
index a135098..683ba46 100644
--- a/setup.py
+++ b/setup.py
@@ -43,7 +43,7 @@ packages = [this for this in packages if this not in ['test']]
 
 on_rtd = os.environ.get('READTHEDOCS', None) == 'True'
 if on_rtd:
-    extra_packages = ["pillow"]  # 
+    extra_packages = ["pillow", "numpydoc", "sphinx"]  # 
 else:
     extra_packages = []
 
@@ -70,11 +70,10 @@ setup(
     # pillow, sphinx-gallery and numpydoc are  for the doc only
     # mock is for the test only
     # qtconsole is required by Sequanix
-    install_requires = ["easydev>=0.9.31", "reports>=0.3.0", "matplotlib>=2.0.0",
+    install_requires = ["easydev>=0.9.32", "reports>=0.3.0", "matplotlib>=2.0.0",
         "pyVCF", "pandas", "cutadapt>=1.9.1", "bioservices>=1.4.14",
-        "biokit>=0.3.4", "pysam", "docutils", "mock", "qtconsole",
-        "numpydoc",  "blist", "sphinx", "ruamel.yaml>=0.13.2",
-        "colorlog"] + extra_packages,
+        "biokit>=0.4.1", "pysam", "docutils", "mock", "qtconsole",
+        "ruamel.yaml>=0.13.2", "colorlog"] + extra_packages,
 
     # here below '': pattern means include that pattern in all packages
     # so '' :['README.rst'] will include all README.rst recursively
