diff --git a/phase_common/src/models/haplotype_segment_single.h b/phase_common/src/models/haplotype_segment_single.h
index 23fab20..f576c4a 100644
--- a/phase_common/src/models/haplotype_segment_single.h
+++ b/phase_common/src/models/haplotype_segment_single.h
@@ -27,7 +27,11 @@
 #include <objects/compute_job.h>
 #include <objects/hmm_parameters.h>
 
+#ifdef __ARM_NEON
+#include "sse2neon.h"
+#else
 #include <immintrin.h>
+#endif
 #include <boost/align/aligned_allocator.hpp>
 
 template <typename T>
diff --git a/phase_common/src/models/haplotype_segment_double.h b/phase_common/src/models/haplotype_segment_double.h
index feb4d52..2a4996f 100644
--- a/phase_common/src/models/haplotype_segment_double.h
+++ b/phase_common/src/models/haplotype_segment_double.h
@@ -27,7 +27,11 @@
 #include <objects/compute_job.h>
 #include <objects/hmm_parameters.h>
 
+#ifdef __ARM_NEON
+#include "sse2neon.h"
+#else
 #include <immintrin.h>
+#endif
 #include <boost/align/aligned_allocator.hpp>
 
 template <typename T>
diff --git a/phase_rare/src/models/hmm_scaffold/hmm_scaffold_header.h b/phase_rare/src/models/hmm_scaffold/hmm_scaffold_header.h
index 02bfdd9..29d255e 100644
--- a/phase_rare/src/models/hmm_scaffold/hmm_scaffold_header.h
+++ b/phase_rare/src/models/hmm_scaffold/hmm_scaffold_header.h
@@ -29,7 +29,11 @@
 #include <objects/hmm_parameters.h>
 #include <containers/state_set.h>
 
+#ifdef __ARM_NEON
+#include "sse2neon.h"
+#else
 #include <immintrin.h>
+#endif
 
 inline
 float horizontal_add (const __m256& a) {
