diff --git a/scripts/compare_equiwidth_masks.pl b/scripts/compare_equiwidth_masks.pl
index 2f01778..05bdc78 100644
--- a/scripts/compare_equiwidth_masks.pl
+++ b/scripts/compare_equiwidth_masks.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 #
 # compare_equiwidth_masks.pl
 # Eric Nawrocki
diff --git a/scripts/rename_mask.pl b/scripts/rename_mask.pl
index bbeb92e..3b64073 100644
--- a/scripts/rename_mask.pl
+++ b/scripts/rename_mask.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 #
 # rename_mask.pl
 # Eric Nawrocki
diff --git a/scripts/stk2fa.pl b/scripts/stk2fa.pl
index fb918e5..870e4be 100644
--- a/scripts/stk2fa.pl
+++ b/scripts/stk2fa.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 #
 # stk2fa.pl 
 # Eric Nawrocki
diff --git a/scripts/summarize_mask.pl b/scripts/summarize_mask.pl
index 2bc5964..1f656fa 100644
--- a/scripts/summarize_mask.pl
+++ b/scripts/summarize_mask.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 #
 # summarize_mask.pl
 # Eric Nawrocki
diff --git a/src/ssu-align b/src/ssu-align
index b845182..cef594d 100755
--- a/src/ssu-align
+++ b/src/ssu-align
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 #
 # ssu-align
 # Eric Nawrocki
diff --git a/src/ssu-build b/src/ssu-build
index 76c14c0..040654a 100755
--- a/src/ssu-build
+++ b/src/ssu-build
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 #
 # ssu-build
 # Eric Nawrocki
diff --git a/src/ssu-draw b/src/ssu-draw
index 5a58b52..aa5bebd 100755
--- a/src/ssu-draw
+++ b/src/ssu-draw
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 #
 # ssu-draw
 # Eric Nawrocki
diff --git a/src/ssu-mask b/src/ssu-mask
index c2b432b..9389567 100755
--- a/src/ssu-mask
+++ b/src/ssu-mask
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 #
 # ssu-mask
 # Eric Nawrocki
diff --git a/src/ssu-merge b/src/ssu-merge
index 3069dc8..26ac288 100755
--- a/src/ssu-merge
+++ b/src/ssu-merge
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 #
 # ssu-merge
 # Eric Nawrocki
diff --git a/src/ssu-prep b/src/ssu-prep
index 0840473..662c902 100755
--- a/src/ssu-prep
+++ b/src/ssu-prep
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 #
 # ssu-prep
 # Eric Nawrocki
diff --git a/src/ssu.pm b/src/ssu.pm
index 82b5f4d..aac21be 100644
--- a/src/ssu.pm
+++ b/src/ssu.pm
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 #
 # ssu.pm
 # Eric Nawrocki
