diff --git a/src/cpp_matching/matching/matcher.cpp b/src/cpp_matching/matching/matcher.cpp
index ca9d34d..a7f37de 100644
--- a/src/cpp_matching/matching/matcher.cpp
+++ b/src/cpp_matching/matching/matcher.cpp
@@ -5,6 +5,7 @@
 #include <vector>
 #include <limits>
 #include <utility>
+#include <algorithm>
 using namespace std;
 
 
@@ -264,4 +265,4 @@ vector<MatchInfo> get_matches(const unordered_map<BGC_Variant_ID , HMM>& hmms,
         matches.push_back(get_full_match_info(match_light, hmms.at(match_light.bgc_variant_id)));
     }
     return matches;
-}
\ No newline at end of file
+}
diff --git a/src/cpp_matching/matching/viterbi_algorithm.cpp b/src/cpp_matching/matching/viterbi_algorithm.cpp
index b1f7a6e..d87c548 100644
--- a/src/cpp_matching/matching/viterbi_algorithm.cpp
+++ b/src/cpp_matching/matching/viterbi_algorithm.cpp
@@ -3,6 +3,7 @@
 #include <vector>
 #include <limits>
 #include <utility>
+#include <algorithm>
 using namespace std;
 
 // Computes the maximum probability path through an HMM given an input sequence.
@@ -136,4 +137,4 @@ get_opt_hmm_path(const HMM& hmm,
     assert(path.back() == final_state && "the path endpoints are wrong!!!");
 
     return { dp[final_state][seq_len], path };
-}
\ No newline at end of file
+}
