Fix ARB refusing to build with `--std=c++17` 
diff -rNu arbsrc_15220.orig/TEMPLATES/cxxforward.h arbsrc_15220/TEMPLATES/cxxforward.h
--- arbsrc_15220.orig/TEMPLATES/cxxforward.h	2016-07-08 03:39:15.000000000 -0600
+++ arbsrc_15220/TEMPLATES/cxxforward.h	2019-02-12 20:01:53.000000000 -0700
@@ -24,7 +24,7 @@
 #   if (__cplusplus == 201103L)
 #    define ARB_ENABLE_Cxx11_FEATURES
 #   else
-#    if (__cplusplus == 201402L)
+#    if (__cplusplus >= 201402L)
 #     define ARB_ENABLE_Cxx11_FEATURES
 // #     define ARB_ENABLE_Cxx14_FEATURES // not needed yet
 #    else
