--- GDE/Makefile.orig	2017-03-13 16:21:54.000000000 -0600
+++ GDE/Makefile	2017-03-13 16:34:08.000000000 -0600
@@ -43,6 +43,10 @@

 ARCHS=$(ARCHS_DEPENDS) $(ARCHS_NO_DEPENDS)

+# Allow disabling build of packages individually via environment variable
+ARB_BUILD_SKIP_PKGS?=""
+ARCHS:=$(filter-out $(patsubst %,%/%.dummy,$(ARB_BUILD_SKIP_PKGS)), $(ARCHS))
+
 # --------------------------------------------------------------------------------
 # modify cflags for submakefiles
