diff -rNu arbsrc_15220.orig/SH/arb_readlink arbsrc_15220/SH/arb_readlink
--- arbsrc_15220.orig/SH/arb_readlink	2016-05-02 05:19:20.000000000 -0600
+++ arbsrc_15220/SH/arb_readlink	2019-02-13 15:03:44.427236093 -0700
@@ -1,9 +1,2 @@
 #!/bin/bash
-
-if [[ `uname` == Darwin* ]]
-then
-    greadlink "$@"
-else
-    readlink "$@"
-fi
-
+exec readlink "$@"
diff -rNu arbsrc_15220.orig/SH/arb_sed arbsrc_15220/SH/arb_sed
--- arbsrc_15220.orig/SH/arb_sed	2016-05-02 05:19:20.000000000 -0600
+++ arbsrc_15220/SH/arb_sed	2017-06-18 19:11:31.000000000 -0600
@@ -1,9 +1,2 @@
 #!/bin/bash
-
-if [[ `uname` == Darwin* ]]
-then
-    gsed "$@"
-else
-    sed "$@"
-fi
-
+exec sed "$@"
