--- Makefile.orig	2017-03-23 13:58:59.000000000 -0600
+++ Makefile	2017-03-23 14:01:56.000000000 -0600
@@ -544,6 +544,9 @@
 SHARED_LIB_SUFFIX = a# static lib suffix
 LINK_SHARED_LIB := $(LINK_STATIC_LIB)
 else
+  ifeq ($(USE_CLANG),1)
+    clflags += -undefined dynamic_lookup
+  endif
 SHARED_LIB_SUFFIX = so# shared lib suffix
 LINK_SHARED_LIB := $(A_CXX) $(clflags) -shared $(GCOVFLAGS) -o# link shared lib
 endif
