diff -rNu arbsrc_15220/AWT/AWT_input_mask.cxx arbsrc_15220.orig/AWT/AWT_input_mask.cxx
--- arbsrc_15220/AWT/AWT_input_mask.cxx	2019-02-13 17:09:54.427542915 -0700
+++ arbsrc_15220.orig/AWT/AWT_input_mask.cxx	2016-08-22 04:27:48.000000000 -0600
@@ -2211,7 +2211,7 @@
     GB_ERROR         error      = 0;
 
     if (registered == registeredTypes.end()) error = GBS_global_string("Type '%s' not registered (yet)", awt_itemtype_names[type]);
-    else registered->second.getOpenCb()(registered->second.getWindow(), (AW_CL)mask_id, (AW_CL)0);
+    else registered->second.getOpenCb()(registered->second.getWindow(), (AW_CL)mask_id, nullptr);
 
     return error;
 }
