diff --git a/scripts/Assemblytics b/scripts/Assemblytics
index 635f0de..3eb0ea2 100755
--- a/scripts/Assemblytics
+++ b/scripts/Assemblytics
@@ -60,7 +60,11 @@ mkdir -p ${OUTPUT_PREFIX%/*}/
 LOG_FILE=${OUTPUT_PREFIX%/*}/progress.log
 >&2 echo "Logging progress updates in $LOG_FILE"
 
-SCRIPT_PATH=$(dirname $BASH_SOURCE)
+SCRIPT_PATH=$(cd "$(dirname $BASH_SOURCE)/../libexec/assemblytics" > /dev/null 2>&1 && pwd)
+if [ $? != 0 ]
+then
+    SCRIPT_PATH=$(dirname $BASH_SOURCE)
+fi
 echo 'script path:' $SCRIPT_PATH
 
 # Check that Rscript is available (needed by R scripts)
