diff --git a/Cargo.toml b/Cargo.toml
index 5a1396a..ba0edb5 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -17,7 +17,7 @@ path = "src/main.rs"
 [dependencies]
 rust-htslib = { version = "0.38.2", features = ["libdeflate", "static"] }
 #bitpacking = "0.8.4"
-stream-vbyte = { version = "0.4.1", features = ["x86_ssse3", "x86_sse41"] }
+stream-vbyte = "0.4.1"
 clap = { version = "~2.27.0", features = ["suggestions"] }
 c2rust-bitfields = "0.3.0"
 libc = "*"
