From 4f74f7814a440103e82f93487693b86f2ec47dbc Mon Sep 17 00:00:00 2001
From: Nick Waters <nickp60@users.noreply.github.com>
Date: Thu, 13 Nov 2025 15:20:05 -0500
Subject: [PATCH] env-config

---
 humann/config.py | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/humann/config.py b/humann/config.py
index 6d556d5..1cf196c 100644
--- a/humann/config.py
+++ b/humann/config.py
@@ -102,10 +102,14 @@ def log_settings():
     logger.info("\nRun config settings: \n\n" + "\n".join(lines))
 
 # User config file
-user_edit_config_file="humann.cfg"
+# User config file
+if not os.environ.get("HUMANN_CONFIG"):
+    user_edit_config_file="humann.cfg"
 
-full_path_user_edit_config_file=os.path.join(os.path.dirname(os.path.abspath(__file__)),
-    user_edit_config_file)
+    full_path_user_edit_config_file=os.path.join(os.path.dirname(os.path.abspath(__file__)),
+                                                 user_edit_config_file)
+else:
+    full_path_user_edit_config_file=os.environ.get("HUMANN_CONFIG")
 
 def update_user_edit_config_file_single_item(section,name,value):
     """
-- 
2.39.5 (Apple Git-154)

