diff --git a/build.py b/build.py
index fec5a56..bab6ca9 100644
--- a/build.py
+++ b/build.py
@@ -47,12 +47,11 @@ def compile_htslib():
             CFLAGS += " -mmacosx-version-min=11.0"
         CFLAGS += "'"
         commands = []
-        if not Path("config.h").exists():
-            commands = [
-                "autoreconf -i",
-                f"./configure --with-libdeflate --enable-lzma --enable-bz2 {CFLAGS}",
-            ]
-        commands.append("make -j")
+        commands = [
+            "autoreconf -i",
+            f"./configure --with-libdeflate --enable-lzma --enable-bz2 {CFLAGS}",
+            "make -j",
+        ]
         for command in commands:
             retcode = subprocess.call(
                 f"{command}", stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True
