--- a/core/jms-implementation/support-mini-x86-32/initial_setup.py
+++ b/core/jms-implementation/support-mini-x86-32/initial_setup.py
@@ -60,9 +60,20 @@ if __name__ == "__main__":
         hmmpress_force = True
     ipr_properties = load_properties('interproscan.properties')
     hmmer3_dir = ipr_properties['binary.hmmer3.path']
+    try:
+        hmmer3_dir
+        os.path.join(hmmer3_dir, '') # will add the trailing slash if it's not already there.
+    except NameError:
+        hmmer3_dir=''
+
     hmmer33_dir = ipr_properties['binary.hmmer33.path']
-    hmmpress_path = hmmer3_dir + '/hmmpress'
-    hmmpress33_path = hmmer33_dir + '/hmmpress'
+    try:
+        hmmer33_dir
+        os.path.join(hmmer33_dir, '') # will add the trailing slash if it's not already there.
+    except NameError:
+        hmmer33_dir=''
+    hmmpress_path = hmmer3_dir + 'hmmpress'
+    hmmpress33_path = hmmer33_dir + 'hmmpress'
     hmm_models_paths = get_hmm_models_props(ipr_properties, hmmpress_force)
     if (len(hmm_models_paths) > 0):
         print("Checking any hmm models that need indexing ... this may take a few minutes")
