--- ./core/jms-implementation/support-mini-x86-32/interproscan.sh	2022-10-14 17:28:01.000000000 +0200
+++ ./core/jms-implementation/support-mini-x86-32/interproscan.sh	2022-12-05 17:42:17.443390814 +0100
@@ -49,10 +49,10 @@
     exit 1
 fi
 
-if ! python3 setup.py "$PROPERTIES"
-then
-    exit 1
-fi
+#if ! python3 setup.py "$PROPERTIES"
+#then
+#    exit 1
+#fi
 
 "$JAVA" \
  -XX:ParallelGCThreads=8 \
