diff --git a/etc/configuration.cfg b/etc/configuration.cfg
index f3d8283..4ead0ea 100755
--- a/etc/configuration.cfg
+++ b/etc/configuration.cfg
@@ -1,38 +1,36 @@
 [paths]
-python = /usr/bin/
+python = /opt/anaconda1anaconda2anaconda3/bin/
 #
 # for changing the shebang for all Python scripts used by FusionCatcher 
 # to '#!/some/other/python'
 # one could run 'fusioncatcher/bin/shebang.py' -s #!/some/other/python
 #
-data = ../data/current/
-scripts = ../bin/
-bowtie = ../tools/bowtie/
-bowtie2 = ../tools/bowtie2/
-blat = ../tools/blat/
-#bwa = ../tools/bwa/
-bbmap = ../tools/bbmap/
-liftover = ../tools/liftover/
-#velvet = ../tools/velvet/
-#oases = ../tools/oases/
-fatotwobit = ../tools/fatotwobit/
-#samtools = ../tools/samtools/
-seqtk = ../tools/seqtk/
-star = ../tools/star/source/
-sra = ../tools/sratoolkit/bin/
-#numpy = ../tools/numpy/
-biopython = ../tools/biopython/
-#xlrd = ../tools/xlrd/
-#openpyxl = ../tools/openpyxl/
-#lzo = ../tools/lzo/
-#lzop = ../tools/lzop/src/
-#coreutils = ../tools/coreutils/src/
-#parallel = ../tools/paralell/src/
-pigz = ../tools/pigz/
-#pxz = ../tools/pxz/
-picard = ../tools/picard/
-#python3 = /usr/bin/
-java = /usr/bin/
+data = ${FC_DB_PATH}/current/
+scripts = /opt/anaconda1anaconda2anaconda3/bin/
+bowtie = /opt/anaconda1anaconda2anaconda3/bin/
+bowtie2 = /opt/anaconda1anaconda2anaconda3/bin/
+blat = /opt/anaconda1anaconda2anaconda3/bin/
+bwa = /opt/anaconda1anaconda2anaconda3/bin/
+bbmap = /opt/anaconda1anaconda2anaconda3/bin/
+liftover = /opt/anaconda1anaconda2anaconda3/bin/
+velvet = /opt/anaconda1anaconda2anaconda3/bin/
+oases = /opt/anaconda1anaconda2anaconda3/bin/
+fatotwobit = /opt/anaconda1anaconda2anaconda3/bin/
+samtools = /opt/anaconda1anaconda2anaconda3/bin/
+seqtk = /opt/anaconda1anaconda2anaconda3/bin/
+star = /opt/anaconda1anaconda2anaconda3/bin/
+sra = /opt/anaconda1anaconda2anaconda3/bin/
+numpy = /opt/anaconda1anaconda2anaconda3/lib/python*/site-packages/numpy/
+biopython = /opt/anaconda1anaconda2anaconda3/lib/python*/site-packages/
+xlrd = /opt/anaconda1anaconda2anaconda3/lib/python*/site-packages/xlrd/
+openpyxl = /opt/anaconda1anaconda2anaconda3/lib/python*/site-packages/openpyxl/
+lzo = /opt/anaconda1anaconda2anaconda3/bin/
+lzop = /opt/anaconda1anaconda2anaconda3/bin/
+coreutils = /opt/anaconda1anaconda2anaconda3/bin/
+pigz = /opt/anaconda1anaconda2anaconda3/bin/
+pxz = /opt/anaconda1anaconda2anaconda3/bin/
+picard = /opt/anaconda1anaconda2anaconda3/bin/
+java = /opt/anaconda1anaconda2anaconda3/bin/
 
 [parameters]
 threads = 0
@@ -40,9 +38,3 @@ aligners = blat,star
 
 [versions]
 fusioncatcher = 1.30
-
-
-
-
-
-
