diff --git a/etc/configuration.cfg b/etc/configuration.cfg
index 936ec10..6d46e14 100755
--- a/etc/configuration.cfg
+++ b/etc/configuration.cfg
@@ -1,30 +1,30 @@
 [paths]
-python = /usr/bin/
+python = /opt/anaconda1anaconda2anaconda3/bin/
 #
 # for changing the shebang for all Python scripts used by FusionCatcher 
 # to '#!/some/other/python'
 # one could run 'fusioncatcher/bin/shebang.py' -s #!/some/other/python
 #
-data = ../data/current/
-scripts = ../bin/
-bowtie = ../tools/bowtie/
-bowtie2 = ../tools/bowtie2/
-blat = ../tools/blat/
-bbmap = ../tools/bbmap/
-liftover = ../tools/liftover/
+data = ${FC_DB_PATH}/current/
+scripts = /opt/anaconda1anaconda2anaconda3/bin/
+bowtie = /opt/anaconda1anaconda2anaconda3/bin/
+bowtie2 = /opt/anaconda1anaconda2anaconda3/bin/
+blat = /opt/anaconda1anaconda2anaconda3/bin/
+bbmap = /opt/anaconda1anaconda2anaconda3/bin/
+liftover = /opt/anaconda1anaconda2anaconda3/bin/
 #velvet = ../tools/velvet/
 #oases = ../tools/oases/
-fatotwobit = ../tools/fatotwobit/
-seqtk = ../tools/seqtk/
-star = ../tools/star/source/
-sra = ../tools/sratoolkit/bin/
-biopython = ../tools/biopython/
-parallel = ../tools/parallel/src/
-pigz = ../tools/pigz/
-picard = ../tools/picard/
-fastqtk = ../tools/fastqtk/
+fatotwobit = /opt/anaconda1anaconda2anaconda3/bin/
+seqtk = /opt/anaconda1anaconda2anaconda3/bin/
+star = /opt/anaconda1anaconda2anaconda3/bin/
+sra = /opt/anaconda1anaconda2anaconda3/bin/
+biopython = /opt/anaconda1anaconda2anaconda3/lib/python*/site-packages/
+parallel = /opt/anaconda1anaconda2anaconda3/bin/
+pigz = /opt/anaconda1anaconda2anaconda3/bin/
+picard = /opt/anaconda1anaconda2anaconda3/bin/
+fastqtk = /opt/anaconda1anaconda2anaconda3/bin/
 #python3 = /usr/bin/
-java = /usr/bin/
+java = /opt/anaconda1anaconda2anaconda3/bin/
 
 [parameters]
 threads = 0
