--- etc/configuration.cfg	2019-10-29 15:29:05.000000000 +0100
+++ etc/configuration.cfg2	2019-11-08 22:52:17.187637905 +0100
@@ -1,37 +1,37 @@
 [paths]
-python = /usr/bin/
+python = /opt/anaconda1anaconda2anaconda3/bin/
 #
 # for changing the shebang for all Python scripts used by FusionCatcher 
 # to '#!/some/other/python'
 # one could run 'fusioncatcher/bin/shebang.py' -s #!/some/other/python
 #
-data = ../data/current/
-scripts = ../bin/
-bowtie = ../tools/bowtie/
-bowtie2 = ../tools/bowtie2/
-blat = ../tools/blat/
-bwa = ../tools/bwa/
-bbmap = ../tools/bbmap/
-liftover = ../tools/liftover/
-velvet = ../tools/velvet/
-oases = ../tools/oases/
-fatotwobit = ../tools/blat/
-samtools = ../tools/samtools/
-seqtk = ../tools/seqtk/
-star = ../tools/star/source/
-sra = ../tools/sratoolkit/bin/
-numpy = ../tools/numpy/
-biopython = ../tools/biopython/
-xlrd = ../tools/xlrd/
-openpyxl = ../tools/openpyxl/
-lzo = ../tools/lzo/
-lzop = ../tools/lzop/src/
-coreutils = ../tools/coreutils/src/
-parallel = ../tools/paralell/src/
-pigz = ../tools/pigz/
-pxz = ../tools/pxz/
-picard = ../tools/picard/
-java = /usr/bin/
+data = ${FC_DB_PATH}/current/
+scripts = /opt/anaconda1anaconda2anaconda3/bin/
+bowtie = /opt/anaconda1anaconda2anaconda3/bin/
+bowtie2 = /opt/anaconda1anaconda2anaconda3/bin/
+blat = /opt/anaconda1anaconda2anaconda3/bin/
+bwa = /opt/anaconda1anaconda2anaconda3/bin/
+bbmap = /opt/anaconda1anaconda2anaconda3/bin/
+liftover = /opt/anaconda1anaconda2anaconda3/bin/
+velvet = /opt/anaconda1anaconda2anaconda3/bin/
+oases = /opt/anaconda1anaconda2anaconda3/bin/
+fatotwobit = /opt/anaconda1anaconda2anaconda3/bin/
+samtools = /opt/anaconda1anaconda2anaconda3/bin/
+seqtk = /opt/anaconda1anaconda2anaconda3/tools/seqtk/
+star = /opt/anaconda1anaconda2anaconda3/bin/
+sra = /opt/anaconda1anaconda2anaconda3/bin/
+numpy = /opt/anaconda1anaconda2anaconda3/lib/python*/site-packages/numpy/
+biopython = /opt/anaconda1anaconda2anaconda3/lib/python*/site-packages/
+xlrd = /opt/anaconda1anaconda2anaconda3/lib/python*/site-packages/xlrd/
+openpyxl = /opt/anaconda1anaconda2anaconda3/lib/python*/site-packages/openpyxl/
+lzo = /opt/anaconda1anaconda2anaconda3/bin/
+lzop = /opt/anaconda1anaconda2anaconda3/bin/
+coreutils = /opt/anaconda1anaconda2anaconda3/bin/
+parallel = /opt/anaconda1anaconda2anaconda3/bin/
+pigz = /opt/anaconda1anaconda2anaconda3/bin/
+pxz = /opt/anaconda1anaconda2anaconda3/bin/
+picard = /opt/anaconda1anaconda2anaconda3/bin/
+java = /opt/anaconda1anaconda2anaconda3/bin/
 
 [parameters]
 threads = 0
