--- etc/configuration.cfg	2018-10-17 18:23:40.361288543 -0500
+++ etc/configuration.cfg.new	2018-10-17 18:46:16.142296785 -0500
@@ -1,37 +1,35 @@
 [paths]
-python = /usr/bin/
+python = ${CONDA_PREFIX}/bin/
 #
 # for changing the shebang for all Python scripts used by FusionCatcher 
 # to '#!/some/other/python'
 # one could run 'fusioncatcher/bin/shebang.py' -s #!/some/other/python
 #
-data = /apps/fusioncatcher/data/current/
-scripts = /apps/fusioncatcher/bin/
-bowtie = /apps/fusioncatcher/tools/bowtie/
-bowtie2 = /apps/fusioncatcher/tools/bowtie2/
-blat = /apps/fusioncatcher/tools/blat/
-bwa = /apps/fusioncatcher/tools/bwa/
-bbmap = /apps/fusioncatcher/tools/bbmap/
-liftover = /apps/fusioncatcher/tools/liftover/
-velvet = /apps/fusioncatcher/tools/velvet/
-oases = /apps/fusioncatcher/tools/oases/
-fatotwobit = /apps/fusioncatcher/tools/blat/
-samtools = /apps/fusioncatcher/tools/samtools/
-seqtk = /apps/fusioncatcher/tools/seqtk/
-star = /apps/fusioncatcher/tools/star/source/
-sra = /apps/fusioncatcher/tools/sratoolkit/bin/
-numpy = /apps/fusioncatcher/tools/numpy/
-biopython = /apps/fusioncatcher/tools/biopython/
-xlrd = /apps/fusioncatcher/tools/xlrd/
-openpyxl = /apps/fusioncatcher/tools/openpyxl/
-lzo = /apps/fusioncatcher/tools/lzo/
-lzop = /apps/fusioncatcher/tools/lzop/src/
-coreutils = /apps/fusioncatcher/tools/coreutils/src/
-parallel = /appsfusioncatcher/tools/paralell/src/
-pigz = /apps/fusioncatcher/tools/pigz/
-pxz = /apps/fusioncatcher/tools/pxz/
-picard = /apps/fusioncatcher/tools/picard/
-java = /usr/bin/
+data = ${FC_DB_PATH}/current/
+scripts = ${CONDA_PREFIX}/bin/
+bowtie =  ${CONDA_PREFIX}/bin/
+bowtie2 =  ${CONDA_PREFIX}/bin/
+blat =  ${CONDA_PREFIX}/bin/
+bwa =  ${CONDA_PREFIX}/bin/
+bbmap =  ${CONDA_PREFIX}/bin/
+liftover = ${CONDA_PREFIX}/bin/  
+velvet =  ${CONDA_PREFIX}/bin/
+oases =  ${CONDA_PREFIX}/bin/
+fatotwobit =  ${CONDA_PREFIX}/bin/
+samtools =  ${CONDA_PREFIX}/bin/
+seqtk =  ${CONDA_PREFIX}/bin/
+star =  ${CONDA_PREFIX}/bin/
+sra = ${CONDA_PREFIX}/bin/
+numpy = ${CONDA_PREFIX}/lib/python*/site-packages/numpy/
+biopython = ${CONDA_PREFIX}/lib/python*/site-packages/
+xlrd = ${CONDA_PREFIX}/lib/python*/site-packages/xlrd/
+openpyxl = ${CONDA_PREFIX}/lib/python*/site-packages/openpyxl/
+lzop = ${CONDA_PREFIX}/bin/
+coreutils = ${CONDA_PREFIX}/bin/
+parallel = ${CONDA_PREFIX}/bin/
+pigz = ${CONDA_PREFIX}/bin/
+picard = ${CONDA_PREFIX}/bin/
+java = ${CONDA_PREFIX}/bin/
 
 [parameters]
 threads = 0
@@ -39,9 +37,3 @@
 
 [versions]
 fusioncatcher = 1.00
-
-
-
-
-
-
--- bin/sort_ttdb.py	2018-10-17 18:23:40.372288599 -0500
+++ bin/sort_ttdb.py.new	2018-10-17 18:23:40.378288630 -0500
@@ -1,4 +1,4 @@
-#!/mnt/software/bin/python2.7-real
+#!/usr/bin/env python
 # -*- coding: utf-8 -*-
 """
 It sorts the input file (text tab separated file) based on the specified columns.
--- bin/add_custom_gene.py	2018-10-17 18:23:40.362288548 -0500
+++ bin/add_custom_gene.py.new	2018-10-17 18:23:40.377288625 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/usr/bin/env python
 # -*- coding: utf-8 -*-
 """
 Add the custom human genes which are missing from the Ensembl database.
--- bin/add_gap_gene.py	2018-10-17 18:23:40.371288594 -0500
+++ bin/add_gap_gene.py.new	2018-10-17 18:23:40.377288625 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/usr/bin/env python
 # -*- coding: utf-8 -*-
 """
 Add the custom human genes which are known to have random inserts at fusion junction.
