diff --git a/bin/add_custom_gene.py b/bin/add_custom_gene.py
index efda63c..6024d38 100755
--- a/bin/add_custom_gene.py
+++ b/bin/add_custom_gene.py
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/usr/bin/env python
 # -*- coding: utf-8 -*-
 """
 Add the custom human genes which are missing from the Ensembl database.
diff --git a/bin/add_gap_gene.py b/bin/add_gap_gene.py
index 06f0bbf..91f6831 100755
--- a/bin/add_gap_gene.py
+++ b/bin/add_gap_gene.py
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/usr/bin/env python
 # -*- coding: utf-8 -*-
 """
 Add the custom human genes which are known to have random inserts at fusion junction.
diff --git a/bin/sort_ttdb.py b/bin/sort_ttdb.py
index d5bfa38..193886f 100755
--- a/bin/sort_ttdb.py
+++ b/bin/sort_ttdb.py
@@ -1,4 +1,4 @@
-#!/mnt/software/bin/python2.7-real
+#!/usr/bin/env python
 # -*- coding: utf-8 -*-
 """
 It sorts the input file (text tab separated file) based on the specified columns.
diff --git a/etc/configuration.cfg b/etc/configuration.cfg
index c217883..4c82a8d 100644
--- a/etc/configuration.cfg
+++ b/etc/configuration.cfg
@@ -1,37 +1,37 @@
 [paths]
-python = /usr/bin/
+python = /opt/anaconda1anaconda2anaconda3/bin/
 #
 # for changing the shebang for all Python scripts used by FusionCatcher 
 # to '#!/some/other/python'
 # one could run 'fusioncatcher/bin/shebang.py' -s #!/some/other/python
 #
-data = /apps/fusioncatcher/data/current/
-scripts = /apps/fusioncatcher/bin/
-bowtie = /apps/fusioncatcher/tools/bowtie/
-bowtie2 = /apps/fusioncatcher/tools/bowtie2/
-blat = /apps/fusioncatcher/tools/blat/
-bwa = /apps/fusioncatcher/tools/bwa/
-bbmap = /apps/fusioncatcher/tools/bbmap/
-liftover = /apps/fusioncatcher/tools/liftover/
-velvet = /apps/fusioncatcher/tools/velvet/
-oases = /apps/fusioncatcher/tools/oases/
-fatotwobit = /apps/fusioncatcher/tools/blat/
-samtools = /apps/fusioncatcher/tools/samtools/
-seqtk = /apps/fusioncatcher/tools/seqtk/
-star = /apps/fusioncatcher/tools/star/source/
-sra = /apps/fusioncatcher/tools/sratoolkit/bin/
-numpy = /apps/fusioncatcher/tools/numpy/
-biopython = /apps/fusioncatcher/tools/biopython/
-xlrd = /apps/fusioncatcher/tools/xlrd/
-openpyxl = /apps/fusioncatcher/tools/openpyxl/
-lzo = /apps/fusioncatcher/tools/lzo/
-lzop = /apps/fusioncatcher/tools/lzop/src/
-coreutils = /apps/fusioncatcher/tools/coreutils/src/
-parallel = /appsfusioncatcher/tools/paralell/src/
-pigz = /apps/fusioncatcher/tools/pigz/
-pxz = /apps/fusioncatcher/tools/pxz/
-picard = /apps/fusioncatcher/tools/picard/
-java = /usr/bin/
+data = ${FC_DB_PATH}//current/
+scripts = /opt/anaconda1anaconda2anaconda3/bin/
+bowtie = /opt/anaconda1anaconda2anaconda3/bin/
+bowtie2 = /opt/anaconda1anaconda2anaconda3/bin/
+blat = /opt/anaconda1anaconda2anaconda3/bin/
+bwa = /opt/anaconda1anaconda2anaconda3/bin/
+bbmap = /opt/anaconda1anaconda2anaconda3/bin/
+liftover = /opt/anaconda1anaconda2anaconda3/bin/
+velvet = /opt/anaconda1anaconda2anaconda3/bin/
+oases = /opt/anaconda1anaconda2anaconda3/bin/
+fatotwobit = /opt/anaconda1anaconda2anaconda3/bin/
+samtools = /opt/anaconda1anaconda2anaconda3/bin/
+seqtk = /opt/anaconda1anaconda2anaconda3/tools/seqtk/
+star = /opt/anaconda1anaconda2anaconda3/bin/
+sra = /opt/anaconda1anaconda2anaconda3/bin/
+numpy = /opt/anaconda1anaconda2anaconda3/lib/python*/site-packages/numpy/
+biopython = /opt/anaconda1anaconda2anaconda3/lib/python*/site-packages/
+xlrd = /opt/anaconda1anaconda2anaconda3/lib/python*/site-packages/xlrd/
+openpyxl = /opt/anaconda1anaconda2anaconda3/lib/python*/site-packages/openpyxl/
+lzo = /opt/anaconda1anaconda2anaconda3/bin/
+lzop = /opt/anaconda1anaconda2anaconda3/bin/
+coreutils = /opt/anaconda1anaconda2anaconda3/bin/
+parallel = /opt/anaconda1anaconda2anaconda3/bin/
+pigz = /opt/anaconda1anaconda2anaconda3/bin/
+pxz = /opt/anaconda1anaconda2anaconda3/bin/
+picard = /opt/anaconda1anaconda2anaconda3/bin/
+java = /opt/anaconda1anaconda2anaconda3/bin/
 
 [parameters]
 threads = 1
