--- setup.py
+++ setup.py
@@ -318 +318 @@
-        cqfcmd = ['bash', '-c', 'cd third-party/cqf && make']
+        cqfcmd = ['bash', '-c', 'cd third-party/cqf && cxx="${CXX} ${CXXFLAGS} ${LDFLAGS} -std=c++11" && make CC="${cxx} -fPIC" CXX="${cxx} -fPIC" LD="${cxx}"']
--- setup.cfg
+++ setup.cfg
@@ -6,5 +6,5 @@
 # docker/Dockerfile
-# libraries = z,bz2
+libraries = z,bz2
 ## if using system libraries
-include-dirs = include:third-party/zlib:third-party/bzip2:third-party/seqan/core/include:third-party/smhasher:third-party/cqf:third-party/rollinghash
+include-dirs = include:third-party/seqan/core/include:third-party/smhasher:third-party/cqf:third-party/rollinghash
 # include-dirs = lib
