--- khmer-2.1.1/setup.py	2017-05-25 20:10:21.000000000 -0400
+++ khmer-2.1.1/setup.py	2024-08-07 17:30:48.930797570 -0400
@@ -259,7 +259,7 @@
         if sys.platform == 'darwin' and 'gcov' in self.libraries:
             self.libraries.remove('gcov')
 
-        if "z" not in self.libraries:
+        if False:
             zcmd = ['bash', '-c', 'cd ' + ZLIBDIR + ' && ( test Makefile -nt'
                     ' configure || bash ./configure --static ) && make -f '
                     'Makefile.pic PIC']
@@ -269,7 +269,7 @@
                     "adler32", "compress", "crc32", "deflate", "gzclose",
                     "gzlib", "gzread", "gzwrite", "infback", "inffast",
                     "inflate", "inftrees", "trees", "uncompr", "zutil"])
-        if "bz2" not in self.libraries:
+        if False:
             bz2cmd = ['bash', '-c', 'cd ' + BZIP2DIR + ' && make -f '
                       'Makefile-libbz2_so all']
             spawn(cmd=bz2cmd, dry_run=self.dry_run)
