diff --git a/third-party/cqf/Makefile b/third-party/cqf/Makefile
index 33769ef..703a36b 100644
--- a/third-party/cqf/Makefile
+++ b/third-party/cqf/Makefile
@@ -31,7 +31,7 @@ CXX = g++ -std=c++11 -fPIC
 CC = g++ -std=c++11 -fPIC
 LD= g++ -std=c++11
 
-CXXFLAGS = -Wall $(DEBUG) $(PROFILE) $(OPT) $(ARCH) -m64 -I. -Wno-unused-result -Wno-strict-aliasing -Wno-unused-function
+CXXFLAGS = -Wall $(DEBUG) $(PROFILE) $(OPT) $(ARCH) -I. -Wno-unused-result -Wno-strict-aliasing -Wno-unused-function
 
 LDFLAGS = $(DEBUG) $(PROFILE) $(OPT) -lpthread -lssl -lcrypto -lm
 
