--- khmer-2.1.1/setup.cfg	2017-05-25 20:10:21.000000000 -0400
+++ khmer-2.1.1/setup.cfg	2024-08-07 22:04:20.100887810 -0400
@@ -3,8 +3,8 @@
 undef = NO_UNIQUE_RC
 # libraries = z,bz2
 ## if using system libraries
-include-dirs = lib:third-party/zlib:third-party/bzip2:third-party/seqan/core/include:third-party/smhasher
-# include-dirs = lib
+# include-dirs = lib:third-party/zlib:third-party/bzip2:third-party/seqan/core/include:third-party/smhasher
+include_dirs = lib:include:include/seqan:third-party/smhasher
 ## if using system libraries (broken)
 
 # define = NDEBUG
