diff --git a/third-party/cqf/Makefile b/third-party/cqf/Makefile
index 33769ef..30792ef 100644
--- a/third-party/cqf/Makefile
+++ b/third-party/cqf/Makefile
@@ -2,7 +2,7 @@ TARGETS=gqf.o
 
 ifdef D
 	DEBUG=-g
-	OPT=
+	OPT=-O3
 else
 	DEBUG=
 	OPT=-Ofast
@@ -24,16 +24,16 @@ ifdef P
 endif
 
 ifneq ($(INCLUDE),)
-	INCLUDE := -I$(INCLUDE)
+	INCLUDE := -I$(PREFIX)/include
 endif
 
-CXX = g++ -std=c++11 -fPIC
-CC = g++ -std=c++11 -fPIC
-LD= g++ -std=c++11
+CXX = $(CXX) -std=c++14 -O3 -fPIC
+CC = $(CXX) -std=c++14 -O3 -fPIC
+LD = $(CXX) -std=c++14
 
-CXXFLAGS = -Wall $(DEBUG) $(PROFILE) $(OPT) $(ARCH) -m64 -I. -Wno-unused-result -Wno-strict-aliasing -Wno-unused-function
+CXXFLAGS = -Wall $(DEBUG) $(PROFILE) $(OPT) $(ARCH) -I. -Wno-unused-result -Wno-strict-aliasing -Wno-unused-function -Wno-format -Wno-deprecated -Wno-unused-command-line-argument
 
-LDFLAGS = $(DEBUG) $(PROFILE) $(OPT) -lpthread -lssl -lcrypto -lm
+LDFLAGS = $(DEBUG) $(PROFILE) $(OPT) -lpthread -lssl -lcrypto -lm -L$(PREFIX)/lib
 
 #
 # declaration of dependencies
