--- setup.py.orig	2025-01-19 00:50:27
+++ setup.py	2025-01-19 00:50:31
@@ -130,10 +130,10 @@
 
 def distutils_dir_name(dname):
     """Returns the name of a distutils build directory"""
-    f = "{dirname}.{platform}-{version[0]}.{version[1]}"
+    f = "{dirname}.{platform}-{cache_tag}"
     return f.format(dirname=dname,
                     platform=sysconfig.get_platform(),
-                    version=sys.version_info)
+                    cache_tag=sys.implementation.cache_tag)
 
 
 def build_dir():
