diff --git a/setup.py b/setup.py
index b6bd3cb..de816be 100755
--- a/setup.py
+++ b/setup.py
@@ -172,7 +172,7 @@ EXTRA_COMPILE_ARGS = ['-O3', '-std=c++14', '-Wno-deprecated-declarations', '-Wno
 EXTRA_LINK_ARGS = []
 
 if sys.platform == 'darwin':
-    EXTRA_COMPILE_ARGS.extend(['-arch', 'x86_64', '-mmacosx-version-min=10.13'])
+    EXTRA_COMPILE_ARGS.extend(['-arch', 'arm64', '-mmacosx-version-min=10.13'])
     EXTRA_LINK_ARGS.append('-mmacosx-version-min=10.13')
 
 if check_for_openmp():
