--- khmer-2.1.1/MANIFEST.in
+++ khmer-2.1.1/MANIFEST.in
@@ -3,12 +3,12 @@
 include LICENSE TODO .ycm_extra_conf.py authors.csv
 recursive-include lib *.hh *.cc [Mm]akefile* get_version.py
 recursive-include khmer *.hh
-recursive-include third-party *.cc *.1 *.xsl README* sample* words* *.sh *.c
-recursive-include third-party manual* [Mm]akefile* *.pl *.dsp CHANGES *.txt *.h
-recursive-include third-party ChangeLog FAQ INDEX configure *.xsl
-recursive-include third-party/bzip2 bzip.css entities.xml libbz2.def LICENSE
-include third-party/zlib/zconf.h.in third-party/zlib/zlib.pc.in
-exclude third-party/zlib/Makefile third-party/zlib/zconf.h
+#recursive-include third-party *.cc *.1 *.xsl README* sample* words* *.sh *.c
+#recursive-include third-party manual* [Mm]akefile* *.pl *.dsp CHANGES *.txt *.h
+#recursive-include third-party ChangeLog FAQ INDEX configure *.xsl
+#recursive-include third-party/bzip2 bzip.css entities.xml libbz2.def LICENSE
+#include third-party/zlib/zconf.h.in third-party/zlib/zlib.pc.in
+#exclude third-party/zlib/Makefile third-party/zlib/zconf.h
 recursive-include scripts README.txt
 graft tests
 include scripts/oxli
