diff --git a/setup.py b/setup.py
index 48a787c..e333005 100755
--- a/setup.py
+++ b/setup.py
@@ -171,9 +171,8 @@ SOURCES.extend(path_join("third-party", "smhasher", bn + ".cc") for bn in [
 EXTRA_COMPILE_ARGS = ['-O3', '-std=c++14', '-Wno-deprecated-declarations', '-Wno-unused-local-typedef', '-Wno-reorder-ctor', '-Wno-unused-but-set-variable', '-Wno-unused', '-Wno-delete-non-abstract-non-virtual-dtor', '-Wno-delete-abstract-non-virtual-dtor', '-Wno-macro-redefined']
 EXTRA_LINK_ARGS = []
 
-if sys.platform == 'darwin':
-    EXTRA_COMPILE_ARGS.extend(['-arch', 'x86_64', '-mmacosx-version-min=10.13'])
-    EXTRA_LINK_ARGS.append('-mmacosx-version-min=10.13')
+if sys.platform == 'linux':
+    EXTRA_COMPILE_ARGS.extend(['-arch', 'aarch64'])
 
 if check_for_openmp():
     EXTRA_COMPILE_ARGS.extend(['-fopenmp'])
