diff --git a/setup.py b/setup.py
index 1ec804f..5199bd7 100755
--- a/setup.py
+++ b/setup.py
@@ -172,8 +172,8 @@ EXTRA_COMPILE_ARGS = ['-O3', '-std=c++11']
 EXTRA_LINK_ARGS = []
 
 if sys.platform == 'darwin':
-    EXTRA_COMPILE_ARGS.extend(['-arch', 'x86_64', '-mmacosx-version-min=10.9'])
-    EXTRA_LINK_ARGS.append('-mmacosx-version-min=10.9')
+    EXTRA_COMPILE_ARGS.extend(['-arch', 'arm64', '-mmacosx-version-min=10.15'])
+    EXTRA_LINK_ARGS.append('-mmacosx-version-min=10.15')
 
 if check_for_openmp():
     EXTRA_COMPILE_ARGS.extend(['-fopenmp'])
