--- Makefile
+++ Makefile
@@ -29,7 +29,7 @@
 #                                    long data type and on Windows.
 #      -D__ZLIB_AVAILABLE__          Enables support for GZIP'ed input files
 # 
-CFLAGS=-O2 -I./libsrc -I./merlin -I./pdf -I./clusters -D_FILE_OFFSET_BITS=64 -D__ZLIB_AVAILABLE__ -Wall
+CFLAGS+=-O2 -I./libsrc -I./merlin -I./pdf -I./clusters -D_FILE_OFFSET_BITS=64 -D__ZLIB_AVAILABLE__ -Wall
 
 # executable file names and locations
 BINDIR = executables
