--- Makefile
+++ Makefile
@@ -179,7 +179,7 @@
 	 $(CXX) $(CFLAGS) -o $@ $(OFFXOBJ) $(CLUSTERXOBJ) $(PDFLIB) $(LIBFILE) -lm -lz
 
 $(PEDSTATS) : pedstats-$(PSVERSION).tar.gz
-	gunzip -c pedstats-$(PSVERSION).tar.gz | tar -xf - 
+	gunzip -c pedstats-$(PSVERSION).tar.gz | tar -xof - 
 	cd pedstats-$(PSVERSION) ; $(MAKE) executables/pedstats
 	cp pedstats-$(PSVERSION)/executables/pedstats executables
 	rm -rf pedstats-$(PSVERSION)
