diff --git a/SibeliaZ-LCB/common/streamfastaparser.h b/SibeliaZ-LCB/common/streamfastaparser.h
index 9f647df..48fdb08 100644
--- a/SibeliaZ-LCB/common/streamfastaparser.h
+++ b/SibeliaZ-LCB/common/streamfastaparser.h
@@ -5,9 +5,9 @@
 #include <fstream>
 #include <stdexcept>
 #include <algorithm>
-#include <tbb/mutex.h>
 #include <iostream>
 #include <memory>
+#include <mutex>
 
 #include "dnachar.h"
 
@@ -131,7 +131,7 @@ namespace TwoPaCo
 		size_t seqId_;
 		size_t start_;
 		size_t pieceId_;
-		tbb::mutex mutex_;
+		std::mutex mutex_;
 		size_t overlapSize_;
 		std::ifstream in_;
 		std::string overlapBuffer_;
diff --git a/SibeliaZ-LCB/distancekeeper.h b/SibeliaZ-LCB/distancekeeper.h
index 5dba162..75ddb89 100644
--- a/SibeliaZ-LCB/distancekeeper.h
+++ b/SibeliaZ-LCB/distancekeeper.h
@@ -2,6 +2,7 @@
 #define _DISTANCE_KEEPER_H_
 
 #include "junctionstorage.h"
+#include <climits>
 
 namespace Sibelia
 {
