diff -rNu octopus-0.5.2-beta.orig/src/CMakeLists.txt octopus-0.5.2-beta/src/CMakeLists.txt
--- octopus-0.5.2-beta.orig/src/CMakeLists.txt	2018-10-29 08:36:32.000000000 -0600
+++ octopus-0.5.2-beta/src/CMakeLists.txt	2018-11-13 15:54:06.000000000 -0700
@@ -639,7 +639,7 @@
 else()
     add_executable(octopus main.cpp ${OCTOPUS_SOURCES} ${INCLUDE_SOURCES})
     target_compile_features(octopus PRIVATE cxx_thread_local)
-    target_compile_options(octopus PRIVATE -ffast-math -funroll-loops -march=native)
+    target_compile_options(octopus PRIVATE -ffast-math -funroll-loops)
     target_include_directories(octopus PUBLIC ${octopus_SOURCE_DIR}/lib ${octopus_SOURCE_DIR}/src)
     target_link_libraries(octopus tandem ranger)
     if (NOT BUILD_SHARED_LIBS)
