diff --git a/src/configure b/src/configure
index be7161b..c4a9983 100755
--- a/src/configure
+++ b/src/configure
@@ -50,19 +50,19 @@ case `(uname -sr || echo unknown) 2>/dev/null` in
 #UNIX_System_V\ 4.2.0)  CFLAGS=${CFLAGS-"-O"} ;;
 #UNIX_SV\ 4.2MP)  CFLAGS=${CFLAGS-"-O"} ;;
 #CYGWIN*) CXX="g++"; CFLAGS=${CFLAGS-"-O3"}  ;;
-Linux*) CFLAGS=${CFLAGS-"-O3 -march=native"}; 
-	CLIB=${CLIB-"clib.a"}; 
-	ILIB=${ILIB-"-lpthread -lm -lz"} ;;
-Darwin*) CFLAGS=${CFLAGS-"-O3 -std=c++11"}; 
-	CLIB=${CLIB-"clib.a libz.a"} ;;
-*)	CFLAGS=${CFLAGS-"-O3"} ;;
+Linux*) CFLAGS=${CFLAGS-"-O3 -march=native"};
+	CLIB=${CLIB-"clib.a"};
+	ILIB=${ILIB-"-pthread -lm -lz"} ;;
+Darwin*) CFLAGS=${CFLAGS-"-O3 -std=c++14"};
+	CLIB=${CLIB-"clib.a"} ;;
+*)	ILIB=${ILIB-"-pthread -lm -lz"} ;;
 esac
 
 if ($CXX -c $CFLAGS $test.cc) 2>/dev/null; then
   echo "OK, CXX=$CXX CFLAGS=$CFLAGS"
 else
-  CXX="g++"
-  CFLASS="-O3"
+  CXX="${CXX}"
+  CFLAGS="-O3 -I${PREFIX}/include"
   echo "use CXX=$CXX CFLAGS=$CFLAGS"
 fi
 
@@ -101,7 +101,7 @@ sed < iolib.src "
 /#table_dir#/s%#table_dir#%$table_dir%
 " > iolib.h
 
-# udpate dbs.h 
+# udpate dbs.h
 sed < dbs.src "
 /#alndbs_dir#/s%#alndbs_dir#%$alndbs_dir%
 " > dbs.h
