diff --git a/src/configure b/src/configure
index b4f79b9..d6678e2 100755
--- a/src/configure
+++ b/src/configure
@@ -8,7 +8,7 @@
 #	Copyright(c) Osamu Gotoh <<gotoh.osamu.67a@st.kyoto-u.ac.jp>>
 
 VER=`sed -n -e '/VERSION "/s/.*"\(.*\)".*/\1/p' < iolib.src`
-AR=${AR-"ar rc"}
+AR=${AR-"ar rcs"}
 RANLIB=${RANLIB-"ranlib"}
 exec_prefix=${exec_prefix-${HOME}/bin}
 table_dir=${table_dir-${HOME}/table}
@@ -50,18 +50,19 @@ case `(uname -sr || echo unknown) 2>/dev/null` in
 #UNIX_System_V\ 4.2.0)  CFLAGS=${CFLAGS-"-O"} ;;
 #UNIX_SV\ 4.2MP)  CFLAGS=${CFLAGS-"-O"} ;;
 #CYGWIN*) CXX="g++"; CFLAGS=${CFLAGS-"-O3"}  ;;
-Linux*) CFLAGS=${CFLAGS-"-O3 -march=native"}; 
-	CLIB=${CLIB-"clib.a"} ;; 
-Darwin*) CFLAGS=${CFLAGS-"-O3 -std=c++11"}; 
-	CLIB=${CLIB-"clib.a libz.a"} ;;
-*)	CFLAGS=${CFLAGS-"-O3"} ;;
+Linux*) CFLAGS=${CFLAGS-"-O3 -march=x86-64-v3"};
+	CLIB=${CLIB-"clib.a"} ;
+	ILIB=${ILIB-"-pthread -lm -lz"} ;;
+Darwin*) CFLAGS=${CFLAGS-"-O3 -std=c++14 -march=x86-64-v3"}; 
+	CLIB=${CLIB-"clib.a"} ;;
+*)	ILIB=${ILIB-"-pthread -lm -lz"} ;;
 esac
 
 if ($CXX -c $CFLAGS $test.cc) 2>/dev/null; then
   echo "OK, CXX=$CXX CFLAGS=$CFLAGS"
 else
-  CXX="g++"
-  CFLASS="-O3"
+  CXX="${CXX}"
+  CFLASS="-O3 -I${PREFIX}/include"
   echo "use CXX=$CXX CFLAGS=$CFLAGS"
 fi
 
