diff --git a/src/simd_functions.h b/src/simd_functions.h
index 63adcbb..3888949 100644
--- a/src/simd_functions.h
+++ b/src/simd_functions.h
@@ -29,7 +29,7 @@
 #if defined(__SSE4_1__)
 #include <x86intrin.h>
 #elif defined(__ARM_NEON)
-#include <arm_neon.h>
+#include "sse2neon.h"
 #endif
 
 #define SIMD_INLINE inline __attribute__((always_inline))
@@ -79,7 +79,7 @@ struct Simd_functions {
 
 // class specialization
 
-#if __SSE4_1__
+#if __SSE4_1__ || __ARM_NEON
 
 template <typename regist_v>
 struct Simd_functions<char, 16, regist_v> {
@@ -387,7 +387,7 @@ struct Simd_functions<float, 4, regist_v> {
 };
 #endif	// __SSE4_1__
 
-#if __ARM_NEON
+#if __ARM_NEON_NO
 
 template <typename regist_v>
 struct Simd_functions<char, 16, regist_v> {
