--- a/src/Makefile.in
+++ b/src/Makefile.in
@@ -1,14 +1,14 @@
 #	Makefile.in for splan/sortgrcd
 #	Copyright(c) Osamu Gotoh <<o.gotoh@asit.go.jp>>
-exec_prefix = /usr/local/bin
-table_dir = /usr/local/bin
-alndbs_dir = /db/seqdb
-CXX  = g++
-CFLAGS = -O3
+exec_prefix = $(PREFIX)/bin
+table_dir = $(PREFIX)/share/spaln/table
+alndbs_dir = $(PREFIX)/share/spaln/alndbs
+CXX  = $(CXX)
+CFLAGS = -O3 -I$(PREFIX)/include
 RANLIB = ranlib
 #CXX = clang++
 #CFLAGS = -Wall
-LD  = g++
+LD  = $(CXX)
 AR = ar ru
 DEL = rm -f
 DEFLT = -DM_THREAD=1
