From b7cbb8885be74844972f4807f3af9ac7628ce764 Mon Sep 17 00:00:00 2001
From: Matus Kosut <matus.kosut@ntnu.no>
Date: Mon, 26 Oct 2020 11:01:35 +0100
Subject: [PATCH] Patch CMakeLists.txt for conda build

---
 CMakeLists.txt | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 75b8238..54436df 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -17,10 +17,13 @@ if(CGET_PREFIX)
     link_directories(${CGET_PREFIX}/lib)
 endif()
 
+include_directories("include" "${SRC_DIR}/include" "${PREFIX}/include" "${PREFIX}/include/shrinkwrap" "${PREFIX}/include/htslib" "${SYS_PREFIX}/include")
+link_directories("lib" "${SRC_DIR}/lib" "${PREFIX}/lib" "${SYS_PREFIX}/lib")
+
 #list(APPEND CMAKE_PREFIX_PATH ${CMAKE_INSTALL_PREFIX})
 #find_package(shrinkwrap CONFIG REQUIRED)
 #if (NOT BUILD_SHARED_LIBS)
-    set(CMAKE_FIND_LIBRARY_SUFFIXES ".a;${CMAKE_FIND_LIBRARY_SUFFIXES}") # Prefer libz.a when both are available
+#    set(CMAKE_FIND_LIBRARY_SUFFIXES ".a;${CMAKE_FIND_LIBRARY_SUFFIXES}") # Prefer libz.a when both are available
 #endif()
 find_library(HTS_LIBRARY hts)
 find_library(ZLIB_LIBRARY z)
-- 
2.23.0

