--- makefile
+++ makefile
@@ -1,4 +1,4 @@
-CC=	g++
+CXX=	g++
 
 BIN = $(DESTDIR)/usr/bin
 FLAGS= -DMAXHITS=1000 -DTHREAD -funroll-loops -Lsamtools -Isamtools -Lgzstream -Igzstream -O3 -m64
@@ -13,11 +13,11 @@
 
 all: bsmap
 %.o:%.cpp
-	$(CC) $(FLAGS) -c $< -o $@
+	$(CXX) $(FLAGS) -c $< -o $@
 bsmap: $(OBJS1)
 	(cd samtools; make)
 	(cd gzstream; make)
-	$(CC) $(FLAGS) $^ -o $@ $(THREAD) -lbam -lz -lgzstream
+	$(CXX) $(FLAGS) $^ -o $@ $(THREAD) -lbam -lz -lgzstream
 	rm -f *.o
 
 clean:
