--- sam2bam.sh	2015-06-08 16:42:55.000000000 +0000
+++ sam2bam.sh.new	2018-05-07 22:04:41.318630981 +0000
@@ -5,12 +5,17 @@
 tmpbam=${1%.*}.tmp.bam
 outbam=${1%.*}.bam
 
+if [ "$1" == "-h" ]; then
+        echo "Usage: sam2bam.sh <infile>"
+        exit 0
+fi
 echo "Converting SAM to BAM ..."
+if []; then
 if [ ! -f $1 ]; then
 	echo "$1 does not exist."
 	exit 1
 fi
-samtools/samtools view -bS $1 > $tmpbam
+samtools view -bS $1 > $tmpbam
 if [ $? -ne 0 ]; then
 	echo "SAM2BAM conversion not sucessful."
 	echo "$1 remains unchanged."
@@ -18,7 +23,7 @@
 	exit 1
 fi
 echo "Sorting BAM ..."
-samtools/samtools sort $tmpbam ${outbam%.*}
+samtools sort $tmpbam ${outbam%.*}
 if [ $? -ne 0 ]; then
 	echo "BAM file sorting not sucessful."
 	echo "$outbam is in unsorted BAM format".
@@ -27,7 +32,7 @@
 fi
 rm $tmpbam
 echo "Indexing BAM ..."
-samtools/samtools index $outbam
+samtools index $outbam
 if [ $? -ne 0 ]; then
 	echo "BAM file indexing not sucessful."
 	exit 1
--- bsp2sam.py	2015-06-08 16:42:59.000000000 +0000
+++ bsp2sam.py.new	2018-05-07 22:07:57.424826000 +0000
@@ -1,4 +1,4 @@
-#! /usr/bin/env python
+#!/opt/anaconda1anaconda2anaconda3/bin/python
 import sys, optparse, time
 
 usage = "usage: %prog [options] BSMAP_MAPPING_FILE"
--- methdiff.py	2015-06-08 16:42:59.000000000 +0000
+++ methdiff.py.new	2018-05-07 22:08:10.822839323 +0000
@@ -1,4 +1,4 @@
-#! /usr/bin/env python
+#!/opt/anaconda1anaconda2anaconda3/bin/python
 import sys, time, os, array, optparse
 usage = "usage: %prog [options] <GROUP1_SAMPLE1,GROUP1_SAMPLE2,...> <GROUP2_SAMPLE1,GROUP2_SAMPLE2,...>"
 parser = optparse.OptionParser(usage=usage)
--- methratio.py	2015-06-08 16:42:59.000000000 +0000
+++ methratio.py.new	2018-05-07 22:08:21.550849990 +0000
@@ -1,4 +1,4 @@
-#! /usr/bin/env python
+#!/opt/anaconda1anaconda2anaconda3/bin/python
 import sys, time, os, array, optparse
 usage = "usage: %prog [options] BSMAP_MAPPING_FILES"
 parser = optparse.OptionParser(usage=usage)
