diff --git a/CMakeLists.txt b/CMakeLists.txt
index bc68486..691fabc 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -191,10 +191,10 @@ add_subdirectory(cobs)
 add_subdirectory(src)
 
 # descend into tests
-add_subdirectory(tests)
+# add_subdirectory(tests)
 
 # descend into python
-add_subdirectory(python)
+# add_subdirectory(python)
 
 ################################################################################
 
