diff --git a/CMakeLists.txt b/CMakeLists.txt
index e370bae..af8d450 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -169,13 +169,12 @@ else()
 endif()
 
 ### use Boost filesystem ###
-set(Boost_USE_STATIC_LIBS   ON)
 find_package(Boost 1.42.0 COMPONENTS system filesystem)
 if(${Boost_FOUND})
   set(COBS_INCLUDE_DIRS ${Boost_INCLUDE_DIRS} ${COBS_INCLUDE_DIRS})
   set(COBS_LINK_LIBRARIES ${Boost_LIBRARIES} ${COBS_LINK_LIBRARIES})
 endif()
-set(COBS_LINK_LIBRARIES stdc++fs ${COBS_LINK_LIBRARIES})
+set(Boost_USE_STATIC_LIBS   ON)
 
 ### use TLX ###
 add_subdirectory(extlib/tlx)
