diff --git a/cobs/util/fs.hpp b/cobs/util/fs.hpp
index feea6fd..ff14d4b 100644
--- a/cobs/util/fs.hpp
+++ b/cobs/util/fs.hpp
@@ -9,19 +9,6 @@
 #ifndef COBS_UTIL_FS_HEADER
 #define COBS_UTIL_FS_HEADER
 
-#if __cplusplus >= 201703L
-
-#include <experimental/filesystem>
-
-namespace cobs {
-
-namespace fs = std::experimental::filesystem;
-using std::error_code;
-
-} // namespace cobs
-
-#else
-
 #include <boost/filesystem.hpp>
 
 namespace cobs {
@@ -31,7 +18,6 @@ using boost::system::error_code;
 
 } // namespace cobs
 
-#endif
 
 #endif // !COBS_UTIL_FS_HEADER
 
