diff -ruN /tmp/ncbi-blast-2.2.31+-src/c++/src/build-system/configure ncbi-blast-2.2.31+-src/c++/src/build-system/configure
--- /tmp/ncbi-blast-2.2.31+-src/c++/src/build-system/configure	2019-12-02 14:17:51.654826167 +0100
+++ ncbi-blast-2.2.31+-src/c++/src/build-system/configure	2019-12-02 14:18:14.486778694 +0100
@@ -5819,14 +5819,10 @@
 
 if test "$GCC" = "yes" ; then
    compiler_ver="`$real_CXX -dumpversion 2>&1`"
-   case "$compiler_ver" in
-     2.95* | 2.96* | 3.* | 4.* )
-       compiler="GCC"
-       ncbi_compiler="GCC"
-       ncbi_compiler_ver="$compiler_ver"
-                 WithFeatures="$WithFeatures${WithFeaturesSep}GCC"; WithFeaturesSep=" "
-       ;;
-   esac
+   compiler="GCC"
+   ncbi_compiler="GCC"
+   ncbi_compiler_ver="$compiler_ver"
+             WithFeatures="$WithFeatures${WithFeaturesSep}GCC"; WithFeaturesSep=" "
 elif test "$KCC" = "yes" ; then
    compiler_ver="$kcc_ver"
    compiler="KCC"
Binary files /tmp/ncbi-blast-2.2.31+-src/c++/src/build-system/.configure.swp and ncbi-blast-2.2.31+-src/c++/src/build-system/.configure.swp differ
