--- Makefile.in.bak	2017-03-09 11:49:34.000000000 -0500
+++ Makefile.in	2017-03-09 12:06:52.000000000 -0500
@@ -112,7 +112,7 @@
 	mkdir -p ${BINDIR}
 	-mkdir -p ${MANDIR}/man${MANSUFFIX}
 	for file in $(PROGS) $(PVMPROGS); do\
-	   cp src/$$file $(BINDIR)/$$file;\
+	   cp src/$$file $(BINDIR)/$${file//2};\
 	done
 	-for file in hmmer $(PROGS); do\
 	   $(INSTMAN) documentation/man/$$file.man $(MANDIR)/man$(MANSUFFIX)/$$file.$(MANSUFFIX);\
