--- src/Deletions.cpp	2019-06-03 09:28:37.570379155 +0200
+++ src/Deletions.cpp	2019-06-03 09:12:57.695330587 +0200
@@ -211,7 +211,7 @@
 	uint medianBpPos1 =  medianBp1.first;
 	uint medianBpPos2 =  medianBp2.first;
 
-	uint svSize = abs(medianBpPos2-(medianBpPos1+1))+1;
+	uint svSize = (uint) abs((long) medianBpPos2-((long) medianBpPos1+1))+1;
 
 	std::stringstream output, bpSeq;
 
--- src/Inversions.cpp	2019-06-03 09:28:37.570379155 +0200
+++ src/Inversions.cpp	2019-06-03 09:30:17.074636998 +0200
@@ -255,7 +255,7 @@
 		bpSeq2 << "";
 	}
 
-	uint svSize = abs(medianBpPos2-(medianBpPos1+1))+1;
+	uint svSize = (uint) abs((long) medianBpPos2-((long) medianBpPos1+1))+1;
 
 	// Outpu format:
 	// Chr	Start	End	Support(PE)	Support(SC)	Upstream-breakpoint-sequence	Downstream-breakpoint-sequence
--- src/Tandems.cpp	2019-06-03 09:28:37.570379155 +0200
+++ src/Tandems.cpp	2019-06-03 09:30:59.878746224 +0200
@@ -174,7 +174,7 @@
 	uint medianBpPos1 =  medianBp1.first;
 	uint medianBpPos2 =  medianBp2.first;
 
-	uint svSize = abs(medianBpPos2-(medianBpPos1+1))+1;
+	uint svSize = (uint) abs((long) medianBpPos2-((long) medianBpPos1+1))+1;
 
 	std::stringstream output, bpSeq;
 
