--- Makefile
+++ Makefile
@@ -27,4 +27,4 @@
-	g++ -O9 -w  -DNO_OPENMP src/cdhit-common.c++ -c
-	g++ -O9 -w  -DNO_OPENMP src/cdhit-utility.c++ -c
-	g++ -O9 -w  -DNO_OPENMP src/cdhit-est.c++ -c
-	g++ -O9 -w  -DNO_OPENMP cdhit-est.o cdhit-common.o cdhit-utility.o -o bin/cd-hit-est
+	$(CXX) -O9 -w  -DNO_OPENMP src/cdhit-common.c++ -c
+	$(CXX) -O9 -w  -DNO_OPENMP src/cdhit-utility.c++ -c
+	$(CXX) -O9 -w  -DNO_OPENMP src/cdhit-est.c++ -c
+	$(CXX) -O9 -w  -DNO_OPENMP cdhit-est.o cdhit-common.o cdhit-utility.o -o bin/cd-hit-est
