--- setup.py.orig	2022-09-09 17:31:39.180256060 +0100
+++ setup.py	2022-09-09 17:49:24.335610213 +0100
@@ -21,11 +21,11 @@
         sources=sourcefiles
         )])
 
-ext_modules = [
+ext_modules = cythonize([
     Extension(
         'BioExt.align._align',
         sources=[
-            os.path.join('BioExt', 'align', '_align.c'),
+            os.path.join('BioExt', 'align', '_align.pyx'),
             os.path.join('BioExt', 'align', 'alignment.c')
             ],
         include_dirs=np_inc,
@@ -35,7 +35,7 @@
     Extension(
         'BioExt.merge._merge',
         sources=[
-            os.path.join('BioExt', 'merge', '_merge.c'),
+            os.path.join('BioExt', 'merge', '_merge.pyx'),
             os.path.join('BioExt', 'merge', 'merge.cpp')
             ],
         extra_compile_args=['-O3', '-I.']
@@ -43,13 +43,13 @@
     Extension(
         'BioExt.rateclass._rateclass',
         sources=[
-            os.path.join('BioExt', 'rateclass', '_rateclass.cpp'),
+            os.path.join('BioExt', 'rateclass', '_rateclass.pyx'),
             os.path.join('BioExt', 'rateclass', 'rateclass.cpp')
             ],
+        language="c++",
         extra_compile_args=['-O3', '-I.']
         )
-    ] + tn93_extension
-
+    ]) + tn93_extension
 
 setup(
     name='bioext',
