--- a/setup.py
+++ b/setup.py
@@ -24,11 +24,11 @@ tn93_extension = cythonize([
         sources=sourcefiles
         )])
 
-ext_modules = [
+ext_modules = cythonize([
     Extension(
         'BioExt.align._align',
         sources=[
-            os.path.join('BioExt', 'align', '_align.c'),
+            os.path.join('BioExt', 'align', '_align.pyx'),
             os.path.join('BioExt', 'align', 'alignment.c')
             ],
         include_dirs=np_inc,
@@ -38,7 +38,7 @@ ext_modules = [
     Extension(
         'BioExt.merge._merge',
         sources=[
-            os.path.join('BioExt', 'merge', '_merge.c'),
+            os.path.join('BioExt', 'merge', '_merge.pyx'),
             os.path.join('BioExt', 'merge', 'merge.cpp')
             ],
         extra_compile_args=['-O3', '-I.']
@@ -46,12 +46,13 @@ ext_modules = [
     Extension(
         'BioExt.rateclass._rateclass',
         sources=[
-            os.path.join('BioExt', 'rateclass', '_rateclass.cpp'),
+            os.path.join('BioExt', 'rateclass', '_rateclass.pyx'),
             os.path.join('BioExt', 'rateclass', 'rateclass.cpp')
             ],
+	language="c++",
         extra_compile_args=['-O3', '-I.']
         )
-    ] + tn93_extension
+    ]) + tn93_extension
 
 _refdir = join(
     split(
