--- BioExt/stats/__init__.py
+++ BioExt/stats/__init__.py
@@ -3,7 +3,10 @@ from __future__ import division, print_function

 import numpy as np

-from scipy.misc import comb, factorial
+try:  # SciPy >= 0.19
+      from scipy.special import comb, factorial
+except ImportError:
+      from scipy.misc import comb, factorial # noqa
 from scipy.optimize import fmin_cobyla
 from scipy.stats import rv_continuous, rv_discrete
 from scipy.special import beta, gamma
