--- ViReMa.py	2014-06-25 12:43:38.000000000 +0000
+++ ViReMa.py	2017-09-14 16:09:13.112629345 +0000
@@ -1,3 +1,4 @@
+#!/usr/bin/env python
 ##      Copyright (c) 2013-2014 Andrew Laurence Routh
 ##      
 ##      Permission is hereby granted, free of charge, to any person obtaining a copy
--- Compiler_Module.py	2014-06-25 13:36:30.000000000 +0000
+++ Compiler_Module.py	2017-09-14 16:09:27.678753883 +0000
@@ -1,3 +1,4 @@
+#!/usr/bin/env python
 ##      Copyright (c) 2013-2014 Andrew Laurence Routh
 ##      
 ##      Permission is hereby granted, free of charge, to any person obtaining a copy
