--- src/cnvnator2VCF.pl.org	2021-12-07 18:22:03.793708582 -0600
+++ src/cnvnator2VCF.pl	2021-12-07 18:22:20.630673226 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 use warnings;
 use strict;
--- src/plotbaf.py.org	2021-12-07 18:22:39.041634568 -0600
+++ src/plotbaf.py	2021-12-07 18:22:57.361596099 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/opt/anaconda1anaconda2anaconda3/bin/python
 
 import matplotlib.pyplot as plt
 import numpy
--- src/plotcircular.py.org	2021-12-07 18:23:02.683584926 -0600
+++ src/plotcircular.py	2021-12-07 18:23:16.121556708 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/opt/anaconda1anaconda2anaconda3/bin/python
 
 import numpy as np
 import argparse
--- src/plotrdbaf.py.org	2021-12-07 18:23:25.113537829 -0600
+++ src/plotrdbaf.py	2021-12-07 18:23:39.033508600 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/opt/anaconda1anaconda2anaconda3/bin/python
 
 import matplotlib.pyplot as plt
 import numpy
