From ad5cec123fdb1c9b0eaf45c06e8a16d622e82032 Mon Sep 17 00:00:00 2001
From: Elmar Pruesse <elmar.pruesse@ucdenver.edu>
Date: Tue, 2 Apr 2019 15:47:38 -0600
Subject: [PATCH] Fix macos build

---
 Makefile | 22 ++++++++++++++--------
 1 file changed, 14 insertions(+), 8 deletions(-)

diff --git a/Makefile b/Makefile
index 3013b6b..572ecd0 100644
--- a/Makefile
+++ b/Makefile
@@ -41,14 +41,17 @@ BINDIR ?= $(PREFIX)/bin
 INCDIR ?= $(PREFIX)/include
 LIBDIR ?= $(PREFIX)/lib
 
-SOVERSION          := 0
+SOVERSION          := .0
 STATIC_LIB_SUFFIX  := .a
-SHARED_LIB_SUFFIX  := .so.$(SOVERSION)
+SHARED_LIB_SUFFIX  := .so
+SHARED_LIB          = libdeflate$(SHARED_LIB_SUFFIX)$(SOVERSION)
 SHARED_LIB_CFLAGS  := -fPIC
-SHARED_LIB_LDFLAGS := -Wl,-soname=libdeflate$(SHARED_LIB_SUFFIX)
+SHARED_LIB_LDFLAGS := -Wl,-soname=$(SHARED_LIB)
 PROG_SUFFIX        :=
 PROG_CFLAGS        :=
 HARD_LINKS         := 1
+ARFLAGS            := cr
+ARCHS              := *
 
 # Compiling for Windows with MinGW?
 ifneq ($(findstring -mingw,$(shell $(CC) -dumpmachine 2>/dev/null)),)
@@ -70,6 +73,10 @@ ifneq ($(findstring -mingw,$(shell $(CC) -dumpmachine 2>/dev/null)),)
         AR := $(shell echo $(CC) | \
                 sed -E 's/g?cc(-?[0-9]+(\.[0-9]+)*)?(\.exe)?$$/ar\3/')
     endif
+else ifeq ($(shell uname),Darwin)
+   SHARED_LIB_SUFFIX  := .dylib
+   SHARED_LIB          = libdeflate$(SOVERSION)$(SHARED_LIB_SUFFIX)
+   SHARED_LIB_LDFLAGS := "-install_name $(SHARED_LIB)"
 endif
 
 ##############################################################################
@@ -95,14 +102,13 @@ DEFAULT_TARGETS :=
 #### Library
 
 STATIC_LIB := libdeflate$(STATIC_LIB_SUFFIX)
-SHARED_LIB := libdeflate$(SHARED_LIB_SUFFIX)
 
 LIB_CFLAGS += $(CFLAGS) -fvisibility=hidden -D_ANSI_SOURCE
 
 LIB_HEADERS := $(wildcard lib/*.h) $(wildcard lib/*/*.h)
 
 LIB_SRC := lib/aligned_malloc.c lib/deflate_decompress.c \
-	   $(wildcard lib/*/cpu_features.c)
+	   $(wildcard lib/$(ARCHS)/cpu_features.c)
 
 DECOMPRESSION_ONLY :=
 ifndef DECOMPRESSION_ONLY
@@ -139,7 +145,7 @@ $(SHARED_LIB_OBJ): %.shlib.o: %.c $(LIB_HEADERS) $(COMMON_HEADERS) .lib-cflags
 
 # Create static library
 $(STATIC_LIB):$(STATIC_LIB_OBJ)
-	$(QUIET_AR) $(AR) cr $@ $+
+	$(QUIET_AR) $(AR) $(ARFLAGS) $@ $+
 
 DEFAULT_TARGETS += $(STATIC_LIB)
 
@@ -152,9 +158,9 @@ DEFAULT_TARGETS += $(SHARED_LIB)
 
 ifdef SOVERSION
 # Create the symlink libdeflate.so => libdeflate.so.$SOVERSION
-libdeflate.so:$(SHARED_LIB)
+libdeflate$(SHARED_LIB_SUFFIX):$(SHARED_LIB)
 	$(QUIET_LN) ln -sf $+ $@
-DEFAULT_TARGETS += libdeflate.so
+DEFAULT_TARGETS += libdeflate$(SHARED_LIB_SUFFIX)
 endif
 
 # Rebuild if CC, LIB_CFLAGS, or CPPFLAGS changed
-- 
2.14.2

