diff --git a/phase/src/models/imputation_hmm.h b/phase/src/models/imputation_hmm.h
index 1bda64c..9e7b157 100644
--- a/phase/src/models/imputation_hmm.h
+++ b/phase/src/models/imputation_hmm.h
@@ -28,7 +28,11 @@
 
 #include <utils/otools.h>
 #include <containers/conditioning_set.h>
+#ifdef __ARM_NEON
+#include "sse2neon.h"
+#else
 #include <immintrin.h>
+#endif
 #include <boost/align/aligned_allocator.hpp>
 
 template <typename T>
diff --git a/phase/src/models/phasing_hmm.h b/phase/src/models/phasing_hmm.h
index 525cb17..ae7c43b 100644
--- a/phase/src/models/phasing_hmm.h
+++ b/phase/src/models/phasing_hmm.h
@@ -28,7 +28,11 @@
 
 #include <utils/otools.h>
 #include <containers/conditioning_set.h>
+#ifdef __ARM_NEON
+#include "sse2neon.h"
+#else
 #include <immintrin.h>
+#endif
 #include <boost/align/aligned_allocator.hpp>
 
 template <typename T>
