diff --git a/phase/src/models/imputation_hmm.h b/phase/src/models/imputation_hmm.h
index 1bda64c..fd05520 100644
--- a/phase/src/models/imputation_hmm.h
+++ b/phase/src/models/imputation_hmm.h
@@ -28,9 +28,14 @@
 
 #include <utils/otools.h>
 #include <containers/conditioning_set.h>
-#include <immintrin.h>
 #include <boost/align/aligned_allocator.hpp>
 
+#if defined(__x86_64__) || defined(__i386__)
+#include <immintrin.h>
+#elif defined(__aarch64__) || defined(__arm64__)
+#include "sse2neon.h"
+#endif
+
 template <typename T>
 using aligned_vector32 = std::vector<T, boost::alignment::aligned_allocator < T, 32 > >;
 
diff --git a/phase/src/models/phasing_hmm.h b/phase/src/models/phasing_hmm.h
index 525cb17..88fdc91 100644
--- a/phase/src/models/phasing_hmm.h
+++ b/phase/src/models/phasing_hmm.h
@@ -28,9 +28,14 @@
 
 #include <utils/otools.h>
 #include <containers/conditioning_set.h>
-#include <immintrin.h>
 #include <boost/align/aligned_allocator.hpp>
 
+#if defined(__x86_64__) || defined(__i386__)
+#include <immintrin.h>
+#elif defined(__aarch64__) || defined(__arm64__)
+#include "sse2neon.h"
+#endif
+
 template <typename T>
 using aligned_vector32 = std::vector<T, boost::alignment::aligned_allocator < T, 32 > >;
 
