--- setup.py	2017-10-23 15:40:26.000000000 -0500
+++ setup.py.new	2019-11-07 12:18:28.757117252 -0600
@@ -35,8 +35,6 @@
       include_package_data=True,
       keywords=["VCF", "variant", "caller", "format", "snps", "genetic", "variation", "genetics"],
       data_files=gen_data_files("static"),
-      setup_requires=['pytest-runner'],
-      tests_require=['pytest', 'coveralls'],
       classifiers=["Development Status :: 4 - Beta","Operating System :: MacOS",
                    "Operating System :: Unix",
                    "Operating System :: POSIX :: Linux",
