diff --git a/Cargo.toml b/Cargo.toml
index 48d819d..f662c63 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -4,7 +4,7 @@ version = "1.1.0"
 authors = ["Ian Fiddes <ian.fiddes@10xgenomics.com>", "Wyatt McDonnell <wyatt.mcdonnell@10xgenomics.com>"]
 
 [dependencies]
-rust-htslib = "0.20"
+rust-htslib = "0.47.0"
 csv = "1"
 failure = "*"
 clap = "*"
@@ -14,7 +14,7 @@ simplelog = "0.5.0"
 log = "*"
 tempfile = "*"
 rayon = "*"
-ring = "0.14.6"
+ring = "0.17.8"
 data-encoding = "*"
 faccess="0.2.4"
 
