From 76b05cabbfda30cfcc81f07081cc1a091127b393 Mon Sep 17 00:00:00 2001
From: Ilya Shlyakhter <ilya_shl@alum.mit.edu>
Date: Thu, 30 Nov 2023 16:34:22 -0500
Subject: [PATCH 3/3] use ncurses

---
 DWGSIMSrc/samtools/Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/DWGSIMSrc/samtools/Makefile b/DWGSIMSrc/samtools/Makefile
index 37d5345..5143709 100644
--- a/DWGSIMSrc/samtools/Makefile
+++ b/DWGSIMSrc/samtools/Makefile
@@ -19,7 +19,7 @@ PROG=		samtools
 INCLUDES=	-I.
 SUBDIRS=	. bcftools misc
 LIBPATH=
-LIBCURSES=	-lcurses # -lXCurses
+LIBCURSES=	-lncurses # -lXCurses
 
 
 .SUFFIXES:.c .o
-- 
2.43.0

