From 3de5f1769e5fca938ef0479bf0f7a670ad737540 Mon Sep 17 00:00:00 2001
From: Ilya Shlyakhter <ilya_shl@alum.mit.edu>
Date: Thu, 30 Nov 2023 17:16:53 -0500
Subject: [PATCH 4/4] add LDFLAGS

---
 DWGSIMSrc/Makefile | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/DWGSIMSrc/Makefile b/DWGSIMSrc/Makefile
index bbb7047..3aea51f 100644
--- a/DWGSIMSrc/Makefile
+++ b/DWGSIMSrc/Makefile
@@ -35,10 +35,10 @@ all:$(PROG)
 .PHONY:all-recur lib-recur clean-recur cleanlocal-recur install-recur
 
 dwgsim:lib-recur $(DWGSIM_AOBJS)
-	$(CC) $(CFLAGS) -o $@ $(DWGSIM_AOBJS) -lm -lz -lpthread
+	$(CC) $(CFLAGS) -o $@ $(DWGSIM_AOBJS) $(LDFLAGS) -lm -lz -lpthread
 
 dwgsim_eval:lib-recur $(DWGSIM_EVAL_AOBJS)
-	$(CC) $(CFLAGS) -o $@ $(DWGSIM_EVAL_AOBJS) -Lsamtools -lm -lz -lpthread
+	$(CC) $(CFLAGS) -o $@ $(DWGSIM_EVAL_AOBJS) $(LDFLAGS) -Lsamtools -lm -lz -lpthread
 
 cleanlocal:
 		rm -vfr gmon.out *.o a.out *.exe *.dSYM razip bgzip $(PROG) *~ *.a *.so.* *.so *.dylib; \
-- 
2.43.0

