From ba85d6222226cc95ff0bbcc33d29806467e3d774 Mon Sep 17 00:00:00 2001
From: Ilya Shlyakhter <ilya_shl@alum.mit.edu>
Date: Thu, 30 Nov 2023 21:28:16 -0500
Subject: [PATCH 5/5] fix perl loc

---
 faFilter.pl            | 2 +-
 simulateLinkedReads.pl | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/faFilter.pl b/faFilter.pl
index d1faa2f..7c11208 100755
--- a/faFilter.pl
+++ b/faFilter.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use strict;
 use warnings;
diff --git a/simulateLinkedReads.pl b/simulateLinkedReads.pl
index 302748a..bb8a0c0 100755
--- a/simulateLinkedReads.pl
+++ b/simulateLinkedReads.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # The MIT License (MIT)
 # Copyright (c) 2016 Ruibang Luo <aquaskyline@gmail.com>
-- 
2.43.0

