--- binReads.pl	2015-09-30 13:21:52.000000000 -0500
+++ binReads.pl.new	2018-12-12 18:03:04.485187656 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ##========================================================================
 ##
@@ -192,4 +192,3 @@
 	$counter++;
     }
 }
-
--- centersToGEM.pl	2015-12-02 14:09:06.000000000 -0600
+++ centersToGEM.pl.new	2018-12-12 18:03:40.014184067 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 use strict;
 use warnings;
--- domainsMergeToBed.pl	2015-09-30 13:21:53.000000000 -0500
+++ domainsMergeToBed.pl.new	2018-12-12 18:05:32.450172703 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ##========================================================================
 ##
--- domainsToBed.pl	2015-09-30 13:21:53.000000000 -0500
+++ domainsToBed.pl.new	2018-12-12 18:10:10.723143778 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ##========================================================================
 ##
--- hiddenDomains	2017-08-08 09:08:30.000000000 -0500
+++ hiddenDomains.new	2018-12-12 18:08:18.529155487 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ##========================================================================
 ##
--- peakCenters	2015-12-03 10:33:22.000000000 -0600
+++ peakCenters.new	2018-12-12 18:08:50.550152145 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ##========================================================================
 ##
