diff --git a/Makefile b/Makefile
index 43a9c91..newhash 100644
--- a/Makefile
+++ b/Makefile
@@ -65,6 +65,8 @@ else
   endif
 endif
 
+CFLAGS += -fPIC
+
 %.o : %.cpp
 	${CC} ${CFLAGS} -c $< -o $@
 
@@ -73,14 +75,18 @@ endif
 LINKER  := g++
 LIBS := 
 OBJS := GBase.o GStr.o GArgs.o
-GCLIBS := GFastaIndex.o GFaSeqGet.o gff.o gdna.o codons.o GBase.o GStr.o GArgs.o
+GCLIBS := GFastaIndex.o GFaSeqGet.o gff.o gdna.o codons.o

 .PHONY : all
-all:    gtest threads
+all:    gtest threads gffcomparelibs
 debug:  gtest threads
 nodebug:  gtest threads
+gffcomparelibs: ${GCLIBS}
+	$(CXX) ${LDFLAGS} $(GCC45OPTS) $(GCC45OPTMAIN) -dynamiclib \
+          -stdlib=libc++ ${GCLIBS} -o libgffc.dylib \
+          -install_name @rpath/libgffc.dylib
 gtest.o : GBase.h GArgs.h GVec.hh GList.hh GBitVec.h
 GArgs.o : GArgs.h
 gtest: $(OBJS) gtest.o
