--- Makefile	2025-10-31 11:04:27.673874940 +0800
+++ Makefile	2025-10-31 11:09:42.921874940 +0800
@@ -38,13 +38,13 @@
 EXE =
 endif
 
-CC      := g++
+CC      := ${CXX}
 
 BASEFLAGS  := -Wall -Wextra ${SEARCHDIRS} $(MARCH) \
  -D_REENTRANT -fno-exceptions -fno-rtti
 
 #add the link-time optimization flag if gcc version > 4.5
-GCC_VERSION:=$(subst ., ,$(shell gcc -dumpversion))
+GCC_VERSION:=$(subst ., ,$(shell ${CC} -dumpversion))
 GCC_MAJOR:=$(word 1,$(GCC_VERSION))
 GCC_MINOR:=$(word 2,$(GCC_VERSION))
 #GCC_SUB:=$(word 3,$(GCC_VERSION))
@@ -70,17 +70,20 @@
 
 # C/C++ linker
 
-LINKER  := g++
+LINKER  := ${CXX}
 LIBS := 
 OBJS := GBase.o GStr.o GArgs.o
+GCLIBS := GFastaIndex.o GFaSeqGet.o gff.o gdna.o codons.o GBase.o GStr.o GArgs.o
 
 .PHONY : all
-all:    gtest threads
+all:    gtest gffcomparelibs
 
 version: ; @echo "GCC Version is: "$(GCC_MAJOR)":"$(GCC_MINOR)":"$(GCC_SUB)
 	@echo "> GCC Opt. string is: "$(GCC45OPTS)
 debug:  gtest threads
 nodebug:  gtest threads
+gffcomparelibs: ${GCLIBS}
+   ${LINKER} ${LDFLAGS} $(GCC45OPTS) $(GCC45OPTMAIN) -dynamiclib ${GCLIBS} -o libgffc.dylib
 gtest.o : GBase.h GArgs.h GVec.hh GList.hh GBitVec.h
 GArgs.o : GArgs.h
 gtest: $(OBJS) gtest.o
