--- groopm/cluster.py.orig	2025-09-12 12:01:22.941099500 +0800
+++ groopm/cluster.py	2025-09-12 14:09:29.921099500 +0800
@@ -92,7 +92,10 @@
 from numpy.linalg import norm as np_norm
 import scipy.ndimage as ndi
 from scipy.spatial.distance import pdist, squareform, cityblock, euclidean
-from scipy.misc import imsave
+try:
+    from scipy.misc import imsave
+except ImportError:
+    from imageio import imwrite as imsave
 
 # GroopM imports
 from profileManager import ProfileManager
