diff --git a/src/Makefile b/src/Makefile
index 1cffe8e..c3eeb77 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -1,13 +1,11 @@
 ADRM_DIR = adapterremoval
 BT2_DIR = bowtie2-2.4.4
 
-CXX = g++
-
 .PHONY: all clean
 
 all:
 	mkdir -p ../inst
-	(cd	$(ADRM_DIR) && ($(MAKE) CXX="$(CXX) -std=gnu++11" -f Makefile))
+	(cd	$(ADRM_DIR) && ($(MAKE) CXX="$(CXX)" -f Makefile))
 	(cd	$(ADRM_DIR) && ($(MAKE) clean_o -f Makefile))
 	(cd $(BT2_DIR) && ($(MAKE) CXX="$(CXX)" -f Makefile))
 	(cd	$(BT2_DIR) && ($(MAKE) move -f Makefile))
