--- Makefile
+++ Makefile
@@ -127,8 +127,8 @@
 ${LIBPCRE2} :
 	cd lib/pcre2; $(MAKE) libpcre2.a || exit 1; 
 	
-$(TARGET) : ${LIBHTS} ${LIBRMATH} ${LIBPCRE2} $(TOOLOBJ)
-	$(CXX) $(CFLAGS) -o $@ $(TOOLOBJ) $(LIBHTS) $(LIBRMATH) ${LIBPCRE2} -lz -lpthread
+$(TARGET) : ${LIBRMATH} $(TOOLOBJ)
+	$(CXX) $(CFLAGS) -o $@ $(LDFLAGS) $(TOOLOBJ) -lhts $(LIBRMATH) -lpcre2-8 -lz -lpthread
 
 $(TOOLOBJ): $(HEADERSONLY)
 
