--- hts_utils.h
+++ hts_utils.h
@@ -393,7 +393,7 @@
 /**
  * Get 0-based position
  */
-#define bcf_get_pos0(v) ((v)->pos)
+#define bcf_get_pos0(v) ((int32_t)(v)->pos)
 
 /**
  * Set 0-based position
@@ -403,17 +403,17 @@
 /**
  * Get 1-based position
  */
-#define bcf_get_pos1(v) ((v)->pos+1)
+#define bcf_get_pos1(v) ((int32_t)(v)->pos+1)
 
 /**
  * Get 1-based end position
  */
-#define bcf_get_end1(v) ((v)->pos + strlen((v)->d.allele[0]))
+#define bcf_get_end1(v) ((int32_t)(v)->pos + strlen((v)->d.allele[0]))
 
 /**
  * Get 1-based end position
  */
-#define bcf_get_end_pos1(v) ((v)->pos + strlen((v)->d.allele[0]))
+#define bcf_get_end_pos1(v) ((int32_t)(v)->pos + strlen((v)->d.allele[0]))
 
 /**
  * Set 1-based position
