--- Makefile	2009-12-03 20:17:51.000000000 +0000
+++ Makefile.new	2015-11-16 15:58:54.219641188 +0000
@@ -3,9 +3,9 @@
 THISCODEDIR := .
 GCLDIR := ./gclib
 # Directory where libz.a can be found
-ZDIR := ../zlib
+ZDIR := ${PREFIX}/lib
 # Directories to search for header files
-SEARCHDIRS := -I${ZDIR} -I${THISCODEDIR} -I${GCLDIR}
+SEARCHDIRS := -I${PREFIX}/include -I${THISCODEDIR} -I${GCLDIR}
 
 
 SYSTYPE :=     $(shell uname)
@@ -56,7 +56,7 @@
 # C/C++ linker
 
 LINKER    := g++
-LDFLAGS    = -lz
+LDFLAGS    = -L${PREFIX}/lib -lz
 #if ENABLE_COMPRESSION is not needed,
 # LDFLAGS = 
 
