diff --git a/rust/pyproject.toml b/rust/pyproject.toml
index fb71d4e..2c59b4f 100644
--- a/rust/pyproject.toml
+++ b/rust/pyproject.toml
@@ -9,4 +9,7 @@ classifiers = [
     "Programming Language :: Rust",
     "Programming Language :: Python :: Implementation :: CPython",
     "Programming Language :: Python :: Implementation :: PyPy",
-]
\ No newline at end of file
+]
+
+[tool.maturin]
+features = ["pyo3/extension-module"]
