diff --git a/pyproject.toml b/pyproject.toml
index 8f4ec2a..458b69a 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,9 +1,10 @@
 [build-system]
-requires = ["maturin>=0.13,<0.14"]
+requires = ["maturin>=1.8.0,<2.0.0"]
 build-backend = "maturin"
 
 [project]
 name = "blobtk"
+version = "0.7.1"
 requires-python = ">=3.8"
 classifiers = [
     "Programming Language :: Rust",
@@ -13,4 +14,4 @@ classifiers = [
 
 [tool.maturin]
 module-name = "blobtk"
-features = ["python-extension"]
\ No newline at end of file
+features = ["pyo3/extension-module"]
diff --git a/Cargo.toml b/Cargo.toml
index 3e64ef4..f637756 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -18,20 +18,14 @@ include = [
     "src/*",
 ]
 
-[features]
-default = []
-python-extension = ["pyo3/extension-module", "pyo3"]
-
 [lib]
 name = "blobtk"
 path = "src/lib.rs"
-crate-type = ["cdylib", "lib"]
+crate-type = ["cdylib"]
 
 [profile.release]
 strip = true
 lto = true
-codegen-units = 1
-panic = "abort"
 
 [dependencies]
 anyhow = "1.0.71"
