diff --git a/rust/pyproject.toml b/rust/pyproject.toml
index fb71d4e..72f9971 100644
--- a/rust/pyproject.toml
+++ b/rust/pyproject.toml
@@ -4,9 +4,13 @@ build-backend = "maturin"
 
 [project]
 name = "blobtk"
+version = "0.6.5"
 requires-python = ">=3.8"
 classifiers = [
     "Programming Language :: Rust",
     "Programming Language :: Python :: Implementation :: CPython",
     "Programming Language :: Python :: Implementation :: PyPy",
-]
\ No newline at end of file
+]
+
+[tool.maturin]
+features = ["pyo3/extension-module"]
