diff --git a/rust/pyproject.toml b/rust/pyproject.toml
index fb71d4e..72f9971 100644
--- a/rust/pyproject.toml
+++ b/rust/pyproject.toml
@@ -4,9 +4,13 @@ build-backend = "maturin"
 
 [project]
 name = "blobtk"
+version = "0.6.5"
 requires-python = ">=3.8"
 classifiers = [
     "Programming Language :: Rust",
     "Programming Language :: Python :: Implementation :: CPython",
     "Programming Language :: Python :: Implementation :: PyPy",
-]
\ No newline at end of file
+]
+
+[tool.maturin]
+features = ["pyo3/extension-module"]
diff --git a/rust/Cargo.toml b/rust/Cargo.toml
index 9a7e615..10590a5 100644
--- a/rust/Cargo.toml
+++ b/rust/Cargo.toml
@@ -21,7 +21,7 @@ include = [
 [lib]
 name = "blobtk"
 path = "src/lib.rs"
-crate-type = ["cdylib", "lib"]
+crate-type = ["cdylib"]
 
 [profile.release]
 strip = true
