diff --git a/src/api/CMakeLists.txt b/src/api/CMakeLists.txt
index 7e3d3ca..cb3ada3 100644
--- src/api/CMakeLists.txt
+++ src/api/CMakeLists.txt
@@ -54,8 +54,8 @@ target_link_libraries( BamTools        ${APILibs} )
 target_link_libraries( BamTools-static ${APILibs} )
 
 # set library install destinations
-install( TARGETS BamTools        LIBRARY DESTINATION "lib/bamtools" RUNTIME DESTINATION "bin")
-install( TARGETS BamTools-static ARCHIVE DESTINATION "lib/bamtools")
+install( TARGETS BamTools        LIBRARY DESTINATION "lib" RUNTIME DESTINATION "bin")
+install( TARGETS BamTools-static ARCHIVE DESTINATION "lib")
 
 # export API headers
 include(../ExportHeader.cmake)
