*** LINKS	2019-07-22 13:32:15.844407640 -0700
--- LINKS_new	2020-02-03 15:03:25.066710359 -0800
***************
*** 22,28 ****
  use strict;
  use POSIX;
  use FindBin;
! use lib "$FindBin::Bin/./lib/bloomfilter/swig";
  use BloomFilter;
  use Time::HiRes;
  use Getopt::Std;
--- 22,28 ----
  use strict;
  use POSIX;
  use FindBin;
! use lib "$FindBin::Bin/../lib/site_perl/BloomFilter";
  use BloomFilter;
  use Time::HiRes;
  use Getopt::Std;
