
--- LINKS.old   2016-02-16 22:12:55.467119807 +0100
+++ LINKS       2016-02-16 22:13:35.715954268 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 ##!/gsc/software/linux-x86_64-centos5/perl-5.16.3/bin/perl

 #AUTHOR
@@ -77,7 +77,7 @@
    #perl 5.8.8
    ##use lib "$FindBin::Bin/./lib/Bloom-Faster-1.6/bloom/lib64/perl5/site_perl/5.8.8/x86_64-linux-thread-multi";
    ##perl 5.16.3
-   use lib "$FindBin::Bin/./lib/Bloom-Faster-1.7/bloom5-16-3/lib/site_perl/5.16.3/x86_64-linux";###rebuild against PERL version or replace by pre-built
+   ##use lib "$FindBin::Bin/./lib/Bloom-Faster-1.7/bloom5-16-3/lib/site_perl/5.16.3/x86_64-linux";###rebuild against PERL version or replace by pre-built
    use Bloom::Faster;
 }

