diff --git a/settings.sh b/settings.sh
index 4afd7bd..191dbf9 100755
--- a/settings.sh
+++ b/settings.sh
@@ -36,11 +36,11 @@ export DV_USE_PREINSTALLED_TF="${DV_USE_PREINSTALLED_TF:-0}"
 
 export TF_NEED_GCP=1
 
-export CUDNN_INSTALL_PATH="/usr/lib/x86_64-linux-gnu"
+export CUDNN_INSTALL_PATH="${PREFIX}/lib"
 
 # The version of bazel we want to build DeepVariant.
 # https://www.tensorflow.org/install/source#tested_build_configurations
-DV_BAZEL_VERSION="5.3.0"
+DV_BAZEL_VERSION="6.3.2"
 
 # We need to make sure that $HOME/bin is first in the binary search path so that
 # `bazel` will find the latest version of bazel installed in the user's home
@@ -95,7 +95,7 @@ export DV_INSTALL_GPU_DRIVERS="${DV_INSTALL_GPU_DRIVERS:-0}"
 export PYTHON_VERSION=3.10
 # shellcheck disable=SC2155
 export PYTHON_BIN_PATH="$(which python${PYTHON_VERSION})"
-export PYTHON_LIB_PATH="/usr/local/lib/python${PYTHON_VERSION}/dist-packages"
+export PYTHON_LIB_PATH="${PREFIX}/lib/python${PYTHON_VERSION}/site-packages"
 export USE_DEFAULT_PYTHON_LIB_PATH=1
 # N.B. The --experimental_build_setting_api had to be added on protobuf
 # upgrade to 3.9.2 to avoid error in bazel_skylib:
@@ -104,7 +104,7 @@ export USE_DEFAULT_PYTHON_LIB_PATH=1
 #    --experimental_build_setting_api"
 # Presumably it won't be needed at some later point when bazel_skylib is
 # upgraded again.
-export DV_COPT_FLAGS="--copt=-march=corei7 --copt=-Wno-sign-compare --copt=-Wno-write-strings --experimental_build_setting_api --java_runtime_version=remotejdk_11"
+export DV_COPT_FLAGS="--copt=-march=corei7 --copt=-Wno-sign-compare --copt=-Wno-write-strings --copt=-Wno-deprecated-declarations --experimental_build_setting_api --java_runtime_version=remotejdk_11"
 
 function note_build_stage {
   echo "========== [$(date)] Stage '${1}' starting"
