--- fasta_unittest.cpp
+++ fasta_unittest.cpp
@@ -94,7 +94,9 @@
     while (fa.fillFromFile(inStream))
     {
         counter++;
-        string expectedName = static_cast<ostringstream*>( &(ostringstream() << counter) )->str();
+        ostringstream stream = ostringstream();
+        stream << counter;
+        string expectedName = stream.str();
         EXPECT_EQ(0, fa.name().compare(expectedName));
         EXPECT_EQ(0, fa.seq().compare("ACGT"));
     }
--- fastq_unittest.cpp
+++ fastq_unittest.cpp
@@ -50,7 +50,9 @@
     while (fq.fillFromFile(inStream))
     {
         counter++;
-        string expectedName = static_cast<ostringstream*>( &(ostringstream() << counter) )->str();
+        ostringstream stream = ostringstream();
+        stream << counter;
+        string expectedName = stream.str();
         EXPECT_EQ(0, fq.name().compare(expectedName));
         EXPECT_EQ(0, fq.seq().compare("ACGT"));
     }
