--- setup.py
+++ setup.py
@@ -100,10 +100,10 @@
     # copy samtools, bcftools and config files
     bin_dir = os.path.join(install_dir, "bin")
     lib_dir = os.path.join(install_dir, "lib")
-    plugins_dir = os.path.join(install_dir, "libexec", "bcftools")
+    libexec_dir = os.path.join(install_dir, "libexec")
     etc_dir = os.path.join(install_dir, "etc")
     config_dir = os.path.join(etc_dir, "gemBS_configs")
-    for dir in [bin_dir, lib_dir, plugins_dir, config_dir]:
+    for dir in [bin_dir, lib_dir, libexec_dir, config_dir]:
         if not os.path.exists(dir):
             os.makedirs(dir)
     if not (inst.minimal or inst.no_samtools):
@@ -124,15 +124,8 @@
                         shutil.copy(os.path.join(htslib,file), lib_dir)
                         os.chmod(os.path.join(lib_dir, file), 0o755)
                                 
-    if os.path.exists("tools/bcftools/bcftools"):
-        # print ("Copy binary: bcftools to {}".format(bin_dir))
-        shutil.copy("tools/bcftools/bcftools", bin_dir)
-        os.chmod(os.path.join(bin_dir, "bcftools"), 0o755)
-    plugins = [x for x in glob.glob("tools/bcftools/plugins/*.so")]
-    for file in plugins:
-        # print ("Copy plugin: {} to {}".format(file, plugins_dir))
-        shutil.copy(file, plugins_dir)
-        os.chmod(os.path.join(plugins_dir,os.path.basename(file)), 0o755)
+    os.symlink(os.path.join(os.environ["PREFIX"], "bin/bcftools"), os.path.join(bin_dir, "bcftools"))
+    os.symlink(os.path.join(os.environ["PREFIX"], "libexec/bcftools"), os.path.join(libexec_dir, "bcftools"))
 
     files = [x for x in os.listdir("gemBS/etc")]
     for file in files:
@@ -168,7 +161,7 @@
         _install.initialize_options(self)
         
     def run(self):
-        options=['setup', '_bcftools', '_utils']
+        options=['setup', '_utils']
         if not self.minimal:
             if not self.no_samtools:
                 options.append('_samtools')
--- tools/utils/Makefile
+++ tools/utils/Makefile
@@ -9,8 +9,6 @@
 
 # Definitions
 
-CC=gcc
-
 ROOT_PATH=..
 
 TOOLS=gemBS_cat readNameClean
@@ -31,4 +29,4 @@
 debug: $(TOOLS_BIN)
 
 $(TOOLS_BIN): $(TOOLS_SRC)
-	$(CC) --std=c99  $(TOOLS_FLAGS) -o $@ $(notdir $@).c $(LIB_PATH_FLAGS) $(INCLUDE_FLAGS) $(LIBS) $(EXTRA_LIBS)
+	$(CC) $(CPPFLAGS) $(CFLAGS) $(LDFLAGS) --std=c99  $(TOOLS_FLAGS) -o $@ $(notdir $@).c $(LIB_PATH_FLAGS) $(INCLUDE_FLAGS) $(LIBS) $(EXTRA_LIBS)
