--- ISESCAN_REPO/constants.py	2019-10-22 13:34:32.879032371 -0400
+++ constants.py	2019-10-22 14:10:57.657185417 -0400
@@ -7,14 +7,14 @@
 # 
 # FragGeneScan
-FragGeneScan = '/apps/inst/FragGeneScan1.30/run_FragGeneScan.pl'
+FragGeneScan = 'run_FragGeneScan.pl'
 # Hmmer
-phmmer = '/apps/inst/hmmer-3.3/bin/phmmer'
-hmmsearch = '/apps/inst/hmmer-3.3/bin/hmmsearch'
+phmmer = 'phmmer'
+hmmsearch = 'hmmsearch'
 # Blast 
-blastn = '/apps/inst/ncbi-blast-2.10.0+/bin/blastn'
-blastp = '/apps/inst/ncbi-blast-2.10.0+/bin/blastp'
-makeblastdb = '/apps/inst/ncbi-blast-2.10.0+/bin/makeblastdb'
+blastn = 'blastn'
+blastp = 'blastp'
+makeblastdb = 'makeblastdb'
 
 # get path where isescan.py is
 import sys
