--- a/ssw201507/ssw.h	2025-04-08 13:18:11
+++ b/ssw201507/ssw.h	2026-09-11 13:11:40
@@ -14,7 +14,11 @@
 #include <stdio.h>
 #include <stdint.h>
 #include <string.h>
+#if defined(__ARM_NEON) || defined(__aarch64__)
+#include <sse2neon.h>
+#else
 #include <emmintrin.h>
+#endif
 
 #ifdef __cplusplus
 extern "C" {
--- a/ssw201507/ssw.c	2025-04-08 13:18:11
+++ b/ssw201507/ssw.c	2026-09-11 13:11:40
@@ -35,7 +35,11 @@
  *
  */
 
+#if defined(__ARM_NEON) || defined(__aarch64__)
+#include <sse2neon.h>
+#else
 #include <emmintrin.h>
+#endif
 #include <stdint.h>
 #include <stdlib.h>
 #include <stdio.h>
