diff --git a/Makefile b/Makefile
index 4ef7ec4..f98c8c3 100644
--- a/Makefile
+++ b/Makefile
@@ -12,17 +12,17 @@ LINKER_COMMANDS =-out:
 LIBRARIES = pthreadVC2S.lib
 INCLUDES= -Isrc/distanceCalculation -Isrc/ -Ilib/includes
 else
-CC = g++
-LINK = g++
-OPTIMIZATION_LEVEL=-O3 -msse2
+CC ?= $(CXX)
+LINK ?= $(CXX)
+OPTIMIZATION_LEVEL=-O3 -msse2 -march=x86-64-v3
 DEBUG= #-Wall #-g #-pg
 OBJECTFLAG =-o
 COMPILEFLAG =-c
-LIBRARIES =-lpthread
+LIBRARIES =-pthread
 SWITCHES =
 LINK_OPTIONS = $(SWITCHES)
 LINKER_COMMANDS =-o 
-INCLUDES= -Isrc/distanceCalculation -Isrc/
+INCLUDES= -Isrc/distanceCalculation -Isrc/ -I$(PREFIX)/include
 endif
 
 SRCPATH=src
