--- c++/src/build-system/configure.old	2025-03-27 23:30:30.871788804 +0000
+++ c++/src/build-system/configure	2025-03-27 23:31:46.143716731 +0000
@@ -27748,7 +27748,7 @@
             linux*:aarch64* ) vdb_libpath=linux/$vdb_mode/arm64/lib  ;;
             * )
                if test -n "$with_vdb"; then
-                  as_fn_error $? "--with-vdb explicitly specified,
+                  as_fn_error $? "$host_os:$host_cpu:$bit64_sfx --with-vdb explicitly specified,
                                 but not supported on this platform." "$LINENO" 5
                fi
                ;;
