--- a/mlst.py
+++ b/mlst.py
@@ -293,7 +293,7 @@ parser.add_argument("-o", "--outdir",
 parser.add_argument("-s", "--species",
                     help="species database used for MLST prediction", required=True)
 parser.add_argument("-p", "--database",
-                    help="Directory containing the databases.", default="/database")
+                    help="Directory containing the databases.", default=os.getenv("MLST_DB"))
 parser.add_argument("-t", "--tmp_dir",
                     help="Temporary directory for storage of the results\
                           from the external software.",
