--- positive_set.pl	2018-02-27 19:30:56.000000000 +0100
+++ new_positive_set.pl	2018-08-29 14:12:45.095414070 +0200
@@ -108,17 +108,13 @@
 
     my $count = 0;
 	my $blast_rpt = $work_dir."tmp/positive_set.bls";
-	my $search = `which usearch 2>&1`;
+	my $search = `which blast 2>&1`;
     my $ublast_cmd = "";
 	chomp($search);
-	if ($search =~ /^which: no/) {
-		$ublast_cmd = $script_dir."/bin/usearch -ublast $positive_fasta -db $blastdb -id $identity -evalue $evalue -maxhits 1 -blast6out $blast_rpt -threads $threads 2> /dev/null";
-	} else {
-		$ublast_cmd = "usearch -ublast $positive_fasta -db $blastdb -id $identity -evalue $evalue -maxhits 1 -blast6out $blast_rpt -threads $threads 2> /dev/null";
-	}
+	blastn -query $positive_fasta -db $blastdb -perc_identity $identity -evalue $evalue -max_target_seqs 1 -outfmt 6 -out $blast_rpt -num_threads $threads
 
     system($ublast_cmd) == 0 
-        or die ("Error running -ublast search. Please ensure USEARCH is properly installed\n");
+        or die ("Error running -blast search. Please ensure blast is properly installed\n");
     
 	# GTF file
 	open (G, ">".$positive_set_gtf) or die "Error creating file $positive_set_gtf\n";
