diff --git i/src/CMakeLists.txt w/src/CMakeLists.txt
index f351aa6..edc0f62 100644
--- i/src/CMakeLists.txt
+++ w/src/CMakeLists.txt
@@ -100,7 +100,7 @@ target_compile_options (HS_call_variants PRIVATE -lm)
 target_compile_options (HS_call_variants PRIVATE -fopenmp)
 target_compile_options (HS_call_variants PRIVATE -std=c++17)
 target_compile_options (HS_call_variants PRIVATE -O3)
-target_compile_options (HS_call_variants PRIVATE -march=x86-64)
+#target_compile_options (HS_call_variants PRIVATE -march=x86-64)
 
 
 file (GLOB SOURCE_SEPARATE_READS "separate_reads.cpp" "input_output.cpp" "cluster_graph.cpp" "robin_hood.h" "tools.cpp" "read.cpp" "sequence.cpp" "Partition.cpp")
@@ -111,7 +111,7 @@ target_compile_options (HS_separate_reads PRIVATE -lm)
 target_compile_options (HS_separate_reads PRIVATE -fopenmp)
 target_compile_options (HS_separate_reads PRIVATE -std=c++17)
 target_compile_options (HS_separate_reads PRIVATE -O3)
-target_compile_options (HS_separate_reads PRIVATE -march=x86-64)
+#target_compile_options (HS_separate_reads PRIVATE -march=x86-64)
 
 file (GLOB SOURCE_CREATE_NEW_CONTIG "create_new_contigs.cpp" "input_output.cpp" "robin_hood.h" "tools.cpp" "read.cpp" "sequence.cpp" "Partition.cpp")
 add_executable(HS_create_new_contigs ${SOURCE_CREATE_NEW_CONTIG} edlib/src/edlib.cpp)
@@ -120,7 +120,7 @@ target_compile_options (HS_create_new_contigs PRIVATE -lm)
 target_compile_options (HS_create_new_contigs PRIVATE -fopenmp)
 target_compile_options (HS_create_new_contigs PRIVATE -std=c++17)
 target_compile_options (HS_create_new_contigs PRIVATE -O3)
-target_compile_options (HS_create_new_contigs PRIVATE -march=x86-64)
+#target_compile_options (HS_create_new_contigs PRIVATE -march=x86-64)
 
 find_package(OpenMP)
 if(OpenMP_CXX_FOUND)
diff --git i/src/HS_GenomeTailor/CMakeLists.txt w/src/HS_GenomeTailor/CMakeLists.txt
index 80ee4dc..f117ecc 100644
--- i/src/HS_GenomeTailor/CMakeLists.txt
+++ w/src/HS_GenomeTailor/CMakeLists.txt
@@ -25,4 +25,4 @@ target_compile_options (HS_GenomeTailor PRIVATE -g)
 target_compile_options (HS_GenomeTailor PRIVATE -lm)
 target_compile_options (HS_GenomeTailor PRIVATE -std=c++17)
 target_compile_options (HS_GenomeTailor PRIVATE -O3)
-target_compile_options (HS_GenomeTailor PRIVATE -march=x86-64)
\ No newline at end of file
+#target_compile_options (HS_GenomeTailor PRIVATE -march=x86-64)
