--- cmake/BuildDeps.cmake	2017-02-24 13:54:38.727623938 +0100
+++ cmake/BuildDeps.cmake	2017-02-24 16:23:35.495579111 +0100
@@ -45,8 +45,9 @@
         URL ${SAMTOOLS_URL}
         SOURCE_DIR ${SAMTOOLS_ROOT}
         BINARY_DIR ${SAMTOOLS_ROOT}
+        PATCH_COMMAND patch -p1 < $ENV{RECIPE_DIR}/samtools.patch
         CONFIGURE_COMMAND echo "Building samtools, build log at ${SAMTOOLS_LOG}"
-        BUILD_COMMAND make > ${SAMTOOLS_LOG} 2>&1
+        BUILD_COMMAND make 2>&1
         INSTALL_COMMAND ""
     )
 
