--- samtools-0.1.19/Makefile	2013-03-19 08:48:09.000000000 +0100
+++ samtools-0.1.19/Makefile	2017-02-24 17:09:11.277467777 +0100
@@ -1,5 +1,5 @@
 CC=			gcc
-CFLAGS=		-g -Wall -O2
+CFLAGS=		-g -Wall -O2 -L${PREFIX}/lib -I${PREFIX}/include
 #LDFLAGS=		-Wl,-rpath,\$$ORIGIN/../lib
 DFLAGS=		-D_FILE_OFFSET_BITS=64 -D_LARGEFILE64_SOURCE -D_USE_KNETFILE -D_CURSES_LIB=1
 KNETFILE_O=	knetfile.o
@@ -12,10 +12,10 @@
 			cut_target.o phase.o bam2depth.o padding.o bedcov.o bamshuf.o \
 			bam_tview_curses.o bam_tview_html.o
 PROG=		samtools
-INCLUDES=	-I.
+INCLUDES=	-I. -I${PREFIX}/include -I${PREFIX}/include/ncurses
 SUBDIRS=	. bcftools misc
-LIBPATH=
-LIBCURSES=	-lcurses # -lXCurses
+LIBPATH=	-L${PREFIX}/lib
+LIBCURSES=	-lncurses -ltinfo # -lXCurses
 
 .SUFFIXES:.c .o
 .PHONY: all lib
--- samtools-0.1.19/misc/Makefile	2013-03-19 08:48:09.000000000 +0100
+++ samtools-0.1.19/misc/Makefile	2017-02-24 17:09:38.417804184 +0100
@@ -1,11 +1,11 @@
 CC=			gcc
 CXX=		g++
-CFLAGS=		-g -Wall -O2 #-m64 #-arch ppc
+CFLAGS=		-g -Wall -O2 -I${PREFIX}/include -L${PREFIX}/lib #-m64 #-arch ppc
 CXXFLAGS=	$(CFLAGS)
 DFLAGS=		-D_FILE_OFFSET_BITS=64
 OBJS=		
 PROG=		md5sum-lite md5fa maq2sam-short maq2sam-long ace2sam wgsim bamcheck
-INCLUDES=	-I..
+INCLUDES=	-I.. -I${PREFIX}/include
 SUBDIRS=	.
 
 .SUFFIXES:.c .o
