--- a/src/SimplifyGraph/src/CS2_stream/parser_cs2.h
+++ b/src/SimplifyGraph/src/CS2_stream/parser_cs2.h
@@ -83,7 +83,7 @@
 		pos_current=0;          /* 2*no_alines */
 
 	char    in_line[MAXLINE],       /* for reading input line */
-		pr_type[3];             /* for reading type of the problem */
+		pr_type[4];             /* for reading type of the problem */
 
 	int     k,                      /* temporary */
 		err_no;                 /* no of detected error */
