--- a/core/mutations.h
+++ b/core/mutations.h
@@ -125,7 +125,7 @@
 class Missation_map {
  private:
   using From_states_allocator =
-      typename std::allocator_traits<Allocator>::template rebind_alloc<Missation>;
+      typename std::allocator_traits<Allocator>::template rebind_alloc<std::pair<Site_index, Real_seq_letter>>;
   using From_states_map =
       boost::container::flat_map<Site_index, Real_seq_letter, std::less<Site_index>, From_states_allocator>;
   
