--- targqc/utilz/setup_utils.py
+++ targqc/utilz/setup_utils.py
@@ -52,16 +52,9 @@
 
 
 def get_reqs():
-    try: # for pip >= 10
-        from pip._internal.req import parse_requirements
-    except ImportError: # for pip <= 9.0.3
-        from pip.req import parse_requirements
-
-    try:
-        install_reqs = parse_requirements('requirements.txt', session=False)
-    except TypeError:
-        install_reqs = parse_requirements('requirements.txt')
-    reqs = [str(ir.req) for ir in install_reqs if ir.req]
+    with open("requirements.txt") as requirements_txt:
+        requirements_lines = [line.strip() for line in requirements_txt.readlines()]
+    reqs = [line for line in requirements_lines if line and not line.startswith("#")]
     return reqs
 
 
