--- Makefile.orig
+++ Makefile
@@ -8,13 +8,13 @@
 EFILE = mph
 
 # Intel oneAPI DPC++/C++ Compiler
-CXX = icpx
+CXX ?= g++
 
 # Compiler flags
-CXXFLAGS = -Wall -O3 -qmkl -qopenmp -std=c++11 -static 
+CXXFLAGS ?= -Wall -O3 -std=c++11
 # Includes
 # Change the EIGEN path before compiling.
-INCLUDES = -I/home/jjiang26/software/eigen-3.4.0
+INCLUDES ?=
 
 SRC_DIR := ./src
 SRCS := $(wildcard $(SRC_DIR)/*.cpp)
@@ -28,7 +28,7 @@
 	$(CXX) $(CXXFLAGS) $(INCLUDES) -c $< -o $@
 
 $(EFILE) : $(OBJS)
-	$(CXX) $(CXXFLAGS) -o $(EFILE) $(OBJS)
+	$(CXX) $(CXXFLAGS) $(LDFLAGS) -o $(EFILE) $(OBJS) $(LDLIBS)
 
 .PHONY: clean
 clean: 
