--- src/hapbin.cpp.	2020-08-25 23:53:06.158942633 +0000
+++ src/hapbin.cpp	2020-08-26 09:33:02.761858651 +0000
@@ -33,9 +33,9 @@
 #elif (_POSIX_C_SOURCE >= 200112L || _XOPEN_SOURCE >= 600) && !defined(_ISOC11_SOURCE)
-void* aligned_alloc(size_t alignment, size_t size)
-{
-    void* ret;
-    if (posix_memalign(&ret, alignment, size) != 0)
-        throw std::bad_alloc();
-    return ret;
-}
+// void* aligned_alloc(size_t alignment, size_t size)
+// {
+//     void* ret;
+//     if (posix_memalign(&ret, alignment, size) != 0)
+//         throw std::bad_alloc();
+//     return ret;
+// }
 #endif
