diff -cpr ./src/draw.c ../../phylip-3.697/src/draw.c
*** ./src/draw.c	2014-09-17 03:38:54.000000000 +0000
--- ../../phylip-3.697/src/draw.c	2024-12-06 07:29:59.908869727 +0000
***************
*** 4,9 ****
--- 4,14 ----
  
  #include "phylip.h"
  #include "draw.h"
+ long treecolor, namecolor, backcolor, bottomcolor, vrmlskycolornear, vrmlskycolorfar,
+      vrmlgroundcolornear, vrmlgroundcolorfar, vrmlplotcolor;
+ 
+ char afmfile[FNMLNGTH];
+ 
  
  #ifdef QUICKC
  struct videoconfig myscreen;
diff -cpr ./src/draw.h ../../phylip-3.697/src/draw.h
*** ./src/draw.h	2014-09-17 03:38:54.000000000 +0000
--- ../../phylip-3.697/src/draw.h	2024-12-06 07:29:44.658996204 +0000
*************** typedef struct vrmllighttype {
*** 52,61 ****
    double intensity, x, y, z;
  } vrmllighttype;
  
! long treecolor, namecolor, backcolor, bottomcolor, vrmlskycolornear, vrmlskycolorfar,
       vrmlgroundcolornear, vrmlgroundcolorfar, vrmlplotcolor;
       
! char afmfile[FNMLNGTH];
  
  double lengthtext(char *, long, char *, fonttype);
  double heighttext(fonttype, char *);
--- 52,61 ----
    double intensity, x, y, z;
  } vrmllighttype;
  
! extern long treecolor, namecolor, backcolor, bottomcolor, vrmlskycolornear, vrmlskycolorfar,
       vrmlgroundcolornear, vrmlgroundcolorfar, vrmlplotcolor;
       
! extern char afmfile[FNMLNGTH];
  
  double lengthtext(char *, long, char *, fonttype);
  double heighttext(fonttype, char *);
diff -cpr ./src/drawtree.c ../../phylip-3.697/src/drawtree.c
*** ./src/drawtree.c	2014-09-19 15:37:55.000000000 +0000
--- ../../phylip-3.697/src/drawtree.c	2024-12-06 07:31:40.318042205 +0000
*************** boolean     canbeplotted, dotmatrix, has
*** 69,75 ****
               uselengths, regular, rotate, empty, rescaled,
               notfirst, improve, nbody, firstscreens, labelavoid;
  boolean     pictbold,pictitalic,pictshadow,pictoutline;
- boolean javarun;
  
  striptype stripe;
  plottertype plotter, oldplotter;
--- 69,74 ----
Only in ../../phylip-3.697/src: libdrawgram.so
Only in ../../phylip-3.697/src: libdrawtree.so
diff -cpr ./src/phylip.c ../../phylip-3.697/src/phylip.c
*** ./src/phylip.c	2014-09-17 03:38:55.000000000 +0000
--- ../../phylip-3.697/src/phylip.c	2024-12-06 07:31:50.357960299 +0000
***************
*** 33,39 ****
  #include <signal.h>
  
  #include "phylip.h"
! 
  #ifdef WIN32
  #include <windows.h>
  /* for console code (clear screen, text color settings) */
--- 33,39 ----
  #include <signal.h>
  
  #include "phylip.h"
! boolean javarun;
  #ifdef WIN32
  #include <windows.h>
  /* for console code (clear screen, text color settings) */
diff -cpr ./src/phylip.h ../../phylip-3.697/src/phylip.h
*** ./src/phylip.h	2024-12-06 07:32:47.867489055 +0000
--- ../../phylip-3.697/src/phylip.h	2024-12-06 07:28:32.939592771 +0000
*************** MALLOCRETURN    *mymalloc(long);
*** 331,337 ****
  /* Lower-triangular format. */
  #define MAT_LOWERTRI    (MAT_LOWER | MAT_MACHINE)
  
! boolean javarun;
  
  typedef long *steptr;
  typedef long longer[6];
--- 331,337 ----
  /* Lower-triangular format. */
  #define MAT_LOWERTRI    (MAT_LOWER | MAT_MACHINE)
  
! extern boolean javarun;
  
  typedef long *steptr;
  typedef long longer[6];
*************** extern FILE *infile, *outfile,  *intree,
*** 351,357 ****
  extern long spp, words, bits;
  extern boolean ibmpc, ansi, tranvsp;
  extern naym *nayme;                     /* names of species */
! boolean firstplotblock; // for debugging BMP output
  
  #define ebcdic          EBCDIC
  
--- 351,357 ----
  extern long spp, words, bits;
  extern boolean ibmpc, ansi, tranvsp;
  extern naym *nayme;                     /* names of species */
! extern boolean firstplotblock; // for debugging BMP output
  
  #define ebcdic          EBCDIC
  
