--- src/phylip.h
+++ src/phylip.h
@@ -96,7 +96,7 @@
 #define EBCDIC false
 #define INFILE "infile"
 #define OUTFILE "outfile"
-#define FONTFILE "fontfile" /* on unix this might be /usr/local/lib/fontfile */
+#define FONTFILE "@@prefix@@/share/phylip-3.696-0/fonts/fontfile" /* on unix this might be /usr/local/lib/fontfile */
 #define PLOTFILE "plotfile"
 #define INTREE "intree"
 #define INTREE2 "intree2"
