diff -Naur phylip-3.696/src/dnacomp.c phylip-3.696_patched/src/dnacomp.c
--- phylip-3.696/src/dnacomp.c	2014-09-17 05:38:54.000000000 +0200
+++ phylip-3.696_patched/src/dnacomp.c	2017-03-24 11:42:35.000000000 +0100
@@ -31,7 +31,7 @@
    POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define maxtrees        100   /* maximum number of tied trees stored     */
+#define maxtrees        10000   /* maximum number of tied trees stored     */
 
 typedef boolean *boolptr;
 
diff -Naur phylip-3.696/src/dnapenny.c phylip-3.696_patched/src/dnapenny.c
--- phylip-3.696/src/dnapenny.c	2014-09-17 05:38:54.000000000 +0200
+++ phylip-3.696_patched/src/dnapenny.c	2017-03-24 11:42:48.000000000 +0100
@@ -30,7 +30,7 @@
    POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define maxtrees        1000  /* maximum number of trees to be printed out   */
+#define maxtrees        10000  /* maximum number of trees to be printed out   */
 #define often           1000  /* how often to notify how many trees examined */
 #define many            10000 /* how many multiples of howoften before stop  */
 
diff -Naur phylip-3.696/src/dollop.c phylip-3.696_patched/src/dollop.c
--- phylip-3.696/src/dollop.c	2014-09-17 05:38:54.000000000 +0200
+++ phylip-3.696_patched/src/dollop.c	2017-03-24 11:43:02.000000000 +0100
@@ -32,7 +32,7 @@
    POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define maxtrees        100  /* maximum number of tied trees stored     */
+#define maxtrees        10000  /* maximum number of tied trees stored     */
 
 #ifndef OLDC
 /* function prototypes */
diff -Naur phylip-3.696/src/dolpenny.c phylip-3.696_patched/src/dolpenny.c
--- phylip-3.696/src/dolpenny.c	2014-09-17 05:38:54.000000000 +0200
+++ phylip-3.696_patched/src/dolpenny.c	2017-03-24 11:43:12.000000000 +0100
@@ -31,7 +31,7 @@
    POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define maxtrees        1000  /* maximum number of trees to be printed out   */
+#define maxtrees        10000  /* maximum number of trees to be printed out   */
 #define often           100   /* how often to notify how many trees examined */
 #define many            1000  /* how many multiples of howoften before stop  */
 
diff -Naur phylip-3.696/src/mix.c phylip-3.696_patched/src/mix.c
--- phylip-3.696/src/mix.c	2014-09-17 05:38:54.000000000 +0200
+++ phylip-3.696_patched/src/mix.c	2017-03-24 11:43:24.000000000 +0100
@@ -33,7 +33,7 @@
 */
 
 
-#define maxtrees        100  /* maximum number of tied trees stored     */
+#define maxtrees        10000  /* maximum number of tied trees stored     */
 
 typedef long *placeptr;
 
diff -Naur phylip-3.696/src/penny.c phylip-3.696_patched/src/penny.c
--- phylip-3.696/src/penny.c	2014-09-17 05:38:54.000000000 +0200
+++ phylip-3.696_patched/src/penny.c	2017-03-24 11:43:52.000000000 +0100
@@ -32,7 +32,7 @@
    POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define maxtrees        1000 /* maximum number of trees to be printed out   */
+#define maxtrees        10000 /* maximum number of trees to be printed out   */
 #define often           100  /* how often to notify how many trees examined */
 #define many            1000 /* how many multiples of howoften before stop  */
 
diff -Naur phylip-3.696/src/protpars.c phylip-3.696_patched/src/protpars.c
--- phylip-3.696/src/protpars.c	2014-09-17 05:38:55.000000000 +0200
+++ phylip-3.696_patched/src/protpars.c	2017-03-24 11:44:03.000000000 +0100
@@ -32,7 +32,7 @@
 */
 
 
-#define maxtrees        100   /* maximum number of tied trees stored */
+#define maxtrees        10000   /* maximum number of tied trees stored */
 
 typedef enum {
   universal, ciliate, mito, vertmito, flymito, yeastmito
