--- Makefile	2017-12-11 14:51:11.000000000 +0000
+++ Makefile	2017-12-14 12:16:14.000000000 +0000
@@ -8,7 +8,7 @@
 #

 #Basic flags every build needs
-LIBS=-lz
+LIBS=-lcrypto -lz -llzma -lbz2 -lcurl
 CXXFLAGS ?= -g -O3
 CXXFLAGS += -std=c++11 -fopenmp -fsigned-char
 CFLAGS ?= -O3 -std=c99
@@ -46,18 +46,6 @@
     EIGEN_CHECK=
 endif

-# Default to build and link the libhts submodule
-ifeq ($(HTS), install)
-    HTS_LIB=./htslib/libhts.a
-    HTS_INCLUDE=-I./htslib
-else
-    # Use system-wide htslib
-    HTS_LIB=-lhts
-endif
-
-# Include the header-only fast5 library
-FAST5_INCLUDE=-I./fast5/include
-
 # Include the header-only eigen library
 EIGEN_INCLUDE=-I./eigen/

--- src/nanopolish_methyltrain.cpp	2017-07-01 15:55:34.000000000 +0200
+++ src/nanopolish_methyltrain.cpp	2017-07-01 15:54:06.000000000 +0200
@@ -42,7 +42,7 @@
 #include "logger.hpp"
 
 #include "nanopolish_scorereads.h"
-#include "Eigen/Dense"
+#include <eigen3/Eigen/Dense>
 
 extern float g_p_skip, g_p_skip_self, g_p_bad, g_p_bad_self;
 
