--- Makefile	2017-12-11 14:51:11.000000000 +0000
+++ Makefile	2017-12-13 15:55:55.000000000 +0000
@@ -16,9 +16,9 @@
 CC ?= gcc

 # Change the value of HDF5, EIGEN, or HTS below to any value to disable compilation of bundled code
-HDF5?=install
-EIGEN?=install
-HTS?=install
+HDF5?=noinstall
+EIGEN?=noinstall
+HTS?=noinstall

 # Check operating system, OSX doesn't have -lrt
 UNAME_S := $(shell uname -s)

--- src/nanopolish_methyltrain.cpp	2017-07-01 15:55:34.000000000 +0200
+++ src/nanopolish_methyltrain.cpp	2017-07-01 15:54:06.000000000 +0200
@@ -42,7 +42,7 @@
 #include "logger.hpp"
 
 #include "nanopolish_scorereads.h"
-#include "Eigen/Dense"
+#include <eigen3/Eigen/Dense>
 
 extern float g_p_skip, g_p_skip_self, g_p_bad, g_p_bad_self;
 
