--- Makefile	2017-02-05 22:49:13.000000000 +0100
+++ Makefile	2017-02-05 22:51:07.000000000 +0100
@@ -35,11 +35,11 @@
 endif
 
 # Bulild and link the libhts submodule
-HTS_LIB=./htslib/libhts.a
-HTS_INCLUDE=-I./htslib
+
+
 
 # Include the header-only fast5 library
-FAST5_INCLUDE=-I./fast5/src
+
 
 # Include the src subdirectories
 NP_INCLUDE=$(addprefix -I./, $(SUBDIRS))
--- src/nanopolish_methyltrain.cpp	2017-02-05 22:48:52.000000000 +0100
+++ src//nanopolish_methyltrain.cpp	2017-02-05 22:49:43.000000000 +0100
@@ -42,7 +42,7 @@
 #include "logger.hpp"
 
 #include "nanopolish_scorereads.h"
-#include "../eigen/Eigen/Dense"
+#include <eigen3/Eigen/Dense>
 
 extern float g_p_skip, g_p_skip_self, g_p_bad, g_p_bad_self;
 
