--- Makefile	2017-07-01 15:51:35.000000000 +0200
+++ Makefile	2017-07-01 15:53:16.000000000 +0200
@@ -46,11 +46,8 @@
 endif
 
 # Build and link the libhts submodule
-HTS_LIB=./htslib/libhts.a
-HTS_INCLUDE=-I./htslib
 
 # Include the header-only fast5 library
-FAST5_INCLUDE=-I./fast5/src
 
 # Include the src subdirectories
 NP_INCLUDE=$(addprefix -I./, $(SUBDIRS))
--- src/nanopolish_methyltrain.cpp	2017-07-01 15:55:34.000000000 +0200
+++ src/nanopolish_methyltrain.cpp	2017-07-01 15:54:06.000000000 +0200
@@ -42,7 +42,7 @@
 #include "logger.hpp"
 
 #include "nanopolish_scorereads.h"
-#include "../eigen/Eigen/Dense"
+#include <eigen3/Eigen/Dense>
 
 extern float g_p_skip, g_p_skip_self, g_p_bad, g_p_bad_self;
 
