--- Makefile	2019-01-22 09:32:40.000000000 -0700
+++ Makefile	2019-01-22 09:34:50.000000000 -0700
@@ -8,7 +8,7 @@
 #
 
 #Basic flags every build needs
-LIBS = -lz
+LIBS=-lcrypto -lz -llzma -lbz2 -lcurl -ldeflate -ldl
 CXXFLAGS ?= -g -O3
 CXXFLAGS += -std=c++11 -fopenmp -fsigned-char -D_FILE_OFFSET_BITS=64 #D_FILE_OFFSET_BITS=64 makes nanopolish work in 32 bit systems
 CFLAGS ?= -O3 -std=c99 -fsigned-char -D_FILE_OFFSET_BITS=64 
@@ -50,19 +50,8 @@
     EIGEN_CHECK =
 endif
 
-# Default to build and link the libhts submodule
-ifeq ($(HTS), install)
-    HTS_LIB = ./htslib/libhts.a
-    HTS_INCLUDE = -I./htslib
-else
-    # Use system-wide htslib
-    HTS_LIB =
-    HTS_INCLUDE =
-    LIBS += -lhts
-endif
-
 # Include the header-only fast5 library
-FAST5_INCLUDE = -I./fast5/include
+#FAST5_INCLUDE = -I./fast5/include
 
 # Include the header-only eigen library
 EIGEN_INCLUDE = -I./eigen/
--- src/nanopolish_methyltrain.cpp	2019-01-22 09:31:59.000000000 -0700
+++ src/nanopolish_methyltrain.cpp	2019-01-22 09:32:29.000000000 -0700
@@ -45,7 +45,7 @@
 #include "logger.hpp"
 
 #include "nanopolish_scorereads.h"
-#include "Eigen/Dense"
+#include <eigen3/Eigen/Dense>
 
 extern float g_p_skip, g_p_skip_self, g_p_bad, g_p_bad_self;
 
