--- Makefile	2017-07-01 15:51:35.000000000 +0200
+++ Makefile	2017-07-01 15:53:16.000000000 +0200
@@ -46,11 +46,8 @@
 endif
 
 # Build and link the libhts submodule
-HTS_LIB=./htslib/libhts.a
-HTS_INCLUDE=-I./htslib
 
 # Include the header-only fast5 library
-FAST5_INCLUDE=-I./fast5/include
 
 # Include the src subdirectories
 NP_INCLUDE=$(addprefix -I./, $(SUBDIRS))
--- src/nanopolish_methyltrain.cpp	2017-07-01 15:55:34.000000000 +0200
+++ src/nanopolish_methyltrain.cpp	2017-07-01 15:54:06.000000000 +0200
@@ -42,7 +42,7 @@
 #include "logger.hpp"
 
 #include "nanopolish_scorereads.h"
-#include "Eigen/Dense"
+#include <eigen3/Eigen/Dense>

extern float g_p_skip, g_p_skip_self, g_p_bad, g_p_bad_self;
--- a/src/nanopolish_call_methylation.cpp
+++ b/src/nanopolish_call_methylation.cpp
@@ -20,7 +20,7 @@
#include <set>
#include <omp.h>
#include <getopt.h>
-#include <htslib/faidx.h>
+#include "htslib/faidx.h"
#include "nanopolish_eventalign.h"
#include "nanopolish_iupac.h"
#include "nanopolish_poremodel.h"
--- a/src/nanopolish_call_variants.cpp
+++ b/src/nanopolish_call_variants.cpp
@@ -22,7 +22,7 @@
#include <omp.h>
#include <getopt.h>
#include <iterator>
-#include <htslib/faidx.h>
+#include "htslib/faidx.h"
#include "nanopolish_poremodel.h"
#include "nanopolish_transition_parameters.h"
#include "nanopolish_matrix.h"
--- a/src/nanopolish_getmodel.cpp
+++ b/src/nanopolish_getmodel.cpp
@@ -21,7 +21,7 @@
#include <omp.h>
#include <getopt.h>
#include <fast5.hpp>
-#include <htslib/faidx.h>
+#include "htslib/faidx.h"
#include "nanopolish_poremodel.h"
#include "profiler.h"
--- a/src/nanopolish_methyltrain.cpp
+++ b/src/nanopolish_methyltrain.cpp
@@ -23,7 +23,7 @@
#include <omp.h>
#include <getopt.h>
#include <cstddef>
-#include <htslib/faidx.h>
+#include "htslib/faidx.h"
#include "nanopolish_methyltrain.h"
#include "nanopolish_eventalign.h"
#include "nanopolish_iupac.h"
--- a/src/nanopolish_phase_reads.cpp
+++ b/src/nanopolish_phase_reads.cpp
@@ -23,7 +23,7 @@
#include <omp.h>
#include <getopt.h>
#include <cstddef>
-#include <htslib/faidx.h>
+#include "htslib/faidx.h"
#include "nanopolish_iupac.h"
#include "nanopolish_poremodel.h"
#include "nanopolish_transition_parameters.h"
--- a/src/nanopolish_read_db.cpp
+++ b/src/nanopolish_read_db.cpp
@@ -12,8 +12,8 @@
#include <iostream>
#include <sys/stat.h>
#include "nanopolish_common.h"
-#include <htslib/kseq.h>
-#include <htslib/bgzf.h>
+#include "htslib/kseq.h"
+#include "htslib/bgzf.h"
#include "nanopolish_read_db.h"

#define READ_DB_SUFFIX ".readdb"
--- a/src/nanopolish_scorereads.cpp
+++ b/src/nanopolish_scorereads.cpp
@@ -22,7 +22,7 @@
#include <omp.h>
#include <getopt.h>
#include <cstddef>
-#include <htslib/faidx.h>
+#include "htslib/faidx.h"
#include "nanopolish_alphabet.h"
#include "nanopolish_methyltrain.h"
#include "nanopolish_eventalign.h"

