diff --git a/CMakeLists.txt b/CMakeLists.txt
index 3e78405..f4ba1bb 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -133,10 +133,10 @@ if (ENABLE_FUZZERS OR NOT POD5_DISABLE_TESTS)
 endif()
 
 # Install licenses.
-install(
-    DIRECTORY ${CMAKE_BINARY_DIR}/pod5_conan_licenses/
-    DESTINATION licenses
-)
+#install(
+    #DIRECTORY ${CMAKE_BINARY_DIR}/pod5_conan_licenses/
+    #DESTINATION licenses
+#)
 install(
     FILES
         LICENSE.md
