diff --git a/c++/pod5_format/types.cpp b/c++/pod5_format/types.cpp
index 50adbde..052b243 100644
--- a/c++/pod5_format/types.cpp
+++ b/c++/pod5_format/types.cpp
@@ -28,8 +28,8 @@ bool UuidType::ExtensionEquals(ExtensionType const & other) const
 
 std::shared_ptr<arrow::Array> UuidType::MakeArray(std::shared_ptr<arrow::ArrayData> data) const
 {
-    DCHECK_EQ(data->type->id(), arrow::Type::EXTENSION);
-    DCHECK_EQ(
+    ARROW_DCHECK_EQ(data->type->id(), arrow::Type::EXTENSION);
+    ARROW_DCHECK_EQ(
         static_cast<arrow::ExtensionType const &>(*data->type).extension_name(), extension_name());
     return std::make_shared<UuidArray>(data);
 }
@@ -78,8 +78,8 @@ bool VbzSignalType::ExtensionEquals(ExtensionType const & other) const
 
 std::shared_ptr<arrow::Array> VbzSignalType::MakeArray(std::shared_ptr<arrow::ArrayData> data) const
 {
-    DCHECK_EQ(data->type->id(), arrow::Type::EXTENSION);
-    DCHECK_EQ(
+    ARROW_DCHECK_EQ(data->type->id(), arrow::Type::EXTENSION);
+    ARROW_DCHECK_EQ(
         static_cast<arrow::ExtensionType const &>(*data->type).extension_name(), extension_name());
     return std::make_shared<VbzSignalArray>(data);
 }
