diff --git a/src/libmaus2/parallel/threadpool/bam/BamHeaderSeqData.hpp b/src/libmaus2/parallel/threadpool/bam/BamHeaderSeqData.hpp
index e740798..2b3b3e9 100644
--- a/src/libmaus2/parallel/threadpool/bam/BamHeaderSeqData.hpp
+++ b/src/libmaus2/parallel/threadpool/bam/BamHeaderSeqData.hpp
@@ -23,6 +23,7 @@
 #include <filesystem>
 #include <libmaus2/exception/LibMausException.hpp>
 #include <libmaus2/aio/InputStreamInstance.hpp>
+#include <map>
 
 namespace libmaus2
 {
