diff --git a/src/libmaus2/util/UnsignedIntegerIndexIterator.hpp b/src/libmaus2/util/UnsignedIntegerIndexIterator.hpp
index f4f8d55..6d40e44 100644
--- a/src/libmaus2/util/UnsignedIntegerIndexIterator.hpp
+++ b/src/libmaus2/util/UnsignedIntegerIndexIterator.hpp
@@ -40,7 +40,7 @@ namespace libmaus2
 
 			using iterator_category = ::std::random_access_iterator_tag;
 			using value_type = data_type;
-			using difference_type = index_type;
+			using difference_type = std::ptrdiff_t;
 			using pointer = data_type *;
 			using reference = data_type &;
 
