diff --git a/thirdparty/makefile b/thirdparty/makefile
index a0f62f1..7b013eb 100644
--- a/thirdparty/makefile
+++ b/thirdparty/makefile
@@ -27,7 +27,7 @@ ${TARGET_DIR}/pigz: ${TARGET_DIR}/../pigz-2.4/pigz
 
 ${TARGET_DIR}/../pigz-2.4/pigz: ./pigz/pigz-2.4.tar.gz
 	tar -xzf $< -C ${TARGET_DIR}/.. || exit 255
-	cd ${TARGET_DIR}/../pigz-2.4 && make
+	cd ${TARGET_DIR}/../pigz-2.4 && make CC=${CC}
 
 
 
