--- musicc/core.py.orig	2017-11-07 18:35:42.669752152 +0100
+++ musicc/core.py	2017-11-07 18:36:42.072269720 +0100
@@ -19,7 +19,10 @@
 from scipy import stats
 import pandas as pd
 from sklearn import cross_validation, linear_model  # for the LASSO
-from sklearn.utils import ConvergenceWarning
+try:
+    from sklearn.exceptions import ConvergenceWarning
+except ImportError:
+    from sklearn.utils import ConvergenceWarning
 
 
 ################################################################################################################
