--- src/Makefile	2020-02-28 15:33:54.000000000 +0100
+++ src/Makefile.mod	2020-09-04 12:59:20.000000000 +0200
@@ -1,10 +1,8 @@
+BAMTOOLS= "${SRC_DIR}"/gargammel/bamtools
+LIBGAB= "${SRC_DIR}"/gargammel/libgab
 
-CXX      = g++ #-g
-BAMTOOLS= $(realpath ../bamtools/)
-LIBGAB= $(realpath ../libgab/)
-
-CXXFLAGS =  -std=c++0x -Wall -lm -O3 -lz -I${LIBGAB} -I${LIBGAB}/gzstream/ -I${BAMTOOLS}/src/  -c
-LDFLAGS  = ${LIBGAB}/FastQObj.o  ${LIBGAB}/FastQParser.o ${LIBGAB}/PutProgramInHeader.o ${BAMTOOLS}/build/src/api/libbamtools.a  ${LIBGAB}/utils.o ${LIBGAB}/gzstream/libgzstream.a -lm -lz
+CXXFLAGS += -std=c++0x -Wall -lm -O3 -lz -I${LIBGAB} -I${LIBGAB}/gzstream/ -I${BAMTOOLS}/  -c
+LDFLAGS  += ${LIBGAB}/FastQObj.o  ${LIBGAB}/FastQParser.o ${LIBGAB}/PutProgramInHeader.o ${BAMTOOLS}/lib/libbamtools.a  ${LIBGAB}/utils.o ${LIBGAB}/gzstream/libgzstream.a -lm -lz
 
 
 all: fragSim deamSim adptSim fasta2fastas misincorporation2prof
