From 3e59259c7b4162ca51080c78e3d65a9cd1e66979 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Felix=20M=C3=B6lder?= <felix.moelder@uni-due.de>
Date: Thu, 1 Oct 2020 14:40:57 +0200
Subject: [PATCH] Add entrypoint

---
 debarcer/debarcer.py | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/debarcer/debarcer.py b/debarcer/debarcer.py
index 96501fb..d86c877 100755
--- a/debarcer/debarcer.py
+++ b/debarcer/debarcer.py
@@ -652,8 +652,8 @@ def generate_bed(args):
     WriteTargetsBed(args.bamfile, args.bed, args.mincov, args.regionsize, args.maxdepth, args.ignoreorphans, args.stepper)
     
 
-if __name__ == '__main__':
-        
+
+def main()
     ## Argument + config parsing and error handling
     parser = argparse.ArgumentParser(prog='debarcer.py', description="A package for De-Barcoding\
                                      and Error Correction of sequencing data containing molecular barcodes")
@@ -819,4 +819,6 @@ def generate_bed(args):
     except AttributeError as e:
         print(e)
         print(parser.format_help())
-    
+
+if __name__ == '__main__':
+    main()
\ No newline at end of file
