diff --git i/Makefile w/Makefile
index c051218..7220284 100755
--- i/Makefile
+++ w/Makefile
@@ -1,25 +1,25 @@
-SAMTOOLS=/scratch/Shared/samtools-0.1.18/
-
-CC = g++
-CFLAGS =  -O3 -Wall -I$(SAMTOOLS) -L$(SAMTOOLS)
-
+CXX ?= g++
+CXXFLAGS ?=  -O3 -Wall -I${PREFIX}/include -I${PREFIX}/include/bam/ -L${PREFIX}/lib
 all: fastSVseq 
 
 Insertion.o : Insertion.cpp Insertion.h
-	$(CC) $(CFLAGS) -c Insertion.cpp
+	$(CXX) $(CXXFLAGS) -c Insertion.cpp
 
 Caller.o : Caller.cpp  Caller.h
-	$(CC) $(CFLAGS) -c Caller.cpp
+	$(CXX) $(CXXFLAGS) -c Caller.cpp
 
 Reference.o : Reference.cpp  Reference.h
-	$(CC) $(CFLAGS) -c Reference.cpp
+	$(CXX) $(CXXFLAGS) -c Reference.cpp
 
 MySamBam.o : MySamBam.cpp  MySamBam.h
-	$(CC) $(CFLAGS) -c MySamBam.cpp
+	$(CXX) $(CXXFLAGS) -c MySamBam.cpp
 
 main.o: main.cpp  
-	$(CC) $(CFLAGS) -c main.cpp
+	$(CXX) $(CXXFLAGS) -c main.cpp
 
 fastSVseq: main.o Insertion.o Caller.o Reference.o MySamBam.o
-	$(CC) $(CFLAGS) -o SVseq2_2 Insertion.o Caller.o main.o Reference.o MySamBam.o \
-	-lbam -lz -lm
+	$(CXX) $(CXXFLAGS) -o SVseq2_2 Insertion.o Caller.o main.o Reference.o MySamBam.o \
+	-lbam -lz -lm -lpthread $(LDFLAGS)
+
+clean:
+	rm -rf *.o SVseq2_2
