From b7d93726bbc46de7d3774bdc4e466101e15900e3 Mon Sep 17 00:00:00 2001
From: Ilya Shlyakhter <ilya_shl@alum.mit.edu>
Date: Fri, 16 Aug 2019 16:21:01 -0400
Subject: [PATCH] change verbosity

---
 spades_compile.sh | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/spades_compile.sh b/spades_compile.sh
index fe33034..5e0795a 100755
--- a/spades_compile.sh
+++ b/spades_compile.sh
@@ -7,7 +7,7 @@
 # See file LICENSE for details.
 ############################################################################
 
-set -e
+set -e -x
 
 if [ "x$PREFIX" = "x" ]; then
   PREFIX=`pwd`
@@ -20,6 +20,6 @@ mkdir -p "$BASEDIR/$BUILD_DIR"
 set -e
 cd "$BASEDIR/$BUILD_DIR"
 cmake -G "Unix Makefiles" -DCMAKE_INSTALL_PREFIX="$PREFIX" $* "$BASEDIR/src"
-make -j 8
+make -j1 VERBOSE=1
 make install
 cd $PREFIX
-- 
2.20.1

