diff --git a/src/Makefile b/src/Makefile
index ea8d562..a702e36 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -477,8 +477,8 @@ endef
 
 ifeq (${OSTYPE}, Darwin)
   $(foreach suffix,mp-9 9  mp-8 8  mp-7 7,$(eval $(call TEST_COMPILER,${suffix})))
-
-  ifneq ($(shell echo `$(CXX) --version 2>&1 | grep -c clang`), 0)
+  # Let's unconditionally try conda-forge's Clang instead of aborting early.
+  ifneq (0, 0)
      CPATH := $(shell echo `which $(CXX)`)
      CLANG := $(shell echo `$(CXX) --version 2>&1 | grep clang`)
      space := 
diff --git a/src/utility/src/Makefile b/src/utility/src/Makefile
index 7cf3f67..ff33d41 100644
--- a/src/utility/src/Makefile
+++ b/src/utility/src/Makefile
@@ -523,8 +523,8 @@ ifeq (${OSTYPE}, Darwin)
       CXX := $(CXX9)
     endif
   endif
-
-  ifneq ($(shell echo `$(CXX) --version 2>&1 | grep -c clang`), 0)
+  # Let's unconditionally try conda-forge's Clang instead of aborting early.
+  ifneq (0, 0)
      CPATH := $(shell echo `which $(CXX)`)
      CLANG := $(shell echo `$(CXX) --version 2>&1 | grep clang`)
      space := 
