From 78c6f10a9e84432c8bf7b999bb30803c0e9a0852 Mon Sep 17 00:00:00 2001
From: Ed Harry <e.f.harry@gmail.com>
Date: Tue, 21 Jul 2020 10:35:57 +0100
Subject: [PATCH] osx build

---
 src/Makefile | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/src/Makefile b/src/Makefile
index 23647a3..60bd671 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -400,10 +400,10 @@ endif
 
 
 ifeq (${OSTYPE}, Darwin)
-  CC       ?= gcc
-  CXX      ?= g++
+  CC = clang
+  CXX = clang++
 
-  CLANG = $(shell echo `${CXX} --version 2>&1 | grep -c clang`)
+  CLANG = 0
 
   ifeq ($(CLANG), 0)
      CXXFLAGS += -D_GLIBCXX_PARALLEL -fopenmp
-- 
2.27.0

