--- a/src/version.h
+++ b/src/version.h
@@ -1,3 +1,10 @@
+// Updated at April 27, 2025
+// Updated by Jiaming Yu
+// version 3.7.3 - 3.7.4
+// Bioinformatics Group, College of Computer Science and Technology, Qingdao University
+// Code by: Yu Zhang, Jiaming Yu, Yongxiang Huang, Minan Wang
+// Define the version of Parallel-Meta Suite
+
 // Updated at July 3, 2024
 // Updated by Haobo Shi
 // version 3.7.2 - 3.7.3
@@ -8,6 +15,6 @@
 #ifndef _VERSION_H
 #define _VERSION_H
 
-#define Version "3.7.3"
+#define Version "3.7.4"
 
 #endif
