diff -r d8c075170bb5 published-packages/CEAS/lib/inout.py
--- a/published-packages/CEAS/lib/inout.py	Mon Sep 29 10:11:06 2014 -0500
+++ b/published-packages/CEAS/lib/inout.py	Tue Feb 27 14:33:54 2018 +0000
@@ -61,7 +61,7 @@ try:
     import MySQLdb
 except ImportError:
     MYSQL=False
-    warnings.warn("sqlite3 is used instead of MySQLdb because MySQLdb is not installed")
+    #warnings.warn("sqlite3 is used instead of MySQLdb because MySQLdb is not installed")
 
 # ------------------------------------
 # some literals
