diff --git a/src/cs/cstranslate_app.h b/src/cs/cstranslate_app.h
index 3f0df410..c2cb3d42 100644
--- a/src/cs/cstranslate_app.h
+++ b/src/cs/cstranslate_app.h
@@ -35,7 +35,9 @@
 #include "ext/fmemopen.h"
 #include "context_data.lib.h"
 //include "context_data.crf.h"
+namespace cs {
 #include "cs219.lib.h"
+}
 
 #ifdef OPENMP
 #include <omp.h>
@@ -473,7 +475,7 @@ namespace cs {
     void SetupAbstractStateEngine() {
       FILE *fin;
       if (opts_.alphabetfile == "internal") {
-        fin = fmemopen((void*)cs219_lib, cs219_lib_len, "r");
+        fin = fmemopen((void*)cs::cs219_lib, cs::cs219_lib_len, "r");
       } else {
         fin = fopen(opts_.alphabetfile.c_str(), "r");
       }
