diff --git a/lib/ffindex/src/ext/fmemopen.h b/lib/ffindex/src/ext/fmemopen.h
index bcf3f90..06deb86 100644
--- a/lib/ffindex/src/ext/fmemopen.h
+++ b/lib/ffindex/src/ext/fmemopen.h
@@ -1,5 +1,10 @@
 #ifndef __FMEMOPEN_H__
 #define __FMEMOPEN_H__
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
@@ -13,4 +18,9 @@ typedef struct fmem fmem_t;

 FILE *fmemopen(void *, size_t, const char *);

+#ifdef __cplusplus
+}
+#endif
+
 #endif /* __FMEMOPEN_H__ */
+
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 12d4fa5..dbdaeaf 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -118,9 +118,6 @@ target_link_libraries(HH_OBJECTS
         hhviterbialgorithm_with_celloff
         hhviterbialgorithm_and_ss
         hhviterbialgorithm_with_celloff_and_ss)
-if (NOT HAVE_FMEMOPEN)
-target_link_libraries(HH_OBJECTS ext)
-endif()

 add_executable(hhblits hhblits_app.cpp)
 target_link_libraries(hhblits HH_OBJECTS)
