From 6f445606ad44fe0199d308fda59efdf4b2ab62e5 Mon Sep 17 00:00:00 2001
From: Fabien Pertuy <Fabien.Pertuy@Sanofi.com>
Date: Tue, 4 May 2021 15:00:49 +0000
Subject: [PATCH] PATCH

---
 src/Makefile      | 2 +-
 src/Utilities.hpp | 1 +
 2 files changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/Makefile b/src/Makefile
index 091270e..3094582 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -1,6 +1,6 @@
 CXX = g++
 
-CXXFLAGS = -O3 -DNDEBUG -W -pipe -Wundef -Winline --param large-function-growth=100000 -Wall
+CXXFLAGS = -O3 -DNDEBUG -W -pipe -Wundef -Winline --param large-function-growth=100000 -Wall -fpermissive
 LINKFLAGS = -lm
 GDLINKFLAGS = -lgd -lpng
 
diff --git a/src/Utilities.hpp b/src/Utilities.hpp
index 7e05001..2c19684 100644
--- a/src/Utilities.hpp
+++ b/src/Utilities.hpp
@@ -15,6 +15,7 @@
 #include <fstream>
 #include <iomanip>
 #include <iostream>
+#include <limits.h>
 #include <numeric>
 #include <map>
 #include <queue>
-- 
2.30.2

