From 841a65c637fe0cd118f6d31605ba5108178b1696 Mon Sep 17 00:00:00 2001
From: Fabien Pertuy <Fabien.Pertuy@Sanofi.com>
Date: Wed, 5 May 2021 07:19:40 +0000
Subject: [PATCH] PATCH legacy

---
 src/Makefile      | 4 +---
 src/Utilities.hpp | 1 +
 2 files changed, 2 insertions(+), 3 deletions(-)

diff --git a/src/Makefile b/src/Makefile
index 091270e..9ae8f88 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -1,6 +1,4 @@
-CXX = g++
-
-CXXFLAGS = -O3 -DNDEBUG -W -pipe -Wundef -Winline --param large-function-growth=100000 -Wall
+CXXFLAGS = -O3 -DNDEBUG -W -pipe -Wundef -Winline --param large-function-growth=100000 -Wall -fpermissive
 LINKFLAGS = -lm
 GDLINKFLAGS = -lgd -lpng
 
diff --git a/src/Utilities.hpp b/src/Utilities.hpp
index 7e05001..2c19684 100644
--- a/src/Utilities.hpp
+++ b/src/Utilities.hpp
@@ -15,6 +15,7 @@
 #include <fstream>
 #include <iomanip>
 #include <iostream>
+#include <limits.h>
 #include <numeric>
 #include <map>
 #include <queue>
-- 
2.30.2

