diff --git a/install_deps.sh b/install_deps.sh
index ae75f62..1095d09 100755
--- a/install_deps.sh
+++ b/install_deps.sh
@@ -9,30 +9,30 @@ wget -c https://zlib.net/fossils/zlib-1.2.13.tar.gz
 tar xvzf zlib-1.2.13.tar.gz
 pushd zlib-1.2.13
 ./configure --prefix `pwd`/..
-make
+make CC=$CC -j$CPU_COUNT
 make install
 popd
 #############################################
 #
-if [ ! -e deps/samtools-0.1.19.tar.bz2  ]; then
+if [[ ! -e deps/samtools-0.1.19.tar.bz2  ]]; then
     wget -c https://sourceforge.net/projects/samtools/files/samtools/0.1.19/samtools-0.1.19.tar.bz2/download -O deps/samtools-0.1.19.tar.bz2
 fi
-tar jxvf deps/samtools-0.1.19.tar.bz2 
-pushd samtools-0.1.19
-make
-make razip
+tar jxvf deps/samtools-0.1.19.tar.bz2
+pushd samtools-0.1.19 && patch -p1 < $RECIPE_DIR/0003-samtools-Makefile.patch
+make CC=$CC -j$CPU_COUNT
+make razip CC=$CC -j$CPU_COUNT
 popd
 
 ###############################################
 # Downloads were failing often so keep a local copy to speedup the installation
-echo "Installing samtools 1.x" 
-if [ ! -e deps/samtools-1.9.tar.bz2 ] ; then
+echo "Installing samtools 1.x"
+if [[ ! -e deps/samtools-1.9.tar.bz2 ]]; then
     wget -c https://sourceforge.net/projects/samtools/files/samtools/1.9/samtools-1.9.tar.bz2/download -O deps/samtools-1.9.tar.bz2
 fi
 
 tar xvjf deps/samtools-1.9.tar.bz2
 pushd samtools-1.9
-./configure  prefix=$TOPLEVEL_DIR
-make -j $J
+./configure prefix=$TOPLEVEL_DIR
+make -j$J
 #prefix=$TOPLEVEL_DIR
 make install prefix:=$TOPLEVEL_DIR
