*** dammit/dependencies.py	2016-04-20 21:25:51.000000000 +0200
--- ../dammit-0.3_patch/dammit/dependencies.py	2016-05-19 18:58:53.004997771 +0200
*************** def check_blast(logger):
*** 118,124 ****
  
  
  def check_busco(logger):
!     busco = which('BUSCO_v1.1b1.py')
      if busco is None:
          return False, 'Not found on $PATH'
      else:
--- 118,124 ----
  
  
  def check_busco(logger):
!     busco = which('busco')
      if busco is None:
          return False, 'Not found on $PATH'
      else:
*** dammit/tests/test_databases.py	2016-04-20 21:25:51.000000000 +0200
--- ../dammit-0.3_patch/dammit/tests/test_databases.py	2016-05-19 18:59:12.232997459 +0200
*************** execs = ['hmmscan',
*** 27,33 ****
           'blastx',
           'tblastn',
           'makeblastdb',
!          'BUSCO_v1.1b1.py',
           'TransDecoder.LongOrfs',
           'TransDecoder.Predict',
           'lastal',
--- 27,33 ----
           'blastx',
           'tblastn',
           'makeblastdb',
!          'busco',
           'TransDecoder.LongOrfs',
           'TransDecoder.Predict',
           'lastal',
*** dammit/tests/test_dependencies.py	2016-04-20 21:25:51.000000000 +0200
--- ../dammit-0.3_patch/dammit/tests/test_dependencies.py	2016-05-19 18:59:36.728997062 +0200
*************** execs = ['hmmscan',
*** 27,33 ****
           'blastx',
           'tblastn',
           'makeblastdb',
!          'BUSCO_v1.1b1.py',
           'TransDecoder.LongOrfs',
           'TransDecoder.Predict',
           'lastal',
--- 27,33 ----
           'blastx',
           'tblastn',
           'makeblastdb',
!          'busco',
           'TransDecoder.LongOrfs',
           'TransDecoder.Predict',
           'lastal',
*** dammit/tasks.py	2016-04-20 21:25:51.000000000 +0200
--- ../dammit-0.3_patch/dammit/tasks.py	2016-05-19 19:00:24.960996280 +0200
*************** def get_busco_task(input_filename, outpu
*** 388,398 ****
      name = 'busco:' + os.path.basename(input_filename) + '-' + os.path.basename(busco_db_dir)
  
      assert input_type in ['genome', 'OGS', 'trans']
!     exc = which('BUSCO_v1.1b1.py')
      # BUSCO chokes on file paths as output names
      output_name = os.path.basename(output_name)
  
!     cmd = 'python3 {exc} -in {input_filename} -f -o {output_name} -l {busco_db_dir} '\
              '-m {input_type} -c {n_threads}'.format(**locals())
  
      return {'name': name,
--- 388,398 ----
      name = 'busco:' + os.path.basename(input_filename) + '-' + os.path.basename(busco_db_dir)
  
      assert input_type in ['genome', 'OGS', 'trans']
!     exc = which('busco')
      # BUSCO chokes on file paths as output names
      output_name = os.path.basename(output_name)
  
!     cmd = 'python {exc} -in {input_filename} -f -o {output_name} -l {busco_db_dir} '\
              '-m {input_type} -c {n_threads}'.format(**locals())
  
      return {'name': name,
