From 2e6c306c4b851751581c599714707a5be3e19e9e Mon Sep 17 00:00:00 2001
From: Mervin Fansler <mmfansler@gmail.com>
Date: Sun, 16 Jul 2023 15:33:11 -0400
Subject: [PATCH] support strict R headers

---
 src/utils.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/utils.cpp b/src/utils.cpp
index 99e199e..c8a64e1 100644
--- a/src/utils.cpp
+++ b/src/utils.cpp
@@ -115,7 +115,7 @@ SEXP ptr_neq_constraints(SEXP x, SEXP constraints, SEXP ratio, SEXP value){
 	double* p_x = ( isNull(x) ? NULL : NUMERIC_POINTER(x) );
 	double d_ratio = isNull(ratio) ? 0 : *NUMERIC_POINTER(ratio);
 	double* p_value = ( isNull(value) ? NULL : NUMERIC_POINTER(value) );
-	double eps = sqrt(DOUBLE_EPS);
+	double eps = sqrt(DBL_EPSILON);
 
 	// get dimensions
 	int ncol = isNull(GET_DIM(x)) ? 1 : INTEGER(GET_DIM(x))[1];