diff --git a/pscan.cpp b/pscan.cpp
index 701447b..ac000df 100644
--- a/pscan.cpp
+++ b/pscan.cpp
@@ -483,6 +483,7 @@ int main(int argc, char **argv)
 
 void display_help()
 {
+	/*
 	ifstream in("HELP.txt");
 
 	if(!in)
@@ -497,8 +498,8 @@ void display_help()
 		cerr << line << endl;
 
 	in.close();
-
-/*	cerr << endl << "SYNOPSIS" << endl
+	*/
+	cerr << endl << "SYNOPSIS" << endl
 	     << "\tpscan -q multifastafile -p multifastafile [options]" << endl
 	     << "\tpscan -p multifastafile [options]" << endl
 	     << "\tpscan -q multifastafile -M matrixfile [options]" << endl << endl
@@ -575,7 +576,7 @@ void display_help()
 	<< "5)\tpscan -q human_nfy_targets.fasta -l matrixfile.wil -N MATRIX1" << endl << endl
 
         << "\tThis command is like Example #3 but it will use the matrix called \"MATRIX1\" contained in the \"matrixfile.wil\" file." << endl << endl;
-*/
+
 	exit(1);
 }
 
