diff --git a/include/common_types.h b/include/common_types.h
index 2e19e54..ec951ab 100644
--- a/include/common_types.h
+++ b/include/common_types.h
@@ -4,6 +4,7 @@
 #include <unordered_set>
 #include <unordered_map>
 #include <vector>
+#include <cstdint>
 
 namespace mantis {
   using KmerHash = uint64_t;
diff --git a/include/kmer.h b/include/kmer.h
index 0922486..eae4eaf 100644
--- a/include/kmer.h
+++ b/include/kmer.h
@@ -21,6 +21,7 @@
 
 #include <stdio.h>
 #include <string>
+#include <cstdint>
 
 #include "common_types.h"
 #include "nonstd/optional.hpp"
diff --git a/include/squeakrconfig.h b/include/squeakrconfig.h
index b830e15..4e9c882 100644
--- a/include/squeakrconfig.h
+++ b/include/squeakrconfig.h
@@ -12,6 +12,7 @@
 #define _SQUEAKR_CONFIG_H_
 
 #include <string>
+#include <cstdint>
 #include <stdio.h>
 
 namespace squeakr {
diff --git a/src/canonicalKmer.cc b/src/canonicalKmer.cc
index fe2fdef..e02bfc0 100644
--- a/src/canonicalKmer.cc
+++ b/src/canonicalKmer.cc
@@ -3,6 +3,9 @@
 //
 
 #include <assert.h>
+#include <cstdint>
+#include <cstdlib>
+#include <string>
 
 #include "canonicalKmer.h"
 
diff --git a/src/squeakrconfig.cc b/src/squeakrconfig.cc
index 3be2693..6817f92 100644
--- a/src/squeakrconfig.cc
+++ b/src/squeakrconfig.cc
@@ -8,6 +8,7 @@
  */
 
 #include <fstream>      // std::ifstream
+#include <cstdint>
 
 #include "squeakrconfig.h"
 
