diff --git a/train/makefile b/train/makefile
index 56eaa13..5f123e3 100644
--- a/train/makefile
+++ b/train/makefile
@@ -2,11 +2,11 @@
 
 # C compiler
 
-C       = gcc 
-CC      = g++ 
+C       = $(CC)
+CC      = $(CXX) 
 #CFLAGS  = -O1 ${SEARCHDIRS}
 #CFLAGS  = -O3 -g -Wall
-CFLAGS = -Wall -g
+CFLAGS = -O3 -Wall -g -Wno-unused-const-variable -Wno-unused-variable -Wno-comment -Wno-deprecated-declarations -Wno-unused-but-set-variable
 LIBS    = -lm
 
 MAKEFILE= makefile
