From 66bf194fd4d61855216bf6ec779304a1d55838ba Mon Sep 17 00:00:00 2001
From: "Alicia A. Evans" <alicia.evans@nih.gov>
Date: Wed, 16 Jul 2025 12:25:36 -0400
Subject: [PATCH] configure

---
 configure | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/configure b/configure
index 2df4a51..5d3798f 100755
--- a/configure
+++ b/configure
@@ -81,8 +81,8 @@ EOF
 	  (cd src/SYMPHONY && \
 		  ./configure \
 			CC="$CC" CXX="$CXX" \
-			CFLAGS="-w -g -O2" \
-			CXXFLAGS="-w -g -O2" \
+			CFLAGS="$CFLAGS -w -g -O2" \
+			CXXFLAGS="$CXXFLAGS -w -g -O2" \
 		  --enable-static --disable-shared --with-pic \
 		  --with-application=no --disable-dependency-tracking \
 		  --disable-zlib --disable-bzlib "${OMP_FLAG}" \
-- 
2.49.0

