diff --git a/src/SYMPHONY/SYMPHONY/configure b/src/SYMPHONY/SYMPHONY/configure
index 6047fce..4aacc6a 100755
--- a/src/SYMPHONY/SYMPHONY/configure
+++ b/src/SYMPHONY/SYMPHONY/configure
@@ -28996,7 +28996,7 @@ case "$CXX" in
   ;;
   cl* | *\cl* | */cl*)
      CLLCFLAGS=$CXXFLAGS
-     CXXFLAGS="$CXXFLAGS -TP"
+     CXXFLAGS="$CXXFLAGS"
      if test $sym_openmp = true; then
          CXXFLAGS="$CXXFLAGS /openmp"
      fi
