--- src/count-kmers.cpp.org     2020-04-05 00:23:45.000000000 -0500
+++ src/count-kmers.cpp 2020-04-05 00:24:29.000000000 -0500
@@ -8,7 +8,7 @@
 using namespace std;
 using namespace HASHMAP;

-typedef hash_map<Mer_t, unsigned int, hash<unsigned long> > MerTable_t;
+typedef hash_map<Mer_t, unsigned int, HASHMAP::hash<unsigned long> > MerTable_t;

 static void CountMers (const string & s, MerTable_t & mer_table);
 static void PrintMers(const MerTable_t & mer_table, int min_count);
--- src/count-qmers.cpp.org     2020-04-05 00:23:55.000000000 -0500
+++ src/count-qmers.cpp 2020-04-05 00:24:39.000000000 -0500
@@ -9,7 +9,7 @@
 using namespace std;
 using namespace HASHMAP;

-typedef hash_map<Mer_t, double, hash<unsigned long> > MerTable_t;
+typedef hash_map<Mer_t, double, HASHMAP::hash<unsigned long> > MerTable_t;

 static void CountMers (const string & s, const string & q, MerTable_t & mer_table);
 static void PrintMers(const MerTable_t & mer_table, int min_count);
