--- gsea.sh.orig	2022-11-30 15:23:38.000000000 +0000
+++ gsea.sh	2022-11-30 15:22:31.000000000 +0000
@@ -16,7 +16,7 @@
     java -version
 fi
 
-exec java -showversion --module-path="${prefix}/modules" -Xmx4g \
+exec java -showversion --module-path="${prefix}/modules" ${JVM_MEM_OPTS:-"-Xmx4g"} \
     @"${prefix}/gsea.args" \
     --patch-module="jide.common=${prefix}/lib/jide-components-3.7.4.jar:${prefix}/lib/jide-dock-3.7.4.jar:${prefix}/lib/jide-grids-3.7.4.jar" \
     -Djava.util.logging.config.file="${prefix}/logging.properties" \
