--- gsea-cli.sh.orig	2022-11-30 15:20:29.000000000 +0000
+++ gsea-cli.sh	2022-11-30 15:21:15.000000000 +0000
@@ -22,7 +22,7 @@
     PREFS_PROP=
 fi;
 
-exec java --module-path="${prefix}/modules" -Xmx4g \
+exec java --module-path="${prefix}/modules" ${JVM_MEM_OPTS:-"-Xmx4g"} \
     -Djava.awt.headless=true $PREFS_PROP \
     -Djava.util.logging.config.file="${prefix}/logging.properties" \
     @"${prefix}/gsea.args" \
