--- gimmemotifs/rocmetrics.py	2019-12-05 12:48:41.162560495 +0100
+++ gimmemotifs/rocmetrics.fix.py	2021-02-02 08:26:53.020308943 +0100
@@ -10,13 +10,13 @@
 """
 
 # External imports
-from scipy.stats import stats, scoreatpercentile, kstest, fisher_exact
 from sklearn.metrics import (
     precision_recall_curve,
     roc_auc_score,
     roc_curve,
     average_precision_score,
 )
+from scipy.stats import stats, scoreatpercentile, kstest, fisher_exact
 import numpy as np
 
 __all__ = [
