--- gimmemotifs/utils.py	2021-02-02 09:42:05.388538632 +0100
+++ gimmemotifs/utils.fix.py	2021-02-02 09:41:52.312091399 +0100
@@ -6,6 +6,7 @@
 
 """ Odds and ends that for which I didn't (yet) find another place """
 # Python imports
+from gimmemotifs.rocmetrics import ks_pvalue
 import os
 import re
 import sys
@@ -32,7 +33,6 @@
 # gimme imports
 from gimmemotifs.fasta import Fasta
 from gimmemotifs.plot import plot_histogram
-from gimmemotifs.rocmetrics import ks_pvalue
 from gimmemotifs.config import MotifConfig
 
 
