--- compile_externals.py	2019-12-05 13:00:52.183937725 +0100
+++ compile_externals.fix.py	2019-12-05 13:01:05.988415174 +0100
@@ -6,8 +6,8 @@
 
 def compile_simple(name, src_dir="src"):
     gcc = "gcc"
-    if "GCC" in os.environ and os.environ["GCC"]:
-        gcc = os.environ["GCC"]
+    if "CC" in os.environ and os.environ["CC"]:
+        gcc = os.environ["CC"]
 
     path = os.path.join(src_dir, "%s" % name)
 
