From 2caff3fbc92c091d9b154fa247889529a44f90ce Mon Sep 17 00:00:00 2001
From: Justin Lecher <jlec@gentoo.org>
Date: Sun, 19 Feb 2017 18:49:48 +0000
Subject: [PATCH] Fix for building with > gcc-5

Fixes #2

Signed-off-by: Justin Lecher <jlec@gentoo.org>
---
 src/SW.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/SW.cpp b/src/SW.cpp
index 57d2773..3a9c225 100644
--- a/src/SW.cpp
+++ b/src/SW.cpp
@@ -796,7 +796,7 @@ int findAGSBackwardExtensionCarefully(QueryState_t * QS,  ROFF rOff, char * qStr
 ////////////////////////
 
 template <BOOL banded, BOOL extension, BOOL reverse, BOOL XCutoff>
-int findAffineGapScore(QueryState_t * QS, char * qStr, SUINT qLen, char * rStr, SUINT rLen, EditOpList_t * list, SUINT * addedQLen = NULL, SUINT * addedRLen = NULL)
+int findAffineGapScore(QueryState_t * QS, char * qStr, SUINT qLen, char * rStr, SUINT rLen, EditOpList_t * list, SUINT * addedQLen, SUINT * addedRLen)
 {
     // As we will go row by row, then column by column, we do not need to store all of the V, E, and F arrays.
     // Instead we will store only the previous row and column for each.
