diff --git a/pyproject.toml b/pyproject.toml
index 1a03623..36102d0 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -49,7 +49,7 @@ HICTK_BUILD_EXAMPLES = "OFF"
 HICTK_BUILD_BENCHMARKS = "OFF"
 HICTK_BUILD_TOOLS = "OFF"
 HICTK_ENABLE_GIT_VERSION_TRACKING = "OFF"
-BUILD_SHARED_LIBS = "OFF"
+BUILD_SHARED_LIBS = "ON"
 CONAN_INSTALL_ARGS = "--settings=compiler.cppstd=17;--build=missing;--update;--options=*/*:shared=False"
 
 [tool.setuptools_scm]
