diff --git a/pyproject.toml b/pyproject.toml
index 5ded5b9..fe8d305 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -74,7 +74,7 @@ wheel.expand-macos-universal-tags = true
 cmake.build-type = "Release"
 
 [tool.scikit-build.cmake.define]
-BUILD_SHARED_LIBS = "OFF"
+BUILD_SHARED_LIBS = "ON"
 CMAKE_OSX_DEPLOYMENT_TARGET = "10.15"
 CMAKE_PROJECT_TOP_LEVEL_INCLUDES = { env="HICTKPY_PROJECT_TOP_LEVEL_INCLUDES", default="cmake/conan_provider.cmake" }
 CONAN_INSTALL_ARGS = { env="HICTKPY_CONAN_INSTALL_ARGS", default="--settings=compiler.cppstd=17;--settings=build_type=Release;--build=missing;--update;--options=*/*:shared=False" }
