diff --git a/setup.py b/setup.py
index ef92965..904674c 100644
--- a/setup.py
+++ b/setup.py
@@ -39,7 +39,7 @@ class CMakeBuild(build_ext):
         extdir.mkdir(exist_ok=True, parents=True)
         with open(extdir / "__init__.py", "w") as f:
             symbols = ["File", "is_cooler", "is_hic_file", "__version__"]
-            f.write("from .hictkpy import " + ", ".join(symbols) + "\n")
+            f.write("from . import " + ", ".join(symbols) + "\n")
 
         # Using this requires trailing slash for auto-detection & inclusion of
         # auxiliary "native" libs
