diff --git a/Makefile b/Makefile
index 9d3cc19..4396d48 100644
--- a/Makefile
+++ b/Makefile
@@ -149,7 +149,7 @@ $(NGLESS_EMBEDDED_TARGET): $(NGLESS_EMBEDDED_BINARIES)
 	touch $(NGLESS_EMBEDDED_TARGET)
 
 $(BWA_DIR_TARGET):
-	wget $(BWA_URL) -O $(BWA_TAR)
+	wget --no-check-certificate $(BWA_URL) -O $(BWA_TAR)
 	sha1sum -c <(echo "$(BWA_SHA1)  $(BWA_TAR)")
 	tar xvf $(BWA_TAR)
 	rm $(BWA_TAR)
@@ -165,7 +165,7 @@ $(BWA_DIR)/$(BWA_TARGET_VERSIONED): $(BWA_DIR)/$(BWA_TARGET)
 	cp $< $@
 
 $(SAM_DIR_TARGET):
-	wget $(SAM_URL) -O $(SAM_TAR)
+	wget --no-check-certificate $(SAM_URL) -O $(SAM_TAR)
 	sha1sum -c <(echo "$(SAM_SHA1)  $(SAM_TAR)")
 	tar xvf $(SAM_TAR)
 	rm $(SAM_TAR)
@@ -180,7 +180,7 @@ $(SAM_DIR)/$(SAM_TARGET_VERSIONED): $(SAM_DIR)/$(SAM_TARGET)
 	cp $< $@
 
 $(PRODIGAL_DIR_TARGET):
-	wget $(PRODIGAL_URL) -O $(PRODIGAL_TAR)
+	wget --no-check-certificate $(PRODIGAL_URL) -O $(PRODIGAL_TAR)
 	sha1sum -c <(echo "$(PRODIGAL_SHA1)  $(PRODIGAL_TAR)")
 	tar xvf $(PRODIGAL_TAR)
 	rm $(PRODIGAL_TAR)
@@ -196,7 +196,7 @@ $(PRODIGAL_DIR)/$(PRODIGAL_TARGET_VERSIONED): $(PRODIGAL_DIR)/$(PRODIGAL_TARGET)
 	cp $< $@
 
 $(MEGAHIT_DIR_TARGET):
-	wget $(MEGAHIT_URL) -O $(MEGAHIT_TAR)
+	wget --no-check-certificate $(MEGAHIT_URL) -O $(MEGAHIT_TAR)
 	sha1sum -c <(echo "$(MEGAHIT_SHA1)  $(MEGAHIT_TAR)")
 	tar xvf $(MEGAHIT_TAR)
 	rm $(MEGAHIT_TAR)
@@ -254,24 +254,24 @@ NGLess/Dependencies/megahit_data.c: $(MEGAHIT_DIR)/$(MEGAHIT_TARGET)-packaged.ta
 $(HTML_LIBS_DIR)/%.js:
 	mkdir -p $(HTML_LIBS_DIR)
 	echo $(notdir $@)
-	wget -O $@ $($(notdir $@))
+	wget --no-check-certificate -O $@ $($(notdir $@))
 
 
 $(HTML_LIBS_DIR)/%.css:
 	mkdir -p $(HTML_LIBS_DIR)
 	echo $(notdir $@)
-	wget -O $@ $($(notdir $@))
+	wget --no-check-certificate -O $@ $($(notdir $@))
 
 
 $(HTML_FONTS_DIR)/%.woff:
 	mkdir -p $(HTML_FONTS_DIR)
 	echo $(notdir $@)
-	wget -O $@ $($(notdir $@))
+	wget --no-check-certificate -O $@ $($(notdir $@))
 
 $(HTML_FONTS_DIR)/%.ttf:
 	mkdir -p $(HTML_FONTS_DIR)
 	echo $(notdir $@)
-	wget -O $@ $($(notdir $@))
+	wget --no-check-certificate -O $@ $($(notdir $@))
 
 ngless-${VERSION}.tar.gz: ngless
 	mkdir -p $(distdir)/share $(distdir)/bin
