From b03aa957e67c28493a04912d3858fd221ae115f1 Mon Sep 17 00:00:00 2001
From: Marcel Bargull <marcel.bargull@udo.edu>
Date: Tue, 28 Jul 2020 17:38:47 +0200
Subject: [PATCH] Remove "docs" from distribution

---
 setup.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/setup.py b/setup.py
index 0dce699..1c2fe62 100644
--- a/setup.py
+++ b/setup.py
@@ -68,7 +68,7 @@
       long_description=long_description,
       long_description_content_type="text/markdown",
       url='https://github.com/goeckslab/Galaxy-ML/',
-      packages=find_packages(exclude=['tests*', 'test-data*', 'tools*']),
+      packages=find_packages(exclude=['docs', 'tests*', 'test-data*', 'tools*']),
       package_data={
           '': ['README.md',
                'requirements.txt']},
