diff --git a/setup.py b/setup.py
index bcbb306..71a7c03 100644
--- a/setup.py
+++ b/setup.py
@@ -3,6 +3,7 @@
 import codecs  # To use a consistent encoding
 import os
 import re
+import sys
 
 import setuptools
 
@@ -31,12 +32,13 @@ INSTALL_REQUIRES = [
     'requests_file',
     'networkx',
     'pyparsing',
-    'configparser',
     'py2neo',
     'click',
     'sqlalchemy',
     'onto2nx',
 ]
+if sys.version_info < (3, ):
+    INSTALL_REQUIRES.append('configparser')
 EXTRAS_REQUIRE = {}
 TESTS_REQUIRE = [
     'tox',
