diff --git a/scripts/R_scripts/titanCNA.R b/scripts/R_scripts/titanCNA.R
index a021420..da3e99c 100755
--- a/scripts/R_scripts/titanCNA.R
+++ b/scripts/R_scripts/titanCNA.R
@@ -161,7 +161,7 @@ if (is.null(outfile)){
 	outfile <- paste0(outdir, "/", id, "_cluster", numClustersStr, ".titan.txt")
 }
 if (is.null(outparam)){
-	outparam <- gsub(".titan.txt", ".param.txt", outfile)
+	outparam <- gsub(".titan.txt", ".params.txt", outfile)
 }
 if (is.null(outseg)){
 	outseg <- gsub(".titan.txt", ".segs.txt", outfile)
diff --git a/scripts/TenX_scripts/titanCNA_v1.15.0_TenX.R b/scripts/TenX_scripts/titanCNA_v1.15.0_TenX.R
index d04d117..80fb0c0 100644
--- a/scripts/TenX_scripts/titanCNA_v1.15.0_TenX.R
+++ b/scripts/TenX_scripts/titanCNA_v1.15.0_TenX.R
@@ -125,7 +125,7 @@ if (is.null(outfile)){
 	outfile <- paste0(outdir, "/", id, "_cluster", numClustersStr, ".titan.txt")
 }
 if (is.null(outparam)){
-	outparam <- gsub(".titan.txt", ".param.txt", outfile)
+	outparam <- gsub(".titan.txt", ".params.txt", outfile)
 }
 if (is.null(outseg)){
 	outseg <- gsub(".titan.txt", ".segs.txt", outfile)
