diff --git a/scripts/freec2bed.pl b/scripts/freec2bed.pl
index 6b74c03..6a18239 100644
--- a/scripts/freec2bed.pl
+++ b/scripts/freec2bed.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl -w
+#!/usr/bin/env "perl -w"
 #translate *_ratio.txt (the out put of FREEC) into a BED track

 use strict;
diff --git a/scripts/freec2circos.pl b/scripts/freec2circos.pl
index d1569e7..464b053 100644
--- a/scripts/freec2circos.pl
+++ b/scripts/freec2circos.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl -w
+#!/usr/bin/env "perl -w"
 #translate "*_ratio.txt" (output of FREEC) into a Circos track

 use strict;
diff --git a/scripts/get_fasta_lengths.pl b/scripts/get_fasta_lengths.pl
index b45ce78..3d6eeeb 100644
--- a/scripts/get_fasta_lengths.pl
+++ b/scripts/get_fasta_lengths.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl -w
+#!/usr/bin/env "perl -w"
 #prints lengths of sequences from a multifasta file

 use strict;
diff --git a/scripts/vcf2snpFreec.pl b/scripts/vcf2snpFreec.pl
index 14465c9..c706b80 100644
--- a/scripts/vcf2snpFreec.pl
+++ b/scripts/vcf2snpFreec.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env "perl -w"
 #translate dbSNP vcf file into the format acceptable by FREEC (for dbSNP files)

 use strict;
