diff --git a/scripts/freec2bed.pl b/scripts/freec2bed.pl
index 6b74c03..a935177 100644
--- a/scripts/freec2bed.pl
+++ b/scripts/freec2bed.pl
@@ -1,6 +1,6 @@
-#!/usr/bin/env perl -w
+#!/usr/bin/env perl
 #translate *_ratio.txt (the out put of FREEC) into a BED track
-
+use warnings;
 use strict;

 my $usage = qq{
diff --git a/scripts/freec2circos.pl b/scripts/freec2circos.pl
index d1569e7..984baee 100644
--- a/scripts/freec2circos.pl
+++ b/scripts/freec2circos.pl
@@ -1,6 +1,6 @@
-#!/usr/bin/env perl -w
+#!/usr/bin/env perl
 #translate "*_ratio.txt" (output of FREEC) into a Circos track
-
+use warnings;
 use strict;

 my $usage = qq{
diff --git a/scripts/get_fasta_lengths.pl b/scripts/get_fasta_lengths.pl
index b45ce78..b7725c0 100644
--- a/scripts/get_fasta_lengths.pl
+++ b/scripts/get_fasta_lengths.pl
@@ -1,6 +1,6 @@
-#!/usr/bin/env perl -w
+#!/usr/bin/env perl
 #prints lengths of sequences from a multifasta file
-
+use warnings;
 use strict;

 my $Filename;
diff --git a/scripts/vcf2snpFreec.pl b/scripts/vcf2snpFreec.pl
index 14465c9..ecada03 100644
--- a/scripts/vcf2snpFreec.pl
+++ b/scripts/vcf2snpFreec.pl
@@ -1,6 +1,6 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl
 #translate dbSNP vcf file into the format acceptable by FREEC (for dbSNP files)
-
+use warnings;
 use strict;

 my $usage = qq{
