--- src/SeekSubclones.cpp
+++ src/SeekSubclones.cpp
@@ -80,6 +80,7 @@
         vector <int> breakpoints = samplecopynumber.getChrCopyNumberAt(index).getBreakPoints() ;
         int bpstart=0;
         int bpend=0;
+        if (breakpoints.empty()) {breakpoints.push_back(length);} // if empty, segfaults on breakpoints.back()
         if (breakpoints.back()<length) {breakpoints.push_back(length);}
         if (breakpoints.size()!=samplecopynumber.getChrCopyNumberAt(index).getMedianValues().size()) {
             cerr << "Problem with the number of breakpoints:" <<breakpoints.size()<< " != " << samplecopynumber.getChrCopyNumberAt(index).getMedianValues().size()<<"\n";
