diff --git a/bloom_filter.hpp b/bloom_filter.hpp
index 555b897..5158782 100644
--- a/bloom_filter.hpp
+++ b/bloom_filter.hpp
@@ -28,7 +28,7 @@
 #include <sdsl/bit_vectors.hpp>
 
 #include "MurmurHash3.hpp"
-#include "kmc_api/kmc_file.h"
+#include "kmc_file.h"
 
 // using namespace std;
 using namespace sdsl;
diff --git a/main.cpp b/main.cpp
index 7b7bff8..d7abd06 100644
--- a/main.cpp
+++ b/main.cpp
@@ -34,7 +34,7 @@
 #include <zlib.h>
 
 #include "hts_log.h"
-#include "kmc_api/kmc_file.h"
+#include "kmc_file.h"
 #include "kseq.h"
 #include "vcf.h"
 
