diff --git a/CMakeLists.txt b/CMakeLists.txt
index 5796d8b..9030264 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -32,9 +32,17 @@ else ()
     message(STATUS "ZSTD found ${ZSTD}")
 endif ()
 
+find_path(ZSTD_INCLUDE_DIR NAMES zstd.h PATHS ${CMAKE_SYSTEM_PREFIX_PATH})
+if (NOT ZSTD_INCLUDE_DIR)
+    message(FATAL_ERROR "zstd.h (part of ZSTD) required")
+else ()
+    message(STATUS "ZSTD includes found ${ZSTD_INCLUDE_DIR}")
+endif ()
+
 target_include_directories(malva-geno
        PRIVATE ${PROJECT_SOURCE_DIR}
        PRIVATE ${KMC_INCLUDE_DIR}
+       PRIVATE ${ZSTD_INCLUDE_DIR}
 )
 
 target_link_libraries(malva-geno
