--- ../isPcrSrc.orig/inc/common.mk      2005-12-22 05:13:30.000000000 +1100
+++ inc/common.mk       2019-08-10 16:08:29.345749714 +1000
@@ -8,9 +8,9 @@
 # add the follow to makefiles to enable stronger warning checks
 # HG_WARN = ${HG_WARN_ERR}
 ifeq (${OSTYPE},darwin)
-    HG_WARN_ERR = -DJK_WARN -Wall -Werror -Wno-unused-variable
+    HG_WARN_ERR = -DJK_WARN -Wall -Wno-unused-variable
 else
-    HG_WARN_ERR = -DJK_WARN -Wall -Werror
+    HG_WARN_ERR = -DJK_WARN -Wall
 endif

 ifeq (${SCRIPTS},)
