diff --git a/release/Linux/bin/applauncher b/release/Linux/bin/applauncher
index 5ac941f..e0a92c2 100755
--- a/release/Linux/bin/applauncher
+++ b/release/Linux/bin/applauncher
@@ -22,7 +22,7 @@ if [ -z "$BEAST" ]; then
 	cd "$saveddir"
 fi
 
-BEAST_LIB="$BEAST/lib"
+BEAST_LIB="${PREFIX}/share/${PKG_NAME}-${PKG_VERSION}-${PKG_BUILDNUM}/lib"
 
 if [ -z "$JAVA_HOME" ]; then
   JAVA=java
diff --git a/release/Linux/bin/beast b/release/Linux/bin/beast
index 73c83b0..3d147b6 100755
--- a/release/Linux/bin/beast
+++ b/release/Linux/bin/beast
@@ -22,7 +22,7 @@ if [ -z "$BEAST" ]; then
 	cd "$saveddir"
 fi
 
-BEAST_LIB="$BEAST/lib"
+BEAST_LIB="${PREFIX}/share/${PKG_NAME}-${PKG_VERSION}-${PKG_BUILDNUM}/lib"
 
 if [ -z "$JAVA_HOME" ]; then
   JAVA=java
@@ -38,15 +38,16 @@ if [ -n "$BEAGLE_LIB" ]; then
 	else
 		BEAST_EXTRA_LIBS=$BEAGLE_LIB
 	fi
+BEAST_EXTRA_LIBS=${BEAST_EXTRA_LIBS}:${PREFIX}/lib
 fi
 
 # Explicitly add /usr/local/lib to library search path to ensure 
 # beast continues to find beagle when installed here.  (This is
 # necessary due to known problems with certain JREs.)
 if [ -z "$LD_LIBRARY_PATH" ]; then
-    export LD_LIBRARY_PATH=/usr/local/lib
+    export LD_LIBRARY_PATH=${PREFIX}/lib
 else
-    export LD_LIBRARY_PATH="$LD_LIBRARY_PATH":/usr/local/lib
+    export LD_LIBRARY_PATH="$LD_LIBRARY_PATH":${PREFIX}/lib
 fi
 
 # use BEAST_EXTRA_LIBS variable to load BEAGLE and other libraries from non-default locations 
diff --git a/release/Linux/bin/beauti b/release/Linux/bin/beauti
index c8046d6..e30ee8b 100755
--- a/release/Linux/bin/beauti
+++ b/release/Linux/bin/beauti
@@ -22,7 +22,7 @@ if [ -z "$BEAST" ]; then
 	cd "$saveddir"
 fi
 
-BEAST_LIB="$BEAST/lib"
+BEAST_LIB="${PREFIX}/share/${PKG_NAME}-${PKG_VERSION}-${PKG_BUILDNUM}/lib"
 
 if [ -z "$JAVA_HOME" ]; then
   JAVA=java
diff --git a/release/Linux/bin/densitree b/release/Linux/bin/densitree
index 97f0ac9..c7f5178 100755
--- a/release/Linux/bin/densitree
+++ b/release/Linux/bin/densitree
@@ -22,7 +22,7 @@ if [ -z "$BEAST" ]; then
 	cd "$saveddir"
 fi
 
-BEAST_LIB="$BEAST/lib"
+BEAST_LIB="${PREFIX}/share/${PKG_NAME}-${PKG_VERSION}-${PKG_BUILDNUM}/lib"
 
 if [ -z "$JAVA_HOME" ]; then
   JAVA=java
diff --git a/release/Linux/bin/loganalyser b/release/Linux/bin/loganalyser
index f5dada1..7f2aa07 100755
--- a/release/Linux/bin/loganalyser
+++ b/release/Linux/bin/loganalyser
@@ -22,7 +22,7 @@ if [ -z "$BEAST" ]; then
 	cd "$saveddir"
 fi
 
-BEAST_LIB="$BEAST/lib"
+BEAST_LIB="${PREFIX}/share/${PKG_NAME}-${PKG_VERSION}-${PKG_BUILDNUM}/lib"
 
 if [ -z "$JAVA_HOME" ]; then
   JAVA=java
diff --git a/release/Linux/bin/logcombiner b/release/Linux/bin/logcombiner
index d227029..dc1e8b3 100755
--- a/release/Linux/bin/logcombiner
+++ b/release/Linux/bin/logcombiner
@@ -22,7 +22,7 @@ if [ -z "$BEAST" ]; then
 	cd "$saveddir"
 fi
 
-BEAST_LIB="$BEAST/lib"
+BEAST_LIB="${PREFIX}/share/${PKG_NAME}-${PKG_VERSION}-${PKG_BUILDNUM}/lib"
 
 if [ -z "$JAVA_HOME" ]; then
   JAVA=java
diff --git a/release/Linux/bin/packagemanager b/release/Linux/bin/packagemanager
index 5329e35..5bf7864 100755
--- a/release/Linux/bin/packagemanager
+++ b/release/Linux/bin/packagemanager
@@ -22,7 +22,7 @@ if [ -z "$BEAST" ]; then
 	cd "$saveddir"
 fi
 
-BEAST_LIB="$BEAST/lib"
+BEAST_LIB="${PREFIX}/share/${PKG_NAME}-${PKG_VERSION}-${PKG_BUILDNUM}/lib"
 
 if [ -z "$JAVA_HOME" ]; then
   JAVA=java
diff --git a/release/Linux/bin/treeannotator b/release/Linux/bin/treeannotator
index 0f178e0..2aa3d6d 100755
--- a/release/Linux/bin/treeannotator
+++ b/release/Linux/bin/treeannotator
@@ -22,7 +22,7 @@ if [ -z "$BEAST" ]; then
 	cd "$saveddir"
 fi
 
-BEAST_LIB="$BEAST/lib"
+BEAST_LIB="${PREFIX}/share/${PKG_NAME}-${PKG_VERSION}-${PKG_BUILDNUM}/lib"
 
 if [ -z "$JAVA_HOME" ]; then
   JAVA=java
