diff --git a/beast/bin/beast b/beast/bin/beast
index eb0fd47..ce9206f 100755
--- a/beast/bin/beast
+++ b/beast/bin/beast
@@ -33,6 +33,13 @@ fi
 
 CPU=`uname -m`
 
+# Sets BEAGLE_LIB to the location of conda libraries
+if [ -z "$BEAGLE_LIB" ]; then
+  BEAGLE_LIB="$BEAST/../../lib"
+else
+  BEAGLE_LIB="$BEAGLE_LIB:$BEAST/../../lib"
+fi
+
 # use BEAGLE_LIB if the BEAGLE library is not in a standard location
 if [ $CPU  = 'x86_64' ]; then
   if [ -n "$BEAGLE_LIB" ]; then
@@ -57,9 +64,9 @@ fi
 # necessary due to known problems with certain JREs.)
 if [ $CPU  = 'x86_64' ]; then
   if [ -z "$LD_LIBRARY_PATH" ]; then
-    export LD_LIBRARY_PATH="/usr/local/lib:$JAVA_HOME/lib/amd64"
+    export LD_LIBRARY_PATH="/usr/local/lib:$JAVA_HOME/lib/amd64:$BEAST_EXTRA_LIBS"
   else
-    export LD_LIBRARY_PATH="$LD_LIBRARY_PATH:/usr/local/lib:$JAVA_HOME/lib/amd64"
+    export LD_LIBRARY_PATH="$LD_LIBRARY_PATH:/usr/local/lib:$JAVA_HOME/lib/amd64:$BEAST_EXTRA_LIBS"
   fi
 else
   if [ -z "$LD_LIBRARY_PATH" ]; then
