--- src/beast/core/MCMC.java	2022-11-13 22:17:22.690664330 +0100
+++ src/beast/core/MCMC.java	2022-11-13 22:17:22.690664330 +0100
@@ -49,9 +49,9 @@
 //                "  evolutionary analysis. PLoS Computational Biology 10(4): e1003537"
 //        , year = 2014, firstAuthorSurname = "bouckaert",
 //        DOI="10.1371/journal.pcbi.1003537")
-@Citation(value="Bouckaert, Remco, Timothy G. Vaughan, Joëlle Barido-Sottani, Sebastián Duchêne, Mathieu Fourment, \n"
-		+ "Alexandra Gavryushkina, Joseph Heled, Graham Jones, Denise Kühnert, Nicola De Maio, Michael Matschiner, \n"
-        + "Fábio K. Mendes, Nicola F. Müller, Huw A. Ogilvie, Louis du Plessis, Alex Popinga, Andrew Rambaut, \n"
+@Citation(value="Bouckaert, Remco, Timothy G. Vaughan, Joelle Barido-Sottani, Sebastian Duchene, Mathieu Fourment, \n"
+		+ "Alexandra Gavryushkina, Joseph Heled, Graham Jones, Denise Kuehnert, Nicola De Maio, Michael Matschiner, \n"
+        + "Fabio K. Mendes, Nicola F. Mueller, Huw A. Ogilvie, Louis du Plessis, Alex Popinga, Andrew Rambaut, \n"
         + "David Rasmussen, Igor Siveroni, Marc A. Suchard, Chieh-Hsi Wu, Dong Xie, Chi Zhang, Tanja Stadler, \n"
         + "Alexei J. Drummond \n"
 		+ "  BEAST 2.5: An advanced software platform for Bayesian evolutionary analysis. \n"
