diff --git a/beast/bin/loganalyser b/beast/bin/loganalyser
index 0d372f4..4c4dd6f 100755
--- a/beast/bin/loganalyser
+++ b/beast/bin/loganalyser
@@ -31,6 +31,12 @@ else
   JAVA="$JAVA_HOME"/bin/java
 fi
 
+# Sets BEAGLE_LIB to the location of conda libraries
+if [ -z "$BEAGLE_LIB" ]; then
+  BEAGLE_LIB="$BEAST/../../lib"
+else
+  BEAGLE_LIB="$BEAGLE_LIB:$BEAST/../../lib"
+fi
 
 # use BEAGLE_LIB if the BEAGLE library is not in a standard location
 if [ -n "$BEAGLE_LIB" ]; then
