--- MANIFEST.in
+++ MANIFEST.in
@@ -4,7 +4,7 @@ include run_antismash.py
 include download_databases.py
 
 recursive-include antismash *.py
-recursive-include antismash *.cfg *.txt *.tsv *.csv
+recursive-include antismash *.info *.cfg *.txt *.tsv *.csv
 
 recursive-include antismash *.hmm  *.hmm2 *.hmm3 *.hmmdb *.h3f *.h3i *.h3m *.h3p
 recursive-include antismash *.pkl
@@ -26,3 +26,4 @@ recursive-include scripts *.sh
 
 prune antismash/*/*/test/
 prune antismash/*/*/*/test/
+prune antismash/output_modules/html/lib/Mac/
