--- antismash/antismash/output_modules/html/structure_drawer.py
+++ antismash/antismash/output_modules/html/structure_drawer.py
@@ -85,7 +85,7 @@ def generate_chemical_structure_preds(pksnrpsvars, seq_record, options):
                     smilesfile = open("genecluster" + str(geneclusternr) + ".smi", "w")
                     smilesfile.write(smiles_string)
                     smilesfile.close()
+                    depictstatus = "failed"
-                    depictstatus = depict_smile(geneclusternr, options.structuresfolder)
                 if depictstatus == "failed":
                     pksnrpsvars.failedstructures.append(geneclusternr)
         elif utils.get_cluster_type(genecluster) == "ectoine":
@@ -94,7 +94,7 @@ def generate_chemical_structure_preds(pksnrpsvars, seq_record, options):
                 smilesfile = open("genecluster" + str(geneclusternr) + ".smi", "w")
                 smilesfile.write(smiles_string)
                 smilesfile.close()
+                depictstatus = "failed"
-                depictstatus = depict_smile(geneclusternr, options.structuresfolder)
             if depictstatus == "failed":
                 pksnrpsvars.failedstructures.append(geneclusternr)
             elif genecluster in pksnrpsvars.failedstructures:
