--- CMakeLists.txt	2018-05-23 03:59:14.000000000 -0500
+++ CMakeLists.txt.new	2018-10-19 15:29:39.782104577 -0500
@@ -57,8 +57,8 @@
 endif(racon_build_tests)
 
 if (racon_build_wrapper)
-    set(racon_path ${PROJECT_BINARY_DIR}/bin/racon)
-    set(rampler_path ${PROJECT_BINARY_DIR}/vendor/rampler/bin/rampler)
+    set(racon_path $ENV{PREFIX}/bin/racon)
+    set(rampler_path $ENV{PREFIX}/bin/rampler)
     configure_file(${PROJECT_SOURCE_DIR}/scripts/racon_wrapper.py
         ${PROJECT_BINARY_DIR}/${CMAKE_FILES_DIRECTORY}/racon_wrapper)
     file(COPY ${PROJECT_BINARY_DIR}/${CMAKE_FILES_DIRECTORY}/racon_wrapper
