diff --git a/lib/Bio/Easel/SqFile.pm b/lib/Bio/Easel/SqFile.pm
index 6dee36a..50b26d0 100644
--- a/lib/Bio/Easel/SqFile.pm
+++ b/lib/Bio/Easel/SqFile.pm
@@ -56,9 +56,7 @@ BEGIN {
     $src_file = __FILE__;
     $src_file =~ s/\.pm/\.c/;
 
-    my $file = __FILE__;
-    ($easel_src_dir) = $file =~ /^(.*)\/blib/;
-    $easel_src_dir = File::Spec->catfile($easel_src_dir, 'src/easel');
+    $easel_src_dir = File::Spec->catfile( $ENV{BIO_EASEL_SHARE_DIR}, 'src/easel' );
 
     $typemaps = __FILE__;
     $typemaps =~ s/\.pm/\.typemap/;
