--- MUMmer3.23/src/kurtz/streesrc/addleafcount.c.old	2024-08-07 13:34:49
+++ MUMmer3.23/src/kurtz/streesrc/addleafcount.c	2024-08-07 13:48:50
@@ -13,6 +13,16 @@
 #include "spacedef.h"
 #include "arraydef.h"
 
+void getbranchinfostree(Suffixtree *stree,Uint whichinfo,
+                        Branchinfo *branchinfo,Bref btptr);
+
+Sint depthfirststree(Suffixtree *stree,Reference *startnode,
+                     Sint (*processleaf)(Uint,Bref,void *),
+                     BOOL (*processbranch1)(Bref,void *),
+                     Sint (*processbranch2)(Bref,void *),
+                     BOOL (*stoptraversal)(void *),
+                     void *stopinfo,void *info);
+
 typedef struct
 {
   Suffixtree *stree;                      // suffix tree info
