--- MUMmer3.23/src/kurtz/mm3src/findmaxmat.c.old	2024-08-07 13:55:45
+++ MUMmer3.23/src/kurtz/mm3src/findmaxmat.c	2024-08-07 14:03:39
@@ -14,7 +14,31 @@
 #include "debugdef.h"
 #include "spacedef.h"
 #include "maxmatdef.h"
+
+void linklocstree(Suffixtree *stree,Location *outloc,Location *inloc);
+
+SYMBOL *scanprefixstree(Suffixtree *stree,Location *outloc,
+                        Location *inloc,SYMBOL *left,
+                        SYMBOL *right,Uint rescanlength);
+
+SYMBOL *findprefixpathstree(Suffixtree *stree,
+                            ArrayPathinfo *path,
+                            Location *outloc,
+                            Location *inloc,
+                            SYMBOL *left,SYMBOL *right,
+                            Uint rescanlength);
 
+Sint depthfirststree(Suffixtree *stree,Reference *startnode,
+                     Sint (*processleaf)(Uint,Bref,void *),
+                     BOOL (*processbranch1)(Bref,void *),
+                     Sint (*processbranch2)(Bref,void *),
+                     BOOL (*stoptraversal)(void *),
+                     void *stopinfo,void *info);
+
+SYMBOL *scanprefixfromnodestree(Suffixtree *stree,Location *loc,
+                                Bref btptr,SYMBOL *left,
+                                SYMBOL *right,Uint rescanlength);
+
 //}
 
 /*EE
