--- MUMmer3.23/scripts/mummerplot.pl	2018-05-22 17:01:01.000000000 +0100
+++ MUMmer3.23/scripts/mummerplot.pl.new	2018-05-22 17:01:52.000000000 +0100
@@ -1151,7 +1151,7 @@
         $P_KEY = "unset key";
         $P_FORMAT .= "\nset mouse format \"$TFORMAT\"";
         $P_FORMAT .= "\nset mouse mouseformat \"$MFORMAT\"";
-        $P_FORMAT .= "\nset mouse clipboardformat \"$MFORMAT\"";
+        $P_FORMAT .= "\nif(GPVAL_VERSION < 5) set mouse clipboardformat \"$MFORMAT\"";
     }
     else {
         $P_LS = "set linestyle";
