--- MUMmer3.23/src/kurtz/streesrc/linkloc.c.old	2024-08-07 12:45:16
+++ MUMmer3.23/src/kurtz/streesrc/linkloc.c	2024-08-07 13:48:38
@@ -11,6 +11,9 @@
 #include "debugdef.h"
 #include "streeacc.h"
 
+void getbranchinfostree(Suffixtree *stree,Uint whichinfo,
+                        Branchinfo *branchinfo,Bref btptr);
+
 void rescanstree(Suffixtree *stree,Location *loc,
                  Bref btptr,SYMBOL *left,SYMBOL *right)
 {
