--- MUMmer3.23/src/kurtz/mm3src/procmaxmat.c.old	2024-08-07 14:05:17
+++ MUMmer3.23/src/kurtz/mm3src/procmaxmat.c	2024-08-07 14:07:03
@@ -19,6 +19,16 @@
 #include "spacedef.h"
 #include "streedef.h"
 #include "maxmatdef.h"
+
+Uint getmaxtextlenstree(void);
+
+Sint constructprogressstree(Suffixtree *stree,SYMBOL *text,
+                            Uint textlen,
+                            void (*progress)(Uint,void *),
+                            void (*finalprogress)(void *),
+                            void *info);
+
+void freestree(Suffixtree *stree);
 
 //}
 
