--- fastqc.orig	2016-11-18 14:12:31.620056634 +0100
+++ fastqc	2016-11-18 14:12:44.151978522 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 use warnings;
 use strict;
 use FindBin qw($RealBin);
@@ -273,4 +273,4 @@ BUGS
     Any bugs in fastqc should be reported either to simon.andrews@babraham.ac.uk
     or in www.bioinformatics.babraham.ac.uk/bugzilla/
                    
-    
\ No newline at end of file
+    
