diff --git a/src/isoformchecklib/Util.py b/src/isoformchecklib/Util.py
index c65913b..8baa69c 100644
--- a/src/isoformchecklib/Util.py
+++ b/src/isoformchecklib/Util.py
@@ -14,7 +14,7 @@ def get_git_version():
 	time = subprocess.check_output(["git", "show", "-s", "--format=%ci"], cwd=os.path.dirname(os.path.abspath(__file__))).strip().decode().strip()
 	return "branch " + branch + " commit " + commit + " " + time
 
-Version = "development " + get_git_version()
+Version = "bioconda 1.0.0"
 
 GlobalVerbosity = 0
 
