diff --git a/lib/common.h b/lib/common.h
index 052452e..c6040a7 100644
--- a/lib/common.h
+++ b/lib/common.h
@@ -3,6 +3,7 @@
 
 #include <string>
 #include <vector>
+#include <cstdint>
 
 // define our program name
 #define PROGRAM_NAME "chips"
diff --git a/lib/fragment.h b/lib/fragment.h
index e91e0aa..88212db 100644
--- a/lib/fragment.h
+++ b/lib/fragment.h
@@ -3,6 +3,7 @@
 
 #include <stdint.h>
 #include <string>
+#include <cstdint>
 
 class Fragment {
  public:
diff --git a/lib/multithread.cpp b/lib/multithread.cpp
index 780003c..f349ae7 100644
--- a/lib/multithread.cpp
+++ b/lib/multithread.cpp
@@ -1,4 +1,5 @@
 #include "multithread.h"
+#include <stdexcept>
 
 template <typename T>
 T TaskQueue <T> ::pop(){
diff --git a/lib/options.h b/lib/options.h
index cf1616d..4519375 100644
--- a/lib/options.h
+++ b/lib/options.h
@@ -3,8 +3,8 @@
 
 #include <vector>
 #include <string>
-
 #include <stdint.h>
+#include <cstdint>
 
 class Options {
  public:
