--- PePr/PePr.py	2016-02-04 18:34:14.651531855 +0000
+++ PePr/PePr.py.new	2016-02-04 20:30:22.435892169 +0000
@@ -6,16 +6,16 @@
 import sys
 import logging
 
-import logConfig
-import optParser
-import fileParser
-import shiftSize
-import windowSize
-import sigTests
-import misc
-from classDef import Parameters
+from pepr import logConfig
+from pepr import optParser
+from pepr import fileParser
+from pepr import shiftSize
+from pepr import windowSize
+from pepr import sigTests
+from pepr import misc
+from pepr.classDef import Parameters
 
-def main(argv):
+def main(argv=sys.argv):
     # initialize the logger 
     root_logger = logging.getLogger("")
     debug = root_logger.debug
