--- setup.py	2015-08-05 16:48:49.000000000 +0000
+++ setup.py.new	2016-02-04 22:14:57.779156880 +0000
@@ -15,8 +15,9 @@
           "Programming Language :: Python",
           "License :: OSI Approved :: GNU General Public License v3 (GPLv3)"
           ],
-      packages=["PePr"],
+      packages=["pepr"],
       package_data={"PePr":['data/*.bed']},
+      package_dir={"pepr":"./PePr"},
       install_requires=[
           'numpy>=1.6.0',
           'scipy>=0.14.0',
