diff --git a/ariba/external_progs.py b/ariba/external_progs.py
index 4f21f12..e2365c5 100644
--- a/ariba/external_progs.py
+++ b/ariba/external_progs.py
@@ -33,7 +33,7 @@ class Error (Exception): pass
     'bowtie2': ('--version', re.compile('.*bowtie2.*version (.*)$')),
     'cdhit': ('', re.compile('CD-HIT version ([0-9\.]+) \(')),
     'nucmer': ('--version', re.compile('([0-9]+\.[0-9\.]+.*$)')),
-    'spades': ('--version', re.compile('SPAdes\s+v([0-9\.]+)'))
+    'spades': ('--version', re.compile('SPAdes.*v([0-9\.]+)'))
 }
 
