diff --git a/ariba/ref_genes_getter.py b/ariba/ref_genes_getter.py
index 2613ffc..b830710 100644
--- a/ariba/ref_genes_getter.py
+++ b/ariba/ref_genes_getter.py
@@ -617,7 +617,6 @@ def _get_from_ncbi(self, outprefix, test=None): ## author github:schultzm
                                                webenv=webenv, query_key=query_key,
                                                idtype="acc")
             #pull out the records as fasta from the genbank
-            from Bio.Alphabet import generic_dna
             from Bio import SeqIO
             from Bio.Seq import Seq
             from Bio.SeqRecord import SeqRecord
@@ -645,7 +644,7 @@ def _get_from_ncbi(self, outprefix, test=None): ## author github:schultzm
                                 except KeyError:
                                     print(f"gb_feature.qualifer not found", file=sys.stderr)
                             accession = gb_record.id
-                            seq_out = Seq(str(gb_feature.extract(gb_record.seq)), generic_dna)
+                            seq_out = Seq(str(gb_feature.extract(gb_record.seq)))
                             record_new.append(SeqRecord(seq_out,
                                          id=f"{id[0]}.{accession}",
                                          description=""))
