From 8a68981d04eb22ad2ab0fbd0e3fc48d7e8012163 Mon Sep 17 00:00:00 2001
From: "Robert A. Petit III" <robbie.petit@gmail.com>
Date: Wed, 12 Aug 2020 14:29:52 -0400
Subject: [PATCH] Solution 2 for fixing CARD downloads with getref

---
 ariba/ref_genes_getter.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ariba/ref_genes_getter.py b/ariba/ref_genes_getter.py
index 0279d71..8c7b9f0 100644
--- a/ariba/ref_genes_getter.py
+++ b/ariba/ref_genes_getter.py
@@ -42,7 +42,7 @@ def __init__(self, ref_db, version=None, debug=False):
     def _get_card_versions(self, tmp_file):
         print('Getting available CARD versions')
         common.download_file('https://card.mcmaster.ca/download', tmp_file, max_attempts=self.max_download_attempts, sleep_time=self.sleep_time, verbose=True)
-        p = re.compile(r'''href="(/download/.*?broad.*?v([0-9]+\.[0-9]+\.[0-9]+)\.tar\.(gz|bz2))"''')
+        p = re.compile(r'''href="(/download/.*?broadstreet.*?v([0-9]+\.[0-9]+\.[0-9]+)\.tar\.(gz|bz2))"''')
         versions = {}
 
         with open(tmp_file) as f:
