--- src/align.cc
+++ src/align.cc
@@ -1299,9 +1299,9 @@
 
       if(entry.value>storingLimit) {
-	matCoord sCoord={entry.x,entry.y};
+	matCoord sCoord={static_cast<unsigned long int>(entry.x),static_cast<unsigned long int>(entry.y)};
 	map<matCoord,pair<matCoord,store> >::iterator prevGood=maxses.find(sCoord);
  	if(prevGood==maxses.end() ||
  	   prevGood->second < entry.value ){  // If new or improved
- 	  matCoord eCoord={x,y};
+ 	  matCoord eCoord={static_cast<unsigned long int>(x),static_cast<unsigned long int>(y)};
 	  if(prevGood==maxses.end()) { // If new
 	    MS_res balg; // ={entry.value,sCoord,eCoord};
