--- scripts/run	2016-05-03 15:09:03.482611039 +0200
+++ scripts/run	2016-05-03 15:24:36.778606792 +0200
@@ -131,7 +131,7 @@
 end
 
 def depot_path
-  File.join(working_dir, 'depot')
+  File.join('depot')
 end
 
 def ra_bin_dir(debug: false)
@@ -139,40 +139,40 @@
 end
 
 def filter_contained_bin(debug: false)
-  File.join(ra_bin_dir, "filter_contained")
+  File.join("filter_contained")
 end
 
 def create_dovetail_bin(debug: false)
-  File.join(ra_bin_dir, "widen_overlaps")
+  File.join("widen_overlaps")
 end
 
 def filter_transitives_bin(debug: false)
-  File.join(ra_bin_dir, "filter_transitive")
+  File.join("filter_transitive")
 end
 
 def unitigger_bin(debug: false)
-  File.join(ra_bin_dir, "unitigger")
+  File.join("unitigger")
 end
 
 def filter_erroneous_overlaps_bin(debug: false)
-  File.join(ra_bin_dir, "filter_erroneous_overlaps")
+  File.join("filter_erroneous_overlaps")
 end
 
 def draw_graph_bin(debug: false)
-  File.join(ra_bin_dir, "overlap2dot")
+  File.join("overlap2dot")
 end
 
 def fill_read_coverage_bin(debug: false)
-  File.join(ra_bin_dir, "fill_read_coverage")
+  File.join("fill_read_coverage")
 end
 
 def depot_bin(debug: false)
-  File.join(ra_bin_dir, "depot")
+  File.join("depot")
 end
 
 def run_graphmap(reads_path)
   overlaps_path = File.join(working_dir, 'overlaps.mhap')
-  binary = File.join(assembler_dir, 'components', 'graphmap', 'bin', 'Linux-x64', 'graphmap')
+  binary = File.join('graphmap')
   cmd = "#{binary} -w owler -B 0 -r #{reads_path} -d #{reads_path} -L mhap -o #{overlaps_path}"
   puts(cmd)
   if !system(cmd)
