--- setup.py
+++ setup.py
@@ -36,11 +36,6 @@ def setup_package():
                 'rsrc/LANL.TN93OUTPUT.csv.gz'
                 ]
             },
-        dependency_links = ['git+git://github.com/veg/hyphy-python.git@0.1.3#egg=HyPhy-0.1.3',
-                            'git+git://github.com/veg/BioExt.git@0.17.2#egg=BioExt-0.17.2',
-                            'git+git://github.com/veg/hppy.git@0.9.6#egg=hppy-0.9.6',
-                            'git+git://github.com/veg/hivclustering.git@1.2.0#egg=hivclustering-1.2.0'
-                            ],
         install_requires=[
             'biopython >= 1.58',
             'BioExt >= 0.17.2',
