diff --git a/Header.h b/Header.h
index c0db3bb..89aaab7 100644
--- a/Header.h
+++ b/Header.h
@@ -49,6 +49,8 @@ SOFTWARE.
 
 #ifdef _WIN32
 #include <intrin.h>
+#elif defined(__ARM_NEON)
+#include "sse2neon.h"
 #else
 #include <x86intrin.h>
 #endif
diff --git a/meson.build b/meson.build
index f8cf109..c48b26f 100644
--- a/meson.build
+++ b/meson.build
@@ -17,11 +17,6 @@ endif
 base_flags += ['-DPV=' + meson.project_version()]
 
 extensions = {
-    # '': [],
-    '.avx2': ['-mavx2'], 
-    '.avx': ['-mavx'], 
-    '.sse42': ['-msse4.2'], 
-    '.sse41': ['-msse4.1']
     }
 
 foreach ext, flag : extensions
