diff --git a/CMakeLists.txt b/CMakeLists.txt
index fb8a42b..7acc580 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -8,7 +8,7 @@ project(
 
 add_compile_definitions(PV="${PROJECT_VERSION}")
 
-set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD 14)
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 set(CMAKE_C_STANDARD 99)
 set(CMAKE_C_STANDARD_REQUIRED ON)
@@ -34,10 +34,10 @@ endif()
 
 # SIMD exntensions
 set(EXTENSIONS
-    ".avx2,-mavx2"
-    ".avx,-mavx"
-    ".sse42,-msse4.2"
-    ".sse41,-msse4.1"
+    ".avx2,"
+    ".avx,"
+    ".sse42,"
+    ".sse41,"
 )
 foreach(ext_pair ${EXTENSIONS})
     string(REPLACE "," ";" ext_list "${ext_pair}")
