diff --git a/meson.build b/meson.build
index 28b7dc1..296a4a4 100644
--- a/meson.build
+++ b/meson.build
@@ -14,7 +14,7 @@ foreach ext, flag : extensions
     executable('PretextGraph' + ext, 'PretextGraph.cpp', dependencies : thread_dep, install : true, cpp_args : flags, link_with : 
         static_library('deflate' + ext, 'libdeflate/lib/deflate_compress.c', 'libdeflate/lib/deflate_decompress.c', 'libdeflate/lib/utils.c', c_args : flags, link_with : 
             [   static_library('deflate_x86' + ext, 'libdeflate/lib/x86/cpu_features.c', c_args : flags),
-                static_library('deflate_arm' + ext, 'libdeflate/lib/arm/cpu_features.c', c_args : flags)]))
+                static_library('deflate_arm' + ext, 'libdeflate/lib/arm/cpu_features.c', c_args : flags)], include_directories : include_directories('libdeflate')), include_directories : include_directories('libdeflate'))
 endforeach
 
 e = executable('PretextGraph', 'wrapper/SIMDExtensionWrapper.cpp', install : true, cpp_args : base_flags + ['-DProgramName=PretextGraph'])
