--- Makefile	2016-09-18 01:24:06.664551971 +0200
+++ Makefile-new	2016-09-18 20:48:15.855363814 +0200
@@ -15,7 +15,7 @@
 # You might change this to /usr/local/bin and /usr/local/man or some such.
 #
 BINDIR 		= $(HOME)/bin
-MANDIR          = $(HOME)/man
+MANDIR          = $(HOME)/share/man
 
 # Compilation flags. Set your compiler and any necessary global
 # switches here. Use an ANSI-compliant C compiler.
@@ -58,7 +58,7 @@
 
 install: $(PROGS)
 	cp $(PROGS)   $(BINDIR)/
-	cp rnabob.man $(MANDIR)/man1/rnabob.1
+	cp rnabob.man $(MANDIR)/rnabob.1
 
 clean:
 	-rm -f *.o *~ *.aux *.dvi *.log core TAGS $(PROGS)
