--- distribute_setup.py	2014-07-31 21:38:58.000000000 +0000
+++ distribute_setup.py.new	2019-07-30 15:59:17.038028111 +0000
@@ -47,7 +47,7 @@
         return os.spawnl(os.P_WAIT, sys.executable, *args) == 0

 DEFAULT_VERSION = "0.6.10"
-DEFAULT_URL = "http://pypi.python.org/packages/source/d/distribute/"
+DEFAULT_URL = "https://pypi.python.org/packages/source/d/distribute/"
 SETUPTOOLS_FAKED_VERSION = "0.6c11"

 SETUPTOOLS_PKG_INFO = """\
--- setup.py	2015-04-15 18:48:57.000000000 +0000
+++ setup.py.new	2019-07-30 16:01:08.941419746 +0000
@@ -61,12 +61,12 @@
             ext_modules = get_extension_modules(),
             py_modules = get_pymodules(),
             test_suite = 'nose.collector',
-            setup_requires = ['nose>=0.10.4','cython>=0.12'],
+#            setup_requires = ['nose>=0.10.4','Cython>=0.12'],
             author = "Liguo Wang",
 			author_email ="wangliguo78@gmail.com",
 			platforms = ['Linux','MacOS'],
-			requires = ['cython (>=0.17)'],
-			install_requires = ['cython>=0.17',],
+#			requires = ['cython (>=0.17)'],
+#			install_requires = ['cython>=0.17',],
             description = "Model based analysis of ChIP-exo",
             url = "http://chipexo.sourceforge.net/",
             zip_safe = False,
