--- auxprogs/filterBam/src/Makefile	2017-01-18 08:45:16.302477496 +0400
+++ auxprogs/filterBam/src/Makefile	2017-01-18 09:08:29.754681493 +0400
@@ -10,7 +10,7 @@
 OBJECTS = $(SOURCES:.cc=.o)
 BAMTOOLS = ${PREFIX}/include/bamtools
 #INCLUDES = -I$(BAMTOOLS) -Iheaders -I./bamtools
-INCLUDES =  -I${PREFIX}/include/bamtools/ -Iheaders -I./bamtools
+INCLUDES =  -I$(PREFIX)/include -I$(BAMTOOLS) -Iheaders -I./bamtools
 #LIBS = -lbamtools -lz
 LIBS = ${PREFIX}/lib/libbamtools.a -lz
 CFLAGS = -std=c++0x
