diff --git a/auxprogs/bam2wig/Makefile b/auxprogs/bam2wig/Makefile
index 8584531b..28e109d8 100644
--- a/auxprogs/bam2wig/Makefile
+++ b/auxprogs/bam2wig/Makefile
@@ -9,7 +9,7 @@
 ifdef TOOLDIR
 	DEF_TOOLDIR=1
 else
-	TOOLDIR=$(HOME)/tools
+	TOOLDIR=${PREFIX}/tools
 	# replace with your own parent directory of htslib
 	# if TOOLDIR is not specified as environment variable
 	# and libhts-dev is not installed
@@ -20,14 +20,14 @@ SOURCES = $(PROGRAM)
 OBJECTS = $(SOURCES:.c=.o)
 
 ifneq ($(wildcard ${TOOLDIR}/htslib/htslib/.),) # if TOOLDIR exists and contains htslib
-	INCLUDES=-I$(TOOLDIR)/htslib -I$(TOOLDIR)/htslib/htslib
-	HTSLIBS=-L$(TOOLDIR)/htslib -Wl,-rpath,$(TOOLDIR)/htslib -lhts -lcurses -lm -lz -lpthread -lcurl -lssl -lcrypto -lbz2 -llzma
+	INCLUDES=-I${PREFIX}/include/htslib
+	HTSLIBS=-L${PREFIX}/lib -Wl,-rpath,${PREFIX}/lib -lhts -lcurses -lm -lz -lpthread -lcurl -lssl -lcrypto -lbz2 -llzma
 	HTSERRHINT="Check if the installed HTSlib is of version 1.10 or higher - see README.txt"
 else
-	INCLUDES=-I/usr/include/htslib
-	HTSLIBS=-lhts -lz -lpthread
+	INCLUDES=-I${PREFIX}/include/htslib
+	HTSLIBS=-L${PREFIX}/lib -lhts -lz -lpthread
 	ifdef DEF_TOOLDIR
-		HTSERRHINT="There is no htslib folder in directory $(TOOLDIR) \n"
+		HTSERRHINT="There is no htslib folder in directory ${PREFIX} \n"
 	endif
 	HTSERRHINT:=$(HTSERRHINT)"Check if HTSlib is installed and of version 1.10 or higher - see README.txt"
 endif
