--- auxprogs/compileSpliceCands/Makefile.old	2019-03-27 13:57:43.900965927 -0500
+++ auxprogs/compileSpliceCands/Makefile	2019-03-27 13:58:52.676961696 -0500
@@ -1,8 +1,8 @@
 compileSpliceCands : compileSpliceCands.o list.h list.o
-	gcc $(LDFLAGS) -o compileSpliceCands compileSpliceCands.o list.o;
+	$(CC) -o compileSpliceCands compileSpliceCands.o list.o;
 #	cp compileSpliceCands ../../bin
 compileSpliceCands.o : compileSpliceCands.c 
-	gcc -Wall -pedantic -ansi $(CPPFLAGS) -c compileSpliceCands.c 
+	$(CC) -Wall -pedantic -ansi $(CPPFLAGS) -c compileSpliceCands.c 
 
 all : compileSpliceCands
 
