diff --git a/auxprogs/compileSpliceCands/Makefile b/auxprogs/compileSpliceCands/Makefile
index dc16beb2..0a9dc1ce 100644
--- a/auxprogs/compileSpliceCands/Makefile
+++ b/auxprogs/compileSpliceCands/Makefile
@@ -1,5 +1,3 @@
-CC?=gcc
-
 compileSpliceCands : compileSpliceCands.o list.h list.o
 	$(CC) $(CFLAGS)	$(CPPFLAGS) $(LDFLAGS) -o compileSpliceCands compileSpliceCands.o list.o
 	cp compileSpliceCands ../../bin/compileSpliceCands
