--- auxprogs/filterBam/Makefile	2017-01-17 16:55:30.649430242 +0400
+++ auxprogs/filterBam/Makefile	2017-01-17 16:56:08.407628338 +0400
@@ -2,11 +2,11 @@
 #
 # 	Created: 28-February-2012
 #	Last modified: 15-October-2015
-INCLUDES =  ${PREFIX}/include/bamtools/
-LIBS = ${PREFIX}/lib/libbamtools.a -lz
+INCLUDES +=  ${PREFIX}/include/bamtools/
+LIBS += ${PREFIX}/lib/libbamtools.a -lz
 SOURCES = bam2hints.cc
 OBJECTS = $(SOURCES:.cc=.o)
-CFLAGS = -Wall -g -p -O0
+CFLAGS += -Wall -g -p -O0
 CXXFLAGS += -Wall -O2 # -g -p -g -ggdb 
 
 all:
