diff --git a/auxprogs/filterBam/Makefile b/auxprogs/filterBam/Makefile
index c2a3e009..3d7d68e3 100644
--- a/auxprogs/filterBam/Makefile
+++ b/auxprogs/filterBam/Makefile
@@ -3,6 +3,12 @@
 # 	Created: 28-February-2012
 #	Last modified: 09-May-2018
 
+INCLUDES =  ${PREFIX}/include/bamtools/
+LIBS = -L${PREFIX}/lib/ -lbamtools -lz
+SOURCES = bam2hints.cc
+OBJECTS = $(SOURCES:.cc=.o)
+CXXFLAGS += -Wall -O2
+
 all: 
 	(cd src;make)
 
