diff -ruN Augustus-3.4.0-org/src/pp_profile.cc Augustus-3.4.0/src/pp_profile.cc
--- Augustus-3.4.0-org/src/pp_profile.cc	2020-12-11 16:40:21.000000000 +0100
+++ Augustus-3.4.0/src/pp_profile.cc	2022-09-08 16:43:50.706953557 +0200
@@ -672,7 +672,7 @@
             // read in the allowed distance range
             istringstream lstrm(readAndConcatPart(strm, type, lineno));
             DistanceType addDist;
-            if(!(lstrm >> addDist >> ws && lstrm.eof()))
+            if(!(lstrm >> addDist && lstrm.peek() == EOF))
                 throw ProfileParseError(lineno - newlinesFromPos(lstrm.str(), lstrm.tellg()) -1);
             finalDist += addDist;
             } else // if dist is not specified, assume arbitrary distance
Binary files Augustus-3.4.0-org/src/.pp_profile.cc.swp and Augustus-3.4.0/src/.pp_profile.cc.swp differ
