diff --git a/auxprogs/utrrnaseq/Debug/makefile b/auxprogs/utrrnaseq/Debug/makefile
index 6232aac0..99cc94c7 100644
--- a/auxprogs/utrrnaseq/Debug/makefile
+++ b/auxprogs/utrrnaseq/Debug/makefile
@@ -49,11 +49,13 @@ utrrnaseq: $(OBJS) $(USER_OBJS)
 	$(CXX)  -o "utrrnaseq" $(OBJS) $(USER_OBJS) $(LIBS)
 	@echo 'Finished building target: $@'
 	@echo ' '
+	cp utrrnaseq ../../../bin/utrrnaseq
 
 # Other Targets
 clean:
 	-$(RM) $(CC_DEPS)$(C++_DEPS)$(EXECUTABLES)$(C_UPPER_DEPS)$(CXX_DEPS)$(OBJS)$(CPP_DEPS)$(C_DEPS) utrrnaseq
 	-@echo ' '
+	rm ../../../bin/utrrnaseq
 
 .PHONY: all clean dependents
 
