diff --git a/auxprogs/utrrnaseq/Debug/src/subdir.mk b/auxprogs/utrrnaseq/Debug/src/subdir.mk
index 2f18f3ca..ac140634 100644
--- a/auxprogs/utrrnaseq/Debug/src/subdir.mk
+++ b/auxprogs/utrrnaseq/Debug/src/subdir.mk
@@ -38,7 +38,7 @@ CPP_DEPS += \
 src/%.o: ../src/%.cpp
 	@echo 'Building file: $<'
 	@echo 'Invoking: GCC C++ Compiler'
-	g++ -I/usr/include/boost -O0 -g3 -pedantic -Wall -c -fmessage-length=0 -MMD -MP -MF"$(@:%.o=%.d)" -MT"$(@)" -o "$@" "$<"
+	$(CXX) -I${PREFIX}/include -O0 -g3 -pedantic -Wall -c -fmessage-length=0 -MMD -MP -MF"$(@:%.o=%.d)" -MT"$(@)" -o "$@" "$<"
 	@echo 'Finished building: $<'
 	@echo ' '
 
