--- a/auxprogs/utrrnaseq/Makefile	2022-09-23 10:06:14.000000000 -0700
+++ b/auxprogs/utrrnaseq/Makefile	2022-10-08 10:15:31.359607884 -0700
@@ -29,6 +29,7 @@
 	$(CXX) $(CPPFLAGS) $(CXXFLAGS) $(LDFLAGS) -o $@ $^ $(LIBS)
 	@echo 'Finished building target: $@'
 	@echo ' '
+	cp -f utrrnaseq ../../bin/utrrnaseq
 
 $(OBJS): $(OBJ_DIR)/%.o: $(SRC_DIR)/%.cpp
 	@echo 'Building file: $<'
@@ -55,5 +56,6 @@
 clean: clean_test
 	rm -rf $(OBJS) $(DEPS) $(OBJ_DIR) utrrnaseq
 	@echo ' '
+	rm ../../bin/utrrnaseq
 
 -include $(DEPS)
