--- auxprogs/utrrnaseq/Debug/makefile.old	2019-03-28 01:10:52.267908615 -0500
+++ auxprogs/utrrnaseq/Debug/makefile	2019-03-28 01:11:16.438920051 -0500
@@ -44,7 +44,7 @@
 utrrnaseq: $(OBJS) $(USER_OBJS)
 	@echo 'Building target: $@'
 	@echo 'Invoking: GCC C++ Linker'
-	g++  -o "utrrnaseq" $(OBJS) $(USER_OBJS) $(LIBS)
+	$(CXX) -o "utrrnaseq" $(OBJS) $(USER_OBJS) $(LIBS)
 	@echo 'Finished building target: $@'
 	@echo ' '
 
