--- auxprogs/utrrnaseq/Debug/src/subdir.mk.old	2019-03-28 00:35:55.658226975 -0500
+++ auxprogs/utrrnaseq/Debug/src/subdir.mk	2019-03-28 00:36:34.518227203 -0500
@@ -38,7 +38,7 @@
 src/%.o: ../src/%.cpp
 	@echo 'Building file: $<'
 	@echo 'Invoking: GCC C++ Compiler'
-	g++ -I/usr/include/boost -O0 -g3 -pedantic -Wall -c -fmessage-length=0 -MMD -MP -MF"$(@:%.o=%.d)" -MT"$(@)" -o "$@" "$<"
+	$(CXX) -I${PREFIX}/include -O0 -g3 -pedantic -Wall -c -fmessage-length=0 -MMD -MP -MF"$(@:%.o=%.d)" -MT"$(@)" -o "$@" "$<"
 	@echo 'Finished building: $<'
 	@echo ' '
 
