diff --git a/spaTyper/utils.py b/spaTyper/utils.py
index c465eaf..0c1df4c 100644
--- a/spaTyper/utils.py
+++ b/spaTyper/utils.py
@@ -189,9 +189,12 @@ def download_file_types(folder, debug):
 ####################################################
 def print_time_stamp (out):
     """Prints out timestamp in a file provided. Format: time.time()"""
-    timefile = open(out, 'w')    
-    string2write = str(time.time())
-    timefile.write(string2write)
+    try:
+        with open(out, 'wt') as timefile:   
+            string2write = str(time.time())
+            timefile.write(string2write)
+    except IOError:
+        print ("\tUnable to write time stamp", file=sys.stderr)
     return()
 
 ####################################################
