diff --git a/main/spaTyper b/main/spaTyper
index 4aeee87..1eeafc2 100755
--- a/main/spaTyper
+++ b/main/spaTyper
@@ -25,10 +25,10 @@ module. Python 3 version only.
 ''', epilog="Original code: mjsull. Modified by: JFSanchezHerrero")
 
 #####################################################
-parser.add_argument('-r', '--repeat_file', action='store', 
-                    help='List of spa repeats (http://spa.ridom.de/dynamic/sparepeats.fasta)')
-parser.add_argument('-o', '--repeat_order_file', action='store', 
-                    help='List spa types and order of repeats (http://spa.ridom.de/dynamic/spatypes.txt)')
+parser.add_argument('-r', '--repeat_file', action='store', help='List of spa repeats (http://spa.ridom.de/dynamic/sparepeats.fasta)',
+                    default=f'{os.environ.get("SPATYPER_SHARE")}/sparepeats.fasta')
+parser.add_argument('-o', '--repeat_order_file', action='store', help='List spa types and order of repeats (http://spa.ridom.de/dynamic/spatypes.txt)', 
+                    default=f'{os.environ.get("SPATYPER_SHARE")}/spatypes.txt)
 parser.add_argument('-d', '--folder', action='store', help='Folder to save downloaded files from Ridom/Spa server')
 parser.add_argument('-f', '--fasta', action='store', nargs='+', help='List of one or more fasta files.')
 parser.add_argument('-g', '--glob', action='store', 