--- src/Util/VariantIndex.cpp.old	2019-05-03 00:03:15.610010581 -0500
+++ src/Util/VariantIndex.cpp	2019-05-03 00:01:16.046013820 -0500
@@ -86,7 +86,7 @@
         for(size_t vi = 0; vi < bucket.size(); ++vi)
         {
             const VariantRecord& record = m_records[bucket[vi]];
-            if(abs(record.position - position) < distance)
+            if(static_cast<int>(record.position - position) < distance)
                 out.push_back(record);
         }
     }
--- src/Algorithm/GapFillProcess.cpp.old	2019-05-03 00:03:48.087009785 -0500
+++ src/Algorithm/GapFillProcess.cpp	2019-05-03 00:04:02.657009430 -0500
@@ -233,7 +233,7 @@
 
     for(size_t i = 0; i < sequences.size(); ++i)
     {
-        int diff = abs(sequences[i].size() - estimatedSize);
+        int diff = static_cast<int>(sequences[i].size() - estimatedSize);
         //printf("ES: %d S: %zu D: %d\n", estimatedSize, sequences[i].size(), diff);
 
         if(diff < selectedSizeDiff)
