--- src/configure.ac.old	2019-05-03 17:14:18.691599053 -0500
+++ src/configure.ac	2019-05-03 17:15:02.736610682 -0500
@@ -92,8 +92,8 @@
 
 # Set compiler flags.
 AC_SUBST(AM_CXXFLAGS, "-Wall -Wextra $fail_on_warning -Wno-unknown-pragmas")
-AC_SUBST(CXXFLAGS, "-std=c++98 -O3")
-AC_SUBST(CFLAGS, "-std=gnu99 -O3")
+AC_SUBST(CXXFLAGS)
+AC_SUBST(CFLAGS)
 AC_SUBST(CPPFLAGS, "$CPPFLAGS $openmp_cppflags $sparsehash_include $bamtools_include")
 AC_SUBST(LDFLAGS, "$openmp_cppflags $external_malloc_ldflags $bamtools_ldflags $LDFLAGS") 
