--- setup.py.org	2023-10-17 17:35:14.702458316 -0500
+++ setup.py	2023-10-17 17:46:27.269258824 -0500
@@ -1,7 +1,8 @@
+import os
 from setuptools import setup
 from dunamai import Version, Style
 
 setup(
     name="nii2dcm",
-    version=Version.from_git().serialize(metadata=False, style=Style.SemVer),
-)
\ No newline at end of file
+    version=Version.parse(os.getenv('PKG_VERSION')).serialize(metadata=False, style=Style.SemVer),
+)
