--- utilities/package_info.py.old	2017-11-17 14:47:21.853391078 +0200
+++ utilities/package_info.py	2017-11-17 14:47:43.069608417 +0200
@@ -7,5 +7,5 @@
 Copyright 2016 Canada's Michael Smith Genome Sciences Centre
 Report bugs to <trans-abyss@googlegroups.com>
 '''
-PACKAGEDIR = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
+PACKAGEDIR = os.environ.get('CONDA_PREFIX', os.path.dirname(os.path.dirname(os.path.realpath(__file__))))
 BINDIR = os.path.join(PACKAGEDIR, 'bin')
