--- transabyss.old	2016-08-02 19:51:40.000000000 +0200
+++ transabyss	2017-11-17 14:20:44.905030270 +0200
@@ -13,25 +13,25 @@
 import sys
 import textwrap
 import time
-from utilities import package_info
-from utilities import psl_cid_extractor
-from utilities.adj_utils import has_edges
-from utilities.adj_utils import remove_redundant_paths
-from utilities.adj_utils import unbraid
-from utilities.adj_utils import walk
-from utilities.common_utils import StopWatch
-from utilities.common_utils import check_env
-from utilities.common_utils import is_empty_txt
-from utilities.common_utils import log
-from utilities.common_utils import path_action
-from utilities.common_utils import paths_action
-from utilities.common_utils import run_shell_cmd
-from utilities.common_utils import threshold_action
-from utilities.common_utils import touch
-from utilities.fasta_utils import abyssmap_merge_fastas
-from utilities.fasta_utils import blat_merge_fastas
-from utilities.fasta_utils import blat_self_align
-from utilities.fasta_utils import filter_fasta
+from transabyss import package_info
+from transabyss import psl_cid_extractor
+from transabyss.adj_utils import has_edges
+from transabyss.adj_utils import remove_redundant_paths
+from transabyss.adj_utils import unbraid
+from transabyss.adj_utils import walk
+from transabyss.common_utils import StopWatch
+from transabyss.common_utils import check_env
+from transabyss.common_utils import is_empty_txt
+from transabyss.common_utils import log
+from transabyss.common_utils import path_action
+from transabyss.common_utils import paths_action
+from transabyss.common_utils import run_shell_cmd
+from transabyss.common_utils import threshold_action
+from transabyss.common_utils import touch
+from transabyss.fasta_utils import abyssmap_merge_fastas
+from transabyss.fasta_utils import blat_merge_fastas
+from transabyss.fasta_utils import blat_self_align
+from transabyss.fasta_utils import filter_fasta
 
 
 TRANSABYSS_VERSION = package_info.VERSION
