--- CMakeLists.txt.old	2018-05-16 16:00:17.834072232 -0400
+++ CMakeLists.txt	2018-05-16 16:01:17.146423987 -0400
@@ -188,8 +188,12 @@
 #-------------------------------------------------------------------------------
 # default installation prefix
 #-------------------------------------------------------------------------------
-set(INSTALL_PREFIX /usr/local CACHE PATH "Installation prefix")
-set(CMAKE_INSTALL_PREFIX ${INSTALL_PREFIX} CACHE INTERNAL "Installation prefix" FORCE)
+if(NOT DEFINED CMAKE_INSTALL_PREFIX)
+    set(CMAKE_INSTALL_PREFIX "/usr/local" CACHE PATH "Installation prefix")
+endif()
+if(NOT DEFINED INSTALL_PREFIX)
+    set(INSTALL_PREFIX ${CMAKE_INSTALL_PREFIX} CACHE INTERNAL "Installation prefix" FORCE)
+endif()
 if(CMAKE_SYSTEM_NAME MATCHES "FreeBSD")
     set(DEFAULT_LIBRARIES pthread)
 else(CMAKE_SYSTEM_NAME MATCHES "FreeBSD")
