diff --git peakachulib/adaptive.py peakachulib/adaptive.py
index 26c701a..e947749 100644
--- peakachulib/adaptive.py
+++ peakachulib/adaptive.py
@@ -192,11 +192,10 @@ class AdaptiveApproach(object):
                 blocks, columns=["blockNb", "blockChrom", "blockStart",
                                  "blockEnd", "blockStrand", "blockExpression",
                                  "readCount"])
-            block_df.loc[:, ["blockNb", "blockStart", "blockEnd",
-                             "blockExpression", "readCount"]] = block_df.loc[
-                                 :, ["blockNb", "blockStart", "blockEnd",
-                                     "blockExpression", "readCount"]].apply(
-                                         pd.to_numeric)
+            block_df[["blockNb", "blockStart", "blockEnd", "blockExpression",
+                      "readCount"]] = block_df[
+                    ["blockNb", "blockStart", "blockEnd", "blockExpression",
+                     "readCount"]].apply(pd.to_numeric)
             peak_df = self._split_cluster_peaks(block_df, cluster_expr,
                                                 peak_df, min_cluster_expr_frac,
                                                 min_block_overlap,
