diff --git a/external/ggcat/crates/capi/ggcat-cpp-api/Makefile b/external/ggcat/crates/capi/ggcat-cpp-api/Makefile
index 5542962..1e1ef35 100644
--- a/external/ggcat/crates/capi/ggcat-cpp-api/Makefile
+++ b/external/ggcat/crates/capi/ggcat-cpp-api/Makefile
@@ -13,4 +13,3 @@
 
 ./lib/libggcat_cpp_bindings.a: ggcat-source
 	cargo build --release --package ggcat-cpp-bindings
-	cp ../../../target/release/libggcat_cpp_bindings.a ./lib/
