From eed45163b0c80ef28c19450bb75301665a6fcf28 Mon Sep 17 00:00:00 2001
From: John Lees <j.lees@imperial.ac.uk>
Date: Mon, 4 Nov 2019 13:25:31 +0000
Subject: [PATCH] Increase DBITS

---
 default.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/default.h b/default.h
index 055b8bc..b513ad7 100644
--- a/default.h
+++ b/default.h
@@ -1,4 +1,4 @@
 
 // Compile time setting, which determines the maximum number of input files.
 // Note: Maximun is 2^DBITS, and increasing DBITS will increase memory usage.
-#define DBITS 12
+#define DBITS 16
-- 
2.14.3 (Apple Git-98)

