From f8c60cd8328f71166c1c3d545a78310e3f5818f8 Mon Sep 17 00:00:00 2001
From: John Lees <lees.john6@gmail.com>
Date: Mon, 4 Nov 2019 13:44:44 +0000
Subject: [PATCH 2/2] CXX in Makefile

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index c7c4809..89510b2 100644
--- a/Makefile
+++ b/Makefile
@@ -14,6 +14,6 @@ clean:
 	$(RM) fsm-lite *.o *~
 
 depend:
-	g++ -MM -std=c++11 -I$(SDSL_INSTALL_PREFIX)/include *.cpp > dependencies.mk
+	$(CXX) -MM -std=c++11 -I$(SDSL_INSTALL_PREFIX)/include *.cpp > dependencies.mk
 
 include dependencies.mk
-- 
2.14.3 (Apple Git-98)

