diff --git a/src/biascorrection.cpp b/src/biascorrection.cpp
index 4a76abc..e3b0806 100644
--- a/src/biascorrection.cpp
+++ b/src/biascorrection.cpp
@@ -15,6 +15,7 @@
 
 #include <iostream>
 #include <fstream>
+#include <cmath>
 
 using namespace std;
 
@@ -425,7 +426,7 @@ void BiasLearner::getBias(const Scaffold& transcript, vector<double>& startBiase
 				}
 			}
 		}
-        assert(finite(startBias) && finite(endBias));
+        assert(std::isfinite(startBias) && std::isfinite(endBias));
 		startBiases[i] = startBias;
 		endBiases[i] = endBias;
 	}
