diff --git a/perl/rnazAnnotate.pl b/perl/rnazAnnotate.pl
index 4ba90e9..a25f88e 100755
--- perl/rnazAnnotate.pl
+++ perl/rnazAnnotate.pl
@@ -1,10 +1,11 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl
 
 # $Id: rnazAnnotate.pl,v 1.3 2006/03/24 15:43:13 wash Exp $
 
 use strict;
+use warnings;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;
diff --git a/perl/rnazBEDsort.pl b/perl/rnazBEDsort.pl
index 187b4fc..199f5ea 100755
--- perl/rnazBEDsort.pl
+++ perl/rnazBEDsort.pl
@@ -1,12 +1,13 @@
-#!/usr/local/bin/perl -w
+#!/usr/bin/env perl
 
 # $Id: rnazBEDsort.pl,v 1.2 2006/03/24 15:43:13 wash Exp $
 
 #missing: re-introduce track line, more intelligent sort of seq-ids (chr9 before chr10)
 
 use strict;
+use warnings;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;
diff --git a/perl/rnazBEDstats.pl b/perl/rnazBEDstats.pl
index 6cef5ae..8b2fd83 100755
--- perl/rnazBEDstats.pl
+++ perl/rnazBEDstats.pl
@@ -1,10 +1,11 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl
 
 # $Id: rnazBEDstats.pl,v 1.3 2008-01-24 10:26:45 wash Exp $
 
 use strict;
+use warnings;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;
diff --git a/perl/rnazBlast.pl b/perl/rnazBlast.pl
index 32ff7aa..75dd237 100755
--- perl/rnazBlast.pl
+++ perl/rnazBlast.pl
@@ -1,10 +1,11 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl
 
 # $Id: rnazBlast.pl,v 1.2 2006/03/24 15:43:13 wash Exp $
 
 use strict;
+use warnings;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;
diff --git a/perl/rnazCluster.pl b/perl/rnazCluster.pl
index d77995e..6a4c1fe 100755
--- perl/rnazCluster.pl
+++ perl/rnazCluster.pl
@@ -1,10 +1,11 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl
 
 # $Id: rnazCluster.pl,v 1.4 2008-01-24 10:26:45 wash Exp $
 
 use strict;
+use warnings;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;
diff --git a/perl/rnazFilter.pl b/perl/rnazFilter.pl
index 1a84aad..7509e3c 100755
--- perl/rnazFilter.pl
+++ perl/rnazFilter.pl
@@ -1,11 +1,12 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl
 
 # -*-Perl-*-
-# Last changed Time-stamp: <2009-02-23 18:28:47 ivo>
+# Last changed Time-stamp: <2018-12-30 12:12:50 fall>
 
 use strict;
+use warnings;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;
diff --git a/perl/rnazIndex.pl b/perl/rnazIndex.pl
index cad8a51..2ca6934 100755
--- perl/rnazIndex.pl
+++ perl/rnazIndex.pl
@@ -1,11 +1,12 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl
 
 # -*-Perl-*-
-# Last changed Time-stamp: <2009-02-23 18:32:12 ivo>
+# Last changed Time-stamp: <2018-12-30 12:12:55 fall>
 
 use strict;
+use warnings;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;
diff --git a/perl/rnazMAF2BED.pl b/perl/rnazMAF2BED.pl
index 5efa075..ad198ef 100755
--- perl/rnazMAF2BED.pl
+++ perl/rnazMAF2BED.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl
 
 # $Id: rnazMAF2BED.pl,v 1.3 2008-01-24 10:26:45 wash Exp $
 
@@ -6,8 +6,9 @@
 #TODO: handle UCSC style a la: hg17.chr17
 
 use strict;
+use warnings;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;
diff --git a/perl/rnazRandomizeAln.pl b/perl/rnazRandomizeAln.pl
index 4b3077b..9a36d42 100755
--- perl/rnazRandomizeAln.pl
+++ perl/rnazRandomizeAln.pl
@@ -1,10 +1,11 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl
 
 # $Id: rnazRandomizeAln.pl,v 1.2 2006/03/24 15:43:14 wash Exp $
 
 use strict;
+use warnings;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;
diff --git a/perl/rnazSelectSeqs.pl b/perl/rnazSelectSeqs.pl
index 662305a..d7c656b 100755
--- perl/rnazSelectSeqs.pl
+++ perl/rnazSelectSeqs.pl
@@ -1,10 +1,11 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl
 
 # $Id: rnazSelectSeqs.pl,v 1.3 2006/03/24 15:43:14 wash Exp $
 
 use strict;
+use warnings;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;
diff --git a/perl/rnazSort.pl b/perl/rnazSort.pl
index ed3c8db..46c0b00 100755
--- perl/rnazSort.pl
+++ perl/rnazSort.pl
@@ -1,10 +1,11 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl
 
 # $Id: rnazSort.pl,v 1.2 2006/03/24 15:43:14 wash Exp $
 
 use strict;
+use warnings;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;
diff --git a/perl/rnazWindow.pl b/perl/rnazWindow.pl
index a3b2dc1..37b24e5 100755
--- perl/rnazWindow.pl
+++ perl/rnazWindow.pl
@@ -1,7 +1,8 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl 
 
 use strict;
+use warnings;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;