--- perl/Makefile.am	2009-10-28 16:53:13.000000000 +0100
+++ perl/Makefile.am	2018-12-28 15:58:36.371446148 +0100
@@ -2,6 +2,6 @@
 
 docu_DATA = INSTALL README AUTHORS COPYING TODO manual.pdf
  
-EXTRA_DIST = INSTALL README AUTHORS COPYING TODO manual.pdf libsvm-2.89
+EXTRA_DIST = INSTALL README AUTHORS COPYING TODO manual.pdf libsvm-3.21
 
 docudir = $(pkgdatadir)

--- perl/rnazAnnotate.pl	2009-08-11 19:29:06.000000000 +0200
+++ perl/rnazAnnotate.pl	2018-12-29 18:26:17.627862153 +0100
@@ -1,10 +1,10 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl -w
 
 # $Id: rnazAnnotate.pl,v 1.3 2006/03/24 15:43:13 wash Exp $
 
 use strict;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;

--- perl/rnazBEDsort.pl	2009-08-11 19:29:06.000000000 +0200
+++ perl/rnazBEDsort.pl	2018-12-29 18:26:23.310800600 +0100
@@ -1,4 +1,4 @@
-#!/usr/local/bin/perl -w
+#!/usr/bin/env perl -w
 
 # $Id: rnazBEDsort.pl,v 1.2 2006/03/24 15:43:13 wash Exp $
 
@@ -6,7 +6,7 @@
 
 use strict;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;

--- perl/rnazBEDstats.pl	2009-08-11 19:29:06.000000000 +0200
+++ perl/rnazBEDstats.pl	2018-12-29 18:26:28.749741689 +0100
@@ -1,10 +1,10 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl -w
 
 # $Id: rnazBEDstats.pl,v 1.3 2008-01-24 10:26:45 wash Exp $
 
 use strict;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;

--- perl/rnazBlast.pl	2009-08-11 19:29:06.000000000 +0200
+++ perl/rnazBlast.pl	2018-12-29 18:26:34.436680090 +0100
@@ -1,10 +1,10 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl -w
 
 # $Id: rnazBlast.pl,v 1.2 2006/03/24 15:43:13 wash Exp $
 
 use strict;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;

--- perl/rnazCluster.pl	2009-08-11 19:37:02.000000000 +0200
+++ perl/rnazCluster.pl	2018-12-29 18:26:39.814619114 +0100
@@ -1,10 +1,10 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl -w
 
 # $Id: rnazCluster.pl,v 1.4 2008-01-24 10:26:45 wash Exp $
 
 use strict;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;

--- perl/rnazFilter.pl	2009-08-11 19:29:07.000000000 +0200
+++ perl/rnazFilter.pl	2018-12-29 18:26:45.423551138 +0100
@@ -1,11 +1,11 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl -w
 
 # -*-Perl-*-
-# Last changed Time-stamp: <2009-02-23 18:28:47 ivo>
+# Last changed Time-stamp: <2018-12-29 18:26:45 fall>
 
 use strict;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;

--- perl/rnazIndex.pl	2009-08-11 19:29:07.000000000 +0200
+++ perl/rnazIndex.pl	2018-12-29 18:26:50.927484436 +0100
@@ -1,11 +1,11 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl -w
 
 # -*-Perl-*-
-# Last changed Time-stamp: <2009-02-23 18:32:12 ivo>
+# Last changed Time-stamp: <2018-12-29 18:26:50 fall>
 
 use strict;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;

--- perl/rnazMAF2BED.pl	2009-08-11 19:29:07.000000000 +0200
+++ perl/rnazMAF2BED.pl	2018-12-29 18:26:56.883412255 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl -w
 
 # $Id: rnazMAF2BED.pl,v 1.3 2008-01-24 10:26:45 wash Exp $
 
@@ -7,7 +7,7 @@
 
 use strict;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;

--- perl/rnazRandomizeAln.pl	2009-08-11 19:29:07.000000000 +0200
+++ perl/rnazRandomizeAln.pl	2018-12-29 18:27:04.425320853 +0100
@@ -1,10 +1,10 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl -w
 
 # $Id: rnazRandomizeAln.pl,v 1.2 2006/03/24 15:43:14 wash Exp $
 
 use strict;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;

--- perl/rnazSelectSeqs.pl	2009-08-11 19:29:07.000000000 +0200
+++ perl/rnazSelectSeqs.pl	2018-12-29 18:27:10.054252636 +0100
@@ -1,10 +1,10 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl -w
 
 # $Id: rnazSelectSeqs.pl,v 1.3 2006/03/24 15:43:14 wash Exp $
 
 use strict;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;

--- perl/rnazSort.pl	2009-08-11 19:29:07.000000000 +0200
+++ perl/rnazSort.pl	2018-12-29 18:26:09.782947122 +0100
@@ -1,10 +1,10 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl -w
 
 # $Id: rnazSort.pl,v 1.2 2006/03/24 15:43:14 wash Exp $
 
 use strict;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;

--- perl/rnazWindow.pl	2010-01-04 16:25:06.000000000 +0100
+++ perl/rnazWindow.pl	2018-12-29 18:26:03.297017372 +0100
@@ -1,8 +1,8 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl -w
 
 use strict;
 use FindBin;
-use lib $FindBin::Bin;
+use lib "$FindBin::Bin/../lib/5.26.2";
 use RNAz;
 use Getopt::Long;
 use Pod::Usage;

--- rnaz/Makefile.am	2009-06-04 13:45:10.000000000 +0200
+++ rnaz/Makefile.am	2018-12-28 15:58:37.744444795 +0100
@@ -2,11 +2,11 @@
 
 RNAz_SOURCES=RNAz.c rnaz_utils.c svm_helper.c rnaz_utils.h svm_helper.h zscore.c zscore.h cmdline.c cmdline.h strand.c strand.h  ../models/mfe_avg.inc  ../models/mfe_stdv.inc  ../models/strand.inc ../models/decision.inc 
 
-nodist_RNAz_SOURCES=../libsvm-2.89/svm.cpp 
+nodist_RNAz_SOURCES=${PREFIX}/lib/libsvm-3.21/svm.cpp 
 
 LDADD = -L../librna -lRNA -lm
  
 RNAz_LINK = $(CXX) -o $@
  
-INCLUDES = -I../librna -I../libsvm-2.89 -I../models
+INCLUDES = -I../librna -I${PREFIX}/lib/libsvm-3.21 -I../models
 
--- rnaz/strand.c	2009-08-19 14:04:39.000000000 +0200
+++ rnaz/strand.c	2018-12-28 15:54:20.928697847 +0100
@@ -123,9 +123,9 @@
 	    ++j;
 	}
 	singleMFE = fold(woGapsSeq, singleStruc);
-	singleZ=mfe_zscore(woGapsSeq,singleMFE,z_score_type,0,warningString_strand);
+	singleZ=mfe_zscore(woGapsSeq,singleMFE,&z_score_type,0,warningString_strand);
 	singleMFE_reverse = fold(woGapsSeq_reverse, singleStruc_reverse);
-	singleZ_reverse=mfe_zscore(woGapsSeq_reverse,singleMFE_reverse,z_score_type,0,warningString_strand);
+	singleZ_reverse=mfe_zscore(woGapsSeq_reverse,singleMFE_reverse,&z_score_type,0,warningString_strand);
 	sumZ+=singleZ;
 	sumMFE+=singleMFE;
 	sumZ_reverse+=singleZ_reverse;

--- rnaz/svm_helper.c	2009-08-17 13:09:57.000000000 +0200
+++ rnaz/svm_helper.c	2018-12-28 15:52:50.052787389 +0100
@@ -251,6 +251,7 @@
 */
 
 
+/*
 struct svm_model{
   struct svm_parameter param;
   int nr_class;
@@ -263,7 +264,8 @@
   int *label;
   int *nSV;
   int free_sv;
-};
+  };
+*/
 
 struct svm_model* svm_load_model_string(char *modelString){
 
--- rnaz/svm_helper.h	2009-08-17 13:09:52.000000000 +0200
+++ rnaz/svm_helper.h	2018-12-28 15:51:32.059864233 +0100
@@ -11,6 +11,7 @@
  *                                                                   *
  *********************************************************************/
 
+#define svm_destroy_model(A) svm_free_model_content((A))
 
 void get_regression_models(struct svm_model** avg_model,
 			   struct svm_model** stdv_model,
