diff --git a/CMakeLists.txt b/CMakeLists.txt
index c008abe..5d0fa41 100755
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -48,9 +48,9 @@ add_custom_command(TARGET copy_data
 
 find_package (Threads REQUIRED)
 
-set(Boost_USE_STATIC_LIBS        ON)  # only find static libs
+set(Boost_USE_STATIC_LIBS        OFF)  # only find static libs
 set(Boost_USE_MULTITHREADED      ON)
-set(Boost_USE_STATIC_RUNTIME     ON)
+set(Boost_USE_STATIC_RUNTIME     OFF)
 find_package(Boost 1.5 COMPONENTS iostreams program_options filesystem system REQUIRED)
 
 include_directories(${ZLIB_INCLUDE_DIR})
