diff --git a/Makefile b/Makefile
index fc81a1a..42d9f17 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
-CC  ?= gcc
-CXX ?= g++
+CC  ?= $(CC)
+CXX ?= $(CXX)
 
-LIBS = -lz -lm -lbz2 -llzma -lpthread -lcurl -lgsl -lgslcblas
+LIBS = -L$(PREFIX)/lib -lz -lm -lbz2 -llzma -pthread -lcurl -lgsl -lgslcblas
 
 CRYPTO_TRY=$(shell echo 'int main(){}'|$(CXX) -x c++ - -lcrypto 2>/dev/null -o /dev/null; echo $$?)
 ifeq "$(CRYPTO_TRY)" "0"
@@ -25,7 +25,7 @@ else
 #if hts source path is given
 # Adjust $(HTSSRC) to point to your top-level htslib directory
 $(info HTSSRC defined: $(HTSSRC))
-CPPFLAGS += -I"$(realpath $(HTSSRC))"
+CPPFLAGS += -I"$(realpath $(HTSSRC))" -I$(PREFIX)/include
 LIBHTS := $(HTSSRC)/libhts.a
 LIBS := $(LIBHTS) $(LIBS)
 
@@ -40,7 +40,7 @@ $(info Use `make HTSSRC=systemwide` to build metadamage using the systemwide hts
 
 
 HTSSRC := $(CURDIR)/htslib
-CPPFLAGS += -I$(HTSSRC)
+CPPFLAGS += -I$(HTSSRC) -I$(PREFIX)/include
 LIBHTS := $(HTSSRC)/libhts.a
 LIBS := $(LIBHTS) $(LIBS)
 
