diff --git a/Makefile b/Makefile
index 0f35f0e..f5376a8 100644
--- a/Makefile
+++ b/Makefile
@@ -1,10 +1,10 @@
 # make FLAGS=-D__REGRESSION__
 #c++ -I /opt/homebrew/Cellar/eigen/3.4.0_1/include/ -c -I/Users/fvr124/metaDMG-cpp/htslib   regression.cpp -std=c++14  -D__REGRESSION__
 
-CC  ?= gcc
-CXX ?= g++
+CC  ?= $(CC)
+CXX ?= $(CXX)
 
-LIBS = -lz -lm -lbz2 -llzma -lpthread -lcurl -lgsl -lgslcblas
+LIBS = -L$(PREFIX)/lib -lz -lm -lbz2 -llzma -pthread -lcurl -lgsl -lgslcblas
 
 CRYPTO_TRY=$(shell echo 'int main(){}'|$(CXX) -x c++ - -lcrypto 2>/dev/null -o /dev/null; echo $$?)
 ifeq "$(CRYPTO_TRY)" "0"
@@ -28,7 +28,7 @@ else
 #if hts source path is given
 # Adjust $(HTSSRC) to point to your top-level htslib directory
 $(info HTSSRC defined: $(HTSSRC))
-CPPFLAGS += -I"$(realpath $(HTSSRC))"
+CPPFLAGS += -I"$(realpath $(HTSSRC))" -I$(PREFIX)/include
 LIBHTS := $(HTSSRC)/libhts.a
 LIBS := $(LIBHTS) $(LIBS)
 
@@ -43,7 +43,7 @@ $(info Use `make HTSSRC=systemwide` to build metadamage using the systemwide hts
 
 
 HTSSRC := $(CURDIR)/htslib
-CPPFLAGS += -I$(HTSSRC)
+CPPFLAGS += -I$(HTSSRC) -I$(PREFIX)/include
 LIBHTS := $(HTSSRC)/libhts.a
 LIBS := $(LIBHTS) $(LIBS)
 
@@ -51,7 +51,7 @@ all: .activate_module
 
 endif
 
-.PHONY: .activate_module 
+.PHONY: .activate_module
 
 .activate_module:
 	git submodule update --init --recursive
@@ -60,7 +60,7 @@ endif
 
 
 #modied from htslib makefile
-FLAGS = -O3 
+FLAGS = -O3
 CPPFLAGS := $(filter-out -DNDEBUG,$(CPPFLAGS))
 FLAGS2 = $(CPPFLAGS) $(FLAGS) $(LDFLAGS)
 
@@ -100,7 +100,7 @@ version.h:
 
 .PHONY: all clean install install-all install-misc misc test
 
-misc: 
+misc:
 	$(MAKE) -C misc HTSSRC="$(realpath $(HTSSRC))"
 
 -include $(OBJ:.o=.d)
@@ -121,7 +121,7 @@ metaDMG-cpp: version.h $(OBJ)
 testclean:
 	rm -f test/acc2taxid.map.gzf570b1db7c.dedup.filtered.rname.bam.bin
 	rm -f test/data/f570b1db7c.dedup.filtered.rname.bam
-	rm -rf test/output test/logfile version.h 
+	rm -rf test/output test/logfile version.h
 
 clean: testclean
 	rm  -f *.o *.d $(PROGRAMS) version.h *~
@@ -129,7 +129,7 @@ clean: testclean
 
 test:
 	echo "Running unit tests for metaDMG"
-	cd test;./testAll.sh 
+	cd test;./testAll.sh
 
 force:
 
