--- setup.py.orig	2016-10-24 13:09:41.968580382 +0000
+++ setup.py	2016-10-24 13:10:06.104581534 +0000
@@ -49,12 +49,11 @@
         'Programming Language :: Python :: 3',
         'Programming Language :: Python :: 3.4'
     ],
-    'install_requires': ['nose',
-                         'pymongo==3.2',
+    'install_requires': [
+                         'pymongo',
                          'pycurl',
                          'ldap3',
-                         'mock',
-                         'py-bcrypt',
+                         'bcrypt',
                          'drmaa',
                          'future',
                          'tabulate',
