--- tksrc/gmtkModelInfo.cc	2019-04-30 17:27:10.721857453 -0400
+++ tksrc/gmtkModelInfo.cc	2019-04-30 17:27:27.429641923 -0400
@@ -635,9 +635,9 @@
 #define BF "%u"
 #define PB(b) \
 switch (mode) {\
- case terse: fprintf(outF, "%s"BF, sep, ((b) ? 1 : 0)); break; \
- case normal: fprintf(outF, "%s%s = "BF, sep, names[outputOrder[i]], ((b) ? 1 : 0)); break; \
- case verbose: fprintf(outF, "%s%s = "BF, sep, names[outputOrder[i]], ((b) ? 1 : 0)); \
+ case terse: fprintf(outF, "%s" BF, sep, ((b) ? 1 : 0)); break; \
+ case normal: fprintf(outF, "%s%s = " BF, sep, names[outputOrder[i]], ((b) ? 1 : 0)); break; \
+ case verbose: fprintf(outF, "%s%s = " BF, sep, names[outputOrder[i]], ((b) ? 1 : 0)); \
 }
 
 #define PU(x) \
