diff --git a/ext/dsk/minia/SortingCount.cpp b/ext/dsk/minia/SortingCount.cpp
index 2ad6e69..e2d7133 100644
--- a/ext/dsk/minia/SortingCount.cpp
+++ b/ext/dsk/minia/SortingCount.cpp
@@ -544,7 +544,7 @@ void sorting_count(Bank *Sequences, char *prefix, int max_memory, int max_disk_s
         // TODO to guillaume: remove that todo above, because it is done, right?
 #if OMP 
         //omp_set_numthreads(2);  //num_threads(2) //if(!output_histo) num_threads(nb_threads)
-#pragma omp parallel for if (! separate_count) private (p)  num_threads(nb_threads)
+#pragma omp parallel for if (! separate_count) num_threads(nb_threads)
 #endif        
         // load, sort each partition to output solid kmers
         for (int p=0;p<nb_partitions;p++)
