--- samtools-0.1.9/misc/Makefile	2015-08-10 21:46:55.000000000 +0000
+++ samtools-0.1.9/misc/Makefile.new	2015-12-17 20:19:24.563431278 +0000
@@ -1,11 +1,11 @@
 CC=			gcc
 CXX=		g++
-CFLAGS=		-g -Wall -O2 #-m64 #-arch ppc
+CFLAGS=		-g -Wall -O2 -I${PREFIX}/include -L${PREFIX}/lib #-m64 #-arch ppc
 CXXFLAGS=	$(CFLAGS)
 DFLAGS=		-D_FILE_OFFSET_BITS=64
 OBJS=		
 PROG=		md5sum-lite md5fa maq2sam-short maq2sam-long wgsim
-INCLUDES=	-I..
+INCLUDES=	-I.. -I${PREFIX}/include
 SUBDIRS=	.
 
 .SUFFIXES:.c .o

--- samtools-0.1.9/Makefile	2015-12-17 20:44:01.578167339 +0000
+++ samtools-0.1.9/Makefile.new	2015-12-17 20:44:18.199175335 +0000
@@ -1,5 +1,5 @@
 CC=			gcc
-CFLAGS=		-g -Wall -O2 #-m64 #-arch ppc
+CFLAGS=		-g -Wall -O2 -L${PREFIX}/lib -I${PREFIX}/include #-m64 #-arch ppc
 DFLAGS=		-D_FILE_OFFSET_BITS=64 -D_USE_KNETFILE -D_CURSES_LIB=1
 KNETFILE_O=	knetfile.o
 LOBJS=		bgzf.o kstring.o bam_aux.o bam.o bam_import.o sam.o bam_index.o	\
@@ -9,10 +9,10 @@
 			bam_rmdup.o bam_rmdupse.o bam_mate.o bam_stat.o bam_color.o	\
 			bamtk.o kaln.o bam2bcf.o errmod.o sample.o
 PROG=		samtools
-INCLUDES=	-I.
+INCLUDES=	-I. -I${PREFIX}/include -I${PREFIX}/include/ncurses
 SUBDIRS=	. bcftools misc
-LIBPATH=
-LIBCURSES=	-lcurses # -lXCurses
+LIBPATH=	-L${PREFIX}/lib
+LIBCURSES=	-lncurses # -lXCurses
 
 .SUFFIXES:.c .o
 
