From d40ca9338aee90ffe8efb68a7a50f0d3e49d1e34 Mon Sep 17 00:00:00 2001
From: Afif Elghraoui <afif@ghraoui.name>
Date: Sat, 20 Feb 2021 15:13:06 -0800
Subject: [PATCH 1/4] central-lastz

---
 cassis.pl | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/cassis.pl b/cassis.pl
index 7d19d49..0ec9167 100755
--- a/cassis.pl
+++ b/cassis.pl
@@ -49,7 +49,8 @@ my $MINIMUM_SEQUENCE_SIZE_BLOCKS = 50000;
 
 # LASTZ - DEFAULT VALUES
 my $LASTZ_LEVEL             = 2;
-my $LASTZ_PATH              = "${Bin}/lastz/lastz";
+my $LASTZ_PATH              = `which lastz`;
+chomp($LASTZ_PATH);
 my $LASTZ_LEVEL1_PARAMETERS = "K=3000 E=30 O=400 L=2200 H=2000 B=2";
 my $LASTZ_LEVEL2_PARAMETERS = "K=3000 E=30 O=400 L=3000 H=2000 B=2";
 my $LASTZ_LEVEL3_PARAMETERS = "K=2200 L=6000 H=2000 B=2";
-- 
2.30.1

