From 71cbb374d0a722529e5938e359e8b469bfeed943 Mon Sep 17 00:00:00 2001
From: Afif Elghraoui <afif@ghraoui.name>
Date: Sat, 20 Feb 2021 15:15:49 -0800
Subject: [PATCH 2/4] cassis-shebang

---
 cassis.pl | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/cassis.pl b/cassis.pl
index 0ec9167..3066744 100755
--- a/cassis.pl
+++ b/cassis.pl
@@ -1,4 +1,6 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl
+
+use warnings;
 
 ###############################################################################
 # Copyright (C) 2010 Université Claude Bernard Lyon 1                         #
-- 
2.30.1

