From a5ad9cc24e20a3311c4aaf5e8bbc8aee2e113274 Mon Sep 17 00:00:00 2001
From: Afif Elghraoui <afif@ghraoui.name>
Date: Sat, 20 Feb 2021 20:32:50 -0800
Subject: [PATCH 3/4] Adapt paths to lib directory

---
 cassis.pl | 19 ++++++++++---------
 1 file changed, 10 insertions(+), 9 deletions(-)

diff --git a/cassis.pl b/cassis.pl
index 3066744..017fc62 100755
--- a/cassis.pl
+++ b/cassis.pl
@@ -28,7 +28,8 @@ use warnings;
 use strict;
 use File::Spec;
 use FindBin qw($Bin);
-require "${Bin}/core/util.pm";
+my $LIB = "${Bin}/../lib/cassis";
+require "${LIB}/core/util.pm";
 
 ###############################################################################
 ## CONSTANTS ##################################################################
@@ -56,9 +57,9 @@ chomp($LASTZ_PATH);
 my $LASTZ_LEVEL1_PARAMETERS = "K=3000 E=30 O=400 L=2200 H=2000 B=2";
 my $LASTZ_LEVEL2_PARAMETERS = "K=3000 E=30 O=400 L=3000 H=2000 B=2";
 my $LASTZ_LEVEL3_PARAMETERS = "K=2200 L=6000 H=2000 B=2";
-my $LASTZ_LEVEL1_MATRIX     = "${Bin}/lastz/matrix-level1.txt";
+my $LASTZ_LEVEL1_MATRIX     = "${LIB}/lastz/matrix-level1.txt";
 my $LASTZ_LEVEL2_MATRIX     = "default";
-my $LASTZ_LEVEL3_MATRIX     = "${Bin}/lastz/matrix-level3.txt";
+my $LASTZ_LEVEL3_MATRIX     = "${LIB}/lastz/matrix-level3.txt";
 
 ###############################################################################
 ## MAIN PROGRAM ###############################################################
@@ -519,13 +520,13 @@ sub validateParameters {
 # This function verifies the environment (scripts, etc...)
 sub verifyEnvironment {
 
-	my $verifyDirScript   = fileExists( "${Bin}/core/verifyFASTAdirectory.pl",       \&printUsage );
-	my $identifyScript    = fileExists( "${Bin}/core/identifyBreakpoints.pl",        \&printUsage );
-	my $createFASTAScript = fileExists( "${Bin}/pipeline/createBreakpointFastas.pl", \&printUsage );
-	my $alignScript       = fileExists( "${Bin}/pipeline/alignBreakpointRegions.pl", \&printUsage );
-	my $dotplotScript = fileExists( "${Bin}/pipeline/dotplotBreakpointRegions.pl", \&printUsage );
+	my $verifyDirScript   = fileExists( "${LIB}/core/verifyFASTAdirectory.pl",       \&printUsage );
+	my $identifyScript    = fileExists( "${LIB}/core/identifyBreakpoints.pl",        \&printUsage );
+	my $createFASTAScript = fileExists( "${LIB}/pipeline/createBreakpointFastas.pl", \&printUsage );
+	my $alignScript       = fileExists( "${LIB}/pipeline/alignBreakpointRegions.pl", \&printUsage );
+	my $dotplotScript = fileExists( "${LIB}/pipeline/dotplotBreakpointRegions.pl", \&printUsage );
 	my $segmentationScript =
-	  fileExists( "${Bin}/pipeline/breakpointSegmentation.pl", \&printUsage );
+	  fileExists( "${LIB}/pipeline/breakpointSegmentation.pl", \&printUsage );
 
 	$LASTZ_PATH = fileExists( $LASTZ_PATH, \&printUsage );
 
-- 
2.30.1

