--- NBICseq-norm.pl.old	2019-04-23 19:33:54.057835803 -0500
+++ NBICseq-norm.pl	2019-04-24 15:03:41.402332319 -0500
@@ -1,23 +1,23 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 use strict;
 
 use FindBin qw($Bin);
 my $path = $Bin;
 
-my $brs = "$path/BRS_1bp/BRS";
-my $randomsample = "$path/randomSample/random_split";
-my $prepPois = "$path/DataPrepare/PrepPois";
-my $prepPoisGAM = "$path/DataPrepareGAM/PrepPoisGAM";
-my $rnormalize = "$path/R/normalize.R";
-my $rrefine = "$path/R/refine.R";
-my $rrefineGAM = "$path/R/refineGAM.R";
-my $rplot_function = "$path/R/plot_RC_vs_GC.R";
-my $compRatio = "$path/R/compRatio.R";
-my $markCNV = "$path/R/purity.R";
-my $purityEM = "$path/purityEM/purityEM";
-my $filterCNV = "$path/Filter/filterCNV";
-my $rnormalizeGAM = "$path/R/normalizeGAM.R";
-my $testmgcv = "$path/R/test.mgcv.installed.R";
+my $brs = "$path/BRS";
+my $randomsample = "$path/random_split";
+my $prepPois = "$path/PrepPois";
+my $prepPoisGAM = "$path/PrepPoisGAM";
+my $rnormalize = "$path/normalize.R";
+my $rrefine = "$path/refine.R";
+my $rrefineGAM = "$path/refineGAM.R";
+my $rplot_function = "$path/plot_RC_vs_GC.R";
+my $compRatio = "$path/compRatio.R";
+my $markCNV = "$path/purity.R";
+my $purityEM = "$path/purityEM";
+my $filterCNV = "$path/filterCNV";
+my $rnormalizeGAM = "$path/normalizeGAM.R";
+my $testmgcv = "$path/test.mgcv.installed.R";
 
 
 if(!(-e $brs)){die"Cannot find the file $brs\n";}
