diff --git a/src/pyOpenMS/CMakeLists.txt b/src/pyOpenMS/CMakeLists.txt
index 480fa9160a..413c24a467 100644
--- a/src/pyOpenMS/CMakeLists.txt
+++ b/src/pyOpenMS/CMakeLists.txt
@@ -477,7 +477,7 @@ ENDIF()
 
 add_dependencies(pyopenms_build OpenMS)
 
-if(APPLE)
+if(FALSE)
 add_custom_command(
   TARGET pyopenms_build POST_BUILD
   COMMAND ${CMAKE_SOURCE_DIR}/cmake/MacOSX/fix_dependencies.rb -l ${CMAKE_BINARY_DIR}/pyOpenMS/pyopenms -e "@loader_path/"
