--- src/pyOpenMS/pyopenms/__init__.py	2020-01-26 17:08:44.000000000 +0100
+++ src/pyOpenMS/pyopenms/__init__.py.patched	2020-01-28 19:17:06.000000000 +0100
@@ -27,20 +27,6 @@
 from .sysinfo import *
 from .version import version as __version__
 
-import os
-here = os.path.abspath(os.path.dirname(__file__))
-os.environ["OPENMS_DATA_PATH"] = os.path.join(here, "share/OpenMS")
-
-import sys
-if sys.platform.startswith("linux"):
-    # load local shared libries before we import pyopenms.so, else
-    # those are not found. setting LD_LIBRARY_PATH does not work,
-    # see: http://stackoverflow.com/questions/1178094
-    import ctypes
-    ctypes.cdll.LoadLibrary(os.path.join(here, "libOpenSwathAlgo.so"))
-    ctypes.cdll.LoadLibrary(os.path.join(here, "libOpenMS.so"))
-    ctypes.cdll.LoadLibrary(os.path.join(here, "libSuperHirn.so"))
-
 try:
     from .all_modules import *
     from .python_extras import *
@@ -80,5 +66,3 @@
     print("="*70)
     print("\n")
     raise e
-
-del os, here, sys
