--- ./src/pyOpenMS/setup.py	2018-10-29 20:07:55.000000000 +0100
+++ ./src/pyOpenMS/setup.py	2019-12-18 12:21:21.000000000 +0100
@@ -63,7 +63,8 @@
     #  - note that we are missing the self.cython_sources line, so this will not work under all circumstances
     # First, sanity-check the 'extensions' list
     self.check_extensions_list(self.extensions)
-    list(multiprocessing.pool.ThreadPool(int(PY_NUM_THREADS)).imap(self.build_extension, self.extensions))
+    mypool = multiprocessing.pool.ThreadPool(int(PY_NUM_THREADS))
+    list(mypool.imap(self.build_extension, self.extensions))
 if not single_threaded:
     import distutils.command.build_ext
     distutils.command.build_ext.build_ext.build_extensions = parallel_build_extensions
@@ -152,6 +153,7 @@
 elif sys.platform == "darwin":
     # we need to manually link to the Qt Frameworks
     extra_compile_args = ["-Qunused-arguments"]
+    extra_compile_args.append("-isysroot"+os.environ['CONDA_BUILD_SYSROOT'])
 
 if IS_DEBUG:
     extra_compile_args.append("-g2")
