--- segemehl/Makefile
+++ segemehl/Makefile
@@ -190,24 +190,24 @@
 
 haarz.x: ${TESTMATOBJ}
-	                gcc $(CFLAGS) -DGSLACTIVE -c -o libs/evalmatchfiles.o libs/evalmatchfiles.c
-					gcc $(CFLAGS) ${TESTMATOBJ} -o $@ $(LDFLAGS) -lform -lmenu -lgsl -lgslcblas
+	                $(CC) $(CFLAGS) -DGSLACTIVE -c -o libs/evalmatchfiles.o libs/evalmatchfiles.c
+					$(CC) $(CFLAGS) ${TESTMATOBJ} -o $@ $(LDFLAGS) -lform -lmenu -lgsl -lgslcblas
 
 
 segemehl.x: ${SEGEMEHLOBJ}
 					echo " " >> src/version.h
-					gcc $(CFLAGS) ${SEGEMEHLOBJ} -o $@ $(LDFLAGS)
+					$(CC) $(CFLAGS) ${SEGEMEHLOBJ} -o $@ $(LDFLAGS)
 
 genfasta.x: ${GENFASTAOBJ}
-					gcc $(CFLAGS) ${GENFASTAOBJ} -o $@ $(LDFLAGS)
+					$(CC) $(CFLAGS) ${GENFASTAOBJ} -o $@ $(LDFLAGS)
 
 
 testrealign.x: ${TESTREALIGNOBJ}
-					gcc $(CFLAGS) -DREALIGNTEST -c -o libs/realign.o libs/realign.c
-					gcc $(CFLAGS) ${TESTREALIGNOBJ} -o $@ $(LDFLAGS)
+					$(CC) $(CFLAGS) -DREALIGNTEST -c -o libs/realign.o libs/realign.c
+					$(CC) $(CFLAGS) ${TESTREALIGNOBJ} -o $@ $(LDFLAGS)
 
 lack.x: ${TESTREMAPPINGOBJ}					
-					gcc $(CFLAGS) -c -o libs/realign.o libs/realign.c
-					gcc $(CFLAGS) -DREMAPPINGTEST -c -o libs/remapping.o libs/remapping.c
-					gcc $(CFLAGS) ${TESTREMAPPINGOBJ} -o $@ $(LDFLAGS)
+					$(CC) $(CFLAGS) -c -o libs/realign.o libs/realign.c
+					$(CC) $(CFLAGS) -DREMAPPINGTEST -c -o libs/remapping.o libs/remapping.c
+					$(CC) $(CFLAGS) ${TESTREMAPPINGOBJ} -o $@ $(LDFLAGS)
 
 clean: 
