--- setup.py    2016-11-27 08:36:20.000000000 +0100
+++ setup.py    2016-11-27 08:40:04.000000000 +0100
@@ -19,7 +19,7 @@
     # Project uses reStructuredText, so ensure that the
     # docutils get installed or upgraded on the target
     # machine
-    install_requires = ['rpy2', 'numpy', 'matplotlib', 'argparse', 'pandas', 'biopython'],
+    install_requires = ['rpy2', 'numpy', 'matplotlib', 'pandas', 'biopython'],

     package_data = {
         # If any package contains *.txt or *.rst files,
