--- a/cmake/lz4.cmake
+++ b/cmake/lz4.cmake
@@ -2,13 +2,14 @@
 
 ExternalProject_Add(LZ4
     PREFIX ${external_bindir}/LZ4
     SOURCE_SUBDIR build/cmake
     SOURCE_DIR ${external_dir}/lz4
     DOWNLOAD_COMMAND ""
     CMAKE_ARGS -DBUILD_STATIC_LIBS=ON -DBUILD_SHARED_LIBS=OFF
                -DCMAKE_BUILD_TYPE=Release -DCMAKE_VERBOSE_MAKEFILE=ON
                -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER} -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
     INSTALL_COMMAND ${CMAKE_COMMAND} -E echo "Using existing lz4 library"
     LOG_CONFIGURE ON
     LOG_BUILD ON
 )
 