--- a/cmake/fmt.cmake
+++ b/cmake/fmt.cmake
@@ -2,10 +2,10 @@
 
 ExternalProject_Add(FMT
   PREFIX ${external_bindir}/FMT
-  SOURCE_DIR ${external_dir}/fmt
-  CMAKE_ARGS -DCMAKE_BUILD_TYPE=Release -DCMAKE_VERBOSE_MAKEFILE=ON
+  SOURCE_DIR ${external_dir}/fmt
+  CMAKE_ARGS -DCMAKE_BUILD_TYPE=Release -DCMAKE_VERBOSE_MAKEFILE=ON
              -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER} -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
-  INSTALL_COMMAND ""
+  INSTALL_COMMAND "${CMAKE_COMMAND}" -E echo "Using existing fmt library"
   LOG_CONFIGURE ON
   LOG_BUILD ON
 )
