*** src/cmake/boost.cmake	2019-06-28 22:04:47.000000000 +0000
--- src/cmake/boost.cmake	2024-12-02 11:51:43.100697447 +0000
*************** macro (initBoostParams)
*** 44,50 ****
      # for test, so we need two lists now:
      set (THIS_BOOST_BUILD_COMPONENTS date_time filesystem program_options
                                       regex serialization system timer chrono test)
!     set (Boost_USE_STATIC_LIBS ON)
      if (NOT WIN32)
          # bjam on windows ignores this setting so skip for win32:
          set (Boost_USE_MULTITHREADED OFF)
--- 44,50 ----
      # for test, so we need two lists now:
      set (THIS_BOOST_BUILD_COMPONENTS date_time filesystem program_options
                                       regex serialization system timer chrono test)
!     set (Boost_USE_STATIC_LIBS OFF)
      if (NOT WIN32)
          # bjam on windows ignores this setting so skip for win32:
          set (Boost_USE_MULTITHREADED OFF)
