diff --git i/Cargo.toml w/Cargo.toml
index eeefa95..e8a7b28 100644
--- i/Cargo.toml
+++ w/Cargo.toml
@@ -35,7 +35,7 @@ categories = ["command-line-utilities", "science"]
 [dependencies]
 # for local development, look in the libradicl git repository
 # but when published, pull the specified version
-libradicl = { git = "https://github.com/COMBINE-lab/libradicl", branch = "develop", version = "0.8.2" }
+libradicl = "0.8.2"
 anyhow = "1.0.80"
 arrayvec = "0.7.4"
 ahash = "0.8.11"
