From 840e8bbe30c22e1221d273b51dc37f50214a1b03 Mon Sep 17 00:00:00 2001
From: John Lees <lees.john6@gmail.com>
Date: Wed, 20 Feb 2019 17:21:28 -0500
Subject: [PATCH 1/2] use no apple in apegrunt

---
 src/CMakeLists.txt | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 1b75c38..4882dd4 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -101,10 +101,10 @@ if( NOT APEGRUNT_NO_TBB )
 	target_link_libraries( apegrunt ${TBB_LIBRARIES} )
 endif()
 
-if( UNIX )
+if( NOT APPLE )
 	target_link_libraries( libapegrunt pthread rt )
 	target_link_libraries( apegrunt pthread rt )
-endif( UNIX )
+endif( NOT APPLE )
 
 # Prevent linking against shared libraries on OS X;
 # Apple gcc always links against a shared version of a library if present,
-- 
2.14.3 (Apple Git-98)

