From 315c05dddb3fff0159b862377dec7ad1a4dbd13a Mon Sep 17 00:00:00 2001
From: John Lees <lees.john6@gmail.com>
Date: Mon, 4 Feb 2019 13:21:42 -0500
Subject: [PATCH] Add librt to linker line

---
 src/CMakeLists.txt | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 90224d5..214b94f 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -83,6 +83,8 @@ endif()
 
 # Add pthreads library if building in UNIX
 if( UNIX )
+	target_link_libraries( SpydrPick rt )
+	target_link_libraries( aracne rt )
 	target_link_libraries( SpydrPick pthread )
 	target_link_libraries( aracne pthread )
 endif( UNIX )
-- 
2.14.3 (Apple Git-98)

