From 0b8cc4b3904a3ab079fc6ebdbbc27cc35dd9cb25 Mon Sep 17 00:00:00 2001
From: John Lees <lees.john6@gmail.com>
Date: Wed, 20 Feb 2019 17:10:47 -0500
Subject: [PATCH 2/2] Use no apple in Cmakelists

---
 src/CMakeLists.txt | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 090e60d..08530e1 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -74,9 +74,9 @@ if( NOT SPYDRPICK_NO_TBB )
 endif()
 
 # Add rt and pthreads library if building in UNIX
-if( UNIX )
-	target_link_libraries( SpydrPick pthread rt )
-endif( UNIX )
+if( NOT APPLE )
+   target_link_libraries( SpydrPick pthread rt )
+endif( NOT APPLE )
 
 # Prevent linking against shared libraries on OS X;
 # Apple gcc always links against a shared version of a library if present,
-- 
2.14.3 (Apple Git-98)

