From 7e2a8159b5091ceaad5c95f8ec3b63ea03147d8b Mon Sep 17 00:00:00 2001
From: John Lees <lees.john6@gmail.com>
Date: Thu, 21 Feb 2019 14:42:11 -0500
Subject: [PATCH 3/3] Change AVX to SSE3

---
 src/CMakeLists.txt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 08530e1..6591ec2 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -30,7 +30,7 @@ set( SPYDRPICK_SOURCES
 set( CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin )
 
 # general optimization flags	
-set( SPYDRPICK_GCC_OPTIMIZATION_FLAGS "${SPYDRPICK_GCC_OPTIMIZATION_FLAGS} -O3 -mavx2 -static-libstdc++" )
+set( SPYDRPICK_GCC_OPTIMIZATION_FLAGS "${SPYDRPICK_GCC_OPTIMIZATION_FLAGS} -O3 -msse3 -static-libstdc++" )
 set( SPYDRPICK_GCC_OPTIMIZATION_FLAGS "${SPYDRPICK_GCC_OPTIMIZATION_FLAGS} -ftree-vectorize -fwhole-program" )
 set( SPYDRPICK_GCC_OPTIMIZATION_FLAGS "${SPYDRPICK_GCC_OPTIMIZATION_FLAGS} -flto -ffat-lto-objects" ) # -flto-report" )
 
-- 
2.14.3 (Apple Git-98)

