diff --git a/externals/boost.cmake b/externals/boost.cmake
index 434ae95..4196e25 100644
--- a/externals/boost.cmake
+++ b/externals/boost.cmake
@@ -32,10 +32,10 @@ if( ${CMAKE_PROJECT_NAME}_ENABLE_BOOST )
 			set( Boost_NO_SYSTEM_PATHS ON )
 		endif()
 	
-		set( Boost_USE_STATIC_LIBS ON )
+		set( Boost_USE_STATIC_LIBS OFF )
 		set( Boost_USE_MULTITHREADED TRUE )
 		#set( Boost_NO_BOOST_CMAKE ON )
-		#set( Boost_DEBUG ON )
+		set( Boost_DEBUG ON )
 	
 		find_package( Boost REQUIRED program_options filesystem iostreams timer chrono ) #system #date_time thread
 		if( Boost_FOUND )
