diff --git a/include.mk b/include.mk
index da3f4386..9a950263 100644
--- a/include.mk
+++ b/include.mk
@@ -37,6 +37,8 @@ include ${sonLibRootDir}/include.mk
 #Turn asserts back on in spite of sonLib
 #https://github.com/ComparativeGenomicsToolkit/cactus/issues/235
 CFLAGS += -UNDEBUG
+CFLAGS += -I${PREFIX}/include -L${PREFIX}/lib
+CPPFLAGS += -I${PREFIX}/include -L${PREFIX}/lib
 
 # Hack to include xml2
 ifeq (${CACTUS_LIBXML2_INCLUDE_PATH},)
