diff --git a/Makefile b/Makefile
index 1934e97c..183bf9fc 100644
--- a/Makefile
+++ b/Makefile
@@ -6,7 +6,7 @@ modules = api setup caf bar hal reference pipeline preprocessor
 
 # submodules are in multiple pass to handle dependencies cactus2hal being dependent on
 # both cactus and sonLib
-submodules1 = sonLib cPecan hal matchingAndOrdering pinchesAndCacti abPOA lastz paffy red collapse-bubble FASTGA
+submodules1 = sonLib cPecan hal matchingAndOrdering pinchesAndCacti abPOA paffy red collapse-bubble FASTGA
 submodules2 = cactus2hal
 submodules = ${submodules1} ${submodules2}
 
@@ -259,10 +259,6 @@ suball.abPOA:
 	ln -f submodules/abPOA/include/*.h ${INCLDIR}
 	rm -fr ${INCLDIR}/simde && cp -r submodules/abPOA/include/simde ${INCLDIR}
 
-suball.lastz:
-	cd submodules/lastz && ${MAKE}
-	ln -f submodules/lastz/src/lastz bin
-
 suball.paffy:
 	cd submodules/paffy && ${MAKE}
 	rm -rf submodules/paffy/bin/*.dSYM
