diff --git a/Makefile b/Makefile
index e9f8fa71..cbfac27d 100644
--- a/Makefile
+++ b/Makefile
@@ -7,7 +7,7 @@ modules = api setup caf bar hal reference pipeline preprocessor
 # submodules are in multiple pass to handle dependencies cactus2hal being dependent on
 # both cactus and sonLib
 # jemalloc is conditionally added based on include.mk settings
-submodules1 = sonLib cPecan hal matchingAndOrdering pinchesAndCacti abPOA lastz paffy red collapse-bubble FASTGA FASTAN alntools
+submodules1 = sonLib cPecan hal matchingAndOrdering pinchesAndCacti abPOA paffy red collapse-bubble FASTGA alntools
 submodules2 = cactus2hal
 submodules = ${submodules1} ${submodules2}
 
@@ -300,11 +300,6 @@ suball.abPOA:
 	ln -f submodules/abPOA/include/*.h ${INCLDIR}
 	rm -fr ${INCLDIR}/simde && cp -r submodules/abPOA/include/simde ${INCLDIR}
 
-suball.lastz:
-	cd submodules/lastz/src && sed -i -e 's/-lm -o/-lm $${LIBS} -o/g' Makefile
-	cd submodules/lastz && LIBS="${jemallocLib}" ${MAKE}
-	ln -f submodules/lastz/src/lastz bin
-
 suball.paffy:
 	git -C submodules/paffy/submodules/sonLib checkout $$(git -C submodules/sonLib rev-parse HEAD)
 	cd submodules/paffy && LIBS="${jemallocLib}" ${MAKE}
@@ -327,9 +322,6 @@ suball.FASTGA:
 	ln -f submodules/FASTGA/FAtoGDB ${BINDIR}
 	ln -f submodules/FASTGA/GIXmake ${BINDIR}
 	ln -f submodules/FASTGA/GIXrm ${BINDIR}
-suball.FASTAN:
-	cd submodules/FASTAN && sed -i -e 's/-lm -lz/-lm -lpthread -lz/g' Makefile && ${MAKE} || true
-	ln -f submodules/FASTAN/FasTAN ${BINDIR}
 suball.alntools:
 	cd submodules/alntools && ${MAKE}
 	ln -f submodules/alntools/tanbed ${BINDIR}
