--- cactus/setup.py	2019-03-18 14:23:36.545920501 -0400
+++ setup.py	2019-03-18 15:51:12.809349183 -0400
@@ -5,7 +5,7 @@
 versionFile = "src/cactus/shared/version.py"
 if os.path.exists(versionFile):
     os.remove(versionFile)
-git_commit = subprocess.check_output(['git', 'rev-parse', 'HEAD']).strip()
+git_commit = '833d8ca015deecdfa5d0aca01211632cdaca9e58'
 with open(versionFile, 'w') as versionFH:
     versionFH.write("cactus_commit = '%s'" % git_commit)
 
@@ -21,15 +21,6 @@
     # We use the __file__ attribute so this package isn't zip_safe.
     zip_safe=False,
 
-    install_requires=[
-        'decorator',
-        'subprocess32',
-        'psutil',
-        'networkx>=2,<3',
-        'cython',
-        # Someone uploaded an old version of sonLib to pyPI, so we have to use this name
-        'actualSonLib'],
-
     entry_points={
         'console_scripts': ['cactus = cactus.progressive.cactus_progressive:main',
                             'cactus_preprocess = cactus.preprocessor.cactus_preprocessor:main']},)
