--- cy_build.py	2017-01-05 15:12:18.000000000 -0600
+++ cy_build.py.new	2017-03-14 12:39:37.000000000 -0500
@@ -77,9 +77,8 @@
             if not ext.extra_link_args:
                 ext.extra_link_args = []
             ext.extra_link_args += ['-dynamiclib',
-                                    '-rpath', get_python_lib(),
                                     '-Wl,-headerpad_max_install_names',
-                                    '-Wl,-install_name,%s' % linker_path,
+                                    '-Wl,-install_name,%s' % relative_module_path.split(os.sep)[1],
                                     '-Wl,-x']
         else:
             if not ext.extra_link_args:
