diff --git a/bcftools/bcftools.pysam.h b/bcftools/bcftools.pysam.h
index b8bf93e..6068d02 100644
--- a/bcftools/bcftools.pysam.h
+++ b/bcftools/bcftools.pysam.h
@@ -57,4 +57,16 @@ void bcftools_set_optind(int);
 
 extern int bcftools_main(int argc, char *argv[]);
   
+#define bam_smpl_destroy bcftools_bam_smpl_destroy
+#define bam_smpl_init bcftools_bam_smpl_init
+#define bcf_call2bcf bcftools_bcf_call2bcf
+#define bcf_callaux_clean bcftools_bcf_callaux_clean
+#define bcf_call_combine bcftools_bcf_call_combine
+#define bcf_call_destroy bcftools_bcf_call_destroy
+#define bcf_call_gap_prep bcftools_bcf_call_gap_prep
+#define bcf_call_glfgen bcftools_bcf_call_glfgen
+#define bcf_call_init bcftools_bcf_call_init
+#define parse_format_flag bcftools_parse_format_flag
+#define read_file_list bcftools_read_file_list
+
 #endif
