--- phist.py.old	2022-10-03 01:27:09.242058987 -0500
+++ phist.py	2022-10-03 01:28:44.800450188 -0500
@@ -1,3 +1,5 @@
+#!/opt/anaconda1anaconda2anaconda3/bin/python
+
 # 
 # PHIST
 # Copyright (C) 2021, A. Zielezinski, S. Deorowicz, and A. Gudys
@@ -53,12 +55,8 @@
     
     PHIST_DIR = Path(__file__).resolve().parent
 
-    if platform.system() == "Windows":
-        kmer_exec = PHIST_DIR.joinpath('kmer-db', 'src', 'x64', 'Release', 'kmer-db.exe')
-        util_exec = PHIST_DIR.joinpath('utils', 'x64', 'Release', 'phist.exe')
-    else:
-        kmer_exec = PHIST_DIR.joinpath('kmer-db', 'kmer-db')
-        util_exec = PHIST_DIR.joinpath('utils', 'phist')
+    kmer_exec = 'kmer-db'
+    util_exec = 'phist'
 
     args = get_arguments()
 
