diff --git a/tools/nasm-filter.sh b/tools/nasm-filter.sh
index 73d23b8..dcd2fec 100755
--- a/tools/nasm-filter.sh
+++ b/tools/nasm-filter.sh
@@ -15,6 +15,13 @@ while [ -n "$*" ]; do
 	    shift
 	    shift
 	    ;;
+    -isystem )  #-isystem *
+        shift
+        shift
+        ;;
+    -f*'='* ) # Long debug prefixes
+        shift
+        ;;
 	-isysroot | -iframeworkwithsysroot | -iwithsysroot | -framework | -arch )
 	    # Unsupported options with arg
 	    shift
