--- Cargo.toml	
+++ Cargo.toml	
@@ -59,10 +59,10 @@
 # nsw_rs = {git = "https://gitlab.com/jpboth/hnswlib-rs.git"}
 #hnsw_rs = {path = "../hnswlib-rs"}
 
-kmerutils = {path = "../kmerutils"}
+# kmerutils = {path = "../kmerutils"}
 # kmerutils = {git = "https://github.com/jean-pierreBoth/kmerutils"}
 # kmerutils = {git = "https://gitlab.com/jpboth/kmerutils.git"}
-# kmerutils = {version = "0.0.9"}
+kmerutils = {version = "0.0.9"}
 
 lazy_static = { version = "1.4"}
 
