--- a/README.md
+++ b/README.md
@@ -1,3 +1,5 @@
+![Docker Pulls](https://img.shields.io/docker/pulls/teambraker/braker3) 
+
 # BRAKER User Guide
 
 News
-- 
2.40.0


--- a/README.md
+++ b/README.md
@@ -1,4 +1,5 @@
-![Docker Pulls](https://img.shields.io/docker/pulls/teambraker/braker3) 
+![Docker Pulls](https://img.shields.io/docker/pulls/teambraker/braker3)  ![GitHub all releases](https://img.shields.io/github/downloads/gaius-augustus/braker/total)
+
 
 # BRAKER User Guide
 
-- 
2.40.0


--- a/README.md
+++ b/README.md
@@ -1,5 +1,4 @@
-![Docker Pulls](https://img.shields.io/docker/pulls/teambraker/braker3)  ![GitHub all releases](https://img.shields.io/github/downloads/gaius-augustus/braker/total)
-
+![Docker Pulls](https://img.shields.io/docker/pulls/teambraker/braker3) 
 
 # BRAKER User Guide
 
-- 
2.40.0


--- a/scripts/braker.pl
+++ b/scripts/braker.pl
@@ -5404,7 +5404,7 @@ sub GeneMark_ETP {
                         . __LINE__ ."\nFailed to execute: $cmdString!\n");
             }
 
-            # concatenate protein files
+            # concatenate protein files and remove '.' at the end of sequences
             open(PROT_ALL, ">", $protein_file) or clean_abort("$AUGUSTUS_CONFIG_PATH/species/$species",
                 $useexisting, "ERROR in file " . __FILE__ ." at line "
                 . __LINE__ ."\nfailed to open file $protein_file!\n");
@@ -5412,8 +5412,12 @@ sub GeneMark_ETP {
                     open(PROT, "<", $_) or clean_abort("$AUGUSTUS_CONFIG_PATH/species/$species",
                         $useexisting, "ERROR in file " . __FILE__ ." at line "
                         . __LINE__ ."\nfailed to open file $_!\n");
-                    while(<PROT>){
-                        print PROT_ALL $_;
+                    while(my $line = <PROT>){
+                        # remove dots at end of sequences
+                        if ($line !~ /^>/ && $line =~ /\.$/) {
+                            $line =~ s/\.$//;
+                        }
+                        print PROT_ALL $line;
                     }
                     close(PROT) or clean_abort("$AUGUSTUS_CONFIG_PATH/species/$species",
                         $useexisting, "ERROR in file " . __FILE__ ." at line "
@@ -7977,7 +7981,7 @@ sub clean_aug_jobs {
             }
         }
     } else {
-        # moving files for AUGUSTUS prallelization to a separate folder
+        # moving files for AUGUSTUS parallelization to a separate folder
         print LOG "\# " . (localtime) . ": moving files from AUGUSTUS "
             . "parallelization to directory "
             . "$otherfilesDir/augustus_files_$hintId\n" if ($v > 3);
@@ -8287,7 +8291,7 @@ sub merge_transcript_sets_with_tsebra {
     my $tsebra_cfg = $_[3];
     my $output = $_[4];
     my $rename = $_[5];
-    #my $localUtr = shift;
+    my $filter_se = $_[6];
 
     print LOG "\# " . (localtime) . ": Trying to create combined gene set "
                                   . "with TSEBRA...\n" if ($v > 3);
@@ -8332,6 +8336,9 @@ sub merge_transcript_sets_with_tsebra {
     if ( !$arg_str eq "" ){
         $cmdString .= "--hintfiles $arg_str ";
     }
+    if ( !$filter_se eq "" ){
+        $cmdString .= "--filter_single_exon_genes ";
+    }
     $cmdString .= "--cfg $tsebra_cfg --out $output -q 2>$errorfile";
     print LOG "$cmdString\n" if ($v > 3);
     system("$cmdString") == 0 or die("ERROR in file " . __FILE__ ." at line ". __LINE__
@@ -9649,37 +9656,19 @@ sub merge_transcript_sets {
             "$genemarkDir/genemark.gtf");
         @forced_gene_sets = ("$genemarkDir/training.gtf");
         @hintfiles = ($hintsfile);
-        merge_transcript_sets_with_tsebra(
-            \@gene_sets, \@forced_gene_sets, \@hintfiles,
-            "braker3.cfg", "$otherfilesDir/braker$genesetId.gtf", "rename");
 
         # filter TSEBRA output for species with 'large' genomes
         if ($genome_size > 300000000){
-            $cmdString = "mv $otherfilesDir/braker$genesetId.gtf "
-                . "$otherfilesDir/braker$genesetId.unsupported.gtf";
-            print LOG "\# "
-              . (localtime)
-              . ": move TSBERA output $otherfilesDir/braker$genesetId.gtf to "
-                  . "$otherfilesDir/braker$genesetId.unsupported.gtf\n" if ($v > 3);
-            print LOG "$cmdString\n" if ($v > 3);
-            system("$cmdString") == 0
-              or die("ERROR in file " . __FILE__ ." at line ". __LINE__
-                  . "\nfailed to execute: $cmdString!\n");
-
-            $cmdString = "$GENEMARK_PATH/selectSupportedSubsets.py "
-                . "$otherfilesDir/braker$genesetId.unsupported.gtf "
-                . "$hintsfile --fullSupport /dev/null --anySupport "
-                . "$otherfilesDir/braker$genesetId.gtf --noSupport /dev/null "
-                . "1> $otherfilesDir/selectSupportedSubsets.stdout "
-                . "2> $errorfilesDir/selectSupportedSubsets.stderr";
-            print LOG "\# "
-              . (localtime)
-              . ": Filter TSEBRA result:\n" if ($v > 3);
-            print LOG "$cmdString\n" if ($v > 3);
-            system("$cmdString") == 0
-              or die("ERROR in file " . __FILE__ ." at line ". __LINE__
-                  . "\nfailed to execute: $cmdString!\n");
+            merge_transcript_sets_with_tsebra(
+                \@gene_sets, \@forced_gene_sets, \@hintfiles,
+                "braker3.cfg", "$otherfilesDir/braker$genesetId.gtf", "rename",
+                "filter_single_exons");
+        } else {
+            merge_transcript_sets_with_tsebra(
+                \@gene_sets, \@forced_gene_sets, \@hintfiles,
+                "braker3.cfg", "$otherfilesDir/braker$genesetId.gtf", "rename", "");
         }
+        
     } elsif($ESmode == 1) {
         # exists only without utr
         if( (-e "$genemarkDir/genemark.gtf") and
@@ -9788,7 +9777,7 @@ sub gtf2gff3 {
         }
         $perlCmdString .= "cat $gtf | $perl -ne 'if(m/\\tAUGUSTUS\\t/ or "
                        .  "m/\\tAnnotationFinalizer\\t/ or m/\\tGUSHR\\t/ or "
-                       .  "m/\\tGeneMark\.hmm\\t/) {"
+                       .  "m/\\tGeneMark\.hmm\\t/ or m/\\tgmst\\t/) {"
                        .  "print \$_;}' | $perl $string --gff3 --out=$gff3 "
                        .  ">> $otherfilesDir/gtf2gff3.log "
                        .  "2>> $errorfilesDir/gtf2gff3.err";
-- 
2.40.0


--- a/README.md
+++ b/README.md
@@ -855,6 +855,10 @@ You may additionally include bam files from unstranded libraries. Those files wi
 
 The native mode for running BRAKER with RNA-Seq and protein data. This will call GeneMark-ETP, which will use RNA-Seq and protein hints for training GeneMark-ETP. Subsequently, AUGUSTUS is trained on 'high-confindent' genes (genes with very high extrinsic evidence support) from the GeneMark-ETP prediction and a set of genes is predicted by AUGUSTUS. In a last step, the predictions of AUGUSTUS and GeneMark-ETP are combined using TSEBRA.
 
+**Alignment of RNA-Seq reads**
+
+GeneMark-ETP utilizes Stringtie2 to assemble RNA-Seq data, which requires that the aligned reads (BAM files) contain the XS (strand) tag for spliced reads. Therefore, if you align your reads with HISAT2, you must enable the --dta option, or if you use STAR, you must use the --outSAMstrandField intronMotif option. TopHat alignments include this tag by default.
+
 To call the pipeline in this mode, you have to provide it with a protein database using `--prot_seq` (as described in [BRAKER with protein data](#braker-with-protein-data)), and RNA-Seq data either by their SRA ID so that they are downloaded by BRAKER, as unaligned reads in `FASTQ` format, and/or as aligned reads in `bam` format (as described in [BRAKER with RNA-Seq data](#braker-with-rna-seq-data)). You could also specify already processed extrinsic evidence using the `--hints` option. However, this is not recommend for a normal BRAKER run in ETPmode, as these hints won't be used in the GeneMark-ETP step. Only use `--hints` when you want to skip the GenMark-ETP step!
 
 Examples of how you could run BRAKER in ETPmode:
-- 
2.40.0


--- a/Dockerfile
+++ b/Dockerfile
@@ -190,7 +190,6 @@ USER root
 RUN cd /opt && \
     git clone      https://github.com/Gaius-Augustus/BRAKER.git && \
     cd BRAKER && \
-    git checkout braker3  && \
     cd example && \
     wget http://bioinf.uni-greifswald.de/augustus/datasets/RNAseq.bam
 
-- 
2.40.0


--- a/README.md
+++ b/README.md
@@ -11,7 +11,7 @@ Contacts for Repository
 
 Lars Gabriel, University of Greifswald, Germany, lars.gabriel@uni-greifswald.de
 
-Katharina J. Hoff, University of Greifswald, Germany, katharina.hoff@uni-greifswald.de, +49 3834 420 4624
+Katharina J. Hoff, University of Greifswald, Germany, katharina.hoff@uni-greifswald.de, +49 3834 420 4624, ![Twitter Follow](https://img.shields.io/twitter/follow/katharina_hoff)
 
 Mark Borodovsky, Georgia Tech, U.S.A., borodovsky@gatech.edu
 
-- 
2.40.0


--- a/README.md
+++ b/README.md
@@ -11,7 +11,8 @@ Contacts for Repository
 
 Lars Gabriel, University of Greifswald, Germany, lars.gabriel@uni-greifswald.de
 
-Katharina J. Hoff, University of Greifswald, Germany, katharina.hoff@uni-greifswald.de, +49 3834 420 4624, ![Twitter Follow](https://img.shields.io/twitter/follow/katharina_hoff)
+Katharina J. Hoff, University of Greifswald, Germany, katharina.hoff@uni-greifswald.de, +49 3834 420 4624, [![https://twitter.com/katharina_hoff](https://img.shields.io/twitter/follow/katharina_hoff)](https://twitter.com/katharina_hoff)
+
 
 Mark Borodovsky, Georgia Tech, U.S.A., borodovsky@gatech.edu
 
-- 
2.40.0
