diff --git a/setup.py b/setup.py
index 0d80f2f..85278a5 100644
--- a/setup.py
+++ b/setup.py
@@ -3,7 +3,6 @@
 import setuptools
 from setuptools.extension import Extension
 from Cython.Build import cythonize
-
 import numpy as np
 
 # I'd prefer not to include static zstd here, but I get build problems on macOS
@@ -71,7 +70,7 @@ ext_modules = [
               # overload of uint32_t operator
               # extra_compile_args = ["-std=c++11", "-Wno-unused-function"],
               # extra_link_args = ["-std=c++11"],
-              extra_compile_args = ["-std=c++98", "-Wno-unused-function", "-Wno-macro-redefined", "-DZSTD_DISABLE_ASM", "-DLIBDEFLATE_STATIC"],
+              extra_compile_args = ["-std=c++98", "-Wno-unused-function", "-Wno-macro-redefined", "-Wno-c++11-extensions", "-Wno-cpp", "-DZSTD_DISABLE_ASM", "-DLIBDEFLATE_STATIC"],
               extra_link_args = ["-std=c++98", "-lz"],
               include_dirs = [np.get_include()] + ['src/plink2/libdeflate']
               )
@@ -98,7 +97,7 @@ setuptools.setup(
         "Operating System :: OS Independent",
     ],
     package_dir={"": "src"},
-    packages=setuptools.find_packages(where="src"),
+    packages=setuptools.find_namespace_packages(where="src"),
     python_requires=">=3.5",
     libraries=[clib],
     ext_modules=cythonize(ext_modules),
