diff --git a/setup.py b/setup.py
index 45677ce..ba1276c 100644
--- a/setup.py
+++ b/setup.py
@@ -1,11 +1,15 @@
 import subprocess
 import sys
+import os
 
 from setuptools import setup
 
+# get compiler path
+gxx = os.environ.get("CXX", "g++")
+
 try:
     cmd = (
-        "g++ -std=c++11 -O3 scib/knn_graph/knn_graph.cpp -o scib/knn_graph/knn_graph.o"
+        f"{gxx} -std=c++11 -O3 scib/knn_graph/knn_graph.cpp -o scib/knn_graph/knn_graph.o"
     )
     sys.stdout.write("Compile knn_graph C++ code for LISI metric...\n")
     sys.stdout.flush()
