--- Mothur.source/makefile	2012-05-01 13:35:08.000000000 +0000
+++ Mothur.source/makefile.new	2016-01-08 21:30:46.604790880 +0000
@@ -28,7 +28,7 @@
 	#CXX = g++44
 	
 	#if you are a mac user use the following line
-	TARGET_ARCH += -arch x86_64
+	#TARGET_ARCH += -arch x86_64
 	
 	#if you using cygwin to build Windows the following line
 	#CXX = x86_64-w64-mingw32-g++
@@ -37,7 +37,7 @@
  	#TARGET_ARCH += -m64 -static
 
 	#if you are a linux user use the following line
-	#CXXFLAGS += -mtune=native -march=native -m64
+	CXXFLAGS += -mtune=core2 -m64
 	
 	CXXFLAGS += -DBIT_VERSION
     FORTRAN_FLAGS = -m64
@@ -61,8 +61,8 @@
 
 
 ifeq  ($(strip $(USEREADLINE)),yes)
-    CXXFLAGS += -DUSE_READLINE
-    LIBS = \
+    CXXFLAGS += -DUSE_READLINE -L${PREFIX}/lib -I${PREFIX}/include
+    LIBS = -L${PREFIX}/lib \
       -lreadline\
       -lncurses
 endif
