diff --git a/Cargo.toml b/Cargo.toml
index 2956bda..cc6629a 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -12,8 +12,8 @@ name = "pycoverm"
 crate-type = ["cdylib"]
 
 [dependencies]
-pyo3 = "0.20.3"
+pyo3 = "0.24.2"
 coverm = "0.7.0"
-ndarray = "0.15.6"
-numpy = "0.20.0"
-rust-htslib = { version = "0.45.0", default-features = false }
+ndarray = "0.16.1"
+numpy = "0.24.0"
+rust-htslib = { version = "0.48.0", default-features = false }
