--- bamtools/src/toolkit/bamtools_resolve.cpp	2023-10-17 13:08:19.202625000 -0400
+++ bamtools/src/toolkit/bamtools_resolve.cpp	2023-10-17 13:08:19.202625000 -0400
@@ -410,7 +410,7 @@
         ReadGroupResolver& resolver = (*rgIter).second;
 
         // store read name with resolver
-        resolver.ReadNames.insert( make_pair<string,bool>(fields[1], true) ) ;
+        resolver.ReadNames.insert( pair<string,bool>(fields[1], true) ) ;
     }
 
     // if here, return success
@@ -607,7 +607,7 @@
     resolver.IsAmbiguous = ( fields.at(6) == TRUE_KEYWORD );
 
     // store RG entry and return success
-    readGroups.insert( make_pair<string, ReadGroupResolver>(name, resolver) );
+    readGroups.insert( pair<string, ReadGroupResolver>(name, resolver) );
     return true;
 }
 
@@ -1014,7 +1014,7 @@
         }
 
         // if read name not found, store new entry
-        else resolver.ReadNames.insert( make_pair<string, bool>(al.Name, isCurrentMateUnique) );
+        else resolver.ReadNames.insert( pair<string, bool>(al.Name, isCurrentMateUnique) );
     }
 
     // close files
@@ -1046,7 +1046,7 @@
     SamReadGroupConstIterator rgEnd  = header.ReadGroups.ConstEnd();
     for ( ; rgIter != rgEnd; ++rgIter ) {
         const SamReadGroup& rg = (*rgIter);
-        m_readGroups.insert( make_pair<string, ReadGroupResolver>(rg.ID, ReadGroupResolver()) );
+        m_readGroups.insert( pair<string, ReadGroupResolver>(rg.ID, ReadGroupResolver()) );
     }
 }
 
