--- rMATS_pipeline/setup.py	2020-06-02 10:25:19.867757586 -0400
+++ rMATS_pipeline/setup.py	2020-06-02 10:25:19.867757586 -0400
@@ -8,8 +8,6 @@
 bamtools_lib = [os.path.abspath('../bamtools/lib'),]
 source_pattern = 'rmatspipeline/%s.pyx'
 
-os.environ['CC'] = 'g++'
-os.environ['CXX'] = 'g++'
 # os.environ['CXXFLAGS'] = '-Wl,-rpath,%s' % (bamtools_lib[0])
 # os.environ['LD_LIBRARY_PATH'] = '%s' % (bamtools_lib[0])
 
