--- rmats.py    2020-06-01 12:39:50.000000000 -0400
+++ rmats.py    2020-06-01 12:39:50.000000000 -0400
@@ -320,7 +320,7 @@
     ostat_fdr = ostat % ('FDR')
     ostat_paired = ostat % ('paired')

-    root_dir = os.path.abspath(os.path.dirname(__file__))
+    root_dir = os.path.abspath(os.path.dirname(os.path.realpath(__file__)))
     rmats_c = os.path.join(root_dir, 'rMATS_C/rMATSexe')
     paired_model = os.path.join(root_dir, 'rMATS_R/paired_model.R')
     pas_out = os.path.join(root_dir, 'rMATS_P/paste.py')
