--- rMATS-turbo-Linux-UCS4/rmats.py	2018-04-19 21:42:16.000000000 +0200
+++ rMATS-turbo-Linux-UCS4/rmats.py	2018-04-19 21:42:16.000000000 +0200
@@ -243,7 +243,7 @@
     resfp = open(ostat % (''), 'w')
     finfp = os.path.join(od, '%s.MATS.%s.txt' % (ase, counttype))
 
-    root_dir = os.path.abspath(os.path.dirname(__file__))
+    root_dir = os.path.abspath(os.path.dirname(os.path.realpath(__file__)))
     rmats_c = os.path.join(root_dir, 'rMATS_C/rMATSexe')
     pas_out = os.path.join(root_dir, 'rMATS_P/paste.py')
     inc_lvl = os.path.join(root_dir, 'rMATS_P/inclusion_level.py')
