--- rMATS_pipeline/setup.py	2024-11-19 14:12:11.355314000 -0500
+++ rMATS_pipeline/setup.py	2024-11-19 14:12:11.355314000 -0500
@@ -1,6 +1,6 @@
 import os
-from distutils.core import setup
-from distutils.extension import Extension
+from setuptools import setup
+from setuptools import Extension
 from Cython import __version__ as cython_version
 from Cython.Build import cythonize
 from Cython.Distutils import build_ext
