--- ./rMATS_pipeline/rmatspipeline/tcx.h	2023-10-16 13:03:27.945584000 -0400
+++ ./rMATS_pipeline/rmatspipeline/tcx.h	2023-10-16 13:03:27.945584000 -0400
@@ -19,7 +19,7 @@
         std::string g_name;
         std::string chrom;
         char strand;
-        void set_info(std::string& iname, std::string& ichrom, std::string& istrand) {
+        void set_info(const std::string& iname, const std::string& ichrom, const std::string& istrand) {
             this->g_name = iname;
             this->chrom = ichrom;
             this->strand = istrand.c_str()[0];
