--- src/kmer_filter.cc.old	2025-01-02 14:40:19.258463783 -0600
+++ src/kmer_filter.cc	2025-01-02 14:41:48.294398484 -0600
@@ -111,12 +111,12 @@
         if (kmer_distr) {
             generate_kmer_dist(kmers);
             if (write_k_freq == false)
-                exit(1);
+                exit(0);
         }
 
         if (write_k_freq) {
             write_kmer_freq(k_freq_path, kmers);
-            exit(1);
+            exit(0);
         }
 
         cerr << "Filtering reads by kmer frequency...\n";
