From 5c03d64a1c99378e31af1e14cb2899e50240742a Mon Sep 17 00:00:00 2001
From: Julian de Ruiter <julianderuiter@gmail.com>
Date: Tue, 4 Apr 2017 14:34:58 +0200
Subject: [PATCH] Use yintercept in geom_hline.

---
 R/plot.R | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/R/plot.R b/R/plot.R
index 37f2e14..6e363dd 100644
--- a/R/plot.R
+++ b/R/plot.R
@@ -111,7 +111,7 @@ segments.as.data.table <- function(map.loc.agr, segments, markers) {
 #'
 #' @param x A vector of numbers.
 #' @return A vector of strings of formatted numbers.
-#' 
+#'
 #' @noRd
 location.format <- function(x) {
   format(x, scientific=F, trim=T)
@@ -174,7 +174,7 @@ plot.genes.bar <- function(genes, min.x, max.x, focal.p=NULL, focal.n=NULL) {
 #' @param steps A boolean value indicating whether to plot the segments as steps
 #'              or horizontal segments.
 #' @param chromosome The chromosome to be plotted.
-#' 
+#'
 #' @noRd
 plot.cna <- function(amp.profile=NULL, del.profile=NULL,
                      segments.p=NULL, segments.n=NULL,
@@ -268,7 +268,7 @@ plot.cna <- function(amp.profile=NULL, del.profile=NULL,
     }
 
   } else {
-    main.plot <- main.plot + geom_hline(aes(0), size=.2, color='grey', linetype="dashed") +
+    main.plot <- main.plot + geom_hline(aes(yintercept=0), size=.2, color='grey', linetype="dashed") +
       scale_y_continuous(name='Aggregate log ratios')
     rect.min.y <- -Inf
     rect.max.y <- +Inf
-- 
2.10.1 (Apple Git-78)

