--- gtdbtk/main.py	2019-11-15 15:39:05.000000000 +1000
+++ gtdbtk/main_new.py	2020-03-26 14:50:30.950119399 +1000
@@ -506,6 +506,7 @@
         options : argparse.Namespace
             The CLI arguments input by the user.
         """
+        self.logger.warning('You are not running the most recent version of GTDB-Tk!')

         if options.subparser_name == 'de_novo_wf':
             check_dependencies(['prodigal', 'hmmalign'])
