From b44f302295f8e5dd7ecb1dd1a787fbcd101cb85b Mon Sep 17 00:00:00 2001
From: Ilya Shlyakhter <ilya_shl@alum.mit.edu>
Date: Mon, 6 Jan 2020 16:20:52 -0500
Subject: [PATCH] add rt lib to easel

---
 ext/src/easel/CMakeLists.txt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ext/src/easel/CMakeLists.txt b/ext/src/easel/CMakeLists.txt
index 85dba29..3deaeee 100644
--- a/ext/src/easel/CMakeLists.txt
+++ b/ext/src/easel/CMakeLists.txt
@@ -73,5 +73,5 @@ set(easel_src
   esl_wuss.c)
 
 add_library(easel STATIC ${easel_src})
-target_link_libraries(easel m)
+target_link_libraries(easel m rt)
 target_include_directories(easel SYSTEM PUBLIC ${EASEL_BUILT_INCLUDE_DIR} ${EASEL_MAIN_SRC_DIR})
-- 
2.20.1

