From 1f1c3b3e55107c3679fdcc8680c9c4c376ab77b1 Mon Sep 17 00:00:00 2001
From: Alexander Chori <alexandrchori@gmail.com>
Date: Tue, 4 Aug 2020 20:17:05 +0300
Subject: [PATCH] Add missing header

---
 ext/src/llvm/raw_ostream.cpp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/ext/src/llvm/raw_ostream.cpp b/ext/src/llvm/raw_ostream.cpp
index 3b5483f..b24c54d 100644
--- a/ext/src/llvm/raw_ostream.cpp
+++ b/ext/src/llvm/raw_ostream.cpp
@@ -10,6 +10,7 @@
 //
 //===----------------------------------------------------------------------===//
 
+#define __STDC_FORMAT_MACROS
 #include "llvm/Support/raw_ostream.h"
 #include "llvm/ADT/STLExtras.h"
 #include "llvm/ADT/SmallVector.h"
-- 
2.11.0

