-- BuildSuperTranscript.py     2016-12-23 15:20:45.000000000 +0200
+++ BuildSuperTranscript.py     2016-12-23 15:22:53.000000000 +0200
@@ -1,3 +1,4 @@
+#!/usr/bin/env python
 # Author: Anthony Hawkins

 #A script to construct a SuperTranscript given a .fasta file of transcript sequence
--- Checker.py  2016-12-23 15:20:54.000000000 +0200
+++ Checker.py  2016-12-23 15:23:01.000000000 +0200
@@ -1,3 +1,4 @@
+#!/usr/bin/env python
 #Author: Anthony Hawkins

 #A script to systematically check for each gene whether the SuperTranscript builder worked ok
--- Lace.py     2016-12-23 15:21:08.000000000 +0200
+++ Lace.py     2016-12-23 15:23:25.000000000 +0200
@@ -1,3 +1,4 @@
+#!/usr/bin/env python
 #Author: Anthony Hawkins

 #Lacing together different transcripts into a SuperTranscript
@@ -14,7 +15,7 @@
 import sys
 import time
 import argparse
-from Checker import Checker
+from Lace_Checker import Checker

 def worker(fname):
        seq =''
--- Mobius.py   2016-12-23 15:21:18.000000000 +0200
+++ Mobius.py   2016-12-23 15:23:33.000000000 +0200
@@ -1,3 +1,4 @@
+#!/usr/bin/env python
 #Author: Anthony Hawkins
 #A little script to construct an annotation from an SJ.out.tab file, a standard STAR output.
 # This creates the dynamic block annotation
--- STViewer.py 2016-12-23 15:21:42.000000000 +0200
+++ STViewer.py 2016-12-23 15:23:42.000000000 +0200
@@ -1,3 +1,4 @@
+#!/usr/bin/env python
 #Author: Anthony Hawkins
 #Visualise a given gene in your super transcript
