diff --git a/bin/adam-shell b/bin/adam-shell
index a1a7d946..b0c50053 100755
--- a/bin/adam-shell
+++ b/bin/adam-shell
@@ -25,11 +25,19 @@ if [[ -z $@ && -n "$ADAM_OPTS" ]]; then
     echo "Run adam-shell instead as adam-shell <spark-args>" 1>&2
 fi
 
-SOURCE_DIR=$(dirname ${BASH_SOURCE[0]})
+# Find original directory of this script, resolving symlinks
+# http://stackoverflow.com/questions/59895/can-a-bash-script-tell-what-directory-its-stored-in/246128#246128
+SOURCE="${BASH_SOURCE[0]}"
+while [ -h "$SOURCE" ]; do # resolve $SOURCE until the file is no longer a symlink
+    SCRIPT_DIR="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
+    SOURCE="$(readlink "$SOURCE")"
+    [[ $SOURCE != /* ]] && SOURCE="$SCRIPT_DIR/$SOURCE" # if $SOURCE was a relative symlink, we need to resolve it relative to the path where the symlink file was located
+done
+SCRIPT_DIR="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
 
-ADAM_CLI_JAR=$(${SOURCE_DIR}/find-adam-assembly.sh)
+ADAM_CLI_JAR=$(${SCRIPT_DIR}/find-adam-assembly.sh)
 
-SPARK_SHELL=$(${SOURCE_DIR}/find-spark.sh spark-shell)
+SPARK_SHELL=$(${SCRIPT_DIR}/find-spark.sh spark-shell)
 echo "Using SPARK_SHELL=$SPARK_SHELL" 1>&2
 
 # submit the job to Spark
