--- kyototycoon/ktulog.h	2017-04-10 04:57:04.000000000 -0400
+++ kyototycoon/ktulog.h	2019-03-13 14:20:05.571610303 -0400
@@ -48,7 +48,7 @@
   /* The accuracy of logical time stamp. */
   static const uint64_t TSLACC = 1000 * 1000;
   /* The waiting seconds of auto flush. */
-  static const double FLUSHWAIT = 0.1;
+  static constexpr double FLUSHWAIT = 0.1;
  public:
   /**
    * Reader of update logs.
