From 70f3bac559b81fa8980c3687e74d3b964ce18433 Mon Sep 17 00:00:00 2001
From: Alexey Gurevich <alexeigurevich@gmail.com>
Date: Thu, 8 Feb 2018 13:14:41 +0300
Subject: [PATCH] Temp: add GAGE to regular installation

---
 setup.py | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/setup.py b/setup.py
index 7e36954..0bc9529 100755
--- a/setup.py
+++ b/setup.py
@@ -169,6 +169,9 @@ if cmd_in(['install', 'develop', 'build', 'build_ext']):
     logger.info('* Compiling read analysis tools *')
     if not compile_reads_analyzer_tools(logger):
         modules_failed_to_install.append('Read analysis tools (affects -1/--reads1 and -2/--reads2 options)')
+    logger.info('* Compiling GAGE *')
+    if not compile_gage():
+        modules_failed_to_install.append('GAGE scripts (affects --gage option [will be deprecated soon])')
     if install_full:
         logger.info('* Downloading Manta *')
         if not download_manta(logger):
@@ -176,9 +179,6 @@ if cmd_in(['install', 'develop', 'build', 'build_ext']):
         logger.info('* Downloading SILVA 16S rRNA gene database and BLAST *')
         if not download_all_blast_binaries(logger) or not download_blastdb(logger):
             modules_failed_to_install.append('SILVA 16S rRNA gene database and BLAST (affects metaquast.py in without references mode)')
-        logger.info('* Compiling GAGE *')
-        if not compile_gage():
-            modules_failed_to_install.append('GAGE scripts (affects --gage option [will be deprecated soon])')
     logger.info('')
 
 
-- 
2.7.2

