diff --git a/MANIFEST.in b/MANIFEST.in
index dfc30a89..e27e22e3 100644
--- a/MANIFEST.in
+++ b/MANIFEST.in
@@ -19,6 +19,8 @@ global-exclude *.pyc
 
 recursive-exclude quast_libs/blast *
 recursive-exclude quast_libs/manta/build *
+recursive-exclude quast_libs/genemark *
+recursive-exclude quast_libs/genemark-es *
 
 # include this file, to ensure we can recreate source distributions
 include MANIFEST.in
diff --git a/quast_libs/genemark.py b/quast_libs/genemark.py
index 474e6c8f..02c91540 100644
--- a/quast_libs/genemark.py
+++ b/quast_libs/genemark.py
@@ -21,7 +21,7 @@ from quast_libs.qutils import is_python2
 
 logger = get_logger(qconfig.LOGGER_DEFAULT_NAME)
 
-LICENSE_LIMITATIONS_MODE = False
+LICENSE_LIMITATIONS_MODE = True
 OUTPUT_FASTA = False  # whether output only .gff or with corresponding .fasta files
 
 
diff --git a/setup.py b/setup.py
index 61265258..f34da099 100755
--- a/setup.py
+++ b/setup.py
@@ -234,9 +234,6 @@ The tool accepts multiple assemblies, thus is suitable for comparison.''',
             ] +
             find_package_files('html_saver') +
             minimap_files +
-            find_package_files('genemark/' + qconfig.platform_name) +
-            find_package_files('genemark-es/' + qconfig.platform_name) +
-            find_package_files('genemark-es/lib') +
             find_package_files('glimmer') +
             bwa_files +
             bedtools_files +
