From 09f54c635f913135c4ede2ec5d03fb90b42ab042 Mon Sep 17 00:00:00 2001
From: almiheenko <al.miheenko@gmail.com>
Date: Mon, 13 Aug 2018 12:25:32 +0300
Subject: [PATCH] Glimmer: do not crush if glimmer failed

---
 quast_libs/glimmer.py | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/quast_libs/glimmer.py b/quast_libs/glimmer.py
index 867711f2..101e5962 100644
--- a/quast_libs/glimmer.py
+++ b/quast_libs/glimmer.py
@@ -80,7 +80,7 @@ def glimmerHMM(tool_dir, fasta_fpath, out_fpath, gene_lengths, err_path, tmp_dir
             contigs[ind] = seq
 
     if not gffs:
-        return None, None, None, None
+        return None, None, None, None, None, None
 
     out_gff_fpath = out_fpath + '_genes.gff' + ('.gz' if not qconfig.no_gzip else '')
     out_gff_path = merge_gffs(gffs, out_gff_fpath)
@@ -198,8 +198,8 @@ def do(contigs_fpaths, gene_lengths, out_dirpath):
             report.add_field(reporting.Fields.PREDICTED_GENES, genes)
         if unique is None and full_genes is None:
             logger.error(
-                'Glimmer failed running Glimmer for %s. ' + ('Run with the --debug option'
-                ' to see the command line.' if not qconfig.debug else '') % label)
+                'Failed running Glimmer for %s. ' % label + ('Run with the --debug option'
+                ' to see the command line.' if not qconfig.debug else ''))
 
     if not qconfig.debug:
         shutil.rmtree(tmp_dirpath)
-- 
2.17.1

