diff --git a/setup.py b/setup.py
index 27443ca5..3cfbf4e9 100755
--- a/setup.py
+++ b/setup.py
@@ -246,10 +246,6 @@ The tool accepts multiple assemblies, thus is suitable for comparison.''',
     include_package_data=True,
     zip_safe=False,
     scripts=['quast.py', 'metaquast.py', 'icarus.py', 'quast-lg.py'],
-    install_requires=[
-        'joblib',
-        'simplejson',
-    ],
     classifiers=[
         'Environment :: Console',
         'Environment :: Web Environment',
