diff --git a/quast_libs/site_packages/simplejson/tool.py b/quast_libs/site_packages/simplejson/tool.py
index 73370db..9676125 100644
--- a/quast_libs/site_packages/simplejson/tool.py
+++ b/quast_libs/site_packages/simplejson/tool.py
@@ -29,7 +29,7 @@ def main():
         obj = json.load(infile,
                         object_pairs_hook=json.OrderedDict,
                         use_decimal=True)
-    except ValueError, e:
+    except ValueError as e:
         raise SystemExit(e)
     json.dump(obj, outfile, sort_keys=True, indent='    ', use_decimal=True)
     outfile.write('\n')
