diff --git a/quast_libs/options_parser.py b/quast_libs/options_parser.py
index d3b8bcf6..54581db8 100644
--- a/quast_libs/options_parser.py
+++ b/quast_libs/options_parser.py
@@ -772,7 +772,7 @@ def parse_options(logger, quast_args):
             if not is_metaquast:
                 qconfig.features = test_features
                 qconfig.operons = test_operons
-                qconfig.glimmer = True
+                qconfig.glimmer = False
                 if not qconfig.large_genome:  # special case -- large mode imposes eukaryote gene finding (GeneMark-ES) and our test data is too small for it.
                     qconfig.gene_finding = True
         if qconfig.test_sv:
