diff --git a/quast_libs/glimmer.py b/quast_libs/glimmer.py
index 101e5962..bf467e23 100644
--- a/quast_libs/glimmer.py
+++ b/quast_libs/glimmer.py
@@ -51,12 +51,13 @@ def parse_gff(gff_path):
 
 
 def glimmerHMM(tool_dir, fasta_fpath, out_fpath, gene_lengths, err_path, tmp_dir, index):
+    import sys
     def run(contig_path, tmp_path):
         with open(err_path, 'a') as err_file:
             return_code = qutils.call_subprocess(
                 [tool_exec, contig_path, '-d', trained_dir, '-g', '-o', tmp_path],
-                stdout=err_file,
-                stderr=err_file,
+                stdout=sys.stdout,
+                stderr=sys.stdout,
                 indent='  ' + qutils.index_to_str(index) + '  ')
             return return_code
 
