diff --git a/quast_libs/qutils.py b/quast_libs/qutils.py
index 808a02f4..fae19310 100644
--- a/quast_libs/qutils.py
+++ b/quast_libs/qutils.py
@@ -859,13 +859,13 @@ def compile_tool(name, dirpath, requirements, just_notice=False, logger=logger,
         if configure_args:
             prev_dir = os.getcwd()
             os.chdir(dirpath)
-            call_subprocess(['./configure'] + configure_args, stdout=open(make_logs_basepath + '.log', 'a'),
-                            stderr=open(make_logs_basepath + '.err', 'a'))
+            call_subprocess(['./configure'] + configure_args, stdout=None,
+                            stderr=None)
             os.chdir(prev_dir)
         try:
             return_code = call_subprocess((['make', make_cmd] if make_cmd else ['make']) + ['-C', dirpath],
-                                      stdout=open(make_logs_basepath + '.log', 'a'),
-                                      stderr=open(make_logs_basepath + '.err', 'a'), logger=logger)
+                                      stdout=None,
+                                      stderr=None, logger=logger)
         except IOError:
             msg = 'Permission denied accessing ' + dirpath + '. Did you forget sudo?'
             if just_notice:
