diff --git a/INSTALL b/INSTALL
index 33a8f20..8371c51 100755
--- a/INSTALL
+++ b/INSTALL
@@ -1,13 +1,13 @@
 #!/bin/sh
 
-DESTDIR=${DESTDIR:="/"}
-
-BINDIR=$DESTDIR/usr/bin
-DOCDIR=$DESTDIR/usr/share/doc/collectl
-SHRDIR=$DESTDIR/usr/share/collectl
-MANDIR=$DESTDIR/usr/share/man/man1
-SYSDDIR=$DESTDIR/usr/lib/systemd/system
-ETCDIR=$DESTDIR/etc
+DESTDIR=${DESTDIR:="${PREFIX}"}
+
+BINDIR=$DESTDIR/bin
+DOCDIR=$DESTDIR/share/doc/collectl
+SHRDIR=$DESTDIR/share/collectl
+MANDIR=$DESTDIR/share/man/man1
+SYSDDIR=$DESTDIR/lib/systemd/system
+ETCDIR=$DESTDIR/bin
 INITDIR=$ETCDIR/init.d
 
 mkdir -p $BINDIR
