diff -u -ur collectl-4.0.4/collectl collectl-4.0.4.patched/collectl
--- collectl-4.0.4/collectl	2016-01-29 15:38:21.000000000 +0100
+++ collectl-4.0.4.patched/collectl	2016-12-19 22:18:03.486134053 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl
 
 # Copyright 2003-2016 Hewlett-Packard Development Company, L.P. 
 #
diff -u -ur collectl-4.0.4/colmux collectl-4.0.4.patched/colmux
--- collectl-4.0.4/colmux	2016-01-29 15:38:21.000000000 +0100
+++ collectl-4.0.4.patched/colmux	2016-12-19 22:18:12.194036392 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env perl
 
 # problems
 #  pattern match for -i recognition/removal (in -test) not robust enough
