--- recipes/longtr/Makefile	2025-09-22 13:15:58
+++ recipes/longtr/Makefile_patched	2025-09-22 13:17:29
@@ -25,10 +25,9 @@
 
 CEPHES_ROOT=lib/cephes
 
-LIBS              = -L./ -lm -Llib/htslib/lib -lhts -ldeflate -lz -L$(CEPHES_ROOT)/ -llzma -lbz2 -lcurl -lcrypto -Llib/spoa/build/lib -lspoa
-INCLUDE           = -Ilib -Ilib/htslib/include -Ilib/spoa/include
+LIBS              = -L./ -lm -lhts -ldeflate -lz -L$(CEPHES_ROOT)/ -llzma -lbz2 -lcurl -lcrypto -lspoa
+INCLUDE           = -Ilib
 CEPHES_LIB        = lib/cephes/libprob.a
-HTSLIB_LIB        = lib/htslib/lib/libhts.a
 
 # For each CPP file, generate an object file
 OBJ_COMMON  := $(SRC_COMMON:.cpp=.o)
