diff --git a/HCCeval.py b/HCCeval.py
index 08dd1b3..2adccbd 100755
--- a/HCCeval.py
+++ b/HCCeval.py
@@ -40,7 +40,7 @@ def get_similarity(method, cluster, stepwise, pool) :
 def get_silhouette(profile, cluster, stepwise, pool) :
     logging.info('Calculating pairwise distance ...')
     dist = getDistance(profile, 'p_dist', pool)
-    with NamedTemporaryFile(dir='.', prefix='HCCeval_') as file :
+    with NamedTemporaryFile(prefix='HCCeval_') as file :
         dist_buf = 'file://{0}.dist'.format(file.name)
         dist2 = sa.create(dist_buf, dist.shape[:2], dist.dtype)
         dist2[:] = dist[:, :, 0] + dist[:, :, 0].T
