diff --git a/src/MinCollector.cpp b/src/MinCollector.cpp
index db3c4ec..c50fa91 100644
--- a/src/MinCollector.cpp
+++ b/src/MinCollector.cpp
@@ -1,5 +1,6 @@
 #include "MinCollector.h"
 #include <algorithm>
+#include <limits>
 
 // utility functions
 
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 650baa9..83adcaa 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 2.8.12)
+cmake_minimum_required(VERSION 3.5)
 
 project(kallisto)
 
@@ -40,8 +40,7 @@ ExternalProject_Add(htslib
     PREFIX ${PROJECT_SOURCE_DIR}/ext/htslib
     SOURCE_DIR ${PROJECT_SOURCE_DIR}/ext/htslib
     BUILD_IN_SOURCE 1
-    CONFIGURE_COMMAND autoheader && autoconf && ${PROJECT_SOURCE_DIR}/ext/htslib/configure 
-        --prefix=${PREFIX} --disable-bz2 --disable-lzma --disable-libcurl
+    CONFIGURE_COMMAND ./configure --prefix=${PREFIX} --host=x86_64 CC=${CC} --disable-bz2 --disable-lzma --disable-libcurl
     BUILD_COMMAND make lib-static
     INSTALL_COMMAND ""
 )
