--- nucleoatac/NucleosomeCalling.py
+++ nucleoatac/NucleosomeCalling.py
@@ -11 +11 @@
-import pyximport; pyximport.install(setup_args={"include_dirs":np.get_include()})
+
--- nucleoatac/Occupancy.py
+++ nucleoatac/Occupancy.py
@@ -11 +11 @@
-import pyximport; pyximport.install(setup_args={"include_dirs":np.get_include()})
+
--- pyatac/chunkmat2d.py
+++ pyatac/chunkmat2d.py
@@ -6 +6 @@
-import pyximport; pyximport.install(setup_args={"include_dirs":np.get_include()})
+
--- pyatac/fragmentsizes.py
+++ pyatac/fragmentsizes.py
@@ -8,2 +8,2 @@
-import pyximport
-pyximport.install(setup_args={"include_dirs":np.get_include()})
+
+
--- pyatac/get_cov.py
+++ pyatac/get_cov.py
@@ -15 +15 @@
-import pyximport; pyximport.install()
+
--- pyatac/get_ins.py
+++ pyatac/get_ins.py
@@ -15 +15 @@
-import pyximport; pyximport.install(setup_args={"include_dirs":np.get_include()})
+
--- pyatac/get_pwm.py
+++ pyatac/get_pwm.py
@@ -16,2 +16,2 @@
-import pyximport
-pyximport.install(setup_args={"include_dirs":np.get_include()})
+
+
--- pyatac/make_vplot.py
+++ pyatac/make_vplot.py
@@ -13 +13 @@
-import pyximport; pyximport.install(setup_args={"include_dirs":np.get_include()})
+
--- pyatac/tracks.py
+++ pyatac/tracks.py
@@ -12 +12 @@
-import pyximport; pyximport.install(setup_args={"include_dirs":np.get_include()})
+
--- setup.py
+++ setup.py
@@ -4,0 +5,3 @@
+from setuptools import Extension
+from Cython.Build import cythonize
+import numpy
@@ -31 +34,5 @@
-    install_requires=['cython >= 0.22','numpy >= 1.9.1', 'scipy >= 0.16.0','pysam >= 0.10.0','matplotlib'],
+    install_requires=['numpy >= 1.9.1', 'scipy >= 0.16.0','pysam >= 0.10.0','matplotlib'],
+    ext_modules=cythonize([
+        Extension("nucleoatac.multinomial_cov", ["nucleoatac/multinomial_cov.pyx"], include_dirs=[numpy.get_include()]),
+        Extension("pyatac.fragments", ["pyatac/fragments.pyx"], include_dirs=[numpy.get_include()]),
+    ]),
