diff --git a/Makefile b/Makefile
index eae151a..16bfda4 100755
--- a/Makefile
+++ b/Makefile
@@ -4,19 +4,19 @@
 all : realign_cm merge_result read_extractor
 
 edlib.o: utils/edlib.cpp utils/edlib.h
-	g++ -std=c++11 -I./ -I./utils -g -c -Wall utils/edlib.cpp -o edlib.o
+	$(CXX) -march=x86-64-v3 -std=c++11 -I./ -I./utils -I$(PREFIX)/include -g -c -Wall utils/edlib.cpp -o edlib.o
 
 realign_cm: realign_cm.cpp  utils/incr_array.h utils/MultiThread.h  edlib.o utils/edlib.cpp utils/edlib.h
-	g++ -std=c++11 -I./ -g -c -Wall realign_cm.cpp -o realign_cm.o
-	g++ -std=c++11 -o realign_cm realign_cm.o -lhts -lz -lpthread edlib.o
+	$(CXX) -march=x86-64-v3 -std=c++11 -I./ -I$(PREFIX)/include -g -c -Wall realign_cm.cpp -o realign_cm.o
+	$(CXX) -march=x86-64-v3 -std=c++11 -I$(PREFIX)/include realign_cm.o edlib.o -o realign_cm -L$(PREFIX)/lib -lhts -lz -pthread
 
 merge_result: merge_result.cpp
-	g++ -std=c++11 -g -c -Wall merge_result.cpp -o merge_result.o
-	g++ -std=c++11 merge_result.o  -lpthread -o merge_result 
+	$(CXX) -march=x86-64-v3 -std=c++11 -g -c -Wall merge_result.cpp -o merge_result.o
+	$(CXX) -march=x86-64-v3 -std=c++11 merge_result.o -o merge_result -L$(PREFIX)/lib -pthread
 
 read_extractor: read_extractor.cpp
-	g++ -std=c++11 -g -c -Wall read_extractor.cpp -o read_extractor.o
-	g++ -std=c++11 read_extractor.o  -lpthread -lhts -o read_extractor
+	$(CXX) -march=x86-64-v3 -std=c++11 -g -c -Wall -I$(PREFIX)/include read_extractor.cpp -o read_extractor.o
+	$(CXX) -march=x86-64-v3 -std=c++11 -I$(PREFIX)/include read_extractor.o -o read_extractor -L$(PREFIX)/lib -pthread -lhts
 
 clean:
-	rm -rf *.o realign_cm merge_result 
+	rm -rf *.o realign_cm merge_result
