diff --git a/install.sh b/install.sh
index e8d9eca..12e5465 100755
--- a/install.sh
+++ b/install.sh
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/bash
 
 set -xe
 
@@ -9,16 +9,16 @@ ROOT=`pwd -P`
 # Check for gmake #
 ###################
 mkdir -p dist-bin
-PATH=$PATH:$ROOT/dist-bin
+PATH="$PATH:$ROOT/dist-bin"
 ln -sf $(which make) $ROOT/dist-bin/gmake
 ln -sf $ROOT/PkgConfig.pm $ROOT/dist-bin/pkg-config
 
 export NUM_THREADS=`grep -c '^processor' /proc/cpuinfo 2>/dev/null || sysctl -n hw.ncpu 2>/dev/null || echo 1`;
-BINDIR=$DEST/bin
-LIBDIR=$DEST/lib
-export PKG_CONFIG_PATH=$LIBDIR/pkgconfig:$PKG_CONFIG_PATH
-(cd global-1 && ./configure --prefix=$DEST --bindir=$BINDIR --libdir=$LIBDIR && make -j $NUM_THREADS install-special)
+export BINDIR="$DEST/bin"
+export LIBDIR="$DEST/lib"
+export PKG_CONFIG_PATH="$PREFIX/lib/pkgconfig:$PKG_CONFIG_PATH"
+(cd global-1 && autoreconf -if && ./configure --prefix="${DEST}" --bindir="${BINDIR}" --libdir="${LIBDIR}" && make -j"${NUM_THREADS}" install-special)
 echo "creating example config file masurca_config_example.txt"
-$BINDIR/masurca -g masurca_config_example.txt
+${SRC_DIR}/global-1/masurca -g masurca_config_example.txt
 
-(cd Flye && make && cp -a ../Flye $DEST);
+(cd Flye && make -j"${CPU_COUNT}" && cp -arf ../Flye "${DEST}");
