diff --git a/install.sh b/install.sh
index c30290e..0bce5d1 100755
--- a/install.sh
+++ b/install.sh
@@ -17,7 +17,7 @@ export NUM_THREADS=`grep -c '^processor' /proc/cpuinfo 2>/dev/null || sysctl -n
 BINDIR=$DEST/bin
 LIBDIR=$DEST/lib
 export PKG_CONFIG_PATH=$LIBDIR/pkgconfig:$PKG_CONFIG_PATH
-(cd global-1 && ./configure --prefix=$DEST --bindir=$BINDIR --libdir=$LIBDIR && make -j $NUM_THREADS install-special)
+(cd global-1 && autoreconf -if && ./configure --prefix=$DEST --bindir=$BINDIR --libdir=$LIBDIR && make -j $NUM_THREADS install-special)
 echo "creating example config file masurca_config_example.txt"
 $BINDIR/masurca -g masurca_config_example.txt
-(cd Flye && make && cp -a ../Flye $DEST);
+(cd Flye && make && cp -arf ../Flye $DEST);
