diff --git a/global-1/MUMmer/include/mummer/48bit_iterator.hpp b/global-1/MUMmer/include/mummer/48bit_iterator.hpp
index 479ae9b..8abe1e2 100644
--- a/global-1/MUMmer/include/mummer/48bit_iterator.hpp
+++ b/global-1/MUMmer/include/mummer/48bit_iterator.hpp
@@ -1,6 +1,7 @@
 #ifndef __48BIT_ITERATOR_H__
 #define __48BIT_ITERATOR_H__
 
+#include <cstdint>
 #include "const_iterator_traits.hpp"
 
 // Define the iterator and its const version
