diff --git a/Makefile b/Makefile
index c7fba73..2c8bd95 100644
--- a/Makefile
+++ b/Makefile
@@ -1,12 +1,12 @@
 TARGET = ema
-LIBS = -L$(BWADIR) -lbwa -lm -lz -lpthread
-CC = gcc
+LIBS = -L$(BWADIR) -lbwa -lm -lz -pthread
+CC = $(CC)
 WARNINGS = -Wall -Wextra
-CFLAGS = -std=gnu99 -march=x86-64 -O3 -fopenmp -flto -fstrict-aliasing $(WARNINGS)
-LFLAGS = -lstdc++ -march=x86-64 -O3 -flto -fopenmp -lpthread
+CFLAGS = -std=gnu11 -march=x86-64-v3 -O3 -fopenmp -flto -fstrict-aliasing -Wno-unused-result -Wno-unused-but-set-variable $(WARNINGS)
+LFLAGS = -lstdc++ -march=x86-64-v3 -O3 -flto -fopenmp -pthread
 
-CXX = g++
-CPPFLAGS = -c -std=c++11 -O3 -march=x86-64 -pthread
+CXX = $(CXX)
+CPPFLAGS = -c -std=c++14 -O3 -march=x86-64-v3 -pthread
 LDFLAGS = -pthread
 
 #CFLAGS = -std=gnu99 -fstrict-aliasing -fopenmp -ggdb -O0 $(WARNINGS)
