diff -u -ur ema/bwa/Makefile ema.patched/bwa/Makefile
--- ema/bwa/Makefile	2019-11-29 17:21:15.076685218 +0100
+++ ema.patched/bwa/Makefile	2018-07-30 17:47:18.000000000 +0200
@@ -3,7 +3,7 @@
 CFLAGS=		-g -Wall -Wno-unused-function -O2
 WRAP_MALLOC=-DUSE_MALLOC_WRAPPERS
 AR=			ar
-DFLAGS=		-DHAVE_PTHREAD $(WRAP_MALLOC)
+DFLAGS=		-DHAVE_PTHREAD $(WRAP_MALLOC) $(LDFLAGS)
 LOBJS=		utils.o kthread.o kstring.o ksw.o bwt.o bntseq.o bwa.o bwamem.o bwamem_pair.o bwamem_extra.o malloc_wrap.o \
 			QSufSort.o bwt_gen.o rope.o rle.o is.o bwtindex.o
 AOBJS=		bwashm.o bwase.o bwaseqio.o bwtgap.o bwtaln.o bamlite.o \
diff -u -ur ema/Makefile ema.patched/Makefile
--- ema/Makefile	2018-07-30 17:47:17.000000000 +0200
+++ ema.patched/Makefile	2019-11-29 17:19:55.357328604 +0100
@@ -38,7 +38,7 @@
 .PRECIOUS: $(TARGET) $(OBJECTS)
 
 $(TARGET): $(OBJECTS) preproc bwa
-	$(CXX) $(LFLAGS) $(OBJECTS) $(LIBS) -o $@
+	$(CXX) $(LDFLAGS) $(LFLAGS) $(OBJECTS) $(LIBS) -o $@
 
 clean:
 	-rm -f $(OBJDIR)/*.o
