From 60999a2c6a7beee4b2ce57610fc77e03501daf58 Mon Sep 17 00:00:00 2001
From: Ed Harry <eh15@sanger.ac.uk>
Date: Tue, 2 Feb 2021 14:46:27 +0000
Subject: [PATCH] ...

---
 Makefile | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/Makefile b/Makefile
index 15e57b0..c7fba73 100644
--- a/Makefile
+++ b/Makefile
@@ -2,11 +2,11 @@ TARGET = ema
 LIBS = -L$(BWADIR) -lbwa -lm -lz -lpthread
 CC = gcc
 WARNINGS = -Wall -Wextra
-CFLAGS = -std=gnu99 -march=native -O3 -fopenmp -flto -fstrict-aliasing $(WARNINGS)
-LFLAGS = -lstdc++ -march=native -O3 -flto -fopenmp -lpthread
+CFLAGS = -std=gnu99 -march=x86-64 -O3 -fopenmp -flto -fstrict-aliasing $(WARNINGS)
+LFLAGS = -lstdc++ -march=x86-64 -O3 -flto -fopenmp -lpthread
 
 CXX = g++
-CPPFLAGS = -c -std=c++11 -O3 -march=native -pthread
+CPPFLAGS = -c -std=c++11 -O3 -march=x86-64 -pthread
 LDFLAGS = -pthread
 
 #CFLAGS = -std=gnu99 -fstrict-aliasing -fopenmp -ggdb -O0 $(WARNINGS)
-- 
2.22.0

