--- gtdbtk/main.py	2018-08-30 16:11:57.000000000 +1000
+++ gtdbtk/main_new.py	2020-03-26 11:06:15.827210774 +1000
@@ -284,6 +284,7 @@

     def parse_options(self, options):
         """Parse user options and call the correct pipeline(s)"""
+        self.logger.warning('You are not running the most recent version of GTDB-Tk!')

         if(options.subparser_name == 'de_novo_wf'):
             check_dependencies(['prodigal', 'hmmalign'])
