diff --git a/Makefile b/Makefile
index 4dd33e6..5f4307f 100644
--- a/Makefile
+++ b/Makefile
@@ -1,13 +1,13 @@
-CC=			gcc
-CFLAGS=		-g -Wall -Wc++-compat -std=c99 -msse4 -O3
-CPPFLAGS=
+CC=			$(CC)
+CFLAGS=		-g -Wall -Wc++-compat -std=c11 -msse4 -O3
+CPPFLAGS=	-I$(PREFIX)/include
 INCLUDES=
 OBJS=		kalloc.o kthread.o algo.o sys.o gfa-base.o gfa-io.o gfa-aug.o gfa-bbl.o gfa-ed.o \
             sketch.o misc.o bseq.o options.o shortk.o miniwfa.o \
 			index.o lchain.o gchain1.o galign.o gcmisc.o map-algo.o cal_cov.o \
 			format.o gmap.o ggsimple.o ggen.o asm-call.o
 PROG=		minigraph
-LIBS=		-lz -lpthread -lm
+LIBS=		-lz -pthread -lm
 
 ifneq ($(asan),)
 	CFLAGS+=-fsanitize=address
