diff --git a/ncbi_genome_download/summary.py b/ncbi_genome_download/summary.py
index 0a9f7e6..a1efda0 100644
--- a/ncbi_genome_download/summary.py
+++ b/ncbi_genome_download/summary.py
@@ -1,6 +1,7 @@
 '''Parse NCBI RefSeq/GenBank summary files'''
 
 import logging
+import re
 
 
 # pylint: disable=too-few-public-methods
@@ -14,8 +15,9 @@ def __init__(self, infile):
             line = self._file.readline().rstrip('\n')
             self._lineno += 1
 
-        if line.startswith('# '):
-            line = line[2:]
+        if re.search(r'^#.*assembly_accession\t', line):
+            res = re.search(r'assembly_accession', line)
+            line = line[res.start():]
 
         self._fields = line.split('\t')
 