diff --git a/src/conditioned_sfs.cpp b/src/conditioned_sfs.cpp
index 2f41f25..47ed25d 100644
--- a/src/conditioned_sfs.cpp
+++ b/src/conditioned_sfs.cpp
@@ -105,7 +105,7 @@ std::vector<Matrix<T> > incorporate_theta(const std::vector<Matrix<T> > &csfs, d
     for (unsigned int i = 0; i < csfs.size(); ++i)
     {
         assert(csfs[i](0, 0) == 0.);
-        assert(csfs[i](2, n) == 0.);
+        assert(csfs[i](2, csfs[i].cols() - 1) == 0.);
         T tauh = csfs[i].sum();
         // if (toDouble(tauh) > 1.0 / theta)
         // {
