diff --git a/include/marginalize_key.h b/include/marginalize_key.h
index 4d00ab5..c81f5f6 100644
--- a/include/marginalize_key.h
+++ b/include/marginalize_key.h
@@ -58,7 +58,7 @@ std::map<block_key, double> marginalize_key<P>::run(
         const Eigen::MatrixBase<Derived3> &na)
 {
     std::map<block_key, double> ret;
-    assert(a(0) >= 0);
+    assert(key(0) >= 0);
     std::map<block_key, double> sub_left = marginalize_key<1>::run(
             key.head(3),
             n.head(1),
