--- Velvet-1.1.06/Makefile	2012-09-26 15:22:55.000000000 +0000
+++ Velvet-1.1.06/Makefile	2016-06-29 08:39:16.309057675 +0000
@@ -1,3 +1,6 @@
+DEBUG = -g
+LIBS = -lm -lz
+INCLUDES ="${PREFIX}/include"
 OBJS = readSet.o tightString.o kmer.o kmerOccurenceTable.o \
 	roadMap.o splay.o splayTable.o \
 	preGraph.o preGraphConstruction.o concatenatedPreGraph.o \
@@ -8,14 +11,12 @@
 	scaffold.o shortReadPairs.o locallyCorrectedGraph.o readCoherentGraph.o \
 	utility.o allocArray.o recycleBin.o
 PROG =
-INCLUDES = 
-LIBS = 
-SUBDIRS = . 
+SUBDIRS = .
 
 .SUFFIXES:.c .o
 
 .c.o:
-	$(CXX) -c $(CXXFLAGS) $(DFLAGS) $(DEF) $(INCLUDES) $< -o $@
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) -c  $(DFLAGS) $(DEF) $(INCLUDES) $< -o $@
 
 lib:libVelvet.a
 
