From 850b11cc50863272c7196228a4fd16353ba433fa Mon Sep 17 00:00:00 2001
From: Marcel Bargull <marcel.bargull@udo.edu>
Date: Thu, 1 Dec 2022 13:59:53 +0100
Subject: [PATCH] Add missing cmath for cos(...) usage

---
 src/lib/RbtDihedralSF.cxx | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/lib/RbtDihedralSF.cxx b/src/lib/RbtDihedralSF.cxx
index 62531c3..49d03f0 100644
--- a/src/lib/RbtDihedralSF.cxx
+++ b/src/lib/RbtDihedralSF.cxx
@@ -11,6 +11,7 @@
 ***********************************************************************/
 
 #include "RbtDihedralSF.h"
+#include <cmath>
 
 RbtDihedral::RbtDihedral(RbtAtom* pAtom1, RbtAtom* pAtom2, RbtAtom* pAtom3, RbtAtom* pAtom4, const prms& dihprms)
   :  m_pAtom1(pAtom1),m_pAtom2(pAtom2),m_pAtom3(pAtom3),m_pAtom4(pAtom4)
