diff --git a/src/cpp/makefile b/src/cpp/makefile
index bb9d77d..53a97cd 100644
--- a/src/cpp/makefile
+++ b/src/cpp/makefile
@@ -1,137 +1,137 @@
-CPPFLAGS=-Wall -O4 --fast-math
+CPPFLAGS=-Wall -O3 --fast-math -std=c++14
 BINPATH=../../bin
 VPATH=../../bin
-TARGETS=Fido FidoChooseParameters 
+TARGETS=Fido FidoChooseParameters
 #TARGETS=ROC ROC-Interpolate ROC50 ROCDifferenceInterpolate WTest MeanStdDev Plottify StringCounter Fido EM Pivdo Fasta2MissingPeptides MCMC2Pivdo CrossValSplitter Rlist2FPLists FPListIntersection FDR FDRDivergence Rank RelativeOptimize SensitivityAtFDR FPlottify FidoDetectability KDTest FidoKnownDetectability FidoMAP FidoChooseParameters
 #TARGETS=FS-LP Grid-LP ROC FP-Princeps ROC-Interpolate Simplex GenerateLP GenerateSparseLP GridSimplex ProtProph2ROC ROC50 ROCDifferenceInterpolate AS-LP WTest exe MeanStdDev Plottify StringCounter Fido EM
 
 all: $(TARGETS)
 
 depend:
-	g++ -MM *.cpp *.h
+	$(CXX) -MM *.cpp *.h
 
 clean:
 	rm *.o;
 
 KDTest: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o ProteinIdentifier.o BasicBigraph.o BasicGroupBigraph.o GroupPowerBigraph.o PeptideFeatureReader.o KDMethod.o mainKDMethod.o PivdoSplitter.o
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
-exe: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o Random.o main.o 
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+exe: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o Random.o main.o
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 MCMC2Pivdo: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o ProteinIdentifier.o BasicBigraph.o BasicGroupBigraph.o  GroupPowerBigraph.o PeptideFeatureReader.o KDMethod.o PivdoSplitter.o mainMCMC2Pivdo.o 
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 CrossValSplitter: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o ProteinIdentifier.o BasicBigraph.o BasicGroupBigraph.o  GroupPowerBigraph.o PeptideFeatureReader.o KDMethod.o PivdoSplitter.o mainCrossValSplit.o 
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 Plottify: Set.o Vector.o Numerical.o Matrix.o Random.o plottify.o 
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 FPlottify: Set.o Vector.o Numerical.o Matrix.o Random.o fplottify.o 
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 MSE: Set.o Vector.o Numerical.o Matrix.o Random.o MSE.o 
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 FDRDivergence: Set.o Vector.o Numerical.o Matrix.o Random.o fdrDivergence.o 
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 ROCDifferenceInterpolate: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o Random.o ROCInterpolator.o rocDifferenceInterpolate.o 
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 WTest: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o Random.o w_test.o 
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 MeanStdDev: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o Random.o meanStdDev.o 
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 FS-LP: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o mainForwardSampler.o 
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 AS-LP: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o AffineScaling.o Random.o mainAffineScaling.o 
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 Simplex: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o Simplex.o Random.o mainSimplex.o 
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 GridSimplex: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o Simplex.o GridSimplex.o Random.o mainGridSimplex.o 
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 GenerateLP: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o Random.o generateLinProg.o 
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 GenerateSparseLP: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o Random.o generateSparseLinProg.o 
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 Grid-LP: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o GridLP.o mainGridLP.o 
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 Rank: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o rank.o
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 ROC: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o rlistROC.o
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 SensitivityAtFDR: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o rlistSensitivityAtFDR.o
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 RelativeOptimize: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o relativeOptimize.o
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 FDR: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o rlistFDR.o
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 Rlist2FPLists: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o rlistFPLists.o
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 FPListIntersection: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o FPListIntersection.o
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 StringCounter: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o Random.o stringCounter.o
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 FP-Princeps: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o firstFPs.o
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 ROC50: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o roc50.o
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 ProtProph2ROC: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o GridLP.o mainROC_ProtProph.o
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 ROC-Interpolate: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o GridLP.o ROCInterpolator.o interpolateROC.o
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 Fasta2MissingPeptides: Set.o Vector.o Numerical.o Matrix.o Random.o FastaReader.o Proteomics.o testFasta.o ProteinBigraph.o BasicBigraph.o BasicGroupBigraph.o   PivdoBigraph.o ProteinIdentifier.o
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 Pivdo: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o ProteinBigraph.o BasicBigraph.o BasicGroupBigraph.o   PivdoBigraph.o ProteinIdentifier.o mainPivdo.o 
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 Fido: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o ProteinIdentifier.o BasicBigraph.o BasicGroupBigraph.o  GroupPowerBigraph.o PeptideFeatureReader.o KDMethod.o mainFido.o PivdoSplitter.o
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 FidoChooseParameters: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o ProteinIdentifier.o BasicBigraph.o BasicGroupBigraph.o  GroupPowerBigraph.o PeptideFeatureReader.o KDMethod.o mainFido_ChooseParameters.o PivdoSplitter.o InputFileShrinker.o
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 FidoMAP: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o ProteinIdentifier.o BasicBigraph.o BasicGroupBigraph.o  GroupPowerBigraph.o PeptideFeatureReader.o KDMethod.o mainFidoMAP.o PivdoSplitter.o
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 Fido_Likelihood: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o ProteinIdentifier.o BasicBigraph.o BasicGroupBigraph.o  GroupPowerBigraph.o PeptideFeatureReader.o KDMethod.o mainFido_Likelihood.o PivdoSplitter.o
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 FidoKnownDetectability: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o ProteinIdentifier.o BasicBigraph.o BasicGroupBigraph.o  GroupPowerBigraph.o PeptideFeatureReader.o KDMethod.o mainFidoLoadDetectability.o PivdoSplitter.o
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 FidoDetectability: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o ProteinIdentifier.o BasicBigraph.o BasicGroupBigraph.o GroupPowerBigraph.o PeptideFeatureReader.o KDMethod.o mainFidoDetectability.o PivdoSplitter.o
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 EM: Set.o Vector.o Numerical.o Matrix.o GramSchmidt.o LPSolver.o ForwardSampler.o Random.o ProteinIdentifier.o BasicBigraph.o BasicGroupBigraph.o EMBigraph.o  mainEM.o 
-	g++ $(CPPFLAGS) $^ -o $(BINPATH)/$@
+	$(CXX) $(CPPFLAGS) $^ -o $(BINPATH)/$@
 
 %.o: %.cpp
-	g++ $(CPPFLAGS) -c $<
+	$(CXX) $(CPPFLAGS) -c $<
 
 #automade
 AffineScaling.o: AffineScaling.cpp AffineScaling.h LPSolver.h Numerical.h \
