diff --git a/setup.py b/setup.py
index f9d07b6..5747298 100644
--- a/setup.py
+++ b/setup.py
@@ -2,6 +2,7 @@
 # -*- coding: utf-8 -*-
 
 from setuptools import setup
+import os
 
 
 requirements = [
@@ -14,13 +15,12 @@ test_requirements = [
 ]
 
 setup_requirements = [
-    'setuptools_scm'
 ]
 
 
 setup(
     name = 'superfocus',
-    use_scm_version = True,
+    version=os.environ['PKG_VERSION'],
     description = 'SUPER-FOCUS 0.30: A tool for agile functional analysis of shotgun metagenomic data',
     author = 'Genivaldo G.Z. Silva',
     author_email = 'genivaldo.gueiros@gmail.com',
