--- COCACOLA-python/cocacola.py	2017-03-06 00:55:57.000000000 +0100
+++ COCACOLA-clean/cocacola.py	2018-04-24 09:52:37.707098074 +0200
@@ -1,3 +1,5 @@
+#!/usr/bin/env python
+
 import time
 import math
 import sys
@@ -193,10 +195,10 @@ if __name__ == '__main__':
     if bestK == 0:
         print("No cluster number specified. Now automatically detect...")
     
-        fragScanURL = os.path.join(os.getcwd(),'auxiliary','FragGeneScan1.19','run_FragGeneScan.pl'); os.system("chmod 777 " + fragScanURL);
-        hmmExeURL = os.path.join(os.getcwd(),'auxiliary','hmmer-3.1b1','bin','hmmsearch'); os.system("chmod 777 " + hmmExeURL);
-        markerExeURL = os.path.join(os.getcwd(),'auxiliary','test_getmarker.pl'); os.system("chmod 777 " + markerExeURL);
-        markerURL = os.path.join(os.getcwd(),'auxiliary','marker.hmm'); 
+        fragScanURL = 'run_FragGeneScan.pl';
+        hmmExeURL = 'hmmsearch';
+        markerExeURL = 'test_getmarker.pl';
+        markerURL = os.path.join(os.path.dirname(os.path.realpath(sys.argv[0])),'auxiliary','marker.hmm');
         seedURL = args.contig_file+".seed";
         fragResultURL = args.contig_file+".frag.faa"; 
         hmmResultURL = args.contig_file+".hmmout";
