diff --git a/Makefile b/Makefile
index 5fb3cbb..93bb002 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 .PHONY=clean
 
-CXX?=g++
-CC?=gcc
+CXX?=$(CXX)
+CC?=$(CC)
 
 CACHE_SIZE?=4194304
 CACHE_SIZE_FLAG:=-DD2_CACHE_SIZE=${CACHE_SIZE}
@@ -14,11 +14,11 @@ UNAME_P := $(shell uname -p)
 ifeq ($(UNAME_P),arm)
 	TARGET_FLAG=-target arm64-apple-macos11 -mmacosx-version-min=11.0
 else
-    TARGET_FLAG=-march=native
+	TARGET_FLAG=-march=native
 endif
 
-LIB=-lz # -lfmt
-INC=-IlibBigWig -Ibonsai/include -Ibonsai -Ibonsai/hll -Ibonsai/hll/include -Ibonsai -I. -Isrc -Ifmt/include
+LIB=-lz -L$(PREFIX)/lib # -lfmt
+INC=-IlibBigWig -Ibonsai/include -Ibonsai -Ibonsai/hll -Ibonsai/hll/include -Ibonsai -I. -Isrc -Ifmt/include -I$(PREFIX)/include
 OPT+= -O3 \
     $(TARGET_FLAG) \
     -fopenmp -pipe $(CACHE_SIZE_FLAG)
