diff --git a/m4/ax_python_devel.m4 b/m4/ax_python_devel.m4
index ffebc82..c065642 100644
--- a/m4/ax_python_devel.m4
+++ b/m4/ax_python_devel.m4
@@ -139,8 +139,8 @@ to something else than an empty string.
 	if test -n "$1"; then
 		AC_MSG_CHECKING([for a version of Python $1])
 		ac_supports_python_ver=`$PYTHON -c "import sys; \
-			ver = sys.version.split ()[[0]]; \
-			print (ver $1)"`
+			ver = sys.version_info[[:2]]; \
+			print (ver >= (3, 5))"`
 
 		if test "$ac_supports_python_ver" = "True"; then
 		   AC_MSG_RESULT([yes])
