--- ./src/junction_filter.cc	2025-04-04 12:12:12.200316679 +0100
+++ ./src/junction_filter.cc	2025-04-04 12:12:12.202316396 +0100
@@ -152,7 +152,7 @@
 
 void portcullis::JunctionFilter::filter() {
     path outputDir = output.parent_path();
-    string outputPrefix = output.leaf().string();
+    string outputPrefix = output.filename().string();
     if (outputDir.empty()) {
         outputDir = ".";
     }
--- ./src/junction_builder.cc	2025-04-04 12:12:12.281305218 +0100
+++ ./src/junction_builder.cc	2025-04-04 12:12:12.285304652 +0100
@@ -63,7 +63,7 @@
 portcullis::JunctionBuilder::JunctionBuilder(const path& _prepDir, const path& _output) {
 	prepData = PreparedFiles(_prepDir);
 	outputDir = _output.empty() ? path(".") : _output.parent_path();
-	outputPrefix = _output.empty() ? "portcullis" : _output.leaf().string();
+	outputPrefix = _output.empty() ? "portcullis" : _output.filename().string();
 	threads = 1;
 	extra = false;
 	useCsi = false;
--- ./lib/include/portcullis/bam/genome_mapper.hpp	2025-04-04 12:12:12.293303520 +0100
+++ ./lib/include/portcullis/bam/genome_mapper.hpp	2025-04-04 12:12:12.297302954 +0100
@@ -79,7 +79,7 @@
 
 
 	path getFastaIndexFile() const {
-		return path(genomeFile.parent_path()) /= path(genomeFile.leaf().string() + ".fai");
+		return path(genomeFile.parent_path()) /= path(genomeFile.filename().string() + ".fai");
 	}
 
 
