--- src/Makefile.am	2025-04-04 22:13:46.590136447 +0100
+++ src/Makefile.am	2025-04-04 22:14:13.339131263 +0100
@@ -6,7 +6,6 @@
 portcullis_noinstdir = $(top_srcdir)/src
 
 portcullis_CXXFLAGS = \
-			-m64 \
 			-Wno-unused-function -Wno-unused-variable -Wno-ignored-qualifiers \
 			-Wall \
 			@AM_CXXFLAGS@
--- lib/Makefile.am	2025-04-04 22:13:52.439135319 +0100
+++ lib/Makefile.am	2025-04-04 22:14:21.828129605 +0100
@@ -72,7 +72,6 @@
 	@AM_CPPFLAGS@
 
 libportcullis_la_CXXFLAGS = \
-	-m64 \
 	-Wno-unused-function -Wno-unused-variable -Wno-ignored-qualifiers \
 	-Wall \
 	-fPIC \
