diff --git i/lib/Makefile.am w/lib/Makefile.am
index fd3f7fc5..e32417ce 100644
--- i/lib/Makefile.am
+++ w/lib/Makefile.am
@@ -72,7 +72,6 @@ libportcullis_la_CPPFLAGS = \
 	@AM_CPPFLAGS@
 
 libportcullis_la_CXXFLAGS = \
-	-m64 \
 	-Wno-unused-function -Wno-unused-variable -Wno-ignored-qualifiers \
 	-Wall \
 	-fPIC \
diff --git i/src/Makefile.am w/src/Makefile.am
index 9b3fb6c1..791c3dae 100644
--- i/src/Makefile.am
+++ w/src/Makefile.am
@@ -6,7 +6,6 @@ bin_PROGRAMS = portcullis
 portcullis_noinstdir = $(top_srcdir)/src
 
 portcullis_CXXFLAGS = \
-			-m64 \
 			-Wno-unused-function -Wno-unused-variable -Wno-ignored-qualifiers \
 			-Wall \
 			@AM_CXXFLAGS@
