--- BlastAlignP.bak	2014-04-08 11:09:42.000000000 +0200
+++ BlastAlignP	2017-02-21 17:18:32.169343005 +0100
@@ -1,4 +1,4 @@
-#! /usr/bin/perl -w
+#! /usr/bin/env perl
 use Getopt::Std; 	# read in data at command line using -i type options
 
 # Written by Robert Belshaw (School of Biomedical & Healthcare Sciences, University of Plymouth) & Aris Katzourakis (Department of Zoology, University of Oxford)
