--- BlastAlign.bak	2017-02-21 08:57:30.000000000 +0100
+++ BlastAlign	2017-02-21 16:50:02.360998557 +0100
@@ -1,4 +1,4 @@
-#! /usr/bin/perl -w
+#! /usr/bin/env perl
 use Getopt::Std; 	# read in data at command line using -i type options
 
 # Written by Robert Belshaw (School of Biomedical & Healthcare Sciences, University of Plymouth) & Aris Katzourakis (Department of Zoology, University of Oxford)
