--- BlastAlign	2017-09-28 17:59:40.094095780 +0200
+++ BlastAlign.bak	2017-09-28 18:03:23.090998508 +0200
@@ -549,6 +549,10 @@
 		$final_name = $_->{"final_name"};
 		$final_string = $_->{"final_string"}; 
 		$number_of_gaps = ($final_string =~ tr/-/-/);
+		if ($length == 0) {
+		    print "Unauthorized division by 0 upcoming ... Exiting BlastAlign - failed to align sequences of this file.";
+		    exit;
+		}
 		$percentage_gaps = sprintf "%.1f", ($number_of_gaps / $length) * 100;
 		if ($percentage_gaps < $max_percentage) {
 			push @nexus_line_array, $final_name, "\t", $final_string, "\n";
