diff --git a/libs/vectorf128.h b/libs/vectorf128.h
index bc4226a..efb81cb 100644
--- a/libs/vectorf128.h
+++ b/libs/vectorf128.h
@@ -136,7 +136,7 @@ public:
     operator __m128() const {
         return xmm;
     }
-#if defined (__clang__) && CLANG_VERSION < 30900 || defined(__apple_build_version__)
+#if defined (__clang__) && CLANG_VERSION < 90000 || defined(__apple_build_version__)
 #define FIX_CLANG_VECTOR_ALIAS_AMBIGUITY  // clang 3.3 - 3.5 has silent conversion between intrinsic vector types.
                                           // I expected this to be fixed in version 3.4 but it still exists!
                                           // http://llvm.org/bugs/show_bug.cgi?id=17164
