diff --git a/src/tree/UPGMA.cpp b/src/tree/UPGMA.cpp
index 4897fc6..1ad7e8e 100644
--- a/src/tree/UPGMA.cpp
+++ b/src/tree/UPGMA.cpp
@@ -104,10 +104,10 @@ void UPGMA::computeDistances(std::vector<CSequence>& sequences, UPGMA_dist_t *di
 }
 
 // *******************************************************************
-template <bool is_modified>
+template <bool MODIFIED_T>
 void UPGMA::computeTree(UPGMA_dist_t* distances, size_t n_seq, tree_structure& tree)
 {
-	Average<UPGMA_dist_t, is_modified> average;
+	Average<UPGMA_dist_t, MODIFIED_T> average;
 	
 	uint64_t g_uLeafCount;
 	uint64_t g_uTriangleSize;
