From 381bead2046a8814b6c7d40290f36bcf595cec6e Mon Sep 17 00:00:00 2001
From: Lightning Auriga <lightning.auriga@gmail.com>
Date: Sun, 17 Sep 2023 22:29:25 -0400
Subject: [PATCH 2/3] enable 2014 standard

---
 CMakeLists.txt | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 76c0720..6388656 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -31,8 +31,8 @@ project(CHAP LANGUAGES CXX C VERSION 0.9.1)
 # follow GNU guideline on install directories:
 include(GNUInstallDirs)
 
-# set C++11 standard and turn of GNU extensions:
-set(CMAKE_CXX_STANDARD 11)
+# set C++14 standard and turn of GNU extensions:
+set(CMAKE_CXX_STANDARD 14)
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 set(CMAKE_CXX_EXTENSIONS OFF)
 
-- 
2.41.0

