--- Makefile	2023-03-24 12:00:55.075873598 +0100
+++ Makefile	2023-03-24 12:03:00.176181845 +0100
@@ -2,13 +2,13 @@
 	@echo "Maast build completed."
 
 callm_db_build:  ./src/callm_db_build.cpp Makefile
-	g++ -std=c++11 ./src/callm_db_build.cpp -o ./bin/callm_db_build -O3 -lpthread
+	$(CXX) -std=c++11 ./src/callm_db_build.cpp -o ./bin/callm_db_build -O3 -lpthread
 
 callm_db_val: ./src/callm_db_val.cpp Makefile
-	g++ -std=c++11 ./src/callm_db_val.cpp -o ./bin/callm_db_val -O3 -lpthread
+	$(CXX) -std=c++11 ./src/callm_db_val.cpp -o ./bin/callm_db_val -O3 -lpthread
 
 iso_gt_mtar: ./src/callm_db_val.cpp Makefile
-	g++ -std=c++11 ./src/iso_gt_mtar.cpp -o ./bin/iso_gt_mtar -O3 -lpthread
+	$(CXX) -std=c++11 ./src/iso_gt_mtar.cpp -o ./bin/iso_gt_mtar -O3 -lpthread
 
 clean:
 	rm ./bin/callm_db_build ./bin/callm_db_val ./bin/iso_gt_mtar
