diff --git a/CMakeLists.txt b/CMakeLists.txt
index 75f5632..bd3e927 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -48,7 +48,7 @@ ExternalProject_Add(utils
     DOWNLOAD_DIR ${LIB_DIR}
     SOURCE_DIR ${LIB_DIR}/utils
     CONFIGURE_COMMAND ""
-    BUILD_COMMAND make libutils.a
+    BUILD_COMMAND make libutils.a CC=$(CC)
     INSTALL_COMMAND ""
     BUILD_IN_SOURCE 1
     )
@@ -60,7 +60,7 @@ ExternalProject_Add(sgsl
     DOWNLOAD_DIR ${LIB_DIR}
     SOURCE_DIR ${LIB_DIR}/sgsl
     CONFIGURE_COMMAND ""
-    BUILD_COMMAND make libgsl.a
+    BUILD_COMMAND make libgsl.a CC=$(CC)
     INSTALL_COMMAND ""
     BUILD_IN_SOURCE 1
     )
@@ -84,7 +84,7 @@ if(APPLE)
             mv -f ${htslib_base} htslib
         SOURCE_DIR ${LIB_DIR}/htslib
         INSTALL_DIR ${LIB_DIR}/install
-        CONFIGURE_COMMAND ${LIB_DIR}/htslib/configure --prefix=<INSTALL_DIR> --disable-bz2 --disable-lzma --disable-libcurl
+        CONFIGURE_COMMAND ${LIB_DIR}/htslib/configure --prefix=<INSTALL_DIR> CC=$(CC) --disable-bz2 --disable-lzma --disable-libcurl
         BUILD_COMMAND make lib-static
         INSTALL_COMMAND ""
         BUILD_IN_SOURCE 1
@@ -100,7 +100,7 @@ else()
             mv -f ${htslib_base} htslib
         SOURCE_DIR ${LIB_DIR}/htslib
         INSTALL_DIR ${LIB_DIR}/install
-        CONFIGURE_COMMAND ${LIB_DIR}/htslib/configure --prefix=<INSTALL_DIR> --disable-bz2 --disable-lzma --disable-libcurl
+        CONFIGURE_COMMAND ${LIB_DIR}/htslib/configure --prefix=<INSTALL_DIR> CC=$(CC) --disable-bz2 --disable-lzma --disable-libcurl
         BUILD_COMMAND make lib-static
         INSTALL_COMMAND ""
         BUILD_IN_SOURCE 1
