diff --git EDeN/Makefile EDeN/Makefile
index 5ecb626..3cf307a 100644
--- EDeN/Makefile
+++ EDeN/Makefile
@@ -22,12 +22,12 @@

 #CHECKLIMITS activates the limit check for graph operations

-CXX=g++
-C=gcc
+#CXX=g++
+#C=gcc
 OPTS=-g -O3 -Wno-deprecated -DLOSS=1 -fopenmp #-static  #-DDEBUGON -DCHECKLIMITS
 #OPTS=-g -Wno-deprecated -DLOSS=1 -DDEBUGON
 #OPTS=-g -Wno-deprecated -static -pg
-CXXFLAGS=${OPTS} -Wall -mtune=native -march=native
+CXXFLAGS=${OPTS} -Wall -mtune=native -march=native -I$(PREFIX)/include -L$(PREFIX)/lib
 #OBABEL=-I /usr/local/openbabel/2.3.1/include/openbabel-2.0  -L /usr/local/openbabel/2.3.1/lib -DUSEOBABEL
 #OBABEL=-I /usr/include/openbabel-2.0 -DUSEOBABEL
 #CXXFLAGS= ${OPTS}
