--- libxsd-frontend/xsd-frontend/semantic-graph/elements.cxx	2025-03-13 10:25:32.537967205 +0000
+++ libxsd-frontend/xsd-frontend/semantic-graph/elements.cxx	2025-03-13 10:25:44.998239553 +0000
@@ -3,7 +3,7 @@
 // license   : GNU GPL v2 + exceptions; see accompanying LICENSE file
 
 #include <algorithm>
-
+#include <ostream>
 #include <cutl/compiler/type-info.hxx>
 
 #include <xsd-frontend/semantic-graph/elements.hxx>
