--- a/src/Snakefiles/functions.sm
+++ b/src/Snakefiles/functions.sm
@@ -148,7 +148,6 @@ def getConsensusMemoryRequest(attempt, jobid):    #  Figure out what partitionin
                      mem = max(mem, 10.0 + 1.10 * float(words[4]))
                   else:
                      mem = max(mem, 5 + 0.8 * float(words[4]))
-                  sys.stderr.write("Processing memory request for consensus and the bam is %s and so request %s from %s\n"%(withBAM, mem, words[4])) 

    return int(math.ceil(math.ceil(mem)*scl))

