diff --git a/src/verkko.sh b/src/verkko.sh
index 6e1d8f8..cd417b5 100755
--- a/src/verkko.sh
+++ b/src/verkko.sh
@@ -881,6 +881,17 @@ if [ "x$cnspaths" != "x" ] ; then
     #  is done - instead of symlinking - to prevent Snakemake from
     #  'accidentally' obliterating precious original files.
 
+    if [ ! -e "${outd}/5-untip" ]; then
+       mkdir ${outd}/5-untip
+       cp -p ${cnsassembly}/5-untip/unitig-popped-unitig-normal-connected-tip.hifi-coverage.csv     ${outd}/5-untip/unitig-popped-unitig-normal-connected-tip.hifi-coverage.csv
+       cp -p ${cnsassembly}/5-untip/unitig-popped-unitig-normal-connected-tip.ont-coverage.csv      ${outd}/5-untip/unitig-popped-unitig-normal-connected-tip.ont-coverage.csv
+       cp -p ${cnsassembly}/5-untip/unitig-popped-unitig-normal-connected-tip.gfa                   ${outd}/5-untip/unitig-popped-unitig-normal-connected-tip.gfa
+       cp -p ${cnsassembly}/5-untip/unitig-popped-unitig-normal-connected-tip.ont-coverage.csv      ${outd}/5-untip/unitig-popped-unitig-normal-connected-tip.ont-coverage.csv
+       cp -p ${cnsassembly}/5-untip/combined-edges-final.gfa                                        ${outd}/5-untip/combined-edges-final.gfa
+       cp -p ${cnsassembly}/5-untip/combined-nodemap-final.txt                                      ${outd}/5-untip/combined-nodemap-final.txt
+       cp -p ${cnsassembly}/5-untip/nodelens-final.txt                                              ${outd}/5-untip/nodelens-final.txt
+    fi
+
     if [ ! -e "${outd}/6-layoutContigs" ] ; then
         mkdir ${outd}/6-layoutContigs
         cp -p ${cnsassembly}/6-layoutContigs/combined-nodemap.txt     ${outd}/6-layoutContigs/combined-nodemap.txt
