diff --git a/src/Makefile b/src/Makefile
index 5f33ff4..25c2285 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -693,7 +693,7 @@ $(eval $(call INCLUDE_SUBMAKEFILE,main.mk))
 # Perform post-processing on global variables as needed.
 DEFS := $(addprefix -D,${DEFS})
 INCDIRS := $(addprefix -I,$(call CANONICAL_PATH,${INCDIRS}))
-VERSION := verkko release v1.0 beta2
+VERSION := bioconda $(PKG_NAME) bioconda $(PKG_VERSION) beta2
 
 # Define the "all" target (which simply builds all user-defined targets) as the
 # default goal.
