diff --git a/src/Makefile b/src/Makefile
index 5f33ff4..25c2285 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -301,7 +301,7 @@
         $${OBJS}: SRC_CFLAGS   := $${$${TGT}_CFLAGS} $${SRC_CFLAGS}
         $${OBJS}: SRC_CXXFLAGS := $${$${TGT}_CXXFLAGS} $${SRC_CXXFLAGS}
         $${OBJS}: SRC_DEFS     := $$(addprefix -D,$${$${TGT}_DEFS} $${SRC_DEFS})
-        $${OBJS}: SRC_INCDIRS  := $$(addprefix -I,\
+        $${OBJS}: SRC_INCDIRS  := $$(addprefix -iquote,\
                                      $${$${TGT}_INCDIRS} $${SRC_INCDIRS})
     endif

@@ -389,7 +389,7 @@
 ALL_TGTS :=
 DEFS :=
 DIR_STACK :=
-INCDIRS :=
+INCDIRS := utgcns/libboost
 TGT_STACK :=

 # Discover our OS and architecture.  These were previously used to set
@@ -703,7 +703,7 @@

 # Perform post-processing on global variables as needed.
 DEFS := $(addprefix -D,${DEFS})
-INCDIRS := $(addprefix -I,$(call CANONICAL_PATH,${INCDIRS}))
-VERSION := verkko release v1.4.1
+INCDIRS := $(addprefix -isystem,$(call CANONICAL_PATH,${INCDIRS}))
+VERSION := bioconda $(PKG_NAME) bioconda $(PKG_VERSION)

 # Define the "all" target (which simply builds all user-defined targets) as the
 # default goal.
