--- bin/unitem.bak	2018-04-16 16:21:45.000000000 +1000
+++ bin/unitem	2020-10-20 11:49:23.000000000 +1000
@@ -1,4 +1,4 @@
-#!/srv/sw/python/2.7.4/bin/python
+#!/usr/bin/env python
 ###############################################################################
 #                                                                             #
 #    This program is free software: you can redistribute it and/or modify     #
