From 83c26834a2620c71f8b3acaf7d97a9d184099c5c Mon Sep 17 00:00:00 2001
From: Matthias De Smet <11850640+matthdsm@users.noreply.github.com>
Date: Tue, 13 Jul 2021 11:23:10 +0200
Subject: [PATCH] Add rscript shebang

---
 PREFACE.R | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/PREFACE.R b/PREFACE.R
index 68c09bd..b0a5cda 100755
--- a/PREFACE.R
+++ b/PREFACE.R
@@ -1,3 +1,5 @@
+#!/usr/bin/env Rscript
+
 version = 'v0.1.2'
 
 # ---
@@ -526,4 +528,4 @@ if (!(args[1] %in% c('train', 'predict'))){
 }
 if (args[1] == 'train') train(args[-1])
 
-if (args[1] == 'predict') predict(args[-1])
\ No newline at end of file
+if (args[1] == 'predict') predict(args[-1])
-- 
2.31.1
