--- MitoSAlt1.1.1.pl.old	2022-02-17 10:01:08.489671417 -0600
+++ MitoSAlt1.1.1.pl	2022-02-17 10:05:46.847662525 -0600
@@ -1,9 +1,25 @@
-#!/usr/bin/perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 use strict;
 use warnings;
 use List::Util qw(first min max sum);
 use Data::Dumper;
 
+#REMOVE EXISTING AND CREATE NEW DIRECTORIES
+rmdir("log");
+rmdir("indel");
+rmdir("bam");
+rmdir("tab");
+rmdir("bw");
+rmdir("plot");
+mkdir("log");
+mkdir("indel");
+mkdir("bam");
+mkdir("tab");
+mkdir("bw");
+mkdir("plot");
+
+my $condaPrefix = $ENV{'CONDA_ENV_BIN'};
+
 #INPUT 
 my $config_file = $ARGV[0];
 my $p1 = $ARGV[1];
@@ -200,7 +216,7 @@
 
   #PLOT DELETIONS
   print scalar(localtime).":Plot deletions/duplications\n";
-  system("R CMD BATCH --no-save --no-restore \'--args $msize $orihs $orihe $orils $orile $sizelimit indel/$tag.cluster indel/$tag.breakpoint $tag $hplimit $MT_fasta $flank\' delplot.R");
+  system("R CMD BATCH --no-save --no-restore \'--args $msize $orihs $orihe $orils $orile $sizelimit indel/$tag.cluster indel/$tag.breakpoint $tag $hplimit $MT_fasta $flank\' $condaPrefix/delplot.R");
 }
 
 if($rmtmp eq 'yes'){
--- MitoSAlt_SE1.1.1.pl.old	2022-02-17 10:07:46.864089868 -0600
+++ MitoSAlt_SE1.1.1.pl	2022-02-17 10:09:04.277364333 -0600
@@ -1,9 +1,25 @@
-#!/usr/bin/perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 use strict;
 use warnings;
 use List::Util qw(first min max sum);
 use Data::Dumper;
 
+#REMOVE EXISTING AND CREATE NEW DIRECTORIES
+rmdir("log");
+rmdir("indel");
+rmdir("bam");
+rmdir("tab");
+rmdir("bw");
+rmdir("plot");
+mkdir("log");
+mkdir("indel");
+mkdir("bam");
+mkdir("tab");
+mkdir("bw");
+mkdir("plot");
+
+my $condaPrefix = $ENV{'CONDA_ENV_BIN'};
+
 #INPUT 
 my $config_file = $ARGV[0];
 my $p1 = $ARGV[1];
@@ -196,7 +212,7 @@
 
   #PLOT DELETIONS
   print scalar(localtime).":Plot deletions/duplications\n";
-  system("R CMD BATCH --no-save --no-restore \'--args $msize $orihs $orihe $orils $orile $sizelimit indel/$tag.cluster indel/$tag.breakpoint $tag $hplimit $MT_fasta $flank\' delplot.R");
+  system("R CMD BATCH --no-save --no-restore \'--args $msize $orihs $orihe $orils $orile $sizelimit indel/$tag.cluster indel/$tag.breakpoint $tag $hplimit $MT_fasta $flank\' $condaPrefix/delplot.R");
 }
 
 if($rmtmp eq 'yes'){
