--- MitoSAlt1.1.1.pl	2021-05-24 17:12:54.000000000 +0000
+++ MitoSAlt1.1.1_modified.pl	2025-02-23 13:54:22.238780500 +0000
@@ -4,6 +4,16 @@
 use List::Util qw(first min max sum);
 use Data::Dumper;
 
+#CREATE NEW DIRECTORIES
+mkdir("log");
+mkdir("indel");
+mkdir("bam");
+mkdir("tab");
+mkdir("bw");
+mkdir("plot");
+
+my $condaPrefix = $ENV{'CONDA_ENV_BIN'};
+
 #INPUT 
 my $config_file = $ARGV[0];
 my $p1 = $ARGV[1];
@@ -56,12 +66,12 @@
 my $bg2bw = $User_Preferences->{bg2bw};
 
 #DATABASES
-my $hsindex = $User_Preferences->{hsindex};
-my $faindex = $User_Preferences->{faindex};
-my $lastindex = $User_Preferences->{lastindex};
-my $mtfaindex = $User_Preferences->{mtfaindex};
-my $gsize = $User_Preferences->{gsize};
-my $MT_fasta = $User_Preferences->{MT_fasta};
+my $hsindex = $ENV{MITOSALT_DATA} . "/" . $User_Preferences->{hsindex};
+my $faindex = $ENV{MITOSALT_DATA} . "/" . $User_Preferences->{faindex};
+my $lastindex = $ENV{MITOSALT_DATA} . "/" . $User_Preferences->{lastindex};
+my $mtfaindex = $ENV{MITOSALT_DATA} . "/" . $User_Preferences->{mtfaindex};
+my $gsize = $ENV{MITOSALT_DATA} . "/" . $User_Preferences->{gsize};
+my $MT_fasta = $ENV{MITOSALT_DATA} . "/" . $User_Preferences->{MT_fasta};
 
 #COMPUTATION
 my $threads = $User_Preferences->{threads};
@@ -200,7 +210,7 @@
 
   #PLOT DELETIONS
   print scalar(localtime).":Plot deletions/duplications\n";
-  system("R CMD BATCH --no-save --no-restore \'--args $msize $orihs $orihe $orils $orile $sizelimit indel/$tag.cluster indel/$tag.breakpoint $tag $hplimit $MT_fasta $flank\' delplot.R");
+  system("R CMD BATCH --no-save --no-restore \'--args $msize $orihs $orihe $orils $orile $sizelimit indel/$tag.cluster indel/$tag.breakpoint $tag $hplimit $MT_fasta $flank\' $condaPrefix/delplot.R");
 }
 
 if($rmtmp eq 'yes'){
@@ -1033,5 +1043,3 @@
   }
   close(CF);
 }
-
-
--- MitoSAlt_SE1.1.1.pl	2021-05-24 17:12:54.000000000 +0000
+++ MitoSAlt_SE1.1.1_modified.pl	2025-02-23 13:44:43.464494300 +0000
@@ -4,6 +4,16 @@
 use List::Util qw(first min max sum);
 use Data::Dumper;
 
+#CREATE NEW DIRECTORIES
+mkdir("log");
+mkdir("indel");
+mkdir("bam");
+mkdir("tab");
+mkdir("bw");
+mkdir("plot");
+
+my $condaPrefix = $ENV{'CONDA_ENV_BIN'};
+
 #INPUT 
 my $config_file = $ARGV[0];
 my $p1 = $ARGV[1];
@@ -52,12 +62,12 @@
 my $bg2bw = $User_Preferences->{bg2bw};
 
 #DATABASES
-my $hsindex = $User_Preferences->{hsindex};
-my $faindex = $User_Preferences->{faindex};
-my $lastindex = $User_Preferences->{lastindex};
-my $mtfaindex = $User_Preferences->{mtfaindex};
-my $gsize = $User_Preferences->{gsize};
-my $MT_fasta = $User_Preferences->{MT_fasta};
+my $hsindex = $ENV{MITOSALT_DATA} . "/" . $User_Preferences->{hsindex};
+my $faindex = $ENV{MITOSALT_DATA} . "/" . $User_Preferences->{faindex};
+my $lastindex = $ENV{MITOSALT_DATA} . "/" . $User_Preferences->{lastindex};
+my $mtfaindex = $ENV{MITOSALT_DATA} . "/" . $User_Preferences->{mtfaindex};
+my $gsize = $ENV{MITOSALT_DATA} . "/" . $User_Preferences->{gsize};
+my $MT_fasta = $ENV{MITOSALT_DATA} . "/" . $User_Preferences->{MT_fasta};
 
 #COMPUTATION
 my $threads = $User_Preferences->{threads};
@@ -196,7 +206,7 @@
 
   #PLOT DELETIONS
   print scalar(localtime).":Plot deletions/duplications\n";
-  system("R CMD BATCH --no-save --no-restore \'--args $msize $orihs $orihe $orils $orile $sizelimit indel/$tag.cluster indel/$tag.breakpoint $tag $hplimit $MT_fasta $flank\' delplot.R");
+  system("R CMD BATCH --no-save --no-restore \'--args $msize $orihs $orihe $orils $orile $sizelimit indel/$tag.cluster indel/$tag.breakpoint $tag $hplimit $MT_fasta $flank\' $condaPrefix/delplot.R");
 }
 
 if($rmtmp eq 'yes'){
