--- src/common/cdalignproc_base.cpp
+++ src/common/cdalignproc_base.cpp
@@ -100,8 +100,6 @@ SeqLen_t CCdAlignProcessor::SortReadingFrames(vector<size_t> rfIndice[READINGFRA
                        //rfIndices[rfidx].emplace_back(idx);
                        rfIndice[rSrc.m_vecAlignments[idx].GetRFIdx(qlen)].emplace_back(idx);
                }
-       else    //should not happen
-               THROW_SIMPLE("Cannot sort reading frames - query sequence length unknown");
        return qlen;
 }
