--- src/common/cdalignproc_base.cpp
+++ src/common/cdalignproc_base.cpp
@@ -101,7 +101,7 @@ SeqLen_t CCdAlignProcessor::SortReadingFrames(vector<size_t> rfIndice[READINGFRA
                        rfIndice[rSrc.m_vecAlignments[idx].GetRFIdx(qlen)].emplace_back(idx);
                }
        else    //should not happen
-               THROW_SIMPLE("Cannot sort reading frames - query sequence length unknown");
+               THROW_SIMPLE("Cannot sort reading frames: query sequence length unknown");
        return qlen;
 }
