--- c++/src/objtools/blast/seqdb_reader/seqdbimpl.cpp.old   2016-10-12 07:27:34.000000000 -0700
+++ c++/src/objtools/blast/seqdb_reader/seqdbimpl.cpp    2017-03-29 11:14:01.852864789 -0700
@@ -1813,6 +1813,9 @@ void CSeqDBImpl::HashToOids(unsigned has

 CSeqDBIdSet CSeqDBImpl::GetIdSet()
 {
+    // RMH: Added to eliminate race condition for vector garbage collection
+    CFastMutexGuard guard(m_MutexForIdSet);
+
     if (m_IdSet.Blank()) {
         if (! m_UserGiList.Empty()) {
             // Note: this returns a 'blank' IdSet list for positive
--- c++/src/objtools/blast/seqdb_reader/seqdbimpl.hpp.old    2016-02-03 07:19:38.000000000 -0800
+++ c++/src/objtools/blast/seqdb_reader/seqdbimpl.hpp    2017-03-29 11:12:33.521480161 -0700
@@ -1133,6 +1133,9 @@ public:
 private:
     CLASS_MARKER_FIELD("IMPL")

+    // RMH
+    CFastMutex m_MutexForIdSet;
+
     /// Adjust string length to offset of first embedded NUL byte.
     ///
     /// This is a work-around for bad data in the database; probably
