--- c++/src/algo/blast/api/blast_usage_report.cpp	2021-04-01 21:34:46.000000000 +0200
+++ c++/src/algo/blast/api/blast_usage_report.cpp	2021-04-01 21:34:46.000000000 +0200
@@ -31,6 +31,7 @@
  *  BLAST usage report api
  */
 
+#include <iostream>
 #include <ncbi_pch.hpp>
 #include <algo/blast/api/blast_usage_report.hpp>
 #include <corelib/ncbifile.hpp>
@@ -85,11 +86,13 @@
 
 CBlastUsageReport::~CBlastUsageReport()
 {
+	/*
 	if (IsEnabled()) {
 		Send(m_Params);
 		Wait();
 		Finish();
 	}
+	*/
 }
 
 string CBlastUsageReport::x_EUsageParmsToString(EUsageParams p)
@@ -159,30 +162,41 @@
 
 void CBlastUsageReport::AddParam(EUsageParams p, int val)
 {
+	/*
 	if (IsEnabled()){
 		string t = x_EUsageParmsToString(p);
 		m_Params.Add(t, NStr::IntToString(val));
 	}
+	*/
 }
 
 void CBlastUsageReport::AddParam(EUsageParams p, const string & val)
 {
+	/*
 	if (IsEnabled()) {
 		string t = x_EUsageParmsToString(p);
 		m_Params.Add(t, val);
 	}
+	*/
 }
 
 void CBlastUsageReport::AddParam(EUsageParams p, const double & val)
 {
+	/*
 	if (IsEnabled()) {
 		string t = x_EUsageParmsToString(p);
 		m_Params.Add(t, val);
 	}
+	*/
 }
 
 void CBlastUsageReport::x_CheckBlastUsageEnv()
 {
+	SetEnabled(false);
+	CUsageReportAPI::SetEnabled(false);
+	LOG_POST(Info <<"Phone home disabled by config setting");
+	return ;
+	/*
 	char * blast_usage_env = getenv("BLAST_USAGE_REPORT");
 	if(blast_usage_env != NULL){
 		bool enable = NStr::StringToBool(blast_usage_env);
@@ -208,22 +222,27 @@
 	CUsageReportAPI::SetEnabled(true);
 	SetEnabled(true);
 	LOG_POST(Info <<"Phone home enabled");
+	*/
 }
 
 void CBlastUsageReport::AddParam(EUsageParams p, Int8 val)
 {
+	/*
 	if (IsEnabled()) {
 		string t = x_EUsageParmsToString(p);
 		m_Params.Add(t, val);
 	}
+	*/
 
 }
 
 void CBlastUsageReport::AddParam(EUsageParams p, bool val)
 {
+	/*
 	if (IsEnabled()) {
 		string t = x_EUsageParmsToString(p);
 		m_Params.Add(t, val);
 	}
 
+	*/
 }
