--- c++/src/build-system/configure	2023-08-08 20:52:02
+++ c++/src/build-system/configure	2025-03-28 08:54:58
@@ -27743,7 +27743,7 @@
          case "$host_os:$host_cpu:$bit64_sfx" in
             darwin*:*86*:64 ) vdb_libpath=mac/$vdb_mode/x86_64/lib   ;;
             darwin*:*86*:* )  vdb_libpath=mac/$vdb_mode/fat86/lib    ;;
-            darwin*:arm*   )  vdb_libpath=mac/$vdb_mode/arm64/lib    ;;
+            darwin*:a*   )  vdb_libpath=mac/$vdb_mode/arm64/lib      ;;
             linux*:*86*:64 )  vdb_libpath=linux/$vdb_mode/x86_64/lib ;;
             linux*:aarch64* ) vdb_libpath=linux/$vdb_mode/arm64/lib  ;;
             * )
