Subject: [PATCH] turned off asmlib for mac to avoid build problems
---
Index: mbgccoder/MBGC_Decoder.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/mbgccoder/MBGC_Decoder.cpp b/mbgccoder/MBGC_Decoder.cpp
--- a/mbgccoder/MBGC_Decoder.cpp	(revision 4d9edd3397b67ca0d077fe50beb966154ffa9160)
+++ b/mbgccoder/MBGC_Decoder.cpp	(date 1776535778636)
@@ -1,7 +1,7 @@
 #include "MBGC_Decoder.h"
 
 #include "../coders/CodersLib.h"
-#if !defined(__arm__) && !defined(__aarch64__) && !defined(__ARM_ARCH)
+#if false
 #include "../libs/asmlib.h"
 #endif
 #include "../matching/SimpleSequenceMatcher.h"
Index: matching/MultipleGenomeMatchingProcessor.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/matching/MultipleGenomeMatchingProcessor.cpp b/matching/MultipleGenomeMatchingProcessor.cpp
--- a/matching/MultipleGenomeMatchingProcessor.cpp	(revision 4d9edd3397b67ca0d077fe50beb966154ffa9160)
+++ b/matching/MultipleGenomeMatchingProcessor.cpp	(date 1776535751951)
@@ -33,7 +33,7 @@
     }
 }
 
-#if !defined(__arm__) && !defined(__aarch64__) && !defined(__ARM_ARCH)
+#if false
 #include "../libs/asmlib.h"
 #endif
 #include <omp.h>
Index: utils/helper.h
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/utils/helper.h b/utils/helper.h
--- a/utils/helper.h	(revision 4d9edd3397b67ca0d077fe50beb966154ffa9160)
+++ b/utils/helper.h	(date 1776535822451)
@@ -340,7 +340,7 @@
 
 }
 
-#if defined(__arm__) || defined(__aarch64__) || defined(__ARM_ARCH)
+#if true
 
 void A_memcpy(void *dest, const void *src, size_t n);
 
Index: matching/SlidingWindowSparseEMMatcher.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/matching/SlidingWindowSparseEMMatcher.cpp b/matching/SlidingWindowSparseEMMatcher.cpp
--- a/matching/SlidingWindowSparseEMMatcher.cpp	(revision 4d9edd3397b67ca0d077fe50beb966154ffa9160)
+++ b/matching/SlidingWindowSparseEMMatcher.cpp	(date 1776535763293)
@@ -28,7 +28,7 @@
 * ================================================================= */
 
 #include "SlidingWindowSparseEMMatcher.h"
-#if !defined(__arm__) && !defined(__aarch64__) && !defined(__ARM_ARCH)
+#if false
 #include "../libs/asmlib.h"
 #endif
 #include <algorithm>
Index: CMakeLists.txt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	(revision 4d9edd3397b67ca0d077fe50beb966154ffa9160)
+++ b/CMakeLists.txt	(date 1776535865465)
@@ -99,17 +99,6 @@
     set(CMAKE_BUILD_TYPE Release)
 endif()
 
-if(CMAKE_SYSTEM_PROCESSOR MATCHES "(x86_64|AMD64)")
-    if(CYGWIN OR MINGW)
-        set(ASM_LIB ${CMAKE_SOURCE_DIR}/libs/libacof64.lib)
-    elseif(UNIX AND NOT APPLE)
-        set(ASM_LIB ${CMAKE_SOURCE_DIR}/libs/libaelf64.a)
-    elseif(APPLE)
-        set(ASM_LIB ${CMAKE_SOURCE_DIR}/libs/libamac64.a)
-    endif()
-    link_libraries(${ASM_LIB})
-endif()
-
 link_libraries(pthread)
 
 if(CYGWIN OR MINGW)
Index: utils/helper.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/utils/helper.cpp b/utils/helper.cpp
--- a/utils/helper.cpp	(revision 4d9edd3397b67ca0d077fe50beb966154ffa9160)
+++ b/utils/helper.cpp	(date 1776535837254)
@@ -595,7 +595,7 @@
 #endif
 }
 
-#if defined(__arm__) || defined(__aarch64__) || defined(__ARM_ARCH)
+#if true
 
 void A_memcpy(void *dest, const void *src, size_t n) {
     memcpy(dest, src, n);
