From 65a77e1b074de96e95e1a22a5336800a90aafcd4 Mon Sep 17 00:00:00 2001
From: Elmar Pruesse <elmar.pruesse@ucdenver.edu>
Date: Wed, 15 May 2019 17:20:36 -0600
Subject: [PATCH 3/3] Remove vendoring of zlib and htslib

---
 CMakeLists.txt     | 5 +++--
 src/CMakeLists.txt | 4 ++--
 2 files changed, 5 insertions(+), 4 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 4491ef1..1720c10 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -4,8 +4,8 @@ project (MetaBAT)
 message("Installing into ${CMAKE_INSTALL_PREFIX}")
 
 include(${CMAKE_ROOT}/Modules/ExternalProject.cmake)
-include(cmake/zlib.cmake)
-include(cmake/htslib.cmake)
+find_library(ZLIB z)
+find_library(HTSLIB hts)
 
 set(CMAKE_CXX_STANDARD 14)
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
@@ -16,6 +16,7 @@ set(CMAKE_C_STANDARD_REQUIRED ON)
 set(CMAKE_C_EXTENSIONS OFF)
 
 add_definitions(-D_XOPEN_SOURCE=700)
+add_definitions(-D_DARWIN_C_SOURCE)
 
 set(PRE_CONFIGURE_FILE "metabat_version.h.in")
 set(POST_CONFIGURE_FILE "metabat_version.h")
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 62555d2..420b165 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -14,8 +14,8 @@ set(targets jgi_summarize_bam_contig_depths contigOverlaps metabat1 metabat2)
 foreach(target ${targets})
 
    add_executable(${target} ${target}.cpp)
-   add_dependencies(${target} htslib zlib check_git_repository)
-   target_link_libraries(${target} ${htslib_LIB} ${zlib_LIB} ${Boost_LIBRARIES}) 
+   add_dependencies(${target} check_git_repository)
+   target_link_libraries(${target} ${ZLIB} ${HTSLIB} ${Boost_LIBRARIES})
 
 endforeach()
 
-- 
2.14.0

