diff --git a/cmake/Modules/GetGitVersion.cmake b/cmake/Modules/GetGitVersion.cmake
index c600b97..a67068f 100755
--- a/cmake/Modules/GetGitVersion.cmake
+++ b/cmake/Modules/GetGitVersion.cmake
@@ -29,8 +29,8 @@ function(GET_GIT_VERSION)
 set(GIT_PREFIX ${ARGV})
 
 # get the base version string from the top level file in the source tree: "VERSION"
-set(${GIT_PREFIX}_BASE_VERSION_FILE "${CMAKE_CURRENT_LIST_DIR}/VERSION") # VER.SUBVER.PATCH - manual change by developer
-set(${GIT_PREFIX}_BASE_VERSION_FILE "${CMAKE_CURRENT_LIST_DIR}/VERSION" PARENT_SCOPE) # VER.SUBVER.PATCH - manual change by developer
+set(${GIT_PREFIX}_BASE_VERSION_FILE "${CMAKE_CURRENT_LIST_DIR}/VERSION.txt") # VER.SUBVER.PATCH - manual change by developer
+set(${GIT_PREFIX}_BASE_VERSION_FILE "${CMAKE_CURRENT_LIST_DIR}/VERSION.txt" PARENT_SCOPE) # VER.SUBVER.PATCH - manual change by developer
 file(READ "${${GIT_PREFIX}_BASE_VERSION_FILE}" ${GIT_PREFIX}_FULL_VERSION_STRING LIMIT_COUNT 1)
 string(REGEX REPLACE "(\n|\r)" "" ${GIT_PREFIX}_FULL_VERSION_STRING "${${GIT_PREFIX}_FULL_VERSION_STRING}")
 string(REGEX REPLACE "[A-Za-z].*" "" ${GIT_PREFIX}_VERSION_STRING "${${GIT_PREFIX}_FULL_VERSION_STRING}")
