--- src/CMakeLists.txt	2022-11-19 11:07:09.838826843 +0100
+++ src/CMakeLists.txt	2022-11-19 11:07:09.838826843 +0100
@@ -1,6 +1,6 @@
 # find and use Boost
 set(Boost_NO_BOOST_CMAKE ON) # for backwards comptibility of cmake and boost >= 1.70
-set(Boost_USE_STATIC_LIBS   ON)
+set(Boost_USE_STATIC_LIBS   OFF)
 find_package(Boost 1.55.0 COMPONENTS program_options filesystem system graph serialization iostreams REQUIRED)
 include_directories( ${Boost_INCLUDE_DIRS} )
 
