--- a/cmake/htslib.cmake	2024-06-20 08:08:51.668000000 +0000
+++ b.cmake/htslib.cmake	2024-06-20 08:21:07.352000000 +0000
@@ -14,7 +14,8 @@
 ExternalProject_Add(htslib
     PREFIX ${htslib_PREFIX}
     GIT_REPOSITORY "https://github.com/samtools/htslib.git"
-    GIT_TAG "1.17"
+    GIT_TAG "1.20"
+    PATCH_COMMAND patch -p1 < $ENV{RECIPE_DIR}/arm_hwcap.patch
     UPDATE_COMMAND ""
     BUILD_IN_SOURCE 1
     #CONFIGURE_COMMAND "${CMAKE_CURRENT_SOURCE_DIR}/contrib/htslib-prefix/src/htslib/configure"
