diff --git a/cmake/htslib.cmake b/cmake/htslib.cmake
index 36c9cca..9396f60 100644
--- a/cmake/htslib.cmake
+++ b/cmake/htslib.cmake
@@ -27,6 +27,7 @@ ExternalProject_Add(htslib
     PREFIX ${htslib_PREFIX}
     GIT_REPOSITORY "https://github.com/samtools/htslib.git"
     GIT_TAG "1.20"
+    PATCH_COMMAND patch -p1 < $ENV{RECIPE_DIR}/arm_hwcap.patch
     UPDATE_COMMAND ""
     BUILD_IN_SOURCE 1
     #CONFIGURE_COMMAND "${CMAKE_CURRENT_SOURCE_DIR}/contrib/htslib-prefix/src/htslib/configure"
