diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 3cf10bc..9b8a50a 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -1,8 +1,12 @@
 # find and use Boost
+set(Boost_NO_SYSTEM_PATHS ON)
+set(Boost_NO_WARN_NEW_VERSION ON)
 set(Boost_NO_BOOST_CMAKE ON) # for backwards comptibility of cmake and boost >= 1.70
-set(Boost_USE_STATIC_LIBS   ON)
-find_package(Boost 1.55.0 COMPONENTS program_options filesystem system graph serialization iostreams REQUIRED)
+set(Boost_USE_STATIC_LIBS OFF)
+set(Boost_USE_MULTITHREADED TRUE)
+find_package(Boost 1.71.0 COMPONENTS program_options filesystem system graph serialization iostreams REQUIRED)
 include_directories( ${Boost_INCLUDE_DIRS} )
+LINK_DIRECTORIES( ${Boost_LIBRARY_DIRS} )
 
 # find and use OpenMP
 find_package(OpenMP)
