diff --git a/src/Makefile b/src/Makefile
index 6a63a87..82e70c1 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -1,8 +1,8 @@
-CC = gcc
-CFLAGS = -g -I./
-CFLAGS += -Wall -Werror -Wno-unused-function -Wno-unused-parameter -Wcast-align
-CFLAGS += -Wshadow -Wpointer-arith -Wwrite-strings -Wunreachable-code -pedantic
-LDFLAGS = -lz -lpthread
+CC := $(CC)
+CFLAGS = -g -I$(PREFIX)/include -I./ -pthread
+CFLAGS += -Wall -Werror -Wno-unused-function -Wno-unused-parameter -Wcast-align -Wno-format -Wno-deprecated-declarations -Wno-newline-eof -Wno-unused-but-set-variable -Wno-implicit-function-declaration
+CFLAGS += -Wshadow -Wpointer-arith -Wwrite-strings -Wno-unreachable-code -pedantic
+LDFLAGS = -lz -lpthread -L$(PREFIX)/lib
 OBJS = hashcounter.o subcontig.o
 
 all: subcontig hashcounter
