diff --git a/js/View/TrackList/Faceted.js b/js/View/TrackList/Faceted.js
index 4d0156b..cfa9c3e 100755
--- a/js/View/TrackList/Faceted.js
+++ b/js/View/TrackList/Faceted.js
@@ -45,7 +45,9 @@ return declare( 'ComboTrackSelector.View.TrackList.Faceted', JBrowse.View.TrackL
         var title_list = titles.join("<br>");
 
         var icon_elem = '<div></div>';
-        if( this.browser.config.combotracksel_icon !== undefined )
+        if( this.config.icon !== undefined )
+            icon_elem = '</div><img width="20px" src="' + this.config.icon + '"><div>'
+        else if( this.browser.config.combotracksel_icon !== undefined )
             icon_elem = '</div><img width="20px" src="' + this.browser.config.combotracksel_icon + '"><div>'
         dojo.create('div',
             {
