diff --git a/css/main.css b/css/main.css
index e8cf905..26b9088 100755
--- a/css/main.css
+++ b/css/main.css
@@ -18,10 +18,6 @@
     height: 100%;
 }
 
-.jbrowse .header {
-    margin-bottom: 5em;
-}
-
 #faceted_tracksel.active {
     -moz-box-shadow: 4px 4px 10px 3px rgba( 30, 30, 50, 0.3 );
     -webkit-box-shadow: 4px 4px 10px 3px rgba( 30, 30, 50, 0.3 );
diff --git a/js/View/TrackList/Faceted.js b/js/View/TrackList/Faceted.js
index cfa9c3e..c9b6db3 100755
--- a/js/View/TrackList/Faceted.js
+++ b/js/View/TrackList/Faceted.js
@@ -21,6 +21,8 @@ return declare( 'ComboTrackSelector.View.TrackList.Faceted', JBrowse.View.TrackL
    {
 
     renderInitial: function() {
+        dojo.query('#hierarchicalTrackPane .header').style('margin-bottom', '5em');
+
         this.containerElem = dojo.create( 'div', {
             id: 'faceted_tracksel',
             className: 'jbrowse',
