--- cmake/Htslib.cmake
+++ cmake/Htslib.cmake
@@ -46,7 +46,8 @@
 
         unset(hts_configure_flags)
         if (hts_cflags)
-            set(hts_configure_flags "CPPFLAGS=${hts_cflags}" "LDFLAGS=${hts_cflags}")
+            # Keep conda-build's zlib search paths.
+            set(hts_configure_flags "CPPFLAGS=$ENV{CPPFLAGS} ${hts_cflags}" "LDFLAGS=$ENV{LDFLAGS} ${hts_cflags}")
         endif()
 
         # Parallelise the htslib build too, though not enough to compete with the main build.
