--- run_resfinder.py.old	2022-03-24 02:18:37.191844209 -0500
+++ run_resfinder.py	2022-03-24 02:19:40.497086660 -0500
@@ -27,24 +27,6 @@
 # #########                                 FUNCTIONS               ######### #
 # ########################################################################### #
 
-def get_git_tag():
-    git_folders_path = os.path.dirname(os.path.realpath(__file__))
-    try:
-        git_tag = str(
-            subprocess.check_output(
-                ['git', 'describe', '--tags', '--abbrev=0'],
-                stderr=subprocess.STDOUT, cwd=git_folders_path,
-            )).strip('\'b\\n')
-    except subprocess.CalledProcessError as exc_info:
-        match = re.search(
-            'fatal: no tag exactly matches \'(?P<commit>[a-z0-9]+)\'', str(exc_info.output))
-        if match:
-            raise Exception(
-                'Bailing: there is no git tag for the current commit, {commit}'.format(
-                commit=match.group('commit')))
-        raise Exception(str(exc_info.output))
-    return git_tag
-
 def eprint(*args, **kwargs):
     print(*args, file=sys.stderr, **kwargs)
 
@@ -178,7 +160,7 @@
                           ResFinder will be used.",
                     type=float,
                     default=None)
-parser.add_argument("-v", "--version", action="version", version=get_git_tag(),
+parser.add_argument("-v", "--version", action="version", version="4.1.11",
                     help="Show program's version number and exit")
 # Temporary option only available temporary
 parser.add_argument("--pickle",
