diff --git a/ra/src/EdgesSet.hpp b/ra/src/EdgesSet.hpp
index 5618f9f..f79be83 100644
--- a/ra/src/EdgesSet.hpp
+++ b/ra/src/EdgesSet.hpp
@@ -5,6 +5,7 @@
 #include <map>
 #include <vector>
 #include <utility>
+#include <cstdint>
 
 namespace Graph {
 
diff --git a/overlap2dot/src/overlap2dot.h b/overlap2dot/src/overlap2dot.h
index 1689662..c6da18e 100644
--- a/overlap2dot/src/overlap2dot.h
+++ b/overlap2dot/src/overlap2dot.h
@@ -6,6 +6,7 @@
 #include <string>
 #include <iostream>
 #include "ra/EdgesSet.hpp"
+#include <cstdint>
 
 using std::endl;
 using std::ostream;
