diff --git a/serotypefinder.py b/serotypefinder.py
index 1a756aa..f7f96c4 100755
--- a/serotypefinder.py
+++ b/serotypefinder.py
@@ -123,7 +123,7 @@ parser.add_argument("-i", "--infile", dest="infile", help="FASTA or FASTQ input
 parser.add_argument("-o", "--outputPath", dest="outdir",help="Path to blast output", default='.')
 parser.add_argument("-tmp", "--tmp_dir", help="Temporary directory for storage of the results from the external software.")
 parser.add_argument("-mp", "--methodPath", dest="method_path",help="Path to method to use (kma or blastn)")
-parser.add_argument("-p", "--databasePath", dest="db_path",help="Path to the databases", default='/database')
+parser.add_argument("-p", "--databasePath", dest="db_path",help="Path to the databases", default=os.getenv("SEROTYPEFINDER_DB"))
 parser.add_argument("-d", "--databases", dest="databases",help="Databases chosen to search in - if non is specified all is used")
 parser.add_argument("-l", "--mincov", dest="min_cov",help="Minimum coverage", default=0.60)
 parser.add_argument("-t", "--threshold", dest="threshold",help="Minimum threshold for identity", default=0.90)
