diff --git a/spine.pl b/spine.pl
index 554fdc4..950d9e7 100644
--- a/spine.pl
+++ b/spine.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 my $version = "0.3.2";
 ## Changes from v0.3.1 -> v0.3.2
@@ -301,9 +301,9 @@ my $home_dir = abs_path($0); #get the absolute path to spine.pl
 $home_dir =~ s/\/[^\/]*$//; #strip off "/spine.pl"
 unless ($web){
     print STDERR "home_dir = $home_dir\n";
-    die "ERROR: Can't find the required file \"nucmer_multi.pl\". Make sure it is in the \"scripts\" directory with spine.pl ($home_dir/scripts) and has not been renamed.\n" unless (-e "$home_dir/scripts/nucmer_multi.pl");
-    die "ERROR: Can't find the required file \"nucmer_backbone.pl\". Make sure it is in the \"scripts\" directory with spine.pl ($home_dir/scripts) and has not been renamed.\n" unless (-e "$home_dir/scripts/nucmer_backbone.pl");
-    my $nbb_vers = `perl $home_dir/scripts/nucmer_backbone.pl -V`;
+    die "ERROR: Can't find the required file \"nucmer_multi.pl\". Make sure it is in the \"scripts\" directory with spine.pl ($home_dir) and has not been renamed.\n" unless (-e "$home_dir/scripts/nucmer_multi.pl");
+    die "ERROR: Can't find the required file \"nucmer_backbone.pl\". Make sure it is in the \"scripts\" directory with spine.pl ($home_dir) and has not been renamed.\n" unless (-e "$home_dir/nucmer_backbone.pl");
+    my $nbb_vers = `perl $home_dir/nucmer_backbone.pl -V`;
     chomp $nbb_vers;
     my $min_nbb_vers = 0.4;
     my $wrong_vers = "unknown";
@@ -320,8 +320,8 @@ unless ($web){
         }
     } 
     die "ERROR: Minimum version of scripts/nucmer_backbone.pl is $min_nbb_vers (detected version is $wrong_vers)\n" if $wrong_vers;
-    print STDERR "nucmer_backbone.pl (version $nbb_vers) found: $home_dir/scripts/nucmer_backbone.pl\n";
-    my $nm_vers = `perl $home_dir/scripts/nucmer_multi.pl -V`;
+    print STDERR "nucmer_backbone.pl (version $nbb_vers) found: $home_dir/nucmer_backbone.pl\n";
+    my $nm_vers = `perl $home_dir/nucmer_multi.pl -V`;
     chomp ($nm_vers);
     my $min_nm_vers = 0.3;
     $wrong_vers = "unknown";
@@ -338,7 +338,7 @@ unless ($web){
         }
     } 
     die "ERROR: Minimum version of scripts/nucmer_multi.pl is $min_nm_vers (detected version is $wrong_vers)\n" if $wrong_vers;
-    print STDERR "nucmer_multi.pl (version $nm_vers) found: $home_dir/scripts/nucmer_multi.pl\n";
+    print STDERR "nucmer_multi.pl (version $nm_vers) found: $home_dir/nucmer_multi.pl\n";
 }
     
 #read in file of files
@@ -568,7 +568,7 @@ my $return;
     local $opt_o = $pref;
     local $opt_n = $nuc_loc;
     local $opt_w = 1 if $web;
-    $return = do "$home_dir/scripts/nucmer_multi.pl";
+    $return = do "$home_dir/nucmer_multi.pl";
 }
 unless ($return){
     die "ERROR: Couldn't run nucmer_multi.pl: $@\n" if $@;
@@ -614,7 +614,7 @@ print STDERR "out_pan = $out_pan\n" if $out_pan and !$web;
     local $opt_g = $out_opt_g;
     local $opt_w = 1 if $web;
     local $opt_z = $version;
-    $return = do "$home_dir/scripts/nucmer_backbone.pl";
+    $return = do "$home_dir/nucmer_backbone.pl";
 }
 unless ($return){
     die "ERROR: Couldn't run nucmer_backbone.pl: $@\n" if $@;
