From 529908558ea26f017990589fc4a89b6963e8494f Mon Sep 17 00:00:00 2001
From: David <david.laehnemann@hhu.de>
Date: Thu, 6 Jul 2017 17:11:55 +0200
Subject: [PATCH] Remove samtools building from setup.py.

---
 setup.py | 12 ------------
 1 file changed, 12 deletions(-)

diff --git a/setup.py b/setup.py
index afc88d1..85dcdd6 100644
--- a/setup.py
+++ b/setup.py
@@ -21,22 +21,10 @@ class MonoVarBuild(build):
 		# run original build code
 		build.run(self)
 
-		# build samtools
-		build_path = os.path.abspath(self.build_temp)
-		cmd = ['make', '-C', 'external/samtools']
-
-		def compile():
-			subprocess.check_call(cmd)
-
-		self.execute(compile, [], 'Compile samtools')
-
 class MonoVarInstall(install):
 
 	def run(self):
 		install.run(self)
-		import shutil
-		shutil.copy2('external/samtools/samtools',
-                     os.path.join(self.install_lib, 'monovar_src'))
 
 cmdclass = {'build': MonoVarBuild, 'install': MonoVarInstall,}
 
-- 
2.7.4

