diff --git Spectacle_python.py Spectacle_python.py
index 32e042f..82b395b 100644
--- Spectacle_python.py
+++ Spectacle_python.py
@@ -1,3 +1,4 @@
+#!/usr/bin/env python
 """
 Spectacle_python : An implementation of a spectral algorithm for Hidden Markov Models. The input file consists of sample (cross) moments computed from chromatin mark datasets (e.g. from ENCODE). The output file consists of Hidden Markov Model parameters (initial state distribution, emission matrix and transition matrix). This is a module inside the main Spectacle program which is written in Java and based on the ChromHMM software. Command line instructions:
 
-- 
1.9.1

