--- ./configure.in
+++ ./configure.in
@@ -24,7 +24,6 @@ dnl Check for ANSI C prototypes, the const and inline keywords,
 dnl and ANSI style stringification
 AC_GNU_SOURCE
 AC_PROG_CC
-AM_C_PROTOTYPES
 AC_PROG_CPP
 AC_C_CONST
 AC_C_INLINE
--- ./src/Makefile.am
+++ ./src/Makefile.am
@@ -1,5 +1,5 @@
 ## Process this file with automake to produce Makefile.in -*-Makefile-*-
-AUTOMAKE_OPTIONS = ansi2knr foreign 1.2h
+AUTOMAKE_OPTIONS = foreign 1.2h
 
 # in the spirit of automake ...
 pkglibexecdir = $(libexecdir)/@PACKAGE@/@VERSION_MAJOR@
