diff --git a/src/Makefile b/src/Makefile
index 9822b6d..65893cb 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -7,8 +7,8 @@
 
 
 
-CC = gcc
-COPTIM = -Wall -O3 -std=gnu99 -DHAVE_ZLIB=1 -DTRACING_ON=0
+CC ?= ${CC}
+COPTIM = -Wall -O3 -std=gnu99 -DHAVE_ZLIB=1 -DTRACING_ON=0 -I${PREFIX}/include -Wno-incompatible-pointer-types
 	#
 	# -march=native -mtune=native
 	# icc options for optimisation.  COPTIM = -xHOST -O3 -no-prec-div -static -DHAVE_ZLIB=1
