diff -u -r a/orthomclSoftware-v2.0.9/bin/orthomclAdjustFasta b/orthomclSoftware-v2.0.9/bin/orthomclAdjustFasta
--- a/orthomclSoftware-v2.0.9/bin/orthomclAdjustFasta	2013-07-18 19:53:22.000000000 +0200
+++ b/orthomclSoftware-v2.0.9/bin/orthomclAdjustFasta	2017-09-12 16:30:35.318256976 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use strict;
 
diff -u -r a/orthomclSoftware-v2.0.9/bin/orthomclBlastParser b/orthomclSoftware-v2.0.9/bin/orthomclBlastParser
--- a/orthomclSoftware-v2.0.9/bin/orthomclBlastParser	2013-07-18 19:53:22.000000000 +0200
+++ b/orthomclSoftware-v2.0.9/bin/orthomclBlastParser	2017-09-12 16:30:29.950278157 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use strict;
 
diff -u -r a/orthomclSoftware-v2.0.9/bin/orthomclDropSchema b/orthomclSoftware-v2.0.9/bin/orthomclDropSchema
--- a/orthomclSoftware-v2.0.9/bin/orthomclDropSchema	2013-07-18 19:53:22.000000000 +0200
+++ b/orthomclSoftware-v2.0.9/bin/orthomclDropSchema	2017-09-12 16:30:25.674295011 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use DBI;
 use FindBin;
diff -u -r a/orthomclSoftware-v2.0.9/bin/orthomclDumpPairsFiles b/orthomclSoftware-v2.0.9/bin/orthomclDumpPairsFiles
--- a/orthomclSoftware-v2.0.9/bin/orthomclDumpPairsFiles	2013-07-18 19:53:22.000000000 +0200
+++ b/orthomclSoftware-v2.0.9/bin/orthomclDumpPairsFiles	2017-09-12 16:30:16.186332361 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use DBI;
 use FindBin;
diff -u -r a/orthomclSoftware-v2.0.9/bin/orthomclExtractProteinIdsFromGroupsFile b/orthomclSoftware-v2.0.9/bin/orthomclExtractProteinIdsFromGroupsFile
--- a/orthomclSoftware-v2.0.9/bin/orthomclExtractProteinIdsFromGroupsFile	2013-07-18 19:53:22.000000000 +0200
+++ b/orthomclSoftware-v2.0.9/bin/orthomclExtractProteinIdsFromGroupsFile	2017-09-12 16:30:10.010356636 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 use strict;
 
 usage() unless scalar(@ARGV) == 1;
diff -u -r a/orthomclSoftware-v2.0.9/bin/orthomclExtractProteinPairsFromGroupsFile b/orthomclSoftware-v2.0.9/bin/orthomclExtractProteinPairsFromGroupsFile
--- a/orthomclSoftware-v2.0.9/bin/orthomclExtractProteinPairsFromGroupsFile	2013-07-18 19:53:22.000000000 +0200
+++ b/orthomclSoftware-v2.0.9/bin/orthomclExtractProteinPairsFromGroupsFile	2017-09-12 16:30:04.966376438 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 use strict;
 
 usage() unless scalar(@ARGV) == 1;
diff -u -r a/orthomclSoftware-v2.0.9/bin/orthomclFilterFasta b/orthomclSoftware-v2.0.9/bin/orthomclFilterFasta
--- a/orthomclSoftware-v2.0.9/bin/orthomclFilterFasta	2013-07-18 19:53:22.000000000 +0200
+++ b/orthomclSoftware-v2.0.9/bin/orthomclFilterFasta	2017-09-12 16:30:02.806384911 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use strict;
 
diff -u -r a/orthomclSoftware-v2.0.9/bin/orthomclInstallSchema b/orthomclSoftware-v2.0.9/bin/orthomclInstallSchema
--- a/orthomclSoftware-v2.0.9/bin/orthomclInstallSchema	2013-07-18 19:53:22.000000000 +0200
+++ b/orthomclSoftware-v2.0.9/bin/orthomclInstallSchema	2017-09-12 16:29:53.934419672 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use DBI;
 use FindBin;
diff -u -r a/orthomclSoftware-v2.0.9/bin/orthomclLoadBlast b/orthomclSoftware-v2.0.9/bin/orthomclLoadBlast
--- a/orthomclSoftware-v2.0.9/bin/orthomclLoadBlast	2013-07-18 19:53:22.000000000 +0200
+++ b/orthomclSoftware-v2.0.9/bin/orthomclLoadBlast	2017-09-12 16:29:46.206449897 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use DBI;
 use FindBin;
diff -u -r a/orthomclSoftware-v2.0.9/bin/orthomclMclToGroups b/orthomclSoftware-v2.0.9/bin/orthomclMclToGroups
--- a/orthomclSoftware-v2.0.9/bin/orthomclMclToGroups	2013-07-18 19:53:22.000000000 +0200
+++ b/orthomclSoftware-v2.0.9/bin/orthomclMclToGroups	2017-09-12 16:29:35.994489754 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 my ($prefix, $startId) = @ARGV;
 
diff -u -r a/orthomclSoftware-v2.0.9/bin/orthomclPairs b/orthomclSoftware-v2.0.9/bin/orthomclPairs
--- a/orthomclSoftware-v2.0.9/bin/orthomclPairs	2013-07-18 19:53:22.000000000 +0200
+++ b/orthomclSoftware-v2.0.9/bin/orthomclPairs	2017-09-12 16:29:32.466503501 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use DBI;
 use FindBin;
diff -u -r a/orthomclSoftware-v2.0.9/bin/orthomclReduceFasta b/orthomclSoftware-v2.0.9/bin/orthomclReduceFasta
--- a/orthomclSoftware-v2.0.9/bin/orthomclReduceFasta	2013-07-18 19:53:22.000000000 +0200
+++ b/orthomclSoftware-v2.0.9/bin/orthomclReduceFasta	2017-09-12 16:29:26.830525437 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use strict;
 
diff -u -r a/orthomclSoftware-v2.0.9/bin/orthomclReduceGroups b/orthomclSoftware-v2.0.9/bin/orthomclReduceGroups
--- a/orthomclSoftware-v2.0.9/bin/orthomclReduceGroups	2013-07-18 19:53:22.000000000 +0200
+++ b/orthomclSoftware-v2.0.9/bin/orthomclReduceGroups	2017-09-12 16:29:22.246543257 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use strict;
 
diff -u -r a/orthomclSoftware-v2.0.9/bin/orthomclRemoveIdenticalGroups b/orthomclSoftware-v2.0.9/bin/orthomclRemoveIdenticalGroups
--- a/orthomclSoftware-v2.0.9/bin/orthomclRemoveIdenticalGroups	2013-07-18 19:53:22.000000000 +0200
+++ b/orthomclSoftware-v2.0.9/bin/orthomclRemoveIdenticalGroups	2017-09-12 16:29:16.998563633 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use strict;
 
diff -u -r a/orthomclSoftware-v2.0.9/bin/orthomclSingletons b/orthomclSoftware-v2.0.9/bin/orthomclSingletons
--- a/orthomclSoftware-v2.0.9/bin/orthomclSingletons	2013-07-18 19:53:22.000000000 +0200
+++ b/orthomclSoftware-v2.0.9/bin/orthomclSingletons	2017-09-12 16:29:12.642580524 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use strict;
 
diff -u -r a/orthomclSoftware-v2.0.9/bin/orthomclSortGroupMembersByScore b/orthomclSoftware-v2.0.9/bin/orthomclSortGroupMembersByScore
--- a/orthomclSoftware-v2.0.9/bin/orthomclSortGroupMembersByScore	2013-07-18 19:53:22.000000000 +0200
+++ b/orthomclSoftware-v2.0.9/bin/orthomclSortGroupMembersByScore	2017-09-12 16:29:06.650603728 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use strict;
 
diff -u -r a/orthomclSoftware-v2.0.9/bin/orthomclSortGroupsFile b/orthomclSoftware-v2.0.9/bin/orthomclSortGroupsFile
--- a/orthomclSoftware-v2.0.9/bin/orthomclSortGroupsFile	2013-07-18 19:53:22.000000000 +0200
+++ b/orthomclSoftware-v2.0.9/bin/orthomclSortGroupsFile	2017-09-12 16:28:57.326639759 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use strict;
 
