diff --git a/Makefile b/Makefile
index c0bbb72..1ab8a9e 100644
--- a/Makefile
+++ b/Makefile
@@ -1,4 +1,4 @@
-BOOST = /usr/include/boost/
+BOOST = $(PREFIX)/include/boost/
 
 DIR_BIN_PROD=$(DIR)
 
@@ -7,15 +7,15 @@ OBJ = ${CLASS_FILE:.cpp=.o}
 MAIN_FILE=gmove.cpp
 BIN=gmove
 
-COMPILER=g++
+COMPILER=$(CXX)
 CPPFLAGS += -I$(BOOST) -I$(BOOST_INCDIR) -I.
-CPPFLAGS += -O3
+CPPFLAGS += -O3 -std=c++14 -march=x86-64-v3
 CPPFLAGS += -W -Wall -Wno-long-long -Wno-variadic-macros -Wno-deprecated
 CPPFLAGS += -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64
 CPPFLAGS += -g
-AR = ar cr
+AR = ar rcs
 
-COMPILER_BIS=gcc
+COMPILER_BIS=$(CC)
 CFLAGS += -I.
 CFLAGS += -O3
 CFLAGS += -c
