--- MergePaths/PathConsensus.cpp.orig	2017-11-22 21:14:18.752421900 +0100
+++ MergePaths/PathConsensus.cpp	2017-11-22 22:06:07.003253446 +0100
@@ -533,6 +533,13 @@
 	return path;
 }
 
+template <typename Seq>
+struct GetLength {
+    size_t operator()(const Seq& sequence) {
+        return sequence.length();
+    }
+};
+
 /* Resolve ambiguous region using multiple alignment of all paths in
  * `solutions'.
  */
@@ -615,7 +622,7 @@
 
 	vector<unsigned> lengths(amb_seqs.size());
 	transform(amb_seqs.begin(), amb_seqs.end(), lengths.begin(),
-			mem_fun_ref(&string::length));
+			GetLength<Sequence>());
 	unsigned minLength = *min_element(lengths.begin(), lengths.end());
 	unsigned maxLength = *max_element(lengths.begin(), lengths.end());
 	float lengthRatio = (float)minLength / maxLength;
