--- a/CMakeLists.txt	2025-04-10 11:55:17.142316229 +0000
+++ b/CMakeLists.txt	2025-04-10 11:59:31.950895379 +0000
@@ -61,7 +61,7 @@
 message (STATUS "Build type: ${CMAKE_BUILD_TYPE}")
 
 set (WARN_FLAGS               "-Wall -Wextra -Wno-unknown-pragmas -pedantic")
-set (CMAKE_CXX_FLAGS          "-std=c++14 ${WARN_FLAGS} -DEPA_VERSION='\"${epa_PROJECT_VERSION}\"'")
+set (CMAKE_CXX_FLAGS          "-fsigned-char -std=c++14 ${WARN_FLAGS} -DEPA_VERSION='\"${epa_PROJECT_VERSION}\"'")
 
 set (CMAKE_CXX_FLAGS_DEBUG    "-O0 -g -ggdb3 -DDEBUG -D_GLIBCXX_DEBUG")
 set (CMAKE_C_FLAGS_DEBUG      "-O0 -g -ggdb3 -D_GLIBCXX_DEBUG")

