--- a/libs/pll-modules/libs/libpll/src/CMakeLists.txt	2025-04-09 16:07:03.103200068 +0000
+++ b/libs/pll-modules/libs/libpll/src/CMakeLists.txt	2020-06-08 17:23:48.000000000 +0000
@@ -167,7 +167,9 @@
   SET_SOURCE_FILES_PROPERTIES( ${LIBPLL_AVX2_SOURCES} PROPERTIES COMPILE_FLAGS ${AVX2_FLAGS} )
 endif ()
 
+if (CMAKE_HOST_SYSTEM_PROCESSOR STREQUAL "x86_64")
 add_definitions(-DHAVE_X86INTRIN_H)
+endif ()
 
 set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${LIBPLL_BASE_FLAGS}")
 
