diff --git a/src/Makefile b/src/Makefile
index e1ef4c3..af7469c 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -7,7 +7,7 @@ PGO_USE_FLAGS=-fprofile-use
 
 
 ctyper: main.cpp CramReader.o  FastaReader.o FastqReader.o KmerHash.o KmerCounter.o KmerMatrix.o KmerWindow.o KtableReader.o PriorData.o Regression.o TreeRound.o Processor.o 
-	g++ $(SAN) $(CPPFLAGS) -o $@ $^ -I $(EIGEN_ROOT)/include/eigen3 -I$(CONDA_PREFIX)/include/ -L/usr/local/lib -L$(CONDA_PREFIX)/lib/ -lhts -lbz2 -lpthread -lz -lm -lstdc++fs  
+	$(CXX) $(SAN) $(CPPFLAGS) -o $@ $^ -I $(EIGEN_ROOT)/include/eigen3 -I$(CONDA_PREFIX)/include/ -L/usr/local/lib -L$(CONDA_PREFIX)/lib/ -lhts -lbz2 -lpthread -lz -lm  
 
 
 ctyper_pgo_generate: CPPFLAGS += $(PGO_GENERATE_FLAGS)
@@ -17,7 +17,7 @@ ctyper_pgo_use: CPPFLAGS += $(PGO_USE_FLAGS)
 ctyper_pgo_use: ctyper
 
 %.o: %.cpp %.hpp FileReader.hpp
-	g++ $(SAN) $(CPPFLAGS) -c $^  -I $(EIGEN_ROOT)/include/eigen3 -I$(CONDA_PREFIX)/include/ -L/usr/local/lib -L$(CONDA_PREFIX)/lib/ -lhts -lbz2 -lpthread -lz -lm -lstdc++fs 
+	$(CXX) $(SAN) $(CPPFLAGS) -c $^  -I $(EIGEN_ROOT)/include/eigen3 -I$(CONDA_PREFIX)/include/ -L/usr/local/lib -L$(CONDA_PREFIX)/lib/ -lhts -lbz2 -lpthread -lz -lm -lstdc++fs
 
 clean:
 	rm *.o *.gch  ctyper
