diff --git a/src/Regression.cpp b/src/Regression.cpp
index 56ea618..9cf316b 100644
--- a/src/Regression.cpp
+++ b/src/Regression.cpp
@@ -6,7 +6,7 @@
 //  If you use this code, please cite our work.
 //   
 
-#include <Eigen/Dense>
+#include <eigen3/Eigen/Dense>
 #include <vector>
 #include <limits>
 #include <algorithm>
diff --git a/src/config.hpp b/src/config.hpp
index 686e1a4..428dc1f 100644
--- a/src/config.hpp
+++ b/src/config.hpp
@@ -2,7 +2,7 @@
 #define config_hpp
 
 #include <stdio.h>
-#include <Eigen/Dense>
+#include <eigen3/Eigen/Dense>
 #include <atomic>
 
 typedef unsigned int uint;
