--- snakemake/conda.py.old	2017-03-15 09:31:40.191297350 +0100
+++ snakemake/conda.py	2017-03-15 09:30:03.075617878 +0100
@@ -114,7 +114,7 @@
                 tmp_file = tmp.name
         else:
             # turn local file url into plain path
-            env_file = strip_prefix("file:")
+            env_file = strip_prefix(env_file, "file:")
 
     env_hash = get_env_hash(env_file)
     env_path = get_env_path(job, env_hash)

