--- a/autogen.sh
+++ b/autogen.sh
@@ -294,7 +294,7 @@
           if [[ "$libadd" == *.la ]]; then
             lib_target="${libadd/\$\(top_srcdir\)/\$(top_builddir)}"
             # Avoid emitting duplicate rules.
-            if [[ ! " ${SUITE_LIB_TARGETS[*]} " =~ " ${lib_target} " ]]; then
+            if [[ ! " ${SUITE_LIB_TARGETS[*]:-} " =~ " ${lib_target} " ]]; then
               SUITE_LIB_TARGETS+=("$lib_target")
               lib_dir="$(dirname "$lib_target")"
               lib_file="$(basename "$lib_target")"
