--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 2.8.12.2 FATAL_ERROR)
+cmake_minimum_required(VERSION 3.18 FATAL_ERROR)

 # Setup cmake policies.
 foreach(policy
@@ -13,6 +13,7 @@
     CMP0075 # CMake 3.12
     CMP0083 # CMake 3.14
     CMP0093 # CMake 3.15
+    CMP0146 # Suppress FindCUDA deprecation/removal warning
   )
   if(POLICY ${policy})
     cmake_policy(SET ${policy} NEW)
@@ -67,6 +68,7 @@
 #------------------------------------------------------------------------------
 # Locate CUDA
 #------------------------------------------------------------------------------
+cmake_policy(SET CMP0146 OLD)
 find_package (CUDA REQUIRED)

 # Set Cufile installation directory to Cuda installation directory by default
