--- a/src/H5FDgds.h
+++ b/src/H5FDgds.h
@@ -30,7 +30,9 @@
  * Application can set these values through the function H5Pset_fapl_gds. */
 #define MBOUNDARY_DEF 4096
 #define FBSIZE_DEF    4096
+#ifndef CBSIZE_DEF
 #define CBSIZE_DEF    16 * 1024 * 1024
+#endif

 herr_t H5Pset_fapl_gds(hid_t fapl_id, size_t alignment, size_t block_size, size_t cbuf_size);
 herr_t H5Pget_fapl_gds(hid_t fapl_id, size_t *boundary /*out*/, size_t *block_size /*out*/,
