diff --git a/meson.build b/meson.build
index 9d727f9..9288fbb 100644
--- a/meson.build
+++ b/meson.build
@@ -23,14 +23,10 @@ static_build = get_option('static')
 
 cc = meson.get_compiler('cpp')
 
-# fastahack_dep = cc.find_library('libfastahack')
 lzma_dep = dependency('liblzma', static: static_build)
 seqlib_dep = dependency('libseqlib', static: static_build, required: false)
 # smithwaterman_dep = cc.find_library('libsmithwaterman')
-tabixpp_dep = cc.find_library('tabixpp', required: false, static: static_build)
 thread_dep = dependency('threads', static: static_build)
-vcflib_dep = cc.find_library('libvcflib')
-wfa2lib_dep = cc.find_library('libwfa2')
 zlib_dep = dependency('zlib', static: static_build)
 
 # to compile htslib use
@@ -41,8 +37,6 @@ if get_option('prefer_system_deps')
 else
   # uses the local git submodules
   htslib_dep = dependency('', required : false)
-  tabixpp_dep = dependency('', required : false)
-  vcflib_dep = dependency('', required : false)
   seqlib_dep = dependency('', required : false)
 endif
 
@@ -123,47 +114,16 @@ if not htslib_dep.found()
     htslib_dep = declare_dependency(link_with : htslib_lib,
                                     include_directories : htslib_inc)
 else
-    htslib_inc = []
-endif
-
-
-if not tabixpp_dep.found()
-    tabixpp_inc = include_directories('contrib/vcflib/contrib/tabixpp/')
-    tabixpp_src = files('contrib/vcflib/contrib/tabixpp/tabix.cpp')
-    tabixpp_lib = static_library('custom_tabixpp',
-                                 tabixpp_src,
-                                 include_directories : [tabixpp_inc, htslib_inc],
-                                 override_options : warn_quiet)
-    tabixpp_dep = declare_dependency(link_with : tabixpp_lib,
-                                     include_directories : tabixpp_inc)
-else
-    tabixpp_inc = []
+    htslib_inc = [get_option('prefix') + '/include']
+    htslib_lib = [get_option('prefix') + '/lib']
 endif
 
-vcflib_inc = include_directories()
-
-# optionally build from source. As we unbundled just check out the vcflib repo in contrib.
-if not vcflib_dep.found()
-  vcflib_inc = include_directories(
-    'contrib/vcflib/src',
-    'contrib/vcflib/contrib/smithwaterman',
-    'contrib/vcflib/contrib/multichoose',
-    'contrib/vcflib/contrib/filevercmp',
-    'contrib/vcflib/contrib/fastahack',
-    )
-  vcflib_src = files(
-    'contrib/vcflib/src/Variant.cpp',
-    'contrib/vcflib/src/allele.cpp',
-    'contrib/vcflib/src/cigar.cpp',
-    'contrib/vcflib/src/split.cpp',
-    )
-  vcflib_lib = static_library('custom_vcflib',
-                              vcflib_src,
-                              include_directories : [htslib_inc, tabixpp_inc, vcflib_inc],
-                              override_options : warn_quiet)
-  vcflib_dep = declare_dependency(link_with : vcflib_lib,
-                                  include_directories : vcflib_inc)
-endif
+tabixpp_inc = [get_option('prefix') + '/include']
+tabixpp_dep = []
+vcflib_inc = [get_option('prefix') + '/include']
+vcflib_lib = [get_option('prefix') + '/lib']
+vcflib_dep = []
+wfa2lib_dep = []
 
 
 if not seqlib_dep.found()
@@ -248,11 +208,11 @@ freebayes_lib = static_library(
 if static_build
   link_arguments = '-static'
 else
-  link_arguments = []
+  link_arguments = ['-ltabixpp', '-lwfa2', '-lvcflib']
 endif
 
 executable('freebayes',
            [freebayes_src,multichoose_src,fastahack_src,smithwaterman_src],
            include_directories : [multichoose_inc, fastahack_inc, smithwaterman_inc, incdir],
            cpp_args : extra_cpp_args,
            link_args: link_arguments,
