From 2ce902eee1f187dab660d07d09cf11223b2c745f Mon Sep 17 00:00:00 2001
From: Lightning Auriga <lightning.auriga@gmail.com>
Date: Sat, 7 Oct 2023 12:26:03 -0400
Subject: [PATCH 2/2] fix intervaltree include path

---
 src/BedReader.cpp | 2 +-
 src/BedReader.h   | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/BedReader.cpp b/src/BedReader.cpp
index dabd0af..ea79235 100644
--- a/src/BedReader.cpp
+++ b/src/BedReader.cpp
@@ -11,7 +11,7 @@
 #include "Utility.h"
 #include "BedReader.h"
 #include "Logging.h"
-#include <IntervalTree.h>
+#include "IntervalTree.h"
 
 using namespace std;
 
diff --git a/src/BedReader.h b/src/BedReader.h
index 5a44f29..8777d53 100644
--- a/src/BedReader.h
+++ b/src/BedReader.h
@@ -9,7 +9,7 @@
 #include <map>
 #include <iterator>
 #include <algorithm>
-#include <IntervalTree.h>
+#include "IntervalTree.h"
 #include "split.h"
 
 using namespace std;
-- 
2.41.0
