diff --git a/meson.build b/meson.build
index 171e407..1e6e019 100644
--- a/meson.build
+++ b/meson.build
@@ -21,7 +21,7 @@
 
 project('freebayes', ['cpp', 'c'],
         version : '1.3.9',
-        default_options : ['warning_level=1', 'cpp_std=c++17', 'optimization=3'])
+        default_options : ['warning_level=1', 'cpp_std=c++14', 'optimization=3'])
 add_global_arguments('-Wno-maybe-uninitialized', language : 'cpp') # for ttmath warnings
 
 static_build = get_option('static')
@@ -170,6 +170,7 @@ extra_cpp_args = cc.get_supported_arguments(
     '-Wno-sign-compare',
     '-Wno-unused-variable',
     '-Wno-unused-but-set-variable',
+    '-Wno-deprecated-declarations',
     )
 
 freebayes_lib = static_library(
@@ -186,7 +187,7 @@ freebayes_lib = static_library(
 if static_build
   link_arguments = '-static'
 else
-  link_arguments = ['-ltabixpp', '-lwfa2', '-lvcflib']
+  link_arguments = ['-ltabixpp', '-lwfa2', '-lvcflib', '-lz', '-lhts', '-pthread']
 endif
 
 executable('freebayes',
