diff --git a/.gitmodules b/.gitmodules
index 7b41c11..fc71644 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -7,9 +7,6 @@
 [submodule "contrib/smithwaterman"]
 	path = contrib/smithwaterman
 	url = https://github.com/ekg/smithwaterman.git
-[submodule "contrib/fastahack"]
-	path = contrib/fastahack
-	url = https://github.com/ekg/fastahack.git
 [submodule "contrib/multichoose"]
 	path = contrib/multichoose
-	url = git@github.com:vcflib/multichoose.git
+    url = https://github.com/vcflib/multichoose.git
