--- freebayes/meson.build	
+++ freebayes/meson.build	
@@ -18,7 +18,7 @@

 htslib_dep = dependency('htslib', required : false)
 tabixpp_dep = cc.find_library('tabixpp', required : false)
-vcflib_dep = dependency('libvcflib', required : false)
+vcflib_dep = cc.find_library('vcflib', required : false)
 seqlib_dep = dependency('libseqlib', required : false)

 # for setting a warning_level on the external code in custom_* targets below
@@ -113,6 +113,8 @@
                               override_options : warn_quiet)
   vcflib_dep = declare_dependency(link_with : vcflib_lib,
                                   include_directories : vcflib_inc)
+else
+    vcflib_inc = []
 endif
