diff --git a/src/Makefile b/src/Makefile
index 5d2748b..477166c 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -5,7 +5,7 @@
 ################################################################################
 
 # Compiler
-CXX=g++
+CXX=g++ ${CXXFLAGS}
 C=gcc
 
 export CXXFLAGS
@@ -47,7 +47,7 @@ $(HTSLIB_ROOT)/libhts.a:
 	cd $(HTSLIB_ROOT) && make
 
 $(SEQLIB_ROOT)/src/libseqlib.a:
-	cd $(SEQLIB_ROOT) && ./configure && make 
+	cd $(SEQLIB_ROOT) && ./configure && make
 
 OBJECTS=BedReader.o \
 		CNV.o \
@@ -117,11 +117,11 @@ alleles.o: alleles.cpp AlleleParser.o Allele.o
 dummy.o: dummy.cpp AlleleParser.o Allele.o
 	$(CXX) $(CXXFLAGS) $(INCLUDE) -c dummy.cpp
 
-freebayes.o: freebayes.cpp TryCatch.h $(HTSLIB_ROOT)/libhts.a 
+freebayes.o: freebayes.cpp TryCatch.h $(HTSLIB_ROOT)/libhts.a
 	$(CXX) $(CXXFLAGS) $(INCLUDE) -c freebayes.cpp
 
 fastlz.o: fastlz.c fastlz.h
-	$(C) $(CFLAGS) $(INCLUDE) -c fastlz.c	
+	$(C) $(CFLAGS) $(INCLUDE) -c fastlz.c
 
 Parameters.o: Parameters.cpp Parameters.h Version.h
 	$(CXX) $(CXXFLAGS) $(INCLUDE) -c Parameters.cpp
@@ -279,4 +279,3 @@ clean:
 	rm -rf *.o *.cgh *~ freebayes alleles ../bin/freebayes ../bin/alleles ../vcflib/*.o ../vcflib/tabixpp/*.{o,a}
 	cd ../vcflib/smithwaterman && make clean
 	cd ../SeqLib && make clean
-
