--- setup.py
+++ setup.py
@@ -40,7 +40,12 @@
     env = os.environ.copy()
 
     if cc := env.get("CC"):
-        if o_flag := compilers.get(cc):
+        o_flag = None
+        if cc.endswith("clang"):
+            o_flag = compilers["clang"]
+        elif cc.endswith(("gcc", "gnu-cc")):
+            o_flag = compilers["gcc"]
+        if o_flag:
             return cc, o_flag
 
     for compiler in compilers.keys():
@@ -203,7 +208,7 @@
         "-O{o}".format(o=o_flag),
         "-lstdc++",
         "-lm",
-        "-pthreads",
+        "-pthread",
         "-fPIC",
     ]
 
