From 52fefb13d652c6e44e8cfee0910b8305d5d60ce6 Mon Sep 17 00:00:00 2001
From: Ed Harry <e.f.harry@gmail.com>
Date: Mon, 23 Mar 2020 15:48:55 +0000
Subject: [PATCH 1/4] tox-conda

---
 tox.ini | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/tox.ini b/tox.ini
index c2f2d15..1eba5de 100644
--- a/tox.ini
+++ b/tox.ini
@@ -7,8 +7,10 @@
 envlist = py38
 
 [testenv]
-deps =
-    pytest
-    numpy
+conda_deps =
+    pytest>=5.3.5
+    numpy>=1.18.1
+conda_channels=
+    conda-forge
 commands =
     pytest
-- 
2.22.0


From 4e0033b02bf26fbc7f302001f59017c7349b0ec1 Mon Sep 17 00:00:00 2001
From: Ed Harry <e.f.harry@gmail.com>
Date: Mon, 23 Mar 2020 16:14:09 +0000
Subject: [PATCH 2/4] tox-conda

---
 tox.ini | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/tox.ini b/tox.ini
index 1eba5de..4bcec53 100644
--- a/tox.ini
+++ b/tox.ini
@@ -10,6 +10,9 @@ envlist = py38
 conda_deps =
     pytest>=5.3.5
     numpy>=1.18.1
+    click>=7.0
+    pandas>=1.0.1
+    biopython>=1.76
 conda_channels=
     conda-forge
 commands =
-- 
2.22.0


From 36901c47d4b4798aca433a9b964ca04b536ca7db Mon Sep 17 00:00:00 2001
From: Ed Harry <e.f.harry@gmail.com>
Date: Mon, 23 Mar 2020 16:53:59 +0000
Subject: [PATCH 3/4] tox-conda

---
 tox.ini | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/tox.ini b/tox.ini
index 4bcec53..2c80a6e 100644
--- a/tox.ini
+++ b/tox.ini
@@ -15,5 +15,7 @@ conda_deps =
     biopython>=1.76
 conda_channels=
     conda-forge
+install_command =
+    python -m pip install {opts} {packages} --no-deps --ignore-installed
 commands =
     pytest
-- 
2.22.0


From 2cb97f20eeecafd4eedd25b3a6d73606baef519b Mon Sep 17 00:00:00 2001
From: Ed Harry <e.f.harry@gmail.com>
Date: Mon, 23 Mar 2020 17:19:04 +0000
Subject: [PATCH 4/4] tox-conda

---
 tox.ini | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/tox.ini b/tox.ini
index 2c80a6e..eb5c21c 100644
--- a/tox.ini
+++ b/tox.ini
@@ -13,7 +13,9 @@ conda_deps =
     click>=7.0
     pandas>=1.0.1
     biopython>=1.76
-conda_channels=
+    bwa>=0.7.17
+    samtools>=1.10
+conda_channels =
     conda-forge
 install_command =
     python -m pip install {opts} {packages} --no-deps --ignore-installed
-- 
2.22.0

