diff --git a/MANIFEST.in b/MANIFEST.in
new file mode 100644
index 0000000..c6daccf
--- /dev/null
+++ b/MANIFEST.in
@@ -0,0 +1,10 @@
+include LICENSE
+include README.md
+
+recursive-include tests *
+recursive-exclude * __pycache__
+recursive-exclude * *.py[co]
+recursive-exclude dist
+
+recursive-include scripts *
+recursive-include svtk *.py
diff --git a/svtk/baf/__init__.py b/svtk/baf/__init__.py
new file mode 100644
index 0000000..e69de29
