diff --git MANIFEST.in MANIFEST.in
new file mode 100644
index 0000000..c6daccf
--- /dev/null
+++ MANIFEST.in
@@ -0,0 +1,10 @@
+include LICENSE
+include README.md
+
+recursive-include tests *
+recursive-exclude * __pycache__
+recursive-exclude * *.py[co]
+recursive-exclude dist
+
+recursive-include scripts *
+recursive-include svtk *.py
diff --git svtk/baf/__init__.py svtk/baf/__init__.py
new file mode 100644
index 000000000..8d242f1f7
--- /dev/null
+++ svtk/baf/__init__.py
@@ -0,0 +1 @@
+# placeholder for macOS patch
