diff --git a/kent/src/hg/lib/straw/straw.cpp b/kent/src/hg/lib/straw/straw.cpp
index bc11152..b4c298b 100644
--- a/kent/src/hg/lib/straw/straw.cpp
+++ b/kent/src/hg/lib/straw/straw.cpp
@@ -21,7 +21,6 @@
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
 */
-#include <cstdint>
 #include <cstring>
 #include <iostream>
 #include <fstream>
@@ -39,6 +38,7 @@
 #include "straw.h"
 extern "C" {
 #include "../../inc/fakeCurl.h"
+#include <cstdint>
 }
 
 using namespace std;