diff --git a/lib/kit.py b/lib/kit.py
index bdc6e97..e94c0dc 100644
--- a/lib/kit.py
+++ b/lib/kit.py
@@ -200,6 +200,8 @@ def latestver(url):
 		return 'Unknown'
 
 def getver(path):
+	"""
+	# remove internet requirement for determining version
 	ver = 'Unknown'
 	readme = path + '/README.md'
 	if os.path.exists(readme):
@@ -210,7 +212,8 @@ def getver(path):
 	latest = latestver('https://api.github.com/repos/Nextomics/NextPolish/releases/latest')
 	if latest != 'Unknown' and ver != latest:
 		print(('\033[35mPlease update to the latest version: %s, current version: %s \033[0m') % (latest, ver))
-	return ver
+	"""
+	return "BIOCONDA_SED_REPLACE"
 
 def pypath():
 	return sys.executable
