--- a/setup.py
+++ b/setup.py
@@ -29,7 +29,7 @@
         'vcfpy==0.13.8',
         'requests',
         'PyYAML>=5.1',
-        'biopython==1.77',
+        'biopython>=1.77',
         'networkx',
         'simanneal',
         'numpy==1.26.4',
@@ -54,7 +54,7 @@
         'openpyxl',
         'deepdiff',
         'scikit-learn>=1.6.0,<1.8.0',
-        'imblearn',
+        'imbalanced-learn',
     ],
     packages=find_namespace_packages(),
     include_package_data=True,
