From ea351f095cfcf4f71f1e51e67fe04b2a656b2264 Mon Sep 17 00:00:00 2001
From: "Kevin R. Thornton" <krthornt@uci.edu>
Date: Mon, 9 Nov 2015 16:05:40 -0800
Subject: [PATCH] fixed bizarre comparison.  How did that ever do what was
 expected...

---
 src/hts/bamreader.cc | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/hts/bamreader.cc b/src/hts/bamreader.cc
index 395794d4..f5c0d147 100644
--- a/src/hts/bamreader.cc
+++ b/src/hts/bamreader.cc
@@ -36,7 +36,7 @@ namespace Sequence
     __htext(nullptr),
     __refdata(std::vector< std::pair<std::string,I32> >())
   {
-    if(gzopen != NULL)
+    if(in != NULL)
       {
 	auto rv = bgzf_read( in, &__magic[0], 4*sizeof(char) );
 	if (!rv) __EOF = true;
