diff -u -ur samblaster-v.0.1.24/samblaster.cpp samblaster-v.0.1.24-b/samblaster.cpp
--- samblaster-v.0.1.24/samblaster.cpp	2016-11-28 22:55:19.000000000 +0100
+++ samblaster-v.0.1.24-b/samblaster.cpp	2019-10-23 16:28:38.306336222 +0200
@@ -1602,28 +1602,28 @@
     if (!state->quiet)
     {
         if (state->discordantFile != NULL)
-            fprintf(stderr, "samblaster: Output %"PRIu64" discordant read pairs to %s\n", discCount/2, state->discordantFileName);
+            fprintf(stderr, "samblaster: Output %" PRIu64 " discordant read pairs to %s\n", discCount/2, state->discordantFileName);
         if (state->splitterFile != NULL)
-            fprintf(stderr, "samblaster: Output %"PRIu64" split reads to %s\n", splitCount, state->splitterFileName);
+            fprintf(stderr, "samblaster: Output %" PRIu64 " split reads to %s\n", splitCount, state->splitterFileName);
         if (state->unmappedClippedFile != NULL)
-            fprintf(stderr, "samblaster: Output %"PRIu64" unmapped/clipped reads to %s\n", unmapClipCount, state->unmappedClippedFileName);
+            fprintf(stderr, "samblaster: Output %" PRIu64 " unmapped/clipped reads to %s\n", unmapClipCount, state->unmappedClippedFileName);
     }
 
     // Output stats.
     if (state->ignoreUnmated)
     {
-        fprintf(stderr, "samblaster: Found %"PRIu64" of %"PRIu64" (%4.2f%%) read ids unmated\n",
+        fprintf(stderr, "samblaster: Found %" PRIu64 " of %" PRIu64 " (%4.2f%%) read ids unmated\n",
                 unmatedCount, idCount, ((double)100)*unmatedCount/idCount);
         if (unmatedCount > 0) fprintf(stderr, "samblaster: Please double check that input file is read-id (QNAME) grouped\n");
     }
     if (state->removeDups)
     {
-        fprintf(stderr, "samblaster: Removed %"PRIu64" of %"PRIu64" (%4.2f%%) read ids as duplicates",
+        fprintf(stderr, "samblaster: Removed %" PRIu64 " of %" PRIu64 " (%4.2f%%) read ids as duplicates",
                 dupCount, idCount, ((double)100)*dupCount/idCount);
     }
     else
     {
-        fprintf(stderr, "samblaster: Marked %"PRIu64" of %"PRIu64" (%4.2f%%) read ids as duplicates",
+        fprintf(stderr, "samblaster: Marked %" PRIu64 " of %" PRIu64 " (%4.2f%%) read ids as duplicates",
                 dupCount, idCount, ((double)100)*dupCount/idCount);
     }
     if ((TIMING == 0) || state->quiet)
