--- kent/src/inc/common.mk	2017-11-07 17:46:00.000000000 -0500
+++ kent/src/inc/common.mk.new	2017-11-13 17:44:51.017090255 -0500
@@ -17,7 +17,7 @@
 endif
 
 HG_DEFS=-D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_GNU_SOURCE -DMACHTYPE_${MACHTYPE}
-HG_INC+=-I../inc -I../../inc -I../../../inc -I../../../../inc -I../../../../../inc -I$(kentSrc)/htslib
+HG_INC+=-I../inc -I../../inc -I../../../inc -I../../../../inc -I../../../../../inc -I$(kentSrc)/htslib -I ${PREFIX}/include
 
 # to check for Mac OSX Darwin specifics:
 UNAME_S := $(shell uname -s)
--- kent/src/utils/bedJoinTabOffset	2018-06-06 02:23:56.000000000 +0200
+++ kent/src/utils/bedJoinTabOffset	2018-06-06 02:23:56.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python2.7
+#!/usr/bin/env python
 
 import logging, sys, optparse, string
 from collections import defaultdict
--- kent/src/utils/expMatrixToBarchartBed/expMatrixToBarchartBed	2018-06-06 02:23:56.000000000 +0200
+++ kent/src/utils/expMatrixToBarchartBed/expMatrixToBarchartBed	2018-06-06 02:23:56.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python2.7
+#!/usr/bin/env python
 # expMatrixToBarchartBed
 """
 Generate a barChart bed6+5 file from a matrix, meta data, and coordinates.
