index 7600442..a1fab79 100644
--- a/wavefront/wavefront_extend_kernels.c
+++ b/wavefront/wavefront_extend_kernels.c
@@ -29,7 +29,11 @@
  * DESCRIPTION: WFA module for the "extension" of exact matches
  */
 
-#include <endian.h>
+#ifdef __APPLE__ || defined(__FreeBSD__)
+ #include <machine/endian.h>  // __BYTE_ORDER
+ #else
+ #include <endian.h>  // __BYTE_ORDER
+ #endif
 
 #include "wavefront_extend_kernels.h"
 #include "wavefront_termination.h"
