diff --git a/alignment/cigar.h b/alignment/cigar.h
index fbc24ce..3780c9c 100644
--- a/alignment/cigar.h
+++ b/alignment/cigar.h
@@ -32,6 +32,7 @@
 #ifndef CIGAR_H_
 #define CIGAR_H_
 
+#include <stdbool.h>
 #include "system/mm_allocator.h"
 #include "alignment/linear_penalties.h"
 #include "alignment/affine_penalties.h"
diff --git a/system/mm_allocator.h b/system/mm_allocator.h
index 1d2cac1..00390fa 100644
--- a/system/mm_allocator.h
+++ b/system/mm_allocator.h
@@ -35,6 +35,7 @@
 #ifndef MM_ALLOCATOR_H_
 #define MM_ALLOCATOR_H_
 
+#include <stdbool.h>
 #include "utils/vector.h"
 
 /*
diff --git a/utils/vector.h b/utils/vector.h
index 4bd4776..6043e22 100644
--- a/utils/vector.h
+++ b/utils/vector.h
@@ -34,6 +34,7 @@
 #define VECTOR_H_
 
 #include <stdint.h>
+#include <stdbool.h>
 
 /*
  * Checkers
