commit 20dc54f50fead46c71a92a2dc60a86fd04f7c808
Author: Marcel Bargull <marcel.bargull@udo.edu>
Date:   Wed Apr 14 22:24:54 2021 +0200

    Fix non-relative includes

diff --git a/src/include/gzIndex.h b/src/include/gzIndex.h
index 9d9e944..20bce11 100644
--- a/src/include/gzIndex.h
+++ b/src/include/gzIndex.h
@@ -59,7 +59,7 @@ using namespace std;
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
-#include "zlib.h"
+#include <zlib.h>
 #include <iostream>
 #include <fstream>
 
diff --git a/src/main.cpp b/src/main.cpp
index de1aa1a..242864d 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -6,7 +6,7 @@
 #include "subcommands/indexFastq.h"
 #include "subcommands/queryFastq.h"
 #include <set>
-#include "string.h"
+#include <string.h>
 
 
 using namespace std;
