From 88a91f2b4076914b780751ed1485e5d0e22946a8 Mon Sep 17 00:00:00 2001
From: Marcel Bargull <marcel.bargull@udo.edu>
Date: Fri, 16 Apr 2021 22:27:14 +0200
Subject: [PATCH 1/4] Makefile: Fix -L./lib flag for sub build

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 2de8868..fe73c0c 100644
--- a/Makefile
+++ b/Makefile
@@ -15,7 +15,7 @@ LIBDIR = $(PREFIX)/lib
 
 CXX ?= g++
 CXXFLAGS += -Wall -pedantic -O3 -m64 -std=c++11 -fPIC
-LDFLAGS += -L./lib -Wl,-rpath,$(LIBDIR) -Wl,--whole-archive -lpthread -Wl,--no-whole-archive
+LDFLAGS += -L$(curDir)/lib -Wl,-rpath,$(LIBDIR) -Wl,--whole-archive -lpthread -Wl,--no-whole-archive
 
 BAMTOOLS_LIB_PREFIX = lrez_
 BAMTOOLS_LIB = lib/lib$(BAMTOOLS_LIB_PREFIX)bamtools$(SHLIB_EXT)
-- 
2.31.1

