From ff744805f35768f2bd11fb7baebdbe0cae0678f9 Mon Sep 17 00:00:00 2001
From: Marcel Bargull <marcel.bargull@udo.edu>
Date: Sat, 17 Apr 2021 08:13:23 +0200
Subject: [PATCH 2/3] Makefile: drop --whole-archive -lpthread

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index fe73c0c..e616adb 100644
--- a/Makefile
+++ b/Makefile
@@ -15,7 +15,7 @@ LIBDIR = $(PREFIX)/lib
 
 CXX ?= g++
 CXXFLAGS += -Wall -pedantic -O3 -m64 -std=c++11 -fPIC
-LDFLAGS += -L$(curDir)/lib -Wl,-rpath,$(LIBDIR) -Wl,--whole-archive -lpthread -Wl,--no-whole-archive
+LDFLAGS += -L$(curDir)/lib -Wl,-rpath,$(LIBDIR)
 
 BAMTOOLS_LIB_PREFIX = lrez_
 BAMTOOLS_LIB = lib/lib$(BAMTOOLS_LIB_PREFIX)bamtools$(SHLIB_EXT)
-- 
2.31.1

