--- Makefile
+++ Makefile
@@ -153 +153 @@
-VERSION = $(shell cat VERSION)
+VERSION = $(shell cat VERSION.txt)
@@ -254 +254 @@
-	VERSION
+	VERSION.txt
@@ -290 +290 @@
-     -DHISAT2_VERSION="\"`cat VERSION`\"" \
+     -DHISAT2_VERSION="\"`cat VERSION.txt`\"" \
--- hisat2.xcodeproj/project.pbxproj
+++ hisat2.xcodeproj/project.pbxproj
@@ -1006 +1006 @@
-					"HISAT2_VERSION2=\"\\\"`cat VERSION`\\\"\"",
+					"HISAT2_VERSION2=\"\\\"`cat VERSION.txt`\\\"\"",
@@ -1138 +1138 @@
-					"HISAT2_VERSION2=\"\\\"`cat VERSION`\\\"\"",
+					"HISAT2_VERSION2=\"\\\"`cat VERSION.txt`\\\"\"",
@@ -1184 +1184 @@
-					"HISAT2_VERSION=\"\\\"`cat VERSION`\\\"\"",
+					"HISAT2_VERSION=\"\\\"`cat VERSION.txt`\\\"\"",
@@ -1206 +1206 @@
-					"HISAT2_VERSION2=\"\\\"`cat VERSION`\\\"\"",
+					"HISAT2_VERSION2=\"\\\"`cat VERSION.txt`\\\"\"",
--- msvcc/codeStubs.vcxproj
+++ msvcc/codeStubs.vcxproj
@@ -75 +75 @@
-      <Inputs>../VERSION;</Inputs>
+      <Inputs>../VERSION.txt;</Inputs>
@@ -97 +97 @@
-      <Inputs>../VERSION;</Inputs>
+      <Inputs>../VERSION.txt;</Inputs>
--- msvcc/SetVersion.vbs
+++ msvcc/SetVersion.vbs
@@ -4 +4 @@
-set FileHandler = FS.GetFile("../VERSION")
+set FileHandler = FS.GetFile("../VERSION.txt")
