--- src/alien/oxygen/src/mcxdeblast.bak 2016-10-28 18:33:42.552891732 +0000
+++ src/alien/oxygen/src/mcxdeblast     2016-10-28 18:34:02.988093841 +0000
@@ -1,4 +1,4 @@
-#!/usr/local/bin/perl -w
+#!/usr/bin/env perl

 use Getopt::Long;

--- scripts/clxdo.bak   2016-10-28 18:36:06.243058808 +0000
+++ scripts/clxdo       2016-10-28 18:38:09.652159565 +0000
@@ -83,7 +83,7 @@
 require_num 1 "<cls-file>+"
 mx=$2
 mcxdump -imx $mx --no-values --dump-lines -o - \
-|  /usr/local/bin/perl -ane '$ct{@F-1}++; END { print map { "$_ $ct{$_}\n" } sort { $a <=> $b } keys %ct; }'
+|  perl -ane '$ct{@F-1}++; END { print map { "$_ $ct{$_}\n" } sort { $a <=> $b } keys %ct; }'
       ;;


@@ -116,7 +116,7 @@
 shift 2
 for mx in $@; do
       mcxdump -imx $mx --no-values --dump-lines -o - \
-   |  /usr/local/bin/perl -ane '$ct{@F-1}++; END {%i = map { ($_, 1) } grep { $_ >= $ENV{CLXDO_VAL1}; } keys %ct; ($lt, $n_lt, $gq, $n_gq) = (0,0,0,0); for my $x (keys %ct) { if ($i{$x}) { $n_gq += $ct{$x}; $gq += $x * $ct{$x}; } else { $n_lt += $ct{$x}; $lt += $x * $ct{$x} } } print "$n_lt $lt $n_gq $gq\n"; }'
+   |  perl -ane '$ct{@F-1}++; END {%i = map { ($_, 1) } grep { $_ >= $ENV{CLXDO_VAL1}; } keys %ct; ($lt, $n_lt, $gq, $n_gq) = (0,0,0,0); for my $x (keys %ct) { if ($i{$x}) { $n_gq += $ct{$x}; $gq += $x * $ct{$x}; } else { $n_lt += $ct{$x}; $lt += $x * $ct{$x} } } print "$n_lt $lt $n_gq $gq\n"; }'
 done
       ;;
