--- mapsplice.py	2015-08-11 20:43:59.000000000 +0000
+++ mapsplice.py.new	2015-12-18 17:39:40.160231178 +0000
@@ -1,4 +1,5 @@
- #    MapSplice
+#!/usr/bin/env python
+ #   MapSplice
  #
  #    Copyright (C) 2012-2013 University of Kentucky 
  #
@@ -105,7 +106,7 @@
 formated_chrom_dir = temp_dir + "formated_chrom/"
 formated_reads_dir = temp_dir + "formated_reads/"
 filter_repeats_dir = temp_dir + "filter_repeats/"
-bin_dir = sys.path[0] + "/bin/"
+bin_dir = sys.path[0] + "/"
 rerun_all = 1
 DEBUG = 0
 max_chromosome = 1000
