--- a/kmerfinder.py
+++ b/kmerfinder.py
@@ -130,7 +130,7 @@ if __name__ == "__main__":
     parser.add_argument('-i', '--infile',  help="FASTA(.gz) or FASTQ(.gz) file(s) to run KmerFinder on.", nargs='+')
     parser.add_argument('-batch', '--batch_file',  help="OPTION NOT AVAILABLE:file with multipe files listed")  # only for local download 
     parser.add_argument('-o', '--output_folder',  help="folder to store the output", default='output')
-    parser.add_argument('-db', '--db_path',  help="path to database and database file") 
+    parser.add_argument('-db', '--db_path',  help="path to database and database file", default=os.getenv("KmerFinder_DB")) 
     parser.add_argument('-db_batch', '--db_batch',  help="OPTION NOT AVAILABLE:file with paths to multiple databases") # only for local download -- not implemented yet
     # parser.add_argument('-kma', '--kma_arguments',  help="Extra arguments for KMA", nargs='+',action=MyAction)
     parser.add_argument('-kma', '--kma_arguments',  help="OPTION NOT AVAILABLE:Extra arguments for KMA", type=str)
