--- BiocGenerics/DESCRIPTION	2019-05-03 02:20:26.000000000 +0200
+++ BiocGenerics/DESCRIPTION	2019-05-03 02:20:26.000000000 +0200
@@ -5,7 +5,7 @@
 Author: The Bioconductor Dev Team
 Maintainer: Bioconductor Package Maintainer <maintainer@bioconductor.org>
 biocViews: Infrastructure
-Depends: R (>= 3.6.0), methods, utils, graphics, stats, parallel
+Depends: R (< 3.6.0), methods, utils, graphics, stats, parallel
 Imports: methods, utils, graphics, stats, parallel
 Suggests: Biobase, S4Vectors, IRanges, GenomicRanges, Rsamtools,
         AnnotationDbi, oligoClasses, oligo, affyPLM, flowClust, affy,
@@ -14,7 +14,7 @@
 Collate: S3-classes-as-S4-classes.R utils.R normarg-utils.R
         hotfix73465.R replaceSlots.R append.R as.data.frame.R as.list.R
         as.vector.R cbind.R do.call.R duplicated.R eval.R Extremes.R
-        funprog.R get.R grep.R is.unsorted.R lapply.R mapply.R match.R
+        funprog.R get.R grep.R is.unsorted.R lapply.R lengths.R mapply.R match.R
         matrix-summary.R mean.R nrow.R order.R paste.R rank.R rep.R
         row_colnames.R sets.R sort.R start.R subset.R t.R table.R
         tapply.R unique.R unlist.R unsplit.R relist.R var.R which.R
--- BiocGenerics/NAMESPACE	2019-05-02 22:58:30.000000000 +0200
+++ BiocGenerics/NAMESPACE	2019-05-02 22:58:30.000000000 +0200
@@ -66,6 +66,9 @@
     ## from R/lapply.R:
     lapply, sapply,
 
+    ## from R/lengths.R:
+    lengths,
+
     ## from R/mapply.R:
     mapply,
 
--- BiocGenerics/R/lengths.R	1970-01-01 01:00:00.000000000 +0100
+++ BiocGenerics/R/lengths.R	1970-01-01 01:00:00.000000000 +0100
@@ -0,0 +1,10 @@
+### =========================================================================
+### The lengths() generic
+### -------------------------------------------------------------------------
+###
+### Need to explicitly define this generic otherwise the implicit generic in
+### package "base" would dispatch on all its arguments. Here we set dispatch
+### on the first arg ('x') only.
+
+setGeneric("lengths", signature="x")
+
