diff --git a/gfalibs/Makefile b/gfalibs/Makefile
index fe029bd..9f4be35 100644
--- a/gfalibs/Makefile
+++ b/gfalibs/Makefile
@@ -1,4 +1,4 @@
-CXX ?= g++
+CXX ?= $(CXX)
 INCLUDE_DIR += -I./include
 WARNINGS = -Wall -Wextra
 
@@ -8,14 +8,13 @@ TARGET = gfalibs
 BUILD = build/bin
 SOURCE = src
 INCLUDE = include
-LDFLAGS =
 
 SOURCES = $(addsuffix .o, input-filters input-gfa input-agp gfa gfa-lines log stream-obj uid-generator struct output memory MinScan)
 
 all: $(SOURCES)
 
 %.o: $(SOURCE)/%.cpp $(INCLUDE)/%.h
-	$(CXX) $(CXXFLAGS) $(CXXFLAGS) $(LDFLAGS) -c $(SOURCE)/$(basename $@).cpp -o $@
+	$(CXX) $(CPPFLAGS) $(CXXFLAGS) $(CXXFLAGS) $(LDFLAGS) -c $(SOURCE)/$(basename $@).cpp -o $@
     
 clean:
 	$(RM) *.o
