diff --git a/Makefile b/Makefile
index 5823f28..cb3efa1 100644
--- a/Makefile
+++ b/Makefile
@@ -1,5 +1,5 @@
-CXX ?= g++
-INCLUDE_DIR = -I./include -I./gfalibs/include
+CXX ?= ${CXX}
+INCLUDE_DIR = -I./include -I./gfalibs/include -I${PREFIX}/include
 WARNINGS = -Wall -Wextra
 
 CXXFLAGS = -g -std=gnu++14 -O3 $(INCLUDE_DIR) $(WARNINGS)
@@ -12,8 +12,8 @@ SOURCE = src
 INCLUDE = include
 BINDIR := $(BUILD)/.o
 
-LDFLAGS := -pthread
-LIBS = -lz
+LDFLAGS := -pthread -L${PREFIX}/lib
+LIBS := -lz
 
 OBJS := main input reads
 BINS := $(addprefix $(BINDIR)/, $(OBJS))
@@ -27,7 +27,7 @@ head: $(BINS) gfalibs | $(BUILD)
 all: head validate regenerate
 
 $(BINDIR)%: $(SOURCE)/%.cpp $(INCLUDE)/%.h | $(BINDIR)
-	$(CXX) $(CXXFLAGS) $(LDFLAGS) -c $(SOURCE)/$(notdir $@).cpp -o $@
+	$(CXX) $(CXXFLAGS) $(LDFLAGS) -c $(SOURCE)/$(notdir $@).cpp -o $@ $(LIBS)
 
 .PHONY: gfalibs
 gfalibs:
