diff --git a/Makefile b/Makefile
index 5823f28..6f4a60f 100644
--- a/Makefile
+++ b/Makefile
@@ -1,5 +1,5 @@
-CXX ?= g++
-INCLUDE_DIR = -I./include -I./gfalibs/include
+CXX ?= ${CXX}
+INCLUDE_DIR = -I${PREFIX}/include -I./include -I./gfalibs/include
 WARNINGS = -Wall -Wextra
 
 CXXFLAGS = -g -std=gnu++14 -O3 $(INCLUDE_DIR) $(WARNINGS)
@@ -12,8 +12,8 @@ SOURCE = src
 INCLUDE = include
 BINDIR := $(BUILD)/.o
 
-LDFLAGS := -pthread
-LIBS = -lz
+LDFLAGS := -pthread -L${PREFIX}/lib
+LIBS := -lz -lpthread
 
 OBJS := main input reads
 BINS := $(addprefix $(BINDIR)/, $(OBJS))
@@ -22,12 +22,12 @@ BINS := $(addprefix $(BINDIR)/, $(OBJS))
 GFALIBS_DIR := $(CURDIR)/gfalibs
 
 head: $(BINS) gfalibs | $(BUILD)
-	$(CXX) $(CXXFLAGS) $(LDFLAGS) -o $(BUILD)/$(TARGET) $(BINDIR)/* $(GFALIBS_DIR)/*.o $(LIBS)
+	$(CXX) $(CXXFLAGS) $(INCLUDE_DIR) $(LDFLAGS) -o $(BUILD)/$(TARGET) $(BINDIR)/* $(GFALIBS_DIR)/*.o $(LIBS)
 	
 all: head validate regenerate
 
 $(BINDIR)%: $(SOURCE)/%.cpp $(INCLUDE)/%.h | $(BINDIR)
-	$(CXX) $(CXXFLAGS) $(LDFLAGS) -c $(SOURCE)/$(notdir $@).cpp -o $@
+	$(CXX) $(CXXFLAGS) $(INCLUDE_DIR) $(LDFLAGS) -c $(SOURCE)/$(notdir $@).cpp -o $@ $(LIBS)
 
 .PHONY: gfalibs
 gfalibs:
