diff --git a/gfalibs/Makefile b/gfalibs/Makefile
index de9ab9b..f291529 100644
--- a/gfalibs/Makefile
+++ b/gfalibs/Makefile
@@ -1,4 +1,3 @@
-CXX = g++
 INCLUDE_DIR = -I./include
 WARNINGS = -Wall -Wextra
 
@@ -9,6 +8,7 @@ BUILD = build/bin
 SOURCE = src
 INCLUDE = include
 LDFLAGS :=
+LIBS := -lz
 
 SOURCES = $(addsuffix .o, input-filters input-gfa input-agp gfa gfa-lines log stream-obj uid-generator struct output memory)
 
@@ -16,7 +16,7 @@ all: $(SOURCES)
 	@
 
 %.o: $(SOURCE)/%.cpp $(INCLUDE)/%.h
-	$(CXX) $(CXXFLAGS) $(LDFLAGS) -c $(SOURCE)/$(basename $@).cpp -o $@
+	$(CXX) $(CXXFLAGS) $(LDFLAGS) -c $(SOURCE)/$(basename $@).cpp -o $@ $(LIBS)
     
 clean:
 	$(RM) *.o
