diff --git a/gfalibs/Makefile b/gfalibs/Makefile
index de9ab9b..90a703c 100644
--- a/gfalibs/Makefile
+++ b/gfalibs/Makefile
@@ -1,5 +1,5 @@
-CXX = g++
-INCLUDE_DIR = -I./include
+CXX = $(CXX)
+INCLUDE_DIR = -I${PREFIX}/include -I./include -I./include/parallel-hashmap
 WARNINGS = -Wall -Wextra
 
 CXXFLAGS = -g -std=gnu++14 -O3 $(INCLUDE_DIR) $(WARNINGS)
@@ -8,7 +8,7 @@ TARGET = gfalibs
 BUILD = build/bin
 SOURCE = src
 INCLUDE = include
-LDFLAGS :=
+LDFLAGS := -L${PREFIX}/lib -pthread
 
 SOURCES = $(addsuffix .o, input-filters input-gfa input-agp gfa gfa-lines log stream-obj uid-generator struct output memory)
 
@@ -16,7 +16,7 @@ all: $(SOURCES)
 	@
 
 %.o: $(SOURCE)/%.cpp $(INCLUDE)/%.h
-	$(CXX) $(CXXFLAGS) $(LDFLAGS) -c $(SOURCE)/$(basename $@).cpp -o $@
+	$(CXX) $(CXXFLAGS) $(INCLUDE_DIR) $(LDFLAGS) -c $(SOURCE)/$(basename $@).cpp -o $@ -lz -lpthread
     
 clean:
 	$(RM) *.o
