--- gfalibs.orig/include/functions.h	2024-11-12 16:11:22.000000000 -0500
+++ gfalibs/include/functions.h	2024-11-13 11:26:47.585980869 -0500
@@ -31,7 +31,7 @@
 #include <direct.h>
 #endif
 
-#include <parallel-hashmap/phmap.h>
+#include "parallel-hashmap/phmap.h"
 #include "global.h"
 #include "bed.h"
 #include "struct.h"
--- gfalibs.orig/include/kmer.h	2024-11-12 16:11:22.000000000 -0500
+++ gfalibs/include/kmer.h	2024-11-13 11:27:01.022050240 -0500
@@ -2,7 +2,7 @@
 #define KMER_H
 
 #include <future>
-#include <parallel-hashmap/phmap.h>
+#include "parallel-hashmap/phmap.h"
 #include "parallel-hashmap/phmap_dump.h"
 
 #include <fastx.h>
--- gfalibs.orig/src/gfa.cpp	2024-11-12 16:11:22.000000000 -0500
+++ gfalibs/src/gfa.cpp	2024-11-13 11:27:44.086272576 -0500
@@ -2,7 +2,7 @@
 #include <string>
 #include <vector>
 
-#include <parallel-hashmap/phmap.h>
+#include "parallel-hashmap/phmap.h"
 
 #include "log.h"
 #include "global.h"
--- gfalibs.orig/src/input-agp.cpp	2024-11-12 16:11:22.000000000 -0500
+++ gfalibs/src/input-agp.cpp	2024-11-13 11:27:10.734100380 -0500
@@ -2,7 +2,7 @@
 #include <istream>
 #include <fstream>
 
-#include <parallel-hashmap/phmap.h>
+#include "parallel-hashmap/phmap.h"
 
 #include "bed.h"
 #include "struct.h"
--- gfalibs.orig/src/input-gfa.cpp	2024-11-12 16:11:22.000000000 -0500
+++ gfalibs/src/input-gfa.cpp	2024-11-13 11:27:25.134174730 -0500
@@ -5,7 +5,7 @@
 #include <vector>
 #include <iostream>
 
-#include <parallel-hashmap/phmap.h>
+#include "parallel-hashmap/phmap.h"
 
 #include "bed.h"
 #include "struct.h"
--- gfalibs.orig/src/output.cpp	2024-11-12 16:11:22.000000000 -0500
+++ gfalibs/src/output.cpp	2024-11-13 11:27:35.606228804 -0500
@@ -5,7 +5,7 @@
 #include <fstream>
 #include <deque>
 
-#include <parallel-hashmap/phmap.h>
+#include "parallel-hashmap/phmap.h"
 
 #include "bed.h"
 #include "struct.h"
