--- old/gfalibs/include/kmer.h
+++ new/gfalibs/include/kmer.h
@@ -2,7 +2,7 @@
 #define KMER_H
 
 #include <future>
-#include <parallel-hashmap/phmap.h>
+#include "parallel-hashmap/phmap.h"
 #include "parallel-hashmap/phmap_dump.h"
 
 #include <fastx.h>
--- old/gfalibs/include/functions.h
+++ new/gfalibs/include/functions.h
@@ -31,7 +31,7 @@
 #include <direct.h>
 #endif
 
-#include <parallel-hashmap/phmap.h>
+#include "parallel-hashmap/phmap.h"
 #include "global.h"
 #include "bed.h"
 #include "struct.h"
--- old/gfalibs/src/gfa.cpp
+++ new/gfalibs/src/gfa.cpp
@@ -2,7 +2,7 @@
 #include <string>
 #include <vector>
 
-#include <parallel-hashmap/phmap.h>
+#include "parallel-hashmap/phmap.h"
 
 #include "log.h"
 #include "global.h"
--- old/gfalibs/src/input-gfa.cpp
+++ new/gfalibs/src/input-gfa.cpp
@@ -5,7 +5,7 @@
 #include <vector>
 #include <iostream>
 
-#include <parallel-hashmap/phmap.h>
+#include "parallel-hashmap/phmap.h"
 
 #include "bed.h"
 #include "struct.h"
--- old/gfalibs/src/input-agp.cpp
+++ new/gfalibs/src/input-agp.cpp
@@ -2,7 +2,7 @@
 #include <istream>
 #include <fstream>
 
-#include <parallel-hashmap/phmap.h>
+#include "parallel-hashmap/phmap.h"
 
 #include "bed.h"
 #include "struct.h"
--- old/gfalibs/src/output.cpp
+++ new/gfalibs/src/output.cpp
@@ -5,7 +5,7 @@
 #include <fstream>
 #include <deque>
 
-#include <parallel-hashmap/phmap.h>
+#include "parallel-hashmap/phmap.h"
 
 #include "bed.h"
 #include "struct.h"
--- old/gfalibs/src/input-filters.cpp
+++ new/gfalibs/src/input-filters.cpp
@@ -2,7 +2,7 @@
 #include <string>
 #include <vector>
 
-#include <parallel-hashmap/phmap.h>
+#include "parallel-hashmap/phmap.h"
 
 #include "bed.h"
 #include "struct.h"
