diff --git a/Makefile b/Makefile
index 5823f28..7e1295c 100644
--- a/Makefile
+++ b/Makefile
@@ -14,6 +14,7 @@ BINDIR := $(BUILD)/.o
 
 LDFLAGS := -pthread
 LIBS = -lz
+-L/usr/lib/x86_64-linux-gnu -lz
 
 OBJS := main input reads
 BINS := $(addprefix $(BINDIR)/, $(OBJS))
diff --git a/src/reads.cpp b/src/reads.cpp
index 866f49b..ac714cf 100644
--- a/src/reads.cpp
+++ b/src/reads.cpp
@@ -14,7 +14,7 @@
 #include "functions.h" // global functions
 #include "stream-obj.h"
 
-#include "zlib.h"
+#include <zlib.h>
 #include "zstream/zstream_common.hpp"
 #include "zstream/ozstream.hpp"
 #include "zstream/ozstream_impl.hpp"
