diff --git a/gfalibs/include/output.h b/gfalibs/include/output.h
index aabbec0..20f0dc1 100644
--- a/gfalibs/include/output.h
+++ b/gfalibs/include/output.h
@@ -12,7 +12,7 @@
 #include "gfa-lines.h"
 #include "gfa.h"
 
-#include "zlib.h"
+#include <zlib.h>
 #include "zstream/zstream_common.hpp"
 #include "zstream/ozstream.hpp"
 #include "zstream/ozstream_impl.hpp"
diff --git a/gfalibs/include/stream-obj.h b/gfalibs/include/stream-obj.h
index 1faa4df..8e28bf6 100644
--- a/gfalibs/include/stream-obj.h
+++ b/gfalibs/include/stream-obj.h
@@ -2,7 +2,7 @@
 #define STREAM_OBJ_H
 
 #include <fstream>
-#include "zlib.h"
+#include <zlib.h>
 
 class membuf : public std::streambuf {
     
diff --git a/gfalibs/src/stream-obj.cpp b/gfalibs/src/stream-obj.cpp
index e9694ee..c6eca26 100644
--- a/gfalibs/src/stream-obj.cpp
+++ b/gfalibs/src/stream-obj.cpp
@@ -5,7 +5,7 @@
 
 #include "bed.h"
 #include "struct.h"
-#include "zlib.h"
+#include <zlib.h>
 #include "global.h"
 #include "log.h"
 #include "threadpool.h"
diff --git a/src/reads.cpp b/src/reads.cpp
index 866f49b..ac714cf 100644
--- a/src/reads.cpp
+++ b/src/reads.cpp
@@ -14,7 +14,7 @@
 #include "functions.h" // global functions
 #include "stream-obj.h"
 
-#include "zlib.h"
+#include <zlib.h>
 #include "zstream/zstream_common.hpp"
 #include "zstream/ozstream.hpp"
 #include "zstream/ozstream_impl.hpp"
