--- configureHomer.pl.ori	2021-07-05 16:37:13.648245955 +0200
+++ configureHomer.pl	2021-07-05 16:35:05.770994077 +0200
@@ -166,7 +166,7 @@
 	my $url = $baseURL . "homer.misc.v1.0.zip";
 	`wget -O homer.package.zip "$url"`;
 	print STDERR "\t\tUnzipping...\n";
-	`unzip -o -d "$homeDir" homer.package.zip`;
+	`unzip -o -d "$homeDir" homer.package.zip -x configureHomer.pl`;
 	`rm homer.package.zip`;
 }
 my $configFile = $homeDir . "/" . "config.txt";
@@ -432,7 +432,7 @@
 					next;
 				}
 				print STDERR "\t\tUnzipping...\n";
-				`unzip -o -d "$homeDir" homer.package.zip`;
+				`unzip -o -d "$homeDir" homer.package.zip -x configureHomer.pl`;
 				`rm homer.package.zip`;
 				$config->{$mode}->{$package} = $update->{$mode}->{$package};
 				if ($wantedVersion ne '') {
@@ -862,9 +862,9 @@
 			} elsif ($count == 2) {
 				print OUT "use warnings;\n";
 			} elsif ($count == 3) {
-				print OUT "use lib \"$homeDir/bin\";\n";
+				print OUT "use lib \'$homeDir/bin\';\n";
 			} elsif ($count == 4) {
-				print OUT 'my $homeDir = "' . $homeDir . "\";\n";
+				print OUT 'my $homeDir = \'' . $homeDir . "\';\n";
 			} else {
 				print OUT $_;
 			}
