diff --git a/recipes/homer/configureHomer.pl b/recipes/homer/configureHomer.pl
index 4f231c3f74..d3871ce9eb 100644
--- a/recipes/homer/configureHomer.pl
+++ b/recipes/homer/configureHomer.pl
@@ -164,7 +164,7 @@ if ($getFactsFlag == 1) {
 	my $url = $baseURL . "homer.misc.v1.0.zip";
 	`wget -O homer.package.zip "$url"`;
 	print STDERR "\t\tUnzipping...\n";
-	`unzip -o -d "$homeDir" homer.package.zip`;
+	`unzip -o -d "$homeDir" homer.package.zip -x configureHomer.pl`;
 	`rm homer.package.zip`;
 }
 my $configFile = $homeDir . "/" . "config.txt";
@@ -430,7 +430,7 @@ if ($listFlag || $updateFlag || $removeFlag || $installFlag) {
 					next;
 				}
 				print STDERR "\t\tUnzipping...\n";
-				`unzip -o -d "$homeDir" homer.package.zip`;
+				`unzip -o -d "$homeDir" homer.package.zip -x configureHomer.pl`;
 				`rm homer.package.zip`;
 				$config->{$mode}->{$package} = $update->{$mode}->{$package};
 				if ($wantedVersion ne '') {
@@ -867,9 +867,9 @@ sub compileSoftware {
 			} elsif ($count == 2) {
 				print OUT "use warnings;\n";
 			} elsif ($count == 3) {
-				print OUT "use lib \"$homeDir/bin\";\n";
+				print OUT "use lib \'$homeDir/bin\';\n";
 			} elsif ($count == 4) {
-				print OUT 'my $homeDir = "' . $homeDir . "\";\n";
+				print OUT 'my $homeDir = \'' . $homeDir . "\';\n";
 			} else {
 				print OUT $_;
 			}
