diff --git a/modules/io/pymod/export_mmcif_io.cc b/modules/io/pymod/export_mmcif_io.cc
index 11357b906..de2f77307 100644
--- a/modules/io/pymod/export_mmcif_io.cc
+++ b/modules/io/pymod/export_mmcif_io.cc
@@ -28,7 +28,7 @@ using namespace boost::python;
 #include <ost/io/mol/mmcif_info.hh>
 #include <ost/io/mol/star_writer.hh>
 #include <ost/io/mol/mmcif_writer.hh>
-#include <ost/io/mmcif_str.hh>
+#include <ost/io/mol/mmcif_str.hh>
 using namespace ost;
 using namespace ost::io;
 using namespace ost::mol;
diff --git a/modules/io/pymod/export_sdf_io.cc b/modules/io/pymod/export_sdf_io.cc
index 08de790e2..377f1dcb7 100644
--- a/modules/io/pymod/export_sdf_io.cc
+++ b/modules/io/pymod/export_sdf_io.cc
@@ -27,7 +27,7 @@ using namespace boost::python;
 #include <ost/io/mol/io_profile.hh>
 #include <ost/io/mol/sdf_reader.hh>
 #include <ost/io/mol/sdf_writer.hh>
-#include <ost/io/sdf_str.hh>
+#include <ost/io/mol/sdf_str.hh>
 using namespace ost;
 using namespace ost::io;
 using namespace ost::mol;
diff --git a/modules/io/pymod/wrap_io.cc b/modules/io/pymod/wrap_io.cc
index b7cade00c..846929999 100644
--- a/modules/io/pymod/wrap_io.cc
+++ b/modules/io/pymod/wrap_io.cc
@@ -32,7 +32,7 @@ using namespace boost::python;
 #include <ost/io/mol/entity_io_mae_handler.hh>
 #include <ost/io/mol/pdb_reader.hh>
 #include <ost/io/mol/dcd_io.hh>
-#include <ost/io/stereochemical_params_reader.hh>
+#include <ost/io/mol/stereochemical_params_reader.hh>
 using namespace ost;
 using namespace ost::io;
 
diff --git a/modules/io/src/mol/mmcif_str.hh b/modules/io/src/mol/mmcif_str.hh
index c348cbe9f..c26298a73 100644
--- a/modules/io/src/mol/mmcif_str.hh
+++ b/modules/io/src/mol/mmcif_str.hh
@@ -24,7 +24,7 @@
 #include <ost/mol/entity_handle.hh>
 #include <ost/io/mmcif_reader.hh>
 #include <ost/io/mol/io_profile.hh>
-#include <ost/io/mmcif_writer.hh>
+#include <ost/io/mol/mmcif_writer.hh>
 #include <ost/io/mol/omf.hh>
 
 namespace ost { namespace io {
diff --git a/modules/io/src/mol/omf.hh b/modules/io/src/mol/omf.hh
index 2fd9c0595..7f394c8a6 100644
--- a/modules/io/src/mol/omf.hh
+++ b/modules/io/src/mol/omf.hh
@@ -26,7 +26,7 @@
 #include <ost/mol/entity_handle.hh>
 #include <ost/geom/mat4.hh>
 #include <ost/io/io_exception.hh>
-#include <ost/io/mmcif_info.hh>
+#include <ost/io/mol/mmcif_info.hh>
 
 namespace ost { namespace io {
 
diff --git a/modules/io/src/mol/stereochemical_params_reader.cc b/modules/io/src/mol/stereochemical_params_reader.cc
index 3737b6ade..5c4996181 100644
--- a/modules/io/src/mol/stereochemical_params_reader.cc
+++ b/modules/io/src/mol/stereochemical_params_reader.cc
@@ -18,7 +18,7 @@
 //------------------------------------------------------------------------------
 #include <boost/filesystem/fstream.hpp>
 #include <ost/platform.hh>
-#include <ost/io/stereochemical_params_reader.hh>
+#include <ost/io/mol/stereochemical_params_reader.hh>
 
 
 namespace ost { namespace io {
diff --git a/modules/mol/alg/src/biounit.hh b/modules/mol/alg/src/biounit.hh
index 9df52c78e..9634ddf2d 100644
--- a/modules/mol/alg/src/biounit.hh
+++ b/modules/mol/alg/src/biounit.hh
@@ -20,7 +20,7 @@
 #define OST_MOL_ALG_BIOUNIT_HH
 
 #include <ost/mol/entity_handle.hh>
-#include <ost/io/mmcif_info.hh>
+#include <ost/io/mol/mmcif_info.hh>
 
 namespace ost { namespace mol { namespace alg {
 
diff --git a/modules/mol/alg/src/lddt.cc b/modules/mol/alg/src/lddt.cc
index 07c01826b..630dae738 100644
--- a/modules/mol/alg/src/lddt.cc
+++ b/modules/mol/alg/src/lddt.cc
@@ -32,7 +32,7 @@
 #include <ost/mol/alg/filter_clashes.hh>
 #include <ost/io/mol/pdb_reader.hh>
 #include <ost/io/io_exception.hh>
-#include <ost/io/stereochemical_params_reader.hh>
+#include <ost/io/mol/stereochemical_params_reader.hh>
 #include <ost/conop/conop.hh>
 #include <ost/conop/compound_lib.hh>
 #include <ost/string_ref.hh>
