diff --git a/pyproject.toml b/pyproject.toml
index 0f09cf0..01e31b8 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -28,7 +28,7 @@ networkx = { version = "^2.6.3", optional = true }
 matplotlib = { version = "^3.5.0", optional = true }
 pyinstaller = {version = "^5.6", optional = true}
 
-[tool.poetry.group.dev.dependencies]
+[tool.poetry.dev-dependencies]
 pytest = "^7.2.0"
 pytest-cov = "^4.0.0"
 pytest-forked = "^1.4.0"
