diff --git configure.ac configure.ac
index 46de9bd..53c5ea0 100644
--- configure.ac
+++ configure.ac
@@ -7,10 +7,10 @@ AC_ARG_WITH(stuff, [  --with-bam_tools_library            specify the location o
 
 # do tilde expansion on the paths given by the user and echo them back
 if test "$with_bam_tools_headers" != ""
-then 
+then
   with_bam_tools_headers=`echo "$with_bam_tools_headers" | sed "s|~|${HOME}|g"`
   echo "provided BAMTools headers directory is $with_bam_tools_headers"
-fi 
+fi
 if test "$with_bam_tools_library" != ""
 then
   with_bam_tools_library=`echo "$with_bam_tools_library" | sed "s|~|${HOME}|g"`
@@ -62,7 +62,7 @@ CXXFLAGS="$CXXFLAGS -Wl,--no-as-needed"
 if test "$BAM_LIBRARIES_DIR" != ""
 then
   LDFLAGS="$LDFLAGS -L$BAM_LIBRARIES_DIR -lbamtools"
-else 
+else
   LDFLAGS="$LDFLAGS -lbamtools"
 fi
 if test "$BAM_HEADERS_DIR" != ""
@@ -77,14 +77,14 @@ AC_LINK_IFELSE(
 LDFLAGS=$SAVED_LDFLAGS
 CXXFLAGS=$SAVED_CXXFLAGS
 
-# check for BAM tools using the default linking order (this works on 
-# MacOS, but not on Ubuntu)  
+# check for BAM tools using the default linking order (this works on
+# MacOS, but not on Ubuntu)
 SAVED_LDFLAGS=$LDFLAGS
 SAVED_CXXFLAGS=$CXXFLAGS
 if test "$BAM_LIBRARIES_DIR" != ""
 then
   LDFLAGS="$LDFLAGS -L$BAM_LIBRARIES_DIR -lbamtools"
-else 
+else
   LDFLAGS="$LDFLAGS -lbamtools"
 fi
 if test "$BAM_HEADERS_DIR" != ""
@@ -113,7 +113,7 @@ AC_LINK_IFELSE(
   [])
 LDFLAGS=$SAVED_LDFLAGS
 if test "$ac_cv_lib_gsl_matrix" != yes
-then 
+then
   echo "checking for linking against gsl matrix... no"
 else
   echo "checking for linking against gsl matrix... yes"
@@ -123,14 +123,14 @@ GSLERROR="could not find or link to GSL -- Piranha requires GSL and \
 cannot be built without it. GSL can be downloaded from \
 http://www.gnu.org/software/gsl/ -- please ensure GSL is \
 correctly installed on your system."
-if test "$ac_cv_lib_m_cos" != yes 
+if test "$ac_cv_lib_m_cos" != yes
 then
   AC_MSG_ERROR([$GSLERROR])
 fi
 if test "$ac_cv_lib_gsl_gsl_blas_dgemm" != yes
 then
   AC_MSG_ERROR([$GSLERROR])
-fi 
+fi
 if test "$ac_cv_lib_gslcblas_cblas_dgemm" != yes
 then
   AC_MSG_ERROR([$GSLERROR])
@@ -151,19 +151,21 @@ else
   echo "checking for BAMTools... yes"
   AC_SUBST(bamsupport, BAM_SUPPORT)
   AC_SUBST(bamlib, -lbamtools)
-  # only if the user told us where the BAM library was do we add it to 
+  # only if the user told us where the BAM library was do we add it to
   # the places where the executable should look for libraries.
   if test "$BAM_LIBRARIES_DIR" != ""
   then
     AC_SUBST(rpathflags, [-Wl,-rpath,`cd $BAM_LIBRARIES_DIR; pwd`])
     AC_SUBST(bamlibdir, -L$BAM_LIBRARIES_DIR)
   fi
-  # this is ok, because we don't add the -I until the makefile, where 
+  # this is ok, because we don't add the -I until the makefile, where
   # it's added to a bunch of things, so this can be empty. Should make -L
-  # and -l work like this too... 
+  # and -l work like this too...
   AC_SUBST(bamhead, $BAM_HEADERS_DIR)
 fi
 
+AC_SUBST(cxxflags, $CXXFLAGS)
+
 AC_OUTPUT(Makefile)
 AC_OUTPUT(src/Makefile)
 AC_OUTPUT(src/test/Makefile)
diff --git src/progs/Makefile.in src/progs/Makefile.in
index a1f9e6c..3925802 100644
--- src/progs/Makefile.in
+++ src/progs/Makefile.in
@@ -24,9 +24,9 @@ endif
 PROGS = Piranha Simulate
 
 CXX = g++
-CFLAGS = -Wall -fPIC -fmessage-length=50
+CFLAGS = -Wall -fPIC -fmessage-length=50 @cxxflags@
 OPTFLAGS = -O3
-DEBUGFLAGS = -g 
+DEBUGFLAGS = -g
 COMMON_DIR = ../common/
 LIBS = -lgsl -lgslcblas @bamlibdir@ @bamlib@
 
@@ -53,9 +53,9 @@ all:    $(PROGS)
 
 %.o: %.cpp %.hpp
 	$(CXX) $(CFLAGS) -c -o $@ $< $(INCLUDEARGS)
-	
+
 %.o: %.cpp %.h
-	$(CXX) $(CFLAGS) -c -o $@ $< $(INCLUDEARGS)	
+	$(CXX) $(CFLAGS) -c -o $@ $< $(INCLUDEARGS)
 
 %: %.cpp
 	$(CXX) $(CFLAGS) @rpathflags@ -o $@ $^ $(INCLUDEARGS) $(LIBS)
@@ -87,7 +87,6 @@ clean:
 distclean : clean
 	@rm Makefile
 .PHONY : distclean
-	
+
 test:
 	make -C ../test SMITHLAB_CPP=$(SMITHLAB_CPP) test; \
-
