diff --git a/src/parasail.h b/src/parasail.h
index 950a4a5..6d58ba2 100644
--- a/src/parasail.h
+++ b/src/parasail.h
@@ -7,7 +7,15 @@ Author jeffrey.daily@gmail.com
 Copyright (c) 2015 Battelle Memorial Institute.
 ***/
 
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+#ifdef __ARM_NEON
+#include "sse2neon.h"
+#else
 #include <immintrin.h>
+#endif
+
 /*                                            3         2         1          */
 /*                                           10987654321098765432109876543210*/
 #define PARASAIL_FLAG_NW          (1 << 0) /*00000000000000000000000000000001*/
