--- trinity-plugins/RSEM-mod/makefile	2017-02-16 15:35:26.000000000 -0500
+++ trinity-plugins/RSEM-mod/makefile	2017-02-15 19:06:41.000000000 -0500
@@ -52,7 +52,7 @@
 
 
 rsem-parse-alignments : parseIt.o sam/libbam.a
-	$(CC) -o rsem-parse-alignments parseIt.o sam/libbam.a -lz 
+	$(CC) -o rsem-parse-alignments parseIt.o sam/libbam.a -lz -Wl,-headerpad_max_install_names
 
 parseIt.o : utils.h GroupInfo.h Read.h SingleRead.h SingleReadQ.h PairedEndRead.h PairedEndReadQ.h SingleHit.h PairedEndHit.h HitContainer.h SamParser.h RefSeq.h Refs.h sam/sam.h sam/bam.h parseIt.cpp
 	$(CC) $(COFLAGS) parseIt.cpp
