--- trinity-plugins/RSEM-mod/sam/Makefile	2016-06-27 15:59:27.000000000 -0400
+++ trinity-plugins/RSEM-mod/sam/Makefile	2016-06-27 16:00:05.000000000 -0400
@@ -1,6 +1,6 @@
 CC=			gcc
-CFLAGS=		-g -Wall -O2 #-m64 #-arch ppc
-DFLAGS=		-D_FILE_OFFSET_BITS=64 -D_USE_KNETFILE -D_CURSES_LIB=1
+CFLAGS=		-Wall
+DFLAGS=		-D_FILE_OFFSET_BITS=64 -D_USE_KNETFILE -D_CURSES_LIB=0
 KNETFILE_O=	knetfile.o
 LOBJS=		bgzf.o kstring.o bam_aux.o bam.o bam_import.o sam.o bam_index.o	\
 			bam_pileup.o bam_lpileup.o bam_md.o glf.o razf.o faidx.o \
@@ -12,7 +12,6 @@
 INCLUDES=	-I.
 SUBDIRS=	. bcftools misc
 LIBPATH=
-LIBCURSES=	-lcurses # -lXCurses
 
 .SUFFIXES:.c .o
 
