--- Trinity.pl	2017-09-11 12:53:54.000000000 -0400
+++ Trinity.pl	2017-09-11 12:54:03.000000000 -0400
@@ -206,13 +206,6 @@
 }
 
 
-## Check Java version:
-my $java_version = `java -version 2>&1 `;
-unless ($java_version =~ /java version \"1\.6\./) {
-    die "Error, Trinity requires access to Java version 1.6.  Currently installed version is: $java_version";
-}
-
-
 if ($USE_MERYL && ! -e "$MERYL_DIR/meryl") {
     die "Error, cannot find meryl at $MERYL_DIR/meryl";
 }
