--- Trinity.pl	2011-11-28 09:14:58.000000000 -0500
+++ Trinity.pl	2015-12-15 16:54:00.000000000 -0500
@@ -208,7 +208,8 @@
 
 ## Check Java version:
 my $java_version = `java -version 2>&1 `;
-unless ($java_version =~ /java version \"1\.6\./) {
+$java_version =~ /^(?:openjdk|java) version "1\.(\d+)\./;
+unless ($1 >= 6) {
     die "Error, Trinity requires access to Java version 1.6.  Currently installed version is: $java_version";
 }
 
