--- src/ssw_lib.py	2019-06-10 13:10:39.667611674 +0200
+++ src/ssw_lib.py	2019-06-10 13:11:08.451687753 +0200
@@ -101,8 +101,8 @@
         """
 # load libssw
         sLibName = 'libssw.so'
-        if not sLibPath:
-# user doesn't give the path explicitly
+        if sLibPath:
+# user does give the path explicitly
             if not op.exists(op.join(sLibPath, sLibName)):
                 print >> sys.stderr, 'libssw.so does not exist in the input path'
                 sys.exit()
