diff --git a/unicycler/misc.py b/unicycler/misc.py
index d202e3d..b3ee0ab 100644
--- a/unicycler/misc.py
+++ b/unicycler/misc.py
@@ -767,24 +767,14 @@ def get_left_arrow():
     """
     This function returns either a Unicode left arrow or '<-', depending on the system encoding.
     """
-    try:
-        '\u2190'.encode(sys.stdout.encoding)
-    except (AttributeError, UnicodeEncodeError):
-        return '<-'
-    else:
-        return '\u2190 '
+    return '<-'


 def get_right_arrow():
     """
     This function returns either a Unicode right arrow or '->', depending on the system encoding.
     """
-    try:
-        '\u2192'.encode(sys.stdout.encoding)
-    except (AttributeError, UnicodeEncodeError):
-        return '->'
-    else:
-        return '\u2192'
+    return '->'


 def get_default_thread_count():
