--- c++/scripts/common/impl/create_flat_makefile.sh	2025-03-27 09:48:27
+++ c++/scripts/common/impl/create_flat_makefile.sh	2025-03-27 09:49:47
@@ -190,7 +190,10 @@
   for dep in $ptbdep; do
     COMMON_Exec cd $builddir
     COMMON_Exec cd $dep
-    COMMON_Exec make
+    COMMON_Exec echo $PWD
+    COMMON_Exec echo building..
+    COMMON_Exec ls -l
+    COMMON_Exec make --verbose
   done
   COMMON_Exec cd $builddir
   ptb="./build-system/project_tree_builder/$ptbname"
@@ -300,7 +303,7 @@
       echo "Generating objects source code. Please wait."
       echo "**********************************************************************"
       echo make -f $solution all_files
-      make -f $solution all_files >/dev/null 2>&1
+      make  -f $solution all_files >/dev/null 2>&1
     fi
   fi
 fi
