--- src/common/normbase.hpp
+++ src/common/normbase.hpp
@@ -142,7 +142,7 @@ private:

 std::ostream & operator << (std::ostream &os, const CSimpleException &e);

-#define THROW_SIMPLE(a) throw CSimpleException(__FILE__, __LINE__, dynamic_cast<const stringstream& > (stringstream() << a))
+#define THROW_SIMPLE(a) throw CSimpleException(__FILE__, __LINE__, static_cast<const stringstream& > (stringstream() << a))
