--- c++/scripts/projects/project_tree_builder/Manifest	2025-03-20 12:06:58
+++ c++/scripts/projects/project_tree_builder/Manifest	2025-03-20 12:07:34
@@ -33,7 +33,7 @@
 
 Win64_19   : static-64 : static 64 ReleaseMT
 
-XCode      : plain  : Xcode.sh 30 --without-debug --without-ncbi-c --without-pcre
+XCode      : plain  : Xcode.sh 30 --without-debug -mmacos-version-min=11.0 --without-ncbi-c --without-pcre
 
 USE_COMPONENTS
 
