--- c++/scripts/common/impl/create_flat_makefile.sh	2023-03-21 19:27:39
+++ c++/scripts/common/impl/create_flat_makefile.sh	2025-03-27 16:17:07
@@ -190,7 +190,8 @@
   for dep in $ptbdep; do
     COMMON_Exec cd $builddir
     COMMON_Exec cd $dep
-    COMMON_Exec make
+
+    COMMON_Exec ${MAKE} CC="${CC}" CXX="${CXX}" CFLAGS="${CFLAGS}" CXXFLAGS="${CXXFLAGS}"
   done
   COMMON_Exec cd $builddir
   ptb="./build-system/project_tree_builder/$ptbname"
@@ -299,8 +300,8 @@
       echo "**********************************************************************"
       echo "Generating objects source code. Please wait."
       echo "**********************************************************************"
-      echo make -f $solution all_files
-      make -f $solution all_files >/dev/null 2>&1
+      echo make -f $solution all_files 
+      ${MAKE}  -f $solution all_files  CC="${CC}" CXX="${CXX}" CFLAGS="${CFLAGS}" CXXFLAGS="${CXXFLAGS}" >/dev/null 2>&1
     fi
   fi
 fi
