--- c++/scripts/common/impl/create_flat_makefile.sh	2025-03-27 09:48:27
+++ c++/scripts/common/impl/create_flat_makefile.sh	2025-03-27 10:29:12
@@ -190,7 +190,13 @@
   for dep in $ptbdep; do
     COMMON_Exec cd $builddir
     COMMON_Exec cd $dep
-    COMMON_Exec make
+    COMMON_Exec echo $PWD
+    COMMON_Exec echo building..
+    COMMON_Exec ls -l
+    COMMON_Exec cat Makefile
+    COMMON_Exec which make 
+
+    COMMON_Exec ${MAKE}
   done
   COMMON_Exec cd $builddir
   ptb="./build-system/project_tree_builder/$ptbname"
