diff --git a/Makefile_full b/Makefile_full
index 00e7ebc..0392065 100644
--- a/Makefile_full
+++ b/Makefile_full
@@ -1,11 +1,11 @@
-CXXFLAGS =  -Wall -std=c++11 -I$${HOME}/include
-LDFLAGS  =  -L$${HOME}/lib -lroaring -lsdsl -ljellyfish-2.0 -lpthread
+CXXFLAGS =  -Wall -std=c++14 -I$(PREFIX)/include -I$(PREFIX)/include/jellyfish-2.3.1
+LDFLAGS  =  -L$(PREFIX)/lib -lroaring -lsdsl -ljellyfish-2.0 -pthread $(PREFIX)/lib/libjellyfish-2.0.a
 CXXFLAGS += -DincludeSecondaryCommands
 
 CPP_FILES := $(wildcard *.cc)
 OBJ_FILES := $(addprefix ./,$(notdir $(CPP_FILES:.cc=.o)))
 
-all:   CXXFLAGS += -DNDEBUG -O3 
+all:   CXXFLAGS += -DNDEBUG -O3
 
 all:   clean howdesbt
 
@@ -18,5 +18,5 @@ howdesbt: $(OBJ_FILES)
 clean: cleano
 	rm -f howdesbt
 
-cleano: 
+cleano:
 	rm -f *.o
