diff --git a/Cargo.toml b/Cargo.toml
index 6da03b7..01e8e06 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -39,7 +39,7 @@ lru_time_cache = "0.11.1"
 num_cpus = "1.13.0"
 rayon = "1.4.0"
 #TODO add  features = ["libdeflate"] when https://github.com/rust-bio/rust-htslib/pull/341 is merged
-rust-htslib = {version = "0.39"}
+rust-htslib = {version = "0.49.0"}
 rust-lapper = "1.0"
 serde = { version = "1.0.116", features = ["derive"] }
 smartstring = { version = "1.0.1", features = ["serde"] }
