*** TERefiner/Makefile	2019-10-08 11:10:12.000000000 +0300
--- TERefiner/Makefile	2019-10-09 11:22:19.000000000 +0300
***************
*** 1,47 ****
! BAMTOOLS=../bamtools-master/include
! BAMTOOLS_LD=../bamtools-master/lib
  
! CC = g++
! CFLAGS =  -O3 -Wall -static -I$(BAMTOOLS) -L$(BAMTOOLS_LD) -Wl,-rpath,$(BAMTOOLS_LD)
  
  all: TERefiner
  
  public_func.o :	public_func.cpp public_func.h
! 	$(CC) $(CFLAGS) -c public_func.cpp
  
  StrOperation.o: StrOperation.cpp StrOperation.h
! 	$(CC) $(CFLAGS) -c StrOperation.cpp
  
  Alignment.o: Alignment.cpp Alignment.h
! 	$(CC) $(CFLAGS) -c Alignment.cpp
  	
  bam_parse.o : bam_parse.cpp bam_parse.h
! 	$(CC) $(CFLAGS) -c bam_parse.cpp
  
  local_alignment.o: ./algorithms/local_alignment.cpp ./algorithms/local_alignment.h
! 	$(CC) $(CFLAGS) -c ./algorithms/local_alignment.cpp
  
  fai_parser.o : fai_parser.cpp fai_parser.h
! 	$(CC) $(CFLAGS) -c fai_parser.cpp
  
  contigs.o : contigs.cpp contigs.h
! 	$(CC) $(CFLAGS) -c contigs.cpp
  
  Coverage.o : Coverage.cpp Coverage.h
! 	$(CC) $(CFLAGS) -c Coverage.cpp
  
  scaffolding.o: scaffolding.cpp scaffolding.h
! 	$(CC) $(CFLAGS) -c scaffolding.cpp
  
  RepeatsClassifier.o: RepeatsClassifier.cpp RepeatsClassifier.h
! 	$(CC) $(CFLAGS) -c RepeatsClassifier.cpp
  
  refiner.o: refiner.cpp refiner.h
! 	$(CC) $(CFLAGS) -c refiner.cpp
  
  main.o: main.cpp 
! 	$(CC) $(CFLAGS) -c main.cpp
  
  TERefiner: public_func.o StrOperation.o local_alignment.o Alignment.o bam_parse.o fai_parser.o contigs.o Coverage.o scaffolding.o RepeatsClassifier.o refiner.o main.o 
! 	$(CC) $(CFLAGS) -o TERefiner_1 public_func.o StrOperation.o local_alignment.o Alignment.o bam_parse.o fai_parser.o contigs.o Coverage.o scaffolding.o RepeatsClassifier.o refiner.o main.o \
! 	-lbamtools -lz -lm
\ No newline at end of file
--- 1,48 ----
! BAMTOOLS=$(PREFIX)/include/bamtools
! BAMTOOLS_LD=$(PREFIX)/lib
  
! #CXX = $(CXX) 
! CFLAGS   = -g -Wall -O2 -I./ -I$(BAMTOOLS) -L$(PREFIX)/lib -L$(BAMTOOLS_LD) -I$(PREFIX)/include  -std=c++11
! #CFLAGS = -g -static -O3 -Wall -I./ -I$(BAMTOOLS) -L$(PREFIX)/lib -L$(BAMTOOLS_LD) -I$(PREFIX)/include -Wl,-rpath,$(BAMTOOLS_LD) -std=c++11 -lbamtools
  
  all: TERefiner
  
  public_func.o :	public_func.cpp public_func.h
! 	$(CXX) $(CFLAGS) -c public_func.cpp
  
  StrOperation.o: StrOperation.cpp StrOperation.h
! 	$(CXX) $(CFLAGS) -c StrOperation.cpp
  
  Alignment.o: Alignment.cpp Alignment.h
! 	$(CXX) $(CFLAGS) -c Alignment.cpp
  	
  bam_parse.o : bam_parse.cpp bam_parse.h
! 	$(CXX) $(CFLAGS) -c bam_parse.cpp
  
  local_alignment.o: ./algorithms/local_alignment.cpp ./algorithms/local_alignment.h
! 	$(CXX) $(CFLAGS) -c ./algorithms/local_alignment.cpp
  
  fai_parser.o : fai_parser.cpp fai_parser.h
! 	$(CXX) $(CFLAGS) -c fai_parser.cpp
  
  contigs.o : contigs.cpp contigs.h
! 	$(CXX) $(CFLAGS) -c contigs.cpp
  
  Coverage.o : Coverage.cpp Coverage.h
! 	$(CXX) $(CFLAGS) -c Coverage.cpp
  
  scaffolding.o: scaffolding.cpp scaffolding.h
! 	$(CXX) $(CFLAGS) -c scaffolding.cpp
  
  RepeatsClassifier.o: RepeatsClassifier.cpp RepeatsClassifier.h
! 	$(CXX) $(CFLAGS) -c RepeatsClassifier.cpp
  
  refiner.o: refiner.cpp refiner.h
! 	$(CXX) $(CFLAGS) -c refiner.cpp
  
  main.o: main.cpp 
! 	$(CXX) $(CFLAGS) -c main.cpp
  
  TERefiner: public_func.o StrOperation.o local_alignment.o Alignment.o bam_parse.o fai_parser.o contigs.o Coverage.o scaffolding.o RepeatsClassifier.o refiner.o main.o 
! 	$(CXX) $(CFLAGS) $(CXXFLAGS) $(CPPFLAGS) $(LDFLAGS) -o TERefiner_1 public_func.o StrOperation.o local_alignment.o Alignment.o bam_parse.o fai_parser.o contigs.o Coverage.o scaffolding.o RepeatsClassifier.o refiner.o main.o \
! 	-lbamtools -lz -lm 
