--- RepEnrich-1.2/RepEnrich.py	2015-07-23 17:47:18.000000000 +0000
+++ RepEnrich-1.2/RepEnrich.py.new	2017-12-12 03:52:26.540367029 +0000
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/opt/anaconda1anaconda2anaconda3/bin/python
 import argparse
 import csv
 import numpy
@@ -384,4 +384,4 @@
 if os.path.exists(outputfolder + os.path.sep + 'sorted_bowtie'):
 	shutil.rmtree(outputfolder + os.path.sep + 'sorted_bowtie')
 
-print "... Done"
\ No newline at end of file
+print "... Done"
--- RepEnrich-1.2/RepEnrich_setup.py	2015-07-23 17:47:18.000000000 +0000
+++ RepEnrich-1.2/RepEnrich_setup.py.new	2017-12-12 03:52:58.747633551 +0000
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/opt/anaconda1anaconda2anaconda3/bin/python
 import argparse
 import csv
 import os
