diff --git a/probeit.py b/probeit.py
index 443cb47..d9355d4 100644
--- a/probeit.py
+++ b/probeit.py
@@ -242,10 +242,8 @@ class ProbeitUtils:
 
     @classmethod
     def setCover(cls, coverage, length, eStop, dist, reps, mapCSV, genome):
-        filePath = os.path.sep.join(os.path.realpath(__file__).split(os.path.sep)[:-1])
-        setcoverPath = '{}{}{}{}{}'.format(filePath, os.path.sep, 'setcover', os.path.sep, 'setcover')
-        command = " -c {} -l {} -p {} -d {} -i {} {} {}".format(coverage, length, eStop, dist, reps, mapCSV, genome)
-        stdOut, stdErr = cls.runCommand(setcoverPath + command, verbose=True)
+        command = "setcover -c {} -l {} -p {} -d {} -i {} {} {}".format(coverage, length, eStop, dist, reps, mapCSV, genome)
+        stdOut, stdErr = cls.runCommand(command, verbose=True)
         return stdOut, stdErr
 
     @classmethod
