diff --git a/probeit/classmodule.py b/probeit/classmodule.py

index 99fc3fa..c18c4c4 100644
--- a/probeit/classmodule.py
+++ b/probeit/classmodule.py
@@ -244,10 +244,8 @@ class ProbeitUtils:
 
     @classmethod
     def setCover(cls, coverage, length, eStop, dist, reps, mapCSV, genome):
-        filePath = os.path.sep.join(os.path.realpath(__file__).split(os.path.sep)[:-1])
-        setcoverPath = '{}{}{}{}{}'.format(filePath, os.path.sep, 'setcover', os.path.sep, 'setcover')
-        command = " -c {} -l {} -p {} -d {} -i {} {} {}".format(coverage, length, eStop, dist, reps, mapCSV, genome)
-        stdOut, stdErr = cls.runCommand(setcoverPath + command, verbose=True)
+        command = "setcover -c {} -l {} -p {} -d {} -i {} {} {}".format(coverage, length, eStop, dist, reps, mapCSV, genome)
+        stdOut, stdErr = cls.runCommand(command, verbose=True)
         return stdOut, stdErr
 
