--- bgzf.c
+++ bgzf.c
@@ -73,7 +73,7 @@
 static const int Z_DEFAULT_MEM_LEVEL = 8;
 
 
-inline
+static inline
 void
 packInt16(uint8_t* buffer, uint16_t value)
 {
@@ -81,14 +81,14 @@
     buffer[1] = value >> 8;
 }
 
-inline
+static inline
 int
 unpackInt16(const uint8_t* buffer)
 {
     return (buffer[0] | (buffer[1] << 8));
 }
 
-inline
+static inline
 void
 packInt32(uint8_t* buffer, uint32_t value)
 {
