--- src/bamstats.h
+++ src/bamstats.h
@@ -412,7 +412,7 @@
 	if (itRg->second.rc.brange[refIndex].find(psId) == itRg->second.rc.brange[refIndex].end()) {
 	  itRg->second.rc.brange[refIndex].insert(std::make_pair(psId, std::make_pair(rec->core.pos, lastAlignedPosition(rec))));
 	} else {
-	  itRg->second.rc.brange[refIndex][psId].first = std::min(rec->core.pos, itRg->second.rc.brange[refIndex][psId].first);
+	  itRg->second.rc.brange[refIndex][psId].first = std::min((int32_t) rec->core.pos, itRg->second.rc.brange[refIndex][psId].first);
 	  itRg->second.rc.brange[refIndex][psId].second = std::max((int32_t) lastAlignedPosition(rec), itRg->second.rc.brange[refIndex][psId].second);
 	}
       }
