diff -ruN beagle2chromopainter.pl beagle2chromopainter.pl
--- beagle2chromopainter.pl     2020-05-22 17:52:34.000000000 +0200
+++ beagle2chromopainter.pl     2023-09-05 17:27:54.477603175 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 ## CONVERTS PHASED BEAGLE OUTPUT TO CHROMOPAINTER-STYLE INPUT FILES

 sub help {
diff -ruN chromopainter2chromopainterv2.pl chromopainter2chromopainterv2.pl
--- chromopainter2chromopainterv2.pl	2020-05-22 17:52:34.000000000 +0200
+++ chromopainter2chromopainterv2.pl	2023-09-08 16:23:44.121888309 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 ## CONVERTS PHASE (CHROMOPAINTER) FORMAT TO (CHROMOPAINTER)V2 
 use strict;
 use warnings;
diff -ruN convertrecfile.pl convertrecfile.pl
--- convertrecfile.pl   2020-05-22 17:52:34.000000000 +0200
+++ convertrecfile.pl   2023-09-05 17:27:54.477603175 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 #   This program is distributed in the hope that it will be useful,
 #   but WITHOUT ANY WARRANTY; without even the implied warranty of
 #   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
diff -ruN impute2chromopainter.pl impute2chromopainter.pl
--- impute2chromopainter.pl     2020-05-22 17:52:34.000000000 +0200
+++ impute2chromopainter.pl     2023-09-05 17:27:54.477603175 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 ## CONVERTS PHASED IMPUTE2 OUTPUT TO CHROMOPAINTER-STYLE INPUT FILES

 sub help {
diff -ruN makeuniformrecfile.pl makeuniformrecfile.pl
--- makeuniformrecfile.pl       2020-05-22 17:52:34.000000000 +0200
+++ makeuniformrecfile.pl       2023-09-05 17:27:54.477603175 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 #
 use strict;
 use warnings;
diff -ruN msms2cp.pl msms2cp.pl
--- msms2cp.pl  2020-05-22 17:52:34.000000000 +0200
+++ msms2cp.pl  2023-09-05 17:27:54.477603175 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 ## CONVERTS PHASED IMPUTE2 OUTPUT TO CHROMOPAINTER-STYLE INPUT FILES
 use Switch;
 use strict;
diff -ruN phasescreen.pl phasescreen.pl
--- phasescreen.pl      2020-05-22 17:52:34.000000000 +0200
+++ phasescreen.pl      2023-09-05 17:27:54.477603175 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 ## REMOVE SINGLETONS OR NON-SNPS FROM PHASE DATA
 use strict;
 use warnings;
diff -ruN phasesubsample.pl phasesubsample.pl
--- phasesubsample.pl   2020-05-22 17:52:34.000000000 +0200
+++ phasesubsample.pl   2023-09-05 17:27:54.477603175 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 ## CONVERTS PHASE (CHROMOPAINTER) FORMAT TO BEAGLE FORMAT
 use strict;
 use warnings;
diff -ruN plink2chromopainter.pl plink2chromopainter.pl
--- plink2chromopainter.pl      2020-05-22 17:52:34.000000000 +0200
+++ plink2chromopainter.pl      2023-09-05 17:27:54.477603175 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 #
 #
 #
