diff --git a/setup.py b/setup.py
index 05d2d92..05d807c 100644
--- a/setup.py
+++ b/setup.py
@@ -29,9 +29,5 @@ setup(
 	"Intended Audience :: Science/Research",
         "License :: OSI Approved :: BSD License",
         "Programming Language :: Python",
-    ],
-    install_requires=['nibabel', 'numpy>=1.7.1', 'six>=1.9.0', 'nose>=1.3.0', 'theano>=0.8.0', 'pp'],
-    dependency_links=[
-	"http://www.parallelpython.com/downloads/pp/pp-1.6.4.tar.gz",
     ]
 )
