--- a/lib/AGAT/AGAT.pm
+++ b/lib/AGAT/AGAT.pm
@@ -19,7 +19,7 @@ use AGAT::Utilities;
 use AGAT::PlotR;
 use Bio::Tools::GFF;
 
-our $VERSION     = "v2.0.0";
+our $VERSION     = "v1.6.1";
 our @ISA         = qw(Exporter);
 our @EXPORT      = qw(get_agat_header print_agat_version initialize_agat handle_levels create_log_file parse_shared_options apply_shared_options split_argv_shared_vs_script);
 sub import {