--- gmvalue	2015-03-05 23:15:26.000000000 -0600
+++ gmvalue.new	2019-03-28 01:47:36.566885232 -0500
@@ -62,11 +62,11 @@
 }
 
 if ($command eq 'contig'){
-    system ("$Bin/Utility_gmvalue/Nucmer_contig_validate.pl $options");
+    system ("$Bin/Nucmer_contig_validate.pl $options");
 }
 elsif ($command eq 'subcon'){
-    system ("$Bin/Utility_gmvalue/Nucmer_subcontig_validate.pl $options");
+    system ("$Bin/Nucmer_subcontig_validate.pl $options");
 }
 elsif ($command eq 'scaf'){
-    system ("$Bin/Utility_gmvalue/Nucmer_scaf_validate.pl $options");
+    system ("$Bin/Nucmer_scaf_validate.pl $options");
 }
