--- setup.py.bak	2017-07-04 18:15:19.000000000 +0200
+++ setup.py	2017-07-04 19:16:48.000000000 +0200
@@ -31,9 +31,7 @@
 
 if sys.platform == 'darwin':              # Mac OS X?
     # force 64bit only builds
-    EXTRA_COMPILE_ARGS.extend(['-arch', 'x86_64', '-mmacosx-version-min=10.7',
-                               '-stdlib=libc++'])
-
+    EXTRA_COMPILE_ARGS.extend(['-arch', 'x86_64'])
 else:                                     # ...likely Linux
    if os.environ.get('SOURMASH_COVERAGE'):
       print('Turning on coverage analysis.')
