diff -u -r a/src/bwa_sam_converter.pl b/src/bwa_sam_converter.pl
--- a/src/bwa_sam_converter.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/bwa_sam_converter.pl	2017-01-09 01:22:33.064966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 use strict;
 use Getopt::Std;
 
diff -u -r a/src/clip_adapters.pl b/src/clip_adapters.pl
--- a/src/clip_adapters.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/clip_adapters.pl	2017-01-09 01:22:57.180966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/collapse_reads_md.pl b/src/collapse_reads_md.pl
--- a/src/collapse_reads_md.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/collapse_reads_md.pl	2017-01-09 01:23:03.948966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/convert_bowtie_output.pl b/src/convert_bowtie_output.pl
--- a/src/convert_bowtie_output.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/convert_bowtie_output.pl	2017-01-09 01:23:10.816966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use strict;
 
diff -u -r a/src/excise_precursors_iterative_final.pl b/src/excise_precursors_iterative_final.pl
--- a/src/excise_precursors_iterative_final.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/excise_precursors_iterative_final.pl	2017-01-09 01:23:21.816966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/excise_precursors.pl b/src/excise_precursors.pl
--- a/src/excise_precursors.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/excise_precursors.pl	2017-01-09 01:23:16.268966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/extract_miRNAs.pl b/src/extract_miRNAs.pl
--- a/src/extract_miRNAs.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/extract_miRNAs.pl	2017-01-09 01:23:27.252966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use strict;
 
diff -u -r a/src/fastaparse.pl b/src/fastaparse.pl
--- a/src/fastaparse.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/fastaparse.pl	2017-01-09 01:23:37.904966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/fastaselect.pl b/src/fastaselect.pl
--- a/src/fastaselect.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/fastaselect.pl	2017-01-09 01:23:43.516966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/fastq2fasta.pl b/src/fastq2fasta.pl
--- a/src/fastq2fasta.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/fastq2fasta.pl	2017-01-09 01:23:48.996966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use strict;
 
diff -u -r a/src/find_read_count.pl b/src/find_read_count.pl
--- a/src/find_read_count.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/find_read_count.pl	2017-01-09 01:23:56.996966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/geo2fasta.pl b/src/geo2fasta.pl
--- a/src/geo2fasta.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/geo2fasta.pl	2017-01-09 01:24:03.192966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/get_mirdeep2_precursors.pl b/src/get_mirdeep2_precursors.pl
--- a/src/get_mirdeep2_precursors.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/get_mirdeep2_precursors.pl	2017-01-09 01:24:08.972966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 #######################################################################
 #
diff -u -r a/src/illumina_to_fasta.pl b/src/illumina_to_fasta.pl
--- a/src/illumina_to_fasta.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/illumina_to_fasta.pl	2017-01-09 01:24:17.336966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/make_html2.pl b/src/make_html2.pl
--- a/src/make_html2.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/make_html2.pl	2017-01-09 01:24:32.948966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 
 use strict;
diff -u -r a/src/make_html.pl b/src/make_html.pl
--- a/src/make_html.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/make_html.pl	2017-01-09 01:24:26.060966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 
 use strict;
diff -u -r a/src/mapper.pl b/src/mapper.pl
--- a/src/mapper.pl	2016-04-21 11:29:52.000000000 +0200
+++ b/src/mapper.pl	2017-01-09 01:24:53.660966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 ## last update 21/04/2016
 
diff -u -r a/src/mirdeep2bed.pl b/src/mirdeep2bed.pl
--- a/src/mirdeep2bed.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/mirdeep2bed.pl	2017-01-09 01:25:00.000966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl 
+#!/usr/bin/env perl
 
 use strict;
 
diff -u -r a/src/miRDeep2_core_algorithm.pl b/src/miRDeep2_core_algorithm.pl
--- a/src/miRDeep2_core_algorithm.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/miRDeep2_core_algorithm.pl	2017-01-09 01:25:12.540966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/miRDeep2.pl b/src/miRDeep2.pl
--- a/src/miRDeep2.pl	2016-04-21 11:24:32.000000000 +0200
+++ b/src/miRDeep2.pl	2017-01-09 02:24:39.440966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
@@ -10,9 +10,6 @@
 use File::Basename;
 use Term::ANSIColor;
 
-##
-check_install();
-##
 ## generate log file for run, all output will be printed to it
 
 my $version="2.0.0.8";
@@ -939,7 +936,7 @@
 
     if($ARGV[0] eq '-u'){system("make_html.pl -u -y 1"); exit;}
 
-    if($ARGV[0] eq '-h' or $ARGV[0] eq '--help'){die $usage;}
+    if($ARGV[0] eq '-h' or $ARGV[0] eq '--help'){print $usage; exit;}
 
     return;
 }
@@ -1098,18 +1095,3 @@
 	return $l;
 }
 
-
-sub check_install{
-	my $a=`which miRDeep2.pl`;
-	my $bn=`dirname $a`;
-	chomp $bn;
-	if(not -f "$bn/../install_successful"){
-		die "Please run the install.pl script first before using the miRDeep2 package
-		The install script is located in ",substr($bn,0,length($bn)-3)," so just do
-
-		cd ",substr($bn,0,length($bn)-3),
-		"\nperl install.pl
-
-		";
-	}
-}
diff -u -r a/src/parse_mappings.pl b/src/parse_mappings.pl
--- a/src/parse_mappings.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/parse_mappings.pl	2017-01-09 01:26:02.016966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/perform_controls.pl b/src/perform_controls.pl
--- a/src/perform_controls.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/perform_controls.pl	2017-01-09 01:26:07.952966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/permute_structure.pl b/src/permute_structure.pl
--- a/src/permute_structure.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/permute_structure.pl	2017-01-09 01:26:13.888966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/prepare_signature.pl b/src/prepare_signature.pl
--- a/src/prepare_signature.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/prepare_signature.pl	2017-01-09 01:26:19.300966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/quantifier.pl b/src/quantifier.pl
--- a/src/quantifier.pl	2016-04-21 11:25:02.000000000 +0200
+++ b/src/quantifier.pl	2017-01-09 01:26:25.052966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 ######################
 
diff -u -r a/src/remove_white_space_in_id.pl b/src/remove_white_space_in_id.pl
--- a/src/remove_white_space_in_id.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/remove_white_space_in_id.pl	2017-01-09 01:26:30.364966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/rna2dna.pl b/src/rna2dna.pl
--- a/src/rna2dna.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/rna2dna.pl	2017-01-09 01:26:37.692966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/samFLAGinfo.pl b/src/samFLAGinfo.pl
--- a/src/samFLAGinfo.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/samFLAGinfo.pl	2017-01-09 01:26:45.736966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 
 ## convert decimal number to binary and then to hexadecimal to get read FLAG identification 
diff -u -r a/src/sam_reads_collapse.pl b/src/sam_reads_collapse.pl
--- a/src/sam_reads_collapse.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/sam_reads_collapse.pl	2017-01-09 01:26:50.964966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 
 if(not $ARGV[1]){
diff -u -r a/src/sanity_check_genome.pl b/src/sanity_check_genome.pl
--- a/src/sanity_check_genome.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/sanity_check_genome.pl	2017-01-09 01:27:02.940966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl -W
+#!/usr/bin/env perl -W
 
 use strict;
 
diff -u -r a/src/sanity_check_mapping_file.pl b/src/sanity_check_mapping_file.pl
--- a/src/sanity_check_mapping_file.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/sanity_check_mapping_file.pl	2017-01-09 01:27:11.132966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl -W
+#!/usr/bin/env perl -W
 
 use strict;
 my $rid;
diff -u -r a/src/sanity_check_mature_ref.pl b/src/sanity_check_mature_ref.pl
--- a/src/sanity_check_mature_ref.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/sanity_check_mature_ref.pl	2017-01-09 01:27:17.288966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl -W
+#!/usr/bin/env perl -W
 
 use strict;
 
diff -u -r a/src/sanity_check_reads_ready_file.pl b/src/sanity_check_reads_ready_file.pl
--- a/src/sanity_check_reads_ready_file.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/sanity_check_reads_ready_file.pl	2017-01-09 01:27:24.576966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl -W
+#!/usr/bin/env perl -W
 
 use strict;
 
diff -u -r a/src/select_for_randfold.pl b/src/select_for_randfold.pl
--- a/src/select_for_randfold.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/select_for_randfold.pl	2017-01-09 01:27:30.496966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/survey.pl b/src/survey.pl
--- a/src/survey.pl	2016-04-20 14:27:48.000000000 +0200
+++ b/src/survey.pl	2017-01-09 01:27:38.132966271 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
