diff --git a/util/support_scripts/trinity_installer.py b/util/support_scripts/trinity_installer.py
index 1d82e63..3ebffcd 100755
--- a/util/support_scripts/trinity_installer.py
+++ b/util/support_scripts/trinity_installer.py
@@ -8,7 +8,7 @@ print("Trinity package dir: {}".format(trinity_package_dir))
 trinity_package_name = os.path.basename(trinity_package_dir)
 print("Trinity package name: {}".format(trinity_package_name))
 
-destination_package_dir = "/usr/local/bin"
+destination_package_dir = os.getenv('PREFIX') + "/bin"
 
 subprocess.check_call("rsync -av --exclude='.*' {}/ {}".format(trinity_package_dir, destination_package_dir), shell=True)
 
