--- Inchworm/CMakeLists.txt.orig	2025-02-24 18:37:08.778178215 +0000
+++ Inchworm/CMakeLists.txt	2025-02-24 18:37:16.638095965 +0000
@@ -13,7 +13,7 @@
     set(CMAKE_BUILD_TYPE "${default_build_type}")
 endif()
 
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pipe -W -Wall -Wpedantic -fopenmp -m64")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pipe -W -Wall -Wpedantic -fopenmp")
 
 set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O2 -DNDEBUG ")
 set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -g -O0")
--- Chrysalis/CMakeLists.txt.orig	2025-02-24 18:47:03.241952176 +0000
+++ Chrysalis/CMakeLists.txt	2025-02-24 18:47:12.251857513 +0000
@@ -13,7 +13,7 @@
     set(CMAKE_BUILD_TYPE "${default_build_type}")
 endif()
 
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pipe -W -Wall -Wpedantic -fopenmp -pthread -m64")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pipe -W -Wall -Wpedantic -fopenmp -pthread")
 
 set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O2  -DNDEBUG ")
 set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -g -O0")
--- trinity-plugins/ParaFly/src/Makefile.orig	2025-02-24 18:48:19.761141657 +0000
+++ trinity-plugins/ParaFly/src/Makefile	2025-02-24 18:48:52.950792914 +0000
@@ -64,7 +64,7 @@
 DISTFILES = $(DIST_COMMON) $(DIST_SOURCES) $(TEXINFOS) $(EXTRA_DIST)
 ACLOCAL = ${SHELL} /home/bhaas/GITHUB/trinityrnaseq/trinity-plugins/ParaFly/missing --run aclocal-1.11
 AMTAR = ${SHELL} /home/bhaas/GITHUB/trinityrnaseq/trinity-plugins/ParaFly/missing --run tar
-AM_CXXFLAGS = -pedantic -Wall -Wextra -Wno-long-long -Wno-deprecated -m64
+AM_CXXFLAGS = -pedantic -Wall -Wextra -Wno-long-long -Wno-deprecated
 AUTOCONF = ${SHELL} /home/bhaas/GITHUB/trinityrnaseq/trinity-plugins/ParaFly/missing --run autoconf
 AUTOHEADER = ${SHELL} /home/bhaas/GITHUB/trinityrnaseq/trinity-plugins/ParaFly/missing --run autoheader
 AUTOMAKE = ${SHELL} /home/bhaas/GITHUB/trinityrnaseq/trinity-plugins/ParaFly/missing --run automake-1.11
--- trinity-plugins/ParaFly/Makefile.orig	2025-02-24 18:48:10.851235192 +0000
+++ trinity-plugins/ParaFly/Makefile	2025-02-24 18:48:42.500902744 +0000
@@ -100,7 +100,7 @@
 distcleancheck_listfiles = find . -type f -print
 ACLOCAL = ${SHELL} /home/bhaas/GITHUB/trinityrnaseq/trinity-plugins/ParaFly/missing --run aclocal-1.11
 AMTAR = ${SHELL} /home/bhaas/GITHUB/trinityrnaseq/trinity-plugins/ParaFly/missing --run tar
-AM_CXXFLAGS = -pedantic -Wall -Wextra -Wno-long-long -Wno-deprecated -m64
+AM_CXXFLAGS = -pedantic -Wall -Wextra -Wno-long-long -Wno-deprecated
 AUTOCONF = ${SHELL} /home/bhaas/GITHUB/trinityrnaseq/trinity-plugins/ParaFly/missing --run autoconf
 AUTOHEADER = ${SHELL} /home/bhaas/GITHUB/trinityrnaseq/trinity-plugins/ParaFly/missing --run autoheader
 AUTOMAKE = ${SHELL} /home/bhaas/GITHUB/trinityrnaseq/trinity-plugins/ParaFly/missing --run automake-1.11
