--- Inchworm/CMakeLists.txt.orig	2025-02-24 18:37:08.778178215 +0000
+++ Inchworm/CMakeLists.txt	2025-02-24 18:37:16.638095965 +0000
@@ -13,7 +13,7 @@
     set(CMAKE_BUILD_TYPE "${default_build_type}")
 endif()
 
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pipe -W -Wall -Wpedantic -fopenmp -m64")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pipe -W -Wall -Wpedantic -fopenmp")
 
 set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O2 -DNDEBUG ")
 set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -g -O0")
--- Chrysalis/CMakeLists.txt.orig	2025-02-24 18:47:03.241952176 +0000
+++ Chrysalis/CMakeLists.txt	2025-02-24 18:47:12.251857513 +0000
@@ -13,7 +13,7 @@
     set(CMAKE_BUILD_TYPE "${default_build_type}")
 endif()
 
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pipe -W -Wall -Wpedantic -fopenmp -pthread -m64")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pipe -W -Wall -Wpedantic -fopenmp -pthread")
 
 set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O2  -DNDEBUG ")
 set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -g -O0")
--- trinity-plugins/ParaFly/configure.ac.orig	2025-02-24 18:59:31.364093231 +0000
+++ trinity-plugins/ParaFly/configure.ac	2025-02-24 18:59:49.403903916 +0000
@@ -3,7 +3,7 @@
 AC_LANG([C++])
 AC_PROG_CXX
 AC_OPENMP
-AC_SUBST([AM_CXXFLAGS], [-m64 $OPENMP_CXXFLAGS])
+AC_SUBST([AM_CXXFLAGS], [$OPENMP_CXXFLAGS])
 case $CXX in
   g++*) AC_SUBST([AM_CXXFLAGS],["-pedantic -Wall -Wextra -Wno-long-long -Wno-deprecated $AM_CXXFLAGS"]);;
   sunCC*) AC_SUBST([AM_CXXFLAGS], ["-library=stlport4 -xopenmp -xvpara -fast $AM_CXXFLAGS"]) ;;
--- trinity-plugins/ParaFly/configure.orig	2025-02-24 18:58:55.394470680 +0000
+++ trinity-plugins/ParaFly/configure	2025-02-24 18:59:15.054264317 +0000
@@ -3112,7 +3112,7 @@
   fi
 
 
-AM_CXXFLAGS=-m64 $OPENMP_CXXFLAGS
+AM_CXXFLAGS=$OPENMP_CXXFLAGS
 
 case $CXX in
   g++*) AM_CXXFLAGS="-pedantic -Wall -Wextra -Wno-long-long -Wno-deprecated $AM_CXXFLAGS"
