--- Makefile.old	2017-11-29 16:36:53.468179187 +0200
+++ Makefile	2017-11-29 16:39:07.070976219 +0200
@@ -6,13 +6,6 @@
 # to build Inchworm and Chrysalis with the Intel compiler.
 #
 
-ifeq ($(TRINITY_COMPILER),intel)
- INCHWORM_CONFIGURE_FLAGS = CXX=icpc CXXFLAGS="-fast"
- CHRYSALIS_MAKE_FLAGS = COMPILER=icpc
-else
- override TRINITY_COMPILER=gnu
-endif
-
 
 all: inchworm_target chrysalis_target trinity_essentials
 	sh ./util/support_scripts/trinity_install_tests.sh
@@ -24,9 +17,10 @@
 inchworm_target:
 	@echo Using $(TRINITY_COMPILER) compiler for Inchworm and Chrysalis
 	cd Inchworm && (test -e configure || autoreconf) \
-                && sh ./configure --prefix=`pwd` $(INCHWORM_CONFIGURE_FLAGS) && $(MAKE) install
+                && sh ./configure --prefix=`pwd` CC=${PREFIX}/bin/gcc $(INCHWORM_CONFIGURE_FLAGS) CXX=${PREFIX}/bin/g++ CFLAGS="-fopenmp" CXXFLAGS="-fopenmp" && $(MAKE) install
 
 chrysalis_target:
+	echo "cd Chrysalis && $(MAKE) UNSUPPORTED=yes $(CHRYSALIS_MAKE_FLAGS)"
 	cd Chrysalis && $(MAKE) UNSUPPORTED=yes $(CHRYSALIS_MAKE_FLAGS)
 
 
@@ -64,9 +58,9 @@
 
 # note 'test_all': ** this is for a more advanced installation including devel features **
 
-test_all:
-	cd sample_data/ && make test_all
-	cd trinity_ext_sample_data/ && make test_all
+#test_all:
+#	cd sample_data/ && make test_all
+#	cd trinity_ext_sample_data/ && make test_all
 
 test_clean:
 	cd sample_data/ && make clean
