diff --git a/trinity-plugins/seqtk-trinity-0.0.2/Makefile b/trinity-plugins/seqtk-trinity-0.0.2/Makefile
index ba6c5e1..ea14efc 100644
--- a/trinity-plugins/seqtk-trinity-0.0.2/Makefile
+++ b/trinity-plugins/seqtk-trinity-0.0.2/Makefile
@@ -1,10 +1,8 @@
-CC ?= gcc
-CFLAGS=-g -Wall -O2 -Wno-unused-function
 
 all:seqtk-trinity
 
 seqtk-trinity:seqtk.c khash.h kseq.h
-		$(CC) $(CFLAGS) seqtk.c -o $@ -lz -lm
+		gcc $(CFLAGS) seqtk.c -o $@ -lz -lm
 
 clean:
 		rm -fr gmon.out *.o ext/*.o a.out seqtk trimadap *~ *.a *.dSYM session*
