diff --git a/trinity-plugins/Makefile b/trinity-plugins/Makefile
index ad6df93..fd5c8ea 100644
--- a/trinity-plugins/Makefile
+++ b/trinity-plugins/Makefile
@@ -20,7 +20,7 @@ seqtk_target:
 
 parafly_target:
 	tar -zxvf ${PARAFLY_CODE}.tar.gz && \
-	cd ${PARAFLY_CODE} && sh ./configure --prefix=`pwd` && $(MAKE) install && \
+	cd ${PARAFLY_CODE} && sh ./configure --prefix=`pwd` CXX=g++ CC=gcc CFLAGS="-fopenmp" CXXFLAGS="-fopenmp" && $(MAKE) install && \
 	cp bin/ParaFly ../BIN/
 
 
@@ -33,7 +33,7 @@ plugins: slclust_target collectl_target
 
 
 slclust_target:
-	cd slclust && $(MAKE) install
+	cd slclust && $(MAKE) CXX=g++ CC=gcc install
 
 collectl_target:
 	cd COLLECTL && tar xvf ${COLLECTL_CODE}.src.tar.gz && ln -sf ${COLLECTL_CODE} collectl
@@ -41,8 +41,7 @@ collectl_target:
 clean:
 	rm -rf ./seqtk-trinity-0.0.2
 	cd scaffold_iworm_contigs && $(MAKE) clean
-	cd ${PARAFLY_CODE} && $(MAKE) clean
-	rm -f ./parafly # rm symlink
+	rm -rf ${PARAFLY_CODE}
 	rm -f ./Trimmomatic # rm symlink
 	cd slclust && $(MAKE) clean
 	cd COLLECTL && rm -rf ${COLLECTL_CODE} && rm -f collectl
