diff --git a/trinity-plugins/Makefile b/trinity-plugins/Makefile
index 97f11cb..68e5841 100644
--- a/trinity-plugins/Makefile
+++ b/trinity-plugins/Makefile
@@ -9,19 +9,18 @@ TRIMMOMATIC_CODE=Trimmomatic-0.36
 
 
 
-trinity_essentials: seqtk_target parafly_target trimmomatic_target
+trinity_essentials: seqtk_target parafly_target
 
 trimmomatic_target:
 	ln -sf ${TRIMMOMATIC_CODE} Trimmomatic
 
-
 seqtk_target:
 	cd seqtk-trinity-0.0.2 && $(MAKE) CXX=$(CXX) CC=$(CC)
 	mv seqtk-trinity-0.0.2/seqtk-trinity ./BIN/.
 
 parafly_target:
 	tar -zxvf ${PARAFLY_CODE}.tar.gz && \
-	cd ${PARAFLY_CODE} && sh ./configure --prefix=`pwd` CXX=$(CXX) CC=$(CC) CFLAGS="-fopenmp" CXXFLAGS="-fopenmp" && $(MAKE) install && \
+	cd ${PARAFLY_CODE} && sh ./configure --prefix=`pwd` CXX=g++ CC=gcc CFLAGS="-fopenmp" CXXFLAGS="-fopenmp" && $(MAKE) install && \
 	cp bin/ParaFly ../BIN/
 
 
@@ -34,7 +33,7 @@ plugins: slclust_target collectl_target
 
 
 slclust_target:
-	cd slclust && $(MAKE) CXX=$(CXX) CC=$(CC) install
+	cd slclust && $(MAKE) CXX=g++ CC=gcc install
 
 collectl_target:
 	cd COLLECTL && tar xvf ${COLLECTL_CODE}.src.tar.gz && ln -sf ${COLLECTL_CODE} collectl
