*** fastqc.orig	2015-10-09 11:59:24.000000000 -0400
--- fastqc	2015-12-16 13:27:53.953520316 -0500
***************
*** 158,167 ****
--- 158,170 ----
  	
  	push @java_args ,"-Dfastqc.threads=$threads";
  	my $memory = 250 * $threads;
+ 	my $stack = 200 * $threads;
  	unshift @java_args,"-Xmx${memory}m";
+ 	unshift @java_args,"-Xms${memory}m";
  }
  else {
  	unshift @java_args,'-Xmx250m';
+ 	unshift @java_args,'-Xms200m';
  }
  
  if ($kmer_size) {
