--- setup.py.orig	2015-05-15 18:14:37.000000000 -0400
+++ setup.py	2015-07-16 10:53:53.813643137 -0400
@@ -65,7 +65,7 @@
             package_dir = { '': 'lib' },
             package_data = { '': ['*.ps'] },
             scripts = glob.glob( "bin/*.py"),
-            ext_modules = get_extension_modules(),
+            ext_modules = [],
             test_suite = 'nose.collector',
             setup_requires = ['nose>=0.10.4','cython>=0.12'],
             author = "Liguo Wang",
