diff --git a/src/toil/batchSystems/gridengine.py b/src/toil/batchSystems/gridengine.py
index 02c2f39..1e9d7d8 100644
--- src/toil/batchSystems/gridengine.py
+++ src/toil/batchSystems/gridengine.py
@@ -206,7 +206,7 @@ class GridengineBatchSystem(BatchSystemSupport):
 
     @classmethod
     def supportsWorkerCleanup(cls):
-        return False
+        return True
 
     @classmethod
     def supportsHotDeployment(cls):
diff --git a/src/toil/batchSystems/lsf.py b/src/toil/batchSystems/lsf.py
index dcd1cfe..b64dee3 100644
--- src/toil/batchSystems/lsf.py
+++ src/toil/batchSystems/lsf.py
@@ -129,7 +129,7 @@ class LSFBatchSystem(BatchSystemSupport):
     """
     @classmethod
     def supportsWorkerCleanup(cls):
-        return False
+        return True
 
     @classmethod
     def supportsHotDeployment(cls):
diff --git a/src/toil/batchSystems/slurm.py b/src/toil/batchSystems/slurm.py
index 9651462..1e9c5ca 100644
--- src/toil/batchSystems/slurm.py
+++ src/toil/batchSystems/slurm.py
@@ -243,7 +243,7 @@ class SlurmBatchSystem(BatchSystemSupport):
 
     @classmethod
     def supportsWorkerCleanup(cls):
-        return False
+        return True
 
     @classmethod
     def supportsHotDeployment(cls):
