diff --git a/setup.py b/setup.py
index e775a6c..354f1c6 100644
--- setup.py
+++ setup.py
@@ -45,7 +45,7 @@
             'gcs_oauth2_boto_plugin==1.9',
             botoVersionRequired],
         'cwl': [
-            'cwltool==1.0.20160425140546']},
+            'cwltool==1.0.20160427142240']},
     package_dir={'': 'src'},
     packages=find_packages('src', exclude=['*.test']),
     entry_points={
diff --git a/src/toil/common.py b/src/toil/common.py
index 0253574..4432e05 100644
--- src/toil/common.py
+++ src/toil/common.py
@@ -762,12 +762,11 @@ def _shutdownBatchSystem(self):
         """
         Shuts down current batch system if it has been created.
         """
-        assert self._batchSystem is not None
-
-        startTime = time.time()
-        logger.debug('Shutting down batch system')
-        self._batchSystem.shutdown()
-        logger.debug('Finished shutting down the batch system in %s seconds.' % (time.time() - startTime))
+        if self._batchSystem is not None:
+            startTime = time.time()
+            logger.debug('Shutting down batch system')
+            self._batchSystem.shutdown()
+            logger.debug('Finished shutting down the batch system in %s seconds.' % (time.time() - startTime))
 
     def _assertContextManagerUsed(self):
         if not self._inContextManager:
