commit 893aa9b3295400829a86054cab028635e9bd2dd2
Author: Michael R. Crusoe <michael.crusoe@gmail.com>
Date:   Mon Jan 22 14:58:45 2024 +0100

    remove ruamel.yaml.string dependency for a simpler solution

diff --git a/requirements.txt b/requirements.txt
index c34aceb56..b0e85191d 100644
--- a/requirements.txt
+++ b/requirements.txt
@@ -11,6 +11,5 @@ pytz>=2012
 enlighten>=1.5.2, <2
 configargparse>=1.7,<2
 ruamel.yaml>=0.15
-ruamel.yaml.string>=0.1.1
 pyyaml>=6,<7
 typing-extensions>=4.6.2, <5
diff --git a/src/toil/common.py b/src/toil/common.py
index d2531d3aa..c8852c8ce 100644
--- a/src/toil/common.py
+++ b/src/toil/common.py
@@ -23,6 +23,7 @@ import tempfile
 import time
 import uuid
 import warnings
+from io import StringIO
 
 from ruamel.yaml import YAML
 from ruamel.yaml.comments import CommentedMap
@@ -580,15 +581,16 @@ def generate_config(filepath: str) -> None:
     with AtomicFileCreate(filepath) as temp_path:
         with open(temp_path, "w") as f:
             f.write("config_version: 1.0\n")
-            yaml = YAML(typ=['rt', 'string'])
+            yaml = YAML(typ='rt')
             for data in all_data:
                 if "config_version" in data:
                     del data["config_version"]
-                for line in yaml.dump_to_string(data).split("\n"):  # type: ignore[attr-defined]
-                    if line:
-                        f.write("#")
-                    f.write(line)
-                    f.write("\n")
+                with StringIO() as data_string:
+                    yaml.dump(data, data_string)
+                    for line in data_string.readline():
+                        if line:
+                            f.write("#")
+                        f.write(f"{line}\n")
 
 
 def parser_with_common_options(
