--- toil-5.6.0-orig/setup.py	2022-01-10 21:11:20.000000000 +0000
+++ toil-5.6.0/setup.py	2022-01-18 19:45:57.000000000 +0000
@@ -52,7 +52,6 @@
         "google",
         "kubernetes",
         "mesos",
-        "wdl",
         "server"
     ]
     for extra in non_htcondor_extras:
@@ -112,7 +111,6 @@
                 '_toil_worker = toil.worker:main',
                 'cwltoil = toil.cwl.cwltoil:cwltoil_was_removed [cwl]',
                 'toil-cwl-runner = toil.cwl.cwltoil:main [cwl]',
-                'toil-wdl-runner = toil.wdl.toilwdl:main',
                 '_toil_mesos_executor = toil.batchSystems.mesos.executor:main [mesos]',
                 '_toil_contained_executor = toil.batchSystems.contained_executor:executor']})
 
