--- toil-releases-5.7.1-orig/setup.py	2022-12-09 09:21:30.676954931 +0000
+++ toil-releases-5.7.1/setup.py	2022-12-09 09:21:44.129209710 +0000
@@ -53,7 +53,6 @@
         "google",
         "kubernetes",
         "mesos",
-        "wdl",
         "server"
     ]
     for extra in non_htcondor_extras:
@@ -113,8 +112,6 @@
                 '_toil_worker = toil.worker:main',
                 'cwltoil = toil.cwl.cwltoil:cwltoil_was_removed [cwl]',
                 'toil-cwl-runner = toil.cwl.cwltoil:main [cwl]',
-                'toil-wdl-runner = toil.wdl.toilwdl:main',
-                'toil-wes-cwl-runner = toil.server.cli.wes_cwl_runner:main [server]',
                 '_toil_mesos_executor = toil.batchSystems.mesos.executor:main [mesos]',
                 '_toil_contained_executor = toil.batchSystems.contained_executor:executor']})
