diff --git a/setup.py b/setup.py
index 67c23b061..32ba9bd85 100755
--- a/setup.py
+++ b/setup.py
@@ -54,7 +54,6 @@ def run_setup():
         "google",
         "kubernetes",
         "mesos",
-        "wdl",
         "server"
     ]
     for extra in non_htcondor_extras:
@@ -112,8 +111,6 @@ def run_setup():
                 '_toil_worker = toil.worker:main',
                 'cwltoil = toil.cwl.cwltoil:cwltoil_was_removed [cwl]',
                 'toil-cwl-runner = toil.cwl.cwltoil:main [cwl]',
-                'toil-wdl-runner = toil.wdl.wdltoil:main [wdl]',
-                'toil-wdl-runner-old = toil.wdl.toilwdl:main [wdl]',
                 'toil-wes-cwl-runner = toil.server.cli.wes_cwl_runner:main [server]',
                 '_toil_mesos_executor = toil.batchSystems.mesos.executor:main [mesos]',
                 '_toil_contained_executor = toil.batchSystems.contained_executor:executor']})
