--- toil.orig/setup.py
+++ toil/setup.py
@@ -43,7 +43,6 @@
     docker = 'docker==4.3.1'
     dateutil = 'python-dateutil'
     enlighten = 'enlighten>=1.5.2, <2'
-    wdlparse = 'wdlparse==0.1.0'
 
     core_reqs = [
         dill,
@@ -74,9 +73,6 @@
     mesos_reqs = [
         pymesos,
         psutil]
-    wdl_reqs = [
-        wdlparse
-    ]
 
     # htcondor is not supported by apple
     # this is tricky to conditionally support in 'all' due
@@ -88,8 +84,7 @@
         encryption_reqs + \
         google_reqs + \
         kubernetes_reqs + \
-        mesos_reqs + \
-        wdl_reqs
+        mesos_reqs
 
     setup(
         name='toil',
@@ -129,7 +124,6 @@
             'htcondor:sys_platform!="darwin"': htcondor_reqs,
             'kubernetes': kubernetes_reqs,
             'mesos': mesos_reqs,
-            'wdl': wdl_reqs,
             'all': all_reqs},
         package_dir={'': 'src'},
         packages=find_packages(where='src',
@@ -148,7 +142,6 @@
                 '_toil_worker = toil.worker:main',
                 'cwltoil = toil.cwl.cwltoil:cwltoil_was_removed [cwl]',
                 'toil-cwl-runner = toil.cwl.cwltoil:main [cwl]',
-                'toil-wdl-runner = toil.wdl.toilwdl:main',
                 '_toil_mesos_executor = toil.batchSystems.mesos.executor:main [mesos]',
                 '_toil_kubernetes_executor = toil.batchSystems.kubernetes:executor [kubernetes]']})