*** tools/writeBloom.pl	2023-03-08 16:04:57.000000000 -0800
--- tools/writeBloom-conda.pl	2023-03-10 09:11:57.000000000 -0800
***************
*** 25,31 ****
  use strict;
  use POSIX;
  use FindBin;
! use lib "$FindBin::RealBin/./libexec";
  use BloomFilter;
  use Getopt::Std;
  use Net::SMTP;
--- 25,31 ----
  use strict;
  use POSIX;
  use FindBin;
! use lib "$FindBin::Bin/../lib/site_perl/BloomFilter";
  use BloomFilter;
  use Getopt::Std;
  use Net::SMTP;
