--- a/src/dock.cpp
+++ b/src/dock.cpp
@@ -31,7 +31,6 @@
 #include <boost/program_options.hpp>
 #include <boost/filesystem/fstream.hpp>
 #include <boost/filesystem/exception.hpp>
-#include <boost/filesystem/convenience.hpp> // filesystem::basename
 #include <boost/thread/thread.hpp> // hardware_concurrency // FIXME rm ?
 #include "parse_pdbqt.h"
 #include "parallel_mc.h"
--- a/src/dockBMPI.cpp
+++ b/src/dockBMPI.cpp
@@ -30,7 +30,6 @@
 #include <boost/program_options.hpp>
 #include <boost/filesystem/fstream.hpp>
 #include <boost/filesystem/exception.hpp>
-#include <boost/filesystem/convenience.hpp> // filesystem::basename
 #include <boost/thread/thread.hpp> // hardware_concurrency // FIXME rm ?
 #include "parse_pdbqt.h"
 #include "parallel_mc.h"
--- a/src/mpiBparser.cpp
+++ b/src/mpiBparser.cpp
@@ -9,7 +9,6 @@
 #include <boost/program_options.hpp>
 #include <boost/filesystem/fstream.hpp>
 #include <boost/filesystem/exception.hpp>
-#include <boost/filesystem/convenience.hpp> // filesystem::basename
 #include "parse_pdbqt.h"
 #include "parallel_mc.h"
 #include "file.h"
--- a/src/mainProcedure.h
+++ b/src/mainProcedure.h
@@ -15,7 +15,6 @@
 #include <boost/program_options.hpp>
 #include <boost/filesystem/fstream.hpp>
 #include <boost/filesystem/exception.hpp>
-#include <boost/filesystem/convenience.hpp> // filesystem::basename
 #include <boost/thread/thread.hpp> // hardware_concurrency // FIXME rm ?
 
 #include "parse_pdbqt.h"
--- a/src/vinalc.cpp
+++ b/src/vinalc.cpp
@@ -37,7 +37,6 @@
 #include <boost/program_options.hpp>
 #include <boost/filesystem/fstream.hpp>
 #include <boost/filesystem/exception.hpp>
-#include <boost/filesystem/convenience.hpp> // filesystem::basename
 
 #include <boost/mpi.hpp>
 #include <boost/mpi/environment.hpp>
--- a/apps/vina.cpp
+++ b/apps/vina.cpp
@@ -16,7 +16,6 @@
 #include <boost/program_options.hpp>
 #include <boost/filesystem/fstream.hpp>
 #include <boost/filesystem/exception.hpp>
-#include <boost/filesystem/convenience.hpp> // filesystem::basename
 #include <boost/thread/thread.hpp> // hardware_concurrency // FIXME rm ?
 #include "parse_pdbqt.h"
 #include "parallel_mc.h"
--- a/apps/mpiparser.cpp
+++ b/apps/mpiparser.cpp
@@ -9,7 +9,6 @@
 #include <boost/program_options.hpp>
 #include <boost/filesystem/fstream.hpp>
 #include <boost/filesystem/exception.hpp>
-#include <boost/filesystem/convenience.hpp> // filesystem::basename
 #include "parse_pdbqt.h"
 #include "parallel_mc.h"
 #include "file.h"
