--- mapdamage/composition.py
+++ mapdamage/composition.py
@@ -35,11 +35,10 @@
     Gets the basecomposition of all the sequences in filename
     and returns the value to destination if given.
     """
-    path_to_seqtk = mapdamage.rscript.construct_path("seqtk",folder="seqtk") 
     bases = {"A":0,"C":0,"G":0,"T":0}
     alp = ["A","C","G","T"]
     try:
-        proc = subprocess.Popen([path_to_seqtk,"comp",filename],stdout=subprocess.PIPE)
+        proc = subprocess.Popen(["seqtk","comp",filename],stdout=subprocess.PIPE)
         out = proc.communicate()[0]
         for li in out.splitlines():
             tabs = li.split() # 1 is the ref, 2 is the total and then the base counts A, C, G and T.
