--- librna/fold.c.orig	2018-04-19 06:33:33.919628807 +0200
+++ librna/fold.c	2018-04-19 06:34:13.091983140 +0200
@@ -797,7 +797,7 @@
 }
 /*---------------------------------------------------------------------------*/
 
-inline int HairpinE(int size, int type, int si1, int sj1, const char *string) {
+int HairpinE(int size, int type, int si1, int sj1, const char *string) {
   int energy;
   energy = (size <= 30) ? P->hairpin[size] :
     P->hairpin[30]+(int)(P->lxc*log((size)/30.));
@@ -867,7 +867,7 @@
 
 /*--------------------------------------------------------------------------*/
 
-inline int LoopEnergy(int n1, int n2, int type, int type_2,
+int LoopEnergy(int n1, int n2, int type, int type_2,
 		      int si1, int sj1, int sp1, int sq1) {
   /* compute energy of degree 2 loop (stack bulge or interior) */
   int nl, ns, energy;
