--- src/head_align_tail_dist.py
+++ src/head_align_tail_dist.py
@@ -4,7 +4,7 @@
 import numpy
 import HTSeq
 from sklearn.neighbors import KernelDensity
-from sklearn.externals import joblib
+import joblib
 
 
 def kde2d(x, y):
--- src/read_analysis.py
+++ src/read_analysis.py
@@ -17,7 +17,7 @@
 import argparse
 import numpy
 from sklearn.neighbors import KernelDensity
-from sklearn.externals import joblib
+import joblib
 import head_align_tail_dist as align
 import get_besthit_maf
 import get_primary_sam
