diff --git a/scripts/scrub.sh b/scripts/scrub.sh
index fc2f35c..1fdfaab 100755
--- a/scripts/scrub.sh
+++ b/scripts/scrub.sh
@@ -27,17 +27,17 @@ done
 shift $((OPTIND-1))
 fastq="$1"
 DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )"  >/dev/null 2>&1 && pwd  )"
-ROOT=$(dirname $DIR)
-DB=$ROOT/data/human_filter.db
+ROOT=${SCRUBBER_ROOT}
+DB=${SCRUBBER_SHARE}/data/human_filter.db
 
-if [ "$1" == "test" ] && [ -e "$ROOT/test/scrubber_test.fastq" ];
+if [ "$1" == "test" ] && [ -e "${SCRUBBER_SHARE}/test/scrubber_test.fastq" ];
   then
     TMP_DIR=$(mktemp -d)
-    cp "$ROOT"/test/* "$TMP_DIR/"
+    cp "${SCRUBBER_SHARE}"/test/* "$TMP_DIR/"
     fastq=$TMP_DIR/scrubber_test.fastq
 fi
-python "$ROOT/scripts/fastq_to_fasta.py" < "$fastq" > "$fastq.fasta"
-${ROOT}/bin/aligns_to -db "$ROOT/data/human_filter.db" "$fastq.fasta" | "$ROOT/scripts/cut_spots_fastq.py" "$fastq" ${OPTS} > "$fastq.clean"
+fastq_to_fasta.py < "$fastq" > "$fastq.fasta"
+aligns_to -db "${DB}" "$fastq.fasta" | cut_spots_fastq.py "$fastq" ${OPTS} > "$fastq.clean"
 if [ "$1" == "test" ];
   then
     if [ -e "$TMP_DIR/scrubber_test.fastq.clean" ] &&
