diff --git a/init_db.sh b/init_db.sh
index de67e29..03a69c5 100755
--- a/init_db.sh
+++ b/init_db.sh
@@ -1,7 +1,10 @@
 #!/bin/bash
 set -eu
 
+SCRUBBER_SHARE=BIOCONDA_SED_REPLACE
 VERSION=$(curl "https://ftp.ncbi.nlm.nih.gov/sra/dbs/human_filter/current/version.txt")
-cd data && curl -vf "https://ftp.ncbi.nlm.nih.gov/sra/dbs/human_filter/${VERSION}.human_filter.db" -o "${VERSION}.human_filter.db"
-ln -s "${VERSION}".human_filter.db human_filter.db
+cd ${SCRUBBER_SHARE}/data && \
+  curl -vf "https://ftp.ncbi.nlm.nih.gov/sra/dbs/human_filter/${VERSION}.human_filter.db" -o "${VERSION}.human_filter.db" &&
+  ln -s -f "${VERSION}".human_filter.db human_filter.db
+
 exit 0
 