diff --git a/scripts/scrub.sh b/scripts/scrub.sh
index f6b310c..6848a94 100755
--- a/scripts/scrub.sh
+++ b/scripts/scrub.sh
@@ -27,7 +27,8 @@ RUNTEST=false
 THREADS=
 DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )"  >/dev/null 2>&1 && pwd  )"
 ROOT=$(dirname $DIR)
-DB=$ROOT/data/human_filter.db
+SCRUBBER_SHARE=BIOCONDA_SED_REPLACE
+DB=${SCRUBBER_SHARE}/data/human_filter.db
 
 #Get input
 while getopts ":i:o:d:p:hxrt" opts; do
@@ -65,10 +66,10 @@ if [ "$INFILE" == "test" ];
 fi
 
 #TESTING create temp dir and set infile
-if $RUNTEST && [ -e "$ROOT/test/scrubber_test.fastq" ];
+if $RUNTEST && [ -e "${SCRUBBER_SHARE}/test/scrubber_test.fastq" ];
   then
     TMP_DIR=$(mktemp -d)
-    cp "$ROOT"/test/* "$TMP_DIR/"
+    cp "${SCRUBBER_SHARE}"/test/* "$TMP_DIR/"
     INFILE=$TMP_DIR/scrubber_test.fastq
 fi
 
@@ -85,17 +86,17 @@ if [ ! -e "${INFILE}" ];
     INFILE="$TMP_F_DIR/temp.fastq"
 fi
 #Use infile or temp fastq and generate fasta
-"$ROOT/scripts/fastq_to_fasta.py" < "${INFILE}" > "$TMP_F_DIR/temp.fasta"
+fastq_to_fasta.py" < "${INFILE}" > "$TMP_F_DIR/temp.fasta"
 
 if [ "$OUTFILE" ] && [ "$OUTFILE" != "-" ];
   then
-    "${ROOT}"/bin/aligns_to -db "${DB}" $(if [[ "$THREADS" =~ ^[0-9]+$ ]]; then printf "%s" "-num_threads $THREADS"; fi) "$TMP_F_DIR/temp.fasta" | "$ROOT/scripts/cut_spots_fastq.py" "$INFILE" "$REPLACEN" "$SAVEIDSPOTS" > "$OUTFILE"
+    aligns_to -db "${DB}" $(if [[ "$THREADS" =~ ^[0-9]+$ ]]; then printf "%s" "-num_threads $THREADS"; fi) "$TMP_F_DIR/temp.fasta" | cut_spots_fastq.py "$INFILE" "$REPLACEN" "$SAVEIDSPOTS" > "$OUTFILE"
     if [ "$SAVEIDSPOTS" ] && [ -e "$TMP_F_DIR/temp.fastq.removed_spots" ];
       then
         cp "$TMP_F_DIR/temp.fastq.removed_spots" "$OUTFILE.removed_spots"
     fi
     else
-      "${ROOT}"/bin/aligns_to -db "${DB}" $(if [[ "$THREADS" =~ ^[0-9]+$ ]]; then printf "%s" "-num_threads $THREADS"; fi) "$TMP_F_DIR/temp.fasta" | "$ROOT/scripts/cut_spots_fastq.py" "$INFILE" "$REPLACEN"
+      aligns_to -db "${DB}" $(if [[ "$THREADS" =~ ^[0-9]+$ ]]; then printf "%s" "-num_threads $THREADS"; fi) "$TMP_F_DIR/temp.fasta" | cut_spots_fastq.py "$INFILE" "$REPLACEN"
 fi
 
 #Check if TESTING was successful
