diff --git a/Makefile b/Makefile
index e821782..27e8342 100644
--- a/Makefile
+++ b/Makefile
@@ -1,5 +1,6 @@
 #
-SHELL=/bin/sh
+SHELL=/bin/bash
+CXXFLAGS += -Os -ffunction-sections -fdata-sections
+LDFLAGS += --gc-sections

 MAKE=make
 MAKEFILE=Makefile

@@ -35,7 +36,7 @@ compile:
 		echo "------------------------------------------------------------"; \
 		echo "**** Making $$libdir ****"; \
 		echo "------------------------------------------------------------"; \
-		(cd $$libdir && $(MAKE) -f $(MAKEFILE) "OPT=-O" install) || exit 1; \
+		(cd $$libdir && $(MAKE) -f $(MAKEFILE) LDFLAGS=$(LDFLAGS) CXXFLAGS=$(CXXFLAGS) "OPT=-O" install) || exit 1; \
 	done
 #
 debug:
