diff --git a/Makefile b/Makefile
index e821782..27e8342 100644
--- a/Makefile
+++ b/Makefile
@@ -1,5 +1,6 @@
 #
-SHELL=/bin/sh
+CFLAGS=-O3
+SHELL=/bin/bash
 MAKE=make
 MAKEFILE=Makefile
 
@@ -35,7 +36,7 @@ compile:
 		echo "------------------------------------------------------------"; \
 		echo "**** Making $$libdir ****"; \
 		echo "------------------------------------------------------------"; \
-		(cd $$libdir && $(MAKE) -f $(MAKEFILE) "OPT=-O" install) || exit 1; \
+		(cd $$libdir && $(MAKE) -f $(MAKEFILE) CC=$(CC) CFLAGS=$(CFLAGS) "OPT=-O" install) || exit 1; \
 	done
 #
 debug:
