diff --git a/binary.sh b/binary.sh
index 0c9772d..30c1a44 100755
--- a/binary.sh
+++ b/binary.sh
@@ -1,10 +1,24 @@
 #!/bin/bash
-#
+
+set -exuo pipefail
+
 export RCSBROOT="$(pwd)"
-#
+
 ./bin/cif2bin
-#
-./bin/connect_main
+
+./bin/connect_main &
+
+connect_main_pid=$!
+
+while kill -0 ${connect_main_pid} 2> /dev/null; do
+  echo "\`./bin/connect_main\` in progress..."
+  sleep 30
+done
+
+wait ${connect_main_pid}
+
+echo "\`./bin/connect_main\` finished."
+
 if [ -e component.odb ]
 then
     mv component.odb ./data/binary
