diff --git a/cifparse-obj-v7.0/Makefile b/cifparse-obj-v7.0/Makefile
index fbe18ab..a1fc2aa 100755
--- a/cifparse-obj-v7.0/Makefile
+++ b/cifparse-obj-v7.0/Makefile
@@ -20,7 +20,7 @@ VPATH = $(OBJ_DIR)
 # LINCLUDES and LDEFINES are appended to CFLAGS and C++FLAGS
 #----------------------------------------------------------------------------
 LDEFINES  =
-LINCLUDES = -I$(L_INCL_DIR) -I$(M_INCL_DIR) 
+LINCLUDES = -I$(L_INCL_DIR) -I$(M_INCL_DIR)
 
 #----------------------------------------------------------------------------
 # Include the appropriate compiler/platform definitions ...
@@ -29,10 +29,10 @@ include ../etc/Makefile.platform
 
 # flex/bison flags
 CifParser_YACC_FLAGS = $(YACCFLAGS) -p cifparser_
-CifScanner_LEX_FLAGS = $(LEXFLAGS) -Pcifparser_ 
+CifScanner_LEX_FLAGS = $(LEXFLAGS) -Pcifparser_
 
 DICParser_YACC_FLAGS = $(YACCFLAGS) -p dicparser_
-DICScanner_LEX_FLAGS = $(LEXFLAGS) -Pdicparser_ 
+DICScanner_LEX_FLAGS = $(LEXFLAGS) -Pdicparser_
 
 # Dependent libraries
 TABLES_LIB        = $(M_LIB_DIR)/tables.a
@@ -177,9 +177,9 @@ $(L_MOD_LIB): $(OBJ_FILES)
 # Specific rules for making object files
 $(OBJ_DIR)/%Parser.o: $(SRC_DIR)/%Parser.y
 	@sh -c 'cd $(SRC_DIR); $(YACC) $(${^F:.y=_YACC_FLAGS}) ../$<'
-	mv $(^:.y=.tab.c) $(^:.y=.c)
+	cp $(^:.y=.tab.c) $(^:.y=.c)
 	cp $(^:.y=.tab.h) $(^:.y=.h)
-	$(CC) $(CFLAGS_NONANSI) -c $(^:.y=.c) -o $@ 
+	$(CC) $(CFLAGS_NONANSI) -c $(^:.y=.c) -o $@
 
 %Scanner.o: $(SRC_DIR)/%Scanner.l $(OBJ_DIR)/%Parser.o
 	$(LEX) $($(@F:.o=_LEX_FLAGS)) -t $(SRC_DIR)/$(@:.o=.l) > $(SRC_DIR)/$(@F:.o=.c)
