diff --git a/Makefile b/Makefile
index e821782..9a74931 100644
--- a/Makefile
+++ b/Makefile
@@ -1,5 +1,5 @@
 #
-SHELL=/bin/sh
+SHELL=/bin/bash
 MAKE=make
 MAKEFILE=Makefile
 
@@ -22,7 +22,7 @@ LIBDIRS = common-v4.5 \
 all:	compile
 
 binary:	compile
-	@sh -c './binary.sh'
+	@prlimit --as=6000000000 -- bash -c './binary.sh'
 	@./bin/DictToSdb -ddlFile ./data/ascii/mmcif_ddl.dic -dictFile ./data/ascii/mmcif_pdbx.dic -dictSdbFile mmcif_pdbx.sdb
 	@mv mmcif_pdbx.sdb ./data/binary
 	@rm -f ./bin/DictToSdb ./bin/cif2bin ./bin/connect_main
@@ -35,7 +35,7 @@ compile:
 		echo "------------------------------------------------------------"; \
 		echo "**** Making $$libdir ****"; \
 		echo "------------------------------------------------------------"; \
-		(cd $$libdir && $(MAKE) -f $(MAKEFILE) "OPT=-O" install) || exit 1; \
+		(cd $$libdir && prlimit --as=1000000000 -- $(MAKE) -f $(MAKEFILE) "OPT=-O -fdata-sections -ffunction-sections" install) || exit 1; \
 	done
 #
 debug:
@@ -48,7 +48,7 @@ debug:
 		(cd $$libdir && $(MAKE) -f $(MAKEFILE)  "OPT=-g" install) || exit 1; \
 	done
 #
-clean:  
+clean:
 	@for libdir in $(LIBDIRS); do \
 		echo cleaning $$libdir; \
 		(cd $$libdir && $(MAKE) -f $(MAKEFILE) clean) || exit 1; \
