diff --git a/etc/make.platform.gnu8 b/etc/make.platform.gnu8
index 64ed481..72bece3 100644
--- a/etc/make.platform.gnu8
+++ b/etc/make.platform.gnu8
@@ -27,14 +27,14 @@ YACCFLAGS=-d -v -l
 ## ABI
 ## This part defines an application binary interface to be used with the
 ## compiler suite.
-ABI= 
+ABI=
 
 ## Warnings flags
 ## This part defines warnings related flags
 
 # WARNINGS_AS_ERRORS defines flags to instruct all compilers to treat all
 # warnings as errors.
-WARNINGS_AS_ERRORS=-Werror
+WARNINGS_AS_ERRORS=
 
 # ALL_WARNINGS defines flags to instruct all compilers to report all
 # warnings.
@@ -54,10 +54,10 @@ WARNINGS=$(WARNINGS_AS_ERRORS) $(ALL_WARNINGS)
 # Should be defined only for big endian platforms. Otherwise
 # it must be left undefined, which indicates little endian platform.
 #ENDIANESS=-DBIG_ENDIAN_PLATFORM
- 
+
 # OS indicator. Should be defined only if SunOS is the platform
 #PLATFORM_OS=-DSUN_OS
- 
+
 # Collect all platform related flags
 PLATFORM=$(ENDIANESS) $(PLATFORM_OS)
 
@@ -90,11 +90,11 @@ GDEFINES=$(PLATFORM) $(COMPILER)
 DEFINES=$(GDEFINES) $(LDEFINES)
 
 ## Global include directories
-GINCLUDES= 
+GINCLUDES=
 
 # Collect all include directories from global include directories and
 # include directories specified in module makefile
-INCLUDES=$(LINCLUDES) $(GINCLUDES) 
+INCLUDES=$(LINCLUDES) $(GINCLUDES)
 
 ## C compiler
 ## This part defines C compiler information
@@ -109,7 +109,7 @@ C_WARNINGS=$(WARNINGS)
 
 # C compiler flags
 CFLAGS=$(OPT) $(ABI) $(PIC) $(ANSI_C_FLAG) $(C_WARNINGS) $(DEFINES) $(INCLUDES)
-CFLAGS_NONANSI=$(OPT) $(ABI) $(PIC) $(NON_ANSI_C_FLAG) $(DEFINES) $(INCLUDES) 
+CFLAGS_NONANSI=$(OPT) $(ABI) $(PIC) $(NON_ANSI_C_FLAG) $(DEFINES) $(INCLUDES)
 
 ## C++ compiler
 ## This part defines C++ compiler information
@@ -192,7 +192,7 @@ RANLIB=true
 #-----Installer----------------------------------------------------------#
 
 INSTALL=../etc/cifinstall
-INSTALLOPTS=-m 0444 
+INSTALLOPTS=-m 0444
 
 #------------------------------------------------------------------------#
 
