diff --git a/etc/platform.sh b/etc/platform.sh
index 1607685..e5f3573 100755
--- a/etc/platform.sh
+++ b/etc/platform.sh
@@ -6,7 +6,7 @@
 #  Copy the platform specific section of the Makefile in
 #                  the current directory to:
 #
-#                      Makefile.platform. 
+#                      Makefile.platform.
 #
 #
 sysid="unknown"
@@ -29,10 +29,10 @@ case $(uname -s) in
 
 #   Check if it is a Mac OS platform
     Darwin)
-#     First check if it is clang-xxxx
-      clang_ver=$(gcc --version | grep -e "clang-")
+#     First check if it is clang
+      clang_ver=$(clang --version | grep -e "clang version")
       if [ ! -z "$clang_ver" ]
-      then 
+      then
         sysid="darwin4"
       else
 #       Check if it is GCC version 4.x
@@ -61,7 +61,7 @@ case $(uname -s) in
 #         It is GCC version 4.x
           sysid="darwin4"
         fi
-      fi 
+      fi
     ;;
 
 #   Check if it is a Linux platform
@@ -85,11 +85,11 @@ case $(uname -s) in
       fi
 #
       if [ -z "$gcc_ver" ]
-      then	
+      then
 	  #     Check if it is GCC version 5.x
 	  gcc_ver=$(gcc --version | grep -e " 5\.")
 	  if [ -z "$gcc_ver" ]
-	  then	
+	  then
 	      #     If not GCC version 5.x
               gcc_ver=$(gcc --version | grep -e " 4\.")
               if [ -z "$gcc_ver" ]
