diff --git a/lib/Makefile b/lib/Makefile
index 30ce1a5..69b71e1 100644
--- a/lib/Makefile
+++ b/lib/Makefile
@@ -1,15 +1,14 @@
-CC=gcc
-INCLUDES =
-CFLAGS = -std=c99 -O3 -Wall -Wno-unused-function
+CC=$(CC)
+CFLAGS = -std=c11 -O3 -Wall -Wno-unused-function
 FPIC = -fPIC -shared
 HTSDIR = ./htslib/
 LIB_OBJECT = nextcorrect.so ovlseq.so ctg_cns.so ckit.so
-CTG_CNS_CFLAGS = -D_SVID_SOURCE -DLGS_CORRECT -I $(HTSDIR) -lm -lpthread -lz $(HTSLIB_CFLAGS)
+CTG_CNS_CFLAGS = -D_SVID_SOURCE -DLGS_CORRECT -I$(HTSDIR) -lcurl -lcrypto -lm -pthread -lz $(HTSLIB_CFLAGS)
 
 all: $(LIB_OBJECT)
 
 $(HTSDIR)/libhts.a:
-	cd  $(HTSDIR) && ./configure  --disable-bz2 --disable-lzma
+	cd $(HTSDIR) && ./configure --disable-bz2 --disable-lzma
 	@+make -C $(HTSDIR) CPPFLAGS+=-fPIC $(INCLUDES)
 
 nextcorrect.so: align.c dag.c nextcorrect.c
@@ -31,7 +30,7 @@ ckit: kit.c bseq.c
 	$(CC) $(CFLAGS) -o $@ $^ -lz
 
 ctg_cns: bseq.c align.c dag.c bsort.c ctg_cns.c $(HTSDIR)/libhts.a
-	$(CC) $(CFLAGS) -o $@ $^ $(CTG_CNS_CFLAGS)
+	$(CC) $(CFLAGS) $(INCLUDES) -o $@ $^ $(CTG_CNS_CFLAGS)
 
 ovlseq: ovlseq.c bseq.c ovl.c index.c
 	$(CC) $(CFLAGS) -lz -Wno-unused-function -o $@ $^
diff --git a/util/Makefile b/util/Makefile
index 11328f0..5805182 100644
--- a/util/Makefile
+++ b/util/Makefile
@@ -1,6 +1,5 @@
-CC = gcc
-CFLAGS = -Wall -O3 -Wno-unused-function 
-INCLUDES =
+CC = $(CC)
+CFLAGS = -Wall -O3 -Wno-unused-function
 PROGS = seq_dump ovl_sort seq_stat ovl_cvt seq_bit bam_sort nextgraph
 HTSDIR = ../lib/htslib/
 NEXTGRPHDIR = ./ng
@@ -8,14 +7,14 @@ NEXTGRPHDIR = ./ng
 all:$(PROGS)
 
 $(HTSDIR)/libhts.a:
-	cd  $(HTSDIR) && ./configure --disable-bz2 --disable-lzma
+	cd $(HTSDIR) && ./configure --disable-bz2 --disable-lzma
 	@+make -C $(HTSDIR) CPPFLAGS+=-fPIC $(INCLUDES)
 
 seq_dump: seq_dump.c ../lib/bseq.c
 	$(CC) $(CFLAGS) $(INCLUDES) -o $@ $^ -lz
 
 ovl_sort: ovl_sort.c thpool.c ../lib/ovl.c
-	$(CC) $(CFLAGS) -std=c99 -pthread $(INCLUDES) -o $@ $^
+	$(CC) $(CFLAGS) -std=c11 -pthread $(INCLUDES) -o $@ $^
 
 ovl_cvt: ovl_cvt.c ../lib/ovl.c
 	$(CC) $(CFLAGS) $(INCLUDES) -o $@ $^
@@ -24,10 +23,10 @@ seq_bit: seq_bit.c ../lib/bseq.c
 	$(CC) $(CFLAGS) $(INCLUDES) -o $@ $^ -lz
 
 seq_stat: seq_stat.c
-	$(CC) $(CFLAGS) -std=c99 $(INCLUDES) -o $@ $^ -lz
+	$(CC) $(CFLAGS) -std=c11 $(INCLUDES) -o $@ $^ -lz
 
 bam_sort: bam_sort.c ../lib/bsort.c $(HTSDIR)/libhts.a
-	$(CC) $(CFLAGS) -o $@ $^ -I $(HTSDIR) $(INCLUDES) -lpthread -lm -lz
+	$(CC) $(CFLAGS) -o $@ $^ -I$(HTSDIR) $(INCLUDES) -pthread -lcurl -lm -lz -lcrypto -ldeflate
 
 # align: realign.c ../lib/ovl.c ../lib/bseq.c ../lib/align.c
 # 	$(CC) $(CFLAGS) -Wno-unused-function -o $@ $^ -lz
