diff --git a/minimap2/Makefile b/minimap2/Makefile
index 64d7ce9..44f7579 100644
--- a/minimap2/Makefile
+++ b/minimap2/Makefile
@@ -1,10 +1,10 @@
-CFLAGS=		-g -Wall -O2 -Wc++-compat -Wno-unused-function #-Wextra
+CFLAGS=		-g -Wall -O3 -Wc++-compat -Wno-unused-function -L$(PREFIX)/lib  #-Wextra
 CPPFLAGS=	-DHAVE_KALLOC
-INCLUDES=	
+INCLUDES=	-I$(PREFIX)/include
 OBJS=		kthread.o kalloc.o misc.o bseq.o sketch.o sdust.o options.o index.o chain.o align.o hit.o map.o format.o pe.o esterr.o splitidx.o ksw2_ll_sse.o
 PROG=		minimap2-nd
 PROG_EXTRA=	sdust minimap2-lite
-LIBS=		-lm -lz -lpthread
+LIBS=		-lm -lz -pthread
 
 ifeq ($(arm_neon),) # if arm_neon is not defined
 ifeq ($(sse2only),) # if sse2only is not defined
