diff --git a/lib/htslib/Makefile b/lib/htslib/Makefile
index ef86c83..ec18aa3 100644
--- a/lib/htslib/Makefile
+++ b/lib/htslib/Makefile
@@ -27,7 +27,7 @@ AR     = ar
 RANLIB = ranlib
 
 # Default libraries to link if configure is not used
-htslib_default_libs = -lz -lm -lbz2 -llzma
+htslib_default_libs = -lz -lm -lbz2 -llzma -ldeflate
 
 CPPFLAGS =
 # TODO: probably update cram code to make it compile cleanly with -Wc++-compat
diff --git a/util/ng/Makefile b/util/ng/Makefile
index 9221492..3014a46 100644
--- a/util/ng/Makefile
+++ b/util/ng/Makefile
@@ -2,10 +2,10 @@ CC=			gcc
 #GENOME_SIZE=0, 1, 2 for small, middle, large genome size
 CPPFLAGS=   -DGENOME_SIZE=0
 CFLAGS=		-Wall -O3  -Wno-unused-function $(CPPFLAGS)
-INCLUDES=
+INCLUDES=	-I$(PREFIX)/include
 OBJS=		asg.o ctg.o edge.o kit.o main.o node.o opt.o out.o
 PROG=		nextgraph
-LIBS=		../../lib/bseq.c ../../lib/index.c ../../lib/ovl.c -lm -lz
+LIBS=		../../lib/bseq.c ../../lib/index.c ../../lib/ovl.c -L$(PREFIX)/lib -lm -lz
 
 .SUFFIXES:.c .o
 
