diff --git a/crates/turbo-picard-cli/Cargo.toml b/crates/turbo-picard-cli/Cargo.toml
index ad3ccef..812ef54 100644
--- a/crates/turbo-picard-cli/Cargo.toml
+++ b/crates/turbo-picard-cli/Cargo.toml
@@ -19,7 +19,7 @@ name = "picard"
 path = "src/bin/picard.rs"
 
 [dependencies]
-rust-htslib = "1.0.0"
+rust-htslib = { version = "1.0.0", features = ["static"] }
 md5 = "0.7.0"
 flate2 = "1.1.5"
 rustc-hash = "2.1"
diff --git a/crates/turbo-picard-core/Cargo.toml b/crates/turbo-picard-core/Cargo.toml
index f0b379a..8a981ad 100644
--- a/crates/turbo-picard-core/Cargo.toml
+++ b/crates/turbo-picard-core/Cargo.toml
@@ -10,4 +10,4 @@ homepage.workspace = true
 description = "Shared Picard argument and MarkDuplicates configuration types for turbo-picard."
 
 [dependencies]
-rust-htslib = "1.0.0"
\ No newline at end of file
+rust-htslib = { version = "1.0.0", features = ["static"] }
\ No newline at end of file
diff --git a/crates/turbo-picard-markdup/Cargo.toml b/crates/turbo-picard-markdup/Cargo.toml
index 457dd20..c842ebd 100644
--- a/crates/turbo-picard-markdup/Cargo.toml
+++ b/crates/turbo-picard-markdup/Cargo.toml
@@ -12,7 +12,7 @@ description = "Native accelerated MarkDuplicates engine for turbo-picard."
 [dependencies]
 turbo-picard-core = { path = "../turbo-picard-core", version = "0.1.5" }
 md5 = "0.8.0"
-rust-htslib = "1.0.0"
+rust-htslib = { version = "1.0.0", features = ["static"] }
 rustc-hash = "1.1.0"
 
 [dev-dependencies]
