diff --git a/Makefile b/Makefile
index 81bf4e9..56eb80d 100644
--- a/Makefile
+++ b/Makefile
@@ -1,13 +1,13 @@
 # Useful directories
 
 # Directories to search for header files
-SEARCHDIRS := -I. -I- 
+SEARCHDIRS := -I. -I$(PREFIX)/include
 
 
 #SYSTYPE :=     $(shell uname)
 
-CC      := gcc
-CFLAGS  = -O2 -Wall ${SEARCHDIRS} -D_REENTRANT
+CC      := $(CC)
+CFLAGS  = -O3 -Wall ${SEARCHDIRS} -D_REENTRANT -Wno-implicit-function-declaration
 
 %.o : %.c
 	${CC} ${CFLAGS} -c $< -o $@
@@ -26,8 +26,8 @@ CFLAGS  = -O2 -Wall ${SEARCHDIRS} -D_REENTRANT
 
 # C/C++ linker
 
-LINKER    := gcc
-LDFLAGS    =
+LINKER    := $(CC)
+LDFLAGS    = -L$(PREFIX)/lib
 LOADLIBES := 
 
 .PHONY : all
