--- setup.py.orig	2017-11-29 23:08:50.331161360 +0100
+++ setup.py	2017-11-29 23:11:03.449353941 +0100
@@ -10,14 +10,10 @@
 
     def run(self):
         mccortex_dir = os.path.dirname(os.path.realpath(__file__))+"/mccortex"
-        subprocess.call(
-            ["make", "clean"], cwd=mccortex_dir)
-        subprocess.call(
-            ["make"], cwd=mccortex_dir)
         print("Moving mccortex from %s to %s" %
-              (mccortex_dir, os.environ.get('VIRTUAL_ENV', '/usr/local/')))
-        subprocess.call(
-            ["cp", "bin/mccortex31", "%s/bin/" % os.environ.get('VIRTUAL_ENV', '/usr/local/')], cwd=mccortex_dir)
+              (mccortex_dir, os.environ['PREFIX']))
+        subprocess.check_call(
+            ["cp", "bin/mccortex31", "%s/bin/" % os.environ['PREFIX']], cwd=mccortex_dir)
         DistutilsInstall.run(self)
 
 setup(
