commit 98c6545e4d2021069e431815a78c302dc0a342fd
Author: Phelim Bradley <wave@phel.im>
Date:   Thu Mar 9 02:51:07 2017 +0000

    min conf 100

diff --git a/mykatlas/base.py b/mykatlas/base.py
index 8578213..a397ab8 100644
--- a/mykatlas/base.py
+++ b/mykatlas/base.py
@@ -140,7 +140,7 @@ genotyping_mixin.add_argument(
 genotyping_mixin.add_argument(
     "--min_variant_conf",
     help="minimum genotype confidence for variant genotyping",
-    default=None, type=int)
+    default=100, type=int)
 genotyping_mixin.add_argument(
     "--min_gene_conf",
     help="minimum genotype confidence for gene genotyping",
