diff --git a/Makefile b/Makefile
index 30d8cd1..1617252 100644
--- a/Makefile
+++ b/Makefile
@@ -4,8 +4,8 @@ LIB_DIR := $(CURDIR)/lib
 
 # Set the library paths for the compiler
 CONDA_PREFIX ?= $(shell echo $$CONDA_PREFIX)
-LIBRARY_PATHS := -L$(LIB_DIR) -L$(CONDA_PREFIX)/lib
-INCLUDE_PATHS := -I$(INCL_DIR) -I$(CONDA_PREFIX)/include
+LIBRARY_PATHS := -L$(LIB_DIR) -L$(CONDA_PREFIX)/lib -L$(PREFIX)/lib
+INCLUDE_PATHS := -I$(INCL_DIR) -I$(CONDA_PREFIX)/include -I$(PREFIX)/include
 
 # All targets
 all: swig_build compile
