diff --git a/Makefile b/Makefile
index 47aa514..5eea781 100644
--- a/Makefile
+++ b/Makefile
@@ -1,5 +1,5 @@
 CFLAGS ?= -Wall -O3
-CFLAGS += -std=c99
+CFLAGS += -std=c99 -I$(PREFIX)/include
 LIBS = bytescale.o cdist.o cmdline.o dat.o datclust.o dbscan.o dbparse.o distcmp.o dist.o dnj.o fbseek.o filebuff.o fsacmp.o fsacmpthrd.o fullphy.o hashmapstr.o hashmapstrindex.o hclust.o jobs.o ltdmatrix.o ltdmatrixthrd.o machines.o makespan.o matcmp.o matparse.o matrix.o merge.o meth.o methparse.o mvjobs.o mvmakespan.o mvtabusearch.o nj.o nwck.o nwck2phy.o pherror.o phy.o phycmp.o qseqs.o rarify.o resparse.o seqparse.o seq2fasta.o stdnuc.o stdstat.o str.o tabusearch.o tmp.o tree.o trim.o tsv.o tsv2nwck.o tsv2phy.o ulist.o union.o unionparse.o vector.o
 PROGS = ccphylo
 
@@ -9,7 +9,7 @@ PROGS = ccphylo
 all: $(PROGS)
 
 ccphylo: main.c libccphylo.a
-	$(CC) $(CFLAGS) -o $@ main.c libccphylo.a -lm -lpthread -lz
+	$(CC) $(CFLAGS) -o $@ main.c libccphylo.a -L$(PREFIX)/lib -lz -pthread -lm
 
 libccphylo.a: $(LIBS)
 	$(AR) -csr $@ $(LIBS)
