--- src/converters/CMakeLists.txt.bk	2016-09-28 14:03:55.306114301 +0200
+++ src/converters/CMakeLists.txt	2016-09-28 14:04:10.078515184 +0200
@@ -142,12 +142,12 @@
   endif()
 endif()
 
-if(BOOSTDB)
-  find_package(Boost ${BOOST_MIN_VERSION} COMPONENTS serialization filesystem system REQUIRED)
-else(BOOSTDB)
-  find_package(Boost ${BOOST_MIN_VERSION} COMPONENTS filesystem system REQUIRED)
-endif()
-include_directories(${Boost_INCLUDE_DIRS})
+#if(BOOSTDB)
+#  find_package(Boost ${BOOST_MIN_VERSION} COMPONENTS serialization filesystem system REQUIRED)
+#else(BOOSTDB)
+#  find_package(Boost ${BOOST_MIN_VERSION} COMPONENTS filesystem system REQUIRED)
+#endif()
+include_directories(${BOOST_INCLUDEDIR})
 
 
 FIND_PACKAGE ( Threads REQUIRED )
