diff --git a/Makefile b/Makefile
index 76824c4..87a9bc6 100644
--- a/Makefile
+++ b/Makefile
@@ -6,7 +6,7 @@
 # 1) Choose C++ compiler.
 ################################################################################
 
-CXX = g++
+CXX = $(CXX)
 
 ################################################################################
 # 2) Set C++ flags.
@@ -17,8 +17,8 @@ CXX = g++
 
 OTHERFLAGS = -DVERSION="\"1.00\""
 
-# debug mode    
-CXXFLAGS = -g -W -Wall -pedantic $(OTHERFLAGS)
+# debug mode
+CXXFLAGS = -g -O3 -W -Wall -pedantic $(OTHERFLAGS) -march=x86-64-v3 -funroll-loops
 
 # profile mode
 #CXXFLAGS = -pg -W -Wall -pedantic $(OTHERFLAGS)
