--- configure.ac
+++ configure.ac
@@ -11,5 +11,5 @@
-CC=`"${R_HOME}/bin/R" CMD config CC`
-CPP=`"${R_HOME}/bin/R" CMD config CPP`
-CFLAGS=`"${R_HOME}/bin/R" CMD config CFLAGS`
-CPPFLAGS=`"${R_HOME}/bin/R" CMD config CPPFLAGS`
-LDFLAGS=`"${R_HOME}/bin/R" CMD config LDFLAGS`
+CC=${CC-`"${R_HOME}/bin/R" CMD config CC`}
+CPP=${CPP-`"${R_HOME}/bin/R" CMD config CPP`}
+CFLAGS="${CFLAGS} `"${R_HOME}/bin/R" CMD config CFLAGS`"
+CPPFLAGS="${CPPFLAGS} `"${R_HOME}/bin/R" CMD config CPPFLAGS`"
+LDFLAGS="${LDFLAGS} `"${R_HOME}/bin/R" CMD config LDFLAGS`"
