diff -Naur sequenticon-0.1.6.orig/MANIFEST.in sequenticon-0.1.6/MANIFEST.in
--- a/sequenticon-0.1.6/MANIFEST.in	2025-04-01 19:03:43.000000000 +0200
+++ b/sequenticon-0.1.6/MANIFEST.in	2025-04-30 18:17:04.333791817 +0200
@@ -1,5 +1,6 @@
 include *.txt
-recursive-include examples *.txt *.py
 recursive-include sequenticon/report_assets *
 include *.rst
 include ez_setup.py
+prune examples
+prune tests
diff -Naur sequenticon-0.1.6.orig/pyproject.toml sequenticon-0.1.6/pyproject.toml
--- a/sequenticon-0.1.6/pyproject.toml	2025-04-01 19:03:43.000000000 +0200
+++ b/sequenticon-0.1.6/pyproject.toml	2025-04-30 18:32:21.576753256 +0200
@@ -25,4 +25,4 @@
 include-package-data = true
 
 [tool.setuptools.packages.find]
-exclude = ["docs"]
+exclude = ["docs", "tests", "tests.*", "examples", "examples.*"]
diff -Naur sequenticon-0.1.6.orig/setup.py sequenticon-0.1.6/setup.py
--- a/sequenticon-0.1.6/setup.py	2025-04-01 19:03:43.000000000 +0200
+++ b/sequenticon-0.1.6/setup.py	2025-04-30 18:17:01.784799389 +0200
@@ -12,7 +12,7 @@
     long_description=open("pypi-readme.rst").read(),
     license="MIT",
     keywords="DNA sequence barcoding sequenticon identicon hash",
-    packages=find_packages(exclude="docs"),
+    packages=find_packages(exclude=["docs","tests*","examples*"]),
     include_package_data=True,
     install_requires=[
         "Biopython",
