--- src/tldevel.h	2019-10-27 20:27:03.000000000 -0400
+++ src/tldevel.h.new	2020-03-04 12:08:26.837784478 -0500
@@ -518,8 +518,15 @@
 
 
 #define DECLARE_TIMER(n) struct timespec ts1_##n; struct timespec ts2_##n;
+
+#ifdef CLOCK_MONOTONIC_RAW
 #define START_TIMER(n) clock_gettime(CLOCK_MONOTONIC_RAW, &ts1_##n);
 #define STOP_TIMER(n) clock_gettime(CLOCK_MONOTONIC_RAW, &ts2_##n);
+#else
+#define START_TIMER(n) gettimeofday(&ts1_##n);
+#define STOP_TIMER(n) gettimeofday(&ts2_##n);
+#endif
+
 #define GET_TIMING(n) (double)(ts2_##n.tv_sec - ts1_##n.tv_sec) + ((double)  ts2_##n.tv_nsec - ts1_##n.tv_nsec) / 1000000000.0
 
 
