--- setup.py.old	2015-11-21 00:52:00.000000000 +0000
+++ setup.py	2017-09-19 15:47:31.218292751 +0000
@@ -3,7 +3,7 @@
 version = "Unknown"
 for line in open("breakseq2/_version.py"):
     if line.startswith("__version__"):
-        version = line.strip().split("=")[1].strip()
+        version = line.strip().split("=")[1].strip()[1:-1]
 
 print version
 setup(
@@ -16,5 +16,4 @@
       packages = find_packages(),
       install_requires = ["cython", "pysam==0.7.7", "biopython==1.65"],
       scripts=['scripts/run_breakseq2.py', 'scripts/breakseq2_gen_bplib.py'],
-      dependency_links = ["https://pypi.python.org/packages/source/p/pysam/pysam-0.7.7.tar.gz"]
       )
