# ggally nowadays provides the ggnet / ggnet2 functions
diff --git a/R/amp_otu_network.R b/R/amp_otu_network.R
index e92b3ce..3f07286 100644
--- a/R/amp_otu_network.R
+++ b/R/amp_otu_network.R
@@ -57,7 +57,7 @@ amp_otu_network <- function(data,
                             detailed_output = FALSE,
                             ...) {
   checkReqPkg(
-    "ggnet",
+    "GGally",
     " Please install with:\n  remotes::install_github(\"kasperskytte/ggnet\")"
   )
   checkReqPkg("network")
