diff --git a/AMULET.py b/AMULET.py
--- a/AMULET.py
+++ b/AMULET.py
@@ -102 +102 @@ def getDoublets(matrix, unionoverlaps, rcelliddict):
-    return np.array(doublets_probas, dtype=np.object)
+    return np.array(doublets_probas, dtype=object)
@@ -133 +133 @@ def getFilteredOverlaps(data, simplerepeats, expectedoverlap):
-                counts = np.ones((len(starts), 2), dtype=np.object)
+                counts = np.ones((len(starts), 2), dtype=object)
@@ -138 +138 @@ def getFilteredOverlaps(data, simplerepeats, expectedoverlap):
-                counts2 = -1*np.ones((len(starts), 2), dtype=np.object)
+                counts2 = -1*np.ones((len(starts), 2), dtype=object)
diff --git a/FragmentFileOverlapCounter.py b/FragmentFileOverlapCounter.py
--- a/FragmentFileOverlapCounter.py
+++ b/FragmentFileOverlapCounter.py
@@ -193 +193 @@ def writeOverlapSummary(filepath, overlapcounts, vreadspercell, readspercell):
-    table = np.empty(shape=(len(overlapcounts),len(colnames)), dtype=np.object)
+    table = np.empty(shape=(len(overlapcounts),len(colnames)), dtype=object)
diff --git a/peakoverlap.py b/peakoverlap.py
--- a/peakoverlap.py
+++ b/peakoverlap.py
@@ -358 +358 @@ def getUnionPeaks(datasets, chridx=0, startidx=1, endidx=2):
-    return np.array(rv, dtype=np.object)
+    return np.array(rv, dtype=object)
