diff --git a/Makefile.am b/Makefile.am
index 2275cd9..bd032c1 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -3,7 +3,7 @@
 ## to provide a way for the user to supply additional arguments.
 ACLOCAL_AMFLAGS = -I m4 ${ACLOCAL_FLAGS}
 
-AM_CPPFLAGS = -fopenmp -I ../htslib/ -I ../../htslib/htslib -pipe -D__STDC_LIMIT_MACROS -Wall -Wno-unused-local-typedefs -Wno-enum-compare -fpic -O2 
+AM_CPPFLAGS = -fopenmp -I${PREFIX}/include/htslib -pipe -D__STDC_LIMIT_MACROS -Wall -Wno-unused-local-typedefs -Wno-enum-compare -fpic -O2 
 
 noinst_HEADERS = \
 	Constant.h pException.h \
@@ -24,4 +24,4 @@ demuxlet_SOURCES = \
 	tsv_reader.cpp \
 	cmd_cram_demuxlet.cpp
 
-demuxlet_LDADD = ../htslib/libhts.a -lpthread -llzma -lz -lbz2 -lgomp -lcurl -lcrypto
+demuxlet_LDADD = -lhts -lpthread -lgomp
