From d379bfa0ab038bc4b2e3b2f473cd266de5c90d2b Mon Sep 17 00:00:00 2001
From: Oliver Schwengers <oliver@schwengers.de>
Date: Wed, 14 Oct 2020 17:24:04 +0200
Subject: [PATCH] fix Makefile

---
 Makefile | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/Makefile b/Makefile
index 763ee1d..c8a426b 100755
--- a/Makefile
+++ b/Makefile
@@ -8,9 +8,8 @@ EXE =
 RM = rm -f
 CP = cp
 
-GPP = g++
-LD = $(GPP) $(CFLAGS)
-CPP = $(GPP) -c $(CFLAGS) 
+LD = $(CXX) $(CFLAGS)
+CPP = $(CXX) -c $(CFLAGS) 
 
 all: pilercr
 
-- 
2.24.0
