--- src/loadmspectrum.cpp
+++ src/loadmspectrum.cpp
@@ -687,7 +687,7 @@
 				while(*pValue != '\0' && !isspace(*pValue))	{
 					pValue++;
 				}
-				if(pValue != '\0' && strlen(pValue) > 2)	{
+				if(*pValue != '\0' && strlen(pValue) > 2)	{
 					specCurrent.m_strDescription = pValue;
 				}
 				specCurrent.m_vMI.push_back(miCurrent);
@@ -878,7 +878,7 @@
 				while(*pValue != '\0' && !isspace(*pValue))	{
 					pValue++;
 				}
-				if(pValue != '\0' && strlen(pValue) > 2)	{
+				if(*pValue != '\0' && strlen(pValue) > 2)	{
 					specCurrent.m_strDescription = pValue;
 				}
 				bFirst = false;
