diff --git a/src/Makefile b/src/Makefile
index 204296e..38da8f9 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -7,16 +7,16 @@ LINKCC = $(CXX)
 
 #CXXFLAGS denotes flags for the C++ compiler
 
-CXX = g++
+CXX = $(CXX)
 
 #uncomment this line if you are using gcc 4.x
-CXXFLAGS = -O2 -DGCC4_3
+CXXFLAGS = -O3 -DGCC4_3 -I$(PREFIX)/include -Wno-conversion-null -Wno-unused-result -Wno-register -Wno-aggressive-loop-optimizations
 #CXXFLAGS = -O2 -DGCC4_3 -DX_P3
 #CXXFLAGS = -O2 -DGCC4_3 -DPLUGGABLE_SCORING -DX_P3
 
 #LDFLAGS = -lpthread -L/usr/lib -lm /usr/lib64/libexpat.a -static-libstdc++
 #LDFLAGS = -lpthread -L/usr/lib -lm /usr/lib64/libexpat.a
-LDFLAGS = -lpthread -L/usr/lib -lm -lexpat
+LDFLAGS = -pthread -L$(PREFIX)/lib -lm -lexpat
 SRCS := $(wildcard *.cpp)
 OBJS := $(patsubst %.cpp,%.o,$(wildcard *.cpp))
 DEPS := $(patsubst %.o,%.d,$(OBJS))
