--- src/loadmspectrum.cpp
+++ src/loadmspectrum.cpp
@@ -687,7 +687,7 @@
 				while(*pValue != '\0' && !isspace(*pValue))	{
 					pValue++;
 				}
-				if(pValue != '\0' && strlen(pValue) > 2)	{
+				if(*pValue != '\0' && strlen(pValue) > 2)	{
 					specCurrent.m_strDescription = pValue;
 				}
 				specCurrent.m_vMI.push_back(miCurrent);
@@ -878,7 +878,7 @@
 				while(*pValue != '\0' && !isspace(*pValue))	{
 					pValue++;
 				}
-				if(pValue != '\0' && strlen(pValue) > 2)	{
+				if(*pValue != '\0' && strlen(pValue) > 2)	{
 					specCurrent.m_strDescription = pValue;
 				}
 				bFirst = false;
--- src/saxhandler.cpp
+++ src/saxhandler.cpp
@@ -316,7 +316,7 @@
 			while(*pValue != '\0' && a < m_peaksCount)	{
 				while(*pValue != '\0' && isspace(*pValue))
 					pValue++;
-				if(pValue == '\0')
+				if(*pValue == '\0')
 					break;
 				m_vfM.push_back((float)atof(pValue));
 				while(*pValue != '\0' && !isspace(*pValue))
@@ -328,7 +328,7 @@
 			while(*pValue != '\0' && a < m_peaksCount)	{
 				while(*pValue != '\0' && isspace(*pValue))
 					pValue++;
-				if(pValue == '\0')
+				if(*pValue == '\0')
 					break;
 				m_vfI.push_back((float)atof(pValue));
 				while(*pValue != '\0' && !isspace(*pValue))
