diff --git a/src/loadmspectrum.cpp b/src/loadmspectrum.cpp
index 8a3b8d4..7903a17 100644
--- a/src/loadmspectrum.cpp
+++ b/src/loadmspectrum.cpp
@@ -687,7 +687,7 @@ bool loadpkl::get(mspectrum &_m)
 				while(*pValue != '\0' && !isspace(*pValue))	{
 					pValue++;
 				}
-				if(pValue != '\0' && strlen(pValue) > 2)	{
+				if(*pValue != '\0' && strlen(pValue) > 2)	{
 					specCurrent.m_strDescription = pValue;
 				}
 				specCurrent.m_vMI.push_back(miCurrent);
@@ -878,7 +878,7 @@ bool loaddta::get(mspectrum &_m)
 				while(*pValue != '\0' && !isspace(*pValue))	{
 					pValue++;
 				}
-				if(pValue != '\0' && strlen(pValue) > 2)	{
+				if(*pValue != '\0' && strlen(pValue) > 2)	{
 					specCurrent.m_strDescription = pValue;
 				}
 				bFirst = false;
diff --git a/src/saxhandler.cpp b/src/saxhandler.cpp
index 7167565..83c2a9d 100644
--- a/src/saxhandler.cpp
+++ b/src/saxhandler.cpp
@@ -316,7 +316,7 @@ void SAXSpectraHandler::pushPeaks(bool bM /*= true*/, bool bI /*= true*/)
 			while(*pValue != '\0' && a < m_peaksCount)	{
 				while(*pValue != '\0' && isspace(*pValue))
 					pValue++;
-				if(pValue == '\0')
+				if(*pValue == '\0')
 					break;
 				m_vfM.push_back((float)atof(pValue));
 				while(*pValue != '\0' && !isspace(*pValue))
@@ -328,7 +328,7 @@ void SAXSpectraHandler::pushPeaks(bool bM /*= true*/, bool bI /*= true*/)
 			while(*pValue != '\0' && a < m_peaksCount)	{
 				while(*pValue != '\0' && isspace(*pValue))
 					pValue++;
-				if(pValue == '\0')
+				if(*pValue == '\0')
 					break;
 				m_vfI.push_back((float)atof(pValue));
 				while(*pValue != '\0' && !isspace(*pValue))
