--- src/makefile.orig	2019-09-28 17:52:05.564806287 +0200
+++ src/makefile	2019-09-28 17:52:52.475714331 +0200
@@ -3,7 +3,7 @@
 all: main
 
 Compiler	= g++
-FLAGS		= -D NDEBUG -O3 -m64
+FLAGS		= -I$(PREFIX)/include -L$(PREFIX)/lib -D NDEBUG -O3 -m64
 LIB		= -lz -lm -lpthread -lstdc++
 SOURCE		= main.cpp GetData.cpp VariantCalling.cpp ReadMapping.cpp AlignmentRescue.cpp ReadAlignment.cpp AlignmentProfile.cpp SamReport.cpp tools.cpp bwt_index.cpp bwt_search.cpp nw_alignment.cpp KmerAnalysis.cpp
 HEADER		= structure.h
