--- src/makefile.orig	2019-10-08 15:25:02.168089499 +0200
+++ src/makefile	2019-10-08 15:25:24.091931743 +0200
@@ -3,7 +3,7 @@
 all: main
 
 CXX		= g++
-FLAGS		= -Wall -D NDEBUG -O3 -m64 -msse4.1 -fPIC
+FLAGS		= -I${PREFIX}/include -L${PREFIX}/lib -Wall -D NDEBUG -O3 -m64 -msse4.1 -fPIC
 LIB		= -lz -lm -lpthread -lstdc++
 HTSLIB		= htslib
 SOURCE		= main.cpp GetData.cpp VariantCalling.cpp ReadMapping.cpp AlignmentRescue.cpp ReadAlignment.cpp AlignmentProfile.cpp SamReport.cpp tools.cpp bwt_index.cpp bwt_search.cpp nw_alignment.cpp ksw2_alignment.cpp KmerAnalysis.cpp
