From 9c78879e39b44892ab20fdd768fc369292e2cea4 Mon Sep 17 00:00:00 2001
From: Ilya Shlyakhter <ilya_shl@alum.mit.edu>
Date: Tue, 7 Jan 2020 16:51:02 -0500
Subject: [PATCH] link rt for static jemalloc

---
 ext/src/jemalloc/src/CMakeLists.txt | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/ext/src/jemalloc/src/CMakeLists.txt b/ext/src/jemalloc/src/CMakeLists.txt
index a3cceb4..30cc73b 100644
--- a/ext/src/jemalloc/src/CMakeLists.txt
+++ b/ext/src/jemalloc/src/CMakeLists.txt
@@ -271,6 +271,10 @@ endif()
 if(JEMALLOC_BUILD_STATIC)
     add_library(${PROJECT_NAME}-static STATIC $<TARGET_OBJECTS:${PROJECT_NAME}>)
 
+    if(clock_gettime_in_librt)
+        target_link_libraries(${PROJECT_NAME}-static rt)
+    endif()
+
     # Set the name of library filename to be "libjemalloc.*"
     # instead of their full CMake libary names of libjemalloc-static.a, etc.
     set_target_properties(${PROJECT_NAME}-static PROPERTIES
-- 
2.20.1

