From 8b1c3af518a8d7f961259174c2a50237b9f28534 Mon Sep 17 00:00:00 2001
From: Ilya Shlyakhter <ilya_shl@alum.mit.edu>
Date: Mon, 6 Jan 2020 14:41:42 -0500
Subject: [PATCH] turn off JEMalloc

---
 src/cmake/options.cmake | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/cmake/options.cmake b/src/cmake/options.cmake
index 20357f5..2512a88 100644
--- a/src/cmake/options.cmake
+++ b/src/cmake/options.cmake
@@ -73,7 +73,7 @@ if (SPADES_USE_TCMALLOC)
     set(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
   endif()
 endif()
-option(SPADES_USE_JEMALLOC "Link spades with JEMalloc" ON)
+option(SPADES_USE_JEMALLOC "Link spades with JEMalloc" OFF)
 if (SPADES_USE_JEMALLOC)
   if (POLICY CMP0077)
     cmake_policy(SET CMP0077 NEW)
-- 
2.20.1

