From 39f24e863713d170427d0b53741fe8bf0e295c90 Mon Sep 17 00:00:00 2001
From: Anders Goncalves da Silva <andersgs@gmail.com>
Date: Wed, 3 Feb 2021 20:30:52 -0800
Subject: Update verbosity

---
 spades_compile.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/spades_compile.sh b/spades_compile.sh
index 8d82c75..095a173 100755
--- a/spades_compile.sh
+++ b/spades_compile.sh
@@ -222,7 +222,7 @@ fi

 cd "$WORK_DIR"
 cmake -G "Unix Makefiles" -DCMAKE_INSTALL_PREFIX="$PREFIX" $ADDITIONAL_FLAGS "$BASEDIR/src"
-make -j $AMOUNT_OF_THREADS
+make -j $AMOUNT_OF_THREADS VERBOSE=1
 make install
 cd "$PREFIX"

--
2.24.3 (Apple Git-128)