From 776f2467bf37bc051a1b8863c1c9b32615517464 Mon Sep 17 00:00:00 2001
From: Ilya Shlyakhter <ilya_shl@alum.mit.edu>
Date: Fri, 16 Aug 2019 13:07:46 -0400
Subject: [PATCH] fix cpu count

---
 spades_compile.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/spades_compile.sh b/spades_compile.sh
index fe33034..16cc01c 100755
--- a/spades_compile.sh
+++ b/spades_compile.sh
@@ -20,7 +20,6 @@ mkdir -p "$BASEDIR/$BUILD_DIR"
 set -e
 cd "$BASEDIR/$BUILD_DIR"
 cmake -G "Unix Makefiles" -DCMAKE_INSTALL_PREFIX="$PREFIX" $* "$BASEDIR/src"
-make -j 8
+make -j${CPU_COUNT} VERBOSE=1
 make install
 cd $PREFIX
-- 
2.20.1

