diff --git a/spades_compile.sh b/spades_compile.sh
index e13ca4f..e068098 100755
--- a/spades_compile.sh
+++ b/spades_compile.sh
@@ -37,10 +37,10 @@ check_whether_OPTARG_is_an_integer() {
 }
 
 # return the argument first character
-str_head() { echo "$(expr substr "$1" 1 1)"; }
+str_head() { echo "${1::1}"; }
 
 # return the argument without the first character
-str_tail() { echo "$(expr substr "$1" 2 $((${#1}-1)))"; }
+str_tail() { echo "${1:1}"; }
 
 print_help() {
   echo
@@ -221,11 +221,34 @@ else
   ADDITIONAL_FLAGS="$ADDITIONAL_FLAGS -DSPADES_BUILD_INTERNAL=OFF"
 fi
 
-cd "$WORK_DIR"
-cmake -G "Unix Makefiles" -DCMAKE_INSTALL_PREFIX="$PREFIX" $ADDITIONAL_FLAGS "$BASEDIR/src"
-make -j $AMOUNT_OF_THREADS
-make install
-cd "$PREFIX"
+case $(uname -m) in
+    aarch64)
+	export CXXFLAGS="${CXXFLAGS} -march=armv8-a"
+	;;
+    arm64)
+	export CXXFLAGS="${CXXFLAGS} -march=armv8.4-a"
+	;;
+    x86_64)
+	export CXXFLAGS="${CXXFLAGS} -march=x86-64-v3"
+	;;
+esac
+
+if [[ `uname -s` == "Darwin" ]]; then
+	export CONFIG_ARGS="-DCMAKE_FIND_FRAMEWORK=NEVER -DCMAKE_FIND_APPBUNDLE=NEVER"
+	export CFLAGS="${CFLAGS} -Wunused-command-line-argument"
+	export CXXFLAGS="${CXXFLAGS} -D_LIBCPP_DISABLE_AVAILABILITY"
+else
+	export CONFIG_ARGS=""
+fi
+
+cd "${WORK_DIR}"
+cmake -S "${BASEDIR}/src" -B . -G "Unix Makefiles" -DCMAKE_INSTALL_PREFIX="${PREFIX}" \
+	-DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_COMPILER="${CXX}" \
+	-DCMAKE_CXX_FLAGS="${CXXFLAGS} -O3" -DCMAKE_C_COMPILER="${CC}" \
+	-DCMAKE_C_FLAGS="${CFLAGS} -O3" -DSPADES_USE_NCBISDK=ON -DSPADES_ENABLE_PROJECTS="all" \
+	-Wno-dev -Wno-deprecated --no-warn-unused-cli "${ADDITIONAL_FLAGS}" \
+	"${CONFIG_ARGS}"
+cmake --build . --clean-first --target install -j "${CPU_COUNT}"
 
 if [ $RUN_TESTS = "y" ]; then
   cd "$BASEDIR"
