--- SPAdes-3.15.5.orig/spades_compile.sh	2022-07-11 15:49:32.000000000 -0600
+++ SPAdes-3.15.5/spades_compile.sh	2022-11-02 13:03:47.000000000 -0600
@@ -36,10 +36,10 @@
 }

 # return the argument first character
-str_head() { echo "$(expr substr "$1" 1 1)"; }
+str_head() { echo "${1::1}"; }

 # return the argument without the first character
-str_tail() { echo "$(expr substr "$1" 2 $((${#1}-1)))"; }
+str_tail() { echo "${1:1}"; }

 print_help() {
   echo
