From 0f95dac3ff1108d5bab8047385df8213a33bc277 Mon Sep 17 00:00:00 2001
From: Ilya Shlyakhter <ilya_shl@alum.mit.edu>
Date: Fri, 16 Aug 2019 18:40:04 -0400
Subject: [PATCH 2/2] disable parallel libstdc++

---
 src/cmake/flags.cmake | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/cmake/flags.cmake b/src/cmake/flags.cmake
index a3f8156..46998d4 100644
--- a/src/cmake/flags.cmake
+++ b/src/cmake/flags.cmake
@@ -6,7 +6,7 @@ if (OPENMP_FOUND)
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
   set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${OpenMP_EXE_LINKER_FLAGS}")
 # Use parallel libstdc++ if possible
-  add_definitions(-DUSE_GLIBCXX_PARALLEL=1)
+#  add_definitions(-DUSE_GLIBCXX_PARALLEL=1)
 else ()
   if (NOT "${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang")
     message(FATAL_ERROR "SPAdes requires OpenMP to be available")
-- 
2.20.1

