--- src/cmake/modules/FindLibExslt.cmake
+++ src/cmake/modules/FindLibExslt.cmake
@@ -51,7 +51,7 @@
 if(PC_LIBEXSLT_VERSION)
   set(LIBEXSLT_VERSION_STRING ${PC_LIBEXSLT_VERSION})
 elseif(LIBEXSLT_INCLUDE_DIR AND EXISTS "${LIBEXSLT_INCLUDE_DIR}/libexslt/exsltconfig.h")
-  file(STRINGS "${LIBEXSLT_INCLUDE_DIR}/libexslt/xsltconfig.h" libexslt_version_str
+  file(STRINGS "${LIBEXSLT_INCLUDE_DIR}/libexslt/exsltconfig.h" libexslt_version_str
     REGEX "^#define[\t ]+LIBEXSLT_DOTTED_VERSION[\t ]+\".*\"")
 
   string(REGEX REPLACE "^#define[\t ]+LIBEXSLT_DOTTED_VERSION[\t ]+\"([^\"]*)\".*" "\\1"
