--- src/c++/lib/alignment/TemplateBuilder.cpp.orig	2020-05-20 21:44:11.573224586 +0100
+++ src/c++/lib/alignment/TemplateBuilder.cpp	2020-05-20 21:17:09.244032702 +0100
@@ -71,7 +71,7 @@
     , scatterRepeats_(scatterRepeats)
     , rescueShadows_(rescueShadows)
     , dodgyAlignmentScore_(dodgyAlignmentScore)
-    , anomalousPairHandicap_(pow10(double(anomalousPairHandicap) / 10.0) - 1.0)//anomalousPairHandicap)
+    , anomalousPairHandicap_(exp10(double(anomalousPairHandicap) / 10.0) - 1.0)//anomalousPairHandicap)
     , smitWatermanGapsMax_(smitWatermanGapsMax)
     , splitAlignments_(splitAlignments)
     , alignmentCfg_(alignmentCfg)
