--- src/c++/lib/io/FileBufWithReopen.cpp.orig	2020-07-24 11:40:39.598063344 +0100
+++ src/c++/lib/io/FileBufWithReopen.cpp	2020-07-24 11:41:20.950206129 +0100
@@ -30,38 +30,38 @@
                         // inspired by /usr/include/c++/4.3.2/fstream
                         // | ios_base Flag combination      |
                         // |binary  in  out  trunc  app     |
-            (0      )   // |                                |
+            (nullptr)   // |                                |
             ("a"    )   // |                         +      |
-            (0      )   // |                   +            |
-            (0      )   // |                   +     +      |
+            (nullptr)   // |                   +            |
+            (nullptr)   // |                   +     +      |
             ("w"    )   // |             +                  |
             ("a"    )   // |             +           +      |
             ("w"    )   // |             +     +            |
-            (0      )   // |             +     +     +      |
+            (nullptr)   // |             +     +     +      |
             ("r"    )   // |         +                      |
             ("a+"   )   // |         +               +      |
-            (0      )   // |         +         +            |
-            (0      )   // |         +         +     +      |
+            (nullptr)   // |         +         +            |
+            (nullptr)   // |         +         +     +      |
             ("r+"   )   // |         +   +                  |
             ("a+"   )   // |         +   +           +      |
             ("w+"   )   // |         +   +     +            |
-            (0      )   // |         +   +     +     +      |
-            (0      )   // |   +                            |
+            (nullptr)   // |         +   +     +     +      |
+            (nullptr)   // |   +                            |
             ("ab"   )   // |   +                     +      |
-            (0      )   // |   +               +            |
-            (0      )   // |   +               +     +      |
+            (nullptr)   // |   +               +            |
+            (nullptr)   // |   +               +     +      |
             ("wb"   )   // |   +         +                  |
             ("ab"   )   // |   +         +           +      |
             ("wb"   )   // |   +         +     +            |
-            (0      )   // |   +         +     +     +      |
+            (nullptr)   // |   +         +     +     +      |
             ("rb"   )   // |   +     +                      |
             ("a+b"  )   // |   +     +               +      |
-            (0      )   // |   +     +         +            |
-            (0      )   // |   +     +         +     +      |
+            (nullptr)   // |   +     +         +            |
+            (nullptr)   // |   +     +         +     +      |
             ("r+b"  )   // |   +     +   +                  |
             ("a+b"  )   // |   +     +   +           +      |
             ("w+b"  )   // |   +     +   +     +            |
-            (0      )   // |   +     +   +     +     +      |
+            (nullptr)   // |   +     +   +     +     +      |
             ;

 } // namespace io
