--- src/configure.orig	2020-05-20 21:40:21.259601460 +0100
+++ src/configure	2020-05-20 21:41:13.987993770 +0100
@@ -270,8 +270,7 @@
 # after they appear there.
 isaac_cmake_install_dir="${isaac_build_dir}/bootstrap_cmake"
 if [ "x${isaac_cmake}"       == "x" ] ; then 
-    "${isaac_bootstrap_dir}"/installCmake.sh "${isaac_redist_dir}" "${isaac_cmake_install_dir}" "${isaac_parallel}"
-    cmake_install_exit_code="$?"
+    cmake_install_exit_code="1"
     if [ "$cmake_install_exit_code" == "1" ]; then
         isaac_cmake=cmake
         echo "Using existing `which cmake`"
