diff --git a/setup.py b/setup.py
index 819e30c..832c108 100644
--- a/setup.py
+++ b/setup.py
@@ -21,9 +21,9 @@ setuptools.setup(
         "python-levenshtein>=0.12.0",
         "scipy>=1.1.0",
         "multiprocess>=0.70.6.1",
-        "umi_tools==1.0.0",
-        "pytest==4.1.0",
-        "pytest-dependency==0.4.0",
+        "umi_tools>=1.1.5",
+        "pytest>=4.1.0",
+        "pytest-dependency>=0.4.0",
         "pandas>=0.23.4",
         "pybktree==1.1",
     ],
