--- setup.py.org	2021-10-01 00:42:34.157107978 -0500
+++ setup.py	2021-10-01 00:42:52.500079923 -0500
@@ -22,8 +22,8 @@
         "scipy>=1.1.0",
         "multiprocess>=0.70.6.1",
         "umi_tools==1.0.0",
-        "pytest==4.1.0",
-        "pytest-dependency==0.4.0",
+        "pytest>=4.1.0",
+        "pytest-dependency>=0.4.0",
         "pandas>=0.23.4",
         "pybktree==1.1",
     ],
--- CITE_seq_Count.egg-info/requires.txt.org	2021-10-01 00:43:01.952065465 -0500
+++ CITE_seq_Count.egg-info/requires.txt	2021-10-01 00:43:10.664052138 -0500
@@ -2,7 +2,7 @@
 scipy>=1.1.0
 multiprocess>=0.70.6.1
 umi_tools==1.0.0
-pytest==4.1.0
-pytest-dependency==0.4.0
+pytest>=4.1.0
+pytest-dependency>=0.4.0
 pandas>=0.23.4
 pybktree==1.1
