--- a/mydbfinder.py
+++ b/mydbfinder.py
@@ -483,7 +483,7 @@ if __name__ == '__main__':
     parser.add_argument("-p", "--databasePath",
                         dest="db_path",
                         help="Path to the databases",
-                        default='')
+                        default=os.getenv("MyDbFinder_DB"))
     parser.add_argument("-d", "--databases",
                         dest="databases",
                         help="Databases chosen to search in - if none are \
