diff --git a/c++/scripts/common/impl/create_flat_makefile.sh b/c++/scripts/common/impl/create_flat_makefile.sh
index e6d3374..e063c44 100755
--- a/c++/scripts/common/impl/create_flat_makefile.sh
+++ b/c++/scripts/common/impl/create_flat_makefile.sh
@@ -190,7 +190,8 @@ if test "$buildptb" = "yes"; then
   for dep in $ptbdep; do
     COMMON_Exec cd $builddir
     COMMON_Exec cd $dep
-    COMMON_Exec make
+
+    COMMON_Exec ${MAKE} CC="${CC}" CXX="${CXX}" CFLAGS="${CFLAGS}" CXXFLAGS="${CXXFLAGS}"
   done
   COMMON_Exec cd $builddir
   ptb="./build-system/project_tree_builder/$ptbname"
@@ -300,7 +301,7 @@ if test "$dtfound" = "yes"; then
       echo "Generating objects source code. Please wait."
       echo "**********************************************************************"
       echo make -f $solution all_files
-      make -f $solution all_files >/dev/null 2>&1
+      ${MAKE} -f $solution all_files CC="${CC}" CXX="${CXX}" CFLAGS="${CFLAGS}" CXXFLAGS="${CXXFLAGS}" >/dev/null 2>&1
     fi
   fi
 fi
