diff --git a/c++/src/build-system/configure b/c++/src/build-system/configure
index a827d9a..564878c 100755
--- a/c++/src/build-system/configure
+++ b/c++/src/build-system/configure
@@ -27738,7 +27738,7 @@ $as_echo "$as_me: WARNING: $message" >&2;}  ;;
          case "$host_os:$host_cpu:$bit64_sfx" in
             darwin*:*86*:64 ) vdb_libpath=mac/$vdb_mode/x86_64/lib   ;;
             darwin*:*86*:* )  vdb_libpath=mac/$vdb_mode/fat86/lib    ;;
-            darwin*:aarch64* ) vdb_libpath=mac/$vdb_mode/aarch64/lib ;;
+            darwin*:a* ) vdb_libpath=mac/$vdb_mode/arm64/lib ;;
             linux*:*86*:64 )  vdb_libpath=linux/$vdb_mode/x86_64/lib ;;
             linux*:aarch64* ) vdb_libpath=linux/$vdb_mode/aarch64/lib ;;
             * )
