diff --git a/Makefile.am b/Makefile.am
index 9a885ca..22d95e4 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -171,10 +171,10 @@ deltafiltername = $(shell echo delta-filter | sed '$(transform)')
 repeatmatchname = $(shell echo repeat-match | sed '$(transform)')
 
 
-script_inst_subst  = sed -e 's,[@]PERL[@],$(PERL),g' \
-                         -e 's,[@]BIN_DIR[@],$(bindir),g' \
-                         -e 's,[@]LIB_DIR[@],$(script_libdir),g' \
-                         -e 's,[@]LIBEXEC_DIR[@],$(pkglibexecdir),g' \
+script_inst_subst = sed -e 's,[@]PERL[@],$(PERL),g' \
+                         -e -e 's,"[@]BIN_DIR[@]",\x27$(bindir)\x27,g' \
+                         -e -e 's,"[@]LIB_DIR[@]",\x27$(script_libdir)\x27,g' \
+                         -e 's,"[@]LIBEXEC_DIR[@]",\x27$(pkglibexecdir)\x27,g' \
                          -e 's,[@]MUMMER[@],$(mummername),g' \
                          -e 's,[@]MGAPS[@],$(mgapsname),g' \
                          -e 's,[@]PRENUC[@],$(prenucname),g' \
@@ -188,9 +188,9 @@ script_inst_subst  = sed -e 's,[@]PERL[@],$(PERL),g' \
                          -e 's,[@]DELTAFILTER[@],$(deltafiltername),g' \
                          -e 's,[@]REPEATMATCH[@],$(repeatmatchname),g'
 script_local_subst = sed -e 's,[@]PERL[@],$(PERL),g' \
-                         -e 's,[@]BIN_DIR[@],$(abs_builddir),g' \
-                         -e 's,[@]LIB_DIR[@],$(abs_srcdir)/scripts,g' \
-                         -e 's,[@]LIBEXEC_DIR[@],$(abs_builddir),g' \
+                         -e 's,"[@]BIN_DIR[@]",\x27$(abs_builddir)\x27,g' \
+                         -e 's,"[@]LIB_DIR[@]",\x27$(abs_srcdir)/scripts\x27,g' \
+                         -e 's,"[@]LIBEXEC_DIR[@]",\x27$(abs_builddir)\x27,g' \
                          -e 's,[@]MUMMER[@],mummer,g' \
                          -e 's,[@]MGAPS[@],mgaps,g' \
                          -e 's,[@]PRENUC[@],prenuc,g' \
