--- mummer-4.0.0beta2/Makefile.in	2017-10-13 19:44:10.000000000 -0500
+++ mummer4/Makefile.in	2018-08-23 16:04:59.272011109 -0500
@@ -959,16 +959,16 @@
 
 # Rules to generate the scripts, in .libs and in build directory.
 script_inst_subst = sed -e 's,[@]PERL[@],$(PERL),g' \
-                       -e 's,[@]BIN_DIR[@],$(bindir),g' \
-                       -e 's,[@]LIB_DIR[@],$(script_libdir),g' \
-	               -e 's,[@]LIBEXEC_DIR[@],$(pkglibexecdir),g' \
-		       -e 's,[@]GNUPLOT_EXE[@],$(GNUPLOT),g'
+                       -e 's,"[@]BIN_DIR[@]",\x27$(bindir)\x27,g' \
+                       -e 's,"[@]LIB_DIR[@]",\x27$(script_libdir)\x27,g' \
+                       -e 's,"[@]LIBEXEC_DIR[@]",\x27$(pkglibexecdir)\x27,g' \
+                       -e 's,"[@]GNUPLOT_EXE[@]",\x27$(GNUPLOT)\x27,g'
 
 script_local_subst = sed -e 's,[@]PERL[@],$(PERL),g' \
-                       -e 's,[@]BIN_DIR[@],$(abs_builddir),g' \
-                       -e 's,[@]LIB_DIR[@],$(abs_srcdir)/scripts,g' \
-	               -e 's,[@]LIBEXEC_DIR[@],$(abs_builddir),g' \
-                       -e 's,[@]GNUPLOT_EXE[@],$(GNUPLOT),g'
+                       -e 's,"[@]BIN_DIR[@]",\x27$(abs_builddir)\x27,g' \
+                       -e 's,"[@]LIB_DIR[@]",\x27$(abs_srcdir)/scripts\x27,g' \
+                       -e 's,"[@]LIBEXEC_DIR[@]",\x27$(abs_builddir)\x27,g' \
+                       -e 's,"[@]GNUPLOT_EXE[@]",\x27$(GNUPLOT)\x27,g'
 
 
 # Supporting modules
