diff --git a/Makefile.am b/Makefile.am
index c823127..1089894 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -154,14 +154,11 @@ CLEANFILES += $(libs_scripts) $(all_scripts)
 script_inst_subst  = sed -e 's,[@]PERL[@],$(PERL),g' \
                        -e 's,[@]BIN_DIR[@],$(bindir),g' \
                        -e 's,[@]LIB_DIR[@],$(script_libdir),g' \
-	               -e 's,[@]LIBEXEC_DIR[@],$(pkglibexecdir),g' \
-		       -e 's,[@]GNUPLOT_EXE[@],$(GNUPLOT),g'
+                       -e 's,[@]LIBEXEC_DIR[@],$(pkglibexecdir),g'
 script_local_subst = sed -e 's,[@]PERL[@],$(PERL),g' \
                        -e 's,[@]BIN_DIR[@],$(abs_builddir),g' \
                        -e 's,[@]LIB_DIR[@],$(abs_srcdir)/scripts,g' \
-	               -e 's,[@]LIBEXEC_DIR[@],$(abs_builddir),g' \
-                       -e 's,[@]GNUPLOT_EXE[@],$(GNUPLOT),g'
-
+                       -e 's,[@]LIBEXEC_DIR[@],$(abs_builddir),g'
 .libs/%: scripts/%.pl
 	@mkdir -p $(dir $@) && $(script_inst_subst) < $< > $@
 %: scripts/%.pl
diff --git a/configure.ac b/configure.ac
index 969ab85..600e2ee 100644
--- a/configure.ac
+++ b/configure.ac
@@ -33,10 +33,6 @@ AS_IF([test "x$enable_openmp" != "xno"],
 AC_ARG_VAR([YAGGO], [Yaggo switch parser generator])
 AS_IF([test "x$YAGGO" = "x"], [AC_PATH_PROG([YAGGO], [yaggo], [false])])
 
-# Check for gnuplot
-AC_ARG_VAR([GNUPLOT], [Gnuplot plotting program])
-AS_IF([test "x$GNUPLOT" = "x"], [AC_PATH_PROGS([GNUPLOT], [gnuplot gnuplot5 gnuplot4], [false])])
-
 # Check that type __int128 is supported and if the
 AC_ARG_WITH([int128],
             [AS_HELP_STRING([--with-int128], [enable int128])],
diff --git a/scripts/mummerplot.pl b/scripts/mummerplot.pl
index ec4d2a5..6798de0 100644
--- a/scripts/mummerplot.pl
+++ b/scripts/mummerplot.pl
@@ -23,7 +23,7 @@ use IO::Socket;
 
 my $BIN_DIR     = "@BIN_DIR@";
 my $LIB_DIR     = "@LIB_DIR@";
-my $GNUPLOT_EXE = "@GNUPLOT_EXE@";
+my $GNUPLOT_EXE = "gnuplot";
 
 
 #================================================================= Globals ====#
