diff --git a/setup.py b/setup.py
index ed97e7a..95938a3 100644
--- a/setup.py
+++ b/setup.py
@@ -331,7 +331,7 @@ def build_parasail(libname):
             print("Prepending {} to PATH".format(newpath))
             os.environ['PATH'] = newpath + os.pathsep + os.environ['PATH']
             if platform.system() == "Darwin" and 'M4' not in os.environ:
-                os.environ['M4'] = '/usr/bin/m4'
+                os.environ['M4'] = os.environ['BUILD_PREFIX'] + '/bin/m4'
             print("PATH={}".format(os.environ['PATH']))
             if not build_autotools():
                 if not build_autotools(True):
