diff --git a/setup.py b/setup.py
index 4465048..f5822c4 100644
--- a/setup.py
+++ b/setup.py
@@ -235,6 +235,8 @@ def build_autotools(patch_m4=False):
         else:
             print("Expanding {}".format(tarball))
             untar(tarball)
+            subprocess.run('cp -rf ${BUILD_PREFIX}/share/gnuconfig/config.* build-aux/',
+                shell=True)
         if os.path.exists(binary):
             print("{} already exists! Skipping build.".format(binary))
         else:
