diff --git a/setup.py b/setup.py
index ed97e7a..17f1ea1 100644
--- a/setup.py
+++ b/setup.py
@@ -202,10 +202,10 @@ def build_autotools(patch_m4=False):
     if not os.path.exists(top):
         os.mkdir(top)
     # we know these versions to work
-    tools = [('m4', '1.4.17'),
-            ('autoconf', '2.69'),
-            ('automake', '1.15.1'),
-            ('libtool', '2.4.6')]
+    tools = [('m4', '1.4.18'),
+            ('autoconf', '2.71'),
+            ('automake', '1.17'),
+            ('libtool', '2.4.7')]
     for tool,version in tools:
         os.chdir(top)
         tdir = '{}-{}'.format(tool, version)
@@ -331,7 +331,7 @@ def build_parasail(libname):
             print("Prepending {} to PATH".format(newpath))
             os.environ['PATH'] = newpath + os.pathsep + os.environ['PATH']
             if platform.system() == "Darwin" and 'M4' not in os.environ:
-                os.environ['M4'] = '/usr/bin/m4'
+                os.environ['M4'] = os.environ['BUILD_PREFIX'] + '/bin/m4'
             print("PATH={}".format(os.environ['PATH']))
             if not build_autotools():
                 if not build_autotools(True):
