diff --git setup.py setup.py
index d8344f8..961803d 100644
--- setup.py
+++ setup.py
@@ -158,9 +158,9 @@ def run_autoreconf(root):
     all_good = True
     tools = ['perl', 'm4', 'autoconf', 'automake', 'libtoolize', 'autoreconf']
     if platform.system() == "Darwin":
-        tools[4] = 'glibtoolize'
-        os.environ['LIBTOOL'] = 'glibtool'
-        os.environ['LIBTOOLIZE'] = 'glibtoolize'
+        tools[4] = 'libtoolize'
+        os.environ['LIBTOOL'] = 'libtool'
+        os.environ['LIBTOOLIZE'] = 'libtoolize'
     for tool in tools:
         try:
             output = subprocess.check_output([tool, '--version'],
