--- gzstream/Makefile
+++ gzstream/Makefile
@@ -35,8 +35,8 @@
 # CXX      = CC -n32 -LANG:std   # for SGI Irix 6.5, MIPSpro CC version 7.30
 CXX      = g++   # for Linux RedHat 6.1, g++ version 2.95.2
 
-CPPFLAGS = -I. -O
-LDFLAGS  = -L. -lgzstream -lz
+CPPFLAGS = $(CXXFLAGS) -I. -O
+LDFLAGS  += -L. -lgzstream -lz
 AR       = ar cr
 
 # ----------------------------------------------------------------------------
