--- libgab/gzstream/Makefile	2020-09-03 20:37:06.000000000 +0200
+++ libgab/gzstream/Makefile	2020-09-03 22:00:28.000000000 +0200
@@ -33,10 +33,9 @@
 # ----------------------------------------------------------------------------
 
 # CXX      = CC -n32 -LANG:std   # for SGI Irix 6.5, MIPSpro CC version 7.30
-CXX      = g++   # for Linux RedHat 6.1, g++ version 2.95.2
 
-CPPFLAGS = -I. -O
-LDFLAGS  = -L. -lgzstream -lz
+CPPFLAGS = $(CXXFLAGS) -I. -O
+LDFLAGS  += -L. -lgzstream -lz
 AR       = ar cr
 
 # ----------------------------------------------------------------------------
