--- libgab/Makefile	2020-04-09 14:43:24.000000000 +0200
+++ libgab/Makefile	2020-09-04 08:32:43.000000000 +0200
@@ -8,18 +8,18 @@
 	BAMTOOLSINCAPI=${BAMTOOLSINC}
 	BAMTOOLSINCSHARED=${BAMTOOLSINC}
 else
-	BAMTOOLSINCAPI=${BAMTOOLS}/src/
-	BAMTOOLSINCSHARED=${BAMTOOLS}/src/
+	BAMTOOLSINCAPI=${BAMTOOLS}/
+	BAMTOOLSINCSHARED=${BAMTOOLS}/
 endif
 
 ifdef BAMTOOLSLIB
 	BAMTOOLSLIBOBJ=${BAMTOOLSLIB}/libbamtools.a 
 else
-	BAMTOOLSLIBOBJ=${BAMTOOLS}/build/src/api/libbamtools.a 
+	BAMTOOLSLIBOBJ=${BAMTOOLS}/lib/libbamtools.a 
 endif
 
 
-CXX      = g++ #-g  -pg
+
 CXXFLAGS += -Wall  -lm -O3 -I. -Igzstream/ -I${BAMTOOLSINCAPI} -I${BAMTOOLSINCSHARED} 
 LDLIBS   += -lz
 LDLIBSHTS = ../lib/samtools/bedidx.o ../lib/htslib/libhts.a ../lib/samtools/libbam.a ../lib/samtools/libst.a 
