diff --git a/src/Makefile b/src/Makefile
index 77d3cbf3..d1bbee15 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -1,8 +1,8 @@
 .PHONY: all clean debug profile
 
-CXXFLAGS += -Wall -Wextra -pthread -std=c++11 -g
+CXXFLAGS += -Wall -Wextra -pthread -O3 -std=c++14 -g -march=x86-64-v3
 CXXFLAGS += -Wno-missing-field-initializers
-LDFLAGS += -pthread -std=c++11 -rdynamic
+LDFLAGS += -pthread -std=c++14 -rdynamic
 
 MODULES_BIN := ${BIN_DIR}/flye-modules
 
