diff --git a/Cargo.toml b/Cargo.toml
index e8d72fe..aea7af5 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -61,7 +61,7 @@ prost = "0.13.4"
 quick_cache = "0.6.12"
 rand = "0.9"
 rayon = "1.10.0"
-rocksdb = { version = "0.22.0", features = ["multi-threaded-cf"] }
+rocksdb = { version = "0.22.0", features = ["multi-threaded-cf", "jemalloc"] }
 rustc-hash = "2.1.0"
 seqrepo = { version = "0.10.2", features = ["cached"] }
 serde_json = "1.0"
