diff --git i/setup.py w/setup.py
index c815a31..fa18eba 100644
--- i/setup.py
+++ w/setup.py
@@ -101,7 +101,7 @@ setup(name='cell2cell',
                         'tensorly == 0.8.1',
                         'kneed',
                         'scanpy',
-                        'gseapy == 1.0.3'
+                        'gseapy >= 1.0.3'
                         ],
       classifiers=classifiers,
       entry_points={},
@@ -109,4 +109,4 @@ setup(name='cell2cell',
       cmdclass={'install': CustomInstallCommand,
                 'develop': CustomDevelopCommand,
                 'egg_info': CustomEggInfoCommand, },
-      zip_safe=False)
\ No newline at end of file
+      zip_safe=False)
