diff --git a/Makefile b/Makefile
index fd6e888..0836d19 100644
--- a/Makefile
+++ b/Makefile
@@ -18,19 +18,19 @@ ifdef P
 	PROFILE=-pg -no-pie # for bug in gprof.
 endif
 
-CXX = g++ -std=c++11
-CC = gcc -std=gnu11
-LD= g++ -std=c++11
+CXX = $(CXX) -std=c++11
+CC = $(CC) -std=gnu11
+LD= $(CXX) -std=c++11
 
 LOC_INCLUDE=include
 LOC_SRC=src
 OBJDIR=obj
 
-CXXFLAGS += -Wall $(DEBUG) $(PROFILE) $(OPT) $(ARCH) -m64 -I. -I$(LOC_INCLUDE)
+CXXFLAGS += -Wall $(DEBUG) $(PROFILE) $(OPT) $(ARCH) -march=x86-64-v3 -m64 -I. -I$(LOC_INCLUDE) -I$(PREFIX)/include
 
-CFLAGS += -Wall $(DEBUG) $(PROFILE) $(OPT) $(ARCH) -m64 -I. -I$(LOC_INCLUDE)
+CFLAGS += -Wall $(DEBUG) $(PROFILE) $(OPT) $(ARCH) -m64 -I. -I$(LOC_INCLUDE) -I$(PREFIX)/include
 
-LDFLAGS += $(DEBUG) $(PROFILE) $(OPT) -lpthread -lboost_system \
+LDFLAGS += $(DEBUG) $(PROFILE) $(OPT) -L$(PREFIX)/lib -pthread -lboost_system \
 -lboost_thread -lm -lbz2 -lz -lrt
 
 #
diff --git a/include/kmer.h b/include/kmer.h
index 2dd90ac..488b47c 100644
--- a/include/kmer.h
+++ b/include/kmer.h
@@ -12,6 +12,7 @@
 #define _KMER_H_
 
 #include <stdio.h>
+#include <cstdint>
 #include <string>
 #include <unordered_set>
 
diff --git a/include/SqueakrFS.h b/include/SqueakrFS.h
index 769f7a9..ca14af8 100644
--- a/include/SqueakrFS.h
+++ b/include/SqueakrFS.h
@@ -9,6 +9,7 @@
 #define __SQUEAKR_FILESYSTEM_HPP__
 
 #include <vector>
+#include <cstdint>
 #include <string>
 
 namespace squeakr {
