--- a/kent/src/hg/lib/straw/makefile    2022-10-26 12:00:00.000000000 +0100
+++ b/kent/src/hg/lib/straw/makefile    2022-10-26 12:00:00.000000000 +0100
@@ -1,4 +1,4 @@
-KENT_INC=-I../../../inc
+KENT_INC=-I../../../inc -I$(PREFIX)/include

 straw: straw.o cStraw.o
        ld -r -o ../straw.o straw.o cStraw.o
--- a/kent/src/utils/expMatrixToBarchartBed/expMatrixToBarchartBed      2018-06-06 02:23:56.000000000 +0200
+++ b/kent/src/utils/expMatrixToBarchartBed/expMatrixToBarchartBed      2018-06-06 02:23:56.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python2.7
+#!/usr/bin/env python
 # expMatrixToBarchartBed
 """
 Generate a barChart bed6+5 file from a matrix, meta data, and coordinates.
