diff --git a/snapatac2-python/Cargo.toml b/snapatac2-python/Cargo.toml
index 9a0acf0..38b1f44 100644
--- a/snapatac2-python/Cargo.toml
+++ b/snapatac2-python/Cargo.toml
@@ -45,7 +45,7 @@ tikv-jemallocator = {version = "0.5", features = ["disable_initial_exec_tls"]}
 
 [dependencies.pyo3]
 version = "0.21.2"
-features = ["extension-module", "anyhow"]
+features = ["anyhow"]
 
 [lib]
 name = "snapatac2"
