diff --git a/snapatac2-python/pyproject.toml b/snapatac2-python/pyproject.toml
index c90b719..0474a9c 100644
--- a/snapatac2-python/pyproject.toml
+++ b/snapatac2-python/pyproject.toml
@@ -6,6 +6,9 @@ build-backend = "maturin"
 features = ["pyo3/extension-module"]
 python-source = "python"
 module-name = "snapatac2._snapatac2"
+strip = true
+profile = "release"
+locked = true
 
 [project]
 name = "snapatac2"
