diff --git a/bin/titan-gc-cli b/bin/titan-gc
similarity index 88%
rename from bin/titan-gc-cli
rename to bin/titan-gc
index 53b821f..54f42e1 100755
--- a/bin/titan-gc-cli
+++ b/bin/titan-gc
@@ -1,7 +1,7 @@
 #! /bin/bash
-# usage: titan-gc-cli [-h] [-i STR] [--inputs STR] [-o STR] [--outdir STR] [--options STR] [--verbose]
+# usage: titan-gc [-h] [-i STR] [--inputs STR] [-o STR] [--outdir STR] [--options STR] [--verbose]
 #
-# titan-gc-cli - Run Titan GC on a set of samples.
+# titan-gc - Run Titan GC on a set of samples.
 #
 # required arguments:
 #   -i STR, --inputs STR  The JSON file to be used with Cromwell for inputs.
@@ -18,22 +18,22 @@ OPTIONS="0"
 QUIET="0"
 PROFILE="docker"
 CONFIG="0"
-TITAN_PATH=$(which titan-gc-cli | sed 's=bin/titan-gc-cli==')
-TITAN_SHARE=${TITAN_PATH}/share/titan-gc-cli-${TITAN_GC_VERSION}
+TITAN_PATH=$(which titan-gc | sed 's=bin/titan-gc==')
+TITAN_SHARE=${TITAN_PATH}/share/titan-gc-${TITAN_GC_VERSION}
 CROMWELL_JAR=$(which cromwell | sed 's=bin/cromwell=share/cromwell/cromwell.jar=')
 LOG_LEVEL=ERROR
 SINGULARITY_CACHE="${HOME}/.singularity/titan-cache"
 CROMWELL_OPTS="${TITAN_SHARE}/conf/options.json"
 
 version() {
-    echo "titan-gc-cli ${TITAN_GC_VERSION}"
+    echo "titan-gc ${TITAN_GC_VERSION}"
     exit 0
 }
 
 usage() {
-    echo "usage: titan-gc-cli [-h] [-i STR] [--inputs STR] [-o STR] [--outdir STR] [--options STR] [--quiet]"
+    echo "usage: titan-gc [-h] [-i STR] [--inputs STR] [-o STR] [--outdir STR] [--options STR] [--quiet]"
     echo ""
-    echo "titan-gc-cli - Run Titan GC on a set of samples."
+    echo "titan-gc - Run Titan GC on a set of samples."
     echo ""
     echo "required arguments:"
     echo "  -i STR, --inputs STR  The JSON file to be used with Cromwell for inputs."
