diff --git a/bin/titan-gc-cli b/bin/titan-gc-cli
index 8ec0ebd..53b821f 100755
--- a/bin/titan-gc-cli
+++ b/bin/titan-gc-cli
@@ -18,16 +18,15 @@ OPTIONS="0"
 QUIET="0"
 PROFILE="docker"
 CONFIG="0"
-VERSION="1.4.4"
 TITAN_PATH=$(which titan-gc-cli | sed 's=bin/titan-gc-cli==')
-TITAN_SHARE=${TITAN_PATH}/share/titan-gc-cli-${VERSION}
+TITAN_SHARE=${TITAN_PATH}/share/titan-gc-${TITAN_GC_VERSION}
 CROMWELL_JAR=$(which cromwell | sed 's=bin/cromwell=share/cromwell/cromwell.jar=')
 LOG_LEVEL=ERROR
 SINGULARITY_CACHE="${HOME}/.singularity/titan-cache"
 CROMWELL_OPTS="${TITAN_SHARE}/conf/options.json"
 
 version() {
-    echo "titan-gc-cli ${VERSION}"
+    echo "titan-gc-cli ${TITAN_GC_VERSION}"
     exit 0
 }
 
