diff --git a/setup.py b/setup.py
index 9ab10d66..543c2502 100644
--- a/setup.py
+++ b/setup.py
@@ -2,7 +2,7 @@
 
 """Setup file for PyPI"""
 
-from setuptools import setup, find_packages
+from setuptools import setup, find_namespace_packages
 from setuptools.extension import Extension
 from distutils.extension import Extension
 from Cython.Build import cythonize
@@ -12,8 +12,10 @@ import glob
 import re
 import sys
 import numpy as np
-from scipy._build_utils import numpy_nodepr_api
 
+### See comment here https://github.com/cython/cython/issues/2498
+numpy_nodepr_api = dict(define_macros=[("NPY_NO_DEPRECATED_API",
+                                        "NPY_1_9_API_VERSION")])
 
 here = path.abspath(path.dirname("__file__"))
 
@@ -79,7 +81,7 @@ setup(
     ext_modules=cythonize(extensions, compiler_directives = {"language_level": "3"}),
     zip_safe=False,
     keywords="rna-seq annotation genomics transcriptomics",
-    packages=find_packages(),
+    packages=find_namespace_packages(),
     scripts=glob.glob("util/*.py"),
     entry_points={"console_scripts": ["mikado = Mikado.__main__:main",
                                       "daijin = Mikado.daijin.__main__:main",
