From c451622094488c65c4b1e58689ce588bdf2fafc2 Mon Sep 17 00:00:00 2001
From: Barry digby <b.digby237@gmail.com>
Date: Thu, 1 Dec 2022 12:07:35 +0000
Subject: [PATCH] try except for container error

---
 CIRIquant/main.py | 5 ++++-
 setup.py          | 4 ++--
 2 files changed, 6 insertions(+), 3 deletions(-)

diff --git a/CIRIquant/main.py b/CIRIquant/main.py
index 9763f7b..795299e 100644
--- a/CIRIquant/main.py
+++ b/CIRIquant/main.py
@@ -123,7 +123,10 @@ def main():
     # Add lib to PATH
     lib_path = os.path.dirname(os.path.split(os.path.realpath(__file__))[0]) + '/libs'
     os.environ['PATH'] = lib_path + ':' + os.environ['PATH']
-    os.chmod(lib_path + '/CIRI2.pl', 0o755)
+    try:
+        os.chmod(lib_path + '/CIRI2.pl', 0o755)
+    except OSError:
+        pass
 
     """Start Running"""
     os.chdir(outdir)
diff --git a/setup.py b/setup.py
index 1e334e9..7c4ee7b 100644
--- a/setup.py
+++ b/setup.py
@@ -42,7 +42,7 @@ setup(
     include_package_data=True,
     zip_safe=False,
     install_requires=[
-        'argparse==1.2.1', 'PyYAML==5.1.1', 'pysam==0.15.2', 'numpy==1.16.4',
-        'scipy==1.2.2', 'scikit-learn==0.20.3', 'numexpr==2.6.9',
+        'argparse', 'PyYAML', 'pysam', 'numpy',
+        'scipy', 'scikit-learn', 'numexpr',
     ],
 )
-- 
2.38.1

