diff --git a/setup.py b/setup.py
index da4c6c4..7c4ee7b 100644
--- a/setup.py
+++ b/setup.py
@@ -42,7 +42,7 @@ setup(
     include_package_data=True,
     zip_safe=False,
     install_requires=[
-        'argparse>=1.2.1', 'PyYAML==5.4', 'pysam==0.15.2', 'numpy==1.16.4',
-        'scipy==1.2.2', 'scikit-learn==0.20.3', 'numexpr==2.6.9',
+        'PyYAML', 'pysam', 'numpy',
+        'scipy', 'scikit-learn'
     ],
 )
