From ac607987362530e32486a05616ed71c41aa2c051 Mon Sep 17 00:00:00 2001
From: BarryDigby <b.digby237@gmail.com>
Date: Wed, 16 Nov 2022 19:02:40 -0800
Subject: [PATCH] >= for argparse and scipy in setup.py

---
 setup.py | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/setup.py b/setup.py
index 1e334e9..71bcad6 100644
--- a/setup.py
+++ b/setup.py
@@ -42,7 +42,7 @@ setup(
     include_package_data=True,
     zip_safe=False,
     install_requires=[
-        'argparse==1.2.1', 'PyYAML==5.1.1', 'pysam==0.15.2', 'numpy==1.16.4',
-        'scipy==1.2.2', 'scikit-learn==0.20.3', 'numexpr==2.6.9',
+        'argparse>=1.2.1', 'PyYAML==5.1.1', 'pysam==0.15.2', 'numpy==1.16.4',
+        'scipy>=1.2.2', 'scikit-learn==0.20.3', 'numexpr==2.6.9',
     ],
 )
-- 
2.38.1

