--- setup.py.bak        2016-09-29 11:03:42.528559057 +0200
+++ setup.py    2016-09-29 11:58:43.226201408 +0200
@@ -15,6 +15,9 @@
 *********************************************************************************
 """

+
+condaprefix= os.getenv("PREFIX")
+
 # Get the version number from __init__.py
 pkg_name = 'nanonet'
 pkg_path = os.path.join(os.path.dirname(__file__), pkg_name)
@@ -58,7 +61,7 @@
 main_include = os.path.join(os.path.dirname(__file__), 'nanonet', 'include')
 include_dirs = [main_include]
 event_detect_include = []
-boost_inc = []
+boost_inc = [(os.path.join(condaprefix, 'include'))]
 boost_lib_path = []
 boost_libs = []
 opencl_include = []
@@ -74,8 +77,8 @@
     optimisation.extend(['-O3', '-fstrict-aliasing'])
     cpp_compile_args.extend(['-std=c++0x','-Wno-unused-local-typedefs'])
     # may wish to edit - required for 2D
-    boost_inc = ['/opt/local/include/']
-    boost_libs.append('boost_python-mt')
+    #boost_inc = ['/opt/local/include/']
+    boost_libs.append('boost_python')
 elif system == 'Windows':
     event_detect_include.append(os.path.join(pkg_path, 'eventdetection'))
     if not mingw:

