--- src/Makefile
+++ src/Makefile
@@ -6,18 +6,18 @@
 #LINK_OPTS = -L$(L_PATH) -lgsl -lgslcblas
 
 #for osx systems
-CC = g++
-G++FLAG = -fast -m64 -mmmx -msse -msse2
-LINK_OPTS2 = -pthread -lz
-I_PATH = -I../include
-L_PATH = ../lib/osx
-
-#for linux systems
 #CC = g++
-#G++FLAG = -O3 -m64 -mmmx -msse -msse2
+#G++FLAG = -fast -m64 -mmmx -msse -msse2
 #LINK_OPTS2 = -pthread -lz
 #I_PATH = -I../include
-#L_PATH = ../lib/linux
+#L_PATH = ../lib/osx
+
+#for linux systems
+CC = $(CXX) $(CXXFLAGS) $(CPPFLAGS) $(LDFLAGS)
+G++FLAG = -O3 -m64 -mmmx -msse -msse2
+LINK_OPTS2 = -pthread -lz
+I_PATH = -I$(PREFIX)/include
+L_PATH = $(PREFIX)/lib
 
 #for windows, using MinGW build environment
 #CC = g++.exe
