--- src/Makefile
+++ src/Makefile
@@ -6,11 +6,11 @@
 #LINK_OPTS = -L$(L_PATH) -lgsl -lgslcblas
 
 #for osx systems
-CC = g++
-G++FLAG = -fast -m64 -mmmx -msse -msse2
+CC = $(CXX) $(CXXFLAGS) $(CPPFLAGS) $(LDFLAGS)
+G++FLAG = -m64 -mmmx -msse -msse2
 LINK_OPTS2 = -pthread -lz
-I_PATH = -I../include
-L_PATH = ../lib/osx
+I_PATH = -I$(PREFIX)/include
+L_PATH = $(PREFIX)/lib
 
 #for linux systems
 #CC = g++
