diff -u a/install.sh b/install.sh
--- a/install.sh	2013-12-02 09:31:16.000000000 +0200
+++ b/install.sh	2017-03-03 04:17:40.000000000 +0200
@@ -4,11 +4,6 @@
 
 #command line option --prefix parsen;
 # read in where to install TSSAR
-echo -e "\nPlease specify where you would like to install TSSAR:"
-read PREFIX
-
-eval PREFIX=$PREFIX
-PREFIX_A=$(readlink -f $PREFIX)
 
 echo ""
 echo "Installing R package 'skellam', 'VGAM' if needed"
@@ -23,14 +18,3 @@
 R CMD BATCH --no-save --no-restore R_lib_install
 rm R_lib_install*
 echo " ... Done"
-
-
-#subdirectory TSSAR
-PREFIX=$PREFIX_A'/TSSAR'
-
-# copies TSSAR to the chosen directory
-echo ""
-echo "cp ./TSSAR to $PREFIX"
-mkdir -p $PREFIX
-cp -f TSSAR $PREFIX
-echo " ... Done"
