--- a/Makefile
+++ b/Makefile
@@ -32,16 +32,18 @@
 D_COMPILER=ldc2
 CC=gcc

+BIOD_PATH=./BioD:./BioD/contrib/msgpack-d/src
+DFLAGS      = -wi -I. -I$(BIOD_PATH) -g -J.
+LDFLAGS     =
+
 UNAME_S := $(shell uname -s)
 ifeq ($(UNAME_S),Darwin)
   SYS = OSX
 else
   SYS = LINUX
+  LDFLAGS += -L=-flto=full
 endif

-BIOD_PATH=./BioD:./BioD/contrib/msgpack-d/src
-DFLAGS      = -wi -I. -I$(BIOD_PATH) -g -J.
-LDFLAGS     = -L=-flto=full

 # DLIBS       = $(LIBRARY_PATH)/libphobos2-ldc.a $(LIBRARY_PATH)/libdruntime-ldc.a
 # DLIBS_DEBUG = $(LIBRARY_PATH)/libphobos2-ldc-debug.a $(LIBRARY_PATH)/libdruntime-ldc-debug.a
