diff --git a/Makefile b/Makefile
index 91f30c3..443e1eb 100644
--- a/Makefile
+++ b/Makefile
@@ -33,7 +33,7 @@ D_COMPILER=ldc2
 CC=$(CC)
 
 BIOD_PATH=./BioD:./BioD/contrib/msgpack-d/src
-DFLAGS      = -wi -I. -I$(BIOD_PATH) -I$(PREFIX)/include -g -O3 -J.
+DFLAGS      = -wi -g -O3 -J. -I. -I$(BIOD_PATH) -I$(PREFIX)/include
 LDFLAGS     =
 
 UNAME_S := $(shell uname -s)
@@ -48,7 +48,7 @@ endif
 # DLIBS       = $(LIBRARY_PATH)/libphobos2-ldc.a $(LIBRARY_PATH)/libdruntime-ldc.a
 # DLIBS_DEBUG = $(LIBRARY_PATH)/libphobos2-ldc-debug.a $(LIBRARY_PATH)/libdruntime-ldc-debug.a
 # LIBS        = -L-L$(LIBRARY_PATH) -L-lpthread -L-lm -L-lz -L-llz4
-LIBS        = -L-lz -L-llz4
+LIBS        = -L-L$(PREFIX)/lib -L-lz -L-llz4 -L-lm
 LIBS_STATIC = -L-lz -L-llz4 -L-L$(LIBRARY_PATH) -L-lphobos2-ldc -L-ldruntime-ldc
 SRC         = $(wildcard main.d utils/*.d thirdparty/*.d) $(wildcard BioD/contrib/undead/*.d BioD/contrib/undead/*/*.d) $(wildcard BioD/bio/*/*.d BioD/bio/*/*/*.d BioD/bio/*/*/*/*.d BioD/bio/*/*/*/*/*.d BioD/bio/*/*/*/*/*/*/*.d BioD/contrib/msgpack-d/src/msgpack/*.d) $(wildcard sambamba/*.d sambamba/*/*.d sambamba/*/*/*.d)
 OBJ         = $(SRC:.d=.o)
