--- sirius-gui-org/bin/sirius	2022-04-25 14:38:48.819219636 +0200
+++ sirius-gui/bin/sirius	2022-04-27 16:42:31.046947862 +0200
@@ -48,7 +48,7 @@
 DEFAULT_JVM_OPTS="-Xms1G -XX:MaxRAMPercentage=85 -XX:+UseG1GC -XX:+UseStringDeduplication -Djava.library.path=\"$GUROBI_HOME/lib\":\"$CPLEX_HOME/bin/x86-64_linux\":\"$APP_HOME/lib/native\":\"$APP_HOME/lib/runtime/lib\""
 
 # Native library path
-export LD_LIBRARY_PATH="$GUROBI_HOME/lib:$CPLEX_HOME/bin/x86-64_linux:$APP_HOME/lib/native:$APP_HOME/lib/runtime/lib:$LD_LIBRARY_PATH"
+export LD_LIBRARY_PATH="$GUROBI_HOME/lib:$CPLEX_HOME/bin/x86-64_linux:$APP_HOME/lib/native:$APP_HOME/lib/runtime/lib:$APP_HOME/../../lib:$LD_LIBRARY_PATH"
 
 # Add location of jar files for Classpath
 JAR_HOME="$APP_HOME/lib/app"
@@ -70,6 +70,6 @@
 APP_ARGS=`save "$@"`
 
 # Collect all arguments for the java command, following the shell quoting and substitution rules
-eval set -- "$DEFAULT_JVM_OPTS" "$JAVA_OPTS" "$SIRIUS_OPTS" -classpath "$CLASSPATH" "$MAIN_CLASS" "$APP_ARGS"
+eval set -- "$DEFAULT_JVM_OPTS" "$JAVA_OPTS" "$SIRIUS_OPTS" -classpath "$CLASSPATH" "$MAIN_CLASS" --loglevel=ALL "$APP_ARGS"
 
-exec "$JAVACMD" "$@"
\ No newline at end of file
+"$JAVACMD" "$@" 2> /tmp/sirius.err > /tmp/sirius.out
