--- bin/sirius-org	2022-04-09 18:10:12.539469624 +0200
+++ bin/sirius	2022-04-11 11:03:20.895870757 +0200
@@ -48,7 +48,8 @@
 DEFAULT_JVM_OPTS="-Xms1G -XX:MaxRAMPercentage=85 -XX:+UseG1GC -XX:+UseStringDeduplication -Djava.library.path=\"$GUROBI_HOME/lib\":\"$CPLEX_HOME/bin/x86-64_linux\":\"$APP_HOME/lib/native\":\"$APP_HOME/lib/runtime/lib\""
 
 # Native library path
-export LD_LIBRARY_PATH="$GUROBI_HOME/lib:$CPLEX_HOME/bin/x86-64_linux:$APP_HOME/lib/native:$APP_HOME/lib/runtime/lib:$LD_LIBRARY_PATH"
+dname=`dirname "$0"`
+export LD_LIBRARY_PATH="$GUROBI_HOME/lib:$CPLEX_HOME/bin/x86-64_linux:$APP_HOME/lib/native:$APP_HOME/lib/runtime/lib:$dname/../lib:$LD_LIBRARY_PATH"
 
 # Add location of jar files for Classpath
 JAR_HOME="$APP_HOME/lib/app"
@@ -72,4 +73,4 @@
 # Collect all arguments for the java command, following the shell quoting and substitution rules
 eval set -- "$DEFAULT_JVM_OPTS" "$JAVA_OPTS" "$SIRIUS_OPTS" -classpath "$CLASSPATH" "$MAIN_CLASS" "$APP_ARGS"
 
-exec "$JAVACMD" "$@"
\ No newline at end of file
+exec "$JAVACMD" "$@"
