--- Contents/MacOS/sirius-org	2022-04-09 18:49:07.031382696 +0200
+++ Contents/MacOS/sirius	2022-04-09 18:50:45.123379043 +0200
@@ -48,7 +48,8 @@
 DEFAULT_JVM_OPTS="-Xms1G -XX:MaxRAMPercentage=85 -XX:+UseG1GC -XX:+UseStringDeduplication -Djava.library.path=\"$GUROBI_HOME/lib\":\"$CPLEX_HOME/bin/x86-64_linux\":\"$APP_HOME/native\":\"$APP_HOME/runtime/Contents/Home/lib\""
 
 # Native library path
-export DYLD_LIBRARY_PATH="$GUROBI_HOME/lib:$CPLEX_HOME/bin/x86-64_osx:$APP_HOME/native:$APP_HOME/runtime/Contents/Home/lib:$DYLD_LIBRARY_PATH"
+dname=`dirname "$0"`
+export DYLD_LIBRARY_PATH="$GUROBI_HOME/lib:$CPLEX_HOME/bin/x86-64_osx:$APP_HOME/native:$APP_HOME/runtime/Contents/Home/lib:$dname/../lib:$DYLD_LIBRARY_PATH"
 
 # Add location of jar files for Classpath
 JAR_HOME="$APP_HOME/app"
@@ -72,4 +73,4 @@
 # Collect all arguments for the java command, following the shell quoting and substitution rules
 eval set -- "$DEFAULT_JVM_OPTS" "$JAVA_OPTS" "$SIRIUS_OPTS" -classpath "$CLASSPATH" "$MAIN_CLASS" "$APP_ARGS"
 
-exec "$JAVACMD" "$@"
\ No newline at end of file
+exec "$JAVACMD" "$@"
