--- gdc_client/download/client.py	2017-08-22 09:31:27.000000000 -0500
+++ gdc_client/download/client.py.new	2018-07-19 21:06:33.763086968 -0500
@@ -259,7 +259,7 @@
 
         # gdc-client calls parcel's parallel_download,
         # which is where most of the downloading takes place
-        file_id = stream.url.split('/')[-1]
+        file_id = stream.uri.split('/')[-1]
         super(GDCDownloadMixin, self).parallel_download(stream)
 
         if download_related_files or \
--- setup.py	2017-08-22 09:31:27.000000000 -0500
+++ setup.py.new	2018-07-19 21:09:01.675877046 -0500
@@ -8,12 +8,12 @@
     package_data={},
     install_requires=[
         'parcel',
-        'lxml==3.5.0b1',
-        'PyYAML==3.11',
-        'jsonschema==2.5.1',
-        'pyOpenSSL==17.1.0',
-        'ndg-httpsclient==0.4.2',
-        'pyasn1==0.2.3',
+        'lxml',
+        'PyYAML',
+        'jsonschema',
+        'pyOpenSSL',
+        'ndg-httpsclient',
+        'pyasn1',
     ],
     dependency_links=[
         'git+https://github.com/LabAdvComp/parcel.git@50d6124a3e3fcd2a234b3373831075390b886a15#egg=parcel',
--- gdc_client/interactive/repl.py	2017-08-22 09:31:27.000000000 -0500
+++ gdc_client/interactive/repl.py.new	2018-07-19 21:33:05.635597734 -0500
@@ -2,7 +2,8 @@
 import os
 import shlex
 
-from cmd2 import Cmd, options, make_option
+from cmd2 import Cmd, options
+from optparse import make_option
 from parcel import const, manifest
 
 from ..download.client import GDCHTTPDownloadClient
