--- setup.py	2017-10-12 14:47:48.687006943 +0200
+++ setup.py-fixed	2017-10-12 14:48:25.240007617 +0200
@@ -8,12 +8,12 @@
     package_data={},
     install_requires=[
         'parcel',
-        'lxml==3.5.0b1',
-        'PyYAML==3.11',
-        'jsonschema==2.5.1',
-        'pyOpenSSL==17.1.0',
-        'ndg-httpsclient==0.4.2',
-        'pyasn1==0.2.3',
+        'lxml',
+        'PyYAML',
+        'jsonschema',
+        'pyOpenSSL',
+        'ndg-httpsclient',
+        'pyasn1',
     ],
     dependency_links=[
         'git+https://github.com/LabAdvComp/parcel.git@50d6124a3e3fcd2a234b3373831075390b886a15#egg=parcel',
