--- a/configure.ac	2025-04-18 23:12:28
+++ b/configure.ac	2025-09-15 22:20:59
@@ -311,7 +311,6 @@
 
 
 ACX_BUILTIN_POPCOUNT
-ACX_ASM_BSR
 
 
 ACX_SSE2_SLLI_CONST_IMM8
@@ -347,6 +346,7 @@
   ;;
 
   i[[3456]]86*|x86_64*|amd64*)
+    ACX_ASM_BSR
     AC_DEFINE(USE_INTEL_INTRINSICS,1,[Define to 1 to use Intel SIMD intrinsics rather than SIMDe])
     if test x"$ax_cv_c_compiler_vendor" = xintel; then
       AX_CPUID_INTEL      # Determine CPU characteristics for an Intel compiler
