diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7543e0e..1745f85 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,9 +1,7 @@
 cmake_minimum_required(VERSION 3.10)
-project(stark VERSION 0.1 LANGUAGES CXX)
+project(stark VERSION 0.1.1 LANGUAGES CXX)
 
-set(CMAKE_CXX_STANDARD 11)
-set(CMAKE_CXX_FLAGS "${CMAKE_C_FLAGS}  -O3")
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}  -O3")
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall")
+set(CMAKE_CXX_STANDARD 14)
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O3 -march=x86-64-v3")
 
-add_executable(stark src/main.cpp src/node.cpp src/node.h src/utils/logger.h src/utils/logger.cpp)
\ No newline at end of file
+add_executable(stark src/main.cpp src/node.cpp src/node.h src/utils/logger.h src/utils/logger.cpp)
