diff --git a/src/main.cpp b/src/main.cpp
index d7746ea..69d4087 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -23,7 +23,7 @@ int max_node_ids = -1; // For debugging purposes
 bool unify_before_run = false;
 unsigned long sequences_last_unfill;
 char *input_file_name, *output_file_name, *sequences[MAX_NODES],
-        *help_str = const_cast<char *>("stark v1.0\nUsage: stark -i input_file_name [-o output_file_name] "
+        *help_str = const_cast<char *>("stark v0.1.1\nUsage: stark -i input_file_name [-o output_file_name] "
                                        "[-m merge_type] [-l log_level] [-u] [-s statistics-level]\n\n"
                                        "    -i,      --input=FILE           use FILE for input\n"
                                        "    -o,      --output=FILE          use FILE for output\n"
@@ -457,4 +457,4 @@ int main(int argc, char *argv[]) {
     }
     if (output_file_name)
         write_to_file(output_file_name);
-}
\ No newline at end of file
+}
