diff --git a/src/formats/libinchi/CMakeLists.txt b/src/formats/libinchi/CMakeLists.txt
index 2a0be6b5..e72083c9 100644
--- a/src/formats/libinchi/CMakeLists.txt
+++ b/src/formats/libinchi/CMakeLists.txt
@@ -6,7 +6,7 @@ set(INCHI_PATCH_VER 1)
 file(GLOB inchi_srcs "*.c")
 
 if(NOT WIN32)
-  set(libs ${libs} c)
+  set(libs ${libs} c m)
 endif(NOT WIN32)
 
 include_directories(${openbabel_SOURCE_DIR}/include/inchi)
