diff --git CMakeLists.txt CMakeLists.txt
index 3d1e223..0f38684 100644
--- CMakeLists.txt
+++ CMakeLists.txt
@@ -118,9 +118,9 @@ SET (CPACK_SOURCE_IGNORE_FILES
 
 # We copy some files to the binary archive
 INSTALL (TARGETS ${PROJECT_NAME} DESTINATION bin)
-INSTALL (DIRECTORY "${PROJECT_SOURCE_DIR}/test/" DESTINATION test)
-INSTALL (FILES scripts/simple_test.sh DESTINATION test/)
-INSTALL (FILES LICENSE README.md DESTINATION bin/..)
+INSTALL (DIRECTORY ${PROJECT_SOURCE_DIR}/test/ DESTINATION $ENV{SHARE_DIR}/test)
+INSTALL (FILES scripts/simple_test.sh DESTINATION $ENV{SHARE_DIR}/test/)
+INSTALL (FILES LICENSE README.md DESTINATION $ENV{SHARE_DIR}/)
 
 # We include the "bin" tag into binary archive file name
 set (CPACK_PACKAGE_FILE_NAME  ${PROJECT_NAME}-${CPACK_PACKAGE_VERSION}-bin-${CMAKE_SYSTEM_NAME})
