diff --git a/Cargo.toml b/Cargo.toml
index ccae9ae..7ad430e 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,3 +1,4 @@
+
 [package]
 name = "alevin-fry"
 version = "0.5.0"
@@ -51,7 +52,7 @@ snap = "1"
 rand = "0.8.5"
 chrono = "0.4.19"
 csv = "1.1.6"
-mimalloc = { version = "0.1.28", default-features = false }
+mimalloc = { version = "0.1.26", default-features = false }
 num-format = "0.4.0"
 num_cpus = "1.13.0"
 bio-types = "0.12.1"
