diff --git a/Cargo.toml b/Cargo.toml
index 55b87ce..b462721 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -75,7 +75,7 @@ rust-htslib = { version = "0.39.5", default-features = false, features = [
 ] }
 sce = { git = "https://github.com/parazodiac/SingleCellExperiment", version = "0.1.2" }
 
-clap = { version = ">=3.1.18", features = ["derive", "wrap_help", "cargo"] }
+clap = { version = "=3.1.18", features = ["derive", "wrap_help", "cargo"] }
 
 [profile.release]
 #debug = true
