--- Cargo.toml.orig	2024-10-14 10:59:48.140000000 +0000
+++ Cargo.toml	2024-10-14 11:03:33.120000000 +0000
@@ -13,9 +13,9 @@
 rayon="1.7"
 rand="0.8"
 rand_core="0.6"
-rust-htslib= { version = "0.44", default-features = false }
+rust-htslib= { version = "0.47", default-features = false }
 clap = { version = "=4.2.0", features = ["derive"] }
-block-aligner = { version = "0.4", default-features = false }
+block-aligner = { version = "0.5", default-features = false }
 debruijn = "0.3"
 simple_logger="4"
 log="0.4"
