diff --git a/Cargo.toml b/Cargo.toml
index 2221e65..a57aefb 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -13,9 +13,9 @@ statrs="0.17"
 rayon="1.7"
 rand="0.8"
 rand_core="0.6"
-rust-htslib= { version = "1.0.1", default-features = false }
+rust-htslib= { version = "0.47", default-features = false }
 clap = { version = "=4.2.0", features = ["derive"] }
-block-aligner = { version = "0.4", default-features = false }
+block-aligner = { version = "0.5", default-features = false }
 debruijn = "0.3"
 simple_logger="4"
 log="0.4"
