#!/usr/bin/perl
use strict;
use warnings;
use Getopt::Long;
use File::Spec;
use File::Basename;

# é»˜è®¤å˜é‡
my $gff_file;
my $fasta_file;
my $output_path = './';
my $help;

# å‘½ä»¤è¡Œå‚æ•°è§£æ
GetOptions(
    "gff=s" => \$gff_file,
	"genome=s" => \$fasta_file,
	"outdir=s" => \$output_path,
    "help|h" => \$help,
) or die "Error in command line arguments.\n";

# å¸®åŠ©ä¿¡æ¯
if ($help or not $gff_file or not $fasta_file) {
    print <<"USAGE";
Usage:
  perl extract_mRNA_protein.info.pl --genome <genome.fa> --gff <gene.gff3>

Description:
  This script extracts the first mRNA ID (or Name), start, end, and strand (+/-)
  from a GFF3 annotation file. It also generates protein sequences for the selected
  mRNAs from the provided genome FASTA file.

Options:
  --gff <file>      Input GFF3 annotation file (required)
  --genome <file>   Input genome FASTA file (required)
  -h, --help        Show this help message

Output:
  <prefix>.bed       mRNA coordinates (ID, start, end, strand) for mRNAs ending with .1 or -1
  <prefix>.list      List of selected mRNA IDs
  <genome_prefix>.protein.fa  Protein sequences of the selected mRNAs
  <genome_prefix>.protein.all.fa  All protein sequences extracted from genome (temporary, deleted after filtering)

Example:
  perl extract_mRNA_protein.info.pl --genome genome.fa --gff gene.gff3
USAGE
    exit;
}
	# å¦‚æœç›®å½•ä¸å­˜åœ¨,åˆ™åˆ›å»ºå®ƒ,å¹¶æ£€éªŒæ˜¯å¦åˆ›å»ºæˆåŠŸ
	unless (-d $output_path) {
		print "ğŸ“ Directory '$output_path' does not exist. Creating it now...\n";
		system("mkdir -p \"$output_path\"");
		die "Directory '$output_path' does not exist and could not be created!\n" unless -d $output_path;
	}
my $prefix1 = basename($gff_file);
my $prefix_g = basename($fasta_file);
$prefix_g =~ s/\.[^.]+$//;
my $genome_id = $prefix_g . '.genome_id.txt';
my $gff_file_gai = 'filter.' . $prefix1;
$prefix1 =~ s/\.[^.]+$//;
my $output1 = $prefix1 . '.bed';
my $output2 = $prefix1 . '.list';
if($output_path ne './'){
	$genome_id = File::Spec->catfile($output_path, $genome_id);
	$gff_file_gai = File::Spec->catfile($output_path, $gff_file_gai);
	$output1 = File::Spec->catfile($output_path, $output1);
	$output2 = File::Spec->catfile($output_path, $output2);
}
# ç”ŸæˆæŸ“è‰²ä½“çš„IDåˆ—è¡¨æ–‡ä»¶
if (!-e $genome_id) {
	system("seqkit seq -n $fasta_file > $genome_id") == 0
		or die "Failed to run seqkit: $!";
} else {
	print "âš  $genome_id already exists, skipping seqkit(seqkit seq -n $fasta_file > $genome_id).\n";
}
# æå–åŸºå› ç»„æœ‰çš„æŸ“è‰²ä½“IDçš„åŸºå› æ³¨é‡Šä¿¡æ¯ï¼Œé¿å…gffreadæŠ¥é”™[gff3æ–‡ä»¶æœ‰scaffoldä½†åŸºå› ç»„fatsaæ²¡æœ‰æ—¶ï¼Œgffreadä¼šæŠ¥é”™]
if (!-e $gff_file_gai) {
	system("grep -Ff $genome_id $gff_file > $gff_file_gai") == 0
		or die "Failed to run grep: $!";
} else {
	print "âš  $gff_file_gai already exists, skipping grep(grep -Ff $genome_id $gff_file > $gff_file_gai).\n";
}
# æ‰“å¼€æ–‡ä»¶
open(my $IN, "<", $gff_file_gai) or die "Cannot open $gff_file_gai: $!\n";
open(my $OUT1, ">", $output1) or die "Cannot open $output1: $!";
open(my $OUT2, ">", $output2) or die "Cannot open $output2: $!";
# è§£æGFFæ–‡ä»¶ï¼Œæå–mRNAä¿¡æ¯
while (<$IN>) {
	chomp;
	next if /^#/;  # è·³è¿‡æ³¨é‡Šè¡Œ
	my @cols = split(/\t/, $_);
	next unless $cols[2] eq "mRNA";  # åªå¤„ç† mRNA è¡Œ
	my ($chr, $start, $end, $strand) = @cols[0,3,4,6];
	my ($id) = $cols[8] =~ /Name=([^;]+)/;
	$id ||= ($cols[8] =~ /ID=([^;]+)/)[0];  # å¦‚æœæ²¡æœ‰Nameåˆ™ç”¨ID
	# åªè¾“å‡ºä»¥ .1 æˆ– -1 ç»“å°¾çš„ mRNA
	if ($id and $id =~ /(?:\.1|-1)$/) {
		print $OUT1 "$id\t$chr\t$start\t$end\t$strand\n";
		print $OUT2 "$id\n";
	}
}
close $IN;
close $OUT1;
close $OUT2;

my $prefix2 = basename($fasta_file);
$prefix2 =~ s/\.[^.]+$//;
my $output3 = $prefix2 . '.protein.all.fa';
my $output4 = $prefix2 . '.protein.first.fa';
my $output5 = $prefix2 . '.protein.fa';
if($output_path ne './'){
	$output3 = File::Spec->catfile($output_path, $output3);
	$output4 = File::Spec->catfile($output_path, $output4);
	$output5 = File::Spec->catfile($output_path, $output5);
}
# ç”Ÿæˆæ‰€æœ‰è›‹ç™½åºåˆ—æ–‡ä»¶
if (!-e $output3) {
	system("gffread $gff_file_gai -g $fasta_file -y $output3") == 0
		or die "Failed to run gffread: $!";
} else {
	print "âš  $output3 already exists, skipping gffread.\n";
}
# æå–ç¬¬ä¸€æ¡è½¬å½•æœ¬çš„è›‹ç™½åºåˆ—
if (!-e $output4) {
	system("seqkit grep -i -f $output2 $output3 > $output4") == 0
		or die "Failed to run seqkit grep: $!";
} else {
	print "âš  $output4 already exists, skipping seqkit grep.\n";
}

# æ£€æŸ¥è¾“å‡ºæ–‡ä»¶æ˜¯å¦å­˜åœ¨
if (-e $output5) {
    print "âš ï¸  Warning: File '$output5' already exists. Deleting old file...\n";
    unlink $output5 or warn "Cannot remove $output5: $!";
}
# # ç¡®è®¤æ–‡ä»¶å·²åˆ é™¤å,åˆ é™¤åºåˆ—ä¸­é™¤æœ€åä¸€ä¸ªå¤–çš„æ‰€æœ‰å…¶ä»–ä½ç½®å«'.'çš„åºåˆ—,å¹¶å°†æœ€åä¸€ä¸ªä½ç½®çš„'.'æ”¹æˆ'*'
if (!-e $output5) {
	open(my $OUT4, "<", $output4) or die "Cannot open $output4: $!\n";
	open(my $OUT5, ">", $output5) or die "Cannot open $output5: $!";
	my $header = '';
	my $seq = '';

	sub flush_seq {
		my ($h, $s) = @_;
		return unless $h;    # nothing to print on first call
		# remove whitespace/newlines in sequence (should be continuous)
		$s =~ s/\s+//g;
		# body = seq except last char (if seq non-empty)
		my $body = length($s) ? substr($s, 0, length($s)-1) : '';
		# å¦‚æœ body ä¸­åŒ…å« '.' åˆ™è·³è¿‡è¯¥åºåˆ—
		unless ($body =~ /\./) {
			print $OUT5 $h, "\n";
			# æŒ‰ 60 é•¿åº¦æŠ˜è¡Œ
			$s =~ s/(.{1,60})/$1\n/g;
			$s =~ s/\./\*/g;
			print $OUT5 $s;
		}
	}

	while (<$OUT4>) {
		chomp;
		if (/^>/) {
			flush_seq($header, $seq) if $header;
			$header = $_;
			$seq = '';
		} else {
			$seq .= $_;
		}
	}
	# flush last
	flush_seq($header, $seq);
	close $OUT4;
	close $OUT5;
}

# åˆ é™¤ä¸´æ—¶æ–‡ä»¶
if (-e $output2) {
	unlink $output2 or warn "Cannot remove $output2: $!";
}
if (-e $output3) {
	unlink $output3 or warn "Cannot remove $output3: $!";
}
if (-e $output4) {
	unlink $output4 or warn "Cannot remove $output4: $!";
}
if (-e $genome_id) {
	unlink $genome_id or warn "Cannot remove $genome_id: $!";
}
if (-e $gff_file_gai) {
	unlink $gff_file_gai or warn "Cannot remove $gff_file_gai: $!";
}

print "âœ… Extraction completed successfully.\n";
print "Output files:\n";
print "  Coordinates: $output1\n";
print "  Protein sequences: $output5\n";
#./gff2bed.pl 1.MH63.gene.gff3 > 1.MH63.mRNA.bed
#awk '{print$1}' 1.MH63.mRNA.bed > 1.MH63.mRNA.txt
#gffread 1.MH63.gene.gff3 -g 1.MH63RS3.fasta -y 1.MH63RS3.proteins.fa
#seqkit grep -i -f 1.MH63.mRNA.txt 1.MH63RS3.proteins.fa > 1.MH63RS3.proteins2.fa
