#!/bin/bash

# å‚æ•°ä¸ªæ•°ä¸è¶³æˆ–è¿‡å¤šæ—¶æç¤ºç”¨æ³•
if [ "$#" -lt 4 ] || [ "$#" -gt 5 ]; then
    echo "Usage: $0 <query_fasta> <ref_fasta> <ref_db> <output_file> [threads_number]"
    echo "Example (default threads=1):"
    echo "    $0 query.pep.fa ref.pep.fa ref_db ref_vs_query.blast"
    echo
    echo "Example (custom):"
    echo "    $0 query.pep.fa ref.pep.fa ref_db ref_vs_query.blast 1"
    echo
    echo "Parameters:"
    echo "  query_fasta    Query protein FASTA file              (required)"
    echo "  ref_fasta      Target protein FASTA file             (required)"
    echo "  ref_db         Prefix for BLAST database & index     (required)"
    echo "  output_file    Output file name                      (required)"
    echo "  threads_number Number of threads to use [default: 1] (optional)"
    # echo "  run_blast     Whether to run BLASTP [default: 1] (0 = skip)"
    exit 1
fi

# å‚æ•°èµ‹å€¼
QUERY=$1
DB_FASTA=$2
DB_PREFIX=$3
OUTPUT=$4
# è‡ªåŠ¨ç”Ÿæˆé»˜è®¤è¾“å‡ºæ–‡ä»¶åï¼ˆå¦‚æœæœªæä¾›ï¼‰
# QUERY_NAME=$(basename "$QUERY" | cut -d. -f1)
# DB_NAME=$(basename "$DB_FASTA" | cut -d. -f1)
# OUTPUT=${4:-${DB_NAME}vs${QUERY_NAME}.blast}
threads_number=${5:-1}
OUTPUT=$(realpath "$OUTPUT")
workdir=$(dirname "$OUTPUT")
db_dir="${DB_PREFIX}_DB"
cd "$workdir"

# æ˜¯å¦è¿è¡Œ BLASTPï¼ˆé»˜è®¤ 1ï¼‰
align_commands=1

# æ£€æŸ¥ BLAST+ å·¥å…·æ˜¯å¦é½å…¨
for tool in blastp makeblastdb; do
    if ! command -v "$tool" &> /dev/null; then
        echo "âš ï¸ Warning: $tool is not installed or not in PATH. Please install BLAST+."
        exit 0
    fi
done

# åˆ¤æ–­æ˜¯å¦æ‰§è¡Œblastpå¯¹é½
for f in "$OUTPUT"; do
    if [ -e "$f" ]; then
        echo "âš ï¸ Warning: File $f exists."
        align_commands=0
    fi
done

echo "ref:${QUERY}"
echo "query:${DB_FASTA}"
for f in "${QUERY}" "${DB_FASTA}"; do
    if [ ! -e "$f" ]; then
        echo "âš ï¸ Warning: Reference genome file $f does not exist."
        align_commands=0
    fi
done

# è¿è¡Œ BLASTPï¼ˆå¯é€‰ï¼‰
if [ "$align_commands" -eq 1 ]; then
    # åˆ›å»º BLAST æ•°æ®åº“
    echo "ğŸ§¬ Building BLAST database: $DB_PREFIX"
    mkdir "$db_dir"
    cd "$workdir/$db_dir"
    makeblastdb -in "$DB_FASTA" -dbtype prot -parse_seqids -out "$DB_PREFIX" -logfile "${DB_PREFIX}.log" -title "$DB_PREFIX"
    echo "ğŸš€ Running BLASTP: $QUERY vs $DB_PREFIX"
    blastp -query "$QUERY" -db "$DB_PREFIX" -out "$OUTPUT" -evalue 1e-10 -num_threads "$threads_number" -outfmt 6 -num_alignments 1 2> >(grep -v "Warning" >&2)
    if [ ! -e "$OUTPUT" ]; then
    echo "âœ… BLASTP completed. Output saved to $OUTPUT"
    fi
else
    echo "â­ï¸ Skipping BLAST alignment."
fi

