#!/bin/bash

# å‚æ•°ä¸ªæ•°ä¸è¶³æ—¶æç¤ºåŸºæœ¬ç”¨æ³•
if [ "$#" -lt 3 ] || [ "$#" -gt 4 ] || [ "$#" -gt 5 ]; then
    echo "Usage: $0 <ref_file> <query_file> <index_name> [threads_number]"
    echo
    echo "Example (default threads=1):"
    echo "    $0 ref.fasta query.fasta ref_vs_query"
    echo
    echo "Example (custom):"
    echo "    $0 ref.fasta query.fasta ref_vs_query 1"
    echo
    echo "Parameters:"
    echo "  ref_file         Reference genome FASTA file            (required)"
    echo "  query_file       Query genome FASTA file                (required)"
    echo "  index_name       Prefix name for output files           (required)"
    echo "  threads_number   Number of threads to use [default: 1]  (optional)"
    # echo "  sv_commands      1 to run SV detection [default: 1]     (optional)"
    exit 1
fi

# å‚æ•°èµ‹å€¼ï¼ˆå¸¦é»˜è®¤å€¼ï¼‰
ref_file=$1
query_file=$2
index_name=$3
threads_number=${4:-1}
sv_commands=${5:-1}
align_commands=1
index_name=$(realpath "$index_name")
workdir=$(dirname "$index_name")
SCRIPT_PATH="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
cd "$workdir"

# æ£€æŸ¥ MUMmer æ˜¯å¦å·²å®‰è£…ï¼ˆé€šè¿‡æ£€æµ‹ nucmer æ˜¯å¦å¯ç”¨ï¼‰
if ! command -v nucmer &> /dev/null; then
	echo "âš ï¸ Warning: MUMmer (nucmer) is not installed or not in PATH. Skipping MUMmer search."
    exit 0
fi

# åˆ¤æ–­æ˜¯å¦æ‰§è¡Œmummerå¯¹é½
for f in "${index_name}.delta" "${index_name}.delta.filter" "${index_name}.coords"; do
    if [ -e "$f" ]; then
	echo "âš ï¸ Warning: File $f exists."
        align_commands=0
    fi
done

echo "ref:${ref_file}"
echo "query:${query_file}"
for f in "${ref_file}" "${query_file}"; do
    if [ ! -e "$f" ]; then
	echo "âš ï¸ Warning: Genome file $f does not exist."
        align_commands=0
    fi
done

if [ "$align_commands" -eq 1 ]; then
    echo "ğŸš€ Running MUMmer alignment..."
    nucmer -g 1000 -c 90 -l 40 -t "$threads_number" -p "$index_name" "$ref_file" "$query_file"
    delta-filter -r -q -l 1000 "$index_name.delta" > "$index_name.delta.filter"
    show-coords -TrHcl "$index_name.delta.filter" > "$index_name.coords"
    $SCRIPT_PATH/mummer_coords2tsv.pl --in "$index_name.coords" --out "$index_name.mummer.tsv"
    if [ -e "$index_name.coords" ]; then
        echo "âœ… Alignment finished. Output: $index_name.coords"
    fi
else
     echo "â­ï¸ Skipping MUMmer alignment."
fi

