diff --git a/Makefile b/Makefile
index 2a5bfd8..6054c6f 100644
--- a/Makefile
+++ b/Makefile
@@ -35,7 +35,7 @@ libhts.a: samtools-$(SAMVER)/Makefile
 	@echo "\x1b[1;33mMaking $(@F)\x1b[0m"
 	cd samtools-${SAMVER}/htslib-${SAMVER}; \
 	if [ "$(OS)" = "Darwin" ]; then \
-		autoheader && autoconf -Wno-syntax && ./configure; \
+		cp -f ${BUILD_PREFIX}/share/gnuconfig/config.* . && autoheader && autoconf -Wno-syntax && ./configure; \
 	else \
 		CFLAGS="${CFLAGS}" LDFLAGS="${LDFLAGS}" ./configure; \
 	fi; \
