diff --git a/Makefile b/Makefile
index 75e68b1..a64a3ee 100644
--- a/Makefile
+++ b/Makefile
@@ -8,12 +8,12 @@ clean : clean_htslib clean_longphase clean_libclair3
 
 SAMVER	=	1.15.1
 LPVER	=	1.7.3
-GCC	?=	gcc
-GXX	?=	g++
-PREFIX	?=	${CONDA_PREFIX}
-LDFLAGS	=	-L ${PREFIX}/lib
-CFLAGS	= -fpic -std=c99 -O3 -I ${PREFIX}/include -L ${PREFIX}/lib
-CPPFLAGS	=	-std=c++11 -Wall -O3 -I ${PREFIX}/include -L ${PREFIX}/lib -Wl,-rpath=${PREFIX}/lib
+GCC	?=	$(CC)
+GXX	?=	$(CXX)
+PREFIX	?=	${PREFIX}
+LDFLAGS	=	-L${PREFIX}/lib
+CFLAGS	= -fpic -std=c99 -O3 -I${PREFIX}/include -L${PREFIX}/lib
+CPPFLAGS	=	-std=c++14 -Wall -O3 -I${PREFIX}/include -L${PREFIX}/lib -Wl,-rpath=${PREFIX}/lib
 
 
 samtools-$(SAMVER)/Makefile:
