diff --git a/Makefile b/Makefile
index 2a5bfd8..b7b7586 100644
--- a/Makefile
+++ b/Makefile
@@ -35,7 +35,7 @@ libhts.a: samtools-$(SAMVER)/Makefile
 	@echo "\x1b[1;33mMaking $(@F)\x1b[0m"
 	cd samtools-${SAMVER}/htslib-${SAMVER}; \
 	if [ "$(OS)" = "Darwin" ]; then \
-		autoheader && autoconf -Wno-syntax && ./configure; \
+		cp -f ${BUILD_PREFIX}/share/gnuconfig/config.* . && autoheader && autoconf -Wno-syntax && ./configure; \
 	else \
 		CFLAGS="${CFLAGS}" LDFLAGS="${LDFLAGS}" ./configure; \
 	fi; \
@@ -47,7 +47,7 @@ longphase:
 	if [ "$(OS)" = "Darwin" ]; then \
 		curl -L -o v${LPVER}.tar.gz https://github.com/twolinin/longphase/archive/refs/tags/v${LPVER}.tar.gz; \
 		tar -zxf v${LPVER}.tar.gz; \
-		cd longphase-${LPVER} && export CC=${CC_PATH} && autoreconf -i && ./configure && make -j; \
+		cd longphase-${LPVER} && autoreconf -i && CC=/usr/bin/clang CXX=/usr/bin/clang++ CPPFLAGS="" CFLAGS="" LDFLAGS="" ./configure && make -j; \
 		cd .. && rm v${LPVER}.tar.gz; \
 		cp longphase-${LPVER}/longphase $@; \
 	else \
