diff --git a/build.py b/build.py
index b8efc1b..02b3ac4 100644
--- a/build.py
+++ b/build.py
@@ -18,9 +18,10 @@ src_dir = os.path.join(file_directory, 'src')
 extra_compile_args = ['-std=c99', '-O3']
 if platform.machine() in {"aarch64", "arm64"}:
     if platform.system() == "Darwin":
-        pass
+        libraries.append('deflate')
     else:
         extra_compile_args.append("-march=armv8-a+simd")
+        libraries.append('deflate')
 else:
     extra_compile_args.append("-mtune=haswell")
     libraries.append('deflate')
