diff --git a/Makefile b/Makefile
index be87069..061f7d7 100644
--- a/Makefile
+++ b/Makefile
@@ -12,7 +12,7 @@ GCC	?=	$(CC)
 GXX	?=	$(CXX)
 PREFIX	?=	$(PREFIX)
 LDFLAGS	=	-L${PREFIX}/lib
-CFLAGS	= -fpic -std=c11 -O3 -I${PREFIX}/include -L${PREFIX}/lib
+CFLAGS	= -std=c11 -O3 -I${PREFIX}/include -L${PREFIX}/lib
 CPPFLAGS	=	-std=c++14 -Wall -O3 -I${PREFIX}/include -L${PREFIX}/lib -Wl,-rpath=${PREFIX}/lib
 
 
