diff --git a/build.py b/build.py
index b8efc1b..beac66b 100644
--- a/build.py
+++ b/build.py
@@ -21,6 +21,13 @@ if platform.machine() in {"aarch64", "arm64"}:
         pass
     else:
         extra_compile_args.append("-march=armv8-a+simd")
+        libraries.append('deflate')
+        try:
+            conda_path = os.environ['CONDA_PREFIX']
+            extra_link_args = ['-Wl,-rpath={}/lib'.format(conda_path)]
+            library_dirs.append(os.path.join(conda_path, 'lib'))
+        except KeyError:
+            print("[WARNING] Conda prefix not found, please activate clair3 conda environment first!\n")
 else:
     extra_compile_args.append("-mtune=haswell")
     libraries.append('deflate')
