--- pyscenic-0.12.1/src/pyscenic/prune.py	2022-11-21 10:38:05.000000000 +0000
+++ pyscenic-0.12.1/src/pyscenic/prune.py	2024-10-02 10:20:58.423708208 +0000
@@ -20,8 +20,8 @@
 from dask import delayed
 from dask.dataframe import from_delayed
 from dask.distributed import Client, LocalCluster
-from multiprocessing_on_dill.connection import Pipe
-from multiprocessing_on_dill.context import Process
+from multiprocessing import Pipe
+from multiprocessing import Process
 
