--- spapros/__main__.py	2024-06-19 18:26:41.062595569 +0200
+++ spapros/__main__.py	2024-06-19 18:31:52.844067995 +0200
@@ -5,7 +5,6 @@
 
 import click
 import rich.logging
-from pypi_latest import PypiLatest
 from rich import print, traceback
 
 import spapros
@@ -13,8 +12,6 @@
 from spapros.selection.selection import run_selection
 
 log = logging.getLogger()
-spapros_pypi_latest = PypiLatest("spapros", spapros.__version__)
-
 
 def main() -> None:
     traceback.install()
@@ -32,9 +29,6 @@
     )
     print("[bold blue]Run [green]spapros --help [blue]for an overview of all commands\n")
 
-    # Is the latest spapros version installed? Upgrade if not!
-    if not spapros_pypi_latest.check_latest():
-        print("[bold blue]Run [green]spapros upgrade [blue]to get the latest version.")
     spapros_cli(prog_name="spapros")
 
 
@@ -122,12 +116,5 @@
         parameters_file=parameters,
     )
 
-
-@spapros_cli.command(short_help="Check for a newer version of ehrapy and upgrade if required.")
-def upgrade() -> None:
-    """Checks whether the locally installed version of spapros is the latest & upgrades if not."""
-    spapros_pypi_latest.check_upgrade()
-
-
 if __name__ == "__main__":
     main()  # pragma: no cover

