--- pyproject.toml	2024-06-19 17:45:38.517620166 +0200
+++ pyproject.toml	2024-06-19 18:13:47.801886617 +0200
@@ -28,12 +28,10 @@
 leidenalg = ">=0.8.7"
 questionary = ">=1.10.0"
 xgboost = ">=1.6.1"
-pypi-latest = ">=0.1.0"
 "ruamel.yaml" = ">=0.17.10"
 jupyter-sphinx = ">=0.3.2"
 nox = ">=2023.04.22"
 nox-poetry = ">=1.0.3"
-pandoc = ">=2.1"
 bandit = ">=1.7.6"
 venndata = ">=0.1.0"
 Pillow = ">=10.0.2"
@@ -51,12 +49,10 @@
 #leidenalg = "^0.8.7"
 #questionary = "^1.10.0"
 #xgboost = "^1.6.1"
-#pypi-latest = "^0.1.0"
 #"ruamel.yaml" = "^0.17.10"
 #jupyter-sphinx = "^0.3.2"
 #nox = "^2023.04.22"
 #nox-poetry = "^1.0.3"
-#pandoc = "^2.1"
 #bandit = "^1.7.6"
 #venndata = "^0.1.0"
 #Pillow = "^10.0.2"

