diff -Naur genedom-0.2.1.orig/MANIFEST.in genedom-0.2.1/MANIFEST.in
--- genedom-0.2.1.orig/MANIFEST.in	2025-04-04 11:46:33.000000000 +0200
+++ genedom-0.2.1/MANIFEST.in	2025-05-06 12:12:51.929993346 +0200
@@ -1,6 +1,5 @@
 include *.txt
 include *.rst
-recursive-include examples *.txt *.py
 recursive-include genedom/reports_assets *
 recursive-include genedom/assembly_standards *
 include ez_setup.py
diff -Naur genedom-0.2.1.orig/pyproject.toml genedom-0.2.1/pyproject.toml
--- genedom-0.2.1.orig/pyproject.toml	2025-04-04 11:46:33.000000000 +0200
+++ genedom-0.2.1/pyproject.toml	2025-05-06 13:46:01.466225799 +0200
@@ -29,4 +29,4 @@
 include-package-data = true
 
 [tool.setuptools.packages.find]
-exclude = ["docs"]
+exclude = ["docs","examples","tests"]
