--- a/eKLIPse_circos.py
+++ b/eKLIPse_circos.py
@@ -1,7 +1,6 @@
 import sys
 import os
 import shutil
-from string import split
 from eKLIPse_fct import *
 
 
@@ -124,7 +123,7 @@ def circosConf(titleBam,dicoInit,lstError):
 
         #***** CREATE Main Circos Configuration File *****#
         IN = open(pathConfTemplate,'r')
-        lst_lines = split(IN.read(),"\n")
+        lst_lines = IN.read().split("\n")
         IN.close()
         OUT = open(pathConf,'w')
         for line in lst_lines:
--- a/eKLIPse_fct.py
+++ b/eKLIPse_fct.py
@@ -2,7 +2,6 @@ import sys
 import os
 import glob
 import json
-from string import split
 
 
 #=================================================================================================================================================#
--- a/eKLIPse_init.py
+++ b/eKLIPse_init.py
@@ -68,12 +68,12 @@ def arg_manager(argv,pathRootDir,boolColor,boolQT,spinner):
             if not os.path.isfile(argv[i+1]): lstErrorDisplay.append("Input file \""+argv[i+1]+"\" not found")
             else:
                 dicoInit['pathInput'] = argv[i+1]
-                IN = open(argv[i+1],'r') ; lstLine = split(IN.read(),"\n") ; IN.close()
+                IN = open(argv[i+1],'r') ; lstLine = IN.read().split("\n") ; IN.close()
                 for j in range(len(lstLine)):
                     if lstLine[j]!="" and not lstLine[j][0]=="#":
                         try:
-                            pathBam = split(lstLine[j],"\t")[0]
-                            titleBam = split(lstLine[j],"\t")[1]
+                            pathBam = lstLine[j].split("\t")[0]
+                            titleBam = lstLine[j].split("\t")[1]
                             if not os.path.isfile(pathBam): lstErrorDisplay.append("Input file at line"+str(j+1)+": \""+pathBam+"\" not found")
                             if titleBam in dicoInit['lstTitleBam']: lstErrorDisplay.append("Input file at line "+str(j+1)+": \""+titleBam+"\" name already used")
                             dicoInit['lstTitleBam'].append(titleBam)
@@ -95,7 +95,7 @@ def arg_manager(argv,pathRootDir,boolColor,boolQT,spinner):
                             if feature.qualifiers['nomenclature'][0].lower().__contains__("trna"): geneType = "trna"
                             elif feature.qualifiers['nomenclature'][0].lower().__contains__("rna"): geneType = "rrna"
                             else: geneType = "protein"
-                            try: name = split(split(feature.qualifiers['nomenclature'][0],"|")[0],":")[1].replace(" ","")
+                            try: name = feature.qualifiers['nomenclature'][0].split("|")[0].split(":")[1].replace(" ","")
                             except: name = feature.qualifiers['gene'][0]
                             dicoInit["dicoGbk"]['lstGene'].append([name,int(feature.location.start)+1,int(feature.location.end),geneType])
             if len(dicoInit["dicoGbk"])==0: lstErrorDisplay.append("Cannot Read Reference GBK file \""+argv[i+1]+"\"")
@@ -115,7 +115,7 @@ def arg_manager(argv,pathRootDir,boolColor,boolQT,spinner):
                         if feature.qualifiers['nomenclature'][0].lower().__contains__("trna"): geneType = "trna"
                         elif feature.qualifiers['nomenclature'][0].lower().__contains__("rna"): geneType = "rrna"
                         else: geneType = "protein"
-                        try: name = split(split(feature.qualifiers['nomenclature'][0],"|")[0],":")[1].replace(" ","")
+                        try: name = feature.qualifiers['nomenclature'][0].split("|")[0].split(":")[1].replace(" ","")
                         except: name = feature.qualifiers['gene'][0]
                         dicoInit["dicoGbk"]['lstGene'].append([name,int(feature.location.start)+1,int(feature.location.end),geneType])
         # Results Directory
@@ -231,9 +231,9 @@ def arg_manager(argv,pathRootDir,boolColor,boolQT,spinner):
             if dicoInit['windows']==False: cmd_idxstats = dicoInit['pathSamtools']+" idxstats "+dicoInit['dicoBam'][titleBam]['path']+" 1>"+path_idxstats+" 2>/dev/null"
             else: cmd_idxstats = dicoInit['pathSamtools']+" idxstats "+dicoInit['dicoBam'][titleBam]['path']+" 1>"+path_idxstats+" 2> nul"
             os.system(cmd_idxstats)
-            IN = open(path_idxstats,'r') ; lst_lines = split(IN.read(),"\n") ; IN.close()
+            IN = open(path_idxstats,'r') ; lst_lines = IN.read().split("\n") ; IN.close()
             for line in lst_lines:
-                split_line = split(line,"\t")
+                split_line = line.split("\t")
                 if len(split_line)>1:
                     refName = split_line[0]
                     refSize = int(split_line[1])
--- a/eKLIPse_sc.py
+++ b/eKLIPse_sc.py
@@ -2,7 +2,6 @@ import json
 import os
 import sys
 import pybam
-from string import split
 from eKLIPse_fct import *
 
 
@@ -91,13 +90,13 @@ def SC_blast(titleBam,dicoInit,lstError):
             os.system(cmd)
             #***** Read Blast output file *****#
             IN = open(pathSCDELBLAST,'r')
-            lst_lines = split(IN.read(),"\n")
+            lst_lines = IN.read().split("\n")
             IN.close()
             for line in lst_lines:
                 if line!="":
                     # Blast line parser
-                    splitLine = split(line,",")
-                    splitQread = split(splitLine[0],"_")
+                    splitLine = line.split(",")
+                    splitQread = splitLine[0].split("_")
                     SCposRead = int(splitQread[0])
                     SCmappedpart = int(splitQread[1])
                     readOrient = splitQread[2]
@@ -188,8 +187,8 @@ def deletionPrediction(titleBam,dicoInit,lstError):
                 initial_SCposRead_F = [] ; initial_SCposRead_R = []
                 # Browse deletion
                 for delName in dico_shift_del[key]:
-                    start = split(delName,"-")[0]
-                    end = split(delName,"-")[1]
+                    start = delName.split("-")[0]
+                    end = delName.split("-")[1]
                     try: delnbblastF = dicoDel[delName]['scrF']['nbBlast']
                     except: delnbblastF = 0
                     try: delnbblastR = dicoDel[delName]['scrR']['nbBlast']
@@ -214,8 +213,8 @@ def deletionPrediction(titleBam,dicoInit,lstError):
                     # Del entry
                     try : del(dicoDel[delName])
                     except: pass
-                if maxBlast_start=="": maxBlast_start = split(delName,"-")[0]
-                if maxBlast_end=="": maxBlast_end = split(delName,"-")[1]
+                if maxBlast_start=="": maxBlast_start = delName.split("-")[0]
+                if maxBlast_end=="": maxBlast_end = delName.split("-")[1]
                 dicoDel[maxBlast_start+"-"+maxBlast_end] = { 'scrF':{'nbBlast':nbBlast_F,'limit':limit_max_start,'initial_SCposRead':initial_SCposRead_F},\
                                                              'scrR':{'nbBlast':nbBlast_R,'limit':limit_max_end,'initial_SCposRead':initial_SCposRead_R},\
                                                              'freqF':0.0, 'freqR':0.0 }
--- a/pybam.py
+++ b/pybam.py
@@ -169,7 +169,7 @@ class read:
             self.file_name = os.path.basename(os.path.realpath(self._file.name))
             self.file_directory = os.path.dirname(os.path.realpath(self._file.name))
 
-            if magic == 'BAM\1':
+            if magic == b'BAM\1':
                 # The user has passed us already unzipped BAM data! Job done :)
                 data = 'BAM\1' + self._file.read(35536)
                 self.file_bytes_read += len(data)
@@ -182,7 +182,7 @@ class read:
                 DEVNULL.close()
                 raise StopIteration
 
-            elif magic == "\x1f\x8b\x08\x04":  # The user has passed us compressed gzip/bgzip data, which is typical for a BAM file
+            elif magic == b"\x1f\x8b\x08\x04":  # The user has passed us compressed gzip/bgzip data, which is typical for a BAM file
                 # use custom decompressor if provided:
                 if decompressor is not False and decompressor is not 'internal':
                     if type(f) is str: self._subprocess = subprocess.Popen(                                    decompressor.replace('{}',f),    shell=True, stdout=subprocess.PIPE, stderr=DEVNULL)
@@ -243,7 +243,7 @@ class read:
                             bs = 0
                         magic = raw_data[bs:bs+4]
                         if not magic: break # a child's heart
-                        if magic != "\x1f\x8b\x08\x04": raise PybamError('\n\nThe input file is not in a format I understand. First four bytes: ' + repr(magic) + '\n')
+                        if magic != b"\x1f\x8b\x08\x04": raise PybamError('\n\nThe input file is not in a format I understand. First four bytes: ' + repr(magic) + '\n')
                         try:
                             more_bs = bs + unpack("<H", raw_data[bs+16:bs+18])[0] +1
                             internal_cache.append(decompress(raw_data[bs+18:more_bs-8],-15))
@@ -278,12 +278,12 @@ class read:
                             '''
                         blocks_left_to_grab -= 1
                         if blocks_left_to_grab == 0:
-                            yield ''.join(internal_cache)
+                            yield b''.join(internal_cache)
                             internal_cache = []
                             blocks_left_to_grab = 50
                     self._file.close()
                     DEVNULL.close()
-                    if internal_cache != '': yield ''.join(internal_cache)
+                    if internal_cache != '': yield b''.join(internal_cache)
                     raise StopIteration
 
             elif decompressor is not False and decompressor is not 'internal':
@@ -311,11 +311,11 @@ class read:
         self._generator = generator()
 
         ## So lets parse the BAM header:
-        header_cache = ''
+        header_cache = b''
         while len(header_cache) < 8: header_cache += next(self._generator)
 
         p_from = 0; p_to = 4
-        if header_cache[p_from:p_to] != 'BAM\1':
+        if header_cache[p_from:p_to] != b'BAM\1':
             raise PybamError('\n\nInput file ' + self.file_name + ' does not appear to be a BAM file.\n')
 
         ## Parse the BAM header:
@@ -334,18 +334,18 @@ class read:
             l_name = unpack('<l',header_cache[p_from:p_to])[0]
             p_from = p_to; p_to += l_name
             while len(header_cache) < p_to: header_cache += next(self._generator)
-            self.file_chromosomes.append(header_cache[p_from:p_to -1])
+            self.file_chromosomes.append(header_cache[p_from:p_to -1].decode('ascii', errors='replace'))
             p_from = p_to; p_to += 4
             while len(header_cache) < p_to: header_cache += next(self._generator)
             self.file_chromosome_lengths[self.file_chromosomes[-1]] = unpack('<l',header_cache[p_from:p_to])[0]
 
         self.file_bytes_read = p_to
-        self.file_binary_header = buffer(header_cache[:p_to])
+        self.file_binary_header = memoryview(bytes(header_cache[:p_to]))
         header_cache = header_cache[p_to:]
 
         # A quick check to make sure the header of this BAM file makes sense:
         chromosomes_from_header = []
-        for line in self.file_header.split('\n'):
+        for line in self.file_header.decode('ascii', errors='replace').split('\n'):
             if line.startswith('@SQ\tSN:'):
                 chromosomes_from_header.append(line.split('\t')[1][3:])
         if chromosomes_from_header != self.file_chromosomes:
