--- a/eKLIPse_circos.py
+++ b/eKLIPse_circos.py
@@ -1,7 +1,6 @@
 import sys
 import os
 import shutil
-from string import split
 from eKLIPse_fct import *
 
 
@@ -123,7 +122,7 @@ def circosConf(titleBam,dicoInit,lstError):
 
         #***** CREATE Main Circos Configuration File *****#
         IN = open(pathConfTemplate,'r')
-        lst_lines = split(IN.read(),"\n")
+        lst_lines = IN.read().split("\n")
         IN.close()
         OUT = open(pathConf,'w')
         for line in lst_lines:
--- a/eKLIPse_fct.py
+++ b/eKLIPse_fct.py
@@ -2,7 +2,6 @@ import sys
 import os
 import glob
 import json
-from string import split
 
 
 #=================================================================================================================================================#
--- a/eKLIPse_init.py
+++ b/eKLIPse_init.py
@@ -68,12 +68,12 @@ def arg_manager(argv,pathRootDir,boolColor,boolQT,spinner):
             if not os.path.isfile(argv[i+1]): lstErrorDisplay.append("Input file \""+argv[i+1]+"\" not found")
             else:
                 dicoInit['pathInput'] = argv[i+1]
-                IN = open(argv[i+1],'r') ; lstLine = split(IN.read(),"\n") ; IN.close()
+                IN = open(argv[i+1],'r') ; lstLine = IN.read().split("\n") ; IN.close()
                 for j in range(len(lstLine)):
                     if lstLine[j]!="" and not lstLine[j][0]=="#":
                         try:
-                            pathBam = split(lstLine[j],"\t")[0]
-                            titleBam = split(lstLine[j],"\t")[1]
+                            pathBam = lstLine[j].split("\t")[0]
+                            titleBam = lstLine[j].split("\t")[1]
                             if not os.path.isfile(pathBam): lstErrorDisplay.append("Input file at line"+str(j+1)+": \""+pathBam+"\" not found")
                             if titleBam in dicoInit['lstTitleBam']: lstErrorDisplay.append("Input file at line "+str(j+1)+": \""+titleBam+"\" name already used")
                             dicoInit['lstTitleBam'].append(titleBam)
@@ -95,7 +95,7 @@ def arg_manager(argv,pathRootDir,boolColor,boolQT,spinner):
                             if feature.qualifiers['nomenclature'][0].lower().__contains__("trna"): geneType = "trna"
                             elif feature.qualifiers['nomenclature'][0].lower().__contains__("rna"): geneType = "rrna"
                             else: geneType = "protein"
-                            try: name = split(split(feature.qualifiers['nomenclature'][0],"|")[0],":")[1].replace(" ","")
+                            try: name = feature.qualifiers['nomenclature'][0].split("|")[0].split(":")[1].replace(" ","")
                             except: name = feature.qualifiers['gene'][0]
                             dicoInit["dicoGbk"]['lstGene'].append([name,int(feature.location.start)+1,int(feature.location.end),geneType])
             if len(dicoInit["dicoGbk"])==0: lstErrorDisplay.append("Cannot Read Reference GBK file \""+argv[i+1]+"\"")
@@ -115,7 +115,7 @@ def arg_manager(argv,pathRootDir,boolColor,boolQT,spinner):
                         if feature.qualifiers['nomenclature'][0].lower().__contains__("trna"): geneType = "trna"
                         elif feature.qualifiers['nomenclature'][0].lower().__contains__("rna"): geneType = "rrna"
                         else: geneType = "protein"
-                        try: name = split(split(feature.qualifiers['nomenclature'][0],"|")[0],":")[1].replace(" ","")
+                        try: name = feature.qualifiers['nomenclature'][0].split("|")[0].split(":")[1].replace(" ","")
                         except: name = feature.qualifiers['gene'][0]
                         dicoInit["dicoGbk"]['lstGene'].append([name,int(feature.location.start)+1,int(feature.location.end),geneType])
         # Results Directory
@@ -231,9 +231,9 @@ def arg_manager(argv,pathRootDir,boolColor,boolQT,spinner):
             if dicoInit['windows']==False: cmd_idxstats = dicoInit['pathSamtools']+" idxstats "+dicoInit['dicoBam'][titleBam]['path']+" 1>"+path_idxstats+" 2>/dev/null"
             else: cmd_idxstats = dicoInit['pathSamtools']+" idxstats "+dicoInit['dicoBam'][titleBam]['path']+" 1>"+path_idxstats+" 2> nul"
             os.system(cmd_idxstats)
-            IN = open(path_idxstats,'r') ; lst_lines = split(IN.read(),"\n") ; IN.close()
+            IN = open(path_idxstats,'r') ; lst_lines = IN.read().split("\n") ; IN.close()
             for line in lst_lines:
-                split_line = split(line,"\t")
+                split_line = line.split("\t")
                 if len(split_line)>1:
                     refName = split_line[0]
                     refSize = int(split_line[1])
--- a/eKLIPse_sc.py
+++ b/eKLIPse_sc.py
@@ -2,7 +2,6 @@ import json
 import os
 import sys
 import pybam
-from string import split
 from eKLIPse_fct import *
 
 
@@ -90,13 +89,13 @@ def SC_blast(titleBam,dicoInit,lstError):
             os.system(cmd)
             #***** Read Blast output file *****#
             IN = open(pathSCDELBLAST,'r')
-            lst_lines = split(IN.read(),"\n")
+            lst_lines = IN.read().split("\n")
             IN.close()
             for line in lst_lines:
                 if line!="":
                     # Blast line parser
-                    splitLine = split(line,",")
-                    splitQread = split(splitLine[0],"_")
+                    splitLine = line.split(",")
+                    splitQread = splitLine[0].split("_")
                     SCposRead = int(splitQread[0])
                     SCmappedpart = int(splitQread[1])
                     readOrient = splitQread[2]
@@ -187,8 +186,8 @@ def deletionPrediction(titleBam,dicoInit,lstError):
                 initial_SCposRead_F = [] ; initial_SCposRead_R = []
                 # Browse deletion
                 for delName in dico_shift_del[key]:
-                    start = split(delName,"-")[0]
-                    end = split(delName,"-")[1]
+                    start = delName.split("-")[0]
+                    end = delName.split("-")[1]
                     try: delnbblastF = dicoDel[delName]['scrF']['nbBlast']
                     except: delnbblastF = 0
                     try: delnbblastR = dicoDel[delName]['scrR']['nbBlast']
@@ -213,8 +212,8 @@ def deletionPrediction(titleBam,dicoInit,lstError):
                     # Del entry
                     try : del(dicoDel[delName])
                     except: pass
-                if maxBlast_start=="": maxBlast_start = split(delName,"-")[0]
-                if maxBlast_end=="": maxBlast_end = split(delName,"-")[1]
+                if maxBlast_start=="": maxBlast_start = delName.split("-")[0]
+                if maxBlast_end=="": maxBlast_end = delName.split("-")[1]
                 dicoDel[maxBlast_start+"-"+maxBlast_end] = { 'scrF':{'nbBlast':nbBlast_F,'limit':limit_max_start,'initial_SCposRead':initial_SCposRead_F},\
                                                              'scrR':{'nbBlast':nbBlast_R,'limit':limit_max_end,'initial_SCposRead':initial_SCposRead_R},\
                                                              'freqF':0.0, 'freqR':0.0 }
--- a/pybam.py
+++ b/pybam.py
@@ -100,7 +100,7 @@ Github:     http://github.com/JohnLonginotto/pybam
 [ Force Internal bgzip Decompressor ]
   my_bam = pybam.read('/my/data.bam',decompressor='internal')
 [ Parse Words (hah) ]'''
-wat += '\n'+''.join([('\n===============================================================================================\n\n  ' if code is 'file_alignments_read' or code is 'sam' else '  ')+(code+' ').ljust(25,'-')+description+'\n' for code,description in sorted(parse_codes.items())]) + '\n'
+wat += '\n'+''.join([('\n===============================================================================================\n\n  ' if code == 'file_alignments_read' or code == 'sam' else '  ')+(code+' ').ljust(25,'-')+description+'\n' for code,description in sorted(parse_codes.items())]) + '\n'
 
 class read:
     '''
@@ -135,7 +135,7 @@ class read:
         self.file_chromosome_lengths = {}
 
         if fields is not False:
-            if type(fields) is not list or len(fields) is 0:
+            if type(fields) is not list or len(fields) == 0:
                 raise PybamError('\n\nFields for the static parser must be provided as a non-empty list. You gave a ' + str(type(fields)) + '\n')
             else:
                 for field in fields:
@@ -147,7 +147,7 @@ class read:
 
         if decompressor:
             if type(decompressor) is str:
-                 if decompressor is not 'internal' and '{}' not in decompressor: raise PybamError('\n\nWhen a custom decompressor is used and the input file is a string, the decompressor string must contain at least one occurence of "{}" to be substituted with a filepath by pybam.\n')
+                 if decompressor != 'internal' and '{}' not in decompressor: raise PybamError('\n\nWhen a custom decompressor is used and the input file is a string, the decompressor string must contain at least one occurence of "{}" to be substituted with a filepath by pybam.\n')
             else: raise PybamError('\n\nUser-supplied decompressor must be a string that when run on the command line decompresses a named file (or stdin), to stdout:\ne.g. "lzma --decompress --stdout {}" if pybam is provided a path as input file, where {} is substituted for that path.\nor just "lzma --decompress --stdout" if pybam is provided a file object instead of a file path, as data from that file object will be piped via stdin to the decompression program.\n')
 
         ## First we make a generator that will return chunks of uncompressed data, regardless of how we choose to decompress:
@@ -184,7 +184,7 @@ class read:
 
             elif magic == "\x1f\x8b\x08\x04":  # The user has passed us compressed gzip/bgzip data, which is typical for a BAM file
                 # use custom decompressor if provided:
-                if decompressor is not False and decompressor is not 'internal':
+                if decompressor is not False and decompressor != 'internal':
                     if type(f) is str: self._subprocess = subprocess.Popen(                                    decompressor.replace('{}',f),    shell=True, stdout=subprocess.PIPE, stderr=DEVNULL)
                     else:              self._subprocess = subprocess.Popen('{ printf "'+magic+'"; cat; } | ' + decompressor, stdin=self._file, shell=True, stdout=subprocess.PIPE, stderr=DEVNULL)
                     self.file_decompressor = decompressor
@@ -211,7 +211,7 @@ class read:
                             use = 'gzip'
                         except OSError: use = 'internal'
 
-                    if use is not 'internal' and decompressor is not 'internal':
+                    if use != 'internal' and decompressor != 'internal':
                         if type(f) is str: self._subprocess = subprocess.Popen([                                   use , '--decompress','--stdout',       f           ], stdout=subprocess.PIPE, stderr=DEVNULL)
                         else:              self._subprocess = subprocess.Popen('{ printf "'+magic+'"; cat; } | ' + use + ' --decompress  --stdout', stdin=f, shell=True, stdout=subprocess.PIPE, stderr=DEVNULL)
                         time.sleep(1)
@@ -286,7 +286,7 @@ class read:
                     if internal_cache != '': yield ''.join(internal_cache)
                     raise StopIteration
 
-            elif decompressor is not False and decompressor is not 'internal':
+            elif decompressor is not False and decompressor != 'internal':
                 # It wouldn't be safe to just print to the shell four random bytes from the beginning of a file, so instead it's
                 # written to a temp file and cat'd. The idea here being that we trust the decompressor string as it was written by 
                 # someone with access to python, so it has system access anyway. The file/data, however, should not be trusted.
