diff --git a/INSTALL.py b/INSTALL.py
index 1441559..34a3943 100644
--- a/INSTALL.py
+++ b/INSTALL.py
@@ -1,12 +1,12 @@
 import os,sys,string
 #harvest utilities INSTALL script
 user_home = os.environ["HOME"]
-print "<<Welcome to harVest utility script install>>"
+print("<<Welcome to harVest utility script install>>")
 
 #check for python version                                                                                                                                                                         
 if (sys.version_info[0] < 2) or (sys.version_info[0] == 2 and sys.version_info[1] < 6):
 
-  print "Python version is %s. harvest-tools requires at least 2.6"%(sys.version)
+  print(("Python version is %s. harvest-tools requires at least 2.6"%(sys.version)))
   sys.exit(1)
 
 #complete shebang
diff --git a/Makefile.in b/Makefile.in
index fdb2ce0..6d75982 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -1,4 +1,4 @@
-CXXFLAGS += -std=c++11 -Isrc -I@protobuf@/include -I@capnp@/include
+CXXFLAGS += -std=c++14 -Isrc -I@protobuf@/include -I@capnp@/include
 
 UNAME_S=$(shell uname -s)
 
diff --git a/configure.ac b/configure.ac
index fc8e61d..c57aa94 100644
--- a/configure.ac
+++ b/configure.ac
@@ -29,7 +29,7 @@ then
 	AC_MSG_ERROR([Cap'n Proto compiler (capnp) not found.])
 fi
 
-CPPFLAGS="-I$with_protobuf/include -I$with_capnp/include -std=c++11"
+CPPFLAGS="-I$with_protobuf/include -I$with_capnp/include -std=c++14"
 
 AC_CHECK_HEADER(google/protobuf/stubs/common.h, [result=1], [result=0])
 
diff --git a/harvest-snps.py b/harvest-snps.py
index db590d5..04bc566 100644
--- a/harvest-snps.py
+++ b/harvest-snps.py
@@ -11,7 +11,7 @@ harvest_found = True
 
 #print help/usage
 if len(sys.argv) == 1 or sys.argv[1] in helpme:
-    print "usage: harvest-snps.py <parsnp.hvt> [outfile]"
+    print("usage: harvest-snps.py <parsnp.hvt> [outfile]")
     sys.exit(1)
 
 #check in path for harvest
diff --git a/harvest-variants.py b/harvest-variants.py
index 6f01960..3c43ecf 100644
--- a/harvest-variants.py
+++ b/harvest-variants.py
@@ -9,7 +9,7 @@ harvest_found = True
 
 #usage/help
 if len(sys.argv) == 1 or sys.argv[1] in helpme:
-    print "usage: harvest-variants.py <parsnp.hvt> [outfile]"
+    print("usage: harvest-variants.py <parsnp.hvt> [outfile]")
     sys.exit(1)
 
 #check in path for harvest
diff --git a/src/harvest-alignments.py b/src/harvest-alignments.py
index 8e6f8a2..54caf3c 100644
--- a/src/harvest-alignments.py
+++ b/src/harvest-alignments.py
@@ -1,4 +1,4 @@
-#!python
+#!python3
 import os,sys,string,subprocess
 
 ##harvest utility file for XMFA file(s) of core genome alignments
@@ -13,7 +13,7 @@ harvest_found = True
 
 #print help/usage
 if len(sys.argv) == 1 or sys.argv[1] in helpme:
-    print "usage: harvest-alignments.py <parsnp.hvt> [outfile]"
+    print("usage: harvest-alignments.py <parsnp.hvt> [outfile]")
     sys.exit(1)
 
 #check in path for harvest
diff --git a/src/harvest-backbone.py b/src/harvest-backbone.py
index 29ac1d8..ba09c93 100644
--- a/src/harvest-backbone.py
+++ b/src/harvest-backbone.py
@@ -1,4 +1,4 @@
-#!python
+#!python3
 import os,sys,string,subprocess
 
 ##harvest utility file for tab-delimited backbone file
@@ -12,7 +12,7 @@ harvest_found = True
 
 #print help/usage
 if len(sys.argv) == 1 or sys.argv[1] in helpme:
-    print "usage: harvest-backbone.py <parsnp.hvt> [outfile]"
+    print("usage: harvest-backbone.py <parsnp.hvt> [outfile]")
     sys.exit(1)
 
 #check in path for harvest
diff --git a/src/harvest-phylogeny.py b/src/harvest-phylogeny.py
index 40fecf8..fac059a 100644
--- a/src/harvest-phylogeny.py
+++ b/src/harvest-phylogeny.py
@@ -1,4 +1,4 @@
-#!python
+#!python3
 import os,sys,string,subprocess
 
 ##harvest utility file for newick file of core genome phylogeny
@@ -12,7 +12,7 @@ harvest_found = True
 
 #print help/usage
 if len(sys.argv) == 1 or sys.argv[1] in helpme:
-    print "usage: harvest-phylogeny.py <parsnp.hvt> [outfile]"
+    print("usage: harvest-phylogeny.py <parsnp.hvt> [outfile]")
     sys.exit(1)
 
 #check in path for harvest
diff --git a/src/harvest-snps.py b/src/harvest-snps.py
index 5a26077..004c91b 100644
--- a/src/harvest-snps.py
+++ b/src/harvest-snps.py
@@ -1,4 +1,4 @@
-#!python
+#!python3
 import os,sys,string,subprocess
 
 ##harvest utility file for multi-fasta file of SNP loci
@@ -12,7 +12,7 @@ harvest_found = True
 
 #print help/usage
 if len(sys.argv) == 1 or sys.argv[1] in helpme:
-    print "usage: harvest-snps.py <parsnp.hvt> [outfile]"
+    print("usage: harvest-snps.py <parsnp.hvt> [outfile]")
     sys.exit(1)
 
 #check in path for harvest
diff --git a/src/harvest-variants.py b/src/harvest-variants.py
index f88a0eb..c08c3a2 100644
--- a/src/harvest-variants.py
+++ b/src/harvest-variants.py
@@ -1,4 +1,4 @@
-#!python
+#!python3
 import os,sys,string,subprocess
 ##harvest utility file for writing VCF file
 #ROADMAP
@@ -10,7 +10,7 @@ harvest_found = True
 
 #usage/help
 if len(sys.argv) == 1 or sys.argv[1] in helpme:
-    print "usage: harvest-variants.py <parsnp.hvt> [outfile]"
+    print("usage: harvest-variants.py <parsnp.hvt> [outfile]")
     sys.exit(1)
 
 #check in path for harvest
diff --git a/src/harvest/LcbList.cpp b/src/harvest/LcbList.cpp
index a99b037..5eb2cbf 100755
--- a/src/harvest/LcbList.cpp
+++ b/src/harvest/LcbList.cpp
@@ -1352,7 +1352,7 @@ void LcbList::writeToXmfa(ostream & out, const ReferenceList & referenceList, co
 				currvar = blockVarStart;
 			}
 			
-			out << ">" << r+1 << ":" << start << "-" << end << " ";
+			out << "> " << r+1 << ":" << start << "-" << end << " ";
 			
 			if ( ! region.reverse )
 			{
