diff --git a/main_src_ont/Makefile b/main_src_ont/Makefile
index c500e3d..be5d65a 100644
--- a/main_src_ont/Makefile
+++ b/main_src_ont/Makefile
@@ -1,16 +1,12 @@
 .PHONY: all clean bin 
 
-CC 		   ?=	gcc
-CXX 	   ?=	g++
+CC 		   ?=	$(CC)
+CXX 	   ?=	$(CXX)
 
-CXXFLAGS   +=	-std=c++11 -no-pie\
-				-I../\
+CXXFLAGS   +=	-std=c++14 -O3 -I..
 
 
-LD_FLAGS   +=    -L../minimap2\
-				-lz\
-				-lminimap2\
-				-lpthread\
+LD_FLAGS   +=    -L../minimap2 -lz -lminimap2 -lpthread
 
 
 DEUBG_CXX  =	${CXXFLAGS} -g
@@ -55,7 +51,7 @@ all :  ${jobs} ${scripts}
 tgsgapcloser_cpp = tgsgapcloser.cpp
 tgsgapcloser_o = tgsgapcloser.o
 tgsgapcloser : clean ${tgsgapcloser_o} ${source_o} ../tgsgapcloserbin
-	${CXX} ${tgsgapcloser_o} ${source_o} ${LD_FLAGS} ${DEUBG_CXX} -o tgsgapcloser
+	${CXX} ${DEUBG_CXX} ${tgsgapcloser_o} ${source_o} ${LD_FLAGS} -o tgsgapcloser ${LD_FLAGS}
 	mv $@ ../tgsgapcloserbin/
 
 
@@ -64,7 +60,7 @@ tgsgapcloser : clean ${tgsgapcloser_o} ${source_o} ../tgsgapcloserbin
 tgsgapcandidate_cpp = tgsgapcandidate.cpp
 tgsgapcandidate_o = tgsgapcandidate.o
 tgsgapcandidate : clean ${tgsgapcandidate_o} ${source_o} ../tgsgapcloserbin
-	${CXX} ${tgsgapcandidate_o} ${source_o} ${LD_FLAGS} ${DEUBG_CXX} -o tgsgapcandidate
+	${CXX} ${DEUBG_CXX} ${tgsgapcandidate_o} ${source_o} -o tgsgapcandidate ${LD_FLAGS}
 	mv $@ ../tgsgapcloserbin/
 
 
@@ -73,7 +69,7 @@ tgsgapcandidate : clean ${tgsgapcandidate_o} ${source_o} ../tgsgapcloserbin
 tgsseqsplit_cpp = tgsseqsplit.cpp
 tgsseqsplit_o = tgsseqsplit.o
 tgsseqsplit : clean ${tgsseqsplit_o} ${source_o} ../tgsgapcloserbin
-	${CXX} ${tgsseqsplit_o} ${source_o} ${LD_FLAGS} ${DEUBG_CXX} -o tgsseqsplit
+	${CXX} ${DEUBG_CXX} ${tgsseqsplit_o} ${source_o} -o tgsseqsplit ${LD_FLAGS}
 	mv $@ ../tgsgapcloserbin/
 
 
@@ -82,7 +78,7 @@ tgsseqsplit : clean ${tgsseqsplit_o} ${source_o} ../tgsgapcloserbin
 tgsseqgen_cpp = tgsseqgen.cpp
 tgsseqgen_o = tgsseqgen.o
 tgsseqgen : clean ${tgsseqgen_o} ${source_o} ../tgsgapcloserbin
-	${CXX} ${tgsseqgen_o} ${source_o} ${LD_FLAGS} ${DEUBG_CXX} -o tgsseqgen
+	${CXX} ${DEUBG_CXX} ${tgsseqgen_o} ${source_o} -o tgsseqgen ${LD_FLAGS}
 	mv $@ ../tgsgapcloserbin/
 
 
