diff --git a/main_src_ont/Makefile b/main_src_ont/Makefile
index c500e3d..4cf4f60 100644
--- a/main_src_ont/Makefile
+++ b/main_src_ont/Makefile
@@ -1,9 +1,9 @@
 .PHONY: all clean bin 
 
-CC 		   ?=	gcc
-CXX 	   ?=	g++
+CC 		   ?=	$(CC)
+CXX 	   ?=	$(CXX)
 
-CXXFLAGS   +=	-std=c++11 -no-pie\
+CXXFLAGS   +=	-std=c++14 -no-pie\
 				-I../\
 
 
@@ -13,7 +13,7 @@ LD_FLAGS   +=    -L../minimap2\
 				-lpthread\
 
 
-DEUBG_CXX  =	${CXXFLAGS} -g
+DEUBG_CXX  =	${CXXFLAGS} -g -O3
 RELEASE_CXX=	${CXXFLAGS}
 
 source_cpp =	../common/files/file_reader.cpp \
