diff --git a/main_src_ont/tgsgapcloser.cpp b/main_src_ont/tgsgapcloser.cpp
index 89094ea..c1d93d0 100644
--- a/main_src_ont/tgsgapcloser.cpp
+++ b/main_src_ont/tgsgapcloser.cpp
@@ -1,23 +1,23 @@
-#include "common/args/argsparser.h"
-#include "common/log/log.h"
-#include "common/log/logfilter.h"
-#include "common/files/file_writer.h"
-#include "common/files/file_reader.h"
-#include "common/string/stringtools.h"
-#include "common/error/Error.h"
-#include "common/stl/mapHelper.h"
-#include "common/freq/freq.h"
-
-#include "biocommon/fastq/fastq.h"
-#include "biocommon/fasta/fasta.h"
-#include "biocommon/paf/PAF.h"
-#include "biocommon/seq/tool_func.h"
-
-#include "appcommon/contigPairInfo.h"
-#include "appcommon/ScaffInfo.h"
-#include "appcommon/ONT2Gap.h"
-#include "appcommon/AlignMinimap2.h"
-#include "appcommon/fileName.h"
+#include "../common/args/argsparser.h"
+#include "../common/log/log.h"
+#include "../common/log/logfilter.h"
+#include "../common/files/file_writer.h"
+#include "../common/files/file_reader.h"
+#include "../common/string/stringtools.h"
+#include "../common/error/Error.h"
+#include "../common/stl/mapHelper.h"
+#include "../common/freq/freq.h"
+
+#include "../biocommon/fastq/fastq.h"
+#include "../biocommon/fasta/fasta.h"
+#include "../biocommon/paf/PAF.h"
+#include "../biocommon/seq/tool_func.h"
+
+#include "../appcommon/contigPairInfo.h"
+#include "../appcommon/ScaffInfo.h"
+#include "../appcommon/ONT2Gap.h"
+#include "../appcommon/AlignMinimap2.h"
+#include "../appcommon/fileName.h"
 
 #include <map>
 #include <vector>
