diff --git a/common/log/log.h b/common/log/log.h
index 37971a4..1989706 100644
--- a/common/log/log.h
+++ b/common/log/log.h
@@ -4,7 +4,7 @@
 #include <string>
 #include <ostream>
 #include <sstream>
-#include "common/time/timetools.h"
+#include "../time/timetools.h"
 
 namespace BGIQD{
 namespace LOG{
diff --git a/biocommon/fastq/fastq.h b/biocommon/fastq/fastq.h
index c2b8908..62a01ad 100644
--- a/biocommon/fastq/fastq.h
+++ b/biocommon/fastq/fastq.h
@@ -6,8 +6,8 @@
 #include <vector>
 #include <iostream>
 
-#include "biocommon/seq/seq.h"
-#include "common/flags/flags.h"
+#include "../seq/seq.h"
+#include "../../common/flags/flags.h"
 
 namespace BGIQD {
     namespace FASTQ {
diff --git a/biocommon/fasta/fasta.h b/biocommon/fasta/fasta.h
index aac3b64..563100c 100644
--- a/biocommon/fasta/fasta.h
+++ b/biocommon/fasta/fasta.h
@@ -6,8 +6,8 @@
 #include <vector>
 #include <cassert>
 
-#include "biocommon/seq/seq.h"
-#include "common/flags/flags.h"
+#include "../seq/seq.h"
+#include "../../common/flags/flags.h"
 
 namespace BGIQD {
 
diff --git a/biocommon/paf/PAF.h b/biocommon/paf/PAF.h
index 39901bd..e050c9c 100644
--- a/biocommon/paf/PAF.h
+++ b/biocommon/paf/PAF.h
@@ -2,7 +2,7 @@
 #define __BIOCOMMON_PAF_PAF_H__
 
 #include <string>
-#include "biocommon/align_common/align_result.h"
+#include "../align_common/align_result.h"
 
 namespace BGIQD {
     namespace PAF {
diff --git a/appcommon/contigPairInfo.h b/appcommon/contigPairInfo.h
index 9ec3c0c..ec18b4e 100644
--- a/appcommon/contigPairInfo.h
+++ b/appcommon/contigPairInfo.h
@@ -3,9 +3,9 @@
 
 #include <cassert>
 
-#include "appcommon/ScaffInfo.h"
+#include "ScaffInfo.h"
 
-#include "biocommon/paf/PAF.h"
+#include "../biocommon/paf/PAF.h"
 
 namespace BGIQD {
     namespace stLFR {
diff --git a/appcommon/ONT2Gap.h b/appcommon/ONT2Gap.h
index 39023d5..56117d9 100644
--- a/appcommon/ONT2Gap.h
+++ b/appcommon/ONT2Gap.h
@@ -1,8 +1,8 @@
 #ifndef __STLFR_ONT2GAP_H__
 #define __STLFR_ONT2GAP_H__
 
-#include "biocommon/paf/PAF.h"
-#include "appcommon/contigPairInfo.h"
+#include "../biocommon/paf/PAF.h"
+#include "contigPairInfo.h"
 
 #include <vector>
 
diff --git a/appcommon/AlignMinimap2.h b/appcommon/AlignMinimap2.h
index 36bb6b3..0f79528 100644
--- a/appcommon/AlignMinimap2.h
+++ b/appcommon/AlignMinimap2.h
@@ -1,7 +1,7 @@
 #ifndef __APPCOMMON_ALIGNMINIMAP2_H__
 #define __APPCOMMON_ALIGNMINIMAP2_H__
 
-#include "minimap2/minimap.h"
+#include "../minimap2/minimap.h"
 #include <string>
 #include <cstring>
 #include <cassert>
diff --git a/appcommon/fileName.h b/appcommon/fileName.h
index c60d885..2e81fd4 100644
--- a/appcommon/fileName.h
+++ b/appcommon/fileName.h
@@ -2,7 +2,7 @@
 #define __SOAP2_FILENAME_H__
 
 #include <string>
-#include "common/string/stringtools.h"
+#include "../common/string/stringtools.h"
 
 namespace BGIQD {
     namespace SOAP2 {
