diff --git a/Makefile b/Makefile
index e9f468c..fcdb8d7 100644
--- a/Makefile
+++ b/Makefile
@@ -5,17 +5,17 @@ CXX ?= $(CXX)
 PREFIX ?= $(PREFIX)
 
 all: minimap2 minimap2/libminimap2.a
-	cd main_src_ont && make CC=${CC} CXX=${CXX}
+	cd main_src_ont && make arm_neon=1 aarch64=1 CC=${CC} CXX=${CXX} 
 
 minimap2:
 	echo "error: please download/link minimap2 into this folder "
 	exit 1
 
 minimap2/minimap2:
-	cd minimap2 && make
+	cd minimap2 && make arm_neon=1 aarch64=1
 
 minimap2/libminimap2.a:
-	cd minimap2 && make
+	cd minimap2 && make arm_neon=1 aarch64=1
 
 condainstall: all
 	echo  "Installed into ${PREFIX}"
