diff --git a/HyPhy/__init__.py b/HyPhy/__init__.py
index 0b0a5c8..1fd1ea3 100644
--- a/HyPhy/__init__.py
+++ b/HyPhy/__init__.py
@@ -10,7 +10,7 @@ from sys import version_info
 if version_info >= (2,6,0):
     def swig_import_helper():
         from os.path import dirname
-        import importlib
+        import importlib.util
         module_spec = None
         try:
             module_spec = importlib.util.find_spec('_HyPhy', [dirname(__file__)])
