diff --git a/bin/configure.sh b/bin/configure.sh
index dd93112..09a92ab 100755
--- bin/configure.sh
+++ bin/configure.sh
@@ -5,7 +5,7 @@
 # Date:   2013/09/03 08:41:11
 #
 # Changes: Jan Engelhardt, 21/1/19
-#
+# 
 
 # Globale Variablen
 SCRIPTNAME=$(basename $0 .sh)
@@ -19,9 +19,9 @@ EXIT_BUG=10
 
 
 ## get the current working directory
-WRK_DIR=`pwd`"/"
-DATA_DIR=`pwd | sed 's/\/[^\/]*$/\/data\//'`
-CONFIG=$WRK_DIR"packages/CONFIG.pm"
+WRK_DIR=$1"/bin/snostrip"
+DATA_DIR=$1"/share/snostrip"
+CONFIG=$1"/lib/snostrip/CONFIG.pm"
 
 echo -n "write header ... "
 
@@ -47,32 +47,31 @@ echo -e "##################################################\n###################
 
 ## write TARGET paths
 echo -e "## TARGET RNAs" >> $CONFIG
-echo -e "\$FUNGI_TARGET_RNA_PATH = \"targetRNAs/fungi/\";" >> $CONFIG
-echo -e "\$PROTOSTOMIA_TARGET_RNA_PATH = \"targetRNAs/protostomia/\";" >> $CONFIG
-echo -e "\$DEUTEROSTOMIA_TARGET_RNA_PATH = \"targetRNAs/deuterostomia/\";" >> $CONFIG
-echo -e "\$PLANT_TARGET_RNA_PATH = \"targetRNAs/plants/\";\n" >> $CONFIG
++echo -e "\$FUNGI_TARGET_RNA_PATH = \"$DATA_DIR/targetRNAs/fungi/\";" >> $CONFIG
++echo -e "\$PROTOSTOMIA_TARGET_RNA_PATH = \"$DATA_DIR/targetRNAs/protostomia/\";" >> $CONFIG
++echo -e "\$DEUTEROSTOMIA_TARGET_RNA_PATH = \"$DATA_DIR/targetRNAs/deuterostomia/\";" >> $CONFIG
++echo -e "\$PLANT_TARGET_RNA_PATH = \"$DATA_DIR/targetRNAs/plants/\";\n" >> $CONFIG
 
 ## write CM MODEL paths
 echo -e "## MODELS" >> $CONFIG
-echo -e "\$FUNGI_CM_PATH = \"models/fungi/\";" >> $CONFIG
-echo -e "\$PROTOSTOMIA_CM_PATH = \"models/protostomia/\";" >> $CONFIG
-echo -e "\$DEUTEROSTOMIA_CM_PATH = \"models/deuterostomia/\";" >> $CONFIG
-echo -e "\$PLANT_CM_PATH = \"models/plants/\";\n" >> $CONFIG
++echo -e "\$FUNGI_CM_PATH = \"$DATA_DIR/models/fungi/\";" >> $CONFIG
++echo -e "\$PROTOSTOMIA_CM_PATH = \"$DATA_DIR/models/protostomia/\";" >> $CONFIG
++echo -e "\$DEUTEROSTOMIA_CM_PATH = \"$DATA_DIR/models/deuterostomia/\";" >> $CONFIG
++echo -e "\$PLANT_CM_PATH = \"$DATA_DIR/models/plants/\";\n" >> $CONFIG
 
 ## write MULTI FASTA paths
 echo -e "## FASTA FILES" >> $CONFIG
-echo -e "\$FUNGI_FASTA_PATH = \"fasta/fungi/\";" >> $CONFIG
-echo -e "\$PROTOSTOMIA_FASTA_PATH = \"fasta/protostomia/\";" >> $CONFIG
-echo -e "\$DEUTEROSTOMIA_FASTA_PATH = \"fasta/deuterostomia/\";" >> $CONFIG
-echo -e "\$PLANT_FASTA_PATH = \"fasta/plants/\";\n" >> $CONFIG
-
++echo -e "\$FUNGI_FASTA_PATH = \"$DATA_DIR/fasta/fungi/\";" >> $CONFIG
++echo -e "\$PROTOSTOMIA_FASTA_PATH = \"$DATA_DIR/fasta/protostomia/\";" >> $CONFIG
++echo -e "\$DEUTEROSTOMIA_FASTA_PATH = \"$DATA_DIR/fasta/deuterostomia/\";" >> $CONFIG
++echo -e "\$PLANT_FASTA_PATH = \"$DATA_DIR/fasta/plants/\";\n" >> $CONFIG
 
 ## write INFORMATION paths
 echo -e "## INFORMATION FILES" >> $CONFIG
-echo -e "\$FUNGI_INFORMATION_FILE = \"information/fungi_information.csv\";" >> $CONFIG
-echo -e "\$PROTOSTOMIA_INFORMATION_FILE = \"information/protostomia_information.csv\";" >> $CONFIG
-echo -e "\$DEUTEROSTOMIA_INFORMATION_FILE = \"information/deuterostomia_information.csv\";" >> $CONFIG
-echo -e "\$PLANT_INFORMATION_FILE = \"information/plant_information.csv\";\n" >> $CONFIG
+echo -e "\$FUNGI_INFORMATION_FILE = \"$DATA_DIR/information/fungi_information.csv\";" >> $CONFIG
+echo -e "\$PROTOSTOMIA_INFORMATION_FILE = \"$DATA_DIR/information/protostomia_information.csv\";" >> $CONFIG
+echo -e "\$DEUTEROSTOMIA_INFORMATION_FILE = \"$DATA_DIR/information/deuterostomia_information.csv\";" >> $CONFIG
+echo -e "\$PLANT_INFORMATION_FILE = \"$DATA_DIR/information/plant_information.csv\";\n" >> $CONFIG
 
 echo -e "done\n"
 
@@ -270,12 +269,12 @@ echo -e "\$FASTACMD = \""$FASTACMD"\";\n" >> $CONFIG
 
 
 ## search for RNAplex
-RNAPLEX=$WRK_DIR"programs/RNAplex"
+RNAPLEX=$WRK_DIR"/programs/RNAplex"
 echo "\$RNAPLEX = \""$RNAPLEX"\";" >> $CONFIG
 
 
 ## search for RNAsnoop
-RNASNOOP=$WRK_DIR"programs/RNAsnoop"
+RNASNOOP=$WRK_DIR"/programs/RNAsnoop"
 echo -e "\$RNASNOOP = \""$RNASNOOP"\";\n\n\n" >> $CONFIG
 
 
@@ -288,25 +287,25 @@ echo -ne "\ncollecting scripts ... "
 echo -e "##################################################\n###################################\n## SCRIPTS\n" >> $CONFIG
 
 ## search for STOCKHOLM
-STK=$WRK_DIR"scripts/stockholm.pl"
+STK=$WRK_DIR"/scripts/stockholm.pl"
 echo "\$STOCKHOLM = \""$STK"\";" >> $CONFIG
 
 ## search for CLUSTER
-CLUSTER=$WRK_DIR"scripts/clusterChains.pl"
+CLUSTER=$WRK_DIR"/scripts/clusterChains.pl"
 echo "\$CLUSTER = \""$CLUSTER"\";" >> $CONFIG
 
 ## search for SVM
-SVM=$WRK_DIR"scripts/apply_svm.pl"
+SVM=$WRK_DIR"/scripts/apply_svm.pl"
 echo "\$SVM = \""$SVM"\";" >> $CONFIG
 
 ## search for INFERNAL
-INF=$WRK_DIR"scripts/infernal.pl"
+INF=$WRK_DIR"/scripts/infernal.pl"
 echo "\$INFERNAL = \""$INF"\";" >> $CONFIG
 
 ## search for BLAST
-BLAST=$WRK_DIR"scripts/blastRelatives.pl"
+BLAST=$WRK_DIR"/scripts/blastRelatives.pl"
 echo -e "\$BLAST = \""$BLAST"\";" >> $CONFIG
 
 ## search for modification mapping
-MAP=$WRK_DIR"scripts/map_modifications.pl"
+MAP=$WRK_DIR"/scripts/map_modifications.pl"
 echo -e "\$MAP_MODIFICATIONS = \""$MAP"\";\n\n" >> $CONFIG
