diff --git a/bin/packages/Fold.pm b/bin/packages/Fold.pm
index 02b93bf..734d540 100755
--- a/bin/packages/Fold.pm
+++ b/bin/packages/Fold.pm
@@ -1,5 +1,15 @@
 package Fold;
 
+use Cwd 'abs_path';
+
+BEGIN{
+    my $path = abs_path($0);
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\//;
+	push @INC, $apath;
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\/packages\//;
+    push @INC, $apath;
+}
+
 use strict;
 use CONFIG;
 use vars qw(@ISA @EXPORT);
diff --git a/bin/packages/Logger.pm b/bin/packages/Logger.pm
index 0605878..cf2dc37 100755
--- a/bin/packages/Logger.pm
+++ b/bin/packages/Logger.pm
@@ -1,5 +1,15 @@
 package Logger;
 
+use Cwd 'abs_path';
+
+BEGIN{
+    my $path = abs_path($0);
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\//;
+	push @INC, $apath;
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\/packages\//;
+    push @INC, $apath;
+}
+
 use strict;
 use vars qw( @ISA @EXPORT );
 
diff --git a/bin/packages/Plexy.pm b/bin/packages/Plexy.pm
index 07fd07e..1dd39d6 100755
--- a/bin/packages/Plexy.pm
+++ b/bin/packages/Plexy.pm
@@ -1,5 +1,14 @@
 package Plexy;
 
+use Cwd 'abs_path';
+
+BEGIN{
+    my $path = abs_path($0);
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\//;
+	push @INC, $apath;
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\/packages\//;
+    push @INC, $apath;
+}
 
 use strict;
 use Bio::SimpleAlign;
diff --git a/bin/packages/Snoopy.pm b/bin/packages/Snoopy.pm
index 1718714..f10b36b 100755
--- a/bin/packages/Snoopy.pm
+++ b/bin/packages/Snoopy.pm
@@ -1,5 +1,14 @@
 package Snoopy;
 
+use Cwd 'abs_path';
+
+BEGIN{
+    my $path = abs_path($0);
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\//;
+	push @INC, $apath;
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\/packages\//;
+    push @INC, $apath;
+}
 
 use strict;
 use Bio::SimpleAlign;
diff --git a/bin/packages/abbreviation.pm b/bin/packages/abbreviation.pm
index 645a6f0..6c855e5 100755
--- a/bin/packages/abbreviation.pm
+++ b/bin/packages/abbreviation.pm
@@ -1,5 +1,14 @@
 package abbreviation;
 
+use Cwd 'abs_path';
+
+BEGIN{
+    my $path = abs_path($0);
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\//;
+	push @INC, $apath;
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\/packages\//;
+    push @INC, $apath;
+}
 
 use strict;
 use vars qw(@ISA @EXPORT);
diff --git a/bin/packages/boxPositions.pm b/bin/packages/boxPositions.pm
index e88d250..7f05da0 100755
--- a/bin/packages/boxPositions.pm
+++ b/bin/packages/boxPositions.pm
@@ -1,5 +1,15 @@
 package boxPositions;
 
+use Cwd 'abs_path';
+
+BEGIN{
+    my $path = abs_path($0);
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\//;
+	push @INC, $apath;
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\/packages\//;
+    push @INC, $apath;
+}
+
 use strict;
 use vars qw(@ISA @EXPORT);
 require Exporter;
diff --git a/bin/packages/boxing.pm b/bin/packages/boxing.pm
index ae4643b..b73866a 100755
--- a/bin/packages/boxing.pm
+++ b/bin/packages/boxing.pm
@@ -1,5 +1,15 @@
 package boxing;
 
+use Cwd 'abs_path';
+
+BEGIN{
+    my $path = abs_path($0);
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\//;
+	push @INC, $apath;
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\/packages\//;
+    push @INC, $apath;
+}
+
 use strict;
 use vars qw(@ISA @EXPORT);
 use Bio::AlignIO;
diff --git a/bin/packages/find_alignment_position.pm b/bin/packages/find_alignment_position.pm
index 63476c5..8b625a0 100755
--- a/bin/packages/find_alignment_position.pm
+++ b/bin/packages/find_alignment_position.pm
@@ -1,5 +1,15 @@
 package find_alignment_position;
 
+use Cwd 'abs_path';
+
+BEGIN{
+    my $path = abs_path($0);
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\//;
+	push @INC, $apath;
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\/packages\//;
+    push @INC, $apath;
+}
+
 use strict;
 use CONFIG;
 use Bio::AlignIO;
diff --git a/bin/packages/makeMuscleAlignment.pm b/bin/packages/makeMuscleAlignment.pm
index 44461ff..12dbbf9 100755
--- a/bin/packages/makeMuscleAlignment.pm
+++ b/bin/packages/makeMuscleAlignment.pm
@@ -2,9 +2,12 @@ package makeMuscleAlignment;
 
 use Cwd 'abs_path';
 
-
 BEGIN{
-    push @INC, abs_path() . "/packages/";
+    my $path = abs_path($0);
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\//;
+	push @INC, $apath;
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\/packages\//;
+    push @INC, $apath;
 }
 
 use strict;
diff --git a/bin/packages/primeBoxes.pm b/bin/packages/primeBoxes.pm
index 91e18f4..9d8f9fb 100755
--- a/bin/packages/primeBoxes.pm
+++ b/bin/packages/primeBoxes.pm
@@ -1,5 +1,14 @@
 package primeBoxes;
 
+use Cwd 'abs_path';
+
+BEGIN{
+    my $path = abs_path($0);
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\//;
+	push @INC, $apath;
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\/packages\//;
+    push @INC, $apath;
+}
 
 use strict;
 use vars qw(@ISA @EXPORT);
diff --git a/bin/packages/searchForTargetSites.pm b/bin/packages/searchForTargetSites.pm
index 9a6b53b..134453b 100755
--- a/bin/packages/searchForTargetSites.pm
+++ b/bin/packages/searchForTargetSites.pm
@@ -1,5 +1,14 @@
 package searchForTargetSites;
 
+use Cwd 'abs_path';
+
+BEGIN{
+    my $path = abs_path($0);
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\//;
+	push @INC, $apath;
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\/packages\//;
+    push @INC, $apath;
+}
 
 use strict;
 use vars qw(@ISA @EXPORT);
diff --git a/bin/packages/sequences.pm b/bin/packages/sequences.pm
index ddcbf37..7bc8b68 100755
--- a/bin/packages/sequences.pm
+++ b/bin/packages/sequences.pm
@@ -1,5 +1,15 @@
 package sequences;
 
+use Cwd 'abs_path';
+
+BEGIN{
+    my $path = abs_path($0);
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\//;
+	push @INC, $apath;
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\/packages\//;
+    push @INC, $apath;
+}
+
 use strict;
 use CONFIG;
 use vars qw(@ISA @EXPORT);
diff --git a/bin/packages/tools.pm b/bin/packages/tools.pm
index 1883d6a..b569bd7 100644
--- a/bin/packages/tools.pm
+++ b/bin/packages/tools.pm
@@ -1,5 +1,15 @@
 package tools;
 
+use Cwd 'abs_path';
+
+BEGIN{
+    my $path = abs_path($0);
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\//;
+	push @INC, $apath;
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\/packages\//;
+    push @INC, $apath;
+}
+
 use strict;
 use Bio::SeqIO;
 use vars qw(@ISA @EXPORT);
diff --git a/bin/snoStrip.pl b/bin/snoStrip.pl
index 158a164..aef209d 100755
--- a/bin/snoStrip.pl
+++ b/bin/snoStrip.pl
@@ -1,12 +1,13 @@
 #!/usr/bin/env perl
 
-
 use Cwd 'abs_path';
 
 BEGIN{
     my $path = abs_path($0);
-    $path =~ s/snoStrip.pl/packages\//;
-    push @INC, $path;
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\//;
+	push @INC, $apath;
+	(my $apath = $path) =~ s/snoStrip.pl/\.\.\/lib\/snostrip\/packages\//;
+    push @INC, $apath;
 }
 
 use warnings;
