diff --git a/setup.py b/setup.py
index 7e8b201..cb9a3b9 100644
--- a/setup.py
+++ b/setup.py
@@ -1,6 +1,7 @@
 import os
 import sys
-from setuptools import Extension, setup, find_packages
+from setuptools import Extension, setup, find_namespace_packages
+from Cython.Build import cythonize
 import numpy as np
 
 
@@ -38,13 +39,13 @@ setup(
         "Topic :: Utilities",
         "License :: OSI Approved :: BSD License",
     ],
-    packages=find_packages(where="."),
+    packages=find_namespace_packages(where="."),
     ext_modules=[
         Extension(name="iced._filter_",
                   sources=["iced/_filter_.pyx"],
                   include_dirs=[np.get_include()]
         ),
-        Extension(name="iced.normalization/_normalization_",
+        Extension(name="iced.normalization._normalization_",
                   sources=["iced/normalization/_normalization_.pyx"],
                   include_dirs=[np.get_include()]
                   )],
