From d6ba42762d6daa716a06b1c99b5660d1a63c58d1 Mon Sep 17 00:00:00 2001
From: Anders Goncalves da Silva <andersgs@gmail.com>
Date: Sun, 28 Jul 2019 22:56:18 -0700
Subject: [PATCH] Fix Makefile

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index e18c3f3..41e747d 100644
--- a/Makefile
+++ b/Makefile
@@ -12,7 +12,7 @@ endif
 all: lighter
 
 lighter: main.o $(OBJECTS)
-	$(CXX) -o $@ $(CXXFLAGS) $(OBJECTS) main.o $(LINKFLAGS)
+	$(CXX) -o $@ $(CXXFLAGS) $(OBJECTS) main.o $(LDFLAGS) $(LINKFLAGS)
 
 main.o: main.cpp utils.h Reads.h Store.h File.h KmerCode.h bloom_filter.hpp
 ErrorCorrection.o: ErrorCorrection.cpp ErrorCorrection.h utils.h
-- 
2.20.1 (Apple Git-117)

