
# Getting Started
```sh
LocaTE -b <sorted.BAM> -g <TE.gff> 
```
# Table of Contents
[toc]

# Users' Guide
LocaTE is a pipeline to quantify locus-specific expression of transposable elements (TEs) in short read RNA-seq. LocaTE is implemented in Bash only till now, and requires no installation. However, other softwares that are part of the pipeline are required. Users could choose the available way to install.

## Installation
### Anaconda
LocaTE is available on conda and can be installed using conda. 
```sh
conda install jiaqiang_987::locate
```

### Dependencies
If conda does not successfully install, LocaTE could be obtained from github (https://github.com/PengjuZ/LocaTE). Then download the required softwares as following, and add them to the `PATH` variable. After that, recalling by absolute path. 
* Glibc == 2.17
* samtools == 1.17
* bedtools == 2.31.1
* python == 3.11
* numpy == 2.1.40
* pandas == 2.2.2
* scikit-learn == 1.5.1

### Test files
Two files `examples/test.bam` & `examples/testTE.gff` are provided within LocaTE for testing its operation. Once LocaTE and its dependences are in your `PATH` variable, you can execute it as follows. Normally, this process takes in seconds. By default, a output file `test.bam_TE` would be generated in the current directory. 
```sh 
LocaTE -b test.bam -g testTE.gff
```

## General usage
Once LocaTE and its dependences are in your `PATH` variable, you can execute it as:
```sh
LocaTE -b <sorted.BAM> -g <TE.gff> 
```
Important considerations:
* `-b` <b>BAM file must be sorted.</b>
* `-g` Notably, our testing has been conducted solely with the <b>GFF file from RepeatMaske.</b> Thus, the TE annotation file shall be generated by RepeatMasker, or as followsï¼š
```
1       RepeatMasker    dispersed_repeat        1604    1768    35.8    +       .       ID=1;Target "Motif:L2c" 2993 3159
1       RepeatMasker    dispersed_repeat        3973    4154    25.9    -       .       ID=2;Target "Motif:MIRb" 46 240
1       RepeatMasker    dispersed_repeat        10184   10306   29.3    -       .       ID=3;Target "Motif:L1MC5" 7810 7936
1       RepeatMasker    dispersed_repeat        11130   11365   30.1    -       .       ID=4;Target "Motif:MIRb" 23 260
```
Complete parameters `-h`:
```sh
        -b <BAM file (sorted)>
        -g <TE annotation file (GFF)>
        -o <Output prefix> (default: working directory)
        -f <Minimum overlap requirement> (default: 0.5)
        -p <Maximum likelihood value for assignment> (default: 0.8)
        -m <Maximum iterations of the outer cycle> (default: 10)
        -M <Maximum iterations of the inner cycle> (default: 10)
        -e <Abnormal ERV detection> (default: Flase)
```
## Advanced features
### 1. Cleaning the TE annotation file (gff)
In order to clean the silent TEs. A TE database is constructed from the long reads of RNA-seq in several domestic animals. Based on this database, types of silent TE from GFF file would be discarded. Then returning another GFF file for quantification. 
Presently, this function `GFFfilter` is limited to a selection of domestic animals, specifically pigs, cattle, sheep, goats, and chickens. For the other species, users could customize the TE annotation file to adjust the active TE subfamilies.
```sh
GFFfilter -s <species> -g <TE.gff>
```

### 2. Detecting anomalous expression of ERVs
LocaTE also comes with a python script that detects the anomalous  expression of ERVs by `-e`. It statistics the ratio of each ERV between RNA-seq and genome. Based on DBSCAN, the ERVs with anomalous value are deemed as exogenous.
For the specific parameters, Euclidean distance is used as the metric for computation. And the eps set to 3. Moreover, users could modify them in that python script file `abnormal_ERVs.py` to increase or decrease its sensitivity. Additionally, users could modify the textfile `ERVFamily` to detect the specific TE subfamilies.
```sh
LocaTE -b <bam> -g <TE.gff> -e
```

### 3. Expected probability threshold
This parameter `-p` determines the assignment of a multi-mapping TE read indicated by its expected probability. Increasing this parameter yields more accurate assignments but extends processing time due to fewer multi-mapped reads being resolved per cycle.
```sh
LocaTE -b <bam> -g <TE.gff> -p 0.8 
```

### 4.Minimum overlap requirement
Due to most TEs are predominantly active in fragment. Therefore, the default standard is relatively loose to prevent the loss of TE reads. Herein, Users have the flexibility to adjust the key parameters, overlapping rate `-f`, to manage the number of TE reads. This step ensures a balance between capturing a comprehensive representation of TE expression and maintaining the integrity of the TE reads.
```sh
# Reporting only those intersections where 50% of the read is overlapped by a TE.gff (default)
LocaTE -b <bam> -g <TE.gff> -o <output> -f 0.5
# Reporting only those intersections where 100% of the read is overlapped by a TE.gff
LocaTE -b <bam> -g <TE.gff> -o <output> -f 1
```

## Citing 
```
TBD
```

