--- src.CMakeLists.txt	2017-01-17 09:36:04.786802923 +0400
+++ src/CMakeLists.txt	2017-01-17 09:38:13.508652701 +0400
@@ -2,20 +2,20 @@
 project (Pre_Alignment)
 
 set(CMAKE_CXX_COMPILER "g++")
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O3")
+set(CMAKE_CXX_FLAGS "$ENV{CMAKE_CXX_FLAGS} -O3")
 
 #Warning: The followings need your settings...(we have shown some examples below them.)
-set(BOOST_LIB_DIR <Path_To_BOOST_LIB_DIR>)
+set(BOOST_LIB_DIR $ENV{BOOST_LIBRARY_DIR})
 #Example: set(BOOST_LIB_DIR /home/juntaosdu/local/boost/lib)
-set(BOOST_INCLUDE_DIR <Path_To_BOOST_INCLUDE_DIR>)
+set(BOOST_INCLUDE_DIR $ENV{BOOST_INCLUDE_DIR})
 #Example: set(BOOST_INCLUDE_DIR /home/juntaosdu/local/boost/include)
-set(BAMTOOLS_LIB_DIR <Path_To_BAMTOOLS_LIB_DIR>)
+set(BAMTOOLS_LIB_DIR $ENV{BAMTOOLS_LIBRARY_DIR})
 #Example: set(BAMTOOLS_LIB_DIR /home/juntaosdu/bamtools/lib)
-set(BAMTOOLS_INCLUDE_DIR <Path_To_BAMTOOLS_INCLUDE_DIR>)
+set(BAMTOOLS_INCLUDE_DIR $ENV{BAMTOOLS_INCLUDE_DIR})
 #Example: set(BAMTOOLS_INCLUDE_DIR /home/juntaosdu/bamtools/include)
 
-include_directories(. ${BOOST_INCLUDE_DIR} ${BAMTOOLS_INCLUDE_DIR})
-link_directories(${BOOST_LIB_DIR} ${BAMTOOLS_LIB_DIR})
+include_directories(. $ENV{BOOST_INCLUDE_DIR} $ENV{BAMTOOLS_INCLUDE_DIR})
+link_directories($ENV{BOOST_LIB_DIR} $ENV{BAMTOOLS_LIB_DIR})
 
 #add_executable(Assemble Graph_NH_old_combine_filter.cc expression_level.h
 #Find_junc_last_map.h Insert.h Sort.h edge_node_pair.h False_junc.h sep_gene.h
