diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index a9afb6f..66ff9d5 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -1,21 +1,20 @@
 cmake_minimum_required(VERSION 2.6)
 project (Pre_Alignment)
 
-set(CMAKE_CXX_COMPILER "g++")
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O3")
+set(CMAKE_CXX_FLAGS "$ENV{CMAKE_CXX_FLAGS} -O3")
 
 #Warning: The followings need your settings...(we have shown some examples below them.)
-set(BOOST_LIB_DIR <Path_To_BOOST_LIB_DIR>)
+set(BOOST_LIB_DIR $ENV{BOOST_LIB_DIR])
 #example: set(BOOST_LIB_DIR /rhome/jliu/bigdata/local/boost/lib)
-set(BOOST_INCLUDE_DIR <Path_To_BOOST_INCLUDE_DIR>)
+set(BOOST_INCLUDE_DIR $ENV{BOOST_INCLUDE_DIR})
 #example: set(BOOST_INCLUDE_DIR /rhome/jliu/bigdata/local/boost/include)
-set(BAMTOOLS_LIB_DIR <Path_To_BAMTOOLS_LIB_DIR>)
+set(BAMTOOLS_LIB_DIR $ENV{BAMTOOLS_LIB_DIR})
 #example: set(BAMTOOLS_LIB_DIR /rhome/jliu/bamtools/lib)
-set(BAMTOOLS_INCLUDE_DIR <Path_To_BAMTOOLS_INCLUDE_DIR>)
+set(BAMTOOLS_INCLUDE_DIR $ENV{BAMTOOLS_INCLUDE_DIR})
 #example: set(BAMTOOLS_INCLUDE_DIR /rhome/jliu/bamtools/include)
 
-include_directories(. ${BOOST_INCLUDE_DIR} ${BAMTOOLS_INCLUDE_DIR})
-link_directories(${BOOST_LIB_DIR} ${BAMTOOLS_LIB_DIR})
+include_directories(. $ENV{BOOST_INCLUDE_DIR} $ENV{BAMTOOLS_INCLUDE_DIR})
+link_directories($ENV{BOOST_LIB_DIR} $ENV{BAMTOOLS_LIB_DIR})
 
 #add_executable(Assemble Graph_NH_old_combine_filter.cc expression_level.h
 #Find_junc_last_map.h Insert.h Sort.h edge_node_pair.h False_junc.h sep_gene.h
