diff --git a/Makefile b/Makefile
index 265c1d3..b52f08a 100644
--- a/Makefile
+++ b/Makefile
@@ -16,11 +16,11 @@ else
 VPATH = $(SRC_DIR)
 
 # Common C and C++ flags
-CCXXFLAGS:=-O3 -DNDEBUG -march=native -Wall -pthread
+CCXXFLAGS:=-Wall -pthread
 # C-only flags
 CFLAGS+= $(CCXXFLAGS)
 # C++-only flags
-CXXFLAGS+= $(CCXXFLAGS) -std=c++1z
+CXXFLAGS+= $(CCXXFLAGS) 
 # Linker flags
 LDFLAGS+=-pthread
 
