--- setup.py.orig	2017-11-07 19:39:09.207844287 +0200
+++ setup.py	2017-11-07 19:41:39.586927513 +0200
@@ -16,10 +16,6 @@
 	print >> sys.stderr, "ERROR: CPAT only works on Python2.x and requires Python 2.7 or greater"
 	sys.exit()
 
-# Automatically download setuptools if not available
-from distribute_setup import use_setuptools
-use_setuptools()
-
 from setuptools import *
 from glob import glob
 
@@ -32,12 +28,12 @@
 def main():
 	setup(  name = "CPAT",
             version = "1.2.3",
-            py_modules = [ 'psyco_full' ],
-            packages = find_packages( 'lib' ),
+            py_modules = [ ],
+            packages = ['cpmodule'],
             package_dir = { '': 'lib' },
             package_data = { '': ['*.ps'] },
             scripts = glob( "bin/*.py"),
-            ext_modules = get_extension_modules(),
+            ext_modules = [],
             test_suite = 'nose.collector',
             setup_requires = ['nose>=0.10.4','cython>=0.12'],
             author = "Liguo Wang",
@@ -51,8 +47,6 @@
 
 # ---- Commands -------------------------------------------------------------
 
-from distutils.core import Command
-
 # Use build_ext from Cython
 command_classes = {}
 
@@ -63,43 +57,6 @@
 except:
 	pass
     
-# Use epydoc if found
-try:
-	import pkg_resources
-	pkg_resources.require( "epydoc" )
-	import epydoc.cli, sys, os, os.path
-	# Create command class to build API documentation
-	class BuildAPIDocs( Command ):
-		user_options = []
-		def initialize_options( self ):
-			pass
-		def finalize_options( self ):
-			pass
-		def run( self ):
-			# Save working directory and args
-			old_argv = sys.argv
-			old_cwd = os.getcwd()
-			# Build command line for Epydoc
-			sys.argv = """epydoc.py bx --verbose --html --simple-term
-                                       --exclude=._
-                                       --exclude=_tests
-                                       --docformat=reStructuredText
-                                       --output=../doc/docbuild/html/apidoc""".split()
-			# Make output directory
-			if not os.path.exists( "./doc/docbuild/html/apidoc" ):
-				os.mkdir( "./doc/docbuild/html/apidoc" )
-			# Move to lib directory (so bx package is in current directory)
-			os.chdir( "./lib" )
-			# Invoke epydoc
-			epydoc.cli.cli()
-			# Restore args and working directory
-			sys.argv = old_argv
-			os.chdir( old_cwd )
-	# Add to extra_commands    
-	command_classes['build_apidocs'] = BuildAPIDocs
-except:
-	pass
-
 # ---- Extension Modules ----------------------------------------------------
 
 def get_extension_modules():
