diff --git a/v-annotate.pl b/v-annotate.pl
index 25227c0..d4f01f1 100755
--- a/v-annotate.pl
+++ b/v-annotate.pl
@@ -160,7 +160,7 @@
 $execs_H{"esl-reformat"}  = $env_vadr_easel_dir    . "/esl-reformat";
 $execs_H{"esl-seqstat"}   = $env_vadr_easel_dir    . "/esl-seqstat";
 $execs_H{"esl-translate"} = $env_vadr_easel_dir    . "/esl-translate";
-$execs_H{"esl-ssplit"}    = $env_vadr_bioeasel_dir . "/scripts/esl-ssplit.pl";
+$execs_H{"esl-ssplit"}    = $env_vadr_bioeasel_dir . "/esl-ssplit.pl";
 $execs_H{"blastx"}        = $env_vadr_blast_dir    . "/blastx";
 $execs_H{"blastn"}        = $env_vadr_blast_dir    . "/blastn";
 $execs_H{"makeblastdb"}   = $env_vadr_blast_dir    . "/makeblastdb";
