diff --git a/testfiles/dengue.r5.local.testin b/testfiles/dengue.r5.local.testin
index 0832237..92fdce4 100644
--- a/testfiles/dengue.r5.local.testin
+++ b/testfiles/dengue.r5.local.testin
@@ -12,7 +12,7 @@
 # 'out:' file is compared against the first 'exp:' file, the second 
 # 'out:' file is compared against the second 'exp:' file, the second 'out:'
 # file, etc.
-command: perl $VADRSCRIPTSDIR/v-annotate.pl -f --mkey flavi --mdir $VADRINSTALLDIR/vadr-models-flavi --alicheck $VADRSCRIPTSDIR/testfiles/dengue.r5.fa va-dengue.r5 > va-dengue.r5.out
+command: perl $VADRSCRIPTSDIR/v-annotate.pl -f --mkey flavi --mdir $VADRMODELDIR --alicheck $VADRSCRIPTSDIR/testfiles/dengue.r5.fa va-dengue.r5 > va-dengue.r5.out
 desc: annotate-dengue-5-local
 out: va-dengue.r5/va-dengue.r5.vadr.pass.tbl 
 out: va-dengue.r5/va-dengue.r5.vadr.fail.tbl 
