--- general/cfareader.hpp
+++ general/cfareader.hpp
@@ -40 +40 @@
-    m_good = std::getline( *m_fin, m_buffer );
+    m_good = static_cast<bool>(std::getline( *m_fin, m_buffer ));
@@ -46 +46 @@
-    m_good = std::getline( *m_fin, m_buffer );
+    m_good = static_cast<bool>(std::getline( *m_fin, m_buffer ));
--- general/of-debug.hpp
+++ general/of-debug.hpp
@@ -10 +10 @@
-#define VT100(c) "\x1b["c
+#define VT100(c) "\x1b[" c
