diff --git a/Makefile b/Makefile
index 5edd6e1..9ee257f 100755
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,6 @@
-CPP := g++
-CC := gcc
-LINK := g++
+CPP := ${CXX}
+CC := ${CC}
+LINK := ${CXX}
 mkfile_dir := $(dir $(abspath $(lastword $(MAKEFILE_LIST))))
 
 CSOURCES= $(wildcard src/*.c)  
@@ -11,12 +11,12 @@ COMMONOBJS= lib/libhts.a lib/libboost_iostreams.a lib/libtcmalloc_minimal.a
 Warnings=-Wreturn-type -Warray-bounds -Wmaybe-uninitialized -Waddress
 WarningsAsErrors=$(Warnings) -Werror=return-type -Werror=array-bounds -Werror=address
 CFLAGS=  $(WarningsAsErrors) -Wno-unused-function
-CPPFLAGS=  $(WarningsAsErrors) -Wno-unused-function -std=c++11
+CPPFLAGS=  $(WarningsAsErrors) -Wno-unused-function -std=c++14
 
 RELEASE_FLAGS= -O3 -g
 
 # Includes
-INCLUDES = -Iinc/
+INCLUDES += -Iinc/
 #
 # Common flags
 COMMONFLAGS += $(INCLUDES)
@@ -24,7 +24,7 @@ COMMONFLAGS += $(INCLUDES)
 CXXFLAGS += $(COMMONFLAGS)
 CFLAGS += $(COMMONFLAGS)
 CPPFLAGS += $(COMMONFLAGS)
-COMMONLIBS= -Llib/ -lz -lm -lpthread -lbz2 -lcurl -lcrypto -llzma -fopenmp
+COMMONLIBS += -Llib/ -lz -lm -lpthread -lbz2 -lcurl -lcrypto -llzma -fopenmp -L${PREFIX}/lib
 
 #LIBS += $(COMMONLIBS) -ltcmalloc
 LIBS += $(COMMONLIBS)
