diff --git a/install_muse.sh b/install_muse.sh
index 7bc2153..cc6f3ac 100755
--- a/install_muse.sh
+++ b/install_muse.sh
@@ -5,10 +5,12 @@ make clean
 mkdir -p lib
 
 # compile boost library
-cd boost_1_70_0/
-./bootstrap.sh
+cd boost-1.86.0/
+./bootstrap.sh --with-toolset=gcc
 ./b2 --clean
-./b2
+./b2 --prefix="${BOOST_ROOT}" toolset=gcc --layout=system link=static \
+	threading=multi strip=on include="${PREFIX}/include" \
+	variant=release address-model=64 install "${BOOST_BUILD_LIBS}" > "${BOOST_LOG}" 2>&1
 
 cp stage/lib/libboost_iostreams.a ../lib/
 cd ..
@@ -22,10 +24,10 @@ fi
 
 git clone https://github.com/gperftools/gperftools.git
 cd gperftools
-git checkout gperftools-2.9.1
+git checkout gperftools-2.16
 ./autogen.sh
 ./configure --libdir="$PWD"
-make -j4
+make -j${CPU_COUNT}
 make install
 cp libtcmalloc_minimal.a ../lib/
 cd ..
@@ -39,30 +41,19 @@ fi
 
 git clone https://github.com/samtools/htslib.git
 cd htslib
-git checkout 1.9
+git checkout 1.21
 #git submodule update --init --recursive
 autoreconf -i
 ./configure
-make -j4
+make -j${CPU_COUNT}
 cp libhts.a ../lib/
 
 cd ..
 
-make
-
-#clean htslib and gperftools and boost libraries
-rm -r boost_1_70_0/stage
-
-chmod -R 755 gperftools
-rm -r gperftools
-
-chmod -R 755 htslib
-rm -r htslib
-
-
-
-
-
+CFLAGS="${CFLAGS} -O3 ${LDFLAGS}" \
+    CC="${CC}" CXX="${CXX}" CXXFLAGS="${CXXFLAGS} -O3 -I${PREFIX}/include" \
+    LDFLAGS="${LDFLAGS} -L${PREFIX}/lib" \
+    make -j"${CPU_COUNT}"
 
 
 
