diff -ru qiime-1.8.0/scripts/parallel_merge_otu_tables.py qiime-1.8.0.patched/scripts/parallel_merge_otu_tables.py
--- qiime-1.8.0/scripts/parallel_merge_otu_tables.py	2013-12-12 02:53:03.000000000 +0000
+++ qiime-1.8.0.patched/scripts/parallel_merge_otu_tables.py	2018-05-17 10:23:27.200856236 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/home/pjb/scratchpad/bioconda-qiime-1.8.0/miniconda2/envs/qiime@1.8.0/bin/python
 
 from __future__ import division
 
@@ -14,7 +14,7 @@
 from qiime.util import parse_command_line_parameters,\
     load_qiime_config, get_options_lookup, get_qiime_scripts_dir
 from os import popen, system, makedirs, mkdir
-from os.path import split, splitext, join
+from os.path import split, splitext, join, exists, dirname
 from subprocess import check_call, CalledProcessError
 from qiime.util import get_tmp_filename
 from cogent.core.tree import TreeNode
@@ -39,13 +39,17 @@
  make_option('-o','--output_dir',type='new_dirpath',
              help='the output otu table directory path')]
 
+default_merge_otu_tables = join(get_qiime_scripts_dir(),'merge_otu_tables.py')
+if not exists(default_merge_otu_tables):
+    default_merge_otu_tables = join(dirname(__file__),'merge_otu_tables.py')
+
 script_info['optional_options'] = [\
     make_option('-C','--cluster',action='store_true', default=False,
            help="Submit to a torque cluster"),
     make_option('-N','--merge_otus_fp',action='store',\
            type='existing_filepath',help='full path to '+\
            'scripts/merge_otu_tables.py [default: %default]',\
-           default=join(get_qiime_scripts_dir(),'merge_otu_tables.py')),\
+           default=default_merge_otu_tables),\
     options_lookup['python_exe_fp'],
     options_lookup['seconds_to_sleep'],
     options_lookup['job_prefix']]   
