--- plasmidfinder.py	2019-07-24 21:11:06.512992219 -0500
+++ plasmidfinder.py.new	2019-07-24 23:27:42.330109551 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/opt/anaconda1anaconda2anaconda3/bin/python
 from __future__ import division
 from argparse import ArgumentParser
 from tabulate import tabulate
@@ -122,7 +122,7 @@
 parser.add_argument("-o", "--outputPath", dest="outdir",help="Path to blast output", default='.')
 parser.add_argument("-tmp", "--tmp_dir", help="Temporary directory for storage of the results from the external software.")
 parser.add_argument("-mp", "--methodPath", dest="method_path",help="Path to method to use (kma or blastn)")
-parser.add_argument("-p", "--databasePath", dest="db_path",help="Path to the databases", default='/database')
+parser.add_argument("-p", "--databasePath", dest="db_path",help="Path to the databases", default='$PLASMID_DB')
 parser.add_argument("-d", "--databases", dest="databases",help="Databases chosen to search in - if non is specified all is used")
 parser.add_argument("-l", "--mincov", dest="min_cov",help="Minimum coverage", default=0.60)
 parser.add_argument("-t", "--threshold", dest="threshold",help="Minimum threshold for identity", default=0.90)
