--- a/plasmidfinder.py
+++ b/plasmidfinder.py
@@ -14,6 +14,8 @@ import json
 import gzip
 import pprint
 
+PLASMIDFINDER_SHARE="BIOCONDA_SED_REPLACE"
+
 ##########################################################################
 # TODO list
 ##########################################################################
@@ -158,7 +160,7 @@ parser.add_argument("-mp", "--methodPath",
 parser.add_argument("-p", "--databasePath",
                     dest="db_path",
                     help="Path to the databases",
-                    default='/database')
+                    default=PLASMIDFINDER_SHARE)
 parser.add_argument("-d", "--databases",
                     help=("Databases chosen to search in - if non is "
                           "specified all is used"))
