--- plasmidfinder.py	2019-08-28 20:35:11.149351164 +0200
+++ plasmidfinder.py2	2019-08-28 20:42:17.282431720 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/opt/anaconda1anaconda2anaconda3/bin/python
 from __future__ import division
 from argparse import ArgumentParser
 from tabulate import tabulate
@@ -158,7 +158,7 @@
 parser.add_argument("-p", "--databasePath",
                     dest="db_path",
                     help="Path to the databases",
-                    default='/database')
+                    default=os.getenv("PLASMID_DB"))
 parser.add_argument("-d", "--databases",
                     help=("Databases chosen to search in - if non is "
                           "specified all is used"))
