diff --git i/src/genoiser.nim w/src/genoiser.nim
index ca248ec..bb29cc8 100644
--- i/src/genoiser.nim
+++ w/src/genoiser.nim
@@ -72,7 +72,7 @@ iterator mranges*[T](depth: var seq[T], values: var seq[T]): mpair {.inline.} =
     yield (last_i, len(depth), last_pair[0].int, last_pair[1].int)
 
 type
-  Fun* {.shallow.} [T]  = ref object
+  Fun*[T] {.shallow.} = ref object
     values*: seq[T]
     f*: proc(aln:Record, posns:var seq[mrange])
