diff --git a/Makefile b/Makefile
index b4661737e..7f9726d3a 100644
--- a/Makefile
+++ b/Makefile
@@ -16,7 +16,7 @@
 # OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 #
 
-INCLUDEOPT:=-I$(CURDIR)/src -I$(CURDIR)/obj
+INCLUDEOPT:=-I$(PREFIX)/include/cairo -I$(PREFIX)/include/pango-1.0 -I$(PREFIX)/include/harfbuzz -I$(CURDIR)/src -I$(CURDIR)/obj
 
 ifeq ($(shell pkg-config --version > /dev/null 2> /dev/null; echo $$?),0)
   HAS_PKGCONFIG:=yes
