diff --git a/gtpython/setup.py b/gtpython/setup.py
index f796f5b87..291786a55 100644
--- a/gtpython/setup.py
+++ b/gtpython/setup.py
@@ -7,7 +7,7 @@ if sys.version_info < (2, 6, 0):
     sys.stderr.write("Requires Python 2.6 or newer.\n")
     sys.exit(-1)
 
-from distutils.core import setup
+from setuptools import setup, find_namespace_packages
 
 def get_version():
     VERSIONFILE="../VERSION"
@@ -20,6 +20,5 @@ setup(
     author='Sascha Steinbiss',
     author_email='steinbiss@zbh.uni-hamburg.de',
     url= 'http://www.genometools.org ',
-    packages=['gt', 'gt.core',
-             'gt.annotationsketch', 'gt.extended']
+    packages=find_namespace_packages(where="gt")
 )
