From cc6099624ed6b5dde851e402b55d57dc133ac6ab Mon Sep 17 00:00:00 2001
From: "Robert A. Petit III" <robbie.petit@gmail.com>
Date: Wed, 1 Jun 2022 19:03:15 -0600
Subject: [PATCH] enforce sample_name and add outdir option

---
 shigatyper/shigatyper.py | 30 ++++++++----------------------
 1 file changed, 8 insertions(+), 22 deletions(-)

diff --git a/shigatyper/shigatyper.py b/shigatyper/shigatyper.py
index 791496b..13d1e72 100755
--- a/shigatyper/shigatyper.py
+++ b/shigatyper/shigatyper.py
@@ -1,21 +1,6 @@
 #!/usr/bin/env python3
 # coding: utf-8
-"""
-usage: shigatyper.py [-h] [--R1 FASTA] [--R2 FASTA] [--SE FASTA] [--ont] [-n SAMPLE_NAME] [--verbose] [--version]
-
-A WGS-based genoserotyping pipeline for Shigella spp.
-
-options:
-  -h, --help            show this help message and exit
-  --R1 FASTA            Input FASTQ is R1 of paired-end reads
-  --R2 FASTA            Input FASTQ is R2 of paired-end reads
-  --SE FASTA            Input FASTQ is contains single-end reads
-  --ont                 The input FASTQ file contains ONT reads
-  -n SAMPLE_NAME, --name SAMPLE_NAME
-  --verbose, -v
-  --version             show program's version number and exit
-"""
-version = "2.0.0"
+version = "2.0.1"
 usage = f"""ShigaTyper v. {version}, 2022
 
 A WGS-based genoserotyping pipeline for Shigella spp.
@@ -85,7 +70,7 @@ def readable(tdelta):
     else:
         return f"{tdelta.seconds // 3600} hours, {tdelta.seconds // 60 : 01} seconds"
 
-def run(reads, tempdir, sample_name='', threshold=50, rlog=rlog, ont=False, *args, **kwargs):
+def run(reads, tempdir, sample_name='', threshold=50, rlog=rlog, ont=False, outdir='./', *args, **kwargs):
     # ## 3. Map Filtered reads to Reference sequence database (ShigellaRef5)
     rlog.critical(f"ShigaTyper v{version}")
 
@@ -506,7 +491,7 @@ def wzxindex(Hits):
     if checkpoint == 1:
         pass
     else:
-        output = f"{sample_name}-hits.tsv"
+        output = f"{outdir}/{sample_name}-hits.tsv"
         log.critical(f"Writing hits to {output}")
         if "ipaH_c" in Maphits.Hit.tolist():
             List2.to_csv(output, sep="\t")
@@ -515,9 +500,9 @@ def wzxindex(Hits):
             Maphits.to_csv(output, sep="\t")
             Maphits.to_csv(sys.stdout, sep="\t")
 
-    serotype_output = f"{sample_name}.tsv"
+    serotype_output = f"{outdir}/{sample_name}.tsv"
     log.critical(f"Writing final serotype prediction to {serotype_output}")
-    with open(f"{sample_name}.tsv", "wt") as fh_out:
+    with open(serotype_output, "wt") as fh_out:
         ipab = ('-','+')[bool(ipaB)]
         results = f"sample\tprediction\tipaB\tnotes\n{sample_name}\t{prediction}\t{ipab}\t{note}"
         print(results)
@@ -556,6 +541,7 @@ def main():
     parse.add_argument('--ont', action="store_true",
                         help="The input FASTQ file contains ONT reads")
     parse.add_argument('-n', '--name', dest='sample_name')
+    parse.add_argument('-o', '--outdir', help='Where to write output files to', default="./")
     parse.add_argument('--verbose', '-v', action='count', dest='log_v', default=0)
     parse.add_argument('--version', action='version', version=f"ShigaTyper {version}")
 
@@ -599,9 +585,9 @@ def main():
         tempdir = tempfile.mkdtemp()
         reads = []
         if is_paired:
-            run([args.R1, args.R2], tempdir=tempdir, ont=False)
+            run([args.R1, args.R2], tempdir=tempdir, ont=False, sample_name=args.sample_name, outdir=args.outdir)
         else:
-            run([args.SE], tempdir=tempdir, ont=args.ont)
+            run([args.SE], tempdir=tempdir, ont=args.ont, sample_name=args.sample_name, outdir=args.outdir)
     except CalledProcessError as e:
         rlog.error(e)
         if e.stderr:
