--- Makefile	2021-08-19 21:31:58.600065725 +0200
+++ Makefile.new	2021-08-19 21:49:45.083756372 +0200
@@ -1,6 +1,6 @@
 CC = gcc 
 
-CFLAGS = -Wall -D_GNU_SOURCE
+CFLAGS = -Wall -D_GNU_SOURCE -fcommon
 LFLAGS = -lm
 
 ifeq ($(mode), debug)
@@ -33,20 +33,20 @@
 prune-objs = rnr-prune.o common.o Tree.o BitVector.o HashTable.o  legacy.o newFunctions.o List.o
 
 rnr-lsi: $(lsi-objs)
-	$(CC) $(CFLAGS) -o $@ $^ $(LFLAGS) 
+	$(CC) $(CFLAGS) -o $@ $^ $(LFLAGS)
 rnr-tii: $(tii-objs)
-	$(CC) $(CFLAGS)  -o $@ $^ $(LFLAGS) 
+	$(CC) $(CFLAGS)  -o $@ $^ $(LFLAGS)
 rnr-mast: $(mast-objs)
-	$(CC) $(CFLAGS) -o $@ $^ $(LFLAGS) 
+	$(CC) $(CFLAGS) -o $@ $^ $(LFLAGS)
 rnr-prune: $(prune-objs)
-	$(CC) $(CFLAGS) -o $@ $^ $(LFLAGS) 
+	$(CC) $(CFLAGS) -o $@ $^ $(LFLAGS)
 
 ifeq ($(mode),parallel)
 RogueNaRok: $(rnr-objs)
-	$(CC) $(CFLAGS) -o $@-parallel $^ $(LFLAGS) 
+	$(CC) $(CFLAGS) -o $@-parallel $^ $(LFLAGS)
 else
 RogueNaRok: $(rnr-objs)
-	$(CC) $(CFLAGS) -o $@ $^ $(LFLAGS) 
+	$(CC) $(CFLAGS) -o $@ $^ $(LFLAGS)
 endif
 
 %.o : %.c $(DEPS)
