--- Makefile
+++ Makefile
@@ -1,10 +1,10 @@
 dnptools=binstrings corrprofile diprofile
 
 all: $(dnptools) fourier
 
 $(dnptools): % : %.cpp
-	gcc -std=c++11 -std=c++14 -pthread -I include/ $< -lstdc++ -lrt -lm -o $@  
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) $(LDFLAGS) -std=c++11 -std=c++14 -pthread -I include/ $< -lstdc++ -lrt -lm -o $@  
 
 fourier: Fourier_Transform.cpp
-	gcc Fourier_Transform.cpp -lm -o fourier
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) $(LDFLAGS) Fourier_Transform.cpp -lm -o fourier
 
