diff --git a/Makefile b/Makefile
index afa36f5..6a8de01 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
-CC=gcc
-CPP=g++
+CC?=$(CC)
+CPP?=$(CPP)
 OPTIMISER = -O3
-CPPFLAGS = $(OPTIMISER) -std=c++14 -Wno-all -fmessage-length=0 -msse2 -mfpmath=sse
+CPPFLAGS = $(OPTIMISER) -std=c++14 -Wno-all -fmessage-length=0
 CFLAGS = $(OPTIMISER) -std=c99 -Wno-all
 #CFLAGS = -g 
 
@@ -9,7 +9,7 @@ CFLAGS = $(OPTIMISER) -std=c99 -Wno-all
 LIB = -lm 
 OBJS = utils.o hmm.o matrices.o
 HDR = utils.h hmm.h matrices.h
-INC = -I/usr/local/include
+INC = -I/usr/local/include -I$(PREFIX)/include
 # CPP code
 CPPOBJS = Cluster.o Tree.o Random.o Divvier.o bionj.o Sequence.o
 CPPHDR = Cluster.h Tree.h Random.h Divvier.h Sequence.h
