diff --git a/Makefile b/Makefile
index afa36f5..a916f14 100644
--- a/Makefile
+++ b/Makefile
@@ -1,5 +1,5 @@
-CC=gcc
-CPP=g++
+#CC=gcc
+#CPP=g++
 OPTIMISER = -O3
 CPPFLAGS = $(OPTIMISER) -std=c++14 -Wno-all -fmessage-length=0 -msse2 -mfpmath=sse
 CFLAGS = $(OPTIMISER) -std=c99 -Wno-all
@@ -31,10 +31,10 @@ fastZorro : $(CPPOBJS) $(OBJS) $(HDR)
 
 #Cluster
 Cluster.o : Cluster.cpp Cluster.h
-	$(CPP) $(CPPFLAGS) $(INC) -c Cluster.cpp
+	$(CXX) $(CPPFLAGS) $(INC) -c Cluster.cpp
 
 Tree.o : Tree.cpp Tree.h
-	$(CPP) $(CPPFLAGS) $(INC) -c Tree.cpp
+	$(CXX) $(CPPFLAGS) $(INC) -c Tree.cpp
 
 Random.o : Random.cpp Random.h
 	$(CPP) $(CPPFLAGS) $(INC) -c Random.cpp
