--- midas/utility.py.org	2019-07-11 17:02:23.000000000 +0200
+++ midas/utility.py	2019-07-11 17:03:36.000000000 +0200
@@ -84,10 +84,10 @@
 	main_dir = os.path.dirname(src_dir)
 	args['stream_bam'] = '/'.join([src_dir, 'run', 'stream_bam.py'])
 	args['stream_seqs'] = '/'.join([src_dir, 'run', 'stream_seqs.py'])
-	args['hs-blastn'] = '/'.join([main_dir, 'bin', platform.system(), 'hs-blastn'])
-	args['bowtie2-build'] = '/'.join([main_dir, 'bin', platform.system(), 'bowtie2-build'])
-	args['bowtie2'] = '/'.join([main_dir, 'bin', platform.system(), 'bowtie2'])
-	args['samtools'] = '/'.join([main_dir, 'bin', platform.system(), 'samtools'])
+	args['hs-blastn'] = 'hs-blastn'
+	args['bowtie2-build'] = 'bowtie2-build'
+	args['bowtie2'] = 'bowtie2'
+	args['samtools'] = 'samtools'
 	for arg in ['hs-blastn', 'stream_seqs', 'bowtie2-build', 'bowtie2', 'samtools', 'stream_bam']:
 		if not os.path.isfile(args[arg]):
 			sys.exit("\nError: File not found: %s\n" % args[arg])
--- setup.py	2017-08-06 08:49:07.000000000 +0200
+++ setup.py	2017-08-06 08:58:01.000000000 +0200
@@ -1,5 +1,5 @@
 try:
-	from setuptools import setup
+	from setuptools import setup, find_packages
 except:
        from distutils.core import setup
 import os
@@ -12,5 +12,7 @@
 	author = 'Stephen Nayfach',
 	author_email='snayfach@gmail.com',
 	url='https://github.com/snayfach/MIDAS',
-	install_requires = ['biopython >= 1.62', 'numpy >= 1.7.0', 'pysam >= 0.8.1', 'pandas >= 0.17.1']
+	install_requires = ['biopython >= 1.62', 'numpy >= 1.7.0', 'pysam >= 0.8.1', 'pandas >= 0.17.1'],
+	packages = find_packages(),
+	scripts = ['scripts/build_midas_db.py', 'scripts/call_consensus.py', 'scripts/compare_genes.py', 'scripts/merge_midas.py', 'scripts/query_by_compound.py', 'scripts/run_midas.py', 'scripts/snp_diversity.py', 'scripts/strain_tracking.py']
 )
