--- src/Makevars	2016-06-08 16:22:44.000000000 -0500
+++ src/Makevars.new	2017-01-31 12:52:54.000000000 -0600
@@ -29,7 +29,7 @@
 	rm -rf hdf5
 	mv hdf5source/hdf5 .
 	cd hdf5 && \
-	   ./configure CC="${CC}" CPP="${CPP}" CXX="${CXX}" CXXCPP="${CXXCPP}" F77="${F77}" --build="${R_ARCH#*/}" --with-pic 
+	   ./configure CC="${CC}" CPP="${CPP}" CXX="${CXX}" CXXCPP="${CXXCPP}" F77="${F77}" LDFLAGS="-Wl,-rpath ${PREFIX}/lib" --build="${R_ARCH#*/}" --with-pic 
 	cd hdf5/src && \
 	   $(MAKE)
 	cp hdf5/src/.libs/libhdf5ForBioC.a .
