--- a/install.sh
+++ b/install.sh
@@ -13,35 +13,3 @@
 echo $cmd
 $cmd
 
-cmd="wget --no-check-certificate https://obj.umiacs.umd.edu/metacompass/markers.tar.gz -P ./src/metaphyler"
-echo $cmd
-$cmd
-
-cmd="cd ./src/metaphyler/"
-echo $cmd
-$cmd
-
-cmd="tar -xzvf ./markers.tar.gz"
-echo $cmd
-$cmd
-
-cmd="rm markers.tar.gz"
-echo $cmd
-$cmd
-
-cmd="cd ../../"
-echo $cmd
-$cmd
-
-cmd="wget --no-check-certificate https://obj.umiacs.umd.edu/metacompass/refseq.tar.gz"
-echo $cmd
-$cmd
-
-cmd="tar -xzvf refseq.tar.gz"
-echo $cmd
-$cmd
-
-cmd="rm refseq.tar.gz"
-#echo $cmd
-$cmd
-
