diff --git a/src/lambda.hpp b/src/lambda.hpp
index 2dcefb6..c149692 100644
--- a/src/lambda.hpp
+++ b/src/lambda.hpp
@@ -1168,7 +1168,7 @@ computeBlastMatch(TBlastMatch         & bm,
         // we want to allow more gaps in longer query sequences
         switch (lH.options.band)
         {
-            case -3: maxDist = ceil(log2(qryLength)); break;
+            case -3: maxDist = ceil(std::log2(qryLength)); break;
             case -2: maxDist = floor(sqrt(qryLength)); break;
             case -1: break;
             default: maxDist = lH.options.band; break;
