--- scripts/DRAM-setup.py.old	2024-03-24 22:09:07.980198244 -0500
+++ scripts/DRAM-setup.py	2024-03-24 22:10:39.107097364 -0500
@@ -121,10 +121,10 @@
                                     help='vog annotations file') # add loc to vog_annotations to match the rest
 
     set_db_locs_parser.add_argument('--viral_loc', default=None,
+                                    help='mmseqs2 database file from ref seq viral gene collection')
     set_db_locs_parser.add_argument('--camper_tar_gz_loc', default=None,
                                     help='The source for the CAMPER database files, this is a tar.gz file downloaded'
                                     ' from the release page https://github.com/WrightonLabCSU/CAMPER/releases')
-                                    help='mmseqs2 database file from ref seq viral gene collection')
     set_db_locs_parser.add_argument('--peptidase_loc', default=None,
                                     help='mmseqs2 database file from MEROPS database')
     set_db_locs_parser.add_argument('--description_db_loc', default=None,
