--- scripts/DRAM-setup.py.old	2024-03-24 21:08:37.875167764 -0500
+++ scripts/DRAM-setup.py	2024-03-24 21:16:33.729641778 -0500
@@ -120,7 +120,7 @@
     set_db_locs_parser.add_argument('--vog_annotations_loc', default=None,
                                     help='vog annotations file') # add loc to vog_annotations to match the rest
 
-    set_db_locs_parser.add_argument('--viral_loc', default=None,
+    set_db_locs_parser.add_argument('--viral_loc', default=None)
     set_db_locs_parser.add_argument('--camper_tar_gz_loc', default=None,
                                     help="The source for the CAMPER database files, this is a tar.gz file downloaded"
                                     " from the release page https://github.com/WrightonLabCSU/CAMPER/releases")
