--- HapCUT2-1.1/Makefile	2019-04-22 22:00:08.000000000 +0200
+++ HapCUT2-1.1.patched/Makefile	2019-10-29 20:48:24.667768884 +0100
@@ -40,40 +40,40 @@
 
 #temporarily removed -O2 flag after -I$(HTSLIB)
 $(B)/extractHAIRS: $(B)/bamread.o $(B)/hashtable.o $(B)/readvariant.o $(B)/readfasta.o $(B)/hapfragments.o $(H)/extracthairs.c $(SAMTOOLS)/libbam.a $(HTSLIB)/libhts.a $(H)/parsebamread.c $(H)/realignbamread.c $(H)/nw.c $(H)/realign_pairHMM.c $(H)/estimate_hmm_params.c | $(B)
-	$(CC) -I$(SAMTOOLS) -I$(HTSLIB) -g $(B)/bamread.o $(B)/hapfragments.o $(B)/hashtable.o $(B)/readfasta.o $(B)/readvariant.o -o $(B)/extractHAIRS $(H)/extracthairs.c  -L$(SAMTOOLS) -L$(HTSLIB) -pthread -lhts -lbam -lm -lz
+	$(CC) $(LDFLAGS) $(CFLAGS) -I$(SAMTOOLS) -I$(HTSLIB) -g $(B)/bamread.o $(B)/hapfragments.o $(B)/hashtable.o $(B)/readfasta.o $(B)/readvariant.o -o $(B)/extractHAIRS $(H)/extracthairs.c  -L$(SAMTOOLS) -L$(HTSLIB) -pthread -lhts -lbam -lm -lz
 #temporarily removed -O2 flag after -I$(HTSLIB)
 
 $(B)/hapfragments.o: $(H)/hapfragments.c $(H)/hapfragments.h $(H)/readvariant.h | $(B)
-	$(CC) -c $(H)/hapfragments.c -o $(B)/hapfragments.o
+	$(CC) $(CFLAGS) -c $(H)/hapfragments.c -o $(B)/hapfragments.o
 
 $(B)/readvariant.o: $(H)/readvariant.c $(H)/readvariant.h $(H)/hashtable.h $(H)/hashtable.c | $(B)
-	$(CC) -c -I$(HTSLIB) $(H)/readvariant.c -o $(B)/readvariant.o
+	$(CC) $(CFLAGS) -c -I$(HTSLIB) $(H)/readvariant.c -o $(B)/readvariant.o
 
 $(B)/bamread.o: $(H)/bamread.h $(H)/bamread.c $(H)/readfasta.h $(H)/readfasta.c $(SAMTOOLS)/libbam.a $(HTSLIB)/libhts.a | $(B)
-	$(CC) -I$(SAMTOOLS) -I$(HTSLIB) -c $(H)/bamread.c -o $(B)/bamread.o
+	$(CC) $(CFLAGS) -I$(SAMTOOLS) -I$(HTSLIB) -c $(H)/bamread.c -o $(B)/bamread.o
 
 $(B)/hashtable.o: $(H)/hashtable.h $(H)/hashtable.c | $(B)
-	$(CC) -c $(H)/hashtable.c -o $(B)/hashtable.o
+	$(CC) $(CFLAGS) -c $(H)/hashtable.c -o $(B)/hashtable.o
 
 $(B)/readfasta.o: $(H)/readfasta.c $(H)/readfasta.h | $(B)
-	$(CC) -c $(H)/readfasta.c -o $(B)/readfasta.o
+	$(CC) $(CFLAGS) -c $(H)/readfasta.c -o $(B)/readfasta.o
 
 # BUILD HAPCUT2
 
 $(B)/HAPCUT2: $(B)/fragmatrix.o $(B)/readinputfiles.o $(B)/pointerheap.o $(B)/common.o $(X)/hapcut2.c $(X)/find_maxcut.c $(X)/post_processing.c| $(B)
-	$(CC) $(B)/common.o $(B)/fragmatrix.o $(B)/readinputfiles.o $(B)/pointerheap.o -o $(B)/HAPCUT2 -lm $(X)/hapcut2.c -L$(HTSLIB) -lhts 
+	$(CC) $(LDFLAGS) $(CFLAGS) $(B)/common.o $(B)/fragmatrix.o $(B)/readinputfiles.o $(B)/pointerheap.o -o $(B)/HAPCUT2 -lm $(X)/hapcut2.c -L$(HTSLIB) -lhts
 
 $(B)/common.o: $(X)/common.h $(X)/common.c | $(B)
-	$(CC) -c $(X)/common.c -o $(B)/common.o
+	$(CC) $(CFLAGS) -c $(X)/common.c -o $(B)/common.o
 
 $(B)/fragmatrix.o: $(X)/fragmatrix.h $(X)/fragmatrix.c $(X)/common.h $(X)/printhaplotypes.c  | $(B)
-	$(CC) -c $(X)/fragmatrix.c -o $(B)/fragmatrix.o
+	$(CC) $(CFLAGS) -c $(X)/fragmatrix.c -o $(B)/fragmatrix.o
 
 $(B)/readinputfiles.o: $(X)/readinputfiles.h $(X)/readinputfiles.c $(X)/common.h $(X)/fragmatrix.h | $(B)
-	$(CC) -c $(X)/readinputfiles.c -o $(B)/readinputfiles.o
+	$(CC) $(CFLAGS) -c $(X)/readinputfiles.c -o $(B)/readinputfiles.o
 
 $(B)/pointerheap.o: $(X)/pointerheap.h $(X)/pointerheap.c $(X)/common.h | $(B)
-	$(CC) -c $(X)/pointerheap.c -o $(B)/pointerheap.o
+	$(CC) $(CFLAGS) -c $(X)/pointerheap.c -o $(B)/pointerheap.o
 
 # create build directory
 $(B):
