diff -Naur src/dnacomp.c src/dnacomp.c
--- src/dnacomp.c	2014-09-17 05:38:54.000000000 +0200
+++ src/dnacomp.c	2018-09-24 10:39:26.415399493 +0200
@@ -31,7 +31,7 @@
    POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define maxtrees        100   /* maximum number of tied trees stored     */
+#define maxtrees        10000   /* maximum number of tied trees stored     */
 
 typedef boolean *boolptr;
 
diff -Naur src/dnapenny.c src/dnapenny.c
--- src/dnapenny.c	2018-09-24 10:40:55.398404128 +0200
+++ src/dnapenny.c	2018-09-24 10:41:20.861405454 +0200
@@ -30,7 +30,7 @@
    POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define maxtrees        1000  /* maximum number of trees to be printed out   */
+#define maxtrees        10000  /* maximum number of trees to be printed out   */
 #define often           1000  /* how often to notify how many trees examined */
 #define many            10000 /* how many multiples of howoften before stop  */
 
diff -Naur src/dollop.c src/dollop.c
--- src/dollop.c	2014-09-17 05:38:54.000000000 +0200
+++ src/dollop.c	2018-09-24 10:41:35.281406206 +0200
@@ -32,7 +32,7 @@
    POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define maxtrees        100  /* maximum number of tied trees stored     */
+#define maxtrees        10000  /* maximum number of tied trees stored     */
 
 #ifndef OLDC
 /* function prototypes */
diff -Naur src/dolpenny.c src/dolpenny.c
--- src/dolpenny.c	2014-09-17 05:38:54.000000000 +0200
+++ src/dolpenny.c	2018-09-24 10:41:58.230407401 +0200
@@ -31,7 +31,7 @@
    POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define maxtrees        1000  /* maximum number of trees to be printed out   */
+#define maxtrees        10000  /* maximum number of trees to be printed out   */
 #define often           100   /* how often to notify how many trees examined */
 #define many            1000  /* how many multiples of howoften before stop  */
 
diff -Naur src/mix.c src/mix.c
--- src/mix.c	2014-09-17 05:38:54.000000000 +0200
+++ src/mix.c	2018-09-24 10:42:07.580407888 +0200
@@ -33,7 +33,7 @@
 */
 
 
-#define maxtrees        100  /* maximum number of tied trees stored     */
+#define maxtrees        10000  /* maximum number of tied trees stored     */
 
 typedef long *placeptr;
 
diff -Naur src/penny.c src/penny.c
--- src/penny.c	2014-09-17 05:38:54.000000000 +0200
+++ src/penny.c	2018-09-24 10:42:17.858408423 +0200
@@ -32,7 +32,7 @@
    POSSIBILITY OF SUCH DAMAGE.
 */
 
-#define maxtrees        1000 /* maximum number of trees to be printed out   */
+#define maxtrees        10000 /* maximum number of trees to be printed out   */
 #define often           100  /* how often to notify how many trees examined */
 #define many            1000 /* how many multiples of howoften before stop  */
 
diff -Naur src/protpars.c src/protpars.c
--- src/protpars.c	2014-09-17 05:38:55.000000000 +0200
+++ src/protpars.c	2018-09-24 10:42:40.101409582 +0200
@@ -32,7 +32,7 @@
 */
 
 
-#define maxtrees        100   /* maximum number of tied trees stored */
+#define maxtrees        10000   /* maximum number of tied trees stored */
 
 typedef enum {
   universal, ciliate, mito, vertmito, flymito, yeastmito
