diff --git a/linearfold b/linearfold
index fca4a30..7e68544 100755
--- a/linearfold
+++ b/linearfold
@@ -38,7 +38,7 @@ def main():
         
     dangles = str(FLAGS.dangles)    
 
-    path = os.path.dirname(os.path.abspath(__file__))
+    path = os.path.dirname(os.path.realpath(__file__))
     cmd = ["%s/%s" % (path, ('bin/linearfold_v' if use_vienna else 'bin/linearfold_c')), beamsize, is_sharpturn, is_verbose, is_eval, is_constraints, zuker_subopt, delta, shape_file_path, is_fasta, dangles]
     subprocess.call(cmd, stdin=sys.stdin)
     
