From bb95a8f98b874fb9061c5ee1b3178f4d40f96416 Mon Sep 17 00:00:00 2001
From: Fabien Pertuy <Fabien.Pertuy@Sanofi.com>
Date: Fri, 21 May 2021 09:23:34 +0000
Subject: [PATCH] FIX command broken with symbolic link

---
 linearfold | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/linearfold b/linearfold
index f5541d2..2c86b63 100755
--- a/linearfold
+++ b/linearfold
@@ -28,7 +28,7 @@ def main():
     zuker_subopt = '1' if FLAGS.zuker else '0'
     delta = str(FLAGS.delta)
 
-    path = os.path.dirname(os.path.abspath(__file__))
+    path = os.path.dirname(os.path.realpath(__file__))
     cmd = ["%s/%s" % (path, ('bin/linearfold_v' if use_vienna else 'bin/linearfold_c')), beamsize, is_sharpturn, is_verbose, is_eval, is_constraints, zuker_subopt, delta]
     subprocess.call(cmd, stdin=sys.stdin)
     
-- 
2.30.2

