diff --git a/gflags.py b/gflags.py
index ac08b87..1e4d190 100644
--- a/gflags.py
+++ b/gflags.py
@@ -581,7 +581,7 @@ def DocToHelp(doc):
 	# 2) keep ws after new lines if not empty line
 	# 3) all other new lines shall be changed to a space
 	# Solution: Match new lines between non white space and replace with space.
-	doc = re.sub('(?<=\S)\n(?=\S)', ' ', doc, re.M)
+	doc = re.sub('(?<=\\S)\n(?=\\S)', ' ', doc, re.M)
 
 	return doc
 
@@ -2314,4 +2314,4 @@ _SPECIAL_FLAGS = FlagValues()
 ##    "comma-separated list of flag names that it is okay to specify "
 ##    "on the command line even if the program does not define a flag "
 ##    "with that name.  IMPORTANT: flags in this list that have "
-##    "arguments MUST use the --flag=value format.", _SPECIAL_FLAGS)gf
\ No newline at end of file
+##    "arguments MUST use the --flag=value format.", _SPECIAL_FLAGS)gf
