diff --git a/kent/src/hg/lib/straw/makefile b/kent/src/hg/lib/straw/makefile
index f07e6f6..d44ae8c 100644
--- a/kent/src/hg/lib/straw/makefile
+++ b/kent/src/hg/lib/straw/makefile
@@ -1,10 +1,10 @@
-KENT_INC=-I../../../inc
+KENT_INC=-I../../../inc -I${PREFIX}/include
 
 straw: straw.o cStraw.o
 	ld -r -o ../straw.o straw.o cStraw.o
 straw.o: straw.cpp straw.h
-	${CXX} straw.cpp ${KENT_INC} -g -c -std=c++11 -o straw.o
+	${CXX} straw.cpp ${KENT_INC} -g -c -std=c++14 -o straw.o
 cStraw.o: cStraw.cpp straw.h
-	${CXX} cStraw.cpp ${KENT_INC} -g -c -std=c++11 -o cStraw.o
+	${CXX} cStraw.cpp ${KENT_INC} -g -c -std=c++14 -o cStraw.o
 clean:
 	rm -f straw.o cStraw.o ../straw.o
diff --git a/kent/src/utils/expMatrixToBarchartBed/expMatrixToBarchartBed b/kent/src/utils/expMatrixToBarchartBed/expMatrixToBarchartBed
index 9b756bf..697ce4d 100755
--- a/kent/src/utils/expMatrixToBarchartBed/expMatrixToBarchartBed
+++ b/kent/src/utils/expMatrixToBarchartBed/expMatrixToBarchartBed
@@ -1,4 +1,4 @@
-#!/usr/bin/env python2.7
+#!/usr/bin/env python
 # expMatrixToBarchartBed
 """
 Generate a barChart bed6+5 file from a matrix, meta data, and coordinates.