--- btrim.cpp
+++ btrim.cpp
@@ -471 +471 @@
-			beginVector.push_back({wordSeq,numberRead});
+			beginVector.push_back(make_pair(wordSeq,numberRead));
@@ -483 +483 @@
-			endVector.push_back({wordSeq,numberRead});
+			endVector.push_back(make_pair(wordSeq,numberRead));
@@ -508 +508 @@
-							coverageComparison.push_back({coverages[beginVector[indiceBegin].second], abs(beginVector[indiceBegin].second)});
+							coverageComparison.push_back(make_pair(coverages[beginVector[indiceBegin].second], beginVector[indiceBegin].second));
@@ -550 +550 @@
-							coverageComparison.push_back({coverages[endVector[indiceEnd].second],abs(endVector[indiceEnd].second)});
+							coverageComparison.push_back(make_pair(coverages[endVector[indiceEnd].second],endVector[indiceEnd].second));
@@ -663 +663 @@
-				beginVector.push_back({wordSeq,numberRead});
+				beginVector.push_back(make_pair(wordSeq,numberRead));
@@ -677 +677 @@
-				endVector.push_back({wordSeq,numberRead});
+				endVector.push_back(make_pair(wordSeq,numberRead));
--- btrimBubble.cpp
+++ btrimBubble.cpp
@@ -396 +396 @@
-			beginVector.push_back({wordSeq,numberRead});
+			beginVector.push_back(make_pair(wordSeq,numberRead));
@@ -408 +408 @@
-			endVector.push_back({wordSeq,numberRead});
+			endVector.push_back(make_pair(wordSeq,numberRead));
@@ -515 +515 @@
-				beginVector.push_back({wordSeq,numberRead});
+				beginVector.push_back(make_pair(wordSeq,numberRead));
@@ -529 +529 @@
-				endVector.push_back({wordSeq,numberRead});
+				endVector.push_back(make_pair(wordSeq,numberRead));
