From 6b1ed06f05676c2de59571e9deb63eee760dab14 Mon Sep 17 00:00:00 2001
From: Jure Zmrzlikar <zmrzlikar.jure@gmail.com>
Date: Mon, 27 Nov 2017 15:54:03 +0100
Subject: [PATCH 01/14] setup: Change `install_requires` to list

---
 setup.py | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/setup.py b/setup.py
index 067dd62..770df06 100755
--- a/setup.py
+++ b/setup.py
@@ -56,7 +56,7 @@ setup(
             'examples/*.sh',
         ]
     },
-    install_requires={
+    install_requires=[
         'numpy',
         'pandas',
         'cutadapt>=1.10',
@@ -65,7 +65,7 @@ setup(
         'numpydoc',
         'sphinx>=1.4',
         'matplotlib',
-    },
+    ],
     extras_require={
         'docs': [
             'docutils',
-- 
2.7.4

