diff --git a/makefile b/makefile
index 5b919e2..e42f9ec 100644
--- a/makefile
+++ b/makefile
@@ -2,11 +2,11 @@ PROGRAM = tedna
 C_FILES := $(wildcard *.cpp)
 OBJS := $(patsubst %.cpp, %.o, $(C_FILES))
 #CC = clang++
-CC = g++
+CC = $(CXX)
 # CFLAGS = -Wall -fstack-protector-all -Wstack-protector -fno-omit-frame-pointer -pedantic -std=c++11 -pthread -fsanitize=thread
-CFLAGS = -Wall -fstack-protector-all -Wstack-protector -fno-omit-frame-pointer -pedantic -std=c++11 -pthread
+CFLAGS = -Wall -fstack-protector-all -Wstack-protector -fno-omit-frame-pointer -pedantic -march=x86-64-v3 -std=c++14 -pthread -I$(PREFIX)/include
 # CFLAGS = -Wall -pedantic -std=c++11 -pthread
-LDFLAGS =
+LDFLAGS = -L$(PREFIX)/lib
 # LDFLAGS = -static -static-libgcc -static-libstdc++
 
 k ?= 61
