--- configure.ac	2016-12-14 10:09:38.761568504 +0100
+++ configure.ac	2016-12-14 10:09:47.509757018 +0100
@@ -6,10 +6,6 @@
 
 AC_CANONICAL_HOST
 
-PKG_CHECK_MODULES([CHECK],[check >= 0.8.2],[have_check="yes"],
-  AC_MSG_WARN(["'Check' unit testing framework not found. It would be impossible to run unit tests!"])
-  [have_check="no"])
-
 AC_CHECK_HEADERS([zlib.h math.h getopt.h])
 AC_CHECK_LIB(m, floor)
 AX_CHECK_ZLIB
