diff --git a/ipk.py b/ipk.py
index fb884f3..71bf356 100755
--- a/ipk.py
+++ b/ipk.py
@@ -268,12 +268,12 @@ def build_database(ar,
         if keep_positions:
             raise RuntimeError("--keep-positions is not supported for DNA.")
         else:
-            bin = f"{current_dir}/bin/ipk/ipk-dna"
+            bin = f"{current_dir}/ipk-dna"
     else:
         if keep_positions:
-            bin = f"{current_dir}/bin/ipk/ipk-aa-pos"
+            bin = f"{current_dir}/ipk-aa-pos"
         else:
-            bin = f"{current_dir}/bin/ipk/ipk-aa"
+            bin = f"{current_dir}/ipk-aa"
 
 
     command = [
