--- src/BuildGraph/Release/makefile
+++ src/BuildGraph/Release/makefile
@@ -47 +47 @@
-	g++ -fopenmp -o "buildG" $(OBJS) $(USER_OBJS) $(LIBS) $(OPTS) $(GZLIB) 
+	$(CXX) -fopenmp -o "buildG" $(OBJS) $(USER_OBJS) $(LIBS) $(OPTS) $(GZLIB) 
--- src/BuildGraph/Release/src/subdir.mk
+++ src/BuildGraph/Release/src/subdir.mk
@@ -35 +35 @@
-	g++ $(OPTS) -Wno-sign-compare -Wmaybe-uninitialized -g3 -c -fmessage-length=0 -fopenmp $(GZLIB) -std=c++11 -O3 -lgomp -MMD -MP -MF"$(@:%.o=%.d)" -MT"$(@)" -o "$@" "$<"
+	$(CXX) $(OPTS) -Wno-sign-compare -Wmaybe-uninitialized -g3 -c -fmessage-length=0 -fopenmp $(GZLIB) -std=c++11 -O3 -lgomp -MMD -MP -MF"$(@:%.o=%.d)" -MT"$(@)" -o "$@" "$<"
--- src/SimplifyGraph/Release/makefile
+++ src/SimplifyGraph/Release/makefile
@@ -52 +52 @@
-	g++ -fopenmp -o "fullsimplify" $(OBJS) $(USER_OBJS) $(LIBS) $(OPTS) $(GZLIB)
+	$(CXX) -fopenmp -o "fullsimplify" $(OBJS) $(USER_OBJS) $(LIBS) $(OPTS) $(GZLIB)
@@ -60 +60 @@
-	g++ -fopenmp -o "parsimplify" $(OBJS2) $(USER_OBJS) $(LIBS) $(OPTS) $(GZLIB)
+	$(CXX) -fopenmp -o "parsimplify" $(OBJS2) $(USER_OBJS) $(LIBS) $(OPTS) $(GZLIB)
--- src/SimplifyGraph/Release/src/subdir1.mk
+++ src/SimplifyGraph/Release/src/subdir1.mk
@@ -45 +45 @@
-	g++ $(OPTS) -O3 -c -fmessage-length=0 -fopenmp -Wmaybe-uninitialized -Wno-sign-compare $(GZLIB) -lgomp -std=c++11 -MMD -MP -MF"$(@:%.o=%.d)" -MT"$(@)" -o "$@" "$<"
+	$(CXX) $(OPTS) -O3 -c -fmessage-length=0 -fopenmp -Wmaybe-uninitialized -Wno-sign-compare $(GZLIB) -lgomp -std=c++11 -MMD -MP -MF"$(@:%.o=%.d)" -MT"$(@)" -o "$@" "$<"
--- src/SimplifyGraph/Release/src/subdir2.mk
+++ src/SimplifyGraph/Release/src/subdir2.mk
@@ -45 +45 @@
-	g++ $(OPTS) -O3 -g3 -c -fmessage-length=0 -fopenmp -Wno-sign-compare $(GZLIB) -lgomp -std=c++11 -MMD -MP -MF"$(@:%.o=%.d)" -MT"$(@)" -o "$@" "$<"
+	$(CXX) $(OPTS) -O3 -g3 -c -fmessage-length=0 -fopenmp -Wno-sign-compare $(GZLIB) -lgomp -std=c++11 -MMD -MP -MF"$(@:%.o=%.d)" -MT"$(@)" -o "$@" "$<"
