--- Makefile.bak	2017-07-29 19:25:54.753863472 +0000
+++ Makefile	2017-07-31 06:01:38.756135444 +0000
@@ -1,6 +1,4 @@
 CC=	gcc
-LIBS= -pthread -lm lib/zlib/libz.a -lm lib/jemalloc/lib/libjemalloc.a -lm lib/hdf5/lib/libhdf5-static.a  -lm lib/hdf5/lib/libhdf5_hl-static.a -ldl -lm lib/libdivsufsort/lib/libdivsufsort64.a
-CFLAGS=	-fgnu89-inline -O2 -D USE_JEMALLOC -D DEBUG
 Hera:	
 	mkdir -p build/
 	$(CC) $(CFLAGS) src/xxhash.c src/bgzf.c src/hash_align.c src/EM.c src/bam_write.c src/fusion.c src/fmindex.c src/genome_map.c src/assembly.c src/main.c $(LIBS) -o build/hera
