--- src/EM.h.bak	2017-07-29 19:25:06.996391671 +0000
+++ src/EM.h	2017-07-29 19:30:16.116392160 +0000
@@ -1,8 +1,8 @@
 #ifndef EM_H_
 #define EM_H_
 
-#include "../lib/hdf5/include/hdf5.h"
-#include "../lib/hdf5/include/hdf5_hl.h"
+#include <hdf5.h>
+#include <hdf5_hl.h>
 
 #include "hash_align.h"
 
--- src/fmindex.h.bak	2017-07-29 19:25:23.328428148 +0000
+++ src/fmindex.h	2017-07-29 19:30:47.831508865 +0000
@@ -1,7 +1,7 @@
 #ifndef FMINDEX_H_
 #define FMINDEX_H_
 
-#include <divsufsort64.h>//"../lib/libdivsufsort/include/divsufsort64.h"
+#include <divsufsort64.h>
 #include "hash_align.h"
 
 typedef struct {
--- src/hash_align.h.bak	2017-07-29 19:25:39.101079888 +0000
+++ src/hash_align.h	2017-07-30 20:18:24.985542981 +0000
@@ -2,11 +2,7 @@
 #define HASH_ALIGN_H_
 
 #if defined(USE_JEMALLOC)
-#include "../lib/jemalloc/include/jemalloc/jemalloc.h"
-#define malloc(size) je_malloc(size)
-#define calloc(count,size) je_calloc(count,size)
-#define realloc(ptr,size) je_realloc(ptr,size)
-#define free(ptr) je_free(ptr)
+#include <jemalloc/jemalloc.h>
 #endif
 
 #if defined(DEBUG)
@@ -26,8 +22,8 @@
 #include <pthread.h>
 #include <math.h>
 
-#include "../lib/zlib/zlib.h"
-#include "../lib/libdivsufsort/include/divsufsort64.h"
+#include <zlib.h>
+#include <divsufsort64.h>
 #include "bgzf.h"
 #include "kseq.h"
 #include "xxhash.h"
