diff --git a/sonicparanoid/sys_tools.py b/sonicparanoid/sys_tools.py
index 5fff04f..4cde5cb 100644
--- a/sonicparanoid/sys_tools.py
+++ b/sonicparanoid/sys_tools.py
@@ -5,6 +5,7 @@ import subprocess
 import logging
 from shutil import copyfileobj
 from filetype import guess_mime
+from typing import List, Tuple
 
 __module_name__ = "System Tools"
 __source__ = "sys_tools.py"
@@ -332,7 +333,7 @@ def getShell():
 
 
 
-def is_conda_env() -> tuple[bool, str]:
+def is_conda_env() -> Tuple[bool, str]:
     """
     Check if a CONDA environment is being used.
     """
@@ -351,14 +352,14 @@ def is_conda_env() -> tuple[bool, str]:
 
 
 
-def is_mamba_env() -> tuple[bool, str]:
+def is_mamba_env() -> Tuple[bool, str]:
     """
     Check if a Mamba environment is being used.
     """
     # The identification is done based on the path of the python binaries
     # Possible paths are chosen based on conda documentation
     # https://docs.anaconda.com/anaconda/user-guide/tasks/integration/python-path
-    mambaTypes: list[str] = ["mamba", "micromamba"]
+    mambaTypes: List[str] = ["mamba", "micromamba"]
     # Extract the path with the binaries
     pyBinPrefix: str = sys.exec_prefix
     # Check if the prefix contains one of the anaconda path keywords
