diff --git i/setup.py w/setup.py
index e99c995..fc79fa0 100755
--- i/setup.py
+++ w/setup.py
@@ -59,12 +59,6 @@ if _CYTHON_INSTALLED:
 
 '''
 
-# Force to use the default version of gcc and g++
-if platform.system() == "Linux":
-    os.environ["CC"] = "gcc"
-    os.environ["CXX"] = "g++"
-
-
 def makedir(path):
     """Create a directory including the intermediate directories in the path if not existing."""
     try:
