diff --git a/setup.py b/setup.py
index 21f5a4e..bca0d6b 100755
--- a/setup.py
+++ b/setup.py
@@ -59,12 +59,6 @@ if _CYTHON_INSTALLED:
 
 '''
 
-# Force to use the default version of gcc and g++
-if platform.system() == "Linux":
-    os.environ["CC"] = "gcc"
-    os.environ["CXX"] = "g++"
-
-
 def makedir(path):
     """Create a directory including the intermediate directories in the path if not existing."""
     try:
