--- a/setup.py
+++ b/setup.py
@@ -10,13 +10,6 @@ from setuptools import setup, Extension, Command
 from setuptools.command.build_ext import build_ext
 from setuptools.command.sdist import sdist
 
-if sys.version_info.major == 2 and sys.version_info.minor != 7:
-    sys.stderr.write(
-        "ERROR: cyvcf2 is only for python 2.7 or greater you are running %d.%d\n",
-        (sys.version_info.major, sys.version_info.minor),
-    )
-    sys.exit(1)
-
 import numpy as np
 
 
@@ -236,6 +229,7 @@ htslib_include_dirs = ["htslib", "htslib/htslib"]
 
 # Build the Cython extension by statically linking to the bundled htslib
 sources = ["cyvcf2/cyvcf2.pyx", "cyvcf2/helpers.c"]
+extra_libs = ['hts']
 
 extensions = [
     Extension(
