diff -Naur probconsRNA/Main.cc probconsRNA_new/Main.cc
--- probconsRNA/Main.cc	2005-03-21 20:51:24.000000000 +0000
+++ probconsRNA_new/Main.cc	2021-01-12 17:10:48.374985934 +0000
@@ -21,6 +21,8 @@
 #include <cstdio>
 #include <cstdlib>
 #include <cerrno>
+#include <cstring>
+#include <climits>
 #include <iomanip>
 
 string parametersInputFilename = "";
diff -Naur probconsRNA/Makefile probconsRNA_new/Makefile
--- probconsRNA/Makefile	2005-06-07 19:43:53.000000000 +0100
+++ probconsRNA_new/Makefile	2021-01-12 17:23:11.023015823 +0000
@@ -6,7 +6,7 @@
 # 1) Choose C++ compiler.
 ################################################################################
 
-CXX = g++
+#CXX = g++
 
 ################################################################################
 # 2) Set C++ flags.
@@ -25,7 +25,7 @@
 
 # release mode
 #CXXFLAGS = -O3 -W -Wall -pedantic -DNDEBUG $(OTHERFLAGS) -mmmx -msse -msse2 -mfpmath=sse -march=pentium4 -mcpu=pentium4 -funroll-loops -fomit-frame-pointer 
-CXXFLAGS = -O3 -W -Wall -pedantic -DNDEBUG $(OTHERFLAGS) -funroll-loops
+CXXFLAGS += -O3 -W -Wall -pedantic -DNDEBUG $(OTHERFLAGS) -funroll-loops
 
 ################################################################################
 # 3) Dependencies
