--- CMakeLists.txt	
+++ CMakeLists.txt	
@@ -27,7 +27,7 @@
     RETURN_VALUE ERROR_GIT)
 
 if(NOT ${ERROR_GIT})
-    add_definitions( -DGIT_SHA1="${VERSION_SHA1}" )
+    add_definitions( -DGIT_SHA1="c8ac60252fa0b2abf511f7363cff7c4342dac2ee" )
 else()
     message("Warning: cannot retrieve git version. Bcalm won't display its version. Error value: ${ERROR_GIT})")
 endif(NOT ${ERROR_GIT})
@@ -94,7 +94,7 @@
 SET (CPACK_PACKAGE_VERSION              "")  
 
 # We have to tell what is the server name
-SET (CPACK_GFORGE_PROJECT_NAME          "gatb-tools")
+#SET (CPACK_GFORGE_PROJECT_NAME          "gatb-tools")
 
 # We set the kind of archive
 SET (CPACK_GENERATOR                    "TGZ")
@@ -114,8 +114,8 @@
 
 # We copy the project binary to the 'bin' directory
 INSTALL (TARGETS ${PROJECT_NAME} DESTINATION bin)
-INSTALL (DIRECTORY "${PROJECT_SOURCE_DIR}/example/" DESTINATION example)
-INSTALL (FILES VERSION LICENSE README.md DESTINATION bin/..)
+#INSTALL (DIRECTORY "${PROJECT_SOURCE_DIR}/example/" DESTINATION example)
+#INSTALL (FILES VERSION LICENSE README.md DESTINATION bin/..)
 
 # cmake_system_name for mac is Darwin, i want to change that
 if (${CMAKE_SYSTEM_NAME} MATCHES "Linux")
