--- bcalm-2.2.0/CMakeLists.txt	2017-12-20 15:51:42.000000000 +0100
+++ CMakeLists.txt	2018-10-13 19:05:16.067524264 +0200
@@ -25,7 +25,7 @@
     ARGS "rev-parse --short HEAD"
     OUTPUT_VARIABLE VERSION_SHA1 )
 
-add_definitions( -DGIT_SHA1="${VERSION_SHA1}" )
+add_definitions( -DGIT_SHA1="c8ac60252fa0b2abf511f7363cff7c4342dac2ee" )
 
 
 ################################################################################
@@ -84,7 +84,7 @@
 SET (CPACK_PACKAGE_VERSION              "")  
 
 # We have to tell what is the server name
-SET (CPACK_GFORGE_PROJECT_NAME          "gatb-tools")
+#SET (CPACK_GFORGE_PROJECT_NAME          "gatb-tools")
 
 # We set the kind of archive
 SET (CPACK_GENERATOR                    "TGZ")
@@ -104,8 +104,8 @@
 
 # We copy the project binary to the 'bin' directory
 INSTALL (TARGETS ${PROJECT_NAME} DESTINATION bin)
-INSTALL (DIRECTORY "${PROJECT_SOURCE_DIR}/example/" DESTINATION example)
-INSTALL (FILES LICENSE README.md DESTINATION bin/..)
+#INSTALL (DIRECTORY "${PROJECT_SOURCE_DIR}/example/" DESTINATION example)
+#INSTALL (FILES LICENSE README.md DESTINATION bin/..)
 
 # cmake_system_name for mac is Darwin, i want to change that
 if (${CMAKE_SYSTEM_NAME} MATCHES "Linux")
