new file mode 100644
--- /dev/null
+++ pyproject.toml
@@ -0,0 +1,34 @@
+[build-system]
+requires = ["setuptools>=61.0"]
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "bamsurgeon"
+description = "Tools for adding mutations to existing bam files, used for testing mutation callers"
+readme = "README.md"
+requires-python = ">=3.6"
+license = {file = "LICENSE.txt"}
+classifiers = [
+    "License :: OSI Approved :: MIT License",
+    "Programming Language :: Python :: 3",
+    "Topic :: Scientific/Engineering",
+    "Topic :: Scientific/Engineering :: Bio-Informatics",
+    "Intended Audience :: Science/Research",
+    "Natural Language :: English",
+]
+dependencies = []
+version = "1.4.1"
+
+[project.urls]
+"Homepage" = "https://github.com/adamewing/bamsurgeon"
+
+[tool.setuptools]
+script-files = [
+    "bin/addindel.py",
+    "bin/addsnv.py",
+    "bin/addsv.py"
+]
+
+[tool.setuptools.packages.find]
+where = ["bin"]
+
