diff --git a/gatb-core/thirdparty/BooPHF/BooPHF.h b/gatb-core/thirdparty/BooPHF/BooPHF.h
index 5b445d5..da1cb95 100644
--- a/gatb-core/thirdparty/BooPHF/BooPHF.h
+++ b/gatb-core/thirdparty/BooPHF/BooPHF.h
@@ -837,7 +837,7 @@ we need this 2-functors scheme because HashFunctors won't work with unordered_ma
 
 			uint64_t totalsize =  totalsizeBitset +  _final_hash.size()*42*8 ;  // unordered map takes approx 42B per elem [personal test] (42B with uint64_t key, would be larger for other type of elem)
 
-			printf("Bitarray    %12" PRIu64 "  bits (%.2f %%)   (array + ranks )\n",
+			printf("Bitarray    %" PRIu64 "  bits (%.2f %%)   (array + ranks )\n",
 				   totalsizeBitset, 100*(float)totalsizeBitset/totalsize);
 			printf("final hash  %12lu  bits (%.2f %%) (nb in final hash %lu)\n",
 				   _final_hash.size()*42*8, 100*(float)(_final_hash.size()*42*8)/totalsize,
