--- JAFFA_stages.groovy	2017-03-14 10:41:40.760151871 +0100
+++ JAFFA_stages.groovy	2017-03-14 10:41:52.320108428 +0100
@@ -14,14 +14,14 @@
 
 /******** Path to reference files ********/
 // leave if references are in the Jaffa folder
-refBase = codeBase
+// refBase = codeBase
 
 // Path to reference files that are elsewhere on the file system
 // refBase = "/path/to/reference/directory"
 
 // Path to reference files 
-// env = System.getenv()
-// refBase = env['GENOMES']
+env = System.getenv()
+refBase = env['JAFFA_REF_BASE']
 
 // Should there be a folder structure for the reference files within the reference folder then
 // the below variables will allow for this flexibility

--- process_transcriptome_blat_table.R.old	2017-03-14 11:00:22.067990258 +0100
+++ process_transcriptome_blat_table.R	2017-03-14 11:00:46.067901275 +0100
@@ -43,7 +43,7 @@
 
    #now get just the non-redundant set of transcripts 
    ranges=IRanges(starts,ends)
-   overs=findOverlaps(ranges,type="within",ignoreSelf=TRUE,ignoreRedundant=TRUE,select="arbitrary")
+   overs=findOverlaps(ranges,type="within",drop.self=TRUE,drop.redundant=TRUE,select="arbitrary")
    reduced=is.na(overs)
    regions=ranges[reduced]
 

