--- configure.ac	2017-04-18 16:38:19.000000000 +0300
+++ configure.ac	2017-04-18 16:38:30.000000000 +0300
@@ -24,9 +24,6 @@
 # Checks for pthread
 AX_PTHREAD
 
-# Ensure zlib is available
-PKG_CHECK_MODULES([zlib], [zlib])
-AC_CHECK_HEADERS([zlib.h math.h])
 
 # Check for Python
 AM_PATH_PYTHON([3.0],

--- python/Makefile.am	2017-09-22 12:57:33.000000000 +0200
+++ python/Makefile.am	2017-10-27 22:46:48.000000000 +0200
@@ -1,11 +1,9 @@
 EXTRA_DIST=gubbins/* setup.py scripts/*

 all-local:
-	${PYTHON} setup.py build


 install-exec-local:
-	${PYTHON} setup.py install

 uninstall-local:
 	rm -rf $(pythondir)/*gubbins*
