*** MoreUtils.xs	2005-11-11 03:23:13.000000000 -0500
--- MoreUtils1.xs	2018-02-19 22:31:44.000000000 -0500
***************
*** 1135,1141 ****
  	HV *hv = newHV();

  	/* don't build return list in scalar context */
! 	if (GIMME == G_SCALAR) {
  	    for (i = 0; i < items; i++) {
  		if (!hv_exists_ent(hv, ST(i), 0)) {
  		    count++;
--- 1135,1141 ----
  	HV *hv = newHV();

  	/* don't build return list in scalar context */
! 	if (GIMME_V == G_SCALAR) {
  	    for (i = 0; i < items; i++) {
  		if (!hv_exists_ent(hv, ST(i), 0)) {
  		    count++;
***************
*** 1269,1272 ****
  	    CvXSUBANY(code).any_ptr = NULL;
  	}
      }
-
--- 1269,1271 ----
