--- make/makedis.csh
+++ make/makedis.csh
@@ -195,0 +196,8 @@
+	case "x86_64":
+		set platform=linux64
+		if ("$?LINUX_MODE" == 1) then
+			if ("$LINUX_MODE" == "icc") then
+				set platform=${platform}_icc9
+			endif
+		endif
+		breaksw
@@ -210 +219 @@
-	set HAVE_MOTIF=0
+	set HAVE_MOTIF=1
--- platform/linux64.ncbi.mk
+++ platform/linux64.ncbi.mk
@@ -7 +7 @@
-NCBI_CC = gcc -pipe -D_GNU_SOURCE
+NCBI_CC = $(BIOCONDA_CC) -pipe -D_GNU_SOURCE
@@ -17 +17 @@
-NCBI_VIBFLAG = -I/usr/X11R6/include -L/usr/X11R6/lib64 -L/usr/X11R6/lib -DWIN_MOTIF
+NCBI_VIBFLAG = -DWIN_MOTIF
@@ -21 +21 @@
-NCBI_DISTVIBLIBS = -L/usr/X11R6/lib64 -L/usr/X11R6/lib -lXmu -Wl,-Bstatic -lXm -Wl,-Bdynamic -lXft -lfontconfig -ljpeg -lpng -lXt -lSM -lICE -lXext -lXp -lX11 -ldl
+NCBI_DISTVIBLIBS = -lXmu -lXm -lXft -lfontconfig -ljpeg -lpng -lXt -lSM -lICE -lXext -lXp -lX11 -ldl
@@ -35,3 +35,3 @@
-OPENGL_INCLUDE = -I/usr/X11R6/include
-OPENGL_LIBS = -L/usr/X11R6/lib64 -L/usr/X11R6/lib -lGL -lGLU
-NCBI_OGLLIBS = -L/usr/X11R6/lib64 -L/usr/X11R6/lib -lGL -lGLU
+OPENGL_INCLUDE =
+OPENGL_LIBS = -lGL -lGLU
+NCBI_OGLLIBS = -lGL -lGLU
--- platform/darwin.ncbi.mk
+++ platform/darwin.ncbi.mk
@@ -11 +11 @@
-NCBI_CC = cc -pipe
+NCBI_CC = $(BIOCONDA_CC) -pipe
