diff -ruN /tmp/ncbi/make/ln-if-absent ncbi/make/ln-if-absent
--- /tmp/ncbi/make/ln-if-absent	2002-06-17 16:27:20.000000000 +0200
+++ ncbi/make/ln-if-absent	2019-03-10 11:58:32.475870278 +0100
@@ -1,4 +1,4 @@
-#!/bin/csh -f
+#!/usr/bin/env tcsh
 # NOTE:  use "/bin/tcsh" above if "/bin/csh" is absent (e.g. on QNX OS)
 #        use a "ln-if-absent.beos" on the BeOS platform (rename it to ln-if-absent)
 #
diff -ruN /tmp/ncbi/make/makedis.csh ncbi/make/makedis.csh
--- /tmp/ncbi/make/makedis.csh	2016-09-02 22:00:27.000000000 +0200
+++ ncbi/make/makedis.csh	2019-03-10 11:59:18.167893129 +0100
@@ -1,4 +1,4 @@
-#!/bin/csh -f
+#!/usr/bin/env tcsh
 #
 # $Id: makedis.csh 512803 2016-09-02 19:57:21Z ucko $
 #
