From 6f79676d9a01d31587acf979a34d58e3710df006 Mon Sep 17 00:00:00 2001
From: Ed Harry <e.f.harry@gmail.com>
Date: Tue, 26 Nov 2019 13:43:40 +0000
Subject: [PATCH] disable auto-vectorisation for conda libm

---
 install | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/install b/install
index b4d39a2..6428191 100755
--- a/install
+++ b/install
@@ -165,7 +165,7 @@ for i in range(nExts):
     outputBinary = "PretextSnapshot.{}".format(buildExts[i])
     flags = args.flags + " {}".format(exts[buildExts[i]])
 
-    command = args.compiler + " PretextSnapshot.cpp " + libdeflate + " " + libmpc + " -Iinclude -lstdc++ -lm " + flags + " -o " + outputBinary
+    command = args.compiler + " PretextSnapshot.cpp " + libdeflate + " " + libmpc + " -Iinclude -fno-builtin-exp -fno-builtin-log -fno-builtin-pow -lstdc++ -lm " + flags + " -o " + outputBinary
     p = Popen(command.strip().split(), stdin=PIPE, stdout=PIPE, stderr=PIPE)
     output, err = p.communicate()
     if p.returncode != 0:
-- 
2.24.0

