diff --git a/meson.build b/meson.build
index f848ab2..55c6940 100644
--- a/meson.build
+++ b/meson.build
@@ -17,9 +17,9 @@ foreach ext, flag : extensions
         [   static_library('deflate' + ext, 'libdeflate/lib/deflate_compress.c', 'libdeflate/lib/deflate_decompress.c', 'libdeflate/lib/utils.c', c_args : flags, link_with : 
             [   static_library('deflate_x86' + ext, 'libdeflate/lib/x86/cpu_features.c', c_args : flags),
                 static_library('deflate_arm' + ext, 'libdeflate/lib/arm/cpu_features.c', c_args : flags),
-                static_library('deflate_zlib' + ext, 'libdeflate/lib/adler32.c', 'libdeflate/lib/zlib_decompress.c', 'libdeflate/lib/zlib_compress.c', c_args : flags)]),
+                static_library('deflate_zlib' + ext, 'libdeflate/lib/adler32.c', 'libdeflate/lib/zlib_decompress.c', 'libdeflate/lib/zlib_compress.c', c_args : flags, include_directories : 'libdeflate/')], include_directories : 'libdeflate/'),
             static_library('mpc' + ext, 'mpc/mpc.c', c_args : flags)],
-        include_directories : include_directories('include'))
+        include_directories : [include_directories('include'), include_directories('libdeflate')])
 endforeach
 
 e = executable('PretextSnapshot', 'wrapper/SIMDExtensionWrapper.cpp', install : true, cpp_args : base_flags + ['-DProgramName=PretextSnapshot'])
