--- admin/Build_Trinotate_Boilerplate_SQLite_db.pl	2020-04-08 20:44:17.000000000 +0200
+++ admin/Build_Trinotate_Boilerplate_SQLite_db.pl	2020-07-28 20:31:06.795248438 +0200
@@ -14,7 +14,7 @@
 my $no_cleanup_flag = $ARGV[1] || 0;
 
 
-my $UTILDIR = "$FindBin::RealBin/util";
+my $UTILDIR = "$FindBin::RealBin";
 
 ## Resources:
 my $SPROT_DAT_URL = "ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.dat.gz";
--- util/report_summary/trinotate_report_summary.pl	2020-04-08 20:44:17.000000000 +0200
+++ util/report_summary/trinotate_report_summary.pl	2020-07-28 20:31:31.003092687 +0200
@@ -17,7 +17,7 @@
 my $TOP_TAX_LEVEL = 6;
 
 
-my $UTILDIR = "$FindBin::Bin/util";
+my $UTILDIR = "$FindBin::Bin";
 
 main: {
 
