--- admin/Build_Trinotate_Boilerplate_SQLite_db.pl	2018-08-02 15:49:56.299807109 +0200
+++ admin/Build_Trinotate_Boilerplate_SQLite_db.pl	2018-08-02 15:50:05.030720666 +0200
@@ -14,13 +14,13 @@
 my $no_cleanup_flag = $ARGV[1] || 0;
 
 
-my $UTILDIR = "$FindBin::RealBin/util";
+my $UTILDIR = "$FindBin::RealBin";
 
 ## Resources:
-my $SPROT_DAT_URL = "ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.dat.gz";
+my $SPROT_DAT_URL = "http://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.dat.gz";
 my $EGGNOG_DAT_URL = "http://eggnogdb.embl.de/download/latest/data/NOG/NOG.annotations.tsv.gz";
 my $GENE_ONTOLOGY_DAT_URL = "http://purl.obolibrary.org/obo/go/go-basic.obo";
-my $PFAM_DAT_URL = "ftp://ftp.ebi.ac.uk/pub/databases/Pfam/current_release/Pfam-A.hmm.gz";
+my $PFAM_DAT_URL = "http://ftp.ebi.ac.uk/pub/databases/Pfam/current_release/Pfam-A.hmm.gz";
 my $PFAM2GO_DAT_URL = "http://www.geneontology.org/external2go/pfam2go"; 
 
 
--- util/report_summary/trinotate_report_summary.pl	2018-08-02 16:01:05.364254161 +0200
+++ util/report_summary/trinotate_report_summary.pl	2018-08-02 16:01:14.624160886 +0200
@@ -17,7 +17,7 @@
 my $TOP_TAX_LEVEL = 6;
 
 
-my $UTILDIR = "$FindBin::Bin/util";
+my $UTILDIR = "$FindBin::Bin";
 
 main: {
 
