--- admin/Build_Trinotate_Boilerplate_SQLite_db.pl	2019-08-07 16:30:32.000000000 +0200
+++ admin/Build_Trinotate_Boilerplate_SQLite_db.pl	2019-08-08 23:25:04.356442144 +0200
@@ -14,10 +14,10 @@
 my $no_cleanup_flag = $ARGV[1] || 0;
 
 
-my $UTILDIR = "$FindBin::RealBin/util";
+my $UTILDIR = "$FindBin::RealBin";
 
 ## Resources:
-my $SPROT_DAT_URL = "ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.dat.gz";
+my $SPROT_DAT_URL = "http://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.dat.gz";
 my $EGGNOG_DAT_URL = "http://eggnogdb.embl.de/download/latest/data/NOG/NOG.annotations.tsv.gz";
 my $GENE_ONTOLOGY_DAT_URL = "http://purl.obolibrary.org/obo/go/go-basic.obo";
 my $PFAM_DAT_URL = "ftp://ftp.ebi.ac.uk/pub/databases/Pfam/current_release/Pfam-A.hmm.gz";
 
 
--- util/report_summary/trinotate_report_summary.pl	2019-08-07 16:30:32.000000000 +0200
+++ util/report_summary/trinotate_report_summary.pl	2019-08-08 23:22:31.824343118 +0200
@@ -17,7 +17,7 @@
 my $TOP_TAX_LEVEL = 6;
 
 
-my $UTILDIR = "$FindBin::Bin/util";
+my $UTILDIR = "$FindBin::Bin";
 
 main: {
 
