--- admin/Build_Trinotate_Boilerplate_SQLite_db.pl	2018-08-02 15:49:56.299807109 +0200
+++ admin/Build_Trinotate_Boilerplate_SQLite_db.pl	2018-08-02 15:50:05.030720666 +0200
@@ -14,7 +14,7 @@
 my $no_cleanup_flag = $ARGV[1] || 0;
 
 
-my $UTILDIR = "$FindBin::RealBin/util";
+my $UTILDIR = "$FindBin::RealBin";
 
 ## Resources:
 my $SPROT_DAT_URL = "ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.dat.gz";
