--- sample_data/runMe.sh	2019-08-07 16:30:32.000000000 +0200
+++ sample_data/runMe.sh	2019-08-08 23:11:46.970453769 +0200
@@ -44,7 +44,7 @@
 echo Loading protein set
 echo "###############################"
 
-../Trinotate ${sqlite_db} init --gene_trans_map data/Trinity.fasta.gene_to_trans_map --transcript_fasta data/Trinity.fasta --transdecoder_pep data/Trinity.fasta.transdecoder.pep
+Trinotate ${sqlite_db} init --gene_trans_map data/Trinity.fasta.gene_to_trans_map --transcript_fasta data/Trinity.fasta --transdecoder_pep data/Trinity.fasta.transdecoder.pep
 
 
 
@@ -52,8 +52,8 @@
 echo Loading swissprot blast results
 echo "##############################"
 
-../Trinotate ${sqlite_db} LOAD_swissprot_blastp data/swissprot.blastp.outfmt6
-../Trinotate ${sqlite_db} LOAD_swissprot_blastx data/swissprot.blastx.outfmt6
+Trinotate ${sqlite_db} LOAD_swissprot_blastp data/swissprot.blastp.outfmt6
+Trinotate ${sqlite_db} LOAD_swissprot_blastx data/swissprot.blastx.outfmt6
 
 
 echo "####################################"
@@ -61,36 +61,36 @@
 echo "####################################"
 
 # blastP
-../Trinotate ${sqlite_db} LOAD_custom_blast --outfmt6 data/custom_pombe.blastp.outfmt6 --prog blastp --dbtype custom_pombe_pep
+Trinotate ${sqlite_db} LOAD_custom_blast --outfmt6 data/custom_pombe.blastp.outfmt6 --prog blastp --dbtype custom_pombe_pep
 # blastX, same database
-../Trinotate ${sqlite_db} LOAD_custom_blast --outfmt6 data/custom_pombe.blastx.outfmt6 --prog blastx --dbtype custom_pombe_pep
+Trinotate ${sqlite_db} LOAD_custom_blast --outfmt6 data/custom_pombe.blastx.outfmt6 --prog blastx --dbtype custom_pombe_pep
  
 
 echo "#############################"
 echo Loading PFAM results
 echo "#############################"
 
-../Trinotate ${sqlite_db} LOAD_pfam data/TrinotatePFAM.out 
+Trinotate ${sqlite_db} LOAD_pfam data/TrinotatePFAM.out 
 
 
 echo "############################"
 echo Loading TMHMM results
 echo "############################"
 
-../Trinotate ${sqlite_db} LOAD_tmhmm data/tmhmm.out
+Trinotate ${sqlite_db} LOAD_tmhmm data/tmhmm.out
 
 echo "###########################"
 echo Loading SignalP results
 echo "###########################"
 
-../Trinotate ${sqlite_db} LOAD_signalp data/signalp.out
+Trinotate ${sqlite_db} LOAD_signalp data/signalp.out
 
 
 echo "###########################"
 echo Loading RNAMMER results
 echo "###########################"
 
-../Trinotate ${sqlite_db} LOAD_rnammer data/Trinity.fasta.rnammer.gff
+Trinotate ${sqlite_db} LOAD_rnammer data/Trinity.fasta.rnammer.gff
 
 
 #################################################################
@@ -105,13 +105,13 @@
 echo "###################################################"
 
 # expression data load for genes
-../util/transcript_expression/import_expression_and_DE_results.pl --sqlite ${sqlite_db} --gene_mode \
+import_expression_and_DE_results.pl --sqlite ${sqlite_db} --gene_mode \
         --samples_file data/samples.txt \
         --count_matrix data/Trinity_genes.counts.matrix \
         --fpkm_matrix data/Trinity_genes.TMM.EXPR.matrix 
 
 # DE results load for genes
-../util/transcript_expression/import_expression_and_DE_results.pl --sqlite ${sqlite_db} --gene_mode \
+import_expression_and_DE_results.pl --sqlite ${sqlite_db} --gene_mode \
         --samples_file data/samples.txt \
         --DE_dir edgeR_genes
 
@@ -121,13 +121,13 @@
 echo "##################################################"
 
 # expression data load for transcripts
-../util/transcript_expression/import_expression_and_DE_results.pl --sqlite ${sqlite_db} --transcript_mode \
+import_expression_and_DE_results.pl --sqlite ${sqlite_db} --transcript_mode \
         --samples_file data/samples.txt \
         --count_matrix data/Trinity_trans.counts.matrix \
         --fpkm_matrix data/Trinity_trans.TMM.EXPR.matrix
 
 # DE results load for transcripts
-../util/transcript_expression/import_expression_and_DE_results.pl --sqlite ${sqlite_db} --transcript_mode \
+import_expression_and_DE_results.pl --sqlite ${sqlite_db} --transcript_mode \
         --samples_file data/samples.txt \
         --DE_dir edgeR_trans
 
@@ -138,29 +138,29 @@
 
 
 # import the transcription profile cluster stuff
-../util/transcript_expression/import_transcript_clusters.pl --group_name DE_all_vs_all --analysis_name diffExpr.P0.1_C1.matrix.RData.clusters_fixed_P_60 --sqlite ${sqlite_db} edgeR_trans/diffExpr.P0.1_C1.matrix.RData.clusters_fixed_P_60/*matrix
+import_transcript_clusters.pl --group_name DE_all_vs_all --analysis_name diffExpr.P0.1_C1.matrix.RData.clusters_fixed_P_60 --sqlite ${sqlite_db} edgeR_trans/diffExpr.P0.1_C1.matrix.RData.clusters_fixed_P_60/*matrix
 
 
 echo "###########################"
 echo Generating report table
 echo "###########################"
 
-../Trinotate ${sqlite_db} report > Trinotate_report.xls
+Trinotate ${sqlite_db} report > Trinotate_report.xls
 
 
 echo "#########################################"
 echo Extracting Gene Ontology Mappings Per Gene
 echo "#########################################"
 
-../util/extract_GO_assignments_from_Trinotate_xls.pl  --Trinotate_xls Trinotate_report.xls -G -I > Trinotate_report.xls.gene_ontology
+extract_GO_assignments_from_Trinotate_xls.pl  --Trinotate_xls Trinotate_report.xls -G -I > Trinotate_report.xls.gene_ontology
 
 # Load annotations
-../util/annotation_importer/import_transcript_names.pl ${sqlite_db} Trinotate_report.xls
+import_transcript_names.pl ${sqlite_db} Trinotate_report.xls
 
 
 
 # Generate trinotate report summary statistics
-../util/report_summary/trinotate_report_summary.pl Trinotate_report.xls Trinotate_report_stats
+trinotate_report_summary.pl Trinotate_report.xls Trinotate_report_stats
 
 
 echo "##########################"
