--- Trinotate	2017-07-24 13:53:21.120002438 +0200
+++ Trinotate	2017-07-24 13:56:46.825089493 +0200
@@ -75,19 +75,19 @@
 
 my %command_to_params = ( 
                           # protein-based
-                          'LOAD_swissprot_blastp' => "$loaders_dir/Trinotate_BLAST_loader.pl --sqlite $sqlite_db --outfmt6 $ARGV[0] --prog blastp --dbtype Swissprot",
-                          'LOAD_trembl_blastp' => "$loaders_dir/Trinotate_BLAST_loader.pl --sqlite $sqlite_db --outfmt6 $ARGV[0] --prog blastp --dbtype TrEMBL",
-                          'LOAD_pfam'  => "$loaders_dir/Trinotate_PFAM_loader.pl --sqlite $sqlite_db --pfam $ARGV[0]",
-                          'LOAD_tmhmm' => "$loaders_dir/Trinotate_TMHMM_loader.pl --sqlite $sqlite_db --tmhmm $ARGV[0]",
-                          'LOAD_signalp' => "$loaders_dir/Trinotate_SIGNALP_loader.pl --sqlite $sqlite_db --signalp $ARGV[0]",
+                          'LOAD_swissprot_blastp' => "Trinotate_BLAST_loader.pl --sqlite $sqlite_db --outfmt6 $ARGV[0] --prog blastp --dbtype Swissprot",
+                          'LOAD_trembl_blastp' => "Trinotate_BLAST_loader.pl --sqlite $sqlite_db --outfmt6 $ARGV[0] --prog blastp --dbtype TrEMBL",
+                          'LOAD_pfam'  => "Trinotate_PFAM_loader.pl --sqlite $sqlite_db --pfam $ARGV[0]",
+                          'LOAD_tmhmm' => "Trinotate_TMHMM_loader.pl --sqlite $sqlite_db --tmhmm $ARGV[0]",
+                          'LOAD_signalp' => "Trinotate_SIGNALP_loader.pl --sqlite $sqlite_db --signalp $ARGV[0]",
                           
                           # transcript-based
-                          'LOAD_swissprot_blastx' => "$loaders_dir/Trinotate_BLAST_loader.pl --sqlite $sqlite_db --outfmt6 $ARGV[0] --prog blastx --dbtype Swissprot",
-                          'LOAD_trembl_blastx' => "$loaders_dir/Trinotate_BLAST_loader.pl --sqlite $sqlite_db --outfmt6 $ARGV[0] --prog blastx --dbtype TrEMBL",
-                          'LOAD_rnammer' => "$loaders_dir/Trinotate_RNAMMER_loader.pl --sqlite $sqlite_db --rnammer $ARGV[0]",
+                          'LOAD_swissprot_blastx' => "Trinotate_BLAST_loader.pl --sqlite $sqlite_db --outfmt6 $ARGV[0] --prog blastx --dbtype Swissprot",
+                          'LOAD_trembl_blastx' => "Trinotate_BLAST_loader.pl --sqlite $sqlite_db --outfmt6 $ARGV[0] --prog blastx --dbtype TrEMBL",
+                          'LOAD_rnammer' => "Trinotate_RNAMMER_loader.pl --sqlite $sqlite_db --rnammer $ARGV[0]",
                           
                           # custom blast searches
-                          'LOAD_custom_blast' => "$loaders_dir/Trinotate_BLAST_loader.pl --sqlite $sqlite_db @ARGV",
+                          'LOAD_custom_blast' => "Trinotate_BLAST_loader.pl --sqlite $sqlite_db @ARGV",
     
     
     );
@@ -99,13 +99,13 @@
 }
 elsif ($command eq "report") {
     ## generate Trinotate report.
-    my $cmd = "$FindBin::RealBin/util/Trinotate_report_writer.pl --sqlite $sqlite_db @ARGV";
+    my $cmd = "Trinotate_report_writer.pl --sqlite $sqlite_db @ARGV";
     &process_cmd($cmd);
     
 }
 elsif ($command eq "init") {
     # populate sequence and gene/trans/orf data
-    my $cmd = "$FindBin::RealBin/util/trinotateSeqLoader/TrinotateSeqLoader.pl --sqlite $sqlite_db @ARGV --bulk_load";
+    my $cmd = "TrinotateSeqLoader.pl --sqlite $sqlite_db @ARGV --bulk_load";
     &process_cmd($cmd);
 }
 else {
