--- transabyss-merge.old	2017-11-17 14:21:01.309192851 +0200
+++ transabyss-merge	2017-11-17 14:21:27.241449996 +0200
@@ -10,16 +10,16 @@
 import shutil
 import sys
 import textwrap
-from utilities import package_info
-from utilities.common_utils import StopWatch
-from utilities.common_utils import check_env
-from utilities.common_utils import log
-from utilities.common_utils import path_action
-from utilities.common_utils import paths_action
-from utilities.common_utils import threshold_action
-from utilities.fasta_utils import abyssmap_merge_fastas
-from utilities.fasta_utils import blat_merge_fastas
-from utilities.fasta_utils import filter_fasta
+from transabyss import package_info
+from transabyss.common_utils import StopWatch
+from transabyss.common_utils import check_env
+from transabyss.common_utils import log
+from transabyss.common_utils import path_action
+from transabyss.common_utils import paths_action
+from transabyss.common_utils import threshold_action
+from transabyss.fasta_utils import abyssmap_merge_fastas
+from transabyss.fasta_utils import blat_merge_fastas
+from transabyss.fasta_utils import filter_fasta
 
 
 TRANSABYSS_VERSION = package_info.VERSION
