diff --git a/mzmine-community/build.gradle b/mzmine-community/build.gradle
index 75714db..46dbbe8 100644
--- a/mzmine-community/build.gradle
+++ b/mzmine-community/build.gradle
@@ -391,7 +391,7 @@ runtime {
         if (OperatingSystem.current().isLinux()) {
             // leave installer type empty to generate all that match
             // for rpm, rpmbuild and rpm tools need to be installed, installerType needs to be set to "rpm"
-            // installerType = "deb"
+            installerType = "rpm"
             // additional options for jpackage
             imageOptions = ["--icon", "src/main/resources/mzmineIcon.png"]
             installerOptions = [
diff --git a/mzmine-community/src/main/resources/mzmineversion.properties b/mzmine-community/src/main/resources/mzmineversion.properties
index a2b2912..d0a7a57 100644
--- a/mzmine-community/src/main/resources/mzmineversion.properties
+++ b/mzmine-community/src/main/resources/mzmineversion.properties
@@ -3,6 +3,6 @@
 version.buildmeta=
 version.major=4
 version.minor=7
-version.patch=9
+version.patch=8
 version.prerelease=
-version.semver=4.7.9
+version.semver=4.7.8
