diff --git a/cassiopeia/solver/HybridSolver.py b/cassiopeia/solver/HybridSolver.py
index 68a0d79..56beb65 100644
--- a/cassiopeia/solver/HybridSolver.py
+++ b/cassiopeia/solver/HybridSolver.py
@@ -13,7 +13,7 @@ from typing import Dict, List, Generator, Optional, Tuple
 import multiprocessing
 import networkx as nx
 import numpy as np
-from numpy.lib.arraysetops import unique
+from numpy import unique
 import pandas as pd
 from tqdm.auto import tqdm
 
