--- a/src/subcommand/similarity_main.cpp
+++ b/src/subcommand/similarity_main.cpp
@@ -284,7 +284,7 @@
         const size_t num_bins = 128;
         const size_t num_threads = args::get(p_num_threads) ? args::get(p_num_threads) : 1;
 
-    const uint64_t max_local_maps = std::min(3UL, num_threads);
+    const uint64_t max_local_maps = std::min(uint64_t(3), uint64_t(num_threads));
 
         if (num_threads > 1) {
             // Index the kmer space of the graph