--- src/main.cpp	2024-10-15 18:27:49.000000000 +0000
+++ src/main.cpp	2025-03-04 18:17:19.563405096 +0000
@@ -29,7 +29,7 @@
 
 // We make sure to compile main for the lowest common denominator architecture.
 // This works on GCC and Clang. But we have to declare main and then define it.
-int main(int argc, char *argv[]) __attribute__((__target__("arch=x86-64")));
+// int main(int argc, char *argv[]) __attribute__((__target__("arch=x86-64")));
 
 int main(int argc, char *argv[]) {
     // set a higher value for tcmalloc warnings
