diff --git i/CMakeLists.txt w/CMakeLists.txt
index dc6e9b44..1c34379c 100644
--- i/CMakeLists.txt
+++ w/CMakeLists.txt
@@ -52,7 +52,7 @@ endif()
 
 # Set optimization through command line; see INSTALL.md
 if (${CMAKE_BUILD_TYPE} MATCHES Release)
-  set(EXTRA_FLAGS "-Ofast -march=native -pipe -msse4.2 -funroll-all-loops") #  -fprofile-generate=../pgo")
+  set(EXTRA_FLAGS "-Ofast -march=native -pipe -funroll-all-loops") #  -fprofile-generate=../pgo")
   set(CMAKE_CXX_FLAGS_RELEASE "-DNDEBUG") # reset CXX_FLAGS to be able to replace -O3 with -Ofast
 endif ()
 
diff --git i/src/main.cpp w/src/main.cpp
index 0300dbd1..68e512d4 100644
--- i/src/main.cpp
+++ w/src/main.cpp
@@ -29,7 +29,7 @@ void odgi_help(char** argv) {
 
 // We make sure to compile main for the lowest common denominator architecture.
 // This works on GCC and Clang. But we have to declare main and then define it.
-int main(int argc, char *argv[]) __attribute__((__target__("arch=x86-64")));
+//int main(int argc, char *argv[]) __attribute__((__target__("arch=x86-64")));
 
 int main(int argc, char *argv[]) {
     // set a higher value for tcmalloc warnings
