diff --git a/conanfile.py b/conanfile.py
index 03f32fcb..bcd9a9b0 100644
--- a/conanfile.py
+++ b/conanfile.py
@@ -245,87 +245,12 @@ class HictkConan(ConanFile):
         self.options["opentelemetry-cpp"].with_jaeger = False
         self.options["opentelemetry-cpp"].with_zipkin = False
 
-    def build_requirements(self):
-        # these are all transitive dependencies.
-        # The only reason why they are defined here is to pin their recipe versions
-        if self._with_abseil:
-            # opentelemetry-cpp, re2
-            self.requires("abseil/20250814.0#4e0fdd34a888b97aca482e648fc27a3b", force=True)
-
-        if self._with_bzip2:
-            # libarchive
-            self.requires("bzip2/1.0.8#00b4a4658791c1f06914e087f0e792f5", force=True)
-
-        if self._with_libcurl:
-            # opentelemetry-cpp
-            self.requires("libcurl/8.16.0#5b85991e709eafd00ab757b79814b3ff", force=True)
-
-        if self._with_lz4:
-            # libarchive
-            self.requires("lz4/1.10.0#59fc63cac7f10fbe8e05c7e62c2f3504", force=True)
-
-        if self._with_lzo:
-            # libarchive
-            self.requires("lzo/2.10#f00b10acc508cea70645727d970a23e1", force=True)
-
-        if self._with_protobuf:
-            # opentelemetry-cpp
-            self.requires("protobuf/6.32.1#f481fd276fc23a33b85a3ed1e898b693", force=True)
-
-        if self._with_re2:
-            # arrow
-            self.requires("re2/20251105#2eec620fc641f812a5837c642ecb7d66", force=True)
-
-        if self._with_xz_utils:
-            # libarchive
-            self.requires("xz_utils/5.8.1#536e4c68ef30b7308b28a1c50e663c31", force=True)
-
-        # hdf5, libarchive, and opentelemetry-cpp
-        self.requires("zlib/1.3.1#b8bc2603263cf7eccbd6e17e66b0ed76", force=True)
-
     def requirements(self):
         self.requires("bshoshany-thread-pool/5.0.0#d94da300363f0c35b8f41b2c5490c94d")
         self.requires("concurrentqueue/1.0.4#1e48e1c712bcfd892087c9c622a51502")
-        self.requires("fast_float/8.1.0#bbf67486bec084d167da0f3e13eee534")
-        self.requires("fmt/12.1.0#50abab23274d56bb8f42c94b3b9a40c7", force=True)
-        self.requires("hdf5/1.14.6#0b780319690d537e6cb0683244919955", force=True)
-        self.requires("highfive/3.1.1#d0c724526ebc8ce396ffa1bf7f3c7b64")
-        self.requires("libdeflate/1.25#49fcd3fe6c130c2ec5a01cabb0481ded")
         self.requires("parallel-hashmap/2.0.0#82acae64ffe2693fff5fb3f9df8e1746")
         self.requires("readerwriterqueue/1.0.6#aaa5ff6fac60c2aee591e9e51b063b83")
         self.requires("span-lite/0.11.0#519fd49fff711674cfed8cd17d4ed422")
-        self.requires("spdlog/1.16.0#942c2c39562ae25ba575d9c8e2bdf3b6")
-        self.requires("zstd/1.5.7#fde461c0d847a22f16d3066774f61b11", force=True)
-
-        if self._with_arrow:
-            self.requires("arrow/22.0.0#e46b173ba20adc478f7926495aeed142")
-
-        if self._with_boost or self._with_boost_header_only:
-            self.requires("boost/1.89.0#010f59feedfd171b15f467b42f723d13", force=True)
-
-        if self._with_catch2:
-            self.requires("catch2/3.11.0#a560b55882ff2deb1f4bafad832a1b98")
-
-        if self._with_cli11:
-            self.requires("cli11/2.6.0#41e625280f9c960e9970c1dc7d2c97b8")
-
-        if self._with_eigen:
-            self.requires("eigen/5.0.0#f7561f543f4aafd6d2dc1f6d677e3075", force=True)
-
-        if self._with_libarchive:
-            self.requires("libarchive/3.8.1#b42b1df243cee62014e0a884a9468b7a")
-
-        if self._with_nlohmann_json:
-            self.requires("nlohmann_json/3.12.0#2d634ab0ec8d9f56353e5ccef6d6612c", force=True)
-
-        if self._with_opentelemetry:
-            self.requires("opentelemetry-cpp/1.22.0#a7deeadbd209a75ccb42aeec309dfb51")
-
-        if self._with_pybind11:
-            self.requires("pybind11/3.0.1#81ed7e3cc5c945080013ca4c5c3abbbb")
-
-        if self._with_tomlplusplus:
-            self.requires("tomlplusplus/3.4.0#85dbfed71376fb8dc23cdcc0570e4727")
 
     def configure(self):
         self._configure_arrow()
