diff --git a/scripts/install.sh b/scripts/install.sh
index d00f98d..1d74855 100755
--- a/scripts/install.sh
+++ b/scripts/install.sh
@@ -20,7 +20,7 @@ else
   fi
 fi
 
-PATH=$PATH:$base/nim-$BRANCH/bin/:$PATH:$base/nimble/src
+PATH=$base/nim-$BRANCH/bin/:$base/nimble/src:$PATH
 cd $base
 echo $PATH
 
@@ -40,14 +40,3 @@ fi
 set -x
 cd $base
 nimble install -y
-
-sudo apt-get -qy install bwa make build-essential cmake libncurses-dev ncurses-dev libbz2-dev lzma-dev liblzma-dev \
-     curl  libssl-dev libtool autoconf automake libcurl4-openssl-dev
-git clone --recursive https://github.com/samtools/htslib.git
-
-cd htslib && git checkout 1.5 && autoheader && autoconf && ./configure --enable-libcurl
-cd ..
-make -j 4 -C htslib
-export LD_LIBRARY_PATH=$base/htslib
-ls -lh $base/htslib/*.so
-echo $base
