diff --git a/Makefile b/Makefile
index 272dba3..124074c 100644
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,6 @@
-CFLAGS	= -DNDEBUG -UDEBUG -Wno-char-subscripts -Wall -O3 -std=c++14 -I. -I./include
+CFLAGS = -DNDEBUG -msse4 -maes -mmmx -mpclmul -mpopcnt -mxsaveopt -mfxsr -Wno-char-subscripts -Wall -O3 -std=c++14 -I. -I${PREFIX}/include
 CXXFLAGS= ${CFLAGS}
-LIBS = -L./lib -lz -lsdsl -pthread
+LIBS = -L${PREFIX}/lib -lz -lsdsl -pthread
 
 .PHONY: all clean
 
