diff --git a/Makefile b/Makefile
index a5c8a6c..e3f8459 100644
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,6 @@
-CFLAGS	= -DNDEBUG -march=native -Wno-char-subscripts -Wall -O3 -std=c++14 -I. -I./include
+CFLAGS	= -DNDEBUG -march=native -Wno-char-subscripts -Wall -O3 -std=c++14 -I. -I${PREFIX}/include
 CXXFLAGS= ${CFLAGS}
-LIBS = -L./lib -lz -lsdsl -ltbb
+LIBS = -L${PREFIX}/lib -lz -lsdsl -ltbb
 
 .PHONY: all clean
 
