--- auxprogs/filterBam/Makefile.old	2019-03-27 14:03:06.878946050 -0500
+++ auxprogs/filterBam/Makefile	2019-03-27 14:04:03.328943096 -0500
@@ -3,6 +3,12 @@
 # 	Created: 28-February-2012
 #	Last modified: 09-May-2018
 
+INCLUDES =  ${PREFIX}/include/bamtools/
+LIBS = -L${PREFIX}/lib/ -lbamtools -lz
+SOURCES = bam2hints.cc
+OBJECTS = $(SOURCES:.cc=.o)
+CXXFLAGS += -Wall -O2 # -g -p -g -ggdb 
+
 all: 
 	(cd src;make)
 
