From 9109d2cd73f8ce118d2852a1536c1a4437891935 Mon Sep 17 00:00:00 2001
From: Mathieu Dubois <mdubois@genoscope.cns.fr>
Date: Mon, 19 Aug 2019 01:47:12 +0200
Subject: [PATCH] Remove direct call to GCC in Makefile.

---
 SELECT/makefile | 2 --
 1 file changed, 2 deletions(-)

diff --git a/SELECT/makefile b/SELECT/makefile
index a1cd3f2..179ec3a 100644
--- a/SELECT/makefile
+++ b/SELECT/makefile
@@ -1,8 +1,6 @@
 # makefile to compile shared objects using the gnu C compiler
 # Stefan Kurtz, October 2000
 
-CC=gcc
-
 # in 64-bit mode add -m64
 
 CFLAGS=-Wall -Werror -O3 -g
-- 
2.21.0

