--- SuffixTools/STCommon.h
+++ SuffixTools/STCommon.h
@@ -99 +99 @@
-        static const uint64_t HIGH_MASK = ~0 << POS_BITS;
+        static const uint64_t HIGH_MASK = ~0ULL << POS_BITS;
--- Util/StdAlnTools.cpp
+++ Util/StdAlnTools.cpp
@@ -122 +122 @@
-        bool success = parser >> code;
+        bool success{parser >> code};
--- StringGraph/SGVisitors.h.   2020-08-23 02:53:32.295545638 +0000
+++ StringGraph/SGVisitors.h    2020-08-23 03:01:23.900861468 +0000
@@ -275 +275 @@
-        if(m_fileHandle!=NULL)
+        if(m_fileHandle)
@@ -327 +327 @@
-           if(m_fileHandle!=NULL)
+           if(m_fileHandle)
@@ -354 +354 @@
-    if(m_fileHandle!=NULL)
+    if(m_fileHandle)
--- Algorithm/KmerOverlaps.cpp
+++ Algorithm/KmerOverlaps.cpp
@@ -209 +209 @@
-        if(abs(pos_0-pos_1) > maxshift)
+        if(abs(static_cast<ssize_t>(pos_0)-static_cast<ssize_t>(pos_1)) > maxshift)
--- StringGraph/SGVisitors.cpp
+++ StringGraph/SGVisitors.cpp
@@ -795 +795 @@
-        if(m_fileHandle!=NULL)
+        if(m_fileHandle)
@@ -802 +802 @@
-            if(m_fileHandle!=NULL) m_fileHandle << " BLACK";
+            if(m_fileHandle) m_fileHandle << " BLACK";
@@ -809 +809 @@
-            if(m_fileHandle!=NULL) m_fileHandle << " BLACK " << avgKmerFreqs ;
+            if(m_fileHandle) m_fileHandle << " BLACK " << avgKmerFreqs ;
@@ -813 +813 @@
-        if(m_fileHandle!=NULL)
+        if(m_fileHandle)
@@ -876 +876 @@
-                               if(m_fileHandle!=NULL)
+                               if(m_fileHandle)
@@ -1002 +1002 @@
-                               if(m_fileHandle!=NULL)
+                               if(m_fileHandle)
--- SuffixTools/SparseGapArray.h
+++ SuffixTools/SparseGapArray.h
@@ -323 +323 @@
-            m_overflow.insert(std::make_pair<size_t, OverflowStorage>(i, c));
+            m_overflow.insert(std::make_pair(i, c));
--- StriDe/overlap.cpp
+++ StriDe/overlap.cpp
@@ -27,13 +27,7 @@
 #include <sys/stat.h>
 
 /*Tatsuki include */
-#include <tr1/unordered_map>
-#include <tr1/unordered_set>
-#define HashMap std::tr1::unordered_map
-#define HashSet std::tr1::unordered_set
-typedef std::tr1::hash<std::string> StringHasher;
-#include <sparsehash/dense_hash_set>
-#define DenseHashSet google::dense_hash_set
+#include "HashMap.h"
 typedef DenseHashSet<std::string, StringHasher> ReadHashSet;
 typedef ReadHashSet::iterator ReadHashSetIter;
 #include <time.h>
