diff --git a/Algorithm/KmerOverlaps.cpp b/Algorithm/KmerOverlaps.cpp
index c2d918d..2bbf976 100755
--- a/Algorithm/KmerOverlaps.cpp
+++ b/Algorithm/KmerOverlaps.cpp
@@ -206,7 +206,7 @@ SequenceOverlapPairVector KmerOverlaps::retrieveMatches(const std::string& query
 
         //The pos of two kmers shouldn't differ more than the min_overlap allowed
         size_t maxshift = query.length()-min_overlap+bandwidth/2;
-        if(abs(pos_0-pos_1) > maxshift)
+        if(abs((long long int) (pos_0-pos_1)) > maxshift)
             continue;
 
         // Check for secondary occurrences
