--- make_pscores.pl     2004-03-22 20:56:57.000000000 +0100
+++ make_pscores.new.pl 2016-01-27 12:58:05.091462094 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 #
 # Usage: make_pscores.pl SEQFILE SCOREFILE
 #
@@ -13,8 +13,8 @@
 $pscore_file = $ARGV[1];

 open(PSCORE_OUT, ">$pscore_file");
-system("./formatdb -i $seq_file -p T");
-system("./blastall -p blastp -d $seq_file -i $seq_file -M BLOSUM80 -o $seq_file.out");
+system("formatdb -i $seq_file -p T");
+system("blastall -p blastp -d $seq_file -i $seq_file -M BLOSUM80 -o $seq_file.out");

 open(BLAST_OUT, "<$seq_file.out");
 while(<BLAST_OUT>){

