diff --git a/pyproject.toml b/pyproject.toml
index b990a07..999a456 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -17,7 +17,7 @@ authors = [
 ]
 dynamic = ["version"]
 readme = "README.md"
-license = {file = "LICENSE"}
+license = "MIT"
 requires-python = ">=3.9"
 
 dependencies = [
@@ -42,6 +42,11 @@ Issues = "https://github.com/nixonlab/stellarscope/issues"
 version_file = 'stellarscope/_version.py'
 
 [tool.setuptools]
+packages = [
+    "stellarscope",
+    "stellarscope.utils",
+    "stellarscope.annotation",
+]
 include-package-data = true
 
 [tool.setuptools.package-data]
