--- Makefile
+++ Makefile
@@ -65,23 +65,23 @@
 	$(CXX) $(CFLAGS) $(OPT) -c -o $(OBJ)ascii.o $(SRC)ascii.cpp
 	
     sffreader.o: $(SRC)sffreader.cpp $(SRC)sff.h
-	g++ $(CFLAGS) $(OPT) -c -o $(OBJ)sffreader.o $(SRC)sffreader.cpp
+	$(CXX) $(CFLAGS) $(OPT) -c -o $(OBJ)sffreader.o $(SRC)sffreader.cpp
     
     sff.o: 
-	g++ $(CFLAGS) $(OPT) -c -o $(OBJ)sff.o $(SRC)sff.c
+	$(CXX) $(CFLAGS) $(OPT) -c -o $(OBJ)sff.o $(SRC)sff.c
 	
     poly.o: 
-	g++ $(CFLAGS) $(OPT) -c -o $(OBJ)poly.o $(SRC)poly.c
+	$(CXX) $(CFLAGS) $(OPT) -c -o $(OBJ)poly.o $(SRC)poly.c
 	
     abi.o: 
-	g++ $(CFLAGS) $(OPT) -c -o $(OBJ)abi.o $(SRC)abi.c
+	$(CXX) $(CFLAGS) $(OPT) -c -o $(OBJ)abi.o $(SRC)abi.c
 	
     libgzstream.a : $(OBJ)gzstream.o $(SRC)gzstream.h
 	${AR} $(OBJ)libgzstream.a $(OBJ)gzstream.o
 	
     gzstream.o : $(SRC)gzstream.C $(SRC)gzstream.h
 	#$(CXX) -I $(SRC) -O -Wall -c -o $(OBJ)gzstream.o $(SRC)gzstream.C 
-	gcc -I $(SRC) $(OPT) -Wall -c -o $(OBJ)gzstream.o $(SRC)gzstream.C 
+	$(CC) -I $(SRC) $(OPT) -Wall -c -o $(OBJ)gzstream.o $(SRC)gzstream.C 
 
 else
 
@@ -142,23 +142,23 @@
 	$(CXX) $(CFLAGS) ${PLATFORM} $(OPT) -c -o $(OBJ)ascii.o $(SRC)ascii.cpp
 
     sffreader.o: $(SRC)sffreader_lin.cpp $(SRC)sff_lin.h
-	g++ $(CFLAGS) ${PLATFORM} $(OPT) -c -o $(OBJ)sffreader.o $(SRC)sffreader_lin.cpp
+	$(CXX) $(CFLAGS) ${PLATFORM} $(OPT) -c -o $(OBJ)sffreader.o $(SRC)sffreader_lin.cpp
 	
     sff.o: 
-	g++ $(CFLAGS) ${PLATFORM} $(OPT) -c -o $(OBJ)sff.o $(SRC)sff_lin.c
+	$(CXX) $(CFLAGS) ${PLATFORM} $(OPT) -c -o $(OBJ)sff.o $(SRC)sff_lin.c
 	
     poly.o: 
-	g++ $(CFLAGS) ${PLATFORM} $(OPT) -c -o $(OBJ)poly.o $(SRC)poly.c
+	$(CXX) $(CFLAGS) ${PLATFORM} $(OPT) -c -o $(OBJ)poly.o $(SRC)poly.c
 	
     abi.o: 
-	g++ $(CFLAGS) ${PLATFORM} $(OPT) -c -o $(OBJ)abi.o $(SRC)abi.c
+	$(CXX) $(CFLAGS) ${PLATFORM} $(OPT) -c -o $(OBJ)abi.o $(SRC)abi.c
 	
     libgzstream.a : $(OBJ)gzstream.o $(SRC)gzstream.h
 	${AR} $(OBJ)libgzstream.a $(OBJ)gzstream.o
 	
     gzstream.o : $(SRC)gzstream.C $(SRC)gzstream.h
 	#$(CXX) -I $(SRC) -O -Wall -c -o $(OBJ)gzstream.o $(SRC)gzstream.C 
-	gcc -I $(SRC) $(OPT) -Wall -c -o $(OBJ)gzstream.o $(SRC)gzstream.C 
+	$(CC) -I $(SRC) $(OPT) -Wall -c -o $(OBJ)gzstream.o $(SRC)gzstream.C 
 
 
 endif
