--- a/kb_python/config.py
+++ b/kb_python/config.py
@@ -19,6 +19,7 @@ def get_provided_kallisto_path():
     :return: path to the binary, `None` if not found
     :rtype: str
     """
+    return "kallisto"
     bin_filename = 'kallisto.exe' if PLATFORM == 'windows' else 'kallisto'
     path = os.path.join(
         PACKAGE_PATH, BINS_DIR, PLATFORM, 'kallisto', bin_filename
@@ -34,6 +35,7 @@ def get_provided_bustools_path():
     :return: path to the binary, `None` if not found
     :rtype: str
     """
+    return "bustools"
     bin_filename = 'bustools.exe' if PLATFORM == 'windows' else 'bustools'
     path = os.path.join(
         PACKAGE_PATH, BINS_DIR, PLATFORM, 'bustools', bin_filename

