diff --git a/CMakeLists.txt b/CMakeLists.txt
index f1f7d73..bc88972 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -10,9 +10,9 @@ set(CMAKE_CXX_STANDARD 17)
 
 enable_testing()
 
-include(CheckIPOSupported) # adds lto
+include(CheckIPOSupported) # adds ltoT
 check_ipo_supported(RESULT ipo_supported OUTPUT output)
-SET(CMAKE_RANLIB "gcc-ranlib") # too try lto with older runtimes
+#SET(CMAKE_RANLIB "gcc-ranlib") # too try lto with older runtimes
 
 # This builds a static version of ./bin/smoothxg
 option(BUILD_STATIC "Build static binary" OFF)
@@ -42,9 +42,7 @@ endif()
 #   cmake -DCMAKE_BUILD_TYPE=Debug ..
 
 if (${CMAKE_BUILD_TYPE} MATCHES Release)
-  if(NOT EXTRA_FLAGS)
-    set(EXTRA_FLAGS "-Ofast -march=native -flto -fno-fat-lto-objects")
-  endif()
+  set(EXTRA_FLAGS "-Ofast -march=native")
   set(CMAKE_CXX_FLAGS_RELEASE "-DNDEBUG") # reset CXX_FLAGS to replace -O3 with -Ofast
 
   # Increase SPOA's performance
