diff --git a/CMakeLists.txt b/CMakeLists.txt
index f1f7d73..96142b2 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -10,9 +10,9 @@ set(CMAKE_CXX_STANDARD 17)
 
 enable_testing()
 
-include(CheckIPOSupported) # adds lto
+include(CheckIPOSupported) # adds ltoT
 check_ipo_supported(RESULT ipo_supported OUTPUT output)
-SET(CMAKE_RANLIB "gcc-ranlib") # too try lto with older runtimes
+#SET(CMAKE_RANLIB "gcc-ranlib") # too try lto with older runtimes
 
 # This builds a static version of ./bin/smoothxg
 option(BUILD_STATIC "Build static binary" OFF)
