--- bowtie2.orig        2017-01-20 19:41:25.706765000 -0500
+++ bowtie2     2017-01-20 16:23:38.574188000 -0500
@@ -38,10 +38,10 @@
  my ($vol,$script_path,$prog);
  $prog = File::Spec->rel2abs( __FILE__ );

-while (-f $prog && -l $prog){
-    my (undef, $dir, undef) = File::Spec->splitpath($prog);
-    $prog = File::Spec->rel2abs(readlink($prog), $dir);
-}
+#while (-f $prog && -l $prog){
+#    my (undef, $dir, undef) = File::Spec->splitpath($prog);
+#    $prog = File::Spec->rel2abs(readlink($prog), $dir);
+#}

  ($vol,$script_path,$prog)
                  = File::Spec->splitpath($prog);
