diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index b2d24ff..007a3ac 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -32,7 +32,7 @@ include_directories(${PROJECT_SOURCE_DIR}/include/)
 
 set_source_files_properties(main.cpp PROPERTIES COMPILE_FLAGS "-DGIT_REVISION=`../../build-support/git_describe.sh`")
 
-link_libraries("-flto -pthread -Wl,--format=binary -Wl,../../LICENSE -Wl,--format=default") # see show-license command in main.cpp
+link_libraries("-flto -Wl,--whole-archive -lrt -lpthread -Wl,--no-whole-archive -Wl,--format=binary -Wl,../../LICENSE -Wl,--format=default") # see show-license command in main.cpp
 
 #add_compile_options(-fsanitize=address)
 #add_link_options(-fsanitize=address)
