--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -5,6 +5,6 @@
 set(gx_VERSION_MINOR 0)

 set(CMAKE_CXX_FLAGS_DEBUG   "${CMAKE_CXX_FLAGS} -O0 -ggdb3 -DDEBUG")
-set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS} -static -O2 -DNDEBUG")  # TODO: investigate whether can/should enable -march=x86-64-v4
+set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS} -O2 -DNDEBUG")  # TODO: investigate whether can/should enable -march=x86-64-v4

 add_subdirectory(src)
