From 9b6b61deb41b9e8d6a564866d2462fb1d52a06be Mon Sep 17 00:00:00 2001
From: Christopher Pockrandt <noreply-github@cpockrandt.de>
Date: Sat, 28 May 2022 00:43:45 +0200
Subject: [PATCH] [FIX] Missing include

---
 src/classify.cpp | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/classify.cpp b/src/classify.cpp
index c0f1130..0bbd58e 100644
--- a/src/classify.cpp
+++ b/src/classify.cpp
@@ -18,6 +18,8 @@
  * along with Kraken.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#define __STDC_FORMAT_MACROS 1 // for PRIu64, etc.
+
 #include "kraken_headers.hpp"
 #include "krakendb.hpp"
 #include "krakenutil.hpp"
@@ -31,6 +33,7 @@
 #include <inttypes.h>
 #include <cassert>
 #include <cstdio>
+#include <inttypes.h>
 
 const size_t DEF_WORK_UNIT_SIZE = 500000;
 int New_taxid_start = 1000000000;
