diff --git a/src/Makefile b/src/Makefile
index 0a4e924..5342bd2 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -1,4 +1,3 @@
-CXX = g++
 FOPENMP?=-fopenmp
 NDEBUG=-D NDEBUG
 
@@ -11,7 +10,7 @@ FOPENMP=
 endif
 
 
-CXXFLAGS = -Wall -Wextra -Wfatal-errors -pipe -O2 -std=c++11 $(FOPENMP) -I./gzstream $(NDEBUG) ${CPPFLAGS} 
+CXXFLAGS += -Wall -Wextra -Wfatal-errors -pipe -O2 -std=c++11 $(FOPENMP) -I./gzstream $(NDEBUG) ${CPPFLAGS} 
 #CXXFLAGS = -Wall -std=c++11 $(FOPENMP) -O3 -Wfatal-errors
 PROGS1 = classify classifyExact db_sort set_lcas db_shrink build_taxdb read_uid_mapping count_unique dump_taxdb query_taxdb
 TEST_PROGS = grade_classification test_hll_on_db dump_db_kmers
