--- auxprogs/filterBam/src/Makefile	2017-01-18 08:45:16.302477496 +0400
+++ auxprogs/filterBam/src/Makefile	2017-01-18 08:49:27.192115685 +0400
@@ -13,8 +13,8 @@
 INCLUDES =  -I${PREFIX}/include/bamtools/ -Iheaders -I./bamtools
 #LIBS = -lbamtools -lz
 LIBS = ${PREFIX}/lib/libbamtools.a -lz
-CFLAGS = -std=c++0x
-CXXFLAGS += -Wall -O2 # -g -p -g -ggdb 
+CFLAGS = -std=c++
+CXXFLAGS += -Wall -O2 -std=c++ # -g -p -g -ggdb 
 VPATH = functions
 
 all : $(PROGRAM) $(OBJECTS) CHECKBAM BIN
