diff -u -r ./cmh-test.pl ../popoolation2_1201/cmh-test.pl
--- ./cmh-test.pl	2011-09-20 13:21:57.000000000 -0500
+++ ../popoolation2_1201/cmh-test.pl	2023-05-22 17:46:15.143501024 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 use strict;
 use warnings;
 use Getopt::Long;
diff -u -r ./create-genewise-sync.pl ../popoolation2_1201/create-genewise-sync.pl
--- ./create-genewise-sync.pl	2011-09-20 13:21:57.000000000 -0500
+++ ../popoolation2_1201/create-genewise-sync.pl	2023-05-22 17:46:15.144501022 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 {
 use strict;
 use warnings;
diff -u -r ./export/cmh2gwas.pl ../popoolation2_1201/export/cmh2gwas.pl
--- ./export/cmh2gwas.pl	2011-09-20 13:21:57.000000000 -0500
+++ ../popoolation2_1201/export/cmh2gwas.pl	2023-05-22 17:46:33.047476877 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 use strict;
 use warnings;
 use Getopt::Long;
diff -u -r ./export/pwc2igv.pl ../popoolation2_1201/export/pwc2igv.pl
--- ./export/pwc2igv.pl	2011-09-20 13:21:57.000000000 -0500
+++ ../popoolation2_1201/export/pwc2igv.pl	2023-05-22 17:46:33.048476876 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 use strict;
 use warnings;
 use Getopt::Long;
diff -u -r ./export/subsample_sync2fasta.pl ../popoolation2_1201/export/subsample_sync2fasta.pl
--- ./export/subsample_sync2fasta.pl	2011-09-20 13:21:57.000000000 -0500
+++ ../popoolation2_1201/export/subsample_sync2fasta.pl	2023-05-22 17:46:33.049476875 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 use strict;
 use warnings;
 use Getopt::Long;
diff -u -r ./export/subsample_sync2GenePop.pl ../popoolation2_1201/export/subsample_sync2GenePop.pl
--- ./export/subsample_sync2GenePop.pl	2011-09-20 13:21:57.000000000 -0500
+++ ../popoolation2_1201/export/subsample_sync2GenePop.pl	2023-05-22 17:46:33.050476873 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 use strict;
 use warnings;
 use Getopt::Long;
diff -u -r ./fisher-test.pl ../popoolation2_1201/fisher-test.pl
--- ./fisher-test.pl	2012-03-22 07:06:54.000000000 -0500
+++ ../popoolation2_1201/fisher-test.pl	2023-05-22 17:46:15.145501021 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 use strict;
 use warnings;
 use Getopt::Long;
diff -u -r ./fst-sliding.pl ../popoolation2_1201/fst-sliding.pl
--- ./fst-sliding.pl	2011-11-23 04:11:55.000000000 -0600
+++ ../popoolation2_1201/fst-sliding.pl	2023-05-22 17:46:15.146501020 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 use strict;
 use warnings;
 use Getopt::Long;
diff -u -r ./indel_filtering/filter-sync-by-gtf.pl ../popoolation2_1201/indel_filtering/filter-sync-by-gtf.pl
--- ./indel_filtering/filter-sync-by-gtf.pl	2011-09-20 13:21:57.000000000 -0500
+++ ../popoolation2_1201/indel_filtering/filter-sync-by-gtf.pl	2023-05-22 17:46:37.694470605 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 use strict;
 use warnings;
 use Data::Dumper;
diff -u -r ./indel_filtering/identify-indel-regions.pl ../popoolation2_1201/indel_filtering/identify-indel-regions.pl
--- ./indel_filtering/identify-indel-regions.pl	2011-09-20 13:21:57.000000000 -0500
+++ ../popoolation2_1201/indel_filtering/identify-indel-regions.pl	2023-05-22 17:46:37.695470604 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 use strict;
 use warnings;
 use Data::Dumper;
diff -u -r ./mpileup2sync.pl ../popoolation2_1201/mpileup2sync.pl
--- ./mpileup2sync.pl	2011-09-20 13:21:57.000000000 -0500
+++ ../popoolation2_1201/mpileup2sync.pl	2023-05-22 17:46:15.147501018 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 use strict;
 use warnings;
 use Getopt::Long;
diff -u -r ./snp-frequency-diff.pl ../popoolation2_1201/snp-frequency-diff.pl
--- ./snp-frequency-diff.pl	2012-02-01 06:11:55.000000000 -0600
+++ ../popoolation2_1201/snp-frequency-diff.pl	2023-05-22 17:46:15.148501017 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 use strict;
 use warnings;
 use Getopt::Long;
diff -u -r ./subsample-synchronized.pl ../popoolation2_1201/subsample-synchronized.pl
--- ./subsample-synchronized.pl	2011-09-20 13:21:57.000000000 -0500
+++ ../popoolation2_1201/subsample-synchronized.pl	2023-05-22 17:46:15.148501017 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 use strict;
 use warnings;
 use Getopt::Long;
diff -u -r ./synchronize-pileup.pl ../popoolation2_1201/synchronize-pileup.pl
--- ./synchronize-pileup.pl	2011-09-20 13:21:57.000000000 -0500
+++ ../popoolation2_1201/synchronize-pileup.pl	2023-05-22 17:46:15.150501014 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 #
 # Authors:
 # Dr. Robert Kofler
