--- src/debug_utils.cc.orig	2024-03-19 15:40:50.492023297 +0200
+++ src/debug_utils.cc	2024-03-19 15:44:34.032276878 +0200
@@ -476,7 +476,10 @@
 void FWrite(FILE* file, const std::string& str) {
   auto simple_fwrite = [&]() {
     // The return value is ignored because there's no good way to handle it.
-    fwrite(str.data(), str.size(), 1, file);
+    size_t ret = fwrite(str.data(), str.size(), 1, file);
+    if (ret != str.size()) {
+      fprintf(stderr, "fwrite returned error : %lu\n", ret);
+    }
   };
 
   if (file != stderr && file != stdout) {
