--- libsequence/src/summstats.cc.old	2019-05-29 23:26:48.240704101 -0500
+++ libsequence/src/summstats.cc	2019-05-29 23:28:11.643748715 -0500
@@ -14,9 +14,9 @@
 #include <Sequence/PolySIM.hpp>
 #include <Sequence/PolyTable.hpp>
 #include <Sequence/SimData.hpp>
-#include <Sequence/SummStats/nSL.hpp>
-#include <Sequence/SummStats/Garud.hpp>
-#include <Sequence/SummStats/lHaf.hpp>
+#include <Sequence/SummStatsDeprecated/nSL.hpp>
+#include <Sequence/SummStatsDeprecated/Garud.hpp>
+#include <Sequence/SummStatsDeprecated/lHaf.hpp>
 #include <Sequence/Recombination.hpp>
 #include <Sequence/stateCounter.hpp>
 
