diff --git a/CMakeLists.txt b/CMakeLists.txt
index 8e7ec080..f1275556 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -146,13 +146,13 @@ if (WIN32)
     add_definitions(-DWIN32)
 elseif (APPLE)
     message("Target OS     : Mac OS X")
-    # to be compatible back to Mac OS X 10.7
+    # to be compatible back to Mac OS X 10.9
     if (IQTREE_FLAGS MATCHES "oldmac")
         add_definitions("-mmacosx-version-min=10.5")
         set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -mmacosx-version-min=10.5")
     else()
-        add_definitions("--target=x86_64-apple-macos10.7")
-        set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} --target=x86_64-apple-macos10.7")
+        add_definitions("--target=x86_64-apple-macos10.9")
+        set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} --target=x86_64-apple-macos10.9")
     endif()
     SET(CMAKE_FIND_LIBRARY_SUFFIXES .a ${CMAKE_FIND_LIBRARY_SUFFIXES})
 elseif (UNIX)
