From b55eea2df0a4d474078bf5c4f7929b6c22d0868e Mon Sep 17 00:00:00 2001
From: Matthew Dillon <matthewrdillon@gmail.com>
Date: Mon, 16 Mar 2020 10:55:28 -0700
Subject: [PATCH] PATCH: macos 10.9

---
 CMakeLists.txt | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 2f2a01ca..16a75924 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -89,13 +89,13 @@ if (WIN32)
     add_definitions(-DWIN32)
 elseif (APPLE)
 	message("Target OS     : Mac OS X")
-	# to be compatible back to Mac OS X 10.7
+	# to be compatible back to Mac OS X 10.9
 	if (IQTREE_FLAGS MATCHES "oldmac")
 		add_definitions("-mmacosx-version-min=10.5")
 		set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -mmacosx-version-min=10.5")
 	else()
-		add_definitions("-mmacosx-version-min=10.7")
-		set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -mmacosx-version-min=10.7")
+		add_definitions("-mmacosx-version-min=10.9")
+		set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -mmacosx-version-min=10.9")
 	endif()
     SET(CMAKE_FIND_LIBRARY_SUFFIXES .a ${CMAKE_FIND_LIBRARY_SUFFIXES})
 elseif (UNIX)
-- 
2.24.1

