--- cime/scripts/lib/CIME/case/case.py.ori	2021-04-05 21:18:32.522432308 +0000
+++ cime/scripts/lib/CIME/case/case.py	2021-04-05 21:20:21.579464813 +0000
@@ -1144,14 +1144,6 @@
                 note = "This component includes user_mods {}".format(user_mods)
                 append_status(note, "README.case", caseroot=self._caseroot)
                 logger.info(note)
-        gittag = run_cmd_no_fail('git remote -v',from_dir=self.get_value("CIMEROOT"))
-        gitbranch = run_cmd_no_fail('git branch -vv',from_dir=self.get_value("CIMEROOT"))
-        gitlog = run_cmd_no_fail('git log -n 1',from_dir=self.get_value("CIMEROOT"))
-        append_status("INFORMATION ABOUT YOUR GIT VERSION CONTROL SYSTEM :","README.case", caseroot=self._caseroot)
-        append_status("remote branch:{}".format(gittag),"README.case", caseroot=self._caseroot)
-        append_status("git branch:{}".format(gitbranch),"README.case", caseroot=self._caseroot)
-        append_status("git log:{}".format(gitlog.encode('utf8')),"README.case", caseroot=self._caseroot)
-#
         if not clone:
             self._create_caseroot_sourcemods()
         self._create_caseroot_tools()
