--- ch4Mod.F90-orig
+++ ch4Mod.F90
@@ -3602,10 +3602,11 @@
                      pondz = h2osfc(c) / 1000._r8 / frac_h2osfc(c) ! Assume all h2osfc corresponds to sat area
                      ! mm      /  mm/m
                      pondres = pondres + pondz / ponddiff
-                  else if (.not. lake .and. sat == 1 .and. frac_h2osfc(c) > 0._r8 .and. &
-                       h2osfc(c)/frac_h2osfc(c) > capthick) then ! Assuming short-circuit logic will avoid FPE here.
-                     ! assume surface ice is impermeable
-                     pondres = 1/smallnumber
+                  else if (.not. lake .and. sat == 1 .and. frac_h2osfc(c) > 0._r8) then
+                          if (h2osfc(c)/frac_h2osfc(c) > capthick) then ! Assuming short-circuit logic will NOT avoid FPE here.
+                             ! assume surface ice is impermeable
+                             pondres = 1/smallnumber
+                          end if
                   end if
 
                   spec_grnd_cond(c,s) = 1._r8/(1._r8/grnd_ch4_cond(c) + snowres(c) + pondres)
