--- seq_infodata_mod.F90-orig
+++ seq_infodata_mod.F90
@@ -2901,7 +2901,7 @@
     write(logunit,F0I) subname,'flux_max_iteration       = ', infodata%flux_max_iteration
     write(logunit,F0R) subname,'gust_fac                 = ', infodata%gust_fac
     write(logunit,F0I) subname,'flux_scheme              = ', infodata%flux_scheme  !+tht COARE
-    write(logunit,F0I) subname,'alb_cosz_avg             = ', infodata%alb_cosz_avg !+tht albedos
+    write(logunit,F0L) subname,'alb_cosz_avg             = ', infodata%alb_cosz_avg !+tht albedos
     write(logunit,F0A) subname,'glc_renormalize_smb      = ', trim(infodata%glc_renormalize_smb)
     write(logunit,F0R) subname,'wall_time_limit          = ', infodata%wall_time_limit
     write(logunit,F0A) subname,'force_stop_at            = ', trim(infodata%force_stop_at)
