--- src/haru/makefile.gcc
+++ src/haru/makefile.gcc
@@ -3 +3 @@
-
+AR?=ar
@@ -36,2 +36,2 @@
-CPPFLAGS=-I.  -O3 -DNDEBUG $(MYCPPFLAGS) $(ENCRYPTFLAGS)
-CFLAGS=-I.  -O3 $(MYCPPFLAGS) $(ENCRYPTFLAGS)
+CPPFLAGS+=-I.  -O3 -DNDEBUG $(MYCPPFLAGS) $(ENCRYPTFLAGS)
+CFLAGS+=-I.  -O3 $(MYCPPFLAGS) $(ENCRYPTFLAGS)
@@ -45 +45 @@
-LDFLAGS=-L. -lharu -lstdc++ $(MYLDFLAGS)
+LDFLAGS+=-L. -lharu -lstdc++ $(MYLDFLAGS)
@@ -126 +126 @@
-	ar rc $@ $(OBJS)
+	$(AR) rc $@ $(OBJS)
--- src/haru/makefile.osx
+++ src/haru/makefile.osx
@@ -3 +3 @@
-
+ar?=ar
@@ -36,2 +36,2 @@
-CPPFLAGS= $(MAINCFLAGS) -I.  -DNDEBUG $(MYCPPFLAGS) $(ENCRYPTFLAGS)
-CFLAGS=   $(MAINCFLAGS) -I.  $(MYCPPFLAGS) $(ENCRYPTFLAGS)
+CPPFLAGS+= $(MAINCFLAGS) -I.  -DNDEBUG $(MYCPPFLAGS) $(ENCRYPTFLAGS)
+CFLAGS+=   $(MAINCFLAGS) -I.  $(MYCPPFLAGS) $(ENCRYPTFLAGS)
@@ -45 +45 @@
-LDFLAGS=-L. -lharu -lstdc++ $(MYLDFLAGS)
+LDFLAGS+=-L. -lharu -lstdc++ $(MYLDFLAGS)
