--- src/Makefile.in
+++ src/Makefile.in
@@ -132 +132 @@
-LDFLAGS           = 
+LDFLAGS           = @LDFLAGS@
@@ -184 +184 @@
-	$(CC) $(CFLAGS) -c $*.cpp 
+	$(CXX) $(CFLAGS) -c $*.cpp 
@@ -257 +257 @@
-	(cd ./haru ; $(GMAKE) -f makefile.$(SYSTEM) "CC=$(CXX)" "CPP=$(CC)" "ARCH=$(ARCH)" "MACOSVERSION=$(MACOSVERSION)" "MYCPPFLAGS=$(PRETTYCFLAGS) $(ZLIBHARUINCL)")
+	(cd ./haru ; $(GMAKE) -f makefile.$(SYSTEM) "CC=$(CC)" "CPP=$(CXX)" "ARCH=$(ARCH)" "MACOSVERSION=$(MACOSVERSION)" "MYCPPFLAGS=$(PRETTYCFLAGS) $(ZLIBHARUINCL)")
@@ -260 +260 @@
-	(cd ./haru ; $(GMAKE) -f makefile.$(SYSTEM) "CC=$(CXX)" "CPP=$(CC)" "ARCH=$(ARCH)" "MACOSVERSION=$(MACOSVERSION)" "MYCPPFLAGS= $(PRETTYCFLAGS) $(ZLIBHARUINCL)")
+	(cd ./haru ; $(GMAKE) -f makefile.$(SYSTEM) "CC=$(CC)" "CPP=$(CXX)" "ARCH=$(ARCH)" "MACOSVERSION=$(MACOSVERSION)" "MYCPPFLAGS= $(PRETTYCFLAGS) $(ZLIBHARUINCL)")
@@ -465 +465 @@
-	$(CC) $(CFLAGS) $(LDFLAGS) -o $@ $(OFILES) $(LIBS) ;
+	$(CXX) $(CFLAGS) $(LDFLAGS) -o $@ $(OFILES) $(LIBS) ;
@@ -468 +468 @@
-	$(CC) $(CFLAGS) $(LDFLAGS) -o $@ $(OFILES) $(LIBS) ;
+	$(CXX) $(CFLAGS) $(LDFLAGS) -o $@ $(OFILES) $(LIBS) ;
@@ -471 +471 @@
-	$(CC) $(CFLAGS) $(LDFLAGS) -o $@ $(OFILES) $(OTHERLIBS) $(WINLIBS) ;
+	$(CXX) $(CFLAGS) $(LDFLAGS) -o $@ $(OFILES) $(OTHERLIBS) $(WINLIBS) ;
@@ -475 +475 @@
-	$(CC) $(CFLAGS) $(LDFLAGS) -o $@ $(OFILES) $(EFENCELIBS) $(LIBS) ;
+	$(CXX) $(CFLAGS) $(LDFLAGS) -o $@ $(OFILES) $(EFENCELIBS) $(LIBS) ;
@@ -479 +479 @@
-	$(CC) $(CFLAGS) $(LDFLAGS) -o $@ $(OFILES) $(MEMDEBUGLIBS) $(LIBS) ;
+	$(CXX) $(CFLAGS) $(LDFLAGS) -o $@ $(OFILES) $(MEMDEBUGLIBS) $(LIBS) ;
@@ -483 +483 @@
-	$(CC) $(CFLAGS) $(LDFLAGS) -o $@ $(OFILES) $(DMALLOCLIBS) $(LIBS) ;
+	$(CXX) $(CFLAGS) $(LDFLAGS) -o $@ $(OFILES) $(DMALLOCLIBS) $(LIBS) ;
