--- src/Makefile.in	2020-05-29 15:25:18.824280442 +0200
+++ src/Makefile.in	2020-05-29 15:25:39.028321238 +0200
@@ -28,7 +28,7 @@
 #######################################################
 MIGRATEVERSION    = 4.4.3
 MIGRATESUBVERSION = March-21-2019
-MIGRATEGITVERSION = $(shell . showgit.sh)
+MIGRATEGITVERSION = ${MIGRATEVERSION}
 #######################################################
 prefix            = /usr/local
 exec_prefix       = ${prefix}
