diff -u -ur ncbi-vdb-2.10.8/py_vdb/vdb.py ncbi-vdb-2.10.8/py_vdb/vdb.py
--- ncbi-vdb-2.10.8/py_vdb/vdb.py 	2020-10-02 10:28:07.023792722 -0700
+++ ncbi-vdb-2.10.8/py_vdb/vdb.py	2020-10-02 10:04:33.701107637 -0700
@@ -1824,7 +1824,8 @@
         else :
             res = path
     else :
-        res = make_lib_name( mode )
+        lib_path = os.path.join(os.path.sep, os.environ['CONDA_PREFIX'], 'lib64')
+        res = make_lib_name( mode, lib_path )
         if res == None :
             raise vdb_error( 0, "cannot load lib: unknow platform", None )
         else :
