diff --git base/Makefile base/Makefile
index 6451479..15e2efa 100644
--- base/Makefile
+++ base/Makefile
@@ -3,7 +3,7 @@ LIB_DBG = lib-dbg-base.a
 BASE = Argument Exception IO OrderedMap Regex TypeConversion Utils Logger RangeList SimpleMatrix
 OBJ = $(BASE:=.o)

-DEFAULT_CXXFLAGS = -I. -I../third/tabix -I../third/pcre/include
+#DEFAULT_CXXFLAGS = -I. -I../third/tabix -I../third/pcre/include

 .PHONY: all release debug
 all: release
