diff --git base/Makefile base/Makefile
index 6451479..37dea81 100644
--- base/Makefile
+++ base/Makefile
@@ -10,10 +10,10 @@ all: release

 # 'make release' will: clean current build; build with -O4 flags; backup executable files to release/ folder
 # NOTE: don't use -j flag!
-release: CXXFLAGS = -O4 $(DEFAULT_CXXFLAGS)
+release: CXXFLAGS += -O4 $(DEFAULT_CXXFLAGS)
 release: $(LIB)

-debug: CXXFLAGS = -Wall -ggdb -O0 $(DEFAULT_CXXFLAGS)
+debug: CXXFLAGS += -Wall -ggdb -O0 $(DEFAULT_CXXFLAGS)
 debug: $(LIB_DBG)

 $(LIB): $(OBJ)
