diff --git Makefile Makefile
index 9d00318..b0061fa 100755
--- Makefile
+++ Makefile
@@ -6,7 +6,7 @@ SUBDIRS=core src base libVcf third
 
 
 all: $(TARGET)
-       (cd third; make)
+       #(cd third; make)
        (cd base; make)
        (cd libVcf; make)
        (cd core; make)
@@ -17,7 +17,7 @@ all: $(TARGET)
        @echo "Type ./$(TARGET) to invoke the command and see the command line options.\n"
 
 noopenmp: $(TARGET)
-       (cd third; make)
+       #(cd third; make)
        (cd base; make)
        (cd libVcf; make)
        (cd core; make)

