diff --git a/CMakeLists.txt b/CMakeLists.txt
index 0ead7e4..d7143e0 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -339,7 +339,7 @@ if (htslib_LOCAL)
   ExternalProject_Add(htslib-EXT
     SOURCE_DIR "${CMAKE_SOURCE_DIR}/${htslib_LOCAL}"
     UPDATE_COMMAND autoreconf -if
-    CONFIGURE_COMMAND ./configure --disable-s3 CC=${CC}
+    CONFIGURE_COMMAND ./configure --disable-s3
     INSTALL_COMMAND ""
     BUILD_IN_SOURCE ON
     BUILD_COMMAND $(MAKE) CFLAGS=${flags} lib-static
