diff --git a/CMakeLists.txt b/CMakeLists.txt
index e4edb6e..f0f85b1 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -335,11 +335,11 @@ if (htslib_LOCAL)
 
   include_directories(${htslib_LOCAL})
 
-  set(flags "-O3 -g -fPIC")
+  set(flags "-O3 -g -pthread -fPIC")
   ExternalProject_Add(htslib-EXT
     SOURCE_DIR "${CMAKE_SOURCE_DIR}/${htslib_LOCAL}"
-    UPDATE_COMMAND autoreconf -if
-    CONFIGURE_COMMAND ./configure --disable-s3 CC=${CC}
+    UPDATE_COMMAND cd htscodecs && autoupdate && autoreconf -i && cd ../ && autoupdate && autoreconf -i
+    CONFIGURE_COMMAND ./configure --disable-s3
     INSTALL_COMMAND ""
     BUILD_IN_SOURCE ON
     BUILD_COMMAND $(MAKE) CFLAGS=${flags} lib-static
