--- a/contrib/tabixpp/tabix.hpp
+++ b/contrib/tabixpp/tabix.hpp
@@ -3,10 +3,10 @@
 #include <string>
 #include <stdlib.h>
 #include <sys/stat.h>
-#include "htslib/bgzf.h"
-#include "htslib/tbx.h"
-#include "htslib/kseq.h"
-#include "htslib/hfile.h"
+#include "htslib/htslib/bgzf.h"
+#include "htslib/htslib/tbx.h"
+#include "htslib/htslib/kseq.h"
+#include "htslib/htslib/hfile.h"
 #include <iostream>
 #include <cstring>
 #include <vector>
