diff --git a/CMakeLists.txt b/CMakeLists.txt
index 088605d..286c9fc 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -445,6 +445,8 @@ if (NOT BUILD_ONLY_LIB)
   target_link_libraries(vcfparsealts PUBLIC ${WFALIB})
   target_link_libraries(vcfroc PUBLIC ${WFALIB})
   target_link_libraries(vcfstats PUBLIC ${WFALIB})
+  target_link_libraries(vcfstreamsort PUBLIC ${WFALIB})
+  target_link_libraries(vcfuniq PUBLIC ${WFALIB})
   target_link_libraries(vcfwave PUBLIC ${WFALIB})
   install(TARGETS ${BINS} RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR})
