diff --git a/CMakeLists.txt b/CMakeLists.txt
index 15e52a5f0b..e10ff8d046 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -462,6 +462,7 @@ if (NOT BUILD_ONLY_LIB)
   target_link_libraries(vcfallelicprimitives PUBLIC ${WFALIB})
   target_link_libraries(vcfcleancomplex PUBLIC ${WFALIB})
   target_link_libraries(vcfparsealts PUBLIC ${WFALIB})
+  target_link_libraries(vcfstreamsort PUBLIC ${WFALIB})
   target_link_libraries(vcfroc PUBLIC ${WFALIB})
   target_link_libraries(vcfstats PUBLIC ${WFALIB})
   target_link_libraries(vcfwave PUBLIC ${WFALIB})
