diff --git a/CMakeLists.txt b/CMakeLists.txt
index 088605d..286c9fc 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -462,6 +462,15 @@ if (NOT BUILD_ONLY_LIB)
   target_link_libraries(vcfallelicprimitives PUBLIC ${WFALIB})
   target_link_libraries(vcfcleancomplex PUBLIC ${WFALIB})
   target_link_libraries(vcfparsealts PUBLIC ${WFALIB})
+  target_link_libraries(vcfstreamsort PUBLIC ${WFALIB})
+  target_link_libraries(vcfuniq PUBLIC ${WFALIB})
+  target_link_libraries(vcfcreatemulti PUBLIC ${WFALIB})
+  target_link_libraries(vcfld PUBLIC ${WFALIB})
+  target_link_libraries(vcfecho PUBLIC ${WFALIB})
+  target_link_libraries(vcfuniqalleles PUBLIC ${WFALIB})
+  target_link_libraries(vcffilter PUBLIC ${WFALIB})
+  target_link_libraries(permuteGPAT++ PUBLIC ${WFALIB})
+  target_link_libraries(vcfintersect PUBLIC ${WFALIB})
   target_link_libraries(vcfroc PUBLIC ${WFALIB})
   target_link_libraries(vcfstats PUBLIC ${WFALIB})
   target_link_libraries(vcfwave PUBLIC ${WFALIB})
