--- mod_seqfiles.py	2005-06-09 09:58:54.000000000 +0000
+++ mod_seqfiles.py.new	2018-09-26 16:49:54.414426921 +0000
@@ -1,4 +1,4 @@
-#!/usr/local/python/bin/python
+#!/usr/bin/env python
 
 #    Copyright 2002,2003 Rasmus Wernersson, Technical University of Denmark
 #
--- mod_translate.py	2005-06-09 09:58:54.000000000 +0000
+++ mod_translate.py.new	2018-09-26 16:50:41.171416216 +0000
@@ -1,4 +1,4 @@
-#!/usr/local/python/bin/python
+#!/usr/bin/env python
 
 #    Copyright 2002,2003,2004,2005 Rasmus Wernersson, Technical University of Denmark
 #
--- revtrans.py	2005-06-09 09:58:54.000000000 +0000
+++ revtrans.py.new	2018-09-26 16:51:29.217405205 +0000
@@ -1,4 +1,4 @@
-#!/usr/local/python/bin/python
+#!/usr/bin/env python
 
 #
 # $Id: revtrans.py,v 1.10 2005/06/09 09:58:54 raz Exp $
--- translate.py	2005-06-09 09:58:54.000000000 +0000
+++ translate.py.new	2018-09-26 16:51:57.975398618 +0000
@@ -1,4 +1,4 @@
-#!/usr/local/python/bin/python
+#!/usr/bin/env python
 
 import sys,mod_seqfiles,mod_translate
 
