diff -u ./fixMAFnames.pl ../mugsy_x86-64-v1r2.3.new/fixMAFnames.pl
--- ./fixMAFnames.pl	2011-12-21 20:23:48.000000000 +0000
+++ ../mugsy_x86-64-v1r2.3.new/fixMAFnames.pl	2015-12-01 21:00:37.779712920 +0000
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use strict;
 
diff -u ./maf2fasta.pl ../mugsy_x86-64-v1r2.3.new/maf2fasta.pl
--- ./maf2fasta.pl	2011-12-21 20:23:48.000000000 +0000
+++ ../mugsy_x86-64-v1r2.3.new/maf2fasta.pl	2015-12-01 21:00:37.779712920 +0000
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 #Convert MAF to FASTA
 #Optionally only convert blocks that contain label 
 #./maf2fasta.pl [label] < maf > fasta
diff -u ./mugsy ../mugsy_x86-64-v1r2.3.new/mugsy
--- ./mugsy	2011-12-21 20:23:48.000000000 +0000
+++ ../mugsy_x86-64-v1r2.3.new/mugsy	2015-12-01 21:00:37.779712920 +0000
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 if(! -d $ENV{'MUGSY_INSTALL'}){
     my $default_install = "/usr/local/projects/angiuoli/mugsy_trunk/";
     if( -d $default_install){
Common subdirectories: ./MUMmer3.20 and ../mugsy_x86-64-v1r2.3.new/MUMmer3.20
diff -u ./plot.pl ../mugsy_x86-64-v1r2.3.new/plot.pl
--- ./plot.pl	2011-12-21 20:23:48.000000000 +0000
+++ ../mugsy_x86-64-v1r2.3.new/plot.pl	2015-12-01 21:00:37.779712920 +0000
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 #./plot.pl outputprefix reforganismname ps,xll
 #Eg. cat /tmp/testfilter.maf | ./plot.pl /tmp/testfilter genome2 mugsy.out > out.gp
diff -u ./splitmaf.pl ../mugsy_x86-64-v1r2.3.new/splitmaf.pl
--- ./splitmaf.pl	2011-12-21 20:23:48.000000000 +0000
+++ ../mugsy_x86-64-v1r2.3.new/splitmaf.pl	2015-12-01 21:00:37.779712920 +0000
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 #Accepts pairwise maf only
 #./splitmaf.pl outputprefix < input.maf
diff -u ./xmfa2maf.pl ../mugsy_x86-64-v1r2.3.new/xmfa2maf.pl
--- ./xmfa2maf.pl	2011-12-21 20:23:48.000000000 +0000
+++ ../mugsy_x86-64-v1r2.3.new/xmfa2maf.pl	2015-12-01 21:00:37.779712920 +0000
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 #Utility for converting output of Mauve XMFA to MAF format
 
 #USAGE: ./xmfa2maf seqs.len < aln.xmfa > aln.maf
