diff --git a/include/geometry/abstract_cubic_spline_interp.hpp b/include/geometry/abstract_cubic_spline_interp.hpp
index ba28ecb..3c20084 100644
--- a/include/geometry/abstract_cubic_spline_interp.hpp
+++ b/include/geometry/abstract_cubic_spline_interp.hpp
@@ -27,7 +27,7 @@
 
 #include <vector>
 
-#include <gromacs/utility/vec.h>
+#include <gromacs/math/vec.h>
 
 #include "geometry/basis_spline.hpp"
 
diff --git a/include/geometry/abstract_spline_curve.hpp b/include/geometry/abstract_spline_curve.hpp
index f7b0858..33163b5 100644
--- a/include/geometry/abstract_spline_curve.hpp
+++ b/include/geometry/abstract_spline_curve.hpp
@@ -27,7 +27,7 @@
 
 #include <vector>
 
-#include <gromacs/utility/vec.h>
+#include <gromacs/math/vec.h>
 
 #include "geometry/bspline_basis_set.hpp"
 
diff --git a/include/geometry/spline_curve_1D.hpp b/include/geometry/spline_curve_1D.hpp
index 7f63033..9ec9e79 100644
--- a/include/geometry/spline_curve_1D.hpp
+++ b/include/geometry/spline_curve_1D.hpp
@@ -30,7 +30,7 @@
 
 #include <gtest/gtest_prod.h>   
 
-#include <gromacs/utility/vec.h>
+#include <gromacs/math/vec.h>
 
 #include "geometry/abstract_spline_curve.hpp"
 
diff --git a/include/geometry/spline_curve_3D.hpp b/include/geometry/spline_curve_3D.hpp
index e01e33f..1c13cc7 100644
--- a/include/geometry/spline_curve_3D.hpp
+++ b/include/geometry/spline_curve_3D.hpp
@@ -30,7 +30,7 @@
 #include <gtest/gtest_prod.h>   
 
 #include <gromacs/utility/real.h> 
-#include <gromacs/utility/vec.h>
+#include <gromacs/math/vec.h>
 
 #include "geometry/abstract_spline_curve.hpp"
 
diff --git a/include/gromacs/math/3dtransforms.h b/include/gromacs/math/3dtransforms.h
index 4446fbd..70e386a 100644
--- a/include/gromacs/math/3dtransforms.h
+++ b/include/gromacs/math/3dtransforms.h
@@ -37,7 +37,7 @@
 #include <cstdio>
 
 #include "gromacs/utility/real.h"
-#include "gromacs/utility/vectypes.h"
+#include "gromacs/math/vectypes.h"
 
 /** Index for the fourth dimension for `vec4`. */
 #define WW 3
diff --git a/include/io/colour.hpp b/include/io/colour.hpp
index 8cd413e..118f380 100644
--- a/include/io/colour.hpp
+++ b/include/io/colour.hpp
@@ -29,7 +29,7 @@
 #include <string>
 #include <vector>
 
-#include <gromacs/utility/vec.h> 
+#include <gromacs/math/vec.h> 
 
 #include "external/rapidjson/document.h"
 
diff --git a/include/io/wavefront_mtl_io.hpp b/include/io/wavefront_mtl_io.hpp
index 5eb42eb..49b1919 100644
--- a/include/io/wavefront_mtl_io.hpp
+++ b/include/io/wavefront_mtl_io.hpp
@@ -29,7 +29,7 @@
 #include <string>
 #include <vector>
 
-#include <gromacs/utility/vec.h>
+#include <gromacs/math/vec.h>
 #include <gromacs/utility/real.h>
 
 
diff --git a/include/io/wavefront_obj_io.hpp b/include/io/wavefront_obj_io.hpp
index 16273c6..e294ca3 100644
--- a/include/io/wavefront_obj_io.hpp
+++ b/include/io/wavefront_obj_io.hpp
@@ -32,7 +32,7 @@
 #include <tuple>
 #include <vector>
 
-#include <gromacs/utility/vec.h>
+#include <gromacs/math/vec.h>
 #include <gromacs/utility/real.h>   
 
 
diff --git a/include/path-finding/molecular_path.hpp b/include/path-finding/molecular_path.hpp
index 0628773..15ae7e4 100644
--- a/include/path-finding/molecular_path.hpp
+++ b/include/path-finding/molecular_path.hpp
@@ -29,7 +29,7 @@
 #include <string>
 #include <vector>
 
-#include <gromacs/utility/vec.h>
+#include <gromacs/math/vec.h>
 #include <gromacs/pbcutil/pbc.h>
 #include <gromacs/utility/real.h>
 #include <gromacs/selection/selection.h>
diff --git a/include/path-finding/naive_cylindrical_path_finder.hpp b/include/path-finding/naive_cylindrical_path_finder.hpp
index 3aeba51..daf4d7b 100644
--- a/include/path-finding/naive_cylindrical_path_finder.hpp
+++ b/include/path-finding/naive_cylindrical_path_finder.hpp
@@ -26,7 +26,7 @@
 #define NAIVE_CYLINDRICAL_PATH_FINDER_HPP
 
 #include <gromacs/utility/real.h> 
-#include <gromacs/utility/vec.h>
+#include <gromacs/math/vec.h>
 
 #include "path-finding/abstract_path_finder.hpp"
 
diff --git a/src/io/molecular_path_obj_exporter.cpp b/src/io/molecular_path_obj_exporter.cpp
index 0dc1808..1cfb311 100644
--- a/src/io/molecular_path_obj_exporter.cpp
+++ b/src/io/molecular_path_obj_exporter.cpp
@@ -28,7 +28,7 @@
 #include <sstream>
 #include <vector>
 
-#include <gromacs/utility/vec.h>
+#include <gromacs/math/vec.h>
 #include <gromacs/utility/real.h> 
 
 #include "io/molecular_path_obj_exporter.hpp"
diff --git a/src/path-finding/inplane_optimised_probe_path_finder.cpp b/src/path-finding/inplane_optimised_probe_path_finder.cpp
index d87b265..2f59bb4 100644
--- a/src/path-finding/inplane_optimised_probe_path_finder.cpp
+++ b/src/path-finding/inplane_optimised_probe_path_finder.cpp
@@ -25,7 +25,7 @@
 #include <iostream>
 #include <limits>
 
-#include <gromacs/utility/vec.h>
+#include <gromacs/math/vec.h>
 
 #include "optim/simulated_annealing_module.hpp"
 #include "optim/nelder_mead_module.hpp"
diff --git a/test/path-finding/ut_molecular_path.cpp b/test/path-finding/ut_molecular_path.cpp
index 183371b..ddea2c0 100644
--- a/test/path-finding/ut_molecular_path.cpp
+++ b/test/path-finding/ut_molecular_path.cpp
@@ -24,7 +24,7 @@
 
 #include <gtest/gtest.h>
 
-#include <gromacs/utility/vec.h>
+#include <gromacs/math/vec.h>
 
 #include "path-finding/molecular_path.hpp"
 
