diff --git a/python3/assemblyGet.py b/python3/assemblyGet.py
index 6c1f2ac..f2dd85d 100644
--- a/python3/assemblyGet.py
+++ b/python3/assemblyGet.py
@@ -1,3 +1,4 @@
+#!/usr/bin/env python3
 #
 # assemblyGet.py
 #
diff --git a/python3/enaGroupGet.py b/python3/enaGroupGet.py
index 55a498c..c953bb2 100644
--- a/python3/enaGroupGet.py
+++ b/python3/enaGroupGet.py
@@ -1,3 +1,4 @@
+#!/usr/bin/env python3
 #
 # enaGroupGet.py
 #
diff --git a/python3/enaDataGet.py b/python3/enaDataGet.py
index ef37afd..4aa7a2d 100644
--- a/python3/enaDataGet.py
+++ b/python3/enaDataGet.py
@@ -1,3 +1,4 @@
+#!/usr/bin/env python3
 #
 # enaDataGet.py
 #
diff --git a/python3/readGet.py b/python3/readGet.py
index e63fcba..695d968 100644
--- a/python3/readGet.py
+++ b/python3/readGet.py
@@ -1,3 +1,4 @@
+#!/usr/bin/env python3
 #
 # readGet.py
 #
diff --git a/python3/sequenceGet.py b/python3/sequenceGet.py
index b6b6a4e..37acfde 100644
--- a/python3/sequenceGet.py
+++ b/python3/sequenceGet.py
@@ -1,3 +1,4 @@
+#!/usr/bin/env python3
 #
 # sequenceGet.py
 #
diff --git a/python3/utils.py b/python3/utils.py
index 5746e5f..573f8b7 100644
--- a/python3/utils.py
+++ b/python3/utils.py
@@ -1,3 +1,4 @@
+#!/usr/bin/env python3
 #
 # utils.py
 #
