--- python3/assemblyGet.py.orig	2018-12-22 21:32:36.841398257 +0200
+++ python3/assemblyGet.py	2018-12-22 21:50:43.884719723 +0200
@@ -17,13 +17,17 @@
 # limitations under the License.
 #
 
+from __future__ import print_function, division
 import os
 import sys
 import argparse
 import gzip
 import xml.etree.ElementTree as ElementTree
 
-import utils
+if sys.version_info.major == 2: 
+    import utils_py2
+else:
+    import utils
 import sequenceGet
 
 REPLICON = 'assembled-molecule'
