--- python3/enaDataGet.py.orig	2018-12-22 21:33:38.950048122 +0200
+++ python3/enaDataGet.py	2018-12-22 21:51:18.592843498 +0200
@@ -17,6 +17,7 @@
 # limitations under the License.
 #
 
+from __future__ import print_function, division
 import argparse
 import os
 import sys
@@ -24,7 +25,10 @@
 import sequenceGet
 import assemblyGet
 import readGet
-import utils
+if sys.version_info.major == 2: 
+    import utils_py2
+else:
+    import utils
 import traceback
 
 def set_parser():
