--- python3/enaGroupGet.py.orig	2018-12-22 21:33:55.862224436 +0200
+++ python3/enaGroupGet.py	2018-12-22 21:52:07.105030148 +0200
@@ -17,6 +17,7 @@
 # limitations under the License.
 #
 
+from __future__ import print_function, division
 import argparse
 import os
 import sys
@@ -24,7 +25,10 @@
 import sequenceGet
 import assemblyGet
 import readGet
-import utils
+if sys.version_info.major == 2: 
+    import utils_py2
+else:
+    import utils
 import traceback
 
 def set_parser():
@@ -113,7 +117,7 @@
                 if data_accession not in update_accs:
                     write_record = True
             if write_record:
-                sequenceGet.write_record(dest_file, data_accession, output_format)
+                sequenceGet.write_record(dest_file, data_accession, output_format, expanded)
                 dest_file.flush()
     os.remove(temp_file_path)
     return update_accs
