--- python3/readGet.py.orig	2018-12-22 21:39:59.792915317 +0200
+++ python3/readGet.py	2018-12-22 21:52:13.097054244 +0200
@@ -17,11 +17,15 @@
 # limitations under the License.
 #
 
+from __future__ import print_function, division
 import os
 import sys
 import argparse
 
-import utils
+if sys.version_info.major == 2: 
+    import utils_py2
+else:
+    import utils
 
 def check_read_format(output_format):
     if output_format not in [utils.SUBMITTED_FORMAT, utils.FASTQ_FORMAT, utils.SRA_FORMAT]:
@@ -118,6 +122,6 @@
         for index_file in indexlist:
             if index_file != '':
                 download_file(index_file, target_dir, None, aspera)
-    if utils.is_empty_dir(target_dir):
-        print('Deleting directory ' + os.path.basename(target_dir))
-        os.rmdir(target_dir)
+        if utils.is_empty_dir(target_dir):
+            print('Deleting directory ' + os.path.basename(target_dir))
+            os.rmdir(target_dir)
