diff --git a/fermi-lite/Makefile b/fermi-lite/Makefile
index 53255ba..bf6c3c9 100644
--- a/fermi-lite/Makefile
+++ b/fermi-lite/Makefile
@@ -1,5 +1,5 @@
-CC=			gcc
-CFLAGS=		-g -Wall -O2 -Wno-unused-function #-fno-inline-functions -fno-inline-functions-called-once
+CC=			$(CC)
+CFLAGS=		-g -Wall -O3 -Wno-unused-function #-fno-inline-functions -fno-inline-functions-called-once
 CPPFLAGS=
 INCLUDES=
 OBJS=		kthread.o misc.o \
