diff --git a/fermi-lite/Makefile b/fermi-lite/Makefile
index 53255ba..2dfc5c8 100644
--- a/fermi-lite/Makefile
+++ b/fermi-lite/Makefile
@@ -1,7 +1,7 @@
-CC=			gcc
-CFLAGS=		-g -Wall -O2 -Wno-unused-function #-fno-inline-functions -fno-inline-functions-called-once
-CPPFLAGS=
-INCLUDES=
+CC=			$(CC)
+CFLAGS=		-g -Wall -O3 -Wno-unused-function #-fno-inline-functions -fno-inline-functions-called-once
+CPPFLAGS=	-I$(PREFIX)/include
+INCLUDES=	
 OBJS=		kthread.o misc.o \
 			bseq.o htab.o bfc.o \
 			rle.o rope.o mrope.o rld0.o \
