--- ncbi-igblast-1.14.0-src/c++/src/app/igblast/Makefile.igblastn.app	2017-12-18 22:24:39.000000000 +0100
+++ ncbi-igblast-1.14.0-src-patched/c++/src/app/igblast/Makefile.igblastn.app	2019-06-26 16:19:39.408248385 +0200
@@ -11,7 +11,7 @@
 CXXFLAGS = $(FAST_CXXFLAGS:ppc=i386)
 LDFLAGS  = $(FAST_LDFLAGS:ppc=i386)
 
-LIBS = $(VDB_STATIC_LIBS) $(CMPRS_LIBS) $(DL_LIBS) $(NETWORK_LIBS) $(BLAST_THIRD_PARTY_LIBS) $(ORIG_LIBS)
+LIBS = $(VDB_LIBS) $(CMPRS_LIBS) $(DL_LIBS) $(NETWORK_LIBS) $(BLAST_THIRD_PARTY_LIBS) $(ORIG_LIBS)
 
 REQUIRES = VDB objects -Cygwin
 
