--- ncbi-igblast-1.22.0-src/c++/src/app/igblast/Makefile.igblastn.app	2023-10-06 15:33:05.000000000 -0400
+++ ncbi-igblast-1.22.0-src-patched/c++/src/app/igblast/Makefile.igblastn.app	2024-01-10 11:35:41.231820985 -0500
@@ -12,7 +12,7 @@
 LDFLAGS  = $(FAST_LDFLAGS:ppc=i386)
 
 LIBS = $(BLAST_THIRD_PARTY_LIBS) $(GENBANK_THIRD_PARTY_LIBS) \
-       $(VDB_STATIC_LIBS) $(CMPRS_LIBS) $(DL_LIBS) $(NETWORK_LIBS) $(ORIG_LIBS)
+       $(VDB_LIBS) $(CMPRS_LIBS) $(DL_LIBS) $(NETWORK_LIBS) $(ORIG_LIBS)
 
 REQUIRES = VDB objects -Cygwin
 
