diff -u -ru ncbi-igblast-1.15.0-src/c++/src/app/igblast/Makefile.igblastn.app ncbi-igblast-1.15.0-src-patched/c++/src/app/igblast/Makefile.igblastn.app
--- ncbi-igblast-1.15.0-src/c++/src/app/igblast/Makefile.igblastn.app	2019-08-04 23:12:21.000000000 +0200
+++ ncbi-igblast-1.15.0-src-patched/c++/src/app/igblast/Makefile.igblastn.app	2020-02-19 14:30:43.571473876 +0100
@@ -11,7 +11,7 @@
 CXXFLAGS = $(FAST_CXXFLAGS:ppc=i386)
 LDFLAGS  = $(FAST_LDFLAGS:ppc=i386)
 
-LIBS = $(GENBANK_THIRD_PARTY_LIBS) $(VDB_STATIC_LIBS) $(CMPRS_LIBS) $(DL_LIBS) $(NETWORK_LIBS) $(BLAST_THIRD_PARTY_LIBS) $(ORIG_LIBS)
+LIBS = $(GENBANK_THIRD_PARTY_LIBS) $(VDB_LIBS) $(CMPRS_LIBS) $(DL_LIBS) $(NETWORK_LIBS) $(BLAST_THIRD_PARTY_LIBS) $(ORIG_LIBS)
 
 REQUIRES = VDB objects -Cygwin
 
