--- pasta/__init__.py	2023-12-04 08:02:14.754964843 +0100
+++ pasta/__init__.condafixed.py	2023-12-04 09:00:09.426457135 +0100
@@ -20,7 +20,7 @@
 PROGRAM_LONG_DESCRIPTION = """
 PASTA performs iterative realignment and tree inference, similar to SATe, but uses a very different merge algorithm which improves running time, memory usage, and accuracy. The current code is heavily based on SATe, with lots of modifications, many related to algorithmic differences between PASTA and SATe, but also many scalability improvements (parallelization, tree parsing, defaults, etc.)
 
-Minimally you must provide a sequence file (with the '--input' option); a starting tree is optional. By default, important algorithmic parameters are set based on automatic rules. 
+Minimally you must provide a sequence file (with the '--input' option); a starting tree is optional. By default, important algorithmic parameters are set based on automatic rules.
 
 The command line allows you to alter the behavior of the algorithm (termination criteria, when the algorithm switches to "Blind" acceptance of new alignments, how the tree is decomposed to find subproblems to be used, and the external tools to use).
 
@@ -269,6 +269,8 @@
     Otherwise, '../../sate-tools-linux', '../../sate-tools-mac', or
     '../../sate-tools-win' is used (path relative to this __file__).
     """
+    return os.path.abspath(os.environ['CONDA_PREFIX'])  # injected via conda patching process
+
     if "PASTA_TOOLS_DEVDIR" in os.environ and platform_name is None:
         bin_path = os.path.expanduser(os.path.expandvars(os.environ["PASTA_TOOLS_DEVDIR"]))
     else:
@@ -288,4 +290,3 @@
         raise ToolsDirNotFoundError(paths_tried=[bin_path],
                 env_var_name="PASTA_TOOLS_DEVDIR")
     return bin_path
-
