--- a/run_pasta.py	2023-12-04 14:25:42.341227081 +0100
+++ b/run_pasta.mpstart.py	2023-12-04 14:25:34.944972464 +0100
@@ -22,6 +22,9 @@
 # Jiaye Yu and Mark Holder, University of Kansas
 
 if __name__ == "__main__":
+    import multiprocessing as mp
+    mp.set_start_method('fork')  # how to create child processes. Does not work with default in OSX py38
+
     import os
     import sys
     import pasta
