diff --git a/Makefile b/Makefile
index 5a1ba0c..6c9e9f4 100644
--- a/Makefile
+++ b/Makefile
@@ -26,13 +26,13 @@ NAME = Distributed Pumpkin
 MPICXX = mpicxx
 
 # CXXFLAGS can be changed by the end user with make CXXFLAGS="..."
-CXXFLAGS = -O3 -std=c++98 -Wall -g
+CXXFLAGS = -O3 -std=c++98 -march=x86-64-v3 -Wall -g
 
 RM = rm
 CD = cd
 MAKE = make
 ECHO = echo
-MKDIR = mkdir
+MKDIR = mkdir -p
 CP = cp
 
 CONFIG_RAY_VERSION = $(VERSION).$(PATCHLEVEL).$(SUBLEVEL)$(EXTRAVERSION)
