--- userApps/kent/src/lib/htmshell.c.old	2022-10-28 08:32:08.344445121 +0000
+++ userApps/kent/src/lib/htmshell.c	2022-10-28 08:32:31.708262420 +0000
@@ -711,11 +711,11 @@
 puts("Status: 400\r");
 puts("Content-Type: text/plain; charset=UTF-8\r");
 puts("\r");
-if (format != NULL && args != NULL)
-    {
-    vfprintf(stdout, format, args);
-    fprintf(stdout, "\n");
-    }
+if (format != NULL)
+    {
+    vfprintf(stdout, format, args);
+    fprintf(stdout, "\n");
+    }
 exit(-1);
 }
 
