--- hapsolo.py.old	2023-01-18 12:21:49.819445632 -0600
+++ hapsolo.py	2023-01-18 12:23:03.919747316 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/opt/anaconda1anaconda2anaconda3/bin/python
 # -*- coding: utf-8 -*-
 """
 @author: Edwin
--- preprocessfasta.py.old	2023-01-18 12:24:31.652104498 -0600
+++ preprocessfasta.py	2023-01-18 12:24:39.732137397 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/opt/anaconda1anaconda2anaconda3/bin/python
 # -*- coding: utf-8 -*-
 """
 @author: Edwin
