--- scripts/bash_buscopreprocess.sh.old	2023-01-18 13:31:50.098495020 -0600
+++ scripts/bash_buscopreprocess.sh	2023-01-18 13:32:00.632537587 -0600
@@ -3,7 +3,7 @@
 REF=YOURASM.FASTA
 
 # make sure to start from your root working directory that contains your assmebly
-python preprocessfasta.py -i ${REF} -m 0
+preprocessfasta.py -i ${REF} -m 0
 
 cd contigs
 # make sure to run in ${WORKDIR}/contigs/ directory
--- scripts/bash_gnuparallelblat.sh.old	2023-01-18 13:14:31.430293417 -0600
+++ scripts/bash_gnuparallelblat.sh	2023-01-18 13:15:53.752626635 -0600
@@ -1,7 +1,7 @@
 #!/bin/bash
 # Please also cite GNU Parallel if you use this script
 
-module load blat
+#module load blat
 
 ### Set number of cores available here
 CORES=48
--- scripts/bash_quastbusco.sh.old	2023-01-18 13:38:41.503157381 -0600
+++ scripts/bash_quastbusco.sh	2023-01-18 13:39:05.599254737 -0600
@@ -3,15 +3,15 @@
 #run example: bash bash_quastbusco.sh MYFASTA.FASTA JOBID
 
 #please load your Quast, BUSCO, Augustus, BRAKER binaries here
-source /networkshare/.mybashrc
-export AUGUSTUS_CONFIG_PATH="/networkshare/bin/augustus-3.2.2/config"
+#source /networkshare/.mybashrc
+#export AUGUSTUS_CONFIG_PATH="/networkshare/bin/augustus-3.2.2/config"
 
 #please copy to contigs directory and run in that directory
 #please run bash_buscopreprocess.sh prior to running this
 
 INPUTTYPE="geno"
 #please enter the directory for your ODB9 libraries here
-MYLIBDIR="/networkshare/bin/busco/lineages/"
+#MYLIBDIR="/networkshare/bin/busco/lineages/"
 
 #drosophila
 MYLIB="diptera_odb9"
--- scripts/qsub_blat.sh.old	2023-01-18 13:14:38.203320834 -0600
+++ scripts/qsub_blat.sh	2023-01-18 13:16:11.066696727 -0600
@@ -6,7 +6,7 @@
 #$ -ckpt blcr
 #$ -q bigmemory
 
-module load blat
+#module load blat
 #PATH=/networkshare/bin/blat:$PATH
 
 REF=primary_new.fasta
--- scripts/qsub_preprocess.sh.old	2023-01-18 12:58:22.520364243 -0600
+++ scripts/qsub_preprocess.sh	2023-01-18 13:16:23.187745790 -0600
@@ -6,14 +6,14 @@
 #$ -ckpt blcr
 #$ -q myqueue
 
-module load blat
+#module load blat
 
 REF=myassemblyfile.fasta
 
 #create this file: faToTwoBit ${REF} $(basename ${REF} .fasta).2bit
 faToTwoBit ${REF} $(basename ${REF} .fasta).2bit
 
-python preprocessfasta.py -i ${REF}
+preprocessfasta.py -i ${REF}
 
 JOBFILE=jobfile.txt
 ls contigs/*.fasta > ${JOBFILE}
--- scripts/sbatch_blat.sh.old	2023-01-18 13:16:35.726796541 -0600
+++ scripts/sbatch_blat.sh	2023-01-18 13:16:46.161838782 -0600
@@ -13,7 +13,7 @@
 ###SBATCH --mail-type=begin	# email me when the job starts
 ###SBATCH --mail-type=end	# email me when the job finishes
 
-module load blat
+#module load blat
 
 REF=myfastafile.fasta
 
--- scripts/sbatch_busco.sh.old	2023-01-18 13:40:32.613606340 -0600
+++ scripts/sbatch_busco.sh	2023-01-18 13:40:45.552658628 -0600
@@ -13,15 +13,15 @@
 ###SBATCH --mail-type=end	# email me when the job finishes
 
 #please load your BUSCO, Augustus, BRAKER binaries here
-source /networkshare/.mybashrc
-export AUGUSTUS_CONFIG_PATH="/networkshare/bin/augustus-3.2.2/config"
+# source /networkshare/.mybashrc
+# export AUGUSTUS_CONFIG_PATH="/networkshare/bin/augustus-3.2.2/config"
 
 #please copy to contigs directory and run in that directory
 #please run bash_buscopreprocess.sh prior to running this
 
 INPUTTYPE="geno"
 #please enter the directory for your ODB9 libraries here
-MYLIBDIR="/networkshare/bin/busco/lineages/"
+# MYLIBDIR="/networkshare/bin/busco/lineages/"
 
 #drosophila
 MYLIB="diptera_odb9"
--- scripts/sbatch_preprocess.sh.old	2023-01-18 12:58:37.318424321 -0600
+++ scripts/sbatch_preprocess.sh	2023-01-18 13:16:58.781889855 -0600
@@ -15,14 +15,14 @@
 # This is the variable that has the total number of cores committed to this job. In this case it is 1
 #echo $SLURM_CPUS_PER_TASK
 
-module load blat
+#module load blat
 
 REF=myassemblyfile.fasta
 
 #create this file: faToTwoBit ${REF} $(basename ${REF} .fasta).2bit
 faToTwoBit ${REF} $(basename ${REF} .fasta).2bit
 
-python preprocessfasta.py -i ${REF}
+preprocessfasta.py -i ${REF}
 
 JOBFILE=jobfile.txt
 ls contigs/*.fasta > ${JOBFILE}
