--- a/fatslimlib/command.py
+++ b/fatslimlib/command.py
@@ -890,9 +890,9 @@
     def prepare_results(self, num_frames):
         super(MembraneAnalysisCommand, self).prepare_results(num_frames)
 
-        self.membrane_nlipids = np.zeros(num_frames, dtype=np.int)
-        self.leaflet1_nlipids = np.zeros(num_frames, dtype=np.int)
-        self.leaflet2_nlipids = np.zeros(num_frames, dtype=np.int)
+        self.membrane_nlipids = np.zeros(num_frames, dtype=np.int64)
+        self.leaflet1_nlipids = np.zeros(num_frames, dtype=np.int64)
+        self.leaflet2_nlipids = np.zeros(num_frames, dtype=np.int64)
 
         for prop in self.properties:
             prop.prepare(num_frames)
