--- primer_core_tools.py	2017-08-11 04:28:32.000000000 -0500
+++ primer_core_tools.py.new	2019-11-26 20:29:24.064830724 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/opt/anaconda1anaconda2anaconda3/bin/python
 ''' RUCS - Rapid Identification of PCR Primers Pairs for Unique Core Sequences
 
 
@@ -2222,7 +2222,7 @@
    return kmers
 
 ############ FUNCTIONS (FOR PCR PRIMER IDENTIFICATION)
-def load_global_settings(settings_file='settings.default.cjson'):
+def load_global_settings(settings_file=str(os.getenv("SETTINGS_FILE"))):
    ''' This method is run at compile time, and will initialise global
    dependencies '''
    global settings
@@ -4171,7 +4171,7 @@
                        help=("File containing pcr primer pair sets (1 pair per "
                              "line, tab-separated sequences, format: forward, "
                              "reverse, probe*) *optional"))
-   parser.add_argument("--settings_file", default="settings.default.cjson",
+   parser.add_argument("--settings_file", default=str(os.getenv("SETTINGS_FILE")),
                        help=("File containing the default settings for this "
                              "program"))
    # Settings-modification options
