diff --git a/seroba/kmc.py b/seroba/kmc.py
index c9510c3..de79fb3 100644
--- a/seroba/kmc.py
+++ b/seroba/kmc.py
@@ -10,9 +10,9 @@ def run_kmc(sequence_file,kmer_size,temp_dir,prefix):
     file_format = common.detect_sequence_format(sequence_file)
     kmer_count_files = os.path.join(temp_dir,prefix)
     if file_format == 'fa':
-        param = ' -ci1 -m1 -t1 -fm'
+        param = ' -ci1 -m2 -t1 -fm'
     elif file_format =='fq':
-        param = ' -ci4  -m1 -t1'
+        param = ' -ci4  -m2 -t1'
     print(param)
     command1=[ext_progs.exe('kmc') + ' -k'+kmer_size,param,sequence_file,kmer_count_files,temp_dir]
     print(' '.join(command1))
