diff -rNu arbsrc_17491.orig/Makefile arbsrc_17491/Makefile
--- arbsrc_17491.orig/Makefile	2018-10-17 04:22:41.000000000 -0600
+++ arbsrc_17491/Makefile	2019-02-16 22:03:51.493258562 -0700
@@ -846,7 +846,7 @@
  endif
 endif
 
-LINK_STATIC_LIB := ar -csq# link static lib
+LINK_STATIC_LIB := $(AR) -csq# link static lib
 LINK_EXECUTABLE := $(A_CXX) $(blflags) -o# link executable (c++)
 
 ifeq ($(LINK_STATIC),1)
