diff -rNu arbsrc_17491.orig/TEMPLATES/cxxforward.h arbsrc_17491/TEMPLATES/cxxforward.h
--- arbsrc_17491.orig/TEMPLATES/cxxforward.h	2018-04-13 02:36:29.000000000 -0600
+++ arbsrc_17491/TEMPLATES/cxxforward.h	2019-02-13 20:30:39.015226745 -0700
@@ -17,13 +17,12 @@
 #include "gccver.h"
 #endif
 #if defined(__cplusplus)
-# if (GCC_VERSION_CODE >= 407)
 #  if (__cplusplus == 199711L)
 #  else
 #   if (__cplusplus == 201103L)
 #    define ARB_ENABLE11_FEATURES
 #   else
-#    if (__cplusplus == 201402L)
+#    if (__cplusplus >= 201402L)
 #     define ARB_ENABLE11_FEATURES
 #     define ARB_ENABLE14_FEATURES
 #    else
@@ -31,7 +30,6 @@
 #    endif
 #   endif
 #  endif
-# endif
 #else
 # error C compilation includes cxxforward.h
 #endif
