diff --git a/Makefile b/Makefile
index d7f412b..472ced8 100644
--- a/Makefile
+++ b/Makefile
@@ -853,6 +853,9 @@ ifeq ($(LINK_STATIC),1)
  SHARED_LIB_SUFFIX = a# static lib suffix
  LINK_SHARED_LIB := $(LINK_STATIC_LIB)
 else
+ ifeq ($(USE_CLANG),1)
+  llflags += -undefined dynamic_lookup
+ endif
  SHARED_LIB_SUFFIX = so# shared lib suffix
  LINK_SHARED_LIB := $(A_CXX) $(llflags) -shared -o# link shared lib
 endif
