Only in b/src: Arc-Ful-SRP
diff -u -r a/src/BPs_to_CT.py b/src/BPs_to_CT.py
--- a/src/BPs_to_CT.py	2015-09-17 04:19:50.000000000 +0200
+++ b/src/BPs_to_CT.py	2017-06-15 21:03:34.000000000 +0200
@@ -1,3 +1,5 @@
+#!/usr/bin/env python
+
 """
     DotKnot: software for RNA pseudoknot prediction
     Copyright (C) 2010-2011 Jana Sperschneider	
Only in b/src: BPs_to_CT.pyc
diff -u -r a/src/cc06.py b/src/cc06.py
--- a/src/cc06.py	2015-09-17 04:19:50.000000000 +0200
+++ b/src/cc06.py	2017-06-15 21:03:30.000000000 +0200
@@ -1,3 +1,5 @@
+#!/usr/bin/env python
+
 """
     DotKnot: software for RNA pseudoknot prediction
     Copyright (C) 2010-2011 Jana Sperschneider	
Only in b/src: cc06.pyc
diff -u -r a/src/cc09.py b/src/cc09.py
--- a/src/cc09.py	2015-09-17 04:19:50.000000000 +0200
+++ b/src/cc09.py	2017-06-15 21:03:28.000000000 +0200
@@ -1,3 +1,5 @@
+#!/usr/bin/env python
+
 """
     DotKnot: software for RNA pseudoknot prediction
     Copyright (C) 2010-2011 Jana Sperschneider	
Only in b/src: cc09.pyc
diff -u -r a/src/dotknot.py b/src/dotknot.py
--- a/src/dotknot.py	2015-09-17 04:19:50.000000000 +0200
+++ b/src/dotknot.py	2017-07-03 17:27:06.891597855 +0200
@@ -1,3 +1,5 @@
+#!/usr/bin/env python
+
 """ 
     DotKnot: software for RNA pseudoknot prediction
     Copyright (C) 2010-2011 Jana Sperschneider	
@@ -26,7 +28,7 @@
 import BPs_to_CT
 
 # Path to RNAfold program which uses low probability threshold
-RNAFOLD_PATH = "./ViennaRNA-1.8.5/Progs/"
+RNAFOLD_PATH = "."
 
 def main(identifier, seq, khp, local, global_structure):
     #---------------------------------------------------------------------------------------------------------                
diff -u -r a/src/functions.py b/src/functions.py
--- a/src/functions.py	2015-09-17 04:19:50.000000000 +0200
+++ b/src/functions.py	2017-06-15 21:03:22.000000000 +0200
@@ -1,3 +1,5 @@
+#!/usr/bin/env python
+
 """
     DotKnot: software for RNA pseudoknot prediction
     Copyright (C) 2010-2011 Jana Sperschneider	
Only in b/src: functions.pyc
diff -u -r a/src/kissing_hairpins.py b/src/kissing_hairpins.py
--- a/src/kissing_hairpins.py	2015-09-17 04:19:50.000000000 +0200
+++ b/src/kissing_hairpins.py	2017-06-15 21:03:18.000000000 +0200
@@ -1,3 +1,5 @@
+#!/usr/bin/env python
+
 """
     DotKnot: software for RNA pseudoknot prediction
     Copyright (C) 2010-2011 Jana Sperschneider	
Only in b/src: kissing_hairpins.pyc
diff -u -r a/src/longPK.py b/src/longPK.py
--- a/src/longPK.py	2015-09-17 04:19:50.000000000 +0200
+++ b/src/longPK.py	2017-06-15 21:03:14.000000000 +0200
@@ -1,3 +1,5 @@
+#!/usr/bin/env python
+
 """ DotKnot: software for RNA pseudoknot prediction
     Copyright (C) 2010-2011 Jana Sperschneider	
 
Only in b/src: longPK.pyc
Only in b/src: MMTV
diff -u -r a/src/mwis_2d.py b/src/mwis_2d.py
--- a/src/mwis_2d.py	2015-09-17 04:19:50.000000000 +0200
+++ b/src/mwis_2d.py	2017-06-15 21:03:10.000000000 +0200
@@ -1,3 +1,5 @@
+#!/usr/bin/env python
+
 """
     DotKnot: software for RNA pseudoknot prediction
     Copyright (C) 2010-2011 Jana Sperschneider	
Only in b/src: mwis_2d.pyc
diff -u -r a/src/mwis.py b/src/mwis.py
--- a/src/mwis.py	2015-09-17 04:19:50.000000000 +0200
+++ b/src/mwis.py	2017-06-15 21:03:06.000000000 +0200
@@ -1,3 +1,5 @@
+#!/usr/bin/env python
+
 """
     DotKnot: software for RNA pseudoknot prediction
     Copyright (C) 2010-2011 Jana Sperschneider	
Only in b/src: mwis.pyc
diff -u -r a/src/pk_construction.py b/src/pk_construction.py
--- a/src/pk_construction.py	2015-09-17 04:19:50.000000000 +0200
+++ b/src/pk_construction.py	2017-06-15 21:03:02.000000000 +0200
@@ -1,3 +1,5 @@
+#!/usr/bin/env python
+
 """
     DotKnot: software for RNA pseudoknot prediction
     Copyright (C) 2010-2011 Jana Sperschneider	
Only in b/src: pk_construction.pyc
diff -u -r a/src/pk_energy.py b/src/pk_energy.py
--- a/src/pk_energy.py	2015-09-17 04:19:50.000000000 +0200
+++ b/src/pk_energy.py	2017-06-15 21:02:52.000000000 +0200
@@ -1,3 +1,5 @@
+#!/usr/bin/env python
+
 """
     DotKnot: software for RNA pseudoknot prediction
     Copyright (C) 2010-2011 Jana Sperschneider	
diff -u -r a/src/pk_interrupted.py b/src/pk_interrupted.py
--- a/src/pk_interrupted.py	2015-09-17 04:19:50.000000000 +0200
+++ b/src/pk_interrupted.py	2017-06-15 21:02:48.000000000 +0200
@@ -1,3 +1,5 @@
+#!/usr/bin/env python
+
 """ 
     DotKnot: software for RNA pseudoknot prediction
     Copyright (C) 2010-2011 Jana Sperschneider	
Only in b/src: pk_interrupted.pyc
diff -u -r a/src/pk_recursive.py b/src/pk_recursive.py
--- a/src/pk_recursive.py	2015-09-17 04:19:50.000000000 +0200
+++ b/src/pk_recursive.py	2017-06-15 21:02:42.000000000 +0200
@@ -1,3 +1,5 @@
+#!/usr/bin/env python
+
 """
     DotKnot: software for RNA pseudoknot prediction
     Copyright (C) 2010-2011 Jana Sperschneider	
Only in b/src: pk_recursive.pyc
diff -u -r a/src/pk_tools.py b/src/pk_tools.py
--- a/src/pk_tools.py	2015-09-17 04:19:50.000000000 +0200
+++ b/src/pk_tools.py	2017-06-15 21:02:38.000000000 +0200
@@ -1,3 +1,5 @@
+#!/usr/bin/env python
+
 """
     DotKnot: software for RNA pseudoknot prediction
     Copyright (C) 2010-2011 Jana Sperschneider	
Only in b/src: pk_tools.pyc
diff -u -r a/src/secondary_elements.py b/src/secondary_elements.py
--- a/src/secondary_elements.py	2015-09-17 04:19:50.000000000 +0200
+++ b/src/secondary_elements.py	2017-06-15 21:02:36.000000000 +0200
@@ -1,3 +1,5 @@
+#!/usr/bin/env python
+
 """
     DotKnot: software for RNA pseudoknot prediction
     Copyright (C) 2010-2011 Jana Sperschneider	
Only in b/src: secondary_elements.pyc
diff -u -r a/src/stems.py b/src/stems.py
--- a/src/stems.py	2015-09-17 04:19:50.000000000 +0200
+++ b/src/stems.py	2017-06-15 21:02:32.000000000 +0200
@@ -1,3 +1,5 @@
+#!/usr/bin/env python
+
 """
     DotKnot: software for RNA pseudoknot prediction
     Copyright (C) 2010-2011 Jana Sperschneider	
Only in b/src: stems.pyc
diff -u -r a/src/user.py b/src/user.py
--- a/src/user.py	2015-09-17 04:19:50.000000000 +0200
+++ b/src/user.py	2017-06-15 21:02:28.000000000 +0200
@@ -1,3 +1,5 @@
+#!/usr/bin/env python
+
 """ 
     DotKnot: software for RNA pseudoknot prediction
     Copyright (C) 2010-2011 Jana Sperschneider	
Only in b/src: user.pyc
