diff --git a/Makefile b/Makefile
index 9c7f457..6d07033 100644
--- a/Makefile
+++ b/Makefile
@@ -1,9 +1,9 @@
-CC     = gcc
+CC     = $(CC)
 AR     = ar
-CFLAGS = -g -Wall
+CFLAGS = -O3 -g -Wall -I${PREFIX}/include -L${PREFIX}/lib
 
 ifeq (1, $(CF_OPTIMIZE))
-	CFLAGS += -O2
+	CFLAGS += -O3
 	CFLAGS := $(filter-out -g, $(CFLAGS))
 endif
 
@@ -14,7 +14,7 @@ LIBUTILS_OBJS = \
 	wzhmm.o
 
 .c.o :
-	$(CC) -c $(CFLAGS) $< -o $@
+	$(CC) -c $(CFLAGS) $< -o $@ -lz
 
 libutils.a: $(LIBUTILS_OBJS)
 	@-rm -f $@
