diff --git a/grabix.cpp b/grabix.cpp
index 2712130..6bfc30a 100644
--- a/grabix.cpp
+++ b/grabix.cpp
@@ -104,10 +104,10 @@ int create_grabix_index(string bgzf_file)
     {
         // grab the next line and store the offset
         eof = bgzf_getline(bgzf_fp, '\n', line);
-        offset = bgzf_tell (bgzf_fp);
+        offset = bgzf_tell(bgzf_fp);
         chunk_count++;
         // stop if we have encountered an empty line
-        if (eof <= 0)
+        if (eof <= 0 || offset == prev_offset)
         {
             if (bgzf_check_EOF(bgzf_fp) == 1) {
                 if (offset > prev_offset) {
diff --git a/grabix.h b/grabix.h
index af94cbd..43d7daa 100644
--- a/grabix.h
+++ b/grabix.h
@@ -7,7 +7,7 @@ using namespace std;
 #include "bgzf.h"
 
 
-#define VERSION "0.1.7"
+#define VERSION "0.1.8"
 // we only want to store the offset for every 10000th
 // line. otherwise, were we to store the position of every
 // line in the file, the index could become very large for
diff --git a/test.sh b/test.sh
index d67a841..e5e46b0 100644
--- a/test.sh
+++ b/test.sh
@@ -9,7 +9,7 @@ echo "indexing"
 time ./grabix index $FQ
 echo "indexed"
 python tests/test-fastq.py $FQ
-a=$(grabix grab test.cnt.gz $(($lines * 4)))
+a=$(./grabix grab test.cnt.gz $(($lines * 4)))
 b=$(zless $FQ | tail -1)
 if [[ "$a" != "$b" ]]; then
 	echo FAIL last record
