diff -ur NASP-69444e55e5b4704a5acae4537123023c28e26d59/nasp/configuration_parser.py NASP/nasp/configuration_parser.py
--- NASP-69444e55e5b4704a5acae4537123023c28e26d59/nasp/configuration_parser.py	2016-08-23 22:07:28.000000000 +0200
+++ NASP/nasp/configuration_parser.py	2016-09-05 14:33:58.474152834 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 # coding=utf-8
 
 __author__ = "Darrin Lemmer"
diff -ur NASP-69444e55e5b4704a5acae4537123023c28e26d59/nasp/convert_external_genome.py NASP/nasp/convert_external_genome.py
--- NASP-69444e55e5b4704a5acae4537123023c28e26d59/nasp/convert_external_genome.py	2016-08-23 22:07:28.000000000 +0200
+++ NASP/nasp/convert_external_genome.py	2016-09-05 14:33:58.402151378 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 
 __author__ = "David Smith"
 __version__ = "1.0.0"
diff -ur NASP-69444e55e5b4704a5acae4537123023c28e26d59/nasp/dispatcher.py NASP/nasp/dispatcher.py
--- NASP-69444e55e5b4704a5acae4537123023c28e26d59/nasp/dispatcher.py	2016-08-23 22:07:28.000000000 +0200
+++ NASP/nasp/dispatcher.py	2016-09-05 14:33:58.522153806 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 
 __author__ = "Darrin Lemmer"
 __version__ = "1.0.0"
diff -ur NASP-69444e55e5b4704a5acae4537123023c28e26d59/nasp/find_duplicates.py NASP/nasp/find_duplicates.py
--- NASP-69444e55e5b4704a5acae4537123023c28e26d59/nasp/find_duplicates.py	2016-08-23 22:07:28.000000000 +0200
+++ NASP/nasp/find_duplicates.py	2016-09-05 14:33:58.522153806 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 
 __author__ = "David Smith"
 __version__ = "1.0.0"
diff -ur NASP-69444e55e5b4704a5acae4537123023c28e26d59/nasp/format_fasta.py NASP/nasp/format_fasta.py
--- NASP-69444e55e5b4704a5acae4537123023c28e26d59/nasp/format_fasta.py	2016-08-23 22:07:28.000000000 +0200
+++ NASP/nasp/format_fasta.py	2016-09-05 14:33:58.522153806 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 
 __author__ = "David Smith"
 __version__ = "1.0.0"
diff -ur NASP-69444e55e5b4704a5acae4537123023c28e26d59/nasp/matrix_DTO.py NASP/nasp/matrix_DTO.py
--- NASP-69444e55e5b4704a5acae4537123023c28e26d59/nasp/matrix_DTO.py	2016-08-23 22:07:28.000000000 +0200
+++ NASP/nasp/matrix_DTO.py	2016-09-05 14:33:58.474152834 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 
 __author__ = "Darrin Lemmer"
 __version__ = "0.9.8"
diff -ur NASP-69444e55e5b4704a5acae4537123023c28e26d59/nasp/nasp_objects.py NASP/nasp/nasp_objects.py
--- NASP-69444e55e5b4704a5acae4537123023c28e26d59/nasp/nasp_objects.py	2016-08-23 22:07:28.000000000 +0200
+++ NASP/nasp/nasp_objects.py	2016-09-05 14:33:58.482152997 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 
 __author__ = "David Smith"
 __version__ = "1.0.0"
diff -ur NASP-69444e55e5b4704a5acae4537123023c28e26d59/nasp/nasp.py NASP/nasp/nasp.py
--- NASP-69444e55e5b4704a5acae4537123023c28e26d59/nasp/nasp.py	2016-08-23 22:07:28.000000000 +0200
+++ NASP/nasp/nasp.py	2016-09-05 14:33:58.402151378 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 
 __author__ = "Darrin Lemmer"
 __version__ = "1.0.0"
diff -ur NASP-69444e55e5b4704a5acae4537123023c28e26d59/nasp/nasptool/command/matrix/README.md NASP/nasp/nasptool/command/matrix/README.md
--- NASP-69444e55e5b4704a5acae4537123023c28e26d59/nasp/nasptool/command/matrix/README.md	2016-08-23 22:07:28.000000000 +0200
+++ NASP/nasp/nasptool/command/matrix/README.md	2016-09-05 14:33:58.462152592 +0200
@@ -38,7 +38,7 @@
         pass
 ````
 
-`time python3 indexFasta.py`
+`time python indexFasta.py`
 
 real    0m4.195s
 user    0m11.078s
diff -ur NASP-69444e55e5b4704a5acae4537123023c28e26d59/nasp/vcf_to_matrix.py NASP/nasp/vcf_to_matrix.py
--- NASP-69444e55e5b4704a5acae4537123023c28e26d59/nasp/vcf_to_matrix.py	2016-08-23 22:07:28.000000000 +0200
+++ NASP/nasp/vcf_to_matrix.py	2016-09-05 14:33:58.474152834 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 
 __author__ = "David Smith"
 __version__ = "0.9.9"
diff -ur NASP-69444e55e5b4704a5acae4537123023c28e26d59/scripts/filter_matrix_by_distance.py NASP/scripts/filter_matrix_by_distance.py
--- NASP-69444e55e5b4704a5acae4537123023c28e26d59/scripts/filter_matrix_by_distance.py	2016-08-23 22:07:28.000000000 +0200
+++ NASP/scripts/filter_matrix_by_distance.py	2016-09-05 14:33:58.390151136 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 
 __author__ = "Darrin Lemmer"
 __version__ = "0.9.9"
diff -ur NASP-69444e55e5b4704a5acae4537123023c28e26d59/tests.bak/nasp/configuration_parser_test.py NASP/tests.bak/nasp/configuration_parser_test.py
--- NASP-69444e55e5b4704a5acae4537123023c28e26d59/tests.bak/nasp/configuration_parser_test.py	2016-08-23 22:07:28.000000000 +0200
+++ NASP/tests.bak/nasp/configuration_parser_test.py	2016-09-05 14:33:59.226168043 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 # coding=utf-8
 
 __author__ = "jtravis"
diff -ur NASP-69444e55e5b4704a5acae4537123023c28e26d59/tests.bak/nasp/find_duplicates_test.py NASP/tests.bak/nasp/find_duplicates_test.py
--- NASP-69444e55e5b4704a5acae4537123023c28e26d59/tests.bak/nasp/find_duplicates_test.py	2016-08-23 22:07:28.000000000 +0200
+++ NASP/tests.bak/nasp/find_duplicates_test.py	2016-09-05 14:33:59.218167881 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 
 '''
 Created on Oct 18, 2013
