diff -ur nasp/configuration_parser.py nasp/configuration_parser.py
--- nasp/configuration_parser.py	2016-08-23 22:07:28.000000000 +0200
+++ nasp/configuration_parser.py	2016-09-05 14:33:58.474152834 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 # coding=utf-8
 
 __author__ = "Darrin Lemmer"
diff -ur nasp/convert_external_genome.py nasp/convert_external_genome.py
--- nasp/convert_external_genome.py	2016-08-23 22:07:28.000000000 +0200
+++ nasp/convert_external_genome.py	2016-09-05 14:33:58.402151378 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 
 __author__ = "David Smith"
 __version__ = "1.0.0"
diff -ur nasp/dispatcher.py nasp/dispatcher.py
--- nasp/dispatcher.py	2016-08-23 22:07:28.000000000 +0200
+++ nasp/dispatcher.py	2016-09-05 14:33:58.522153806 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 
 __author__ = "Darrin Lemmer"
 __version__ = "1.0.0"
diff -ur nasp/find_duplicates.py nasp/find_duplicates.py
--- nasp/find_duplicates.py	2016-08-23 22:07:28.000000000 +0200
+++ nasp/find_duplicates.py	2016-09-05 14:33:58.522153806 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 
 __author__ = "David Smith"
 __version__ = "1.0.0"
diff -ur nasp/format_fasta.py nasp/format_fasta.py
--- nasp/format_fasta.py	2016-08-23 22:07:28.000000000 +0200
+++ nasp/format_fasta.py	2016-09-05 14:33:58.522153806 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 
 __author__ = "David Smith"
 __version__ = "1.0.0"
diff -ur nasp/matrix_DTO.py nasp/matrix_DTO.py
--- nasp/matrix_DTO.py	2016-08-23 22:07:28.000000000 +0200
+++ nasp/matrix_DTO.py	2016-09-05 14:33:58.474152834 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 
 __author__ = "Darrin Lemmer"
 __version__ = "0.9.8"
diff -ur nasp/nasp_objects.py nasp/nasp_objects.py
--- nasp/nasp_objects.py	2016-08-23 22:07:28.000000000 +0200
+++ nasp/nasp_objects.py	2016-09-05 14:33:58.482152997 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 
 __author__ = "David Smith"
 __version__ = "1.0.0"
diff -ur nasp/nasp.py nasp/nasp.py
--- nasp/nasp.py	2016-08-23 22:07:28.000000000 +0200
+++ nasp/nasp.py	2016-09-05 14:33:58.402151378 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 
 __author__ = "Darrin Lemmer"
 __version__ = "1.0.0"
diff -ur nasp/nasptool/command/matrix/README.md nasp/nasptool/command/matrix/README.md
--- nasp/nasptool/command/matrix/README.md	2016-08-23 22:07:28.000000000 +0200
+++ nasp/nasptool/command/matrix/README.md	2016-09-05 14:33:58.462152592 +0200
@@ -38,7 +38,7 @@
         pass
 ````
 
-`time python3 indexFasta.py`
+`time python indexFasta.py`
 
 real    0m4.195s
 user    0m11.078s
diff -ur nasp/vcf_to_matrix.py nasp/vcf_to_matrix.py
--- nasp/vcf_to_matrix.py	2016-08-23 22:07:28.000000000 +0200
+++ nasp/vcf_to_matrix.py	2016-09-05 14:33:58.474152834 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 
 __author__ = "David Smith"
 __version__ = "0.9.9"
diff -ur scripts/filter_matrix_by_distance.py scripts/filter_matrix_by_distance.py
--- scripts/filter_matrix_by_distance.py	2016-08-23 22:07:28.000000000 +0200
+++ scripts/filter_matrix_by_distance.py	2016-09-05 14:33:58.390151136 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 
 __author__ = "Darrin Lemmer"
 __version__ = "0.9.9"
diff -ur tests.bak/nasp/configuration_parser_test.py tests.bak/nasp/configuration_parser_test.py
--- tests.bak/nasp/configuration_parser_test.py	2016-08-23 22:07:28.000000000 +0200
+++ tests.bak/nasp/configuration_parser_test.py	2016-09-05 14:33:59.226168043 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 # coding=utf-8
 
 __author__ = "jtravis"
diff -ur tests.bak/nasp/find_duplicates_test.py tests.bak/nasp/find_duplicates_test.py
--- tests.bak/nasp/find_duplicates_test.py	2016-08-23 22:07:28.000000000 +0200
+++ tests.bak/nasp/find_duplicates_test.py	2016-09-05 14:33:59.218167881 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env python
 
 '''
 Created on Oct 18, 2013
