--- a/pyproject.toml
+++ b/pyproject.toml
@@ -29,8 +29,6 @@
     'pyarrow~=20.0.0',
     'rapidfuzz~=3.13.0',
     'statsmodels~=0.14.6',
-    'tkinterweb~=3.23.5',
-    'ttkthemes~=3.2.2',
     'typing_extensions~=4.14.0',
 ]
 
@@ -78,9 +76,6 @@
 ranking = "countess.plugins.ranking:RankingPlugin"
 random_effects = "countess.plugins.random_effects:RandomEffectsPlugin"
 
-[project.entry-points.gui_scripts]
-countess_gui = "countess.gui.main:main"
-
 [project.entry-points.console_scripts]
 countess_cmd = "countess.core.cmd:main"
 
