diff --git a/r/configure b/r/configure
index d6af517..da143e2 100755
--- a/r/configure
+++ b/r/configure
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-if [ -z $BPCELLS_DEBUG_INSTALL ]; then 
+if [ -z $BPCELLS_DEBUG_INSTALL ]; then
     exec 3>/dev/null
 else
     exec 3>&1
@@ -12,7 +12,7 @@ fi
 # No identifiable information or IP addresses are saved, and server logs are
 # deleted every 14 days. More information on data privacy policy: https://plausible.io/data-policy
 
-# To skip counting your installation, either delete the following line, 
+# To skip counting your installation, either delete the following line,
 # or set the CI or ENABLE_INSTALL_COUNTING environment variables ahead of installation.
 if [ -z "$CI" ]; then ENABLE_INSTALL_COUNTING=${ENABLE_INSTALL_COUNTING:-yes}; fi
 
@@ -166,8 +166,8 @@ sed \
   -e "s/@HWY_MIN_VERSION@/$HWY_MIN_VERSION/g" \
   tools/hwy-test.cpp.in > tools/hwy-test.cpp
 
-HWY_CFLAGS="-Ibpcells-cpp"
-HWY_LIBS="-lhwy"
+HWY_CFLAGS="-I$PREFIX/include -Ibpcells-cpp"
+HWY_LIBS="-L$PREFIX/lib -lhwy"
 # Use a compile log to keep error/pragma messages
 HWY_COMPILE_LOG=$(mktemp)
 # Try to compile with hardcoded flags first
