--- src/poa-graph/Makefile
+++ src/poa-graph/Makefile
@@ -1,2 +1,2 @@
-
-AR=ar rc
+AR ?= ar
+AR += rc
@@ -29,3 +29,3 @@
-CC = gcc
-#CFLAGS= -g -ansi-strict -W -Wall -DUSE_WEIGHTED_LINKS -DUSE_PROJECT_HEADER -I.
-CFLAGS= -g -Ofast -DUSE_WEIGHTED_LINKS -DUSE_PROJECT_HEADER -I. -pthread
+CC ?= gcc
+#CFLAGS+= -g -ansi-strict -W -Wall -DUSE_WEIGHTED_LINKS -DUSE_PROJECT_HEADER -I.
+CFLAGS+= -g -Ofast -DUSE_WEIGHTED_LINKS -DUSE_PROJECT_HEADER -I. -pthread
@@ -47 +47 @@
-	$(CC) -o $@ $(OBJECTS) -lm liblpo.a -lpthread
+	$(CC) $(LDFLAGS) -o $@ $(OBJECTS) -lm liblpo.a -lpthread
--- src/simulator/makefile
+++ src/simulator/makefile
@@ -1,4 +1,4 @@
-#~ CC=/usr/bin/g++
-CC=g++
-CFLAGS=  -Wall -Wextra  -Ofast -std=c++11 -march=native -pthread -pipe -Isparsepp
-LDFLAGS=-pthread -Isparsepp
+#~ CXX=/usr/bin/g++
+CXX?=g++
+CXXFLAGS+=  -Wall -Wextra  -Ofast -std=c++11 -pthread -pipe -Isparsepp
+LDFLAGS+=-pthread -Isparsepp
@@ -8,2 +8,2 @@
-CFLAGS=-std=c++0x -pg -O3  -march=native
-LDFLAGS=-pg
+CXXFLAGS+=-std=c++0x -pg -O3
+LDFLAGS+=-pg
@@ -13,2 +13,2 @@
-CFLAGS=-std=c++0x -O3 -g
-LDFLAGS=-g
+CXXFLAGS+=-std=c++0x -O3 -g
+LDFLAGS+=-g
@@ -24 +24 @@
-	$(CC) -o $@ $^ $(LDFLAGS)
+	$(CXX) $(LDFLAGS) -o $@ $^
@@ -27 +27 @@
-	$(CC) -o $@ -c $< $(CFLAGS)
+	$(CXX) -o $@ -c $< $(CXXFLAGS)
--- src/split/makefile
+++ src/split/makefile
@@ -1,3 +1,3 @@
-CC=g++
-CFLAGS=  -Wall  -Ofast -std=c++11  -pthread -pipe -Wfatal-errors -fopenmp
-LDFLAGS=-pthread -pipe -flto -fopenmp
+CXX?=g++
+CXXFLAGS+=  -Wall  -Ofast -std=c++11  -pthread -pipe -Wfatal-errors -fopenmp
+LDFLAGS+=-pthread -pipe -flto -fopenmp
@@ -12 +12 @@
-	$(CC) -o $@ $^ $(LDFLAGS)
+	$(CXX) $(LDFLAGS) -o $@ $^
@@ -15 +15 @@
-	$(CC) -o $@ -c $< $(CFLAGS)
+	$(CXX) -o $@ -c $< $(CXXFLAGS)
@@ -18 +18 @@
-	$(CC) -o $@ $^ $(LDFLAGS)
+	$(CXX) $(LDFLAGS) -o $@ $^
@@ -21 +21 @@
-	$(CC) -o $@ -c $< $(CFLAGS)
+	$(CXX) -o $@ -c $< $(CXXFLAGS)
@@ -24 +24 @@
-	$(CC) -o $@ $^ $(LDFLAGS)
+	$(CXX) $(LDFLAGS) -o $@ $^
--- src/utils/makefile
+++ src/utils/makefile
@@ -1,3 +1,3 @@
-CC=g++
-CFLAGS=  -Wall -Wextra  -Ofast -std=c++11 -march=native  -pipe -Isparsepp
-LDFLAGS= -Isparsepp
+CXX?=g++
+CXXFLAGS+=  -Wall -Wextra  -Ofast -std=c++11  -pipe -Isparsepp
+LDFLAGS+= -Isparsepp
@@ -12 +12 @@
-	$(CC) -o $@ $^ $(LDFLAGS)
+	$(CXX) $(LDFLAGS) -o $@ $^
@@ -15 +15 @@
-	$(CC) -o $@ -c $< $(CFLAGS)
+	$(CXX) -o $@ -c $< $(CXXFLAGS)
