--- batindel/src/penguin.cpp
+++ batindel/src/penguin.cpp
@@ -78,7 +78,7 @@ int SW_THRESHOLD=290;
 int READS_TO_ESTIMATE=100000;
 int MISINSCAN=0;
 int SEG_SIZE=18;
-const int ScaleQ[]={0,1.5,1.75,2,3,4};
+
 extern const Alignment Default_Alignment={0};
 const int ORGSTRINGLENGTH=2200; 
 bool PAIRED=FALSE;
--- batindel/src/penguin-s.cpp
+++ batindel/src/penguin-s.cpp
@@ -49,7 +49,7 @@ int MODE=INT_MAX;
 int SEEDSIZE=INT_MAX;
 int INSERT=5;
 int DELETE=5;
-const int ScaleQ[]={0,1.5,1.75,2,3,4};
+
 extern const Alignment Default_Alignment={0};
 //extern const int QUALITYCONVERSIONFACTOR=64;
 //extern const int QUALITYSCALEFACTOR=33;
