--- edge.pl	2013-05-31 12:47:40.000000000 +0100
+++ edge.pl.new	2017-10-04 19:05:50.887525991 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 die <<ENDARGS
 Usage
@@ -10,6 +10,7 @@
 use warnings;
 use strict;
 use Switch;
+use FindBin qw($Bin);
 
 #####Default values
 my $genome="none";
@@ -18,7 +19,7 @@
 my $reads="none";
 my $reads2="none";
 my $threads=1;
-my $sourceDir=".";
+my $sourceDir=$Bin;
 my $prefix="out";
 my $minins=0;
 my $maxins=500;
