diff --git a/Makefile b/Makefile
index a7e0ef8..8686c1a 100644
--- a/Makefile
+++ b/Makefile
@@ -31,12 +31,12 @@ endif
 SVNREV := -D'SVN_REV="$(VERSION_STRING)"'
 
 # Define default paths
-prefix=/usr/local
+prefix=${PREFIX}
 bindir = $(prefix)/bin
 datadir=$(prefix)/share
 BLAST_BIN=$(BIN)
 HMMER_BIN=$(BIN)
-INSTALL_DIR = $(datadir)/amrfinder
+INSTALL_DIR=$(bindir)
 INSTALL_PROGRAM=install
 
 CPPFLAGS = -std=gnu++11 -pthread -malign-double -fno-math-errno -O3 $(SVNREV)
