--- Makefile
+++ Makefile
@@ -1,7 +1,7 @@
-CFLAGS=-c -Wall -O0 -g -std=c99
+CFLAGS=-c -Wall -O0 -g -std=c99 -I${PREFIX}/include
 #recommended options: -ffast-math -ftree-vectorize -march=core2 -mssse3 -O3
-COPTS=
-LDFLAGS=-lz -lm
+COPTS=-ffast-math -ftree-vectorize -mtune=core2 -mssse3 -O3
+LDFLAGS=-L${PREFIX}/lib -lz -lm
 SOURCES=SeqPrep.c utils.c stdaln.c
 OBJECTS=$(SOURCES:.c=.o)
 EXECUTABLE=SeqPrep
