--- a/SpoTyping-3.0/SpoTyping-v3.0-commandLine/SpoTyping.py
+++ b/SpoTyping-3.0/SpoTyping-v3.0-commandLine/SpoTyping.py
@@ -29,7 +29,7 @@
 import urllib.request, urllib.parse, urllib.error

 ## Global variables
-dir = os.path.split(os.path.realpath(__file__))[0] # script directory
+dir = os.path.join(sys.prefix, 'share', 'SpoTyping')# script directory
 setlength = 50*5000000                             # base input cut-off for swift mode

 ## Option variables
