diff --git a/VERSION b/VERSION.txt
rename from VERSION
rename to VERSION.txt
diff --git a/Makefile b/Makefile
--- a/Makefile
+++ b/Makefile
@@ -131 +131 @@
-VERSION = $(shell cat VERSION)
+VERSION = $(shell cat VERSION.txt)
@@ -188 +188 @@
-               VERSION
+               VERSION.txt
@@ -218 +218 @@
-     -DBOWTIE_VERSION="\"`cat VERSION`\"" \
+     -DBOWTIE_VERSION="\"`cat VERSION.txt`\"" \
