--- configure.orig	2015-04-20 22:45:13.803069541 +0000
+++ configure	2015-04-20 22:46:30.082986916 +0000
@@ -13373,8 +13373,9 @@
 FFI_PC_CFLAGS=""
 FFI_PC_LIBS=""
 FFI_PC_PACKAGES=""
-if test x"$have_ffi_pkgconfig" = xyes ; then
-    FFI_PC_PACKAGES="libffi"
+if true ; then
+  FFI_PC_CFLAGS="$FFI_CFLAGS"
+  FFI_PC_LIBS="$FFI_LIBS"
 else
   { $as_echo "$as_me:${as_lineno-$LINENO}: checking for ffi.h" >&5
 $as_echo_n "checking for ffi.h... " >&6; }
