diff --git a/Cargo.toml b/Cargo.toml
index 3f9624e..219948e 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -6,7 +6,7 @@ rust-version = "1.91"
 edition = "2021"
 
 [dependencies]
-htslib = { package = "rust-htslib", version = "1.0", default-features = false, features = ["bzip2", "curl", "lzma"] }
+htslib = { package = "rust-htslib", version = "0.40.2", default-features = false }
 nalgebra = { version = "0.34", default-features = false, features = ["std"] } # Possibly remove later.
 statrs = { version = "0.18", default-features = false } # Possibly remove later.
 nohash = { package = "nohash-hasher", version = "0.2.0", default-features = false, features = ["std"] }
