diff -u -ur ncbi-vdb-2.9.6/build/ld.linux.slib.sh ncbi-vdb-2.9.6/build/ld.linux.slib.sh
--- ncbi-vdb-2.9.6/build/ld.linux.slib.sh	2019-03-18 17:57:05.000000000 +0100
+++ ncbi-vdb-2.9.6/build/ld.linux.slib.sh	2019-06-24 10:55:35.020645910 +0200
@@ -41,14 +41,14 @@
 source "${0%slib.sh}cmn.sh"
 
 # initialize command
-CMD="ar -rc"
+CMD="$AR -rc"
 
 # function to convert an archive into individual object files
 convert-static ()
 {
     # list members
     local path="$1"
-    local mbrs="$(ar -t $path)"
+    local mbrs="$($AR -t $path)"
 
     # unpack archive into temporary directory
     mkdir -p ld-tmp
@@ -57,7 +57,7 @@
         echo "$SELF_NAME: failed to cd to ld-tmp"
         exit 5
     fi
-    ar -x "$path"
+    $AR -x "$path"
 
     # rename and add to source files list
     local m=
