diff --git a/build/env.cmake b/build/env.cmake
index 9527451..27f62ab 100644
--- a/build/env.cmake
+++ b/build/env.cmake
@@ -106,17 +106,14 @@ else ()
 endif ()
 
 # create variables based entirely upon OS
-if ( "mac" STREQUAL ${OS} )
+if ( CMAKE_C_COMPILER_ID STREQUAL "AppleClang" )
     add_compile_definitions( MAC BSD UNIX )
     set(CMAKE_C_ARCHIVE_CREATE   "<CMAKE_AR> Scr <TARGET> <LINK_FLAGS> <OBJECTS>")
-    set(CMAKE_CXX_ARCHIVE_CREATE "<CMAKE_AR> Scr <TARGET> <LINK_FLAGS> <OBJECTS>")
     set(CMAKE_C_ARCHIVE_FINISH   "<CMAKE_RANLIB> -no_warning_for_no_symbols -c <TARGET>")
+endif()
+if ( CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang" )
+    set(CMAKE_CXX_ARCHIVE_CREATE "<CMAKE_AR> Scr <TARGET> <LINK_FLAGS> <OBJECTS>")
     set(CMAKE_CXX_ARCHIVE_FINISH "<CMAKE_RANLIB> -no_warning_for_no_symbols -c <TARGET>")
-elseif( "linux" STREQUAL ${OS} )
-    add_compile_definitions( LINUX UNIX )
-    set( LMCHECK -lmcheck )
-elseif( "windows" STREQUAL ${OS} )
-    add_compile_definitions( WINDOWS _WIN32_WINNT=0x0502 )
 endif()
 
 # create variables based entirely upon ARCH
