diff -u -ur ncbi-vdb-2.9.6/build/ld.linux.dlib.sh ncbi-vdb-2.9.6/build/ld.linux.dlib.sh
--- ncbi-vdb-2.9.6/build/ld.linux.dlib.sh	2019-03-18 17:57:05.000000000 +0100
+++ ncbi-vdb-2.9.6/build/ld.linux.dlib.sh	2019-06-24 10:42:33.167872909 +0200
@@ -46,18 +46,7 @@
 # parameters and common functions
 source "${0%dlib.sh}cmn.sh"
 
-# discover tool chain
-case "$LD" in
-g*)
-    source "${0%dlib.sh}gcc.sh"
-    ;;
-i*)
-    source "${0%dlib.sh}icc.sh"
-    ;;
- *)
-    echo "$SELF_NAME: unrecognized ld tool - '$LD'"
-    exit 5
-esac
+source "${0%dlib.sh}gcc.sh"
 
 # DLIB_CMD was started in tool-specific source
 CMD="$DLIB_CMD $LDFLAGS"
 
