--- ncbi-vdb/setup/konfigure.perl
+++ ncbi-vdb/setup/konfigure.perl
@@ -336,8 +336,8 @@
     $SHLX = 'so';
     $EXEX = '';
     $OSINC = 'unix';
-    $TOOLS = 'gcc' unless ($TOOLS);
+    $TOOLS = $ENV{'CC'};
     $PYTHON = 'python';
 } elsif ($OSTYPE =~ /darwin/i) {
     $LPFX = 'lib';
     $OBJX = 'o';
@@ -366,12 +366,13 @@
 $CPP     = $OPT{CXX    } if ($OPT{CXX    });
 $LDFLAGS = $OPT{LDFLAGS} if ($OPT{LDFLAGS});

-if ($TOOLS =~ /gcc$/) {
-    $CPP  = 'g++' unless ($CPP);
-    $CC   = "$TOOLS -c";
+println "TOOLS = $TOOLS";
+if ($TOOLS =~ /cc$/) {
+    $CPP  = $ENV{'CPP'} unless ($CPP);
+    $CC   = $ENV{'CC'} . " -c";
     $CP   = "$CPP -c";
-    $AR   = 'ar rc';
-    $ARX  = 'ar x';
-    $ARLS = 'ar t';
+    $AR   = $ENV{'AR'};
+    $ARX  = "$AR x";
+    $ARLS = "$AR t";
     $LD   = $TOOLS;
     $LP   = $CPP;
