--- a/ncbi-vdb/setup/konfigure.perl
+++ b/ncbi-vdb/setup/konfigure.perl
@@ -394,12 +394,12 @@ $CC      = $OPT{CC     } if ($OPT{CC     });
 $CPP     = $OPT{CXX    } if ($OPT{CXX    });
 $LDFLAGS = $OPT{LDFLAGS} if ($OPT{LDFLAGS});
 
-if ($TOOLS =~ /gcc$/) {
-    $CPP  = 'g++' unless ($CPP);
-    $CP   = "$CPP -c";
-    $AR   = 'ar rc';
-    $ARX  = 'ar x';
-    $ARLS = 'ar t';
+if ($TOOLS =~ /cc$/) {
+    $CPP  = $ENV{'CPP'} unless ($CPP);
+    $CP   = $ENV{'CC'} . " -c";
+    $AR   = $ENV{'AR'};
+    $ARX  = "$AR x";
+    $ARLS = "$AR t";
     $LP   = $CPP;
 
     $CC   = $TOOLS unless ($CC);
