diff --git a/R/build.R b/R/build.R
index b4ed0ed..56844a1 100644
--- a/R/build.R
+++ b/R/build.R
@@ -55,7 +55,7 @@ cytolibLdFlags <- function() {
 cytolibLibPath <- function(suffix = "") {
    sysname <- Sys.info()['sysname']
    cytolibSupported <- list(
-      "Darwin" = paste("cytolib", suffix, ".so", sep = ""), 
+      "Darwin" = paste("cytolib", suffix, ".dylib", sep = ""),
       "Linux" = paste("cytolib", suffix, ".so", sep = ""), 
       "Windows" = paste("libcytolib", suffix, ".a", sep = ""),
       "SunOS" = paste("cytolib", suffix, ".so", sep = "")
