--- scripts/hal2maf_split.pl	2015-07-16 14:38:10.000000000 +0400
+++ scripts/hal2maf_split.pl	2016-05-25 17:29:23.628965671 +0400
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # this script works on top of the 'halTools' toolbox
 # and exports a hal alignment to maf by splitting the alignment into several smaller
 # alignment chunks of a certain size. An overlap between two consecutive alignment chunks can
