--- scripts/selectLongestORFs.pl	2015-04-24 16:23:45.000000000 +0400
+++ scripts/selectLongestORFs.pl	2016-05-25 17:29:23.661965672 +0400
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # Jonas Behr's tool Transcript Skimmer will report two ORFs in assembled transcripts if strand unspecific RNA-Seq data was used for assembly. This script selects and reports the longer of those two transcripts.
 
