From 2966bfa62f06ba304e7f7617baaa26182938463a Mon Sep 17 00:00:00 2001
From: Matus Kosut <matus.kosut@ntnu.no>
Date: Thu, 29 Oct 2020 18:28:49 +0100
Subject: [PATCH] patch cmakelists for conda build

---
 CMakeLists.txt | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 4d7a7a4..8854db6 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -17,10 +17,13 @@ if(CGET_PREFIX)
     link_directories(${CGET_PREFIX}/lib)
 endif()
 
+include_directories("include" "${SRC_DIR}/include" "${CMAKE_INSTALL_PREFIX}/include" "${CMAKE_INSTALL_PREFIX}/include/shrinkwrap" "${CMAKE_INSTALL_PREFIX}/include/htslib")
+link_directories("lib" "${SRC_DIR}/lib" "${CMAKE_INSTALL_PREFIX}/lib")
+
 #list(APPEND CMAKE_PREFIX_PATH ${CMAKE_INSTALL_PREFIX})
 #find_package(shrinkwrap CONFIG REQUIRED)
 #if (NOT BUILD_SHARED_LIBS)
-    set(CMAKE_FIND_LIBRARY_SUFFIXES ".a;${CMAKE_FIND_LIBRARY_SUFFIXES}") # Prefer libz.a when both are available
+#    set(CMAKE_FIND_LIBRARY_SUFFIXES ".a;${CMAKE_FIND_LIBRARY_SUFFIXES}") # Prefer libz.a when both are available
 #endif()
 find_library(HTS_LIBRARY hts)
 find_library(ZLIB_LIBRARY z)
@@ -52,7 +55,7 @@ add_library(savvy
         src/savvy/vcf_reader.cpp include/savvy/vcf_reader.hpp)
 
 target_link_libraries(savvy ${HTS_LIBRARY} ${ZLIB_LIBRARY} ${ZSTD_LIBRARY} ${CMAKE_THREAD_LIBS_INIT})
-target_include_directories(savvy PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include> $<INSTALL_INTERFACE:include>)
+target_include_directories(savvy PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include> $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}> $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/shrinkwrap> $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/htslib>)
 
 add_executable(sav
         src/sav/main.cpp
@@ -115,13 +118,13 @@ if(BUILD_SLR_EXAMPLES)
     target_link_libraries(linreg-ttest savvy ${OPENCL_LIB})
 endif()
 
-install(DIRECTORY include/savvy COMPONENT api DESTINATION include)
+install(DIRECTORY include/savvy COMPONENT api DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
 install(TARGETS savvy COMPONENT api EXPORT ${PROJECT_NAME}-config
         LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
         ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
 
 install(TARGETS sav COMPONENT cli
-        RUNTIME DESTINATION bin
+        RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
         OPTIONAL)
 
 install(EXPORT ${PROJECT_NAME}-config COMPONENT api DESTINATION share/${PROJECT_NAME})
-- 
2.23.0

