diff --git a/CMakeLists.txt b/CMakeLists.txt
index b94227d..a731d3d 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -138,7 +138,7 @@ if(NOT mcfp_FOUND)
 	FetchContent_MakeAvailable(mcfp)
 endif()
 
-find_package(zeep 6.0.16 QUIET)
+find_package(zeep 6.0.16 QUIET COMPONENTS http)
 
 if(NOT zeep_FOUND)
 	FetchContent_Declare(
@@ -148,9 +148,17 @@ if(NOT zeep_FOUND)
 	FetchContent_MakeAvailable(zeep)
 endif()
 
+if(TARGET zeep AND NOT TARGET zeep::zeep)
+  add_library(zeep::zeep ALIAS zeep)
+endif()
+
+if(TARGET zeep-http AND NOT TARGET zeep::http)
+  add_library(zeep::http ALIAS zeep-http)
+endif()
+
 if(BUILD_WEB_APPLICATION)
 	find_package(libpqxx 7.8.0 QUIET)
-	
+
 	if(NOT libpqxx_FOUND)
 		FetchContent_Declare(
 			libpqxx
