diff --git a/src/Makefile b/src/Makefile
index df902f3..76b2771 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -1,8 +1,8 @@
-CXX ?= g++
+CXX ?= $(CXX)
 KRAKEN2_SKIP_FOPENMP ?= -fopenmp
-CXXFLAGS = $(KRAKEN2_SKIP_FOPENMP) -Wall -std=c++11 -O3 -fPIC
-CXXFLAGS += -DLINEAR_PROBING
-CFLAGS = -Wall -std=c99 -O0 -g3
+CXXFLAGS = $(KRAKEN2_SKIP_FOPENMP) -Wall -std=c++14 -O3 -fPIC
+CXXFLAGS += -DLINEAR_PROBING $(LDFLAGS)
+CFLAGS = -Wall -std=c11 -O3 -g3
 
 .PHONY: all clean install
 
@@ -11,7 +11,7 @@ PROGS = estimate_capacity build_db classify dump_table lookup_accession_numbers
 all: $(PROGS)
 
 install: $(PROGS)
-	cp $(PROGS) "$(KRAKEN2_DIR)/"
+	install -v -m 0755 $(PROGS) "$(KRAKEN2_DIR)"
 
 clean:
 	rm -f *.o $(PROGS)
