diff --git a/src/Makefile b/src/Makefile
index df902f3..39be19e 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -1,17 +1,17 @@
-CXX ?= g++
+CXX ?= $(CXX)
 KRAKEN2_SKIP_FOPENMP ?= -fopenmp
-CXXFLAGS = $(KRAKEN2_SKIP_FOPENMP) -Wall -std=c++11 -O3 -fPIC
-CXXFLAGS += -DLINEAR_PROBING
-CFLAGS = -Wall -std=c99 -O0 -g3
+CXXFLAGS = $(KRAKEN2_SKIP_FOPENMP) -Wall -std=c++14 -O3 -fPIC
+CXXFLAGS += -DLINEAR_PROBING $(LDFLAGS)
+CFLAGS = -Wall -std=c11 -O3 -g3
 
 .PHONY: all clean install
 
-PROGS = estimate_capacity build_db classify dump_table lookup_accession_numbers k2mask blast_to_fasta libtax
+PROGS = estimate_capacity build_db classify dump_table lookup_accession_numbers k2mask blast_to_fasta libtax.so
 
 all: $(PROGS)
 
 install: $(PROGS)
-	cp $(PROGS) "$(KRAKEN2_DIR)/"
+	install -v -m 0755 $(PROGS) "$(KRAKEN2_DIR)"
 
 clean:
 	rm -f *.o $(PROGS)
@@ -61,5 +61,5 @@ blast_to_fasta: blast_to_fasta.o blast_defline.o blast_utils.o
 
 libtax.o: libtax.cc
 
-libtax: taxonomy.o mmap_file.o libtax.o
+libtax.so: taxonomy.o mmap_file.o libtax.o
 	$(CXX) $(CXXFLAGS) -shared -o libtax.so taxonomy.o mmap_file.o libtax.o
