diff --git a/src/Makefile b/src/Makefile
index c340d06..4c5d169 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -1,8 +1,8 @@
-CXX ?= g++
+CXX ?= $(CXX)
 KRAKEN2_SKIP_FOPENMP ?= -fopenmp
-CXXFLAGS = $(KRAKEN2_SKIP_FOPENMP) -Wall -std=c++11 -O3 -fPIC
-CXXFLAGS += -DLINEAR_PROBING
-CFLAGS = -Wall -std=c99 -O0 -g3
+CXXFLAGS = $(KRAKEN2_SKIP_FOPENMP) -Wall -std=c++14 -O3 -fPIC
+CXXFLAGS += -DLINEAR_PROBING $(LDFLAGS)
+CFLAGS = -Wall -std=c11 -O3 -g3
 
 .PHONY: all clean install
 
@@ -11,7 +11,7 @@ PROGS = estimate_capacity build_db classify dump_table lookup_accession_numbers
 all: $(PROGS)
 
 install: $(PROGS)
-	cp $(PROGS) "$(KRAKEN2_DIR)/"
+	install -v -m 0755 $(PROGS) "$(KRAKEN2_DIR)"
 
 clean:
 	rm -f *.o $(PROGS)
diff --git a/install_kraken2.sh b/install_kraken2.sh
index 248b553..750428a 100755
--- a/install_kraken2.sh
+++ b/install_kraken2.sh
@@ -28,10 +28,10 @@ fi
 # Perl cmd used to canonicalize dirname - "readlink -f" doesn't work
 # on OS X.
 export KRAKEN2_DIR
-KRAKEN2_DIR=$(perl -MCwd=abs_path -le 'print abs_path(shift)' "$1")
+KRAKEN2_DIR="${PREFIX}/share/${PKG_NAME}-${PKG_VERSION}/libexec"
 
 mkdir -p "$KRAKEN2_DIR"
-make -C src install
+make -C src install CXX="$CXX" CC="$CC" -j"$CPU_COUNT"
 for file in scripts/*
 do
   destination_file="$KRAKEN2_DIR/$(basename "$file")"
