diff --git a/scripts/k2 b/scripts/k2
old mode 100755
new mode 100644
index 04160c2..7d5a574
--- a/scripts/k2
+++ b/scripts/k2
@@ -431,7 +431,7 @@ def download_and_process_blast_volumes(args):
     ) as pool:
         with open("manifest.txt", "r") as in_file:
             tarballs = in_file.readlines()
-        f = functools.partial(
+        wrapped_func = functools.partial(
             wrap_with_globals, extract_blast_db_files,
             LOG.get_queue(), LOG.get_level(),
             SCRIPT_PATHNAME
@@ -442,7 +442,7 @@ def download_and_process_blast_volumes(args):
         )
         for tarball in tarballs:
             tarball = os.path.abspath(tarball)
-            f = pool.submit(extract_blast_db_files, tarball.strip())
+            f = pool.submit(wrapped_func, tarball.strip())
             extraction_futures.append(f)
         for future in concurrent.futures.as_completed(extraction_futures):
             result = future.result()
@@ -4214,12 +4214,17 @@ def merge_classification_output_parallel(
         input = list(zip(file1.readlines(), file2.readlines()))
     input_len = len(input)
     partition_ranges = list(range(0, input_len, int(input_len / args.threads)))
-    partition_ranges[-1] = input_len
+    partition_ranges.append(input_len)
     job_number = 0
     futures = []
+    wrapped_func = functools.partial(
+        wrap_with_globals, merge_classification_output2,
+        LOG.get_queue(), LOG.get_level(),
+        SCRIPT_PATHNAME
+    )
     for start, end in zip(partition_ranges, partition_ranges[1:]):
         future = pool.submit(
-            merge_classification_output2, taxonomy_pathname,
+            wrapped_func, taxonomy_pathname,
             input[start:end], job_number, use_names, args,
             save_seq_names, final
         )
