diff --git a/linearpartition b/linearpartition
index f6c2a4e..7b0fe1d 100755
--- a/linearpartition
+++ b/linearpartition
@@ -80,7 +80,7 @@ def main():
             print("Exit!\n");
             exit();
 
-    path = os.path.dirname(os.path.abspath(__file__))
+    path = os.path.dirname(os.path.realpath(__file__))
     cmd = ["%s/%s" % (path, ('bin/linearpartition_v' if use_vienna else 'bin/linearpartition_c')), beamsize, is_sharpturn, is_verbose, bpp_file, bpp_prefix, pf_only, bpp_cutoff, forest_file, mea, gamma, TK, threshold, ThreshKnot_prefix, MEA_prefix, MEA_bpseq, shape_file_path, is_fasta, dangles]
     subprocess.call(cmd, stdin=sys.stdin)
     
