From 7413bfa18dddf47924be0839df0147045db0d045 Mon Sep 17 00:00:00 2001
From: Fabien Pertuy <Fabien.Pertuy@Sanofi.com>
Date: Fri, 21 May 2021 10:39:10 +0000
Subject: [PATCH] FIX command broken with symbolic link

---
 linearpartition | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/linearpartition b/linearpartition
index 6c9d877..9d8a5b6 100755
--- a/linearpartition
+++ b/linearpartition
@@ -87,7 +87,7 @@ def main():
     #     if os.path.exists(ThreshKnot_output): os.remove(ThreshKnot_output)
 
 
-    path = os.path.dirname(os.path.abspath(__file__))
+    path = os.path.dirname(os.path.realpath(__file__))
     cmd = ["%s/%s" % (path, ('bin/linearpartition_v' if use_vienna else 'bin/linearpartition_c')), beamsize, is_sharpturn, is_verbose, bpp_file, bpp_prefix, pf_only, bpp_cutoff, forest_file, mea, gamma, TK, threshold, ThreshKnot_prefix, MEA_prefix, MEA_bpseq]
     subprocess.call(cmd, stdin=sys.stdin)
     
-- 
2.30.2

