From e0f4ac49aa6f6d3446eed5183ea585f7bea31c34 Mon Sep 17 00:00:00 2001
From: nikostr <nikostr@users.noreply.github.com>
Date: Fri, 31 May 2024 21:50:45 +0200
Subject: [PATCH] Replaces fname char with std::string


diff --git a/dif_cover_scripts/from_unionbed_to_ratio_per_window_CC0 b/dif_cover_scripts/from_unionbed_to_ratio_per_window_CC0
old mode 100644
new mode 100755
index ee868ce..6c0a513
Binary files a/dif_cover_scripts/from_unionbed_to_ratio_per_window_CC0 and b/dif_cover_scripts/from_unionbed_to_ratio_per_window_CC0 differ
diff --git a/dif_cover_scripts/from_unionbed_to_ratio_per_window_CC0.cpp b/dif_cover_scripts/from_unionbed_to_ratio_per_window_CC0.cpp
index 88aa296..40152a5 100644
--- a/dif_cover_scripts/from_unionbed_to_ratio_per_window_CC0.cpp
+++ b/dif_cover_scripts/from_unionbed_to_ratio_per_window_CC0.cpp
@@ -20,7 +20,7 @@ int main( int argc , char** argv ) {
 
 
    int opt, a=1, A=max_cov, b=1, B=max_cov, v=1000, l=0; //!!! Default values for parameters !!!!
-   char fname[1000], fff[1000];
+   char fff[1000];
  
    while ((opt = getopt(argc, argv, "a:A:b:B:v:l:")) != -1) {
         switch (opt) {
@@ -54,12 +54,12 @@ int main( int argc , char** argv ) {
         exit(EXIT_FAILURE);
     }
 
-   strncat(fname,argv[optind],strlen(argv[optind]));
+   std::string fname(argv[optind]);
    //printf ("fname=%s\n",fname);
      
    std::ifstream unionbed_file( fname ) ;
    if (!unionbed_file) {
-	fprintf(stderr, "Can't open union.bed file %s\n",fname);	
+	fprintf(stderr, "Can't open union.bed file %s\n",fname.c_str());
         exit(EXIT_FAILURE);
     }
  
-- 
2.44.0

