diff -rupN graphviz-2.38.0.orig/lib/gvc/gvconfig.c graphviz-2.38.0/lib/gvc/gvconfig.c
--- lib/gvc/gvconfig.c	2014-04-14 04:40:25.000000000 +0800
+++ lib/gvc/gvconfig.c	2014-05-06 21:02:14.000000000 +0800
@@ -311,7 +311,7 @@ char * gvconfig_libdir(GVC_t * gvc)
 			char* s = tmp-1;
 			/* back up to previous slash (or head of string) */
 			while ((*s != '/') && (s > path)) s--;
-			if (strncmp (s, DOTLIBS, STRLEN(DOTLIBS)) == 0);
+			if (strncmp (s, DOTLIBS, STRLEN(DOTLIBS)) == 0)
 			    continue;
 		    }
 
