diff --git ema/Makefile ema.patched/Makefile
index 15e57b0..c7fba73 100644
--- ema/Makefile
+++ ema.patched/Makefile
@@ -2,11 +2,11 @@ TARGET = ema
 LIBS = -L$(BWADIR) -lbwa -lm -lz -lpthread
 CC = gcc
 WARNINGS = -Wall -Wextra
-CFLAGS = -std=gnu99 -march=native -O3 -fopenmp -flto -fstrict-aliasing $(WARNINGS)
-LFLAGS = -lstdc++ -march=native -O3 -flto -fopenmp -lpthread
+CFLAGS = -std=gnu99 -march=x86-64 -O3 -fopenmp -flto -fstrict-aliasing $(WARNINGS)
+LFLAGS = -lstdc++ -march=x86-64 -O3 -flto -fopenmp -lpthread
 
 CXX = g++
-CPPFLAGS = -c -std=c++11 -O3 -march=native -pthread
+CPPFLAGS = -c -std=c++11 -O3 -march=x86-64 -pthread
 LDFLAGS = -pthread
 
 #CFLAGS = -std=gnu99 -fstrict-aliasing -fopenmp -ggdb -O0 $(WARNINGS)
