From db2b15e2d9829a34fb12225ad2bcc8a3ea2e4eb9 Mon Sep 17 00:00:00 2001
From: David Laehnemann <david.laehnemann@hhu.de>
Date: Wed, 19 Jul 2017 17:03:06 +0200
Subject: [PATCH 1/5] Compile using openmp (mpic++) and pthread

---
 Makefile | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/Makefile b/Makefile
index 5f105a1..e015f01 100644
--- a/Makefile
+++ b/Makefile
@@ -32,18 +32,18 @@
 
 #Linux, icc with openmp & pthread
 #CPP = g++
-#CPP = mpic++
+CPP = mpic++
 #CPPFLAGS = -m64 -fopenmp -O3 -s -static
-#MPICPPFLAGS = -I/usr/lib/openmpi/include -I/usr/lib/openmpi/include/openmpi -pthread -L/usr/lib/openmpi/lib -lmpi_cxx -lmpi -lopen-rte -lopen-pal -ldl -lnsl -lutil -lm -ldl -Wl,--export-dynamic
-#CPPFLAGS = -m64 -openmp -pthread -O3 -DMAXK=$(MAXK) -DFIXEDPOINTPROB=1 $(MPICPPFLAGS)
+MPICPPFLAGS = -I/usr/lib/openmpi/include -I/usr/lib/openmpi/include/openmpi -pthread -L/usr/lib/openmpi/lib -lmpi_cxx -lmpi -lopen-rte -lopen-pal -ldl -lnsl -lutil -lm -ldl -Wl,--export-dynamic
+CPPFLAGS = -m64 -openmp -pthread -O3 -DMAXK=$(MAXK) -DFIXEDPOINTPROB=1 $(MPICPPFLAGS)
 
 #Linux, gcc with openmp
-CPP = g++
+#CPP = g++
 #CPPFLAGS = -m64 -fopenmp -O3 -s -static
 #PROFILINGFLAG = -pg
 #VALGRINDFLAG = -g
 #DEBUGFLAG = -g
-CPPFLAGS = -m64 -fopenmp -O3 $(PROFILINGFLAG) $(DEBUGFLAG) -s
+#CPPFLAGS = -m64 -fopenmp -O3 $(PROFILINGFLAG) $(DEBUGFLAG) -s
 LIBEXT = a
 LIBCREATE = rcs
 LIBTOOL = ar 
-- 
2.7.4


From 2a01140eb64ab5a3ee8e5a6feb9f8b79396af224 Mon Sep 17 00:00:00 2001
From: David Laehnemann <david.laehnemann@hhu.de>
Date: Wed, 19 Jul 2017 17:12:19 +0200
Subject: [PATCH 2/5] rm absolute openmpi paths, use any present openmpi

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index e015f01..e6f901b 100644
--- a/Makefile
+++ b/Makefile
@@ -34,7 +34,7 @@
 #CPP = g++
 CPP = mpic++
 #CPPFLAGS = -m64 -fopenmp -O3 -s -static
-MPICPPFLAGS = -I/usr/lib/openmpi/include -I/usr/lib/openmpi/include/openmpi -pthread -L/usr/lib/openmpi/lib -lmpi_cxx -lmpi -lopen-rte -lopen-pal -ldl -lnsl -lutil -lm -ldl -Wl,--export-dynamic
+MPICPPFLAGS = -pthread -lmpi_cxx -lmpi -lopen-rte -lopen-pal -ldl -lnsl -lutil -lm -ldl -Wl,--export-dynamic
 CPPFLAGS = -m64 -openmp -pthread -O3 -DMAXK=$(MAXK) -DFIXEDPOINTPROB=1 $(MPICPPFLAGS)
 
 #Linux, gcc with openmp
-- 
2.7.4


From f20cdeba2e4fbd8e74a2590b58719dec8d508315 Mon Sep 17 00:00:00 2001
From: David Laehnemann <david.laehnemann@hhu.de>
Date: Wed, 19 Jul 2017 17:17:06 +0200
Subject: [PATCH 3/5] bug fix: include getopt.h in mdasim.h (args pars)

---
 include/mdasim.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/include/mdasim.h b/include/mdasim.h
index 17b3620..622ff53 100644
--- a/include/mdasim.h
+++ b/include/mdasim.h
@@ -43,6 +43,7 @@ Copyright 2012- Zeinab Taghavi (ztaghavi@wayne.edu)
 #include <utility>
 
 #include "commonMDA.h"
+#include "getopt.h"
 
 #define CoverageToPrimerConst 1000
 
-- 
2.7.4


From 13fc043a40da352c96a1c89ab81fbf139b919a3e Mon Sep 17 00:00:00 2001
From: David Laehnemann <david.laehnemann@hhu.de>
Date: Tue, 8 Aug 2017 19:13:56 +0200
Subject: [PATCH 4/5] use prefix in installation

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index e6f901b..d1c8a7d 100644
--- a/Makefile
+++ b/Makefile
@@ -50,7 +50,7 @@ LIBTOOL = ar
 
 PACKAGENAME = mdasim
 
-BASE = .
+BASE = $(prefix)
 OBJDIR = $(BASE)/obj
 BINDIR = $(BASE)/bin
 SRCDIR = $(BASE)/src
-- 
2.7.4


From 509fe2d04191591ad80fb9eafee4c42c9df341b4 Mon Sep 17 00:00:00 2001
From: David Laehnemann <david.laehnemann@hhu.de>
Date: Wed, 9 Aug 2017 13:59:10 +0200
Subject: [PATCH 5/5] differentiate between BASE and TARGET dirs

---
 Makefile | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/Makefile b/Makefile
index d1c8a7d..7bc8743 100644
--- a/Makefile
+++ b/Makefile
@@ -50,11 +50,13 @@ LIBTOOL = ar
 
 PACKAGENAME = mdasim
 
-BASE = $(prefix)
-OBJDIR = $(BASE)/obj
-BINDIR = $(BASE)/bin
+TARGET = $(prefix)
+OBJDIR = $(TARGET)/obj
+BINDIR = $(TARGET)/bin
+LIBDIR = $(TARGET)/lib
+
+BASE = ./
 SRCDIR = $(BASE)/src
-LIBDIR = $(BASE)/lib
 INCLUDEDIR = $(BASE)/include
 
 HYDALIB = $(LIBDIR)/lib$(PACKAGENAME).${LIBEXT}
-- 
2.7.4

