diff --git a/src/lib/map/Mapper.cpp b/src/lib/map/Mapper.cpp
index 6eaa2c5..781988c 100644
--- a/src/lib/map/Mapper.cpp
+++ b/src/lib/map/Mapper.cpp
@@ -22,6 +22,7 @@
 //#include "common/Crc32Hw.hpp"
 #include "common/DragenLogger.hpp"
 #include "map/Mapper.hpp"
+#include <boost/range/iterator_range.hpp>
 
 namespace dragenos {
 namespace map {
diff --git a/thirdparty/sswlib/ssw/ssw_avx2.cpp b/thirdparty/sswlib/ssw/ssw_avx2.cpp
index 8464392..50e3542 100644
--- a/thirdparty/sswlib/ssw/ssw_avx2.cpp
+++ b/thirdparty/sswlib/ssw/ssw_avx2.cpp
@@ -47,7 +47,8 @@ constexpr int AVX2_BYTE_ELEMS = 32;
 static inline __attribute__((always_inline)) void* memalign_local(const size_t alignment, const size_t size)
 {
   void* ptr = NULL;
-  BOOST_ASSERT(0 == posix_memalign(&ptr, alignment, size));
+  int rv = posix_memalign(&ptr, alignment, size);
+  BOOST_ASSERT(0 == rv);
   BOOST_ASSERT(ptr != nullptr);
   return ptr;
 }
