diff --git a/src/include/sam/SamGenerator.hpp b/src/include/sam/SamGenerator.hpp
index fe877fe..cca0f27 100644
--- a/src/include/sam/SamGenerator.hpp
+++ b/src/include/sam/SamGenerator.hpp
@@ -169,7 +169,7 @@ public:
       const std::string                 rgsm)
   {
     os << "@HD\tVN:1.4\tSO:unsorted\n";
-    os << "@PG\tID: DRAGEN-OS\tVN:" DRAGEN_OS_VERSION "\tCL:" << commandLine << "\n";
+    os << "@PG\tID:DRAGEN-OS\tVN:" DRAGEN_OS_VERSION "\tCL:" << commandLine << "\n";
     os << "@RG\tID:" << rgid << "\tLB:LB0\tPL:PL0\tPU:PU0\tSM:" << rgsm << "\n";
 
     // sequences must be generated in the original order but they are internally sorted by increasing
