diff --git a/ext/external.cmake b/ext/external.cmake
index ee3722a..8b2580b 100644
--- a/ext/external.cmake
+++ b/ext/external.cmake
@@ -8,7 +8,7 @@ if(NOT EXISTS ${PROJECT_SOURCE_DIR}/include/gum/internal/gfakluge.hpp)
   file(MAKE_DIRECTORY ${PROJECT_SOURCE_DIR}/include/gum/internal)
   set(GFAKluge_SOURCE_DIR ${PROJECT_SOURCE_DIR}/ext/gfakluge)
   ExternalProject_Add(gfakluge_git
-    DOWNLOAD_COMMAND git -C ${PROJECT_SOURCE_DIR} submodule update --init --recursive -- ${GFAKluge_SOURCE_DIR}
+    DOWNLOAD_COMMAND ""
     SOURCE_DIR ${GFAKluge_SOURCE_DIR}
     BUILD_IN_SOURCE 1
     CONFIGURE_COMMAND ${CMAKE_COMMAND} -E echo "Skipping configure step..."
@@ -23,7 +23,7 @@ if(NOT EXISTS ${PROJECT_SOURCE_DIR}/include/gum/internal/sdsl/config.hpp)
   file(MAKE_DIRECTORY ${PROJECT_SOURCE_DIR}/include/gum/internal/sdsl)
   set(SDSL_SOURCE_DIR ${PROJECT_SOURCE_DIR}/ext/sdsl-lite)
   ExternalProject_Add(sdsl_git
-    DOWNLOAD_COMMAND git -C ${PROJECT_SOURCE_DIR} submodule update --init --recursive -- ${SDSL_SOURCE_DIR}
+    DOWNLOAD_COMMAND ""
     SOURCE_DIR ${SDSL_SOURCE_DIR}
     CONFIGURE_COMMAND ${CMAKE_COMMAND} -E echo "Skipping configure step..."
     BUILD_COMMAND ${CMAKE_COMMAND} -E echo "Skipping build step..."
@@ -36,7 +36,7 @@ if(NOT EXISTS ${PROJECT_SOURCE_DIR}/include/gum/internal/parallel_hashmap/phmap.
   file(MAKE_DIRECTORY ${PROJECT_SOURCE_DIR}/include/gum/internal/parallel_hashmap)
   set(ParallelHashmap_SOURCE_DIR ${PROJECT_SOURCE_DIR}/ext/parallel-hashmap)
   ExternalProject_Add(parallelhashmap_git
-    DOWNLOAD_COMMAND git -C ${PROJECT_SOURCE_DIR} submodule update --init --recursive -- ${ParallelHashmap_SOURCE_DIR}
+    DOWNLOAD_COMMAND ""
     SOURCE_DIR ${ParallelHashmap_SOURCE_DIR}
     CONFIGURE_COMMAND ${CMAKE_COMMAND} -E echo "Skipping configure step..."
     BUILD_COMMAND ${CMAKE_COMMAND} -E echo "Skipping build step..."
