--- a/scripts/sf_gain_loss.py
+++ b/scripts/sf_gain_loss.py
@@ -2,10 +2,10 @@ from __future__ import print_function, division
 import os,sys,copy;import numpy as np
 from collections import defaultdict
 sys.path.append('./')
-from treetime.treetime import treeanc as ta
-from treetime.treetime.gtr import GTR
-from treetime.treetime import io
-from treetime.treetime import seq_utils
+from treetime import treeanc as ta
+from treetime.gtr import GTR
+from treetime import io
+from treetime import seq_utils
 from Bio import Phylo, AlignIO
 from sf_miscellaneous import write_json, write_pickle
 from sf_geneCluster_align_makeTree import load_sorted_clusters
