diff --git a/docs/databases.rst b/docs/databases.rst
index a28db18..aeea24a 100644
--- a/docs/databases.rst
+++ b/docs/databases.rst
@@ -32,16 +32,16 @@ This will download and format the databases, they can be displayed like so:
 	Funannotate Databases currently installed:
 
 	  Database          Type        Version      Date         Num_Records   Md5checksum                     
-	  pfam              hmmer3      32.0         2018-08            17929   de7496fad69c1040fd74db1cb5eef0fc
-	  gene2product      text        1.45         2019-07-31         30103   657bb30cf3247fcb74ca4f51a4ab7c18
-	  interpro          xml         76.0         2019-09-18         37113   328f66a791f9866783764f24a74a5aa3
-	  dbCAN             hmmer3      8.0          2019-08-08           607   51c724c1f9ac45687f08d0faa689ed58
-	  busco_outgroups   outgroups   1.0          2019-10-20             7   6795b1d4545850a4226829c7ae8ef058
-	  merops            diamond     12.0         2017-10-04          5009   a6dd76907896708f3ca5335f58560356
-	  mibig             diamond     1.4          2019-10-20         31023   118f2c11edde36c81bdea030a0228492
-	  uniprot           diamond     2019_09      2019-10-16        561176   9fc7871b8c4e3b755fe2086d77ed0645
-	  go                text        2019-10-07   2019-10-07         47375   3bc9ba43a98bf8fcd01db6e7e7813dd2
-	  repeats           diamond     1.0          2019-10-20         11950   4e8cafc3eea47ec7ba505bb1e3465d21
+          merops            diamond     12.5         2023-01-19          5098   6cd3c3dd85650394ce4e3dacb591f2a5
+          uniprot           diamond     2024_01      2024-01-24        570830   c7507ea16b3c4807971c663994cad329
+          dbCAN             hmmer3      11.0         2022-08-09           699   fb112af319a5001fbf547eac29e7c3b5
+          pfam              hmmer3      36.0         2023-07            20795   0725495ccf049a4f198fcc0a92f7f38c
+          repeats           diamond     1.0          2022-03-13         11950   4e8cafc3eea47ec7ba505bb1e3465d21
+          go                text        2024-01-17   2024-01-17         47729   7e6b9974184dda306e6e07631f1783af
+          mibig             diamond     1.4          2022-03-13         31023   118f2c11edde36c81bdea030a0228492
+          interpro          xml         98.0         2024-01-25         40768   502ea05009761b893dedb56d5ea89c48
+          busco_outgroups   outgroups   1.0          2024-03-04             8   6795b1d4545850a4226829c7ae8ef058
+          gene2product      text        1.92         2023-10-02         34459   32a4a80987720e0872377de3207dc0f5
 
 	To update a database type:
 		funannotate setup -i DBNAME -d $HOME/funannotate_db --force
diff --git a/funannotate/downloads.json b/funannotate/downloads.json
index b317d03..5d5df23 100644
--- a/funannotate/downloads.json
+++ b/funannotate/downloads.json
@@ -1,20 +1,20 @@
 {
   "downloads": {
-    "uniprot": "http://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz",
-    "uniprot-release": "http://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/complete/reldate.txt",
-    "merops": "http://ftp.ebi.ac.uk/pub/databases/merops/current_release/merops_scan.lib",
+    "uniprot": "https://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz",
+    "uniprot-release": "https://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/complete/reldate.txt",
+    "merops": "https://ftp.ebi.ac.uk/pub/databases/merops/current_release/meropsscan.lib",
     "dbCAN": "https://bcb.unl.edu/dbCAN2/download/Databases/V11/dbCAN-HMMdb-V11.txt",
     "dbCAN-tsv": "https://bcb.unl.edu/dbCAN2/download/Databases/V11/CAZyDB.08062022.fam-activities.txt",
     "dbCAN-log": "https://bcb.unl.edu/dbCAN2/download/Databases/V11/readme.txt",
-    "pfam": "http://ftp.ebi.ac.uk/pub/databases/Pfam/current_release/Pfam-A.hmm.gz",
-    "pfam-tsv": "http://ftp.ebi.ac.uk/pub/databases/Pfam/current_release/Pfam-A.clans.tsv.gz",
-    "pfam-log": "http://ftp.ebi.ac.uk/pub/databases/Pfam/current_release/Pfam.version.gz",
+    "pfam": "https://ftp.ebi.ac.uk/pub/databases/Pfam/current_release/Pfam-A.hmm.gz",
+    "pfam-tsv": "https://ftp.ebi.ac.uk/pub/databases/Pfam/current_release/Pfam-A.clans.tsv.gz",
+    "pfam-log": "https://ftp.ebi.ac.uk/pub/databases/Pfam/current_release/Pfam.version.gz",
     "outgroups": "https://osf.io/r9sne/download?version=1",
     "repeats": "https://osf.io/vp87c/download?version=1",
-    "go-obo": "http://purl.obolibrary.org/obo/go.obo",
+    "go-obo": "https://purl.obolibrary.org/obo/go.obo",
     "mibig": "https://dl.secondarymetabolites.org/mibig/mibig_prot_seqs_1.4.fasta",
-    "interpro": "http://ftp.ebi.ac.uk/pub/databases/interpro/current_release/interpro.xml.gz",
-    "interpro-tsv": "http://ftp.ebi.ac.uk/pub/databases/interpro/current_release/entry.list",
+    "interpro": "https://ftp.ebi.ac.uk/pub/databases/interpro/current_release/interpro.xml.gz",
+    "interpro-tsv": "https://ftp.ebi.ac.uk/pub/databases/interpro/current_release/entry.list",
     "gene2product": "https://raw.githubusercontent.com/nextgenusfs/gene2product/master/ncbi_cleaned_gene_products.txt"
   },
   "busco": {
@@ -131,4 +131,4 @@
       "embryophyta_odb9"
     ]
   }
-}
\ No newline at end of file
+}
diff --git a/funannotate/library.py b/funannotate/library.py
index 7df3688..073a8f0 100755
--- a/funannotate/library.py
+++ b/funannotate/library.py
@@ -64,7 +64,7 @@ class suppress_stdout_stderr(object):
     A context manager for doing a "deep suppression" of stdout and stderr in
     Python, i.e. will suppress all print, even if the print originates in a
     compiled C/Fortran sub-function.
-       This will not suppress raised exceptions, since exceptions are printed
+    This will not suppress raised exceptions, since exceptions are printed
     to stderr just before a script exits, and after the context manager has
     exited (at least, I think that is why it lets exceptions through).
 
@@ -3010,9 +3010,13 @@ def annotation_summary(
         stats["annotation"]["transcript-level"]["avg_protein_length"] = round(
             sum(protLengths) / float(len(protLengths)), 2
         )
-        stats["annotation"]["transcript-level"]["avg_exon_length"] = round(
-            sum(exonLengths) / float(len(exonLengths)), 2
-        )
+        if len(exonLengths) > 0:
+            stats["annotation"]["transcript-level"]["avg_exon_lenth"] = round(
+                sum(exonLengths) / float(len(exonLengths)), 2
+            )
+        else:
+            stats["annotation"]["transcript-level"]["avg_exon_length"] = -1
+
         exonBED = "tmp.exon.{}.bed".format(os.getpid())
         if transcripts or proteins:
             exonCount = 0
@@ -5322,7 +5326,7 @@ def gff2dict(file, fasta, Genes, debug=False, gap_filter=False):
         for i in range(0, len(v["ids"])):
             if k not in Genes:  # capture if gene was removed
                 continue
-            if v["type"] in ["mRNA", "tRNA", "ncRNA", "rRNA"]:
+            if v["type"] in ["mRNA", "tRNA", "ncRNA", "rRNA", "transcript"]:
                 if v["strand"] == "+":
                     sortedExons = sorted(v["mRNA"][i], key=lambda tup: tup[0])
                 else:
@@ -7396,12 +7400,13 @@ def parsePhobiusSignalP(phobius, sigP, membrane_annot, secretome_annot):
                     if "\t" in line:
                         cols = line.split("\t")
                         if cols[1] != "OTHER":  # then signal peptide
-                            ID, prediction, score1, score2, position = cols[:5]
-                            ID = ID.split(" ")[0]
-                            components = position.split()
-                            pos = components[2].split("-")[0]
-                            prob = components[-1]
-                            sigpDict[ID] = [pos, "", prob]
+                            if len(cols) >= 5:
+                                ID, prediction, score1, score2, position = cols[:5]
+                                ID = ID.split(" ")[0]
+                                components = position.split()
+                                pos = components[2].split("-")[0]
+                                prob = components[-1]
+                                sigpDict[ID] = [pos, "", prob]
     else:
         sigpDict = pSecDict
     # write annotation files
@@ -8278,9 +8283,16 @@ def runSnap(fasta, gff3, minintron, maxintron, dir, output):
         trainingFasta = os.path.join(dir, "snap-training.scaffolds.fasta")
         with open(trainingFasta, "w") as outfile:
             with open(os.path.abspath(fasta), "r") as infile:
+                seqids = {}
                 for title, seq in SimpleFastaParser(infile):
                     if title in list(scaff2genes.keys()):
-                        outfile.write(">{:}\n{:}\n".format(title, softwrap(seq)))
+                        seqids[title] = seq
+            # need to write out the fasta file in a sorted way otherwise the ZFF and FASTA don't match
+            # cannot assume the input genome is actual in ID sorted order
+            # see bug #1060
+            for title in natsorted(seqids.keys()):
+                outfile.write(">{:}\n{:}\n".format(title, softwrap(seqids[title])))
+
 
         # convert to ZFF format
         origzff = os.path.join(dir, "snap.training.zff")
diff --git a/funannotate/resources.py b/funannotate/resources.py
index a3ceb14..3367cce 100644
--- a/funannotate/resources.py
+++ b/funannotate/resources.py
@@ -196,7 +196,7 @@ COGS = {
 DBURL = {
     "uniprot": "https://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz",
     "uniprot-release": "https://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/complete/reldate.txt",
-    "merops": "https://ftp.ebi.ac.uk/pub/databases/merops/current_release/merops_scan.lib",
+    "merops": "https://ftp.ebi.ac.uk/pub/databases/merops/current_release/meropsscan.lib",
     "dbCAN": "https://bcb.unl.edu/dbCAN2/download/Databases/V11/dbCAN-HMMdb-V11.txt",
     "dbCAN-tsv": "https://bcb.unl.edu/dbCAN2/download/Databases/V11/CAZyDB.08062022.fam-activities.txt",
     "dbCAN-log": "https://bcb.unl.edu/dbCAN2/download/Databases/V11/readme.txt",
diff --git a/funannotate/setupDB.py b/funannotate/setupDB.py
index 5b77c2d..b9389f7 100755
--- a/funannotate/setupDB.py
+++ b/funannotate/setupDB.py
@@ -138,7 +138,7 @@ def wget(url, name):
 
 
 def meropsDB(info, force=False, args={}):
-    fasta = os.path.join(FUNDB, 'merops_scan.lib')
+    fasta = os.path.join(FUNDB, 'meropsscan.lib')
     filtered = os.path.join(FUNDB, 'merops.formatted.fa')
     database = os.path.join(FUNDB, 'merops.dmnd')
     if os.path.isfile(fasta) and args.update and not force:
@@ -170,8 +170,8 @@ def meropsDB(info, force=False, args={}):
                'merops.formatted.fa', '--db', 'merops']
         lib.runSubprocess(cmd, os.path.join(FUNDB), lib.log)
         num_records = lib.countfasta(filtered)
-        info['merops'] = ('diamond', database, '12.0',
-                          '2017-10-04', num_records, md5)
+        info['merops'] = ('diamond', database, '12.5',
+                          '2023-01-19', num_records, md5)
     type, name, version, date, records, checksum = info.get('merops')
     lib.log.info('MEROPS Database: version={:} date={:} records={:,}'.format(
         version, date, records))
diff --git a/setup.py b/setup.py
index 3044062..f69caba 100755
--- a/setup.py
+++ b/setup.py
@@ -17,7 +17,7 @@ DESCRIPTION = "funannotate: eukaryotic genome annotation pipeline"
 URL = "https://github.com/nextgenusfs/funannotate"
 EMAIL = "nextgenusfs@gmail.com"
 AUTHOR = "Jon Palmer"
-REQUIRES_PYTHON = ">=3.6.0, <3.10"
+REQUIRES_PYTHON = ">=3.6.0, <3.12"
 VERSION = None
 
 # What packages are required for this module to be executed?
diff --git a/funannotate/aux_scripts/funannotate-BUSCO2-py2.py b/funannotate/aux_scripts/funannotate-BUSCO2-py2.py
index 538cf96..424ffd1 100755
--- a/funannotate/aux_scripts/funannotate-BUSCO2-py2.py
+++ b/funannotate/aux_scripts/funannotate-BUSCO2-py2.py
@@ -62,7 +62,7 @@ def which_path(file_name):
 try:
     AUGUSTUS = os.environ["AUGUSTUS_CONFIG_PATH"]
 except KeyError:
-    print "AUGUSTUS_CONFIG_PATH environmental variable not set, exiting"
+    print("AUGUSTUS_CONFIG_PATH environmental variable not set, exiting")
     sys.exit(1)
 
 if AUGUSTUS.endswith('config'):
