diff --git a/Cluster.h b/Cluster.h
index c792138..5758542 100644
--- a/Cluster.h
+++ b/Cluster.h
@@ -32,6 +32,7 @@
 #include <map>
 #include <climits>
 #include <Read.h>
+#include <cstdint>

 using namespace std;

diff --git a/Read.h b/Read.h
index 592409b..49a15b8 100644
--- a/Read.h
+++ b/Read.h
@@ -23,6 +23,7 @@
 #include <fstream>
 #include <iostream>
 #include <Transcript.h>
+#include <cstdint>

 using namespace std;
