diff --git a/trtools/testsupport/test_trtools.sh b/trtools/testsupport/test_trtools.sh
index 67eb46d5..09e1bf3b 100755
--- a/trtools/testsupport/test_trtools.sh
+++ b/trtools/testsupport/test_trtools.sh
@@ -15,7 +15,7 @@ if [ ! -d "$TMP" ] ; then
 	echo "Downloading test data ..."
 	mkdir $TMP
 	pushd $TMP
-	git clone -b v$(dumpSTR --version) https://github.com/gymrek-lab/TRTools.git
+	git clone -b v$(dumpSTR --version) https://github.com/gymrek-lab/TRTools.git .
 	popd
 	echo "Download done"
 else
