*** src/map/include/parseCmdArgs.hpp	2023-07-28 21:41:45.000000000 +0100
--- src/map/include/parseCmdArgs.hpp.new	2024-11-25 21:28:40.000000000 +0000
*************** namespace skch
*** 193,199 ****
  
      if (versioncheck)
      {
!       std::cerr << "version 1.33\n\n";
        exit(0);
      }
  
--- 193,199 ----
  
      if (versioncheck)
      {
!       std::cerr << "version 1.34\n\n";
        exit(0);
      }
  
