diff --git i/Makefile.in w/Makefile.in
index e789517..9682125 100644
--- Makefile.in
+++ Makefile.in
@@ -2,10 +2,11 @@ CXXFLAGS += -O3 -DNDEBUG -std=c++11 -Isrc -I @mathinc@ @OPENMP_CXXFLAGS@
 CPPFLAGS += @amcppflags@
 
 UNAME_S=$(shell uname -s)
+UNAME_M=$(shell uname -m)
 
 ifeq ($(UNAME_S),Darwin)  #macOS clang
        CXXFLAGS += -mmacosx-version-min=10.7 -stdlib=libc++ -Xpreprocessor -fopenmp -lomp
-else
+elif ($(UNAME_M),x86_64)
        CXXFLAGS += -include src/common/memcpyLink.h -Wl,--wrap=memcpy
        CFLAGS += -include src/common/memcpyLink.h
 endif
