--- Makefile.in	2024-04-27 14:58:06.238060909 +0300
+++ Makefile.in.new	2024-04-27 14:58:01.910293221 +0300
@@ -2,10 +2,11 @@
 CPPFLAGS += @amcppflags@
 
 UNAME_S=$(shell uname -s)
+UNAME_M=$(shell uname -m)
 
 ifeq ($(UNAME_S),Darwin)  #macOS clang
 	CXXFLAGS += -mmacosx-version-min=10.7 -stdlib=libc++ -Xpreprocessor -fopenmp -lomp
-else
+elif ($(UNAME_M),x86_64)
 	CXXFLAGS += -include src/common/memcpyLink.h -Wl,--wrap=memcpy
 	CFLAGS += -include src/common/memcpyLink.h
 endif

