diff --git a/cmake/xxdi.pl b/cmake/xxdi.pl
index acc974c4..f6116b6c 100755
--- a/cmake/xxdi.pl
+++ b/cmake/xxdi.pl
@@ -14,7 +14,15 @@
 
 use strict;
 use warnings;
-use File::Slurp qw(slurp);
+
+sub slurp {
+    my $file = shift;
+    open my $fh, '<', $file or die;
+    local $/ = undef;
+    my $cont = <$fh>;
+    close $fh;
+    return $cont;
+}
 
 my $indata = slurp(@ARGV ? $ARGV[0] : \*STDIN);
 my $len_data = length($indata);
