--- Makefile.orig	2017-03-23 18:26:58.000000000 -0600
+++ Makefile	2017-03-23 19:01:36.000000000 -0600
@@ -213,9 +213,9 @@
 #----------------------

 ifeq ($(DARWIN),1)
- LINK_STATIC=1# link static
+ LINK_STATIC?=1# link static
 else
- LINK_STATIC=0# link dynamically
+ LINK_STATIC?=0# link dynamically
 #  LINK_STATIC=1# link static (testing only)
 endif

@@ -770,6 +770,10 @@

 #---------------------- differences between linking executables and shared libs:

+cflags += $(CFLAGS)
+clflags += $(LDFLAGS)
+cxxflags += $(CXXFLAGS)
+
 # executables:
 ifeq ($(DARWIN),1)
  blflags:=$(clflags)
