--- Makefile.orig	2017-03-23 18:26:58.000000000 -0600
+++ Makefile	2017-03-23 19:01:36.000000000 -0600
@@ -213,9 +213,9 @@
 #----------------------

 ifeq ($(DARWIN),1)
- LINK_STATIC=1# link static
+ LINK_STATIC?=1# link static
 else
- LINK_STATIC=0# link dynamically
+ LINK_STATIC?=0# link dynamically
 #  LINK_STATIC=1# link static (testing only)
 endif

@@ -797,6 +797,10 @@
  cxxflags += $(CPPFLAGS)
 endif

+cflags += $(CFLAGS)
+clflags += $(LDFLAGS)
+cxxflags += $(CXXFLAGS)
+
 ifeq ('$(USE_CLANG)','1')
 # none of the following standards works with clang3.3 (under linux) => dont use standard (as done before r15516)
 #  cxxflags += -std=gnu++11

