diff --git a/src/juncs_db.cpp b/src/juncs_db.cpp
index ecdeba1..2efcf91 100644
--- a/src/juncs_db.cpp
+++ b/src/juncs_db.cpp
@@ -338,7 +338,7 @@ void driver(const vector<FILE*>& splice_coords_files,
 			uint32_t left_coord = atoi(scan_left_coord);
 			uint32_t right_coord = atoi(scan_right_coord);
 			bool antisense = *orientation == '-';
-			junctions.insert(make_pair<Junction, JunctionStats>(Junction(ref_id, left_coord, right_coord, antisense), JunctionStats()));
+			junctions.insert(make_pair(Junction(ref_id, left_coord, right_coord, antisense), JunctionStats()));
 		}
 	}
 
diff --git a/src/tophat_reports.cpp b/src/tophat_reports.cpp
index 50d2165..ce68ace 100644
--- a/src/tophat_reports.cpp
+++ b/src/tophat_reports.cpp
@@ -2705,7 +2705,7 @@ void driver(const string& bam_output_fname,
 				junction_stat.gtf_match = true;
 				junction_stat.accepted = true;
 
-				gtf_junctions.insert(make_pair<Junction, JunctionStats>(Junction(ref_id, left_coord, right_coord, antisense), junction_stat));
+				gtf_junctions.insert(make_pair(Junction(ref_id, left_coord, right_coord, antisense), junction_stat));
 			}
 		}
 		fprintf(stderr, "Loaded %d GFF junctions from %s.\n", (int)(gtf_junctions.size()), gtf_juncs.c_str());
