diff --git a/src/samtools-0.1.18/Makefile b/src/samtools-0.1.18/Makefile
index 00927f8..68ae65c 100644
--- a/src/samtools-0.1.18/Makefile
+++ b/src/samtools-0.1.18/Makefile
@@ -1,5 +1,5 @@
-CC=			gcc
-CFLAGS=		-g -Wall -O2 #-m64 #-arch ppc
+CC?=			gcc
+CFLAGS=		-g -Wall -O3 -I$(PREFIX)/include -L$(PREFIX)/lib  #-m64 #-arch ppc
 DFLAGS=		-D_FILE_OFFSET_BITS=64 -D_LARGEFILE64_SOURCE -D_USE_KNETFILE -D_CURSES_LIB=0
 KNETFILE_O=	knetfile.o
 LOBJS=		bgzf.o kstring.o bam_aux.o bam.o bam_import.o sam.o bam_index.o	\
@@ -10,10 +10,10 @@ AOBJS=		bam_tview.o bam_plcmd.o sam_view.o	\
 			bamtk.o kaln.o bam2bcf.o bam2bcf_indel.o errmod.o sample.o \
 			cut_target.o phase.o bam2depth.o
 PROG=		samtools_0.1.18
-INCLUDES=	-I.
+INCLUDES=	-I. -I$(PREFIX)/include -I$(PREFIX)/include/ncurses
 SUBDIRS=	. bcftools
-LIBPATH=
-LIBCURSES=	-lcurses
+LIBPATH=	-L$(PREFIX)/lib
+LIBCURSES=	-L$(PREFIX)/lib -lncurses -ltinfo
 
 .SUFFIXES:.c .o
 
