From 76ca9a56c9d95d50a3bd254807b2df6f3c963c25 Mon Sep 17 00:00:00 2001
From: John Lees <lees.john6@gmail.com>
Date: Tue, 4 Oct 2022 17:02:10 +0100
Subject: [PATCH] patch for conda OS X

---
 CMakeLists.txt | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 297f258..c3b28e1 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -5,7 +5,9 @@ add_compile_definitions(PYTHON_EXT)
 
 set(CMAKE_CXX_STANDARD 17)
 # older compilers use experimental::filesystem
-if(NOT $ENV{UC_INSTALL} STREQUAL "conda" AND CMAKE_COMPILER_IS_GNUCC)
+if($ENV{UC_INSTALL} STREQUAL "conda" AND CMAKE_COMPILER_IS_CLANG)
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DFS_EXP")
+elseif(NOT $ENV{UC_INSTALL} STREQUAL "conda" AND CMAKE_COMPILER_IS_GNUCC)
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DFS_EXP -fconcepts")
 endif()
 
@@ -71,6 +73,8 @@ else()
     endif()
 endif()
 # older compilers use experimental::filesystem
-if(NOT $ENV{UC_INSTALL} STREQUAL "conda" AND CMAKE_COMPILER_IS_GNUCC)
+if($ENV{UC_INSTALL} STREQUAL "conda" AND CMAKE_COMPILER_IS_CLANG)
+    target_link_libraries("${TARGET_NAME}" PRIVATE c++fs)
+elseif(NOT $ENV{UC_INSTALL} STREQUAL "conda" AND CMAKE_COMPILER_IS_GNUCC)
     target_link_libraries("${TARGET_NAME}" PRIVATE stdc++fs)
 endif()
-- 
2.30.1 (Apple Git-130)

