From 2c01cf4c535882bf1f773e0f128970ea05885777 Mon Sep 17 00:00:00 2001
From: John Lees <lees.john6@gmail.com>
Date: Thu, 6 Oct 2022 10:14:59 +0100
Subject: [PATCH 2/2] Change linker for filesystem path

---
 CMakeLists.txt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index cd50eca..f1dcf9e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -75,7 +75,7 @@ else()
 endif()
 # older compilers use experimental::filesystem
 if("$ENV{UC_INSTALL}" STREQUAL "conda" AND CMAKE_CXX_COMPILER_ID MATCHES "Clang")
-    target_link_libraries("${TARGET_NAME}" PRIVATE c++fs)
+    target_link_libraries("${TARGET_NAME}" PRIVATE c++experimental)
 elseif(NOT "$ENV{UC_INSTALL}" STREQUAL "conda" AND CMAKE_CXX_COMPILER_ID MATCHES "GCC")
     target_link_libraries("${TARGET_NAME}" PRIVATE stdc++fs)
 endif()
\ No newline at end of file
-- 
2.30.1 (Apple Git-130)

