--- c++/augment-bam.cpp    2016-09-17 15:55:19.584598580 +0000
+++ c++/augment-bam.cpp.new    2016-09-17 15:54:45.333095246 +0000
@@ -44,10 +44,10 @@
 #include "pstreams/pstream.h"
 #endif
 
-#include "cansam/sam/alignment.h"
-#include "cansam/sam/header.h"
-#include "cansam/sam/stream.h"
-#include "cansam/exception.h"
+#include <cansam/sam/alignment.h>
+#include <cansam/sam/header.h>
+#include <cansam/sam/stream.h>
+#include <cansam/exception.h>
 
 #include "version.inc"
 
