--- src/c++/programs/fnj/PhylipDmInputStream.cpp
+++ src/c++/programs/fnj/PhylipDmInputStream.cpp
@@ -48 +48 @@
-	if (getline(*fp,line)==0)
+	if (!getline(*fp,line))

--- src/c++/programs/fnj/BinaryInputStream.cpp
+++ src/c++/programs/fnj/BinaryInputStream.cpp
@@ -81 +81 @@
-			if (fp->read( reinterpret_cast<char*>( &f ), sizeof(f))==0)
+			if (!fp->read( reinterpret_cast<char*>( &f ), sizeof(f)))
