*** T-COFFEE_distribution_Version_13.46.0.919e8c6b/t_coffee_source/util_lib/util.c	2023-07-07 22:06:54.000000000 +0000
--- T-COFFEE_distribution_Version_13.46.0.919e8c6b/t_coffee_source/util_lib/util.c	2025-01-16 00:27:11.959164302 +0000
*************** int getNumCores() {
*** 5628,5634 ****
  
  static int nproc;
  
! int set_nproc(int np)
  {
  
    //set to 0 if to be reset by environement
--- 5628,5634 ----
  
  static int nproc;
  
! void set_nproc(int np)
  {
  
    //set to 0 if to be reset by environement
*** T-COFFEE_distribution_Version_13.46.0.919e8c6b/t_coffee_source/util_lib/util.h	2023-07-07 22:06:54.000000000 +0000
--- T-COFFEE_distribution_Version_13.46.0.919e8c6b/t_coffee_source/util_lib/util.h	2025-01-16 00:28:11.108005343 +0000
*************** cache: idem
*** 586,592 ****
  methods: idem
  mcoffee: idem
  */
! int set_nproc (int nproc);
  int get_nproc ();
  char *get_os();
  char *get_lockdir_4_tcoffee ();
--- 586,592 ----
  methods: idem
  mcoffee: idem
  */
! void set_nproc (int nproc);
  int get_nproc ();
  char *get_os();
  char *get_lockdir_4_tcoffee ();
