diff -ur a/t_coffee_source/util_lib/util.c b/t_coffee_source/util_lib/util.c
--- a/t_coffee_source/util_lib/util.c	2023-07-07 22:06:54.000000000 +0000
+++ b/t_coffee_source/util_lib/util.c	2025-01-16 00:27:11.959164302 +0000
@@ -5628,7 +5628,7 @@
 
 static int nproc;
 
-int set_nproc(int np)
+void set_nproc(int np)
 {
 
   //set to 0 if to be reset by environement
diff -ur a/t_coffee_source/util_lib/util.h b/t_coffee_source/util_lib/util.h
--- a/t_coffee_source/util_lib/util.h	2023-07-07 22:06:54.000000000 +0000
+++ b/t_coffee_source/util_lib/util.h	2025-01-16 00:28:11.108005343 +0000
@@ -586,7 +586,7 @@
 methods: idem
 mcoffee: idem
 */
-int set_nproc (int nproc);
+void set_nproc (int nproc);
 int get_nproc ();
 char *get_os();
 char *get_lockdir_4_tcoffee ();
