--- CMakeLists.txt	2019-06-08 10:35:50.000000000 +0200
+++ CMakeLists.txt	2019-07-24 14:42:32.000000000 +0200
@@ -1,8 +1,6 @@
 cmake_minimum_required (VERSION 3.2)
 project (HTStream)

-option(BUILD_STATIC_BIN "builds static binary (linux only)" OFF)
-
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 ")
 #set(CMAKE_FIND_LIBRARY_SUFFIXES ".a")
 set(BUILD_SHARED_LIBS OFF)
@@ -10,9 +8,7 @@

 if (APPLE)
 elseif (UNIX)
-  if (BUILD_STATIC_BIN)
-    set(EXTRA_BUILD_FLAGS "-static")
-  endif()
+  set(EXTRA_BUILD_FLAGS "-static")
 endif()

 if (${CMAKE_CXX_COMPILER_ID} STREQUAL "GNU")
