--- src/dawg.cpp
+++ src/dawg.cpp
@@ -140,3 +140,3 @@
-	bool split  = arg.split || (indeterminate(arg.split) && glopts.output_split);
-	bool append = arg.append || (indeterminate(arg.append) && glopts.output_append);
-	bool label  = arg.label || (indeterminate(arg.label) && glopts.output_label);
+	bool split  = static_cast<bool>(arg.split) || (indeterminate(arg.split) && glopts.output_split);
+	bool append = static_cast<bool>(arg.append) || (indeterminate(arg.append) && glopts.output_append);
+	bool label  = static_cast<bool>(arg.label) || (indeterminate(arg.label) && glopts.output_label);
