--- Makefile.PL	2016-06-21 12:50:27.000000000 +0200
+++ Makefile.PL	2018-08-07 14:25:24.383061252 +0200
@@ -70,7 +70,7 @@
           
           runtime => {
             requires => {
-              'XML::Parser'              =>    '2.23',
+              'XML::Parser'              =>    '2.44',
             },
             recommends => {
               'Scalar::Util'             =>    '1.23',
@@ -98,7 +98,7 @@
     LICENSE      => 'perl',
     EXE_FILES    => [ map {"tools/$_/$_"} @programs],
     VERSION_FROM => 'Twig.pm', 
-    PREREQ_PM    => { 'XML::Parser' => '2.23' },
+    PREREQ_PM    => { 'XML::Parser' => '2.44' },
     dist         => { COMPRESS    => 'gzip -9f',
                       SUFFIX	     => 'gz',
                     },
--- META.json	2016-11-23 18:19:22.000000000 +0100
+++ META.json	2018-08-07 14:25:24.384061241 +0200
@@ -32,7 +32,7 @@
       },
       "runtime" : {
          "requires" : {
-            "XML::Parser" : "2.23"
+            "XML::Parser" : "2.44"
          }
       }
    },
--- META.yml	2016-11-23 18:19:22.000000000 +0100
+++ META.yml	2018-08-07 14:25:24.385061231 +0200
@@ -18,7 +18,7 @@
     - t
     - inc
 requires:
-  XML::Parser: '2.23'
+  XML::Parser: '2.44'
 resources:
   repository: http://github.com/mirod/xmltwig
 version: '3.52'
--- t/test_errors.t	2016-06-21 12:50:27.000000000 +0200
+++ t/test_errors.t	2018-08-07 14:25:20.216103809 +0200
@@ -21,7 +21,7 @@
 my( $q, $q2) = ( ($^O eq "MSWin32") || ($^O eq 'VMS') ) ? ('"', "'") : ("'", '"');
 
 { # test insufficient version of XML::Parser (not that easy, it is already too late here)
-my $need_version= 2.23;
+my $need_version= 2.44;
 
 
 use Config;
--- Twig.pm	2016-11-23 18:19:10.000000000 +0100
+++ Twig.pm	2018-08-07 15:01:47.359248242 +0200
@@ -149,7 +149,7 @@
 $VERSION = '3.52';
 
 use XML::Parser;
-my $needVersion = '2.23';
+my $needVersion = '2.44';
 ($parser_version= $XML::Parser::VERSION)=~ s{_\d+}{}; # remove _<n> from version so numeric tests do not warn
 croak "need at least XML::Parser version $needVersion" unless $parser_version >= $needVersion;
 
--- Twig_pm.slow	2016-11-23 18:17:39.000000000 +0100
+++ Twig_pm.slow	2018-08-07 15:01:49.899222499 +0200
@@ -149,7 +149,7 @@
 $VERSION = '3.52';
 
 use XML::Parser;
-my $needVersion = '2.23';
+my $needVersion = '2.44';
 ($parser_version= $XML::Parser::VERSION)=~ s{_\d+}{}; # remove _<n> from version so numeric tests do not warn
 croak "need at least XML::Parser version $needVersion" unless $parser_version >= $needVersion;
 
