--- Makefile
+++ Makefile
@@ -8 +8 @@
-	g++ -g -pthread -I$(SRCDIR) -I$(JAVA_HOME)/include -I$(JAVA_HOME)/include/linux -I$(SAMTOOLS) -I$(HTSLIB)  $(SRCDIR)/assembler2_vdj.c $(SRCDIR)/seq_score.c $(SRCDIR)/vj_filter.c $(SRCDIR)/seq_to_kmer.c $(SRCDIR)/hash_utils.c $(SRCDIR)/bam_read.c $(SRCDIR)/quick_map3.c $(SRCDIR)/coverage.c $(SRCDIR)/status.c $(SRCDIR)/params.c $(SAMTOOLS)/libbam.a $(HTSLIB)/libhts.a -lz -lpthread -o vdjer
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) $(LDFLAGS) -g -pthread -I$(SRCDIR) -I$(JAVA_HOME)/include -I$(JAVA_HOME)/include/linux -I$(SAMTOOLS) -I$(HTSLIB)  $(SRCDIR)/assembler2_vdj.c $(SRCDIR)/seq_score.c $(SRCDIR)/vj_filter.c $(SRCDIR)/seq_to_kmer.c $(SRCDIR)/hash_utils.c $(SRCDIR)/bam_read.c $(SRCDIR)/quick_map3.c $(SRCDIR)/coverage.c $(SRCDIR)/status.c $(SRCDIR)/params.c $(SAMTOOLS)/libbam.a $(HTSLIB)/libhts.a -lz -lpthread -o vdjer
