diff -ru ViennaRNA-2.7.0/interfaces/Python/Makefile.am ViennaRNA-2.7.0/interfaces/Python/Makefile.am
--- ViennaRNA-2.7.0/interfaces/Python/Makefile.am	2024-10-19 22:19:42.000000000 +0200
+++ ViennaRNA-2.7.0/interfaces/Python/Makefile.am	2024-10-21 22:32:45.262117727 +0200
@@ -160,9 +160,7 @@
   PYTHONPATH=".:$$PYTHONPATH" $(PYTHON3) -O -c "from RNA import RNA" )
 
 RNA/_RNA$(PYTHON3_SO): _RNA.la
-	$(AM_V_GEN)( $(LIBTOOL) --config > libtoolconfig.tmp ) && \
-  ( . ./libtoolconfig.tmp; cp $$objdir/_RNA$(PYTHON3_SO) RNA ) && \
-  rm -f libtoolconfig.tmp
+	$(AM_V_GEN)( ditto `$(LIBTOOL) --config | $(SED) -n -e 's/^objdir=\(.*\)$$/\1/p'`/_RNA$(PYTHON3_SO) RNA/_RNA$(PYTHON3_SO) || cp `$(LIBTOOL) --config | $(SED) -n -e 's/^objdir=\(.*\)$$/\1/p'`/_RNA$(PYTHON3_SO) RNA/_RNA$(PYTHON3_SO) )
 
 
 CLEANFILES = \
diff -ru ViennaRNA-2.7.0_orig/interfaces/Python/Makefile.in ViennaRNA-2.7.0/interfaces/Python/Makefile.in
--- ViennaRNA-2.7.0/interfaces/Python/Makefile.in	2024-10-20 12:27:41.000000000 +0200
+++ ViennaRNA-2.7.0/interfaces/Python/Makefile.in	2024-10-21 22:31:59.809208993 +0200
@@ -1224,9 +1224,7 @@
   PYTHONPATH=".:$$PYTHONPATH" $(PYTHON3) -O -c "from RNA import RNA" )
 
 RNA/_RNA$(PYTHON3_SO): _RNA.la
-	$(AM_V_GEN)( $(LIBTOOL) --config > libtoolconfig.tmp ) && \
-  ( . ./libtoolconfig.tmp; cp $$objdir/_RNA$(PYTHON3_SO) RNA ) && \
-  rm -f libtoolconfig.tmp
+	$(AM_V_GEN)( ditto `$(LIBTOOL) --config | $(SED) -n -e 's/^objdir=\(.*\)$$/\1/p'`/_RNA$(PYTHON3_SO) RNA/_RNA$(PYTHON3_SO) || cp `$(LIBTOOL) --config | $(SED) -n -e 's/^objdir=\(.*\)$$/\1/p'`/_RNA$(PYTHON3_SO) RNA/_RNA$(PYTHON3_SO) )
 
 all-local:  RNA/_RNA$(PYTHON3_SO)
 
