diff --git a/src/RNAlocmin/hash_util.h b/src/RNAlocmin/hash_util.h
index e5dcfd3..50a0138 100644
--- a/src/RNAlocmin/hash_util.h
+++ b/src/RNAlocmin/hash_util.h
@@ -79,10 +79,10 @@ struct hash_eq {
 struct hash_fncts{
   size_t operator()(const Structure &x) const {
 
-  register short *k;        /* the key */
-  register unsigned  length;   /* the length of the key */
-  register unsigned  initval=0;  /* the previous hash, or an arbitrary value */
-  register unsigned a,b,c,len;
+  short *k;        /* the key */
+  unsigned  length;   /* the length of the key */
+  unsigned  initval=0;  /* the previous hash, or an arbitrary value */
+  unsigned a,b,c,len;
 
   /* Set up the internal state */
   k = x.str;
@@ -123,10 +123,10 @@ struct hash_fncts{
 
   size_t operator()(const struct_en &x) const {
 
-  register short *k;        /* the key */
-  register unsigned  length;   /* the length of the key */
-  register unsigned  initval=0;  /* the previous hash, or an arbitrary value */
-  register unsigned a,b,c,len;
+  short *k;        /* the key */
+  unsigned  length;   /* the length of the key */
+  unsigned  initval=0;  /* the previous hash, or an arbitrary value */
+  unsigned a,b,c,len;
 
   /* Set up the internal state */
   k = x.structure;
@@ -165,10 +165,10 @@ struct hash_fncts{
   return (c & HASHSIZE);
   }
   size_t operator()(const Structure *x) const {
-  register short *k;        /* the key */
-  register unsigned  length;   /* the length of the key */
-  register unsigned  initval=0;  /* the previous hash, or an arbitrary value */
-  register unsigned a,b,c,len;
+  short *k;        /* the key */
+  unsigned  length;   /* the length of the key */
+  unsigned  initval=0;  /* the previous hash, or an arbitrary value */
+  unsigned a,b,c,len;
 
   /* Set up the internal state */
   k = x->str;
@@ -209,10 +209,10 @@ struct hash_fncts{
 
   size_t operator()(const struct_en *x) const {
 
-  register short *k;        /* the key */
-  register unsigned  length;   /* the length of the key */
-  register unsigned  initval=0;  /* the previous hash, or an arbitrary value */
-  register unsigned a,b,c,len;
+  short *k;        /* the key */
+  unsigned  length;   /* the length of the key */
+  unsigned  initval=0;  /* the previous hash, or an arbitrary value */
+  unsigned a,b,c,len;
 
   /* Set up the internal state */
   k = x->structure;
