From 28d007d44f4683ce430e8b29d184977ff488007c Mon Sep 17 00:00:00 2001
From: Chris Markiewicz <effigies@gmail.com>
Date: Sun, 13 Oct 2024 11:29:03 -0400
Subject: [PATCH] Import cstdint into libCZI.h

---
 src/CZIlib/CZI/libCZI.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/CZIlib/CZI/libCZI.h b/src/CZIlib/CZI/libCZI.h
index 7799f1aec..6a5c3e047 100644
--- a/src/CZIlib/CZI/libCZI.h
+++ b/src/CZIlib/CZI/libCZI.h
@@ -22,6 +22,7 @@
 
 #pragma once
 
+#include <cstdint>
 #include "ImportExport.h"
 
 #include "priv_guiddef.h"
-- 
2.45.2

