From cc3d4d74235978b136f637620032595cb388933b Mon Sep 17 00:00:00 2001
From: Chris Markiewicz <effigies@gmail.com>
Date: Mon, 14 Oct 2024 13:49:05 -0400
Subject: [PATCH] Fix unsafe narrowing

---
 src/Algorithms/AlgorithmCiftiSeparate.cxx | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/Algorithms/AlgorithmCiftiSeparate.cxx b/src/Algorithms/AlgorithmCiftiSeparate.cxx
index 9a64a262e..f49c04e73 100644
--- a/src/Algorithms/AlgorithmCiftiSeparate.cxx
+++ b/src/Algorithms/AlgorithmCiftiSeparate.cxx
@@ -654,7 +654,7 @@ void AlgorithmCiftiSeparate::getCroppedVolSpace(const CiftiFile* ciftiIn, const
     int64_t numVoxels = (int64_t)myMap.size();
     if (numVoxels > 0)
     {//make a voxel bounding box to minimize memory usage
-        int extrema[6] = { myMap[0].m_ijk[0],
+        int64_t extrema[6] = { myMap[0].m_ijk[0],
             myMap[0].m_ijk[0],
             myMap[0].m_ijk[1],
             myMap[0].m_ijk[1],
@@ -699,7 +699,7 @@ void AlgorithmCiftiSeparate::getCroppedVolSpaceAll(const CiftiFile* ciftiIn, con
     int64_t numVoxels = (int64_t)myMap.size();
     if (numVoxels > 0)
     {//make a voxel bounding box to minimize memory usage
-        int extrema[6] = { myMap[0].m_ijk[0],
+        int64_t extrema[6] = { myMap[0].m_ijk[0],
             myMap[0].m_ijk[0],
             myMap[0].m_ijk[1],
             myMap[0].m_ijk[1],
-- 
2.45.2

