--- CRISPR_Studio_1.0.py.orig	2019-03-12 13:43:36.644438096 -0500
+++ CRISPR_Studio_1.0.py	2019-04-10 13:12:03.262862265 -0500
@@ -771,15 +771,8 @@
         writeFasta(spacerDict, outFasta)
         checkSpacer(spacerDict, averageLength, outFasta)
     elif args.checkFasta:
-        print('\n\tTHE FLAG -f (--checkFasta) WAS USED, NO VERIFICATION OF THE FASTA FILE WILL BE RUN')
-        print('\tIf you want CRISPR_Studio to extract the spacers from the gff file')
-        print('\twithout runing the verification step, just hit enter (leave the field empty)')
-        correctedFasta = input('\n\tPlease enter the name of the corrected fasta file: ')
-        if len(correctedFasta) > 1:
-            spacerDict, isolateList, averageLength, crArrayList = readCorrectedFasta(correctedFasta)
-        else:
-            spacerDict, isolateList, averageLength, crArrayList = spacerExtract(inFile)
-            writeFasta(spacerDict, outFasta)
+        spacerDict, isolateList, averageLength, crArrayList = spacerExtract(inFile)
+        writeFasta(spacerDict, outFasta)
     
     if args.gray and args.unique:
         print('\n\n\n\
