--- CMakeLists.txt
+++ CMakeLists.txt
@@ -27,8 +27,8 @@
 # Build dependencies (samtools and boost)
-include(BuildHtslib)
-include_directories(${HTSlib_INCLUDE_DIRS})
-
-set(REQUIRED_BOOST_LIBS thread program_options timer system chrono)
-include(BuildBoost)
-include_directories(${Boost_INCLUDE_DIRS})
+find_package(HTSlib 1.9)
+set(HTSlib_LIBRARIES hts)
+
+
+
+find_package(Boost 1.59.0 COMPONENTS thread program_options timer system chrono)
 
