--- plugins/rsem/sam/Makefile	2017-01-24 08:02:17.000000000 +1000
+++ plugins/rsem/sam/Makefile.new	2017-01-24 08:08:35.000000000 +1000
@@ -2,11 +2,11 @@
 
 CFLAGS_EXTRA=
 #CFLAGS_EXTRA= -L/usr/include/ncurses
-CFLAGS=		-g -Wall -O2 $(CFLAGS_EXTRA) #-m64 #-arch ppc
+CFLAGS=		-Wall
 
 DFLAGS_EXTRA=
 #DFLAGS_EXTRA= -Dexpl=exp -Dlogl=log
-DFLAGS=		-D_FILE_OFFSET_BITS=64 -D_LARGEFILE64_SOURCE -D_USE_KNETFILE -D_CURSES_LIB=1 $(DFLAGS_EXTRA)
+DFLAGS=		-D_FILE_OFFSET_BITS=64 -D_LARGEFILE64_SOURCE -D_USE_KNETFILE -D_CURSES_LIB=0 $(DFLAGS_EXTRA)
 
 KNETFILE_O=	knetfile.o
 LOBJS=		bgzf.o kstring.o bam_aux.o bam.o bam_import.o sam.o bam_index.o	\
@@ -20,7 +20,7 @@
 INCLUDES=	-I.
 SUBDIRS=	. bcftools misc
 LIBPATH=
-LIBCURSES=	-lcurses # -lXCurses
+
 
 .SUFFIXES:.c .o
 
