--- NAMESPACE
+++ NAMESPACE
@@ -13,7 +13,7 @@
 importFrom(exomeCopy, subdivideGRanges)
 importFrom(SomaticSignatures, seqchar)
 importFrom(VGAM, logit)
-importFrom(dplyr, rbind_all)
+importFrom(dplyr, bind_rows)
 importFrom(reshape2, melt)
 
 export(acCi, nhsCi)
--- R/rariantFromBam.R
+++ R/rariantFromBam.R
@@ -47,7 +47,7 @@
     
     res = NULL ## must be set
     if(value & length(val) > 0) {
-        res = rbind_all(val) ## or unlist on a GRL
+        res = bind_rows(val) ## or unlist on a GRL
         #if(!is.null(res)) { ## TODO: needed?
         res = df2gr(res)
         #}
