From edce335b33b62b6d6667eaa1862ce1af15021f3e Mon Sep 17 00:00:00 2001
From: 4less <joachim.fritscher@gmail.com>
Date: Tue, 29 Jul 2025 18:50:44 +0200
Subject: [PATCH] Change versioning in wfa2

---
 lib/wfa2-lib/VERSION | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/wfa2-lib/VERSION b/lib/wfa2-lib/VERSION
index e70ef1d..60b4912 100644
--- a/lib/wfa2-lib/VERSION
+++ b/lib/wfa2-lib/VERSION
@@ -1 +1 @@
-v2.3
+#define WFA_VERSION "v2.3"
\ No newline at end of file
-- 
2.34.1

