From 355b3e27aa24335cb730af44e3b6accecf5f7b60 Mon Sep 17 00:00:00 2001
From: 4less <joachim.fritscher@gmail.com>
Date: Tue, 29 Jul 2025 20:12:12 +0200
Subject: [PATCH] Change include path robinset for MACOS

---
 src/Utils/Utils.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/Utils/Utils.h b/src/Utils/Utils.h
index dd2e697..06fee0f 100755
--- a/src/Utils/Utils.h
+++ b/src/Utils/Utils.h
@@ -10,7 +10,7 @@
 #include <sys/stat.h>
 #include <unordered_map>
 #include <cstring>
-#include <robin_set.h>
+#include <robin/robin_set.h>
 #include <iostream>
 #include <emmintrin.h>
 
-- 
2.34.1

