From 7c560629f99536e1b8a5eebe84bfd914df690f24 Mon Sep 17 00:00:00 2001
From: 4less <joachim.fritscher@gmail.com>
Date: Sat, 26 Jul 2025 12:47:10 +0200
Subject: [PATCH] Comment out RTLIB

---
 src/CMakeLists.txt | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 51af233..1e8f190 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -201,14 +201,13 @@ message("CPMML: ${cPMML}")
 
 
 # Add this before target_link_libraries
-find_library(RT_LIBRARY rt)
-
-message("RT LIB: ${RT_LIBRARY}")
+#find_library(RT_LIBRARY rt)
+#message("RT LIB: ${RT_LIBRARY}")
 
 # Necessary to link to stdc++fs (filesystem experimental support)
 target_link_libraries(protal_lib
         LINK_PUBLIC
-        ${RT_LIBRARY}
+        #${RT_LIBRARY}
         cPMML
         stdc++fs
         wfa_lib
-- 
2.34.1

