diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index a064e94..e9e3446 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -201,22 +201,14 @@ message("Find cPMML: ${PROJECT_SOURCE_DIR}/../")
 message("Build dir: ${CMAKE_BINARY_DIR}/src/cPMML/")
 message("CPMML: ${cPMML}")
 
-
-# Add this before target_link_libraries
-find_library(RT_LIBRARY rt)
-
-message("RT LIB: ${RT_LIBRARY}")
-
 # Necessary to link to stdc++fs (filesystem experimental support)
 target_link_libraries(protal_lib
         LINK_PUBLIC
-        ${RT_LIBRARY}
         cPMML
         stdc++fs
         wfa_lib
         gzstream_lib
         ZLIB::ZLIB
-        -static # WITHOUT THIS IT WILL NOT COMPILE INTO A STATIC LIBRARY
 )
 
 message("CMAKE_CURRENT_SOURCE_DIR: ${CMAKE_CURRENT_SOURCE_DIR}")
