diff -ru po/src/Core/AlignmentStrategy.h pn/src/Core/AlignmentStrategy.h
--- a/src/Core/AlignmentStrategy.h	2026-08-10 12:04:03.000000000 +0200
+++ b/src/Core/AlignmentStrategy.h	2026-08-10 13:54:22.216600878 +0200
@@ -4,6 +4,7 @@
 
 #pragma once
 
+#include <cassert>
 #include "WFA2Wrapper2.h"
 #include "AlignmentOutputHandler.h"
 #include "GenomeLoader.h"
diff -ru po/src/Core/ChainAnchorFinder.h pn/src/Core/ChainAnchorFinder.h
--- a/src/Core/ChainAnchorFinder.h	2026-08-10 12:04:03.000000000 +0200
+++ b/src/Core/ChainAnchorFinder.h	2026-08-10 13:54:22.217051377 +0200
@@ -3,6 +3,7 @@
 //
 
 #pragma once
+#include <cassert>
 #include "ChainingStrategy.h"
 #include "Constants.h"
 #include "Benchmark.h"
diff -ru po/src/Core/ChainingStrategy.h pn/src/Core/ChainingStrategy.h
--- a/src/Core/ChainingStrategy.h	2026-08-10 12:04:03.000000000 +0200
+++ b/src/Core/ChainingStrategy.h	2026-08-10 13:54:22.216935128 +0200
@@ -5,6 +5,7 @@
 #pragma once
 
 #define __STDC_LIMIT_MACROS
+#include <cassert>
 #include <stdint.h>
 #include <vector>
 #include "KmerLookup.h"
diff -ru po/src/Core/SeedingStrategy.h pn/src/Core/SeedingStrategy.h
--- a/src/Core/SeedingStrategy.h	2026-08-10 12:04:03.000000000 +0200
+++ b/src/Core/SeedingStrategy.h	2026-08-10 13:54:22.216456376 +0200
@@ -4,6 +4,7 @@
 
 #pragma once
 
+#include <cassert>
 #include "KmerLookup.h"
 
 namespace protal {
diff -ru po/src/Hash/compact_map.h pn/src/Hash/compact_map.h
--- a/src/Hash/compact_map.h	2026-08-10 12:04:03.000000000 +0200
+++ b/src/Hash/compact_map.h	2026-08-10 13:54:22.216704072 +0200
@@ -6,6 +6,7 @@
 
 
 #define __STDC_LIMIT_MACROS
+#include <cassert>
 #include <stdint.h>
 #include <iostream>
 #include <err.h>
diff -ru po/src/IO/BufferedOutput.h pn/src/IO/BufferedOutput.h
--- a/src/IO/BufferedOutput.h	2026-08-10 12:04:03.000000000 +0200
+++ b/src/IO/BufferedOutput.h	2026-08-10 13:54:22.216659268 +0200
@@ -5,6 +5,7 @@
 #ifndef VARKIT_BUFFEREDOUTPUT_H
 #define VARKIT_BUFFEREDOUTPUT_H
 
+#include <cassert>
 #include <bits/stdc++.h>
 //#include <fmt/core.h>
 //#include <fmt/os.h>
diff -ru po/src/Profiling/Profiler.h pn/src/Profiling/Profiler.h
--- a/src/Profiling/Profiler.h	2026-08-10 12:04:03.000000000 +0200
+++ b/src/Profiling/Profiler.h	2026-08-10 13:54:22.217399322 +0200
@@ -4,6 +4,7 @@
 
 #pragma once
 
+#include <cassert>
 #include "Strain.h"
 #include "AlignmentUtils.h"
 #include <vector>
diff -ru po/src/SequenceUtils/GenomeLoader.h pn/src/SequenceUtils/GenomeLoader.h
--- a/src/SequenceUtils/GenomeLoader.h	2026-08-10 12:04:03.000000000 +0200
+++ b/src/SequenceUtils/GenomeLoader.h	2026-08-10 13:54:22.216984902 +0200
@@ -4,6 +4,7 @@
 
 #pragma once
 
+#include <cassert>
 #include <Constants.h>
 #include <sparse_map.h>
 #include <string>
diff -ru po/src/Taxonomy/Taxonomy.cpp pn/src/Taxonomy/Taxonomy.cpp
--- a/src/Taxonomy/Taxonomy.cpp	2026-08-10 12:04:03.000000000 +0200
+++ b/src/Taxonomy/Taxonomy.cpp	2026-08-10 13:54:22.216818698 +0200
@@ -2,6 +2,7 @@
 //// Created by fritsche on 17/05/2021.
 ////
 //
+#include <cassert>
 #include "Taxonomy.h"
 #include "Utilities.h"
 #include <iostream>
