diff --git a/dbcan/_version.py b/dbcan/_version.py
index 8eb2617..88f5f73 100644
--- a/dbcan/_version.py
+++ b/dbcan/_version.py
@@ -17,5 +17,5 @@ __version__: str
 __version_tuple__: VERSION_TUPLE
 version_tuple: VERSION_TUPLE
 
-__version__ = version = '5.0.1.dev4+g10d7889.d20250323'
-__version_tuple__ = version_tuple = (5, 0, 1, 'dev4', 'g10d7889.d20250323')
+__version__ = version = '5.1.2'
+__version_tuple__ = version_tuple = (5, 1, 2)
