diff --git a/simfold.cpp b/simfold.cpp
index 08cffe2..0ca03e3 100644
--- a/simfold.cpp
+++ b/simfold.cpp
@@ -64,7 +64,7 @@ void usage (const char *prog)
     printf ("Examples:\n");
     printf ("\t%s -s \"GCAACGAUGACAUACAUCGCUAGUCGACGC\" -n 10 \n", prog);
     printf ("\t%s -s \"GCAACGAUGACAUACAUCGCUAGUCGACGC\" -n 10 -r \"(____________________________)\"\n", prog);
-    printf ("\t%s -s \"GCAACGAUGACAUACAUCGCUAGUCGACGC\" -p params/CG_best_parameters_ISMB2007.txt\n");
+    printf ("\t%s -s \"GCAACGAUGACAUACAUCGCUAGUCGACGC\" -p params/CG_best_parameters_ISMB2007.txt\n", prog);
     exit (0);
 }
 
diff --git a/simfold_pf.cpp b/simfold_pf.cpp
index e3ba56e..424ee85 100644
--- a/simfold_pf.cpp
+++ b/simfold_pf.cpp
@@ -35,7 +35,7 @@ void usage (const char *prog)
     printf ("Examples:\n");
     printf ("\t%s -s \"GCAACGAUGACAUACAUCGCUAGUCGACGC\"\n", prog);
     printf ("\t%s -s \"GCAACGAUGACAUACAUCGCUAGUCGACGC\" -t 0.5\n", prog);
-    printf ("\t%s -s \"GCAACGAUGACAUACAUCGCUAGUCGACGC\" -p params/CG_best_parameters_ISMB2007.txt\n");
+    printf ("\t%s -s \"GCAACGAUGACAUACAUCGCUAGUCGACGC\" -p params/CG_best_parameters_ISMB2007.txt\n", prog);
     exit (0);
 }
 
