commit 625122c01dbf740131c2419a5d2bfd65d8d315d8
Author: Marco Matthies <71844+marcom@users.noreply.github.com>
Date:   Thu Feb 8 16:12:41 2024 +0100

    Remove unneeded include of malloc.h

diff --git a/src/multifold/m_energy_matrix.cpp b/src/multifold/m_energy_matrix.cpp
index 4d47d90..4030513 100644
--- a/src/multifold/m_energy_matrix.cpp
+++ b/src/multifold/m_energy_matrix.cpp
@@ -18,7 +18,6 @@
 #include <stdio.h>
 #include <string.h>
 #include <stdlib.h>
-#include <malloc.h>
 
 #include "constants.h"
 #include "structs.h"
diff --git a/src/multifold/m_multi_loop.cpp b/src/multifold/m_multi_loop.cpp
index a017da8..382d57a 100644
--- a/src/multifold/m_multi_loop.cpp
+++ b/src/multifold/m_multi_loop.cpp
@@ -16,7 +16,6 @@
  ***************************************************************************/
 
 #include <stdio.h>
-#include <malloc.h>
 
 #include "externs.h"
 #include "common.h"
diff --git a/src/pairfold/p_energy_matrix.cpp b/src/pairfold/p_energy_matrix.cpp
index 42c15c2..dc23e29 100644
--- a/src/pairfold/p_energy_matrix.cpp
+++ b/src/pairfold/p_energy_matrix.cpp
@@ -18,7 +18,6 @@
 #include <stdio.h>
 #include <string.h>
 #include <stdlib.h>
-#include <malloc.h>
 
 #include "constants.h"
 #include "structs.h"
diff --git a/src/pairfold/p_multi_loop.cpp b/src/pairfold/p_multi_loop.cpp
index f562ce0..267035a 100644
--- a/src/pairfold/p_multi_loop.cpp
+++ b/src/pairfold/p_multi_loop.cpp
@@ -16,7 +16,6 @@
  ***************************************************************************/
 
 #include <stdio.h>
-#include <malloc.h>
 
 #include "externs.h"
 #include "common.h"
diff --git a/src/pairfold/p_multi_loop_sub.cpp b/src/pairfold/p_multi_loop_sub.cpp
index 467fb53..a418218 100644
--- a/src/pairfold/p_multi_loop_sub.cpp
+++ b/src/pairfold/p_multi_loop_sub.cpp
@@ -16,7 +16,6 @@
  ***************************************************************************/
 
 #include <stdio.h>
-#include <malloc.h>
 
 #include "externs.h"
 #include "common.h"
diff --git a/src/simfold/s_energy_matrix.cpp b/src/simfold/s_energy_matrix.cpp
index cafc1d7..ce09162 100644
--- a/src/simfold/s_energy_matrix.cpp
+++ b/src/simfold/s_energy_matrix.cpp
@@ -20,7 +20,6 @@
 #include <stdio.h>
 #include <string.h>
 #include <stdlib.h>
-#include <malloc.h>
 
 #include "constants.h"
 #include "structs.h"
diff --git a/src/simfold/s_multi_loop.cpp b/src/simfold/s_multi_loop.cpp
index 6c5998a..de7991e 100644
--- a/src/simfold/s_multi_loop.cpp
+++ b/src/simfold/s_multi_loop.cpp
@@ -18,7 +18,6 @@
 // The main class for multi-loop related functions, the mfe case 
  
 #include <stdio.h>
-#include <malloc.h>
 
 #include "externs.h"
 #include "common.h"
diff --git a/src/simfold/s_multi_loop_sub.cpp b/src/simfold/s_multi_loop_sub.cpp
index e8f5219..bbdeb98 100644
--- a/src/simfold/s_multi_loop_sub.cpp
+++ b/src/simfold/s_multi_loop_sub.cpp
@@ -18,7 +18,6 @@
 // this class represents multi-loops for suboptimal pairfold, it implements the Wuchty complete suboptimal paper 
  
 #include <stdio.h>
-#include <malloc.h>
 
 #include "externs.h"
 #include "common.h"
diff --git a/src/simfold/s_partition_function.cpp b/src/simfold/s_partition_function.cpp
index c551b6d..c63e8b4 100644
--- a/src/simfold/s_partition_function.cpp
+++ b/src/simfold/s_partition_function.cpp
@@ -18,7 +18,6 @@
 #include <stdio.h>
 #include <string.h>
 #include <stdlib.h>
-#include <malloc.h>
 #include <math.h>
 #include <ctype.h>
 
diff --git a/src/simfold/s_partition_function_complex.cpp b/src/simfold/s_partition_function_complex.cpp
index 2654f25..2fb2a00 100644
--- a/src/simfold/s_partition_function_complex.cpp
+++ b/src/simfold/s_partition_function_complex.cpp
@@ -18,7 +18,6 @@
 #include <stdio.h>
 #include <string.h>
 #include <stdlib.h>
-#include <malloc.h>
 #include <math.h>
 #include <ctype.h>
 
