--- samtools-0.1.18/Makefile	2018-11-02 13:21:46.000000000 -0500
+++ samtools-0.1.18/Makefile.new	2019-03-15 17:12:00.000000000 -0500
@@ -1,5 +1,5 @@
 CC=			gcc
-CFLAGS=		-g -Wall -fsigned-char -O2 #-m64 #-arch ppc
+CFLAGS=		-g -fPIC -Wall -fsigned-char -O2
 DFLAGS=		-D_FILE_OFFSET_BITS=64 -D_LARGEFILE64_SOURCE -D_CURSES_LIB=0 #-D_USE_KNETFILE 
 KNETFILE_O=	knetfile.o
 LOBJS=		bgzf.o kstring.o bam_aux.o bam.o bam_import.o sam.o bam_index.o	\
