--- samtools-0.1.18/Makefile	2019-05-02 16:25:17.000000000 +0100
+++ samtools-0.1.18/Makefile.new	2019-05-02 23:10:50.364617064 +0100
@@ -1,5 +1,4 @@
-CC=			gcc
-CFLAGS=		-g -Wall -fsigned-char -O2 #-m64 #-arch ppc
+CFLAGS+=	-g -fPIC -Wall -fsigned-char -O2
 DFLAGS=		-D_FILE_OFFSET_BITS=64 -D_LARGEFILE64_SOURCE -D_CURSES_LIB=0 #-D_USE_KNETFILE 
 KNETFILE_O=	knetfile.o
 LOBJS=		bgzf.o kstring.o bam_aux.o bam.o bam_import.o sam.o bam_index.o	\
