From a74e3908015719c088f4797a21fd698f26008144 Mon Sep 17 00:00:00 2001
From: Ilya Shlyakhter <ilya@broadinstitute.org>
Date: Tue, 17 Jul 2018 22:10:51 -0400
Subject: [PATCH 3/3] automake option fixes

---
 Makefile.am | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/Makefile.am b/Makefile.am
index 86784e8..407b43d 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -7,7 +7,7 @@ EXTRA_DIST=
 TESTS=
 
 AM_CPPFLAGS = -DDISABLE_ASMLIB
-AM_CXXFLAGS = -O3 -fPIC $(PTHREAD_CFLAGS)
+AM_CXXFLAGS = -O3 -fPIC $(PTHREAD_CFLAGS) -std=c++14
 AM_LDFLAGS = $(PTHREAD_LIBS)
 
 AM_DEFAULT_SOURCE_EXT = .cpp
@@ -76,16 +76,16 @@ KMC_CPPS = \
 	$(KMC_MAIN_DIR)/timer.h
 
 libraduls_sse2_a_SOURCES = $(KMC_MAIN_DIR)/raduls_sse2.cpp
-libraduls_sse2_a_CXXFLAGS = -msse2
+libraduls_sse2_a_CXXFLAGS = $(AM_CXXFLAGS) -msse2
 
 libraduls_sse41_a_SOURCES = $(KMC_MAIN_DIR)/raduls_sse41.cpp
-libraduls_sse41_a_CXXFLAGS = -msse4.1
+libraduls_sse41_a_CXXFLAGS = $(AM_CXXFLAGS) -msse4.1
 
 libraduls_avx_a_SOURCES = $(KMC_MAIN_DIR)/raduls_avx.cpp
-libraduls_avx_a_CXXFLAGS = -mavx -fabi-version=0
+libraduls_avx_a_CXXFLAGS = $(AM_CXXFLAGS) -mavx -fabi-version=0
 
 libraduls_avx2_a_SOURCES = $(KMC_MAIN_DIR)/raduls_avx2.cpp
-libraduls_avx2_a_CXXFLAGS = -mavx2 -mfma -fabi-version=0
+libraduls_avx2_a_CXXFLAGS = $(AM_CXXFLAGS) -mavx2 -mfma -fabi-version=0
 
 noinst_LIBRARIES = libraduls_sse2.a libraduls_sse41.a libraduls_avx.a libraduls_avx2.a
 
-- 
2.14.4

