From bdfb67f3dbabee32c9a409ce4e03d5cbb39188f9 Mon Sep 17 00:00:00 2001
From: Ilya Shlyakhter <ilya@broadinstitute.org>
Date: Tue, 17 Jul 2018 22:52:39 -0400
Subject: [PATCH 5/5] fixed stat_func

---
 kmer_counter/binary_reader.h | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/kmer_counter/binary_reader.h b/kmer_counter/binary_reader.h
index ce4af0e..21be080 100644
--- a/kmer_counter/binary_reader.h
+++ b/kmer_counter/binary_reader.h
@@ -16,6 +16,7 @@ Date   : 2018-05-10
 #include "queues.h"
 #include "percent_progress.h"
 #include "bam_utils.h"
+#include <sys/types.h>
 #include <sys/stat.h>
 
 class CBinaryFilesReader
@@ -27,7 +28,7 @@ class CBinaryFilesReader
 		const auto& stat_func = _stat64;
 #else
 		typedef struct stat stat_struct;
-		const auto& stat_func = stat;
+		const auto stat_func = stat;
 #endif
 		stat_struct buf;
 		if (stat_func(path, &buf) == -1)
@@ -418,4 +419,4 @@ public:
 	}
 };
 
-#endif
\ No newline at end of file
+#endif
-- 
2.14.4

