--- a/Makefile
+++ b/Makefile
@@ -220,6 +220,8 @@
 
 $(HF_OBJ_DIR)/%.o: $(SRC_DIR)/%.cpp | $(HF_OBJ_DIR)
 	$(HF_CC) \
+		$(CPPFLAGS) \
+		$(CXXFLAGS) \
 		$(HF_CXXFLAGS) \
 		$(DEPFLAGS) \
 		$(HF_INCLUDES) \
@@ -228,6 +230,8 @@
 
 $(HF_OBJ_DIR)/%.o: $(SRC_DIR)/%.c | $(HF_OBJ_DIR)
 	$(HF_CC) \
+		$(CPPFLAGS) \
+		$(CXXFLAGS) \
 		$(HF_CXXFLAGS) \
 		$(DEPFLAGS) \
 		$(HF_INCLUDES) \
@@ -238,6 +242,8 @@
                 $(HIFIASM_DIR)/hifiasm_entry.h \
                 | $(HF_OBJ_DIR)
 	$(CXX) \
+		$(CPPFLAGS) \
+		$(CXXFLAGS) \
 		$(HF_CXXFLAGS) \
 		$(DEPFLAGS) \
 		$(HF_INCLUDES) \
@@ -251,6 +257,8 @@
 
 $(HA_OBJ_DIR)/%.o: $(HIFIASM_DIR)/%.cpp | $(HA_OBJ_DIR)
 	$(CXX) \
+		$(CPPFLAGS) \
+		$(CXXFLAGS) \
 		$(HA_CXXFLAGS) \
 		$(DEPFLAGS) \
 		$(HA_INCLUDES) \
@@ -259,6 +267,8 @@
 
 $(HA_OBJ_DIR)/%.o: $(HIFIASM_DIR)/%.c | $(HA_OBJ_DIR)
 	$(CC) \
+		$(CPPFLAGS) \
+		$(CFLAGS) \
 		$(HA_CXXFLAGS) \
 		$(DEPFLAGS) \
 		$(HA_INCLUDES) \
@@ -269,6 +279,8 @@
                 $(HIFIASM_DIR)/hifiasm_entry.h \
                 | $(HA_OBJ_DIR)
 	$(CXX) \
+		$(CPPFLAGS) \
+		$(CXXFLAGS) \
 		$(HA_CXXFLAGS) \
 		$(DEPFLAGS) \
 		$(HA_INCLUDES) \
@@ -281,7 +293,7 @@
 # ============================================================
 
 $(HA_SHARED_LIB): $(HA_OBJS) $(HA_MAIN_OBJ) | $(BUILD_DIR)
-	$(CXX) -shared -Wl,-Bsymbolic \
+	$(CXX) $(LDFLAGS) -shared -Wl,-Bsymbolic \
 		$(HA_OBJS) $(HA_MAIN_OBJ) \
 		-lz -lpthread -lm \
 		-o $@
@@ -297,6 +309,7 @@
 	$(HA_SHARED_LIB)
 
 	$(CXX) \
+		$(LDFLAGS) \
 		$(HF_CXXFLAGS) \
 		$(HF_OBJS) \
 		$(HF_MAIN_OBJ) \
