--- setup.py	2015-08-21 14:03:40.000000000 +0000
+++ setup.py.new	2016-06-03 18:48:10.665449998 +0000
@@ -23,7 +23,6 @@
       url='http://github.com/soravux/scoop',
       install_requires=['greenlet>=0.3.4',
                         'pyzmq>=13.1.0',
-                        'argparse>=1.1',
                         ],
       extras_require = {'nice': ['psutil>=0.6.1'],
                         },
