From d4176e12f4abb756777e30790505a200ff89cc1e Mon Sep 17 00:00:00 2001
From: Anders Goncalves da Silva <andersgs@gmail.com>
Date: Sun, 28 Jul 2019 22:39:57 -0700
Subject: [PATCH] Update Makefile

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index abbe42f..5478194 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 CC=			gcc
 CFLAGS=		-g -Wall -O2
 AR=			ar
-DFLAGS=		-DHAVE_PTHREAD
+DFLAGS=		-DHAVE_PTHREAD $(LDFLAGS)
 LOBJS=		utils.o kstring.o ksw.o bwt.o bntseq.o bwa.o bwamem.o bwamem_pair.o
 AOBJS=		QSufSort.o bwt_gen.o bwase.o bwaseqio.o bwtgap.o bwtaln.o bamlite.o \
 			is.o bwtindex.o bwape.o kopen.o pemerge.o \
-- 
2.20.1 (Apple Git-117)

