From 986dab0d7ffa8000d2943851396b53e155b8f1f8 Mon Sep 17 00:00:00 2001
From: Vivek Rai <vivekraiiitkgp@gmail.com>
Date: Sun, 7 Feb 2016 00:24:47 +0530
Subject: [PATCH] Fix hardcoded include, library paths

Signed-off-by: Vivek Rai <vivekraiiitkgp@gmail.com>
---
 setup.py | 130 ++++++++++++++++++++++++++++++++-------------------------------
 1 file changed, 66 insertions(+), 64 deletions(-)

diff --git setup.py setup.py
index b223b9d..4fdc7e5 100644
--- setup.py
+++ setup.py
@@ -35,68 +35,70 @@ from distutils.sysconfig import get_config_vars
 # remove -Wstrict-prototypes
 (opt,) = get_config_vars('OPT')
 if opt != None:
-  os.environ['OPT'] = " ".join(
-      flag for flag in opt.split() if flag != '-Wstrict-prototypes'
-  )
+    os.environ['OPT'] = " ".join(
+        flag for flag in opt.split() if flag != '-Wstrict-prototypes'
+    )
 
 # we need to switch the __init__.py file based on the python version
 # as python 3 uses a different syntax for metaclasses
 if sys.version_info >= (3,0):
-  # this is python 3.x
-  if (os.path.exists('./libsbml/__init__.py')):
-    os.remove('./libsbml/__init__.py')
-  shutil.copyfile('./script/libsbml3.py', './libsbml/__init__.py')
-else:
-  # this is an older python
-  if (os.path.exists('./libsbml/__init__.py')):
-    os.remove('./libsbml/__init__.py')
-  shutil.copyfile('./script/libsbml2.py', './libsbml/__init__.py')
+    # this is python 3.x
+    if (os.path.exists('./libsbml/__init__.py')):
+        os.remove('./libsbml/__init__.py')
+        shutil.copyfile('./script/libsbml3.py', './libsbml/__init__.py')
+    else:
+        # this is an older python
+        if (os.path.exists('./libsbml/__init__.py')):
+            os.remove('./libsbml/__init__.py')
+            shutil.copyfile('./script/libsbml2.py', './libsbml/__init__.py')
 
 # figure out the os
 basepath = './base/'
 current_os = 'LINUX'
-lib_dirs =  ['/usr/lib']
-inc_dirs = ['/usr/include/libxml2']
+lib_dirs = [get_config_vars()['LIBDIR']]
+inc_dirs = [os.path.join(get_config_vars()['INCLUDEDIR'], 'libxml2')]
 libs = ['xml2', 'bz2', 'z']
 package_name = '"libsbml"'
 definitions = []
 packages = [
-  ('USE_COMP', None),
-  ('USE_QUAL', None),
-  ('USE_FBC', None),
-  ('USE_LAYOUT', None)
+    ('USE_COMP', None),
+    ('USE_QUAL', None),
+    ('USE_FBC', None),
+    ('USE_LAYOUT', None)
 ]
-if platform.system() == 'Darwin': 
-  current_os = 'DARWIN'
+
+if platform.system() == 'Darwin':
+    current_os = 'DARWIN'
 elif platform.system() == 'Windows':
-  current_os = 'WIN32'
-  package_name = '\\"libsbml\\"'
-  definitions = [
-    ('LIBSBML_EXPORTS', None),
-	('LIBLAX_STATIC', None)
-  ]
-  dependency_dir = 'D:\Development\libsbml_dependencies\install_vs9_release_x86'
-  inc_dirs= [ dependency_dir + '/include']
-  lib_dirs= [ dependency_dir + '/lib']
-  libs = ['libxml2', 'libbz2', 'zdll', 'libiconv', 'ws2_32']
+    current_os = 'WIN32'
+    package_name = '\\"libsbml\\"'
+    definitions = [
+        ('LIBSBML_EXPORTS', None),
+        ('LIBLAX_STATIC', None)
+    ]
+    # FIXME: Hardcoded file path
+    dependency_dir = 'D:\Development\libsbml_dependencies\install_vs9_release_x86'
+    inc_dirs= [ dependency_dir + '/include']
+    lib_dirs= [ dependency_dir + '/lib']
+    libs = ['libxml2', 'libbz2', 'zdll', 'libiconv', 'ws2_32']
 
 cfiles = [ basepath + 'libsbml_wrap.cpp' ]
 for root, dirs, files in os.walk(basepath + 'sbml'):
-  for file in files:
-    if file.endswith('.c') or file.endswith('.cpp'):
-      cfiles.append(os.path.join(root, file))
+    for file in files:
+        if file.endswith('.c') or file.endswith('.cpp'):
+            cfiles.append(os.path.join(root, file))
 
 from distutils.core import setup, Extension
 
-setup(name             = "python-libsbml", 
+setup(name             = "python-libsbml",
       version          = "5.12.0",
       description      = "LibSBML Python API",
-      long_description = ("LibSBML is a library for reading, writing and "+
-                          "manipulating the Systems Biology Markup Language "+
-                          "(SBML).  It is written in ISO C and C++, supports "+
-                          "SBML Levels 1, 2 and 3, and runs on Linux, Microsoft "+
-                          "Windows, and Apple MacOS X.  For more information "+
-                          "about SBML, please see http://sbml.org."),
+      long_description = ("LibSBML is a library for reading, writing and"
+                          " manipulating the Systems Biology Markup Language"
+                          " (SBML). It is written in ISO C and C++, supports"
+                          " SBML Levels 1, 2 and 3, and runs on Linux, Microsoft"
+                          " Windows, and Apple MacOS X. For more information "
+                          " about SBML, please see http://sbml.org."),
       license          = "LGPL",
       author           = "SBML Team",
       author_email     = "libsbml-team@caltech.edu",
@@ -105,27 +107,27 @@ setup(name             = "python-libsbml",
       package_dir      = {'libsbml': 'libsbml'},
       #data_files       = [('lib/site-packages', ['libsbml.pth'])],
       ext_package      = "libsbml",
-      ext_modules      = [Extension("_libsbml", 
-                            sources = cfiles,
-                            define_macros =  definitions 
-							  +  [(current_os, None), 
-                              ('USE_LIBXML', None),
-                              ('USE_ZLIB', None),
-                              ('USE_BZ2', None),
-                              ('_GLIBCXX_USE_CXX11_ABI', 0)
-                              ] 
-							  + packages,
-                            include_dirs = inc_dirs +
-							  [                              
-                              basepath + "/", 
-                              basepath + "/sbml", 
-                              basepath + "/sbml/compress", 
-                              basepath + "/sbml/validator/constraints", 
-                              basepath + "/sbml/packages/comp/validator", 
-                              basepath + "/sbml/packages/comp/validator/constraints", 
-                              "."],
-                            libraries = libs,
-                            library_dirs = lib_dirs
-                            )
-                         ]
-)
+      ext_modules      = [Extension("_libsbml",
+                                    sources = cfiles,
+                                    define_macros =  definitions
+                                    +  [(current_os, None),
+                                        ('USE_LIBXML', None),
+                                        ('USE_ZLIB', None),
+                                        ('USE_BZ2', None),
+                                        ('_GLIBCXX_USE_CXX11_ABI', 0)
+                                        ] \
+                                    + packages,
+                                    include_dirs = inc_dirs +
+                                    [
+                                        basepath + "/",
+                                        basepath + "/sbml",
+                                        basepath + "/sbml/compress",
+                                        basepath + "/sbml/validator/constraints",
+                                        basepath + "/sbml/packages/comp/validator",
+                                        basepath + "/sbml/packages/comp/validator/constraints",
+                                        "."],
+                                    libraries = libs,
+                                    library_dirs = lib_dirs
+                                    )
+                          ]
+      )
-- 
1.9.1

