--- CMakeLists.txt
+++ CMakeLists.txt
@@ -30,9 +30,9 @@
   if(CMAKE_CXX_COMPILER_VERSION VERSION_LESS "4.7.0")
     message(FATAL_ERROR "Insufficient gcc version.  CMake will exit.")
   endif()
-  if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER "7.1")
-    message(FATAL_ERROR "Unsupported gcc version. CMake will exit.")
-  endif()
+#  if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER "7.1")
+#    message(FATAL_ERROR "Unsupported gcc version. CMake will exit.")
+#  endif()
 else ()
   message(STATUS "Couldn't determine gcc version!")
 endif ()
