--- CMakeLists.txt
+++ CMakeLists.txt
@@ -40,9 +40,9 @@
 include( CheckCXXCompilerFlag )
-check_cxx_compiler_flag("-static-libgcc" CXX_HAS_STATIC_LIBGCC)
-check_cxx_compiler_flag("-static-libstdc++" CXX_HAS_STATIC_LIBSTDCXX)
-if (CXX_HAS_STATIC_LIBSTDCXX)
-    set( CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -static-libgcc -static-libstdc++" )
-else ()
-    set( CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -static" )
-endif ()
+#check_cxx_compiler_flag("-static-libgcc" CXX_HAS_STATIC_LIBGCC)
+#check_cxx_compiler_flag("-static-libstdc++" CXX_HAS_STATIC_LIBSTDCXX)
+#if (CXX_HAS_STATIC_LIBSTDCXX)
+#    set( CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -static-libgcc -static-libstdc++" )
+#else ()
+#    set( CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -static" )
+#endif ()
 
--- src/c/CMakeLists.txt
+++ src/c/CMakeLists.txt
@@ -54,5 +54,5 @@
 set(Boost_DEBUG 1)
-set(Boost_USE_STATIC_LIBS   ON)
+set(Boost_USE_STATIC_LIBS   OFF)
 set(Boost_USE_MULTITHREADED OFF)
-set(Boost_USE_STATIC_RUNTIME ON )
+set(Boost_USE_STATIC_RUNTIME OFF )
 
