--- Meraculous-v2.2.6/src/c/Compat.h	2018-08-10 22:43:06.115287000 +0000
+++ Meraculous-v2.2.6.working/src/c/Compat.h	2024-11-25 14:49:34.912000000 +0000
@@ -82,11 +82,7 @@
 #include <sched.h>  // For sched_setaffinity
 #endif
 
-#ifndef __APPLE__
-#include <xmmintrin.h>  // This is currently (in Dec 2013) broken on Mac OS X 10.9 (Apple clang-500.2.79)
-#else
 #define _mm_prefetch(...) {}
-#endif
 
 typedef int64_t _int64;
 typedef uint64_t _uint64;
