diff --git a/src/bowtie/Makefile.orig b/src/bowtie/Makefile
index d79334a..e9db6b0 100644
--- a/src/bowtie/Makefile.orig
+++ b/src/bowtie/Makefile
@@ -5,11 +5,6 @@
 SEQAN_DIR = SeqAn-1.1
 SEQAN_INC = -I $(SEQAN_DIR)
 INC = $(SEQAN_INC)
-GCC_PREFIX = $(shell dirname `which gcc`)
-GCC_SUFFIX =
-CC = $(GCC_PREFIX)/gcc$(GCC_SUFFIX)
-CPP = $(GCC_PREFIX)/g++$(GCC_SUFFIX)
-CXX = $(CPP)
 HEADERS = $(wildcard *.h)
 BOWTIE_PTHREADS = 1
 BOWTIE_MM = 1
