diff --git a/src/MapSplice/Makefile.orig b/src/MapSplice/Makefile
index b04953c..99aefef 100644
--- a/src/MapSplice/Makefile.orig
+++ b/src/MapSplice/Makefile
@@ -5,17 +5,12 @@
 SEQAN_DIR = SeqAn-1.1
 SEQAN_INC = -I $(SEQAN_DIR)
 INC = $(SEQAN_INC)
-GCC_PREFIX = $(shell dirname `which gcc`)
-GCC_SUFFIX =
-CC = $(GCC_PREFIX)/gcc$(GCC_SUFFIX)
-CPP = $(GCC_PREFIX)/g++$(GCC_SUFFIX)
-CXX = $(CPP)
 HEADERS = $(wildcard *.h)
 BOWTIE_PTHREADS = 1
 BOWTIE_MM = 1
 BOWTIE_SHARED_MEM = 1
 EXTRA_FLAGS =
-EXTRA_CFLAGS = 
+EXTRA_CFLAGS =
 EXTRA_CXXFLAGS =
 CFLAGS += $(EXTRA_CFLAGS)
 CXXFLAGS += $(EXTRA_CXXFLAGS)
@@ -47,11 +42,11 @@ LINUX = 1
 EXTRA_FLAGS += -Wl,--hash-style=both
 endif
 
-MM_DEF = 
+MM_DEF =
 ifeq (1,$(BOWTIE_MM))
 MM_DEF = -DBOWTIE_MM
 endif
-SHMEM_DEF = 
+SHMEM_DEF =
 ifeq (1,$(BOWTIE_SHARED_MEM))
 SHMEM_DEF = -DBOWTIE_SHARED_MEM
 endif
@@ -73,7 +68,7 @@ endif
 PREFETCH_LOCALITY = 2
 PREF_DEF = -DPREFETCH_LOCALITY=$(PREFETCH_LOCALITY)
 
-LIBS = 
+LIBS =
 SEARCH_LIBS = $(PTHREAD_LIB)
 BUILD_LIBS =
 
@@ -266,7 +261,7 @@ bowtie-inspect: bowtie_inspect.cpp $(HEADERS) $(OTHER_CPPS)
 		$(OTHER_CPPS) \
 		$(LIBS)
 
-bowtie-inspect-debug: bowtie_inspect.cpp $(HEADERS) $(OTHER_CPPS) 
+bowtie-inspect-debug: bowtie_inspect.cpp $(HEADERS) $(OTHER_CPPS)
 	$(checksum)
 	$(CXX) $(DEBUG_FLAGS) \
 		$(DEBUG_DEFS) $(ALL_FLAGS) \
@@ -291,7 +286,7 @@ bowtie-src.zip: $(SRC_PKG_LIST)
 	cp .src.tmp/$@ .
 	rm -rf .src.tmp
 
-bowtie-bin.zip: $(BIN_PKG_LIST) $(BIN_LIST) $(BIN_LIST_AUX) 
+bowtie-bin.zip: $(BIN_PKG_LIST) $(BIN_LIST) $(BIN_LIST_AUX)
 	chmod a+x scripts/*.sh scripts/*.pl
 	rm -rf .bin.tmp
 	mkdir .bin.tmp
