diff --git a/mapsplice.py b/mapsplice.py
index 96e17df..1271680 100644
--- a/mapsplice.py
+++ b/mapsplice.py
@@ -1,3 +1,4 @@
+#!/usr/bin/env python
  #    MapSplice
  #
  #    Copyright (C) 2012-2013 University of Kentucky
@@ -105,7 +106,7 @@ cluster_data_parsedPER_dir = cluster_data_dir + "parsedPER/"
 formated_chrom_dir = temp_dir + "formated_chrom/"
 formated_reads_dir = temp_dir + "formated_reads/"
 filter_repeats_dir = temp_dir + "filter_repeats/"
-bin_dir = sys.path[0] + "/bin/"
+bin_dir = sys.path[0] + "/"
 rerun_all = 1
 DEBUG = 0
 max_chromosome = 1000
