diff --git a/samtools-0.1.9/misc/Makefile b/samtools-0.1.9/misc/Makefile
index 6c25c78..4f3307b 100644
--- a/samtools-0.1.9/misc/Makefile
+++ b/samtools-0.1.9/misc/Makefile
@@ -1,11 +1,11 @@
 CC=			gcc
 CXX=		g++
-CFLAGS=		-g -Wall -O2 #-m64 #-arch ppc
+CFLAGS=		-g -Wall -O2 -L${PREFIX}/lib -I${PREFIX}/include
 CXXFLAGS=	$(CFLAGS)
 DFLAGS=		-D_FILE_OFFSET_BITS=64
 OBJS=
 PROG=		md5sum-lite md5fa maq2sam-short maq2sam-long wgsim
-INCLUDES=	-I..
+INCLUDES=	-I.. -I${PREFIX}/include
 SUBDIRS=	.

 .SUFFIXES:.c .o
