diff --git a/samtools-0.1.9/Makefile b/samtools-0.1.9/Makefile
index 77cda86..610c2a0 100644
--- a/samtools-0.1.9/Makefile
+++ b/samtools-0.1.9/Makefile
@@ -1,5 +1,5 @@
 CC=			gcc
-CFLAGS=		-g -Wall -O2 #-m64 #-arch ppc
+CFLAGS=		-g -Wall -O2 -L${PREFIX}/lib -I${PREFIX}/include
 DFLAGS=		-D_FILE_OFFSET_BITS=64 -D_USE_KNETFILE -D_CURSES_LIB=1
 KNETFILE_O=	knetfile.o
 LOBJS=		bgzf.o kstring.o bam_aux.o bam.o bam_import.o sam.o bam_index.o	\
@@ -9,10 +9,10 @@ AOBJS=		bam_tview.o bam_maqcns.o bam_plcmd.o sam_view.o	\
 			bam_rmdup.o bam_rmdupse.o bam_mate.o bam_stat.o bam_color.o	\
 			bamtk.o kaln.o bam2bcf.o errmod.o sample.o
 PROG=		samtools
-INCLUDES=	-I.
+INCLUDES=	-I. -I${PREFIX}/include -I${PREFIX}/include/ncurses
 SUBDIRS=	. bcftools misc
-LIBPATH=
-LIBCURSES=	-lcurses # -lXCurses
+LIBPATH=	-L${PREFIX}/lib
+LIBCURSES=	-lncurses

 .SUFFIXES:.c .o

