--- exec/MaaslinToGraphlanAnnotation.py	2017-08-16 01:43:33.000000000 -0500
+++ exec/MaaslinToGraphlanAnnotation.py.new	2019-04-09 19:29:41.486849520 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/opt/anaconda1/anaconda2anaconda3/bin/python
 #####################################################################################
 #Copyright (C) <2012>
 #
--- exec/merge_metadata.py	2017-08-16 01:43:33.000000000 -0500
+++ exec/merge_metadata.py.new	2019-04-09 19:29:56.588850141 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/opt/anaconda1/anaconda2anaconda3/bin/python
 #####################################################################################
 #Copyright (C) <2012>
 #
--- exec/PCLToGraphlanCoreGene.py	2017-08-16 01:43:33.000000000 -0500
+++ exec/PCLToGraphlanCoreGene.py.new	2019-04-09 19:30:09.053850661 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/opt/anaconda1/anaconda2anaconda3/bin/python
 #####################################################################################
 #Copyright (C) <2012>
 #
--- exec/transpose.py	2017-08-16 01:43:33.000000000 -0500
+++ exec/transpose.py.new	2019-04-09 19:30:20.677851136 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/opt/anaconda1/anaconda2anaconda3/bin/python
 #######################################################################################
 # This file is provided under the Creative Commons Attribution 3.0 license.
 #
