diff -ruN metaxa2 metaxa2
--- metaxa2	2018-09-21 16:39:40.796620142 +0200
+++ metaxa2	2018-06-15 17:25:00.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/usr/bin/perl
 # Metaxa 2.2
 $app_title     = "Metaxa2 -- Improved Identification and Classification of barcoding genes in environmental datasets";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_c metaxa2_c
--- metaxa2_c	2018-09-21 16:39:40.812620094 +0200
+++ metaxa2_c	2018-03-01 16:30:09.000000000 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/usr/bin/perl
 # Metaxa Classifier
 $app_title     = "Metaxa Classifier -- Classifies (predicted) barcoding genes";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_dbb metaxa2_dbb
--- metaxa2_dbb	2018-09-21 16:39:40.824620059 +0200
+++ metaxa2_dbb	2018-06-15 17:22:57.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/usr/bin/perl
 # Metaxa2 Database Builder
 $app_title     = "Metaxa2 Database Builder -- Automatic database creation for Metaxa2";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_dc metaxa2_dc
--- metaxa2_dc	2018-09-21 16:39:40.828620047 +0200
+++ metaxa2_dc	2018-06-15 17:23:28.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/usr/bin/perl
 # Metaxa 2.2
 $app_title     = "Metaxa2 Diversity Tools - Data Collector";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_install_database metaxa2_install_database
--- metaxa2_install_database	2018-09-21 16:39:40.828620047 +0200
+++ metaxa2_install_database	2018-06-15 17:25:31.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/usr/bin/perl
 # Metaxa 2.2
 $app_title     = "Metaxa2 -- Database installer";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_rf metaxa2_rf
--- metaxa2_rf	2018-09-21 16:39:40.832620034 +0200
+++ metaxa2_rf	2018-06-15 17:25:22.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/usr/bin/perl
 # Metaxa 2.2
 $app_title     = "Metaxa2 Diversity Tools - Rarefaction Analysis tool";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_si metaxa2_si
--- metaxa2_si	2018-09-21 16:39:40.836620022 +0200
+++ metaxa2_si	2018-06-15 17:23:55.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/usr/bin/perl
 # Metaxa 2.2
 $app_title     = "Metaxa2 Diversity Tools - Species Inference tool";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_ttt metaxa2_ttt
--- metaxa2_ttt	2018-09-21 16:39:40.836620022 +0200
+++ metaxa2_ttt	2018-06-15 17:23:19.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/usr/bin/perl
 # Metaxa Taxonomic Traversal Tool
 $app_title     = "Metaxa Taxonomic Traversal Tool -- Processes Taxonomic Output From Metaxa";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_uc metaxa2_uc
--- metaxa2_uc	2018-09-21 16:39:40.840620010 +0200
+++ metaxa2_uc	2018-06-15 17:24:49.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/usr/bin/perl
 # Metaxa 2.2
 $app_title     = "Metaxa2 Diversity Tools - Uniqueness of Community Analyzer";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_x metaxa2_x
--- metaxa2_x	2018-09-21 16:39:40.844619999 +0200
+++ metaxa2_x	2018-06-15 17:24:16.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/usr/bin/perl
 # Metaxa Extractor
 $app_title     = "Metaxa Extractor -- Identifies barcoding genes in (meta)genomic data and extracts them";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
