diff -ruN install_metaxa2 install_metaxa2
--- install_metaxa2	2015-10-02 09:32:59.000000000 +0300
+++ install_metaxa2	2022-01-24 16:06:00.636515373 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 ## This little neat script installs metaxa in the default way
 print "Metaxa2 Installer\n";
diff -ruN metaxa2 metaxa2
--- metaxa2	2021-02-13 21:35:42.000000000 +0200
+++ metaxa2	2022-01-24 16:06:23.812508227 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # Metaxa 2.2
 $app_title     = "Metaxa2 -- Improved Identification and Classification of barcoding genes in environmental datasets";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_c metaxa2_c
--- metaxa2_c	2021-02-13 21:20:18.000000000 +0200
+++ metaxa2_c	2022-01-24 16:06:18.428509801 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # Metaxa Classifier
 $app_title     = "Metaxa Classifier -- Classifies (predicted) barcoding genes";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_dbb metaxa2_dbb
--- metaxa2_dbb	2020-06-10 15:55:35.000000000 +0300
+++ metaxa2_dbb	2022-01-24 16:06:13.740511213 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # Metaxa2 Database Builder
 $app_title     = "Metaxa2 Database Builder -- Automatic database creation for Metaxa2";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_dc metaxa2_dc
--- metaxa2_dc	2018-06-15 17:23:28.000000000 +0300
+++ metaxa2_dc	2022-01-24 16:06:05.156513901 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # Metaxa 2.2
 $app_title     = "Metaxa2 Diversity Tools - Data Collector";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_install_database metaxa2_install_database
--- metaxa2_install_database	2020-08-13 15:35:50.000000000 +0300
+++ metaxa2_install_database	2022-01-24 16:06:16.108510494 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # Metaxa 2.2
 $app_title     = "Metaxa2 -- Database installer";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_rf metaxa2_rf
--- metaxa2_rf	2018-07-26 23:43:57.000000000 +0300
+++ metaxa2_rf	2022-01-24 16:06:11.676511847 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # Metaxa 2.2
 $app_title     = "Metaxa2 Diversity Tools - Rarefaction Analysis tool";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_si metaxa2_si
--- metaxa2_si	2018-06-15 17:23:55.000000000 +0300
+++ metaxa2_si	2022-01-24 16:06:07.204513247 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # Metaxa 2.2
 $app_title     = "Metaxa2 Diversity Tools - Species Inference tool";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_ttt metaxa2_ttt
--- metaxa2_ttt	2018-06-15 17:23:19.000000000 +0300
+++ metaxa2_ttt	2022-01-24 16:06:02.828514654 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # Metaxa Taxonomic Traversal Tool
 $app_title     = "Metaxa Taxonomic Traversal Tool -- Processes Taxonomic Output From Metaxa";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_uc metaxa2_uc
--- metaxa2_uc	2018-06-15 17:24:49.000000000 +0300
+++ metaxa2_uc	2022-01-24 16:06:09.388512558 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # Metaxa 2.2
 $app_title     = "Metaxa2 Diversity Tools - Uniqueness of Community Analyzer";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_x metaxa2_x
--- metaxa2_x	2021-02-13 21:20:12.000000000 +0200
+++ metaxa2_x	2022-01-24 16:06:20.908509068 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # Metaxa Extractor
 $app_title     = "Metaxa Extractor -- Identifies barcoding genes in (meta)genomic data and extracts them";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
