diff -ruN install_metaxa2 install_metaxa2
--- install_metaxa2	2015-10-02 09:32:59.000000000 +0200
+++ install_metaxa2	2018-09-21 17:16:40.630102936 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 ## This little neat script installs metaxa in the default way
 print "Metaxa2 Installer\n";
diff -ruN metaxa2 metaxa2
--- metaxa2	2018-09-21 17:04:47.706053280 +0200
+++ metaxa2	2018-09-21 17:09:35.230232749 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # Metaxa 2.2
 $app_title     = "Metaxa2 -- Improved Identification and Classification of barcoding genes in environmental datasets";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_c metaxa2_c
--- metaxa2_c	2018-09-21 17:04:47.706053280 +0200
+++ metaxa2_c	2018-09-21 17:09:35.254232475 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # Metaxa Classifier
 $app_title     = "Metaxa Classifier -- Classifies (predicted) barcoding genes";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_dbb metaxa2_dbb
--- metaxa2_dbb	2018-09-21 17:04:47.710053218 +0200
+++ metaxa2_dbb	2018-09-21 17:09:35.262232384 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # Metaxa2 Database Builder
 $app_title     = "Metaxa2 Database Builder -- Automatic database creation for Metaxa2";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_dc metaxa2_dc
--- metaxa2_dc	2018-09-21 17:04:47.714053156 +0200
+++ metaxa2_dc	2018-09-21 17:09:35.266232337 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # Metaxa 2.2
 $app_title     = "Metaxa2 Diversity Tools - Data Collector";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_install_database metaxa2_install_database
--- metaxa2_install_database	2018-09-21 17:04:47.714053156 +0200
+++ metaxa2_install_database	2018-09-21 17:09:35.266232337 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # Metaxa 2.2
 $app_title     = "Metaxa2 -- Database installer";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_rf metaxa2_rf
--- metaxa2_rf	2018-09-21 17:04:47.714053156 +0200
+++ metaxa2_rf	2018-09-21 17:09:35.270232292 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # Metaxa 2.2
 $app_title     = "Metaxa2 Diversity Tools - Rarefaction Analysis tool";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_si metaxa2_si
--- metaxa2_si	2018-09-21 17:04:47.718053094 +0200
+++ metaxa2_si	2018-09-21 17:09:35.274232246 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # Metaxa 2.2
 $app_title     = "Metaxa2 Diversity Tools - Species Inference tool";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_ttt metaxa2_ttt
--- metaxa2_ttt	2018-09-21 17:04:47.718053094 +0200
+++ metaxa2_ttt	2018-09-21 17:09:35.278232200 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # Metaxa Taxonomic Traversal Tool
 $app_title     = "Metaxa Taxonomic Traversal Tool -- Processes Taxonomic Output From Metaxa";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_uc metaxa2_uc
--- metaxa2_uc	2018-09-21 17:04:47.722053031 +0200
+++ metaxa2_uc	2018-09-21 17:09:35.282232155 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # Metaxa 2.2
 $app_title     = "Metaxa2 Diversity Tools - Uniqueness of Community Analyzer";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
diff -ruN metaxa2_x metaxa2_x
--- metaxa2_x	2018-09-21 17:04:47.722053031 +0200
+++ metaxa2_x	2018-09-21 17:09:35.286232109 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # Metaxa Extractor
 $app_title     = "Metaxa Extractor -- Identifies barcoding genes in (meta)genomic data and extracts them";
 $app_author    = "Johan Bengtsson-Palme, University of Gothenburg";
