diff --git a/src/runsimd.cpp b/src/runsimd.cpp
index 10c0b83..54202d5 100644
--- a/src/runsimd.cpp
+++ b/src/runsimd.cpp
@@ -152,6 +152,7 @@ static int exe_path(const char *exe, int max, char buf[], int *base_st)
 static void test_and_launch(char *argv[], char *prefix, const char *simd) // we assume prefix is long enough
 {
 	struct stat st;
+    int prefix_len = strlen(prefix);
 	strcat_s(prefix, PATH_MAX, simd);
     fprintf(stderr, "Looking to launch executable \"%s\", simd = %s\n", prefix, simd);
     fprintf(stderr, "stat(prefix, &st) = %d\n", stat(prefix, &st));
@@ -171,6 +172,7 @@ static void test_and_launch(char *argv[], char *prefix, const char *simd) // we
     {
         fprintf(stderr, "Can not run executable: %s\n", prefix);
     }
+    prefix[prefix_len] = 0;
 }
 
 int main(int argc, char *argv[])
