diff --git a/Makefile b/Makefile
index ad06b39..ceb836a 100644
--- a/Makefile
+++ b/Makefile
@@ -38,7 +38,7 @@ endif
 ARCH_FLAGS=    -msse4.1
 MEM_FLAGS=     -DSAIS=1
 CPPFLAGS=      -DENABLE_PREFETCH -DV17=1 $(MEM_FLAGS) 
-INCLUDES=   -Isrc -Iext/safestringlib/include
+INCLUDES=   -Isrc -Iext/safestringlib/include $(LDFLAGS) $(CFLAGS)
 LIBS=          -lpthread -lm -lz -L. -lbwa  -Lext/safestringlib -lsafestring
 OBJS=          src/fastmap.o src/bwtindex.o src/utils.o src/kthread.o \
                        src/kstring.o src/ksw.o src/bntseq.o src/bwamem.o src/profiling.o src/bandedSWA.o \

