--- ext/safestringlib/makefile	2025-03-05 13:14:04
+++ ext/safestringlib/makefile	2025-03-05 13:13:53
@@ -1,7 +1,7 @@
 IDIR = include
 MKDIR_P = mkdir -p
 CC=gcc
-CFLAGS=-I$(IDIR) -fstack-protector-strong -fPIE -fPIC -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security
+CFLAGS+=-I$(IDIR) -fstack-protector-strong -fPIE -fPIC -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security
 LDFLAGS=-z noexecstack -z relo -z now
 
 ODIR=obj
