diff --git a/src/bntseq.cpp b/src/bntseq.cpp
index 13aceb4..b1dca18 100644
--- a/src/bntseq.cpp
+++ b/src/bntseq.cpp
@@ -31,7 +31,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
-#include <zlib.h>
+#include <zlib-ng.h>
 #include <unistd.h>
 #include <errno.h>
 #include <limits.h>
diff --git a/src/bntseq.h b/src/bntseq.h
index b241a3a..a50fcb2 100644
--- a/src/bntseq.h
+++ b/src/bntseq.h
@@ -31,7 +31,7 @@
 #include <assert.h>
 #include <stdint.h>
 #include <stdio.h>
-#include <zlib.h>
+#include <zlib-ng.h>
 
 #ifndef BWA_UBYTE
 #define BWA_UBYTE
diff --git a/src/bwa.cpp b/src/bwa.cpp
index b764381..260fabb 100644
--- a/src/bwa.cpp
+++ b/src/bwa.cpp
@@ -30,7 +30,7 @@ Authors: Vasimuddin Md <vasimuddin.md@intel.com>; Sanchit Misra <sanchit.misra@i
 
 #include <string.h>
 #include <stdio.h>
-#include <zlib.h>
+#include <zlib-ng.h>
 #include <assert.h>
 #include "bntseq.h"
 #include "bwa.h"
diff --git a/src/bwa.h b/src/bwa.h
index 877f00c..97b2220 100644
--- a/src/bwa.h
+++ b/src/bwa.h
@@ -32,7 +32,7 @@ Authors: Vasimuddin Md <vasimuddin.md@intel.com>; Sanchit Misra <sanchit.misra@i
 #define BWA_H_
 
 #include <stdint.h>
-#include <zlib.h>
+#include <zlib-ng.h>
 #include "bntseq.h"
 #include "bwt.h"
 #include "macro.h"
diff --git a/src/bwtindex.cpp b/src/bwtindex.cpp
index 4a48b20..8d94050 100644
--- a/src/bwtindex.cpp
+++ b/src/bwtindex.cpp
@@ -33,7 +33,7 @@
 #include <string.h>
 #include <unistd.h>
 #include <time.h>
-#include <zlib.h>
+#include <zlib-ng.h>
 #include "bntseq.h"
 #include "bwa.h"
 #include "bwt.h"
diff --git a/src/fastmap.h b/src/fastmap.h
index e94022f..c1ee9d5 100644
--- a/src/fastmap.h
+++ b/src/fastmap.h
@@ -32,7 +32,7 @@ Authors: Vasimuddin Md <vasimuddin.md@intel.com>; Sanchit Misra <sanchit.misra@i
 #define FASTMAP_HPP
 
 #include <ctype.h>
-#include <zlib.h>
+#include <zlib-ng.h>
 #include <stdio.h>
 #include <unistd.h>
 #include <stdlib.h>
diff --git a/src/ksw.cpp b/src/ksw.cpp
index 4b93b8a..03397bd 100644
--- a/src/ksw.cpp
+++ b/src/ksw.cpp
@@ -684,7 +684,7 @@ int ksw_global(int qlen, const uint8_t *query, int tlen, const uint8_t *target,
 
 #include <unistd.h>
 #include <stdio.h>
-#include <zlib.h>
+#include <zlib-ng.h>
 #include "kseq.h"
 KSEQ_INIT(gzFile, err_gzread)
 
diff --git a/src/utils.cpp b/src/utils.cpp
index d90d4f7..54c2940 100644
--- a/src/utils.cpp
+++ b/src/utils.cpp
@@ -30,7 +30,7 @@
 #include <stdarg.h>
 #include <stdlib.h>
 #include <string.h>
-#include <zlib.h>
+#include <zlib-ng.h>
 #include <errno.h>
 #ifdef FSYNC_ON_FLUSH
 #include <sys/types.h>
@@ -77,7 +77,7 @@ gzFile err_xzopen_core(const char *func, const char *fn, const char *mode)
 	gzFile fp;
 	if (strcmp(fn, "-") == 0) {
 		fp = gzdopen(fileno((strstr(mode, "r"))? stdin : stdout), mode);
-		/* According to zlib.h, this is the only reason gzdopen can fail */
+		/* According to zlib-ng.h, this is the only reason gzdopen can fail */
 		if (!fp) err_fatal(func, "Out of memory");
 		return fp;
 	}
diff --git a/src/utils.h b/src/utils.h
index d26209d..55049d5 100644
--- a/src/utils.h
+++ b/src/utils.h
@@ -30,7 +30,7 @@
 
 #include <stdint.h>
 #include <stdio.h>
-#include <zlib.h>
+#include <zlib-ng.h>
 
 #ifdef __GNUC__
 // Tell GCC to validate printf format string and args
