Index: src/CMakeLists.txt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
--- a/src/CMakeLists.txt	(revision b08613bc947a3df628b2c2f018414e0e6d6f8593)
+++ b/src/CMakeLists.txt	(date 1668010688378)
@@ -12,7 +12,7 @@
   STAMP_DIR ${PROJECT_SOURCE_DIR}/Bifrost/build
   SOURCE_DIR ${PROJECT_SOURCE_DIR}/Bifrost
   PREFIX ${PROJECT_SOURCE_DIR}/Bifrost
-  CMAKE_ARGS -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER} -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER} -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE} -DMAX_KMER_SIZE=${MAX_KMER_SIZE} -DCOMPILATION_ARCH=OFF DENABLE_AVX2=OFF
+  CMAKE_ARGS -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER} -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER} -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE} -DMAX_KMER_SIZE=${MAX_KMER_SIZE} -DCOMPILATION_ARCH=OFF -DENABLE_AVX2=OFF
   UPDATE_COMMAND ""	
   INSTALL_COMMAND "")	
 
