diff --git a/makefile b/makefile
index a240538..be0d861 100644
--- a/makefile
+++ b/makefile
@@ -1,7 +1,7 @@
 PLATFORM=$(shell uname -s)

 GPP=$(CXX)
-CPPFLAGS=-Wall -Wextra -std=c++17 -O3 -g -Izstr/src -Iparallel-hashmap/parallel_hashmap/ -Wno-unused-parameter -Icxxopts/include -Iconcurrentqueue
+CPPFLAGS=-Wall -Wextra -std=c++17 -O3 -g -Izstr/src -Iparallel-hashmap/parallel_hashmap/ -Wno-unused-parameter -Icxxopts/include -Iconcurrentqueue -D_LIBCPP_DISABLE_AVAILABILITY

 ODIR=obj
 BINDIR=bin