diff --git a/makefile b/makefile
index 0c0113b..ddbd6b1 100644
--- a/makefile
+++ b/makefile
@@ -22,7 +22,9 @@ else
    LINKFLAGS = $(CPPFLAGS) $(LIBS) -lpthread -pthread -static-libstdc++
 endif
 
-VERSION := Branch $(shell git rev-parse --abbrev-ref HEAD) commit $(shell git rev-parse HEAD) $(shell git show -s --format=%ci)
+CPPFLAGS := $(CXXFLAGS) $(CPPFLAGS)
+LINKFLAGS := $(LDFLAGS) $(LINKFLAGS)
+VERSION := bioconda $(PKG_VERSION)
 
 $(shell mkdir -p bin)
 $(shell mkdir -p obj)
