diff --git a/makefile b/makefile
index 1f409f7..f1cfd7a 100644
--- a/makefile
+++ b/makefile
@@ -15,7 +15,9 @@ OBJ = $(patsubst %, $(ODIR)/%, $(_OBJ))
 
 LINKFLAGS = $(CPPFLAGS) -Wl,-Bstatic $(LIBS) -Wl,-Bdynamic -Wl,--as-needed -lpthread -pthread -static-libstdc++
 
-VERSION := Branch $(shell git rev-parse --abbrev-ref HEAD) commit $(shell git rev-parse HEAD) $(shell git show -s --format=%ci)
+CPPFLAGS := $(CXXFLAGS) $(CPPFLAGS)
+LINKFLAGS := $(LDFLAGS) $(LINKFLAGS)
+VERSION := bioconda $(PKG_VERSION)
 
 $(shell mkdir -p bin)
 $(shell mkdir -p obj)
 