diff --git a/setup.py b/setup.py
index 829a542..df15e64 100644
--- a/setup.py
+++ b/setup.py
@@ -1,22 +1,22 @@
-from setuptools import Extension, setup
+from setuptools import Extension, setup, find_namespace_packages
 from Cython.Build import cythonize
 import subprocess
 from os import environ
 
-CONDA_PREFIX = environ.get('CONDA_PREFIX', None)
+CONDA_PREFIX = environ.get('PREFIX', None)
 if CONDA_PREFIX == None:
     print("Can't find conda installation- aborting.")
     exit(1)
 
 #build the protoc reader files for this system and put them in the src
-subprocess.run("protoc --cpp_out=../ parsimony.proto",shell=True,check=True,cwd='src/usher')
+subprocess.run(["protoc --cpp_out=../ parsimony.proto"], shell=True, check=True, cwd="src/usher")
 
 extensions = [Extension("bte",["src/bte.pyx"],
-    include_dirs=[CONDA_PREFIX+"/include/google/protobuf", CONDA_PREFIX+"/include/"],
-    library_dirs=[CONDA_PREFIX+'/lib/'],
-    libraries=['tbb', 'protobuf','boost_system','boost_iostreams'],
+    include_dirs=[CONDA_PREFIX+"/include/google/protobuf", CONDA_PREFIX+"/include"],
+    library_dirs=[CONDA_PREFIX+'/lib'],
+    libraries=['tbb', 'protobuf', 'boost_system', 'boost_iostreams'],
     language='c++',
-    extra_compile_args=["-std=c++17"]
+    extra_compile_args=["-std=c++14", "-O3"]
     )]
 
-setup(name='bte',ext_modules=cythonize(extensions,language_level=3,annotate=True))
+setup(name = 'bte', packages=find_namespace_packages(where="."), ext_modules = cythonize(extensions, language_level=3))
