diff --git pyDNase/scripts/dnase_average_profile.py pyDNase/scripts/dnase_average_profile.py
index c7876ab..f7427e4 100755
--- pyDNase/scripts/dnase_average_profile.py
+++ pyDNase/scripts/dnase_average_profile.py
@@ -1,3 +1,4 @@
+#!/usr/bin/env python
 import argparse
 import pyDNase
 import numpy as np
diff --git pyDNase/scripts/dnase_bias_estimator.py pyDNase/scripts/dnase_bias_estimator.py
index 9e142fd..e6fbb41 100755
--- pyDNase/scripts/dnase_bias_estimator.py
+++ pyDNase/scripts/dnase_bias_estimator.py
@@ -1,3 +1,4 @@
+#!/usr/bin/env python
 import os, argparse, tempfile, operator, pickle
 from collections import defaultdict
 import pybedtools, pysam
diff --git pyDNase/scripts/dnase_cut_counter.py pyDNase/scripts/dnase_cut_counter.py
index b2ba1b8..abc8184 100755
--- pyDNase/scripts/dnase_cut_counter.py
+++ pyDNase/scripts/dnase_cut_counter.py
@@ -1,3 +1,4 @@
+#!/usr/bin/env python
 import argparse
 import pyDNase
 from clint.textui import progress, puts
diff --git pyDNase/scripts/dnase_ddhs_scorer.py pyDNase/scripts/dnase_ddhs_scorer.py
index 1e1a38c..f7e0104 100755
--- pyDNase/scripts/dnase_ddhs_scorer.py
+++ pyDNase/scripts/dnase_ddhs_scorer.py
@@ -1,3 +1,4 @@
+#!/usr/bin/env python
 import argparse
 import pyDNase
 import math
diff --git pyDNase/scripts/dnase_to_JSON.py pyDNase/scripts/dnase_to_JSON.py
index 691d310..74cad6a 100755
--- pyDNase/scripts/dnase_to_JSON.py
+++ pyDNase/scripts/dnase_to_JSON.py
@@ -1,3 +1,4 @@
+#!/usr/bin/env python
 import argparse, pyDNase
 from clint.textui import puts, progress
 parser = argparse.ArgumentParser(description='Writes a JSON file of DNase I cuts for regions from a BED file')
diff --git pyDNase/scripts/dnase_to_javatreeview.py pyDNase/scripts/dnase_to_javatreeview.py
index 7858383..99ff957 100755
--- pyDNase/scripts/dnase_to_javatreeview.py
+++ pyDNase/scripts/dnase_to_javatreeview.py
@@ -1,3 +1,4 @@
+#!/usr/bin/env python
 import argparse, pyDNase, csv, random
 import numpy as np
 from clint.textui import puts, progress
diff --git pyDNase/scripts/dnase_wig_tracks.py pyDNase/scripts/dnase_wig_tracks.py
index 67a728f..f121af7 100755
--- pyDNase/scripts/dnase_wig_tracks.py
+++ pyDNase/scripts/dnase_wig_tracks.py
@@ -1,3 +1,4 @@
+#!/usr/bin/env python
 import argparse
 from clint.textui import progress, puts
 import pyDNase
diff --git pyDNase/scripts/wellington_bootstrap.py pyDNase/scripts/wellington_bootstrap.py
index fb809ac..bd56921 100755
--- pyDNase/scripts/wellington_bootstrap.py
+++ pyDNase/scripts/wellington_bootstrap.py
@@ -1,3 +1,4 @@
+#!/usr/bin/env python
 import pyDNase, pyDNase.footprinting
 import numpy as np
 from clint.textui import progress
diff --git pyDNase/scripts/wellington_footprints.py pyDNase/scripts/wellington_footprints.py
index 384fb39..545c4f3 100755
--- pyDNase/scripts/wellington_footprints.py
+++ pyDNase/scripts/wellington_footprints.py
@@ -1,3 +1,4 @@
+#!/usr/bin/env python
 import multiprocessing as mp
 import argparse, os
 from clint.textui import progress, puts_err
