--- src/Makefile.nocuda	2016-11-17 13:20:18.000000000 +0100
+++ src/Makefile.nocuda	2017-10-13 13:37:44.849366669 +0200
@@ -1,5 +1,7 @@
 CXX	:= g++
 CC := gcc
+#CC := clang-omp
+#CXX := clang-omp++
 GPUCC := nvcc
 
 # 32bit
@@ -11,10 +13,10 @@
 #GPUFLAGS := -arch=sm_20 -O2 -m32
 
 CXXFLAGS := -Wall -Wextra -Weffc++ -pedantic -std=c++98 -pipe -fopenmp -g -O2
-CFLAGS := -g -O3 -Wall -pipe -fopenmp
+CFLAGS := -g -O2 -Wall -pipe -fopenmp
 LDFLAGS := `gsl-config --libs`
 INCLUDES := -I. `gsl-config --cflags`
-LIBS := -lgomp
+LIBS := -liomp5
 GPUFLAGS := -arch=sm_20 -O2
 
 OBJECTS	:= \
