From a66ffa0d4b5e0aca5a1de1ffad387eadef49447f Mon Sep 17 00:00:00 2001
From: Travis Wrightsman <tw493@cornell.edu>
Date: Mon, 5 Feb 2024 12:21:17 -0500
Subject: [PATCH] Comment out SWIG package search

---
 CMakeLists.txt | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 80455af..0e0c66c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -71,11 +71,11 @@ endif (BUILD_UNIVERSAL)
 #for finding external libraries. 
 #see /usr/local/share/cmake-2.6/Modules/Find*.cmake for more examples
 
-find_package(SWIG REQUIRED)
-include(${SWIG_USE_FILE})
+#find_package(SWIG REQUIRED)
+#include(${SWIG_USE_FILE})
 
 include_directories(${CMAKE_CURRENT_SOURCE_DIR})
-set(CMAKE_SWIG_FLAGS "")
+#set(CMAKE_SWIG_FLAGS "")
 
 set(CMAKE_MODULE_PATH "${PEER_SOURCE_DIR}/cmake" ${CMAKE_MODULE_PATH})
 include_directories(./include)
-- 
2.39.2

