--- a/src/busco/run_BUSCO.py
+++ b/src/busco/run_BUSCO.py
@@ -126,7 +126,7 @@ class BuscoMaster:
         elif os.path.exists("/.singularity.d"):
             self.run_data["distribution"] = "singularity"
         else:
-            self.run_data["distribution"] = "manual"
+            self.run_data["distribution"] = "conda"
 
     def get_download_url(self):
         if self.config.getboolean("busco_run", "offline"):
