--- a/src/busco/ConfigManager.py
+++ b/src/busco/ConfigManager.py
@@ -42,10 +42,12 @@ class BuscoConfigManager:
                 return
         except KeyError:
             pass
-        if os.environ.get("BUSCO_CONFIG_FILE") and os.access(
-            os.environ.get("BUSCO_CONFIG_FILE"), os.R_OK
-        ):
-            self.config_file = os.environ.get("BUSCO_CONFIG_FILE")
+        config_file = (
+            os.environ.get("BUSCO_CONFIG_FILE")
+            or "/opt/anaconda1anaconda2anaconda3/share/busco/config.ini"
+        )
+        if config_file and os.access(config_file, os.R_OK):
+            self.config_file = config_file
         else:
             self.config_file = "local environment"
         return self.config_file

