--- a/src/busco/run_BUSCO.py
+++ b/src/busco/run_BUSCO.py
@@ -203,7 +203,7 @@ class BuscoMaster:
         elif os.path.exists("/.singularity.d"):
             BuscoMaster.run_stats["distribution"] = "singularity"
         else:
-            BuscoMaster.run_stats["distribution"] = "manual"
+            BuscoMaster.run_stats["distribution"] = "conda"
 
     def get_download_url(self):
         if self.config.getboolean("busco_run", "offline"):
