--- a/src/busco/run_BUSCO.py
+++ b/src/busco/run_BUSCO.py
@@ -198,7 +198,7 @@ class BuscoMaster:
         elif os.path.exists("/.singularity.d"):
             type(self).run_stats["distribution"] = "singularity"
         else:
-            type(self).run_stats["distribution"] = "manual"
+            type(self).run_stats["distribution"] = "conda"
 
     def get_download_url(self):
         if self.config.getboolean("busco_run", "offline"):
