--- a/src/busco/run_BUSCO.py
+++ b/src/busco/run_BUSCO.py
@@ -121,18 +121,12 @@ class BuscoMaster:
             logger.debug("File upload failed. Status code: {}".format(response.status))
 
     def get_dist_info(self):
-        if (  # todo: fix
-            "conda" in __file__
-            and os.environ("CONDA_DEFAULT_ENV")
-            and os.environ("CONDA_PREFIX")
-        ):
-            self.run_data["distribution"] = "conda"
-        elif os.path.exists("/.dockerenv"):
+        if os.path.exists("/.dockerenv"):
             self.run_data["distribution"] = "docker"
         elif os.path.exists("/.singularity.d"):
             self.run_data["distribution"] = "singularity"
         else:
-            self.run_data["distribution"] = "manual"
+            self.run_data["distribution"] = "conda"
 
     def get_download_url(self):
         if self.config.getboolean("busco_run", "offline"):
