diff --git a/src/tokenst.cpp b/src/tokenst.cpp
index 2ab6454..badfade 100644
--- src/tokenst.cpp
+++ src/tokenst.cpp
@@ -206,7 +206,13 @@ namespace OpenBabel
     string file;
     const char* datadir = getenv(envvar.c_str());
     if(!datadir)
+    {
       datadir = BABEL_DATADIR;
+    }
+    //puts(datadir);
+    std::string tempstring = string(datadir);
+    datadir = tempstring.c_str();
+    //puts(datadir);
 
     // check the subdirectory for this version number
     file = datadir;