From 44c0593fa49e1e7281024ae2229e07c11515ead7 Mon Sep 17 00:00:00 2001
From: Robert McGibbon <rmcgibbo@gmail.com>
Date: Tue, 13 Oct 2015 18:02:09 -0700
Subject: [PATCH] patch

---
 setup.py | 2 ++
 1 file changed, 2 insertions(+)

diff --git scripts/python/setup.py scripts/python/setup.py
index ae2f5f4..af286d2 100644
--- scripts/python/setup.py
+++ scripts/python/setup.py
@@ -97,6 +97,8 @@ class CustomBuildExt(build_ext):
         include_dirs, library_dirs = locate_ob()
         self.include_dirs.append(include_dirs)
         self.library_dirs.append(library_dirs)
+        if 'OPENBABEL_INCLUDE_DIRS' in os.environ:
+            self.include_dirs.extend(os.environ['OPENBABEL_INCLUDE_DIRS'].split(':'))
         self.swig_opts = ['-c++', '-small', '-O', '-templatereduce', '-naturalvar']
         self.swig_opts += ['-I%s' % i for i in self.include_dirs]
         print('- include_dirs: %s\n- library_dirs: %s' % (self.include_dirs, self.library_dirs))
--
2.3.4

