--- 1.9/plink_common.c	2025-03-17 19:37:00
+++ 1.9/plink_common.c	2025-03-17 19:37:41
@@ -5889,11 +5889,11 @@
 void bitvec_ornot(const uintptr_t* __restrict inverted_or_bitvec, uintptr_t word_ct, uintptr_t* __restrict main_bitvec) {
   // main_bitvec := main_bitvec OR (~inverted_or_bitvec)
 #ifdef __LP64__
-#ifdef __APPLE__
-  const __m128i all1 = {0xffffffffffffffffLLU, 0xffffffffffffffffLLU};
-#else
+//#ifdef __APPLE__
+//  const __m128i all1 = {0xffffffffffffffffLLU, 0xffffffffffffffffLLU};
+//#else
   const __m128i all1 = {-1LL, -1LL};
-#endif
+// #endif
   __m128i* vv128 = (__m128i*)main_bitvec;
   const __m128i* ev128 = (const __m128i*)inverted_or_bitvec;
   __m128i* vv128_end = &(vv128[word_ct / 2]);
