--- makefile	2015-07-27 14:18:26.000000000 +0000
+++ makefile.new	2015-12-07 19:48:19.942611888 +0000
@@ -11,7 +11,7 @@
 USEMPI ?= no
 64BIT_VERSION ?= yes
 USEREADLINE ?= yes
-USECOMPRESSION ?= no
+USECOMPRESSION ?= yes
 USEBOOST ?= yes
 MOTHUR_FILES="\"Enter_your_default_path_here\""
 RELEASE_DATE = "\"7/27/2015\""
@@ -23,7 +23,7 @@

 ifeq  ($(strip $(64BIT_VERSION)),yes)
     #if you are a mac user use the following line
-    TARGET_ARCH += -arch x86_64
+    #TARGET_ARCH += -arch x86_64

     #if you using cygwin to build Windows the following line
     #CXX = x86_64-w64-mingw32-g++
@@ -31,7 +31,7 @@
     #TARGET_ARCH += -m64 -static

     #if you are a linux user use the following line
-    #CXXFLAGS += -mtune=native -march=native
+    CXXFLAGS += -mtune=core2

     CXXFLAGS += -DBIT_VERSION
 endif
@@ -64,14 +64,14 @@

 #The boost libraries allow you to read gz files.
 ifeq  ($(strip $(USEBOOST)),yes)
-    BOOST_INCLUDE_DIR="/usr/local/include"
-    BOOST_LIBRARY_DIR="/usr/local/lib"
-
-    CXXFLAGS += -DUSE_BOOST
+    BOOST_INCLUDE_DIR=${PREFIX}/include
+    BOOST_LIBRARY_DIR=${PREFIX}/lib

+    CXXFLAGS += -DUSE_BOOST -I${BOOST_INCLUDE_DIR} -L${BOOST_LIBRARY_DIR}
+    LDFLAGS += -L${BOOST_LIBRARY_DIR}
     LIBS += \
-    ${BOOST_LIBRARY_DIR}/libboost_iostreams.a \
-    ${BOOST_LIBRARY_DIR}/zlib.a
+    -lboost_iostreams \
+    -lz

     #if linux or windows then ${BOOST_LIBRARY_DIR}/libz.a
 endif
