diff --git i/CMakeLists.txt w/CMakeLists.txt
index c587912..4b2a2a9 100644
--- i/CMakeLists.txt
+++ w/CMakeLists.txt
@@ -150,13 +150,8 @@ IF(USE_SSE_SUPPORT)
 
 	add_definitions("-DSSE3_ENABLED")
 
-	IF(APPLE AND CMAKE_SYSTEM_PROCESSOR STREQUAL "arm64")
-		MESSAGE(STATUS "Using sse2neon for arm64 Apple M1")
-		SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -march=armv8-a+fp+simd+crypto+crc")
-	ELSE()
-		MESSAGE(STATUS "Using SSE ${SSE_LEVEL}")
-		SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -m${SSE_LEVEL}")
-	ENDIF()
+		MESSAGE(STATUS "Using sse2neon for aarch64")
+		SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -march=native")
 
 	SET(${LIBRARY_NAME}_SSE_ENABLED "1")
 	SET(${LIBRARY_NAME}_SSE_LEVEL ${SSE_LEVEL})
@@ -401,4 +396,4 @@ ENDIF (BUILD_TESTING)
 
 IF (BUILD_BENCHMARKING)
 	add_subdirectory(examples)
-ENDIF (BUILD_BENCHMARKING)
\ No newline at end of file
+ENDIF (BUILD_BENCHMARKING)
