diff --git a/CMakeLists.txt b/CMakeLists.txt
index fe8b10b..dd5e92a 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -150,13 +150,8 @@ IF(USE_SSE_SUPPORT)
 
 	add_definitions("-DSSE3_ENABLED")
 
-	IF(APPLE AND CMAKE_SYSTEM_PROCESSOR STREQUAL "arm64")
-		MESSAGE(STATUS "Using sse2neon for arm64 Apple M1")
-		SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -march=armv8-a+fp+simd+crypto+crc")
-	ELSE()
-		MESSAGE(STATUS "Using SSE ${SSE_LEVEL}")
-		SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -m${SSE_LEVEL}")
-	ENDIF()
+		MESSAGE(STATUS "Using sse2neon for aarch64")
+		SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -march=native")
 
 	SET(${LIBRARY_NAME}_SSE_ENABLED "1")
 	SET(${LIBRARY_NAME}_SSE_LEVEL ${SSE_LEVEL})
