diff --git a/build/ld.linux.exe.sh b/build/ld.linux.exe.sh
index e8b7580..62ffb72 100755
--- a/build/ld.linux.exe.sh
+++ b/build/ld.linux.exe.sh
@@ -296,7 +296,7 @@ fi
 # add in xml
 if grep -q HAVE_XML2 ${BUILD_DIR}/Makefile.config.linux.$ARCH > /dev/null
 then
-    if grep -q XML2_LIBDIR ${BUILD_DIR}/Makefile.config.linux.$ARCH > /dev/null
+    if true || grep -q XML2_LIBDIR ${BUILD_DIR}/Makefile.config.linux.$ARCH > /dev/null
     then
         if grep -q 'OS_DISTRIBUTOR = Ubuntu' ${BUILD_DIR}/Makefile.config.linux.$ARCH > /dev/null
         then
diff --git a/setup/konfigure.perl b/setup/konfigure.perl
index 3ec705d..31f1579 100644
--- a/setup/konfigure.perl
+++ b/setup/konfigure.perl
@@ -1697,7 +1697,7 @@ sub check_compiler {
             $flags = $n;
             $log = '                      int main() {                     }\n'
         } elsif ($n eq 'hdf5') {
-            $library = '-Wl,-Bstatic -lhdf5 -Wl,-Bdynamic -ldl -lm -lz';
+            $library = '-Wl,-Bstatic -lhdf5 -ldl -lm -lz';
             $log = '#include <hdf5.h>  \n int main() { H5close         (); }\n'
         } elsif ($n eq 'fuse') {
             $flags = '-D_FILE_OFFSET_BITS=64';
