--- setup/konfigure.perl
+++ setup/konfigure.perl
@@ -1655,7 +1655,7 @@
             $flags = $n;
             $log = '                      int main() {                     }\n'
         } elsif ($n eq 'hdf5') {
-            $library = '-Wl,-Bstatic -lhdf5 -Wl,-Bdynamic -ldl -lm -lz';
+            $library = '-Wl,-Bdynamic -lhdf5 -ldl -lm -lz';
             $log = '#include <hdf5.h>  \n int main() { H5close         (); }\n'
         } elsif ($n eq 'fuse') {
             $flags = '-D_FILE_OFFSET_BITS=64';
--- setup/package.prl
+++ setup/package.prl
@@ -44,7 +44,7 @@
               pkgpath => '/usr',
               usrpath => '$HOME',
               include => 'hdf5.h',
-              lib     => 'libhdf5.a',
+              lib     => 'libhdf5.$SHLX',
             },
             { name    => 'magic',
               option  => 'with-magic-prefix',
@@ -53,7 +53,7 @@
               pkgpath => '/usr',
               usrpath => '$HOME',
               include => 'magic.h',
-              lib     => 'libmagic.so',
+              lib     => 'libmagic.$SHLX',
             },
             { name    => 'xml2',
               option  => 'with-xml2-prefix',
@@ -62,6 +62,6 @@
               pkgpath => '/usr',
               usrpath => '$HOME',
               include => 'libxml2/xmlreader.h',
-              lib     => 'libxml2.a',
+              lib     => 'libxml2.$SHLX',
         } ) }
 1
