diff --git a/tools/driver-tool/sratools.cpp b/tools/driver-tool/sratools.cpp
index 7fbe07c..8c200a9 100644
--- a/tools/driver-tool/sratools.cpp
+++ b/tools/driver-tool/sratools.cpp
@@ -54,6 +54,10 @@
 #include <sysexits.h>
 #endif
 
+#if MAC
+#include <mach-o/dyld.h>
+#endif
+
 #include "sratools2.hpp"
 #include "support2.hpp"
 #include "globals.hpp"
@@ -228,13 +232,18 @@ namespace sratools {
     std::string ToolPath::get_exec_path(std::string const &argv0, char *extra[])
     {
 #if MAC
-        if (extra) {
-            auto path = extra[0];
-            for (auto i = extra; *i; ++i) {
-                if (starts_with("executable_path=", *i)) {
-                    path = *i + 16;
-                    break;
-                }
+        {
+            std::string path(128, '\0');
+            uint32_t bufsize = path.size();
+            auto result = _NSGetExecutablePath(&path[0], &bufsize);
+            if (result < 0) {
+                path.resize(bufsize);
+                result = _NSGetExecutablePath(&path[0], &bufsize);
+            }
+            if (result == 0) {
+                path.resize(path.find('\0'));
+                return path;
+
             }
             if (path) {
                 if (path[0] == '/')
