diff -Naur a/tools/driver-tool/sratools.cpp b/tools/driver-tool/sratools.cpp
--- tools/driver-tool/sratools.cpp    2020-06-02 13:15:04.000000000 -0400
+++ tools/driver-tool/sratools.cpp   2020-06-02 13:17:49.000000000 -0400
@@ -152,9 +152,6 @@

             config = new Config(toolpath);
             struct Defer { ~Defer() { delete config; config = nullptr; } } freeConfig;
-            if (config->noInstallID()) {
-                printInstallMessage();
-            }

             auto const &what = sratools2::WhatImposter(toolpath);
             auto const &args = sratools2::Args(argc, argv, EnvironmentVariables::impersonate());
