diff --git a/tools/driver-tool/sratools.cpp b/tools/driver-tool/sratools.cpp
index 6b183afb..fdd65a13 100644
--- a/tools/driver-tool/sratools.cpp
+++ b/tools/driver-tool/sratools.cpp
@@ -143,9 +143,6 @@ namespace sratools {
         setenv(ENV_VAR_SESSION_ID, sessionID.c_str(), 1);
 
         config = new Config(toolpath);
-        if (config->noInstallID()) {
-            printInstallMessage();
-        }
         try {
             result = sratools2::main2( argc, argv, toolpath );
         }
