diff --git a/src/Makefile.MacOS b/src/Makefile.MacOS
index 5c96c75..e33aa51 100644
--- a/src/Makefile.MacOS
+++ b/src/Makefile.MacOS
@@ -3,7 +3,7 @@ include makefile.version
 
 CCFLAGS = ${MACOS}  -O9 -w -DMAKE_FOR_EXON  -D MAKE_STANDALONE -D SUBREAD_VERSION=\"${SUBREAD_VERSION}\" -D_FILE_OFFSET_BITS=64 ${WARNING_LEVEL}
 LDFLAGS =  -pthread -lz -lm ${MACOS} -DMAKE_FOR_EXON -D MAKE_STANDALONE # -DREPORT_ALL_THE_BEST
-CC = $(C) ${CCFLAGS} ${STATIC_MAKE} -ggdb -fomit-frame-pointer -O3 -ffast-math -funroll-loops -fmessage-length=0 
+CC = $(C) ${CCFLAGS} ${STATIC_MAKE} -ggdb -fomit-frame-pointer -O3 -mmmx -msse -msse2 -msse3 -ffast-math -funroll-loops -fmessage-length=0 
 
 
 ALL_LIBS= core core-junction core-indel sambam-file sublog gene-algorithms hashtable input-files sorted-hashtable gene-value-index exon-algorithms HelperFunctions interval_merge core-bigtable seek-zlib input-blc
