diff -urNp a/src/fullscan.c b/src/fullscan.c
--- a/src/fullscan.c	2023-05-07 08:53:37
+++ b/src/fullscan.c	2024-03-21 23:29:44
@@ -157,7 +157,7 @@ int main (int argc , char ** argv)
 {
 	char index_name [MAX_FILE_NAME_LENGTH];
 	char read_str [1208];
-	char c;
+	int c;
 	int i;
 
 	index_name[0]=0;
diff -urNp a/src/read-repair.c b/src/read-repair.c
--- a/src/read-repair.c	2023-05-07 08:53:37
+++ b/src/read-repair.c	2024-03-21 23:32:42
@@ -50,7 +50,7 @@ int main_read_repair(int argc, char ** argv)
 {
 	double t0 = miltime();
 	int threads = 8, is_BAM = 1;
-	char c;
+	int c;
 	char in_BAM_file[MAX_FILE_NAME_LENGTH+1];
 	char out_BAM_file[MAX_FILE_NAME_LENGTH+1];
 	char rand_prefix[40];
diff -urNp a/src/removeDupReads.c b/src/removeDupReads.c
--- a/src/removeDupReads.c	2023-05-07 08:53:37
+++ b/src/removeDupReads.c	2024-03-21 23:34:56
@@ -326,7 +326,7 @@ int main_repeated_test(int argc,char ** argv)
 int main_repeated_test(int argc,char ** argv)
 #endif
 {
-	char c;
+	int c;
 	char input_SAM_file[MAX_FILE_NAME_LENGTH];
 	char output_SAM_file[MAX_FILE_NAME_LENGTH];
 	char temp_path[MAX_FILE_NAME_LENGTH];
