--- gtdbtk/main.py	2019-06-21 09:38:14.000000000 +1000
+++ gtdbtk/main_new.py	2020-03-26 14:31:17.161971469 +1000
@@ -390,6 +390,7 @@

     def parse_options(self, options):
         """Parse user options and call the correct pipeline(s)"""
+        self.logger.warning('You are not running the most recent version of GTDB-Tk!')

         if(options.subparser_name == 'de_novo_wf'):
             check_dependencies(['prodigal', 'hmmalign'])
