From 48eeb27c8841133401aeff93e016bd93d16b4e8f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Tommi=20M=C3=A4klin?= <tommi@maklin.fi>
Date: Thu, 26 Oct 2023 13:33:57 +0300
Subject: [PATCH] Add missing #include <limits>

---
 src/mash/robin_hood.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/mash/robin_hood.h b/src/mash/robin_hood.h
index 573607f..9141848 100644
--- a/src/mash/robin_hood.h
+++ b/src/mash/robin_hood.h
@@ -42,6 +42,7 @@
 #include <cstdlib>
 #include <cstring>
 #include <functional>
+#include <limits>
 #include <memory> // only to support hash of smart pointers
 #include <stdexcept>
 #include <string>
