--- Makefile.in.old	2025-02-04 11:25:53.218617071 +0000
+++ Makefile.in	2021-02-26 23:13:33.000000000 +0000
@@ -5,9 +5,9 @@
 
 ifeq ($(UNAME_S),Darwin)
 	CXXFLAGS += -mmacosx-version-min=10.7 -stdlib=libc++
+#else
+#	CXXFLAGS += -include src/mash/memcpyLink.h -Wl,--wrap=memcpy
+#	CFLAGS += -include src/mash/memcpyLink.h
-else
-	CXXFLAGS += -include src/mash/memcpyLink.h -Wl,--wrap=memcpy
-	CFLAGS += -include src/mash/memcpyLink.h
 endif
 
 SOURCES=\
