diff --git a/CircSimu.pl b/CircSimu.pl
index 5a86f48..f5651fa 100644
--- a/CircSimu.pl
+++ b/CircSimu.pl
@@ -1,3 +1,5 @@
+#!/usr/bin/env perl
+
 use 5.012;
 use Getopt::Long;
 my ($fq1, $fq2, $out, $gtf, $coverage, $coverage2, $rand_mode, $rand_mode2, $read_length, $seq_err, $insert_length, $insert_length2, $perc_minor, $sigma, $sigma2, $ref_dir, $help, $if_chr, $exon_skipping, $psi, $circ_read);
diff --git a/JCcirc.pl b/JCcirc.pl
index 9c9bfd7..14f8dc4 100644
--- a/JCcirc.pl
+++ b/JCcirc.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 =begin SUMMARY
 Main steps:
