diff --git a/CMakeLists.txt b/CMakeLists.txt
index 28dcc13..3fb6974 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -41,14 +41,14 @@ OPTION(USE_BEAGLE "Use BEAGLE library." ON)
 include_directories(SYSTEM lib/boostable)
 
 # lcfit
-include_directories(lib/lcfit/lcfit_src)
-include_directories(lib/lcfit/lcfit_cpp_src)
-add_subdirectory(lib/lcfit)
+# include_directories(lib/lcfit/lcfit_src)
+# include_directories(lib/lcfit/lcfit_cpp_src)
+# add_subdirectory(lib/lcfit)
 
 # SMCTC
 add_definitions(-DSMCTC_HAVE_BGL)
 include_directories(SYSTEM lib/smctc/include)
-add_subdirectory(lib/smctc)
+# add_subdirectory(lib/smctc)
 
 # JsonCpp
 include_directories(lib/jsoncpp/include)
@@ -69,7 +69,7 @@ find_package(PkgConfig REQUIRED)
 IF(USE_BEAGLE)
 	pkg_check_modules(HMS_BEAGLE hmsbeagle-1 REQUIRED)
 	include_directories(${HMS_BEAGLE_INCLUDE_DIRS})
-	link_directories(${HMS_BEAGLE_LIBRARY_DIRS})
+#	link_directories(${HMS_BEAGLE_LIBRARY_DIRS})
 ELSE(USE_BEAGLE)
 	SET(NO_BEAGLE "1")
 	add_definitions("-DNO_BEAGLE")
