--- garnett_check_markers.R	2020-01-31 11:10:14.000000000 -0600
+++ garnett_check_markers.R.new	2020-02-08 03:52:06.674451167 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env Rscript
+#!/opt/anaconda1anaconda2anaconda3/bin/Rscript
 
 # Load optparse we need to check inputs
 suppressPackageStartupMessages(require(optparse))
--- garnett_classify_cells.R	2020-01-31 11:10:14.000000000 -0600
+++ garnett_classify_cells.R.new	2020-02-08 03:52:22.498500202 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env Rscript
+#!/opt/anaconda1anaconda2anaconda3/bin/Rscript
 
 # Load optparse we need to check inputs
 suppressPackageStartupMessages(require(optparse))
--- garnett_get_feature_genes.R	2020-01-31 11:10:14.000000000 -0600
+++ garnett_get_feature_genes.R.new	2020-02-08 03:52:53.058594901 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env Rscript
+#!/opt/anaconda1anaconda2anaconda3/bin/Rscript
 
 # Load optparse we need to check inputs
 suppressPackageStartupMessages(require(optparse))
--- garnett_train_classifier.R	2020-01-31 11:10:14.000000000 -0600
+++ garnett_train_classifier.R.new	2020-02-08 03:53:08.134641615 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env Rscript
+#!/opt/anaconda1anaconda2anaconda3/bin/Rscript
 
 # Load optparse we need to check inputs
 suppressPackageStartupMessages(require(optparse))
--- get_package_data.R	2020-01-31 11:10:14.000000000 -0600
+++ get_package_data.R.new	2020-02-08 03:53:22.850687224 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env Rscript 
+#!/opt/anaconda1anaconda2anaconda3/bin/Rscript 
 
 # Load optparse we need to check inputs
 suppressPackageStartupMessages(require(optparse))
--- transform_marker_file.R	2020-01-31 11:10:14.000000000 -0600
+++ transform_marker_file.R.new	2020-02-08 03:53:37.657733106 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env Rscript
+#!/opt/anaconda1anaconda2anaconda3/bin/Rscript
 
 # parse and transform marker file from SCXA's format into that accepted by Garnett
 # Marker gene files from SCXA come in tabular format. They are filtered according to p-value 
--- update_marker_file.R	2020-01-31 11:10:14.000000000 -0600
+++ update_marker_file.R.new	2020-02-08 03:53:48.969768155 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env Rscript 
+#!/opt/anaconda1anaconda2anaconda3/bin/Rscript 
 
 # Load optparse we need to check inputs
 suppressPackageStartupMessages(require(optparse))
