--- a/metagraph/CMakeLists.txt
+++ b/metagraph/CMakeLists.txt
@@ -345,6 +345,9 @@ add_subdirectory(external-libraries/zlib)
 target_compile_options(zlib
   PRIVATE
     -Wno-unused-variable
     -Wno-sign-compare
     -Wno-implicit-fallthrough
     -Wno-deprecated-non-prototype
     -Wno-unknown-warning-option
     -Wno-unused-parameter
     -Wno-maybe-uninitialized
 )
+
+# Add zlib include directory to sshash so it can find zlib.h
+target_include_directories(sshash_static PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/external-libraries/zlib)
