diff --git a/metagraph/external-libraries/htslib/htscodecs/htscodecs/rANS_static4x16pr.c b/metagraph/external-libraries/htslib/htscodecs/htscodecs/rANS_static4x16pr.c
index b6d5580..4d7e259 100644
--- a/metagraph/external-libraries/htslib/htscodecs/htscodecs/rANS_static4x16pr.c
+++ b/metagraph/external-libraries/htslib/htscodecs/htscodecs/rANS_static4x16pr.c
@@ -1010,6 +1010,8 @@ unsigned char *(*rans_dec_func(int do_simd, int order))
 
 #if defined(__linux__) || defined(__FreeBSD__)
 #include <sys/auxv.h>
+// Ensure ARM HWCAP_* values are defined even on old glibc
+#include <asm/hwcap.h>
 #elif defined(_WIN32)
 #include <processthreadsapi.h>
 #endif
