--- src/py/valet.py	2017-08-01 23:04:55.000000000 +0200
+++ src/py/valet.py	2020-07-29 13:56:29.524490605 +0200
@@ -13,7 +13,7 @@
 import subprocess
 from subprocess import Popen, PIPE
 
-BASE_PATH = os.path.dirname(sys.argv[0])[:-len('src/py/')]
+BASE_PATH = os.path.dirname(os.path.realpath(__file__))[:-len('src/py/')]
 FILE_LIMIT = resource.getrlimit(resource.RLIMIT_NOFILE)[1]
 FNULL = open('/dev/null', 'w')
 COMMANDS_FILE = None
@@ -524,7 +524,7 @@
     call_arr = ["samtools", "view", "-F", "0x100", "-bS", sam_output_location]
     run(call_arr, bam_fp, error_fp)
 
-    call_arr = ["samtools", "sort", "-@ 16", "-o", sorted_bam_location+".bam", bam_location]
+    call_arr = ["samtools", "sort", "-o", sorted_bam_location+".bam", bam_location]
     run(call_arr, stderr=FNULL)
 
     coverage_file_dir = output_dir + "/coverage/"
