diff --git a/scripts/estimate_tumor_normal_contamination.py b/scripts/estimate_tumor_normal_contamination.py
index b1af6e7..c9bb93a 100755
--- a/scripts/estimate_tumor_normal_contamination.py
+++ b/scripts/estimate_tumor_normal_contamination.py
@@ -13,7 +13,6 @@
 import sys
 import os
 import optparse
-import imp
 from collections import defaultdict
 import numpy as np
 from math import pow
@@ -38,10 +37,10 @@ if opts.conpair_dir:
 else:
     CONPAIR_DIR = os.environ['CONPAIR_DIR']
 
-ContaminationModel = imp.load_source('/ContaminationModel', CONPAIR_DIR + '/modules/ContaminationModel.py')
-ContaminationMarker = imp.load_source('/ContaminationMarker', CONPAIR_DIR + '/modules/ContaminationMarker.py')
-MathOperations = imp.load_source('/MathOperations', CONPAIR_DIR + '/modules/MathOperations.py')
-Genotypes = imp.load_source('/Genotypes', CONPAIR_DIR + '/modules/Genotypes.py')
+import ContaminationModel
+import ContaminationMarker
+import MathOperations
+import Genotypes
 
 if not opts.tumor_pileup or not opts.normal_pileup:
     parser.print_help()
