diff --git a/scripts/estimate_tumor_normal_contamination.py b/scripts/estimate_tumor_normal_contamination.py
index c9bb93a..8fa4473 100755
--- a/scripts/estimate_tumor_normal_contamination.py
+++ b/scripts/estimate_tumor_normal_contamination.py
@@ -1,4 +1,4 @@
-#!/usr/bin/env python2.7
+#!/usr/bin/env python
 
 # New York Genome Center
 # SOFTWARE COPYRIGHT NOTICE AGREEMENT
diff --git a/scripts/run_gatk_pileup_for_sample.py b/scripts/run_gatk_pileup_for_sample.py
index 66dbba2..20587d9 100755
--- a/scripts/run_gatk_pileup_for_sample.py
+++ b/scripts/run_gatk_pileup_for_sample.py
@@ -1,4 +1,4 @@
-#!/usr/bin/env python2.7
+#!/usr/bin/env python
 
 # New York Genome Center
 # SOFTWARE COPYRIGHT NOTICE AGREEMENT
diff --git a/scripts/verify_concordance.py b/scripts/verify_concordance.py
index 3bf6071..b75f7f9 100755
--- a/scripts/verify_concordance.py
+++ b/scripts/verify_concordance.py
@@ -1,4 +1,4 @@
-#!/usr/bin/env python2.7
+#!/usr/bin/env python
 
 # New York Genome Center
 # SOFTWARE COPYRIGHT NOTICE AGREEMENT
