diff --git a/src/branch_and_bound.h b/src/branch_and_bound.h
index d6e0102..3431300 100644
--- a/src/branch_and_bound.h
+++ b/src/branch_and_bound.h
@@ -41,7 +41,7 @@ private:
 public:
 	typedef uint64_t id_type;
 	typedef list<uint64_t, Pool_alloc<id_type> > id_list_type;
-	typedef map<int, id_list_type, std::less<int>, Pool_alloc<std::pair<int, id_list_type > > > mapType;
+	typedef map<int, id_list_type, std::less<int>, Pool_alloc<std::pair<const int, id_list_type > > > mapType;
 	bool debug;
 
 	static Kmer_Generator_Reloaded& getInstance() {
