--- Makefile.linux	2017-10-09 10:59:03.917147807 +0200
+++ Makefile	2017-10-09 14:25:31.597690686 +0200
@@ -3,13 +3,13 @@
 #.SUFFIXES:      .C $(SUFFIXES)
 
 #For Linux and DEC Alpha, do this:
-SYS= -D_SYS_OSF
+#SYS= -D_SYS_OSF
 
 #for Solaris (SunOS), do this (uncomment out this):
 #SYS= -D_SYS_SOLARIS
 
 #For Mac OS X
-#SYS= -D_SYS_OSX
+SYS= -D_SYS_OSX
 
 #if above doesnt work, you might try this:
 #SYS= -D_SYS_$(shell echo `uname | sed -e s/SunOS/SOLARIS/ -e s/Linux/OSF/`)
@@ -24,13 +24,13 @@
 SRC= ./src/
 SYS_LIBS= -lm
 
-CC = gcc
+#CC = gcc
 
 #for DEC Alpha, or if you dont have gcc, comment out above and uncomment this:
 #CC = cc
 
 #For Mac OSX (must have developers tools), can do
-#CC = cc
+CC = cc
 
 # (if you have gcc on the Alpha or Mac, you can leave CC=gcc)
 
