--- userApps/kent/src/utils/expMatrixToBarchartBed/expMatrixToBarchartBed	(original)
+++ userApps/kent/src/utils/expMatrixToBarchartBed/expMatrixToBarchartBed	(refactored)
@@ -179,7 +179,7 @@
                 for group in open(groupOrder, "r"): 
                     bedInfo += group.strip("\n") + " " 
             else: 
-                for key, value in sorted(groupAverages.iteritems()):
+                for key, value in sorted(groupAverages.items()):
                     bedInfo += key + " "
                 bedInfo = bedInfo[:-1] + "\t_offset\t_lineLength"
 
@@ -201,7 +201,7 @@
                 count += 1.0
                 fullAverage += value
         else: 
-            for key, value in sorted(groupAverages.iteritems()): 
+            for key, value in sorted(groupAverages.items()): 
                 if (useMean):
                     bedLine = bedLine + "," + floatRound(str(value))
                     fullAverage += value
@@ -240,7 +240,7 @@
         count +=1 
         splitLine = item.strip("\n").split()
         if (len(splitLine) is not 2):
-            print ("There was an error reading the sample file at line " + str(count))
+            print(("There was an error reading the sample file at line " + str(count)))
             exit(1)
         sampleToGroup.setdefault(splitLine[0], splitLine[1])
 
@@ -327,7 +327,7 @@
     cmd = "cat " + indexedBedFile.name + " >> " + options.outputFile.name
     os.system(cmd)
 
-    print ("The columns and order of the groups are; \n" + bedInfo)
+    print(("The columns and order of the groups are; \n" + bedInfo))
 
 def main(args):
     """
