From 29e33e4636163d5bcb6c96fbfae8a1b9ad251008 Mon Sep 17 00:00:00 2001
From: astaric <anze.staric@gmail.com>
Date: Thu, 18 Jul 2019 10:03:06 +0200
Subject: [PATCH] fix tests

---
 t/02_diff_fully-pp.t | 2 +-
 t/02_diff_fully.t    | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/t/02_diff_fully-pp.t b/t/02_diff_fully-pp.t
index 69a96ab..27e32bd 100644
--- a/t/02_diff_fully-pp.t
+++ b/t/02_diff_fully-pp.t
@@ -30,7 +30,7 @@ new: perl
 --- data1
 -
  - ['-', 'perl']
--
+- []
 --- data2
 old: perl
 new: 
diff --git a/t/02_diff_fully.t b/t/02_diff_fully.t
index f816a89..6c4c8d8 100644
--- a/t/02_diff_fully.t
+++ b/t/02_diff_fully.t
@@ -29,7 +29,7 @@ new: perl
 --- data1
 -
  - ['-', 'perl']
--
+- []
 --- data2
 old: perl
 new: 
-- 
2.20.1 (Apple Git-117)

