diff --git a/CMakeLists.txt b/CMakeLists.txt
index 12d1881..d64a284 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.0)
+cmake_minimum_required(VERSION 3.10)
 cmake_policy(SET CMP0048 NEW)
 project(DR VERSION 0.1 LANGUAGES CXX)
 set(CMAKE_CXX_STANDARD 17)
@@ -22,7 +22,7 @@ set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${OpenMP_EXE_LINKER_FLAGS}
 set(CMAKE_COMPILE_OPTIONS "-Wall ${OpenMP_CXX_FLAGS}")
 set(CMAKE_CXX_FLAGS_RELEASE "-O3")
 set(CMAKE_CXX_FLAGS_DEBUG "-g")
-set(CMAKE_SHARED_LINKER_FLAGS "-Wall -Wc++-compat -O2 -msse4.1 -DHAVE_KALLOC -DKSW_CPU_DISPATCH -D_FILE_OFFSET_BITS=64 -ltbb -fsigned-char -fsanitize=address")
+set(CMAKE_SHARED_LINKER_FLAGS "-Wall -Wc++-compat -O3 -D_FILE_OFFSET_BITS=64 -ltbb -fsigned-char -fsanitize=address")
 
 include_directories(${CMAKE_CURRENT_SOURCE_DIR}/src/tools)
 include_directories(${CMAKE_CURRENT_SOURCE_DIR}/src/lib)
