--- bam/SamInterface.cpp	2012-10-24 15:53:29.000000000 -0500
+++ bam/SamInterface.cpp.new	2018-11-27 02:53:32.764220114 -0600
@@ -15,12 +15,16 @@
  *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#include <iostream>
+#define __STDC_LIMIT_MACROS
+#include <stdint.h>
+#include <limits>
+
+// using namespace std;
+
 #include "SamInterface.h"
 #include "SamRecordHelper.h"
 
-#include <limits>
-#include <stdint.h>
-
 SamInterface::SamInterface()
 {
 }
