--- bam/SamInterface.cpp	2012-10-24 15:53:29.000000000 -0500
+++ bam/SamInterface.cpp.new	2018-11-27 02:53:32.764220114 -0600
@@ -15,12 +15,16 @@
  *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#include <iostream>
+#define __STDC_LIMIT_MACROS
+#include <stdint.h>
+#include <limits>
+
+// using namespace std;
+
 #include "SamInterface.h"
 #include "SamRecordHelper.h"
 
-#include <limits>
-#include <stdint.h>
-
 SamInterface::SamInterface()
 {
 }
--- Makefiles/Makefile.include	2018-11-27 03:39:26.853405131 -0600
+++ Makefiles/Makefile.include.new	2018-11-27 03:40:14.109391148 -0600
@@ -44,7 +44,7 @@
 ZLIB_AVAIL ?= 1
 
 USE_ZLIB = -D__ZLIB_AVAILABLE__
-ZLIB_LIB = -lz
+ZLIB_LIB = -lz -L${PREFIX}/lib
 ifeq ($(ZLIB_AVAIL), 0)
   USE_ZLIB = 
   ZLIB_LIB = 
