diff --git a/src/Makefile b/src/Makefile
index 42ff113..12572c6 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -1,6 +1,6 @@
-override CFLAGS += -I../include -I/usr/include/openblas
+override CFLAGS += -I../include -I/usr/include/openblas -I$(PREFIX)/include
 #LDLIBS += -lgsl -lopenblas -lrt -lm
-override LDLIBS += -lgsl -lopenblas -lm -lpthread
+override LDLIBS += -lgsl -lopenblas -lgfortran -lm -pthread -L$(PREFIX)/lib
 # Some Linux distributions require separate lapacke library
 # override LDLIBS += -llapacke
 # Mac additions using homebrew installations
@@ -15,7 +15,7 @@ endif
 ifeq ($(DEBUG), 1)
 	override CFLAGS += -g # enable debugging
 else
-	override CFLAGS += -O2
+	override CFLAGS += -O3
 endif
 
 ifeq ($(PROFILING), 1)
