diff --git a/src/nicksrc/Makefile b/src/nicksrc/Makefile
index 2d1733c..cc678c0 100644
--- a/src/nicksrc/Makefile
+++ b/src/nicksrc/Makefile
@@ -1,10 +1,10 @@
-override CFLAGS += -c -O3 -p -Wimplicit -I../../include
+override CFLAGS += -c -O3 -Wimplicit -I../../include -I$(PREFIX)/include
 
 all: libnick.a
 
 libnick.a: strsubs.o sortit.o vsubs.o statsubs.o linsubs.o getpars.o xsearch.o gauss.o gds.o
-	ar -r libnick.a $^
-	ranlib libnick.a
+	$(AR) -rcs libnick.a $^
+	$(RANLIB) libnick.a
 
 clean:
 	rm -f *.o
