diff --git a/src/nicksrc/strsubs.c b/src/nicksrc/strsubs.c
index ffe0b88..de840f1 100644
--- a/src/nicksrc/strsubs.c
+++ b/src/nicksrc/strsubs.c
@@ -10,12 +10,14 @@
 #include <errno.h>
 #include <sys/types.h>
 #include <sys/stat.h>
+#ifdef __linux__
 #include <sys/prctl.h>
-#include <xsearch.h>   
+#endif
+#include <xsearch.h>
 
 
 #define MAXSTR 10000
-#define MAXFF  500 
+#define MAXFF  500
 
 #include "strsubs.h"
 #include "vsubs.h"
@@ -227,9 +229,9 @@ fatalx (char *fmt, ...)
 
 void setdump(int cdump) 
 {
-
-  prctl(PR_SET_DUMPABLE, cdump) ; 
-
+#ifdef __linux__
+    prctl(PR_SET_DUMPABLE, cdump);
+#endif
 
 }
 
