diff -u -r a/versions.py b/versions.py
--- a/versions.py	2016-09-18 17:16:43.000000000 +0200
+++ b/versions.py	2017-07-18 16:02:50.796251586 +0200
@@ -15,7 +15,7 @@
 __license__ = "MIT"
 
 
-SOFTWARE_DIR = '/work/odosenl/software/'
+SOFTWARE_DIR = './'
 if getpass.getuser() == 'tim':
     SOFTWARE_DIR = '/home/tim/Documents/dev/odosenl/software/'
 if not os.path.isdir(SOFTWARE_DIR):
@@ -107,29 +107,29 @@
     logging.info('BioPython\t%s', Bio.__version__)
 
     # Blast
-    logging.info('makeblastdb\t%s', _call_program(MAKEBLASTDB, '-version').split('\n')[1])
-    logging.info('blastp\t%s', _call_program(BLASTP, '-version').split('\n')[1])
-    logging.info('blastn\t%s', _call_program(BLASTN, '-version').split('\n')[1])
+    #logging.info('makeblastdb\t%s', _call_program(MAKEBLASTDB, '-version').split('\n')[1])
+    #logging.info('blastp\t%s', _call_program(BLASTP, '-version').split('\n')[1])
+    #logging.info('blastn\t%s', _call_program(BLASTN, '-version').split('\n')[1])
 
     # Life Science Grid Portal
-    logging.info('LSGP %s', LSGP_MAKEBLASTDB.replace('/', '\t'))
-    logging.info('LSGP %s', LSGP_BLASTP.replace('/', '\t'))
-    logging.info('LSGP %s', LSGP_BLASTN.replace('/', '\t'))
+    #logging.info('LSGP %s', LSGP_MAKEBLASTDB.replace('/', '\t'))
+    #logging.info('LSGP %s', LSGP_BLASTP.replace('/', '\t'))
+    #logging.info('LSGP %s', LSGP_BLASTN.replace('/', '\t'))
 
     # OrthoMCL & mcl
-    logging.info('OrthoMCL\t%s', _grep_version(ORTHOMCL_DIR + '../doc/OrthoMCLEngine/Main/releaseNotes.txt', pattern='v2'))
-    logging.info('MCL\t%s', _call_program(MCL, '--version').split('\n')[0])
+    #logging.info('OrthoMCL\t%s', _grep_version(ORTHOMCL_DIR + '../doc/OrthoMCLEngine/Main/releaseNotes.txt', pattern='v2'))
+    #logging.info('MCL\t%s', _call_program(MCL, '--version').split('\n')[0])
 
     # PAML codeml
-    logging.info('PAML codeml\t%s', _grep_version(PAML_DIR + 'src/paml.h'))
+    #logging.info('PAML codeml\t%s', _grep_version(PAML_DIR + 'src/paml.h'))
 
     # PHYLIP dnadist & neighbor
-    logging.info('PHYLIP dnadist\t%s', _grep_version(PHYLIP_DIR + 'src/dnadist.c')[3:])
-    logging.info('PHYLIP neighbor\t%s', _grep_version(PHYLIP_DIR + 'src/neighbor.c')[3:])
+    #logging.info('PHYLIP dnadist\t%s', _grep_version(PHYLIP_DIR + 'src/dnadist.c')[3:])
+    #logging.info('PHYLIP neighbor\t%s', _grep_version(PHYLIP_DIR + 'src/neighbor.c')[3:])
 
     # TranslatorX calls muscle internally
-    logging.info('translatorx\t%s', _grep_version(TRANSLATORX, pattern='TranslatorX v')[28:-6])
-    logging.info('Muscle\t%s', _call_program('muscle', '-version'))
+    #logging.info('translatorx\t%s', _grep_version(TRANSLATORX, pattern='TranslatorX v')[28:-6])
+    #logging.info('Muscle\t%s', _call_program('muscle', '-version'))
 
 if __name__ == '__main__':
     # Parse arguments to setup logging; not in main for testing
