diff -ruN PhaBOX-0.1.0-beta-org/Cherry_single.py PhaBOX-0.1.0-beta/Cherry_single.py
--- PhaBOX-0.1.0-beta-org/Cherry_single.py	2024-05-19 09:22:13.000000000 +0200
+++ PhaBOX-0.1.0-beta/Cherry_single.py	2024-09-27 15:57:51.264652843 +0200
@@ -15,21 +15,21 @@
 from torch import optim
 from torch.nn import functional as F
 
-from scripts.ulity import *
-from scripts.preprocessing import *
-from scripts.cnnscript import *
+from phabox.scripts.ulity import *
+from phabox.scripts.preprocessing import *
+from phabox.scripts.cnnscript import *
 from shutil import which
 from collections import Counter
 from Bio import SeqIO
 from Bio.SeqRecord import SeqRecord
 from Bio.Blast.Applications import NcbiblastnCommandline
-from models.phamer import Transformer
-from models.CAPCNN import WCNN
-from models.PhaGCN import GCN
-from models import Cherry
-from draw import draw_network, drop_network
+from phabox.models.phamer import Transformer
+from phabox.models.CAPCNN import WCNN
+from phabox.models.PhaGCN import GCN
+from phabox.models import Cherry
+from phabox.draw import draw_network, drop_network
 from scipy.special import softmax
-from scripts.data import load_data, preprocess_features, preprocess_adj, sample_mask
+from phabox.scripts.data import load_data, preprocess_features, preprocess_adj, sample_mask
 
 
 
diff -ruN PhaBOX-0.1.0-beta-org/draw.py PhaBOX-0.1.0-beta/draw.py
--- PhaBOX-0.1.0-beta-org/draw.py	2024-05-19 09:22:13.000000000 +0200
+++ PhaBOX-0.1.0-beta/draw.py	2024-09-27 16:42:00.561315534 +0200
@@ -4,8 +4,8 @@
 import numpy as np
 import argparse
 import subprocess
-from scripts.ulity import *
-from scripts.preprocessing import *
+from phabox.scripts.ulity import *
+from phabox.scripts.preprocessing import *
 import math
 import os
 import pandas as pd
diff -ruN PhaBOX-0.1.0-beta-org/main.py PhaBOX-0.1.0-beta/main.py
--- PhaBOX-0.1.0-beta-org/main.py	2024-05-19 09:22:13.000000000 +0200
+++ PhaBOX-0.1.0-beta/main.py	2024-09-27 15:58:22.416637217 +0200
@@ -1,3 +1,5 @@
+#!/usr/bin/env python
+
 import os
 import shutil
 import argparse
@@ -15,22 +17,22 @@
 from torch import optim
 from torch.nn import functional as F
 
-from scripts.ulity import *
-from scripts.preprocessing import *
-from scripts.cnnscript import *
+from phabox.scripts.ulity import *
+from phabox.scripts.preprocessing import *
+from phabox.scripts.cnnscript import *
 from shutil import which
 from collections import Counter
 from Bio import SeqIO
 from Bio.SeqRecord import SeqRecord
 from Bio.Blast.Applications import NcbiblastnCommandline
-from models.phamer import Transformer
-from models.CAPCNN import WCNN
-from models.PhaGCN import GCN
-from models import Cherry
-from draw import draw_network, drop_network
+from phabox.models.phamer import Transformer
+from phabox.models.CAPCNN import WCNN
+from phabox.models.PhaGCN import GCN
+from phabox.models import Cherry
+from phabox.draw import draw_network, drop_network
 from collections import Counter
 from scipy.special import softmax
-from scripts.data import load_data, preprocess_features, preprocess_adj, sample_mask
+from phabox.scripts.data import load_data, preprocess_features, preprocess_adj, sample_mask
 
 
 
Binary files PhaBOX-0.1.0-beta-org/.main.py.swp and PhaBOX-0.1.0-beta/.main.py.swp differ
diff -ruN PhaBOX-0.1.0-beta-org/PhaGCN_single.py PhaBOX-0.1.0-beta/PhaGCN_single.py
--- PhaBOX-0.1.0-beta-org/PhaGCN_single.py	2024-05-19 09:22:13.000000000 +0200
+++ PhaBOX-0.1.0-beta/PhaGCN_single.py	2024-09-27 15:58:44.480625708 +0200
@@ -12,21 +12,21 @@
 import scipy.sparse as sparse
 
 
-from scripts.ulity import *
-from scripts.preprocessing import *
-from scripts.cnnscript import *
+from phabox.scripts.ulity import *
+from phabox.scripts.preprocessing import *
+from phabox.scripts.cnnscript import *
 from shutil import which
 from collections import Counter
 from Bio import SeqIO
 from Bio.SeqRecord import SeqRecord
 from Bio.Blast.Applications import NcbiblastnCommandline
-from models.phamer import Transformer
-from models.CAPCNN import WCNN
-from models.PhaGCN import GCN
-from models import Cherry
-from draw import draw_network, drop_network
+from phabox.models.phamer import Transformer
+from phabox.models.CAPCNN import WCNN
+from phabox.models.PhaGCN import GCN
+from phabox.models import Cherry
+from phabox.draw import draw_network, drop_network
 from scipy.special import softmax
-from scripts.data import load_data, preprocess_features, preprocess_adj, sample_mask
+from phabox.scripts.data import load_data, preprocess_features, preprocess_adj, sample_mask
 
 from torch import nn
 from torch import optim
diff -ruN PhaBOX-0.1.0-beta-org/PhaMer_single.py PhaBOX-0.1.0-beta/PhaMer_single.py
--- PhaBOX-0.1.0-beta-org/PhaMer_single.py	2024-05-19 09:22:13.000000000 +0200
+++ PhaBOX-0.1.0-beta/PhaMer_single.py	2024-09-27 15:59:08.424612831 +0200
@@ -16,20 +16,20 @@
 from torch import optim
 from torch.nn import functional as F
 
-from scripts.ulity import *
-from scripts.preprocessing import *
-from scripts.cnnscript import *
+from phabox.scripts.ulity import *
+from phabox.scripts.preprocessing import *
+from phabox.scripts.cnnscript import *
 from shutil import which
 from collections import Counter
 from Bio import SeqIO
 from Bio.SeqRecord import SeqRecord
 from Bio.Blast.Applications import NcbiblastnCommandline
-from models.phamer import Transformer
-from models.CAPCNN import WCNN
-from models.PhaGCN import GCN
-from models import Cherry
+from phabox.models.phamer import Transformer
+from phabox.models.CAPCNN import WCNN
+from phabox.models.PhaGCN import GCN
+from phabox.models import Cherry
 from scipy.special import softmax
-from scripts.data import load_data, preprocess_features, preprocess_adj, sample_mask
+from phabox.scripts.data import load_data, preprocess_features, preprocess_adj, sample_mask
 
 
 
diff -ruN PhaBOX-0.1.0-beta-org/PhaTYP_single.py PhaBOX-0.1.0-beta/PhaTYP_single.py
--- PhaBOX-0.1.0-beta-org/PhaTYP_single.py	2024-05-19 09:22:13.000000000 +0200
+++ PhaBOX-0.1.0-beta/PhaTYP_single.py	2024-09-27 15:59:23.632604453 +0200
@@ -15,20 +15,20 @@
 from torch import optim
 from torch.nn import functional as F
 
-from scripts.ulity import *
-from scripts.preprocessing import *
-from scripts.cnnscript import *
+from phabox.scripts.ulity import *
+from phabox.scripts.preprocessing import *
+from phabox.scripts.cnnscript import *
 from shutil import which
 from collections import Counter
 from Bio import SeqIO
 from Bio.SeqRecord import SeqRecord
 from Bio.Blast.Applications import NcbiblastnCommandline
-from models.phamer import Transformer
-from models.CAPCNN import WCNN
-from models.PhaGCN import GCN
-from models import Cherry
+from phabox.models.phamer import Transformer
+from phabox.models.CAPCNN import WCNN
+from phabox.models.PhaGCN import GCN
+from phabox.models import Cherry
 from scipy.special import softmax
-from scripts.data import load_data, preprocess_features, preprocess_adj, sample_mask
+from phabox.scripts.data import load_data, preprocess_features, preprocess_adj, sample_mask
 
 
 
diff -ruN PhaBOX-0.1.0-beta-org/scripts/cnnscript.py PhaBOX-0.1.0-beta/scripts/cnnscript.py
--- PhaBOX-0.1.0-beta-org/scripts/cnnscript.py	2024-05-19 09:22:13.000000000 +0200
+++ PhaBOX-0.1.0-beta/scripts/cnnscript.py	2024-09-27 15:55:28.964713910 +0200
@@ -3,7 +3,7 @@
 import  numpy as np
 import  torch
 import  torch.utils.data as Data
-from    models.CAPCNN import WCNN
+from    phabox.models.CAPCNN import WCNN
 from    Bio import SeqIO
 from    Bio.SeqRecord import SeqRecord   
 from    torch import nn
diff -ruN PhaBOX-0.1.0-beta-org/scripts/data.py PhaBOX-0.1.0-beta/scripts/data.py
--- PhaBOX-0.1.0-beta-org/scripts/data.py	2024-05-19 09:22:13.000000000 +0200
+++ PhaBOX-0.1.0-beta/scripts/data.py	2024-09-27 16:59:29.216910933 +0200
@@ -2,7 +2,7 @@
 import  pickle as pkl
 import  networkx as nx
 import  scipy.sparse as sp
-from    scipy.sparse.linalg.eigen.arpack import eigsh
+from scipy.sparse.linalg import eigsh
 import  sys
 
 
