diff -ruN PhaBOX-1.0.0-org/Cherry_single.py PhaBOX-1.0.0/Cherry_single.py
--- PhaBOX-1.0.0-org/Cherry_single.py	2024-09-30 10:20:05.496290778 +0200
+++ PhaBOX-1.0.0/Cherry_single.py	2024-09-30 13:26:23.735605652 +0200
@@ -16,21 +16,21 @@
 from torch import optim
 from torch.nn import functional as F
 
-from scripts.ulity import *
-from scripts.preprocessing import *
-from scripts.cnnscript import *
+from phabox.scripts.ulity import *
+from phabox.scripts.preprocessing import *
+from phabox.scripts.cnnscript import *
 from shutil import which
 from collections import Counter
 from Bio import SeqIO
 from Bio.SeqRecord import SeqRecord
 from Bio.Blast.Applications import NcbiblastnCommandline
-from models.phamer import Transformer
-from models.CAPCNN import WCNN
-from models.PhaGCN import GCN
-from models import Cherry
-from draw import draw_network, drop_network
+from phabox.models.phamer import Transformer
+from phabox.models.CAPCNN import WCNN
+from phabox.models.PhaGCN import GCN
+from phabox.models import Cherry
+from phabox.draw import draw_network, drop_network
 from scipy.special import softmax
-from scripts.data import load_data, preprocess_features, preprocess_adj, sample_mask
+from phabox.scripts.data import load_data, preprocess_features, preprocess_adj, sample_mask
 
 
 
diff -ruN PhaBOX-1.0.0-org/draw.py PhaBOX-1.0.0/draw.py
--- PhaBOX-1.0.0-org/draw.py	2024-09-30 10:20:05.500290882 +0200
+++ PhaBOX-1.0.0/draw.py	2024-09-30 13:27:20.231425936 +0200
@@ -4,8 +4,8 @@
 import numpy as np
 import argparse
 import subprocess
-from scripts.ulity import *
-from scripts.preprocessing import *
+from phabox.scripts.ulity import *
+from phabox.scripts.preprocessing import *
 import math
 import os
 import pandas as pd
diff -ruN PhaBOX-1.0.0-org/main.py PhaBOX-1.0.0/main.py
--- PhaBOX-1.0.0-org/main.py	2024-09-30 10:20:05.500290882 +0200
+++ PhaBOX-1.0.0/main.py	2024-09-30 13:27:33.651385105 +0200
@@ -16,22 +16,22 @@
 from torch import optim
 from torch.nn import functional as F
 
-from scripts.ulity import *
-from scripts.preprocessing import *
-from scripts.cnnscript import *
+from phabox.scripts.ulity import *
+from phabox.scripts.preprocessing import *
+from phabox.scripts.cnnscript import *
 from shutil import which
 from collections import Counter
 from Bio import SeqIO
 from Bio.SeqRecord import SeqRecord
 from Bio.Blast.Applications import NcbiblastnCommandline
-from models.phamer import Transformer
-from models.CAPCNN import WCNN
-from models.PhaGCN import GCN
-from models import Cherry
-from draw import draw_network, drop_network
+from phabox.models.phamer import Transformer
+from phabox.models.CAPCNN import WCNN
+from phabox.models.PhaGCN import GCN
+from phabox.models import Cherry
+from phabox.draw import draw_network, drop_network
 from collections import Counter
 from scipy.special import softmax
-from scripts.data import load_data, preprocess_features, preprocess_adj, sample_mask
+from phabox.scripts.data import load_data, preprocess_features, preprocess_adj, sample_mask
 
 
 
diff -ruN PhaBOX-1.0.0-org/PhaGCN_single.py PhaBOX-1.0.0/PhaGCN_single.py
--- PhaBOX-1.0.0-org/PhaGCN_single.py	2024-09-30 10:20:05.496290778 +0200
+++ PhaBOX-1.0.0/PhaGCN_single.py	2024-09-30 13:26:46.099532952 +0200
@@ -13,21 +13,21 @@
 import scipy.sparse as sparse
 
 
-from scripts.ulity import *
-from scripts.preprocessing import *
-from scripts.cnnscript import *
+from phabox.scripts.ulity import *
+from phabox.scripts.preprocessing import *
+from phabox.scripts.cnnscript import *
 from shutil import which
 from collections import Counter
 from Bio import SeqIO
 from Bio.SeqRecord import SeqRecord
 from Bio.Blast.Applications import NcbiblastnCommandline
-from models.phamer import Transformer
-from models.CAPCNN import WCNN
-from models.PhaGCN import GCN
-from models import Cherry
-from draw import draw_network, drop_network
+from phabox.models.phamer import Transformer
+from phabox.models.CAPCNN import WCNN
+from phabox.models.PhaGCN import GCN
+from phabox.models import Cherry
+from phabox.draw import draw_network, drop_network
 from scipy.special import softmax
-from scripts.data import load_data, preprocess_features, preprocess_adj, sample_mask
+from phabox.scripts.data import load_data, preprocess_features, preprocess_adj, sample_mask
 
 from torch import nn
 from torch import optim
diff -ruN PhaBOX-1.0.0-org/PhaMer_single.py PhaBOX-1.0.0/PhaMer_single.py
--- PhaBOX-1.0.0-org/PhaMer_single.py	2024-09-30 10:20:05.500290882 +0200
+++ PhaBOX-1.0.0/PhaMer_single.py	2024-09-30 13:26:53.231510206 +0200
@@ -16,20 +16,20 @@
 from torch import optim
 from torch.nn import functional as F
 
-from scripts.ulity import *
-from scripts.preprocessing import *
-from scripts.cnnscript import *
+from phabox.scripts.ulity import *
+from phabox.scripts.preprocessing import *
+from phabox.scripts.cnnscript import *
 from shutil import which
 from collections import Counter
 from Bio import SeqIO
 from Bio.SeqRecord import SeqRecord
 from Bio.Blast.Applications import NcbiblastnCommandline
-from models.phamer import Transformer
-from models.CAPCNN import WCNN
-from models.PhaGCN import GCN
-from models import Cherry
+from phabox.models.phamer import Transformer
+from phabox.models.CAPCNN import WCNN
+from phabox.models.PhaGCN import GCN
+from phabox.models import Cherry
 from scipy.special import softmax
-from scripts.data import load_data, preprocess_features, preprocess_adj, sample_mask
+from phabox.scripts.data import load_data, preprocess_features, preprocess_adj, sample_mask
 
 
 
diff -ruN PhaBOX-1.0.0-org/PhaTYP_single.py PhaBOX-1.0.0/PhaTYP_single.py
--- PhaBOX-1.0.0-org/PhaTYP_single.py	2024-09-30 10:20:05.500290882 +0200
+++ PhaBOX-1.0.0/PhaTYP_single.py	2024-09-30 13:27:03.755477016 +0200
@@ -16,20 +16,20 @@
 from torch import optim
 from torch.nn import functional as F
 
-from scripts.ulity import *
-from scripts.preprocessing import *
-from scripts.cnnscript import *
+from phabox.scripts.ulity import *
+from phabox.scripts.preprocessing import *
+from phabox.scripts.cnnscript import *
 from shutil import which
 from collections import Counter
 from Bio import SeqIO
 from Bio.SeqRecord import SeqRecord
 from Bio.Blast.Applications import NcbiblastnCommandline
-from models.phamer import Transformer
-from models.CAPCNN import WCNN
-from models.PhaGCN import GCN
-from models import Cherry
+from phabox.models.phamer import Transformer
+from phabox.models.CAPCNN import WCNN
+from phabox.models.PhaGCN import GCN
+from phabox.models import Cherry
 from scipy.special import softmax
-from scripts.data import load_data, preprocess_features, preprocess_adj, sample_mask
+from phabox.scripts.data import load_data, preprocess_features, preprocess_adj, sample_mask
 
 
 
diff -ruN PhaBOX-1.0.0-org/scripts/cnnscript.py PhaBOX-1.0.0/scripts/cnnscript.py
--- PhaBOX-1.0.0-org/scripts/cnnscript.py	2024-09-30 10:20:05.608293722 +0200
+++ PhaBOX-1.0.0/scripts/cnnscript.py	2024-09-30 13:25:46.367731972 +0200
@@ -3,7 +3,7 @@
 import  numpy as np
 import  torch
 import  torch.utils.data as Data
-from    models.CAPCNN import WCNN
+from    phabox.models.CAPCNN import WCNN
 from    Bio import SeqIO
 from    Bio.SeqRecord import SeqRecord   
 from    torch import nn
diff -ruN PhaBOX-1.0.0-org/scripts/data.py PhaBOX-1.0.0/scripts/data.py
--- PhaBOX-1.0.0-org/scripts/data.py	2024-09-30 10:20:05.608293722 +0200
+++ PhaBOX-1.0.0/scripts/data.py	2024-09-30 13:26:08.807655370 +0200
@@ -2,6 +2,7 @@
 import  pickle as pkl
 import  networkx as nx
 import  scipy.sparse as sp
+from scipy.sparse.linalg import eigsh
 import  sys
 
 
diff -ruN PhaBOX-1.0.0-org/setup.py PhaBOX-1.0.0/setup.py
--- PhaBOX-1.0.0-org/setup.py	2024-09-30 10:20:05.500290882 +0200
+++ PhaBOX-1.0.0/setup.py	2024-09-30 10:21:12.297733200 +0200
@@ -3,7 +3,20 @@
 setup(
     name='phabox',
     version='1.0',
-    packages=find_packages(),
+    packages=["phabox"],
+    package_dir={'phabox': '.'},
+    package_data={
+        '': ['scripts/*.py', 'models/*.py', 'GTDB/*.py']
+    },
+    entry_points={
+        'console_scripts': [
+            'phabox=phabox.main:main',  # Allows running the script with `phabox` command
+            'Cherry_single=phabox.Cherry_single:main',
+            'PhaGCN_single=phabox.PhaGCN_single:main',
+            'PhaMer_single=phabox.PhaMer_single:main',
+            'PhaTYP_single=phabox.PhaTYP_single:main',
+        ]
+    },
     install_requires=[
         'numpy==1.21.2',
         'pandas==1.3.4',
