--- TeloSearchLR.py-old	2024-10-17 23:03:55.738331016 +0400
+++ TeloSearchLR.py	2024-10-17 23:04:03.533242382 +0400
@@ -818,7 +818,7 @@
     for opt, arg in opts:
         if opt in ("-h", "--help"):
             print(arg_help)  # print the help message
-            sys.exit(2)
+            sys.exit(0)
         elif opt in ("-v", "--version"):
             print(version_number)
             sys.exit(2)
