diff -u a/Makefile b/Makefile
--- a/Makefile	2016-09-20 00:24:55.885441679 +0200
+++ b/Makefile	2016-09-20 00:35:54.414954373 +0200
@@ -11,7 +11,7 @@
 ###############################################################
 
 
-CXXFLAGS        = -O4 -Wall -g -I$(VIENNA)/include/ViennaRNA 
+CXXFLAGS        = -O4 -Wall -g -fopenmp -I$(VIENNA)/include/ViennaRNA -I$(VIENNA)/include 
 LDFLAGS         = -L$(VIENNA)/lib -lRNA 
 CXX     	= g++
 
