diff --git a/CMakeLists.txt b/CMakeLists.txt
index 373e42a..e3848ed 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -73,13 +73,9 @@ target_link_libraries(wfmash
 
 install(TARGETS wfmash DESTINATION bin)
 
-install(TARGETS wfa2cpp
-        LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
-        PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
+install(TARGETS wfa2cpp)
 
-install(TARGETS wfa2
-        LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
-        PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
+install(TARGETS wfa2)
 
 
 # version stuff