diff --git a/CMakeLists.txt b/CMakeLists.txt
index 373e42a..6bfaf58 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -73,14 +73,6 @@ target_link_libraries(wfmash
 
 install(TARGETS wfmash DESTINATION bin)
 
-install(TARGETS wfa2cpp
-        LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
-        PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
-
-install(TARGETS wfa2
-        LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
-        PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
-
 
 # version stuff
 file(MAKE_DIRECTORY ${CMAKE_SOURCE_DIR}/include)