--- scripts/CARNAC_to_fasta.py	2019-02-01 05:47:49.000000000 -0600
+++ scripts/CARNAC_to_fasta.py.new	2019-07-27 00:38:27.793424131 -0500
@@ -1,5 +1,5 @@
+#!/opt/anaconda1anaconda2anaconda3/bin/python
 
-#/local/python/3.3.2/bin/python
 import sys
 import os
 import shlex, subprocess
--- scripts/paf_to_CARNAC.py	2019-02-01 05:47:49.000000000 -0600
+++ scripts/paf_to_CARNAC.py.new	2019-07-27 00:43:49.937519650 -0500
@@ -1,4 +1,5 @@
-#!/usr/bin/env python3
+#!/opt/anaconda1anaconda2anaconda3/bin/python
+
 import sys
 import os
 
