diff --git a/CMakeLists.txt b/CMakeLists.txt
index d8aeb45..792851c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -108,6 +108,10 @@ SET(ABPOA_INCLUDE_DIR ${ABPOA_SOURCE_DIR}/include)
 # set_target_properties(ABPOA PROPERTIES IMPORTED_LOCATION ${ABPOA_SOURCE_DIR}/lib/libabpoa.a)
 
 # ksw2
+SET(KSW2_MAKE make)
+if(DEFINED CONDAPREFIX)
+  SET(KSW2_MAKE make "CC=${CMAKE_C_COMPILER}" "CFLAGS=-I${CONDAPREFIX}/include" "LDFLAGS=-L${CONDAPREFIX}/lib")
+endif ()
 message(STATUS "ksw2 will be built from source")
 ExternalProject_Add(ksw2
     GIT_REPOSITORY https://github.com/lh3/ksw2.git
@@ -115,7 +119,7 @@ ExternalProject_Add(ksw2
     BUILD_IN_SOURCE 1
     UPDATE_COMMAND ""
     CONFIGURE_COMMAND ""
-    BUILD_COMMAND make
+    BUILD_COMMAND ${KSW2_MAKE}
     INSTALL_COMMAND ""
 )
 ExternalProject_Get_Property(ksw2 SOURCE_DIR)
