--- libhmsbeagle/JNI/Makefile.am	2014-02-10 07:27:03.000000000 -0600
+++ libhmsbeagle/JNI/Makefile.am.new	2017-03-10 10:21:24.000000000 -0600
@@ -3,7 +3,7 @@
 
 libhmsbeagle_jni_la_SOURCES = beagle_BeagleJNIWrapper.h beagle_BeagleJNIWrapper.cpp
 libhmsbeagle_jni_la_LDFLAGS = -module -avoid-version $(JNI_EXTRA_LDFLAGS)
-libhmsbeagle_jni_la_LIBADD = -L$(top_builddir)/libhmsbeagle -lhmsbeagle
+libhmsbeagle_jni_la_LIBADD = -L$(top_builddir)/libhmsbeagle -lhmsbeagle -L$(PREFIX)/lib -lltdl
 libhmsbeagle_jni_la_CXXFLAGS = $(AM_CXXFLAGS)
 endif
 
