--- m4/ac_jni_include_dir.m4	2014-02-10 07:27:03.000000000 -0600
+++ m4/ac_jni_include_dir.m4.new	2017-03-03 15:51:54.000000000 -0600
@@ -58,7 +58,7 @@
 _JTOPDIR=`echo "$_ACJNI_FOLLOWED" | sed -e 's://*:/:g' -e 's:/[[^/]]*$::'`
 case "$host_os" in
         darwin*)        _JTOPDIR=`echo "$_JTOPDIR" | sed -e 's:/[[^/]]*$::'`
-                        _JINC="$_JTOPDIR/Headers";;
+                        _JINC="$_JTOPDIR/include";;
         *)              _JINC="$_JTOPDIR/include";;
 esac
 if test -f "$_JINC/jni.h"; then
@@ -83,6 +83,7 @@
 solaris*)       _JNI_INC_SUBDIRS="solaris";;
 mingw*)			_JNI_INC_SUBDIRS="win32";;
 cygwin*)		_JNI_INC_SUBDIRS="win32";;
+darwin*)                _JNI_INC_SUBDIRS="darwin";;
 *)              _JNI_INC_SUBDIRS="genunix";;
 esac
 
