diff --git a/src/Makefile b/src/Makefile
index 4be7a16..20a616d 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -1,4 +1,4 @@
-CXXFLAGS := -c -g -O3 -pedantic -std=c++11
+CXXFLAGS := -g -O3 -pedantic -std=c++14 -march=x86-64-v3 -L$(PREFIX)/lib
 
 UNAME_S := $(shell uname -s)
 ifeq ($(UNAME_S),Linux)
@@ -13,17 +13,16 @@ ifeq ($(IS_CLANG),clang++)
 	LDFLAGS += -lomp
 else
 	CXXFLAGS += -fopenmp
-	LDFLAGS += -lgomp
+	LDFLAGS += -lomp
 endif
 
 all: kmer2read_distr
 
 kmer2read_distr: kmer2read_distr.o ctime.o taxonomy.o kraken_processing.o
-	$(CXX) -o $@ $^ $(LDFLAGS)
+	$(CXX)  -o $@ $^ $(LDFLAGS)
 
 clean:
 	rm -f *.o
 
 %.o: %.cpp
 	$(CXX) $(CXXFLAGS) $<
-
