diff --git a/src/Makefile b/src/Makefile
index 8622014..cb21be9 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -4,7 +4,7 @@ CFLAGS = -c -g -O3 -pedantic -std=c++11 -fopenmp
 all: kmer_counter
 
 kmer_counter: 	kmer2read_distr.o ctime.o taxonomy.o kraken_processing.o
-		$(C++) -o kmer2read_distr ctime.o kmer2read_distr.o taxonomy.o kraken_processing.o -lgomp
+		$(C++) -o kmer2read_distr ctime.o kmer2read_distr.o taxonomy.o kraken_processing.o -lomp
 
 clean:
 		rm -f *.o
