--- src/Makefile	2019-08-08 05:03:07.000000000 +0200
+++ src/Makefile2	2019-08-09 09:39:07.007473559 +0200
@@ -4,7 +4,7 @@
 all: kmer_counter
 
 kmer_counter: 	kmer2read_distr.o time.o taxonomy.o kraken_processing.o
-		$(C++) -o kmer2read_distr time.o kmer2read_distr.o taxonomy.o kraken_processing.o -lgomp
+		$(C++) -o kmer2read_distr time.o kmer2read_distr.o taxonomy.o kraken_processing.o -lomp
 
 clean:
 		rm -f *.o
