--- gtdbtk/main.py	2019-07-12 17:46:04.000000000 +1000
+++ gtdbtk/main_new.py	2020-03-26 14:37:03.725568674 +1000
@@ -528,6 +528,7 @@
         options : argparse.Namespace
             The CLI arguments input by the user.
         """
+        self.logger.warning('You are not running the most recent version of GTDB-Tk!')

         if options.subparser_name == 'de_novo_wf':
             check_dependencies(['prodigal', 'hmmalign'])
