--- Makefile	2016-07-12 12:02:59.002916140 +0400
+++ Makefile	2016-07-12 12:03:18.846994692 +0400
@@ -7,7 +7,7 @@
 mandir = 
 
 CC = gcc
-CCOPT = -O3 -minline-all-stringops -funsafe-loop-optimizations -fprefetch-loop-arrays -funroll-loops -fomit-frame-pointer -fno-bounds-check -Wall -W #-pg -g -D_DEBUG -O0 -m32 -march=i386 -mtune=i386
+CCOPT = -O3 -minline-all-stringops  -fprefetch-loop-arrays -funroll-loops -fomit-frame-pointer -fno-bounds-check -Wall -W #-pg -g -D_DEBUG -O0 -m32 -march=i386 -mtune=i386
 #CCOPT = -O0 -g -D_DEBUG -Wall
 LIBS = -lm -lpthread
 INCDIRS = -Iinc/
