commit 07932751e776dd71b224dadc94aea0922d08e653
Author: Milot Mirdita <milot@mirdita.de>
Date:   Mon Jan 20 19:42:31 2025 +0900

    Reduce ProstT5 memory requirements

diff --git a/src/strucclustutils/ProstT5.cpp b/src/strucclustutils/ProstT5.cpp
index 605ce300..28bc6c4b 100644
--- a/src/strucclustutils/ProstT5.cpp
+++ b/src/strucclustutils/ProstT5.cpp
@@ -168,9 +168,9 @@ ProstT5::ProstT5(ProstT5Model& model, int threads) : model(model) {
     auto cparams = llama_context_default_params();
     cparams.n_threads = threads;
     cparams.n_threads_batch = threads;
-    cparams.n_ubatch = 4096;
-    cparams.n_batch = 4096;
-    cparams.n_ctx = 4096;
+    cparams.n_ubatch = 2048;
+    cparams.n_batch = 2048;
+    cparams.n_ctx = 2048;
     cparams.embeddings = true;
     cparams.attention_type = LLAMA_ATTENTION_TYPE_NON_CAUSAL;
 
