diff -u a/targetfinder.pl b/targetfinder.pl
--- a/targetfinder.pl	2015-02-19 17:56:53.000000000 +0100
+++ b/targetfinder.pl	2016-11-25 14:26:22.951456869 +0100
@@ -738,7 +738,7 @@
 	print STDERR "         -r           Search reverse strand for targets?. Use this option if the database is genomic DNA.\n";
 	print STDERR "         -h           Print this menu\n";
 	print STDERR "\n\n";
-	exit 1;
+	exit 0;
 }
 
 ################################################################################
