diff --git i/FindGUROBI.cmake w/FindGUROBI.cmake
index adfe9df..f121ec7 100644
--- i/FindGUROBI.cmake
+++ w/FindGUROBI.cmake
@@ -22,7 +22,7 @@ if(EXISTS ${GUROBI_HOME} )
 
     message( "-- The retrieved name of version-specific library is " gurobi ${GUROBI_VER_LIB} )
 
-    file( GLOB GUROBI_LIB_FILE ${GUROBI_HOME}/linux64/lib/libgurobi${GUROBI_VER_LIB}.* )
+    file( GLOB GUROBI_LIB_FILE ${GUROBI_HOME}/*linux64/lib/libgurobi${GUROBI_VER_LIB}.* )
     if( GUROBI_LIB_FILE )
     	message( "-- Gurobi library " ${GUROBI_LIB_FILE} " found in " ${GUROBI_HOME} "/linux64/lib/" )
     else()
