diff --git a/inst/SBML_odeSolver-1.7.0beta/configure.in b/inst/SBML_odeSolver-1.7.0beta/configure.in
index 6cf8dba..1424902 100644
--- a/inst/SBML_odeSolver-1.7.0beta/configure.in
+++ b/inst/SBML_odeSolver-1.7.0beta/configure.in
@@ -66,7 +66,6 @@ dnl ---------------------------------------------------------------------------
 
 AC_PROG_CXX
 AC_PROG_CC
-AC_PROG_CC_FPIC
 AC_SUBST(FPIC)
 AC_ARG_VAR(FPIC,[compiler flag for generating PIC code that can be used in a shared library])
 AC_PROG_CPP
