--- a/scripts/braker.pl
+++ b/scripts/braker.pl
@@ -3171,16 +3171,6 @@
     }
     $cmdString = "java -version 2>&1 | awk -F[\\\"\\\.] -v OFS=. 'NR==1{print \$2,\$3}'";
     my @javav = `$cmdString` or die("Failed to execute: $cmdString");
-    if(not ($javav[0] =~ m/1\.8/ )){
-        $prtStr = "\# " . (localtime) . " ERROR: in file " . __FILE__
-            ." at line ". __LINE__ ."\n"
-            . "You have installed java version $javav[0]. GUSHR requires version 1.8!\n"
-            . "You can switch between java versions on your system with:\n"
-            . "sudo update-alternatives --config java\n";
-        $logString .= $prtStr;
-        print STDERR $logString;
-        exit(1);
-    }
 
 }
 
