--- scripts/align2hints.pl	2018-09-10 08:36:32.000000000 -0500
+++ scripts/align2hints.pl.new	2019-04-13 00:45:20.199730500 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ####################################################################################################
 #                                                                                                  #
--- scripts/braker.pl	2018-09-10 08:36:32.000000000 -0500
+++ scripts/braker.pl.new	2019-04-13 01:02:02.299535048 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ####################################################################################################
 #                                                                                                  #
@@ -2626,7 +2626,7 @@
     }
 
     #    check whether bam2wig is executable
-    $bam2wig = "$AUGUSTUS_BIN_PATH/../auxprogs/bam2wig/bam2wig";
+    $bam2wig = "$AUGUSTUS_BIN_PATH/bam2wig";
     if ( $UTR eq "on" && $skipAllTraining == 0 )
     {    # MIGHT WANT TO CHANGE THIS!
         if ( not( -x $bam2wig ) ) {
--- scripts/downsample_traingenes.pl	2018-09-10 08:36:32.000000000 -0500
+++ scripts/downsample_traingenes.pl.new	2019-04-13 00:55:29.606607045 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 # Author: Katharina J. Hoff and Mario Stanke
 # Contact: katharina.hoff@uni-greifswald.de
--- scripts/filterGenemark.pl	2018-09-10 08:36:32.000000000 -0500
+++ scripts/filterGenemark.pl.new	2019-04-13 00:56:53.596590032 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ####################################################################################################
 #                                                                                                  #
--- scripts/filterIntronsFindStrand.pl	2018-09-10 08:36:32.000000000 -0500
+++ scripts/filterIntronsFindStrand.pl.new	2019-04-13 00:59:33.829557574 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ####################################################################################################
 #                                                                                                  #
--- scripts/findGenesInIntrons.pl	2018-09-10 08:36:32.000000000 -0500
+++ scripts/findGenesInIntrons.pl.new	2019-04-13 00:58:55.047565429 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 # ==============================================================
 # Katharina J. Hoff
 # When two gene sets are merged with joingenes, genes that
--- scripts/startAlign.pl	2018-09-10 08:36:32.000000000 -0500
+++ scripts/startAlign.pl.new	2019-04-13 00:58:15.723573394 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ####################################################################################################
 #                                                                                                  #
