diff --git a/scripts/align2hints.pl b/scripts/align2hints.pl
index aeacb40..b64ae25 100755
--- a/scripts/align2hints.pl
+++ b/scripts/align2hints.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ####################################################################################################
 #                                                                                                  #
diff --git a/scripts/braker.pl b/scripts/braker.pl
index bf8ed55..cacf37b 100755
--- a/scripts/braker.pl
+++ b/scripts/braker.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ####################################################################################################
 #                                                                                                  #
diff --git a/scripts/braker_cleanup.pl b/scripts/braker_cleanup.pl
index 0cea977..68df4c7 100755
--- a/scripts/braker_cleanup.pl
+++ b/scripts/braker_cleanup.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ####################################################################################################
 #                                                                                                  #
diff --git a/scripts/downsample_traingenes.pl b/scripts/downsample_traingenes.pl
index 13292ee..14897e2 100755
--- a/scripts/downsample_traingenes.pl
+++ b/scripts/downsample_traingenes.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 # Author: Katharina J. Hoff and Mario Stanke
 # Contact: katharina.hoff@uni-greifswald.de
diff --git a/scripts/filterGenemark.pl b/scripts/filterGenemark.pl
index 9f75b44..5b8e464 100755
--- a/scripts/filterGenemark.pl
+++ b/scripts/filterGenemark.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ####################################################################################################
 #                                                                                                  #
diff --git a/scripts/filterIntronsFindStrand.pl b/scripts/filterIntronsFindStrand.pl
index d489dec..be55264 100755
--- a/scripts/filterIntronsFindStrand.pl
+++ b/scripts/filterIntronsFindStrand.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ####################################################################################################
 #                                                                                                  #
diff --git a/scripts/filter_augustus_gff.pl b/scripts/filter_augustus_gff.pl
index c06e766..414320d 100755
--- a/scripts/filter_augustus_gff.pl
+++ b/scripts/filter_augustus_gff.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ####################################################################################################
 #                                                                                                  #
@@ -141,4 +141,4 @@ if(defined($out_file)){
 }else{
 	my $n_tx = keys %keep;
 	print "$n_tx\n";
-}
\ No newline at end of file
+}
diff --git a/scripts/findGenesInIntrons.pl b/scripts/findGenesInIntrons.pl
index 0cf4476..450f0a4 100755
--- a/scripts/findGenesInIntrons.pl
+++ b/scripts/findGenesInIntrons.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 # ==============================================================
 # Katharina J. Hoff
 # When two gene sets are merged with joingenes, genes that
diff --git a/scripts/gatech_pmp2hints.pl b/scripts/gatech_pmp2hints.pl
index 22fc880..a00edd5 100755
--- a/scripts/gatech_pmp2hints.pl
+++ b/scripts/gatech_pmp2hints.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ####################################################################################################
 #                                                                                                  #
@@ -22,4 +22,4 @@ use warnings;
 while(<STDIN>){
 	my @t = split(/\t/);
 	print "$t[0]\t$t[1]\tintron\t$t[3]\t$t[4]\t$t[5]\t$t[6]\t$t[7]\tsrc=P;mult=$t[5];\n";
-}
\ No newline at end of file
+}
diff --git a/scripts/startAlign.pl b/scripts/startAlign.pl
index df15a5a..dc62e41 100755
--- a/scripts/startAlign.pl
+++ b/scripts/startAlign.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ####################################################################################################
 #                                                                                                  #
