--- scripts/align2hints.pl	2018-09-10 08:36:32.000000000 -0500
+++ scripts/align2hints.pl.new	2019-03-29 23:56:43.905797287 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ####################################################################################################
 #                                                                                                  #
--- scripts/downsample_traingenes.pl	2018-09-10 08:36:32.000000000 -0500
+++ scripts/downsample_traingenes.pl.new	2019-03-29 23:57:37.947801370 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 # Author: Katharina J. Hoff and Mario Stanke
 # Contact: katharina.hoff@uni-greifswald.de
--- scripts/findGenesInIntrons.pl	2018-09-10 08:36:32.000000000 -0500
+++ scripts/findGenesInIntrons.pl.new	2019-03-29 23:58:16.815804305 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 # ==============================================================
 # Katharina J. Hoff
 # When two gene sets are merged with joingenes, genes that
