--- scripts/align2hints.pl	2020-02-20 12:05:38.000000000 -0600
+++ scripts/align2hints.pl.new	2020-02-21 01:12:00.016871286 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ####################################################################################################
 #                                                                                                  #
--- scripts/braker_cleanup.pl	2020-02-20 12:05:38.000000000 -0600
+++ scripts/braker_cleanup.pl.new	2020-02-21 01:12:09.345900490 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ####################################################################################################
 #                                                                                                  #
--- scripts/braker.pl	2020-02-20 12:05:38.000000000 -0600
+++ scripts/braker.pl.new	2020-02-21 01:12:19.010930749 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ####################################################################################################
 #                                                                                                  #
--- scripts/downsample_traingenes.pl	2020-02-20 12:05:38.000000000 -0600
+++ scripts/downsample_traingenes.pl.new	2020-02-21 01:12:27.649957796 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 # Author: Katharina J. Hoff and Mario Stanke
 # Contact: katharina.hoff@uni-greifswald.de
--- scripts/filter_augustus_gff.pl	2020-02-20 12:05:38.000000000 -0600
+++ scripts/filter_augustus_gff.pl.new	2020-02-21 01:12:41.149000064 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ####################################################################################################
 #                                                                                                  #
@@ -141,4 +141,4 @@
 }else{
 	my $n_tx = keys %keep;
 	print "$n_tx\n";
-}
\ No newline at end of file
+}
--- scripts/filterGenemark.pl	2020-02-20 12:05:38.000000000 -0600
+++ scripts/filterGenemark.pl.new	2020-02-21 01:12:53.230037887 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ####################################################################################################
 #                                                                                                  #
--- scripts/filterIntronsFindStrand.pl	2020-02-20 12:05:38.000000000 -0600
+++ scripts/filterIntronsFindStrand.pl.new	2020-02-21 01:13:05.924077627 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ####################################################################################################
 #                                                                                                  #
--- scripts/findGenesInIntrons.pl	2020-02-20 12:05:38.000000000 -0600
+++ scripts/findGenesInIntrons.pl.new	2020-02-21 01:13:17.087112574 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 # ==============================================================
 # Katharina J. Hoff
 # When two gene sets are merged with joingenes, genes that
--- scripts/gatech_pmp2hints.pl	2020-02-20 12:05:38.000000000 -0600
+++ scripts/gatech_pmp2hints.pl.new	2020-02-21 01:13:30.435154370 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ####################################################################################################
 #                                                                                                  #
@@ -22,4 +22,4 @@
 while(<STDIN>){
 	my @t = split(/\t/);
 	print "$t[0]\t$t[1]\tintron\t$t[3]\t$t[4]\t$t[5]\t$t[6]\t$t[7]\tsrc=P;mult=$t[5];\n";
-}
\ No newline at end of file
+}
--- scripts/log_reg_prothints.pl	2020-02-20 12:05:38.000000000 -0600
+++ scripts/log_reg_prothints.pl.new	2020-02-21 01:13:41.295188367 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 # Katharina J. Hoff
 # Parse ProtHint prothint.gff (file for GeneMark-EX) to produce two classes of hints for AUGUSTUS
@@ -93,4 +93,4 @@
 }
 
 close(OUT) or die("ERROR: in file " . __FILE__ ." at line "
-    . __LINE__ ."\nFailed to close file $out_file!\n");
\ No newline at end of file
+    . __LINE__ ."\nFailed to close file $out_file!\n");
--- scripts/merge_transcript_sets.pl	2020-02-20 12:05:38.000000000 -0600
+++ scripts/merge_transcript_sets.pl.new	2020-02-21 01:13:51.698220936 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ####################################################################################################
 #                                                                                                  #
@@ -109,4 +109,4 @@
     foreach(@{$txid_to_elements{$value}}){
         print $_;
     }    
-}
\ No newline at end of file
+}
--- scripts/startAlign.pl	2020-02-20 12:05:38.000000000 -0600
+++ scripts/startAlign.pl.new	2020-02-21 01:14:01.115250415 -0600
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 
 ####################################################################################################
 #                                                                                                  #
