--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,4 +1,4 @@
 [build-system]
 requires = ["setuptools>=45", "setuptools_scm[toml]>=6.2"]
-build-backend = "setuptools.backends.legacy:build"
+build-backend = "setuptools.build_meta"

--- a/setup.py
+++ b/setup.py
@@ -13,7 +13,7 @@
         ],
     },
     author_email = 'pierre.peterlongo@inria.fr',
-    maintainer = logan_blaster.__author__,
+    maintainer = 'Pierre Peterlongo',
     maintainer_email = 'pierre.peterlongo@inria.fr',
     download_url = 'https://github.com/pierrepeterlongo/logan_blaster',
     url = 'https://github.com/pierrepeterlongo/logan_blaster',
