--- Makefile.old	2017-09-15 14:09:55.000000000 -0500
+++ Makefile	2017-09-15 14:10:33.000000000 -0500
@@ -1,5 +1,5 @@
 VERSION = v0.4
-DEFAULT_FLAGS = -DAGE_VERSION=\"$(VERSION)\" -DAGE_TIME -static -O3
+DEFAULT_FLAGS = -DAGE_VERSION=\"$(VERSION)\" -DAGE_TIME -O3
 CXX	= g++ -fopenmp -DOMP $(DEFAULT_FLAGS)
 
 MESS = "Compiling with parallel support."
@@ -29,7 +29,7 @@
 	@echo ""
 
 age_align: $(OBJS)
-	$(CXX) -o $@ $(OBJS) -static
+	$(CXX) -o $@ $(OBJS)
 
 $(OBJDIR)/%.o: %.cpp
 	@mkdir -p $(OBJDIR)
