--- setup.py.old	2023-04-18 17:09:16.851099242 -0500
+++ setup.py	2023-04-18 17:09:39.426191218 -0500
@@ -23,7 +23,7 @@
     except Exception:
         print("\nError: gzip cannot be found. gzip is necessary for unzipping one of the files. Exiting.")
         exit(1)
-    subprocess.run(f'gunzip models/vRhyme_machine_model_ET.sav.gz 2> /dev/null', shell=True)
+    subprocess.run(f'gunzip vRhyme/models/vRhyme_machine_model_ET.sav.gz 2> /dev/null', shell=True)
 
     aux = [f'vRhyme/aux/{f}' for f in os.listdir('vRhyme/aux') if f.endswith('.py')]
     threaders = [f'vRhyme/scripts/{f}' for f in os.listdir('vRhyme/scripts') if f.endswith('.py')]
@@ -53,4 +53,4 @@
 
 
 if __name__ == "__main__":
-    do_setup(get_version(), get_descript())
\ No newline at end of file
+    do_setup(get_version(), get_descript())
