diff --git rdkit/RDConfig.py rdkit/RDConfig.py
index 9b685e2..7d022d5 100755
--- rdkit/RDConfig.py
+++ rdkit/RDConfig.py
@@ -25,11 +25,11 @@ elif 'CONDA_DEFAULT_ENV' in os.environ:
   # we are running in a conda environ.
   RDCodeDir=os.path.dirname(__file__)
   splitdir = RDCodeDir.split(os.path.sep)
-  condaDir = splitdir[:-4]
+  condaDir = splitdir[:-3]
   if condaDir[0]=='':
     condaDir[0] = os.path.sep
-  condaDir += ['share','RDKit']
-  _share = os.path.join(*condaDir)
+  condaDir += ['Library','share','RDKit']
+  _share = os.path.sep.join(condaDir)
   RDDataDir=os.path.join(_share,'Data')
   RDDocsDir=os.path.join(_share,'Docs')
   RDProjDir=os.path.join(_share,'Projects')
