diff --git rdkit/CMakeLists.txt rdkit/CMakeLists.txt
index 4ccb94f..a68ee75 100644
--- rdkit/CMakeLists.txt
+++ rdkit/CMakeLists.txt
@@ -1,8 +1,8 @@
 file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/RDPaths.py 
 "import os
-# unset so to trigger exceptions and track use: RDBaseDir=os.environ['RDBASE']
-RDCodeDir=os.path.join(r'${PYTHON_INSTDIR}','rdkit')
-# not really hard-coded alternative RDCodeDir=os.path.dirname(__file__)
+# do not set RDBaseDir, so to trigger exceptions and track use:
+# RDBaseDir=os.environ['RDBASE']
+RDCodeDir=os.path.dirname(__file__)
 _share = os.path.join(r'${CMAKE_INSTALL_PREFIX}', r'${RDKit_ShareDir}')
 RDDataDir=os.path.join(_share,'Data')
 RDDocsDir=os.path.join(_share,'Docs')
