--- lib/R/CheckDB.R.orig	2017-09-17 18:54:13.000000000 -0400
+++ lib/R/CheckDB.R	2017-11-10 16:10:31.353243513 -0500
@@ -28,34 +28,34 @@
   }
 
 ## check bwa version
-yy <- file.exists(file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version"))
-if (yy) {
-  prev.version.bwa <- scan(file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version"), what = "", quiet = T, sep = "\n")
-  system(paste("bwa", "2>&1", "|", "grep ersion", ">", file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version.tmp")))
-  curr.version.bwa <- scan(file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version.tmp"), what = "", quiet = T, sep = "\n")
-  if (curr.version.bwa != prev.version.bwa) {
-    cat("[EricScript] Updating BWA indexes for", refid,"... ")    
-    system(paste("bwa index", file.path(file.path(dbfolder, "data", refid, "EnsemblGene.Reference.fa")), "1>>", file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version.tmp"), "2>>", file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version.tmp")))    
-    cat("done.\n")
-    cat(curr.version.bwa, file = file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version"))
-  }
-} else {
-  system(paste("bwa", "2>&1", "|", "grep ersion", ">", file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version.tmp")))
-  curr.version.bwa <- scan(file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version.tmp"), what = "", quiet = T, sep = "\n")
-  version.a <- gsub("Version: ", "", strsplit(curr.version.bwa, ".", fixed = T)[[1]][1])
-  version.b <- strsplit(curr.version.bwa, ".", fixed = T)[[1]][2]
-  version.c <- gsub("[a-z]", "", strsplit(strsplit(curr.version.bwa, ".", fixed = T)[[1]][3], "-")[[1]][1])
-  version.tot <- as.numeric(paste(version.a, version.b, version.c, sep = ""))
-  if (version.tot >= 74) {
-    cat("[EricScript] Updating BWA indexes for", refid, "...")    
-    system(paste("bwa index", file.path(file.path(dbfolder, "data", refid, "EnsemblGene.Reference.fa")), "1>>", file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version.tmp"), "2>>", file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version.tmp")))    
-    cat("done.\n")
-    system(paste("bwa", "2>&1", "|", "grep ersion", ">", file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version")))
-   } else {
-    flag.dbexists <- 0
-    cat("[EricScript] BWA version >= 0.7.4 is required. Exit.\n") 
-  }
-}
+# yy <- file.exists(file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version"))
+# if (yy) {
+#   prev.version.bwa <- scan(file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version"), what = "", quiet = T, sep = "\n")
+#   system(paste("bwa", "2>&1", "|", "grep ersion", ">", file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version.tmp")))
+#   curr.version.bwa <- scan(file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version.tmp"), what = "", quiet = T, sep = "\n")
+#   if (curr.version.bwa != prev.version.bwa) {
+#     cat("[EricScript] Updating BWA indexes for", refid,"... ")    
+#     system(paste("bwa index", file.path(file.path(dbfolder, "data", refid, "EnsemblGene.Reference.fa")), "1>>", file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version.tmp"), "2>>", file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version.tmp")))    
+#     cat("done.\n")
+#     cat(curr.version.bwa, file = file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version"))
+#   }
+# } else {
+#   system(paste("bwa", "2>&1", "|", "grep ersion", ">", file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version.tmp")))
+#   curr.version.bwa <- scan(file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version.tmp"), what = "", quiet = T, sep = "\n")
+#   version.a <- gsub("Version: ", "", strsplit(curr.version.bwa, ".", fixed = T)[[1]][1])
+#   version.b <- strsplit(curr.version.bwa, ".", fixed = T)[[1]][2]
+#   version.c <- gsub("[a-z]", "", strsplit(strsplit(curr.version.bwa, ".", fixed = T)[[1]][3], "-")[[1]][1])
+#   version.tot <- as.numeric(paste(version.a, version.b, version.c, sep = ""))
+#   if (version.tot >= 74) {
+#     cat("[EricScript] Updating BWA indexes for", refid, "...")    
+#     system(paste("bwa index", file.path(file.path(dbfolder, "data", refid, "EnsemblGene.Reference.fa")), "1>>", file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version.tmp"), "2>>", file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version.tmp")))    
+#     cat("done.\n")
+#     system(paste("bwa", "2>&1", "|", "grep ersion", ">", file.path(ericscriptfolder, "lib", "data", "_resources", ".bwa.version")))
+#    } else {
+#     flag.dbexists <- 0
+#     cat("[EricScript] BWA version >= 0.7.4 is required. Exit.\n") 
+#   }
+# }
 
 mydbdata.bwa <- c("EnsemblGene.Reference.fa.bwt", "EnsemblGene.Reference.fa.pac", "EnsemblGene.Reference.fa.ann", "EnsemblGene.Reference.fa.amb", "EnsemblGene.Reference.fa.sa")
 
