diff --git a/Makefile b/Makefile
index 8eb055f..36eb3ea 100644
--- a/Makefile
+++ b/Makefile
@@ -20,21 +20,21 @@
 # GSL_INCLUDE_DIR - location of GSL include files
 # MOTHUR_FILES - The MOTHUR_FILES parameter is optional, but allows you to set a default location for mothur to look for input files it can't find. This is often used for reference files you want to store in one location separate from your data.
 
-INSTALL_DIR ?= "\"Enter_your_mothur_install_path_here\""
+INSTALL_DIR ?= "$(PREFIX)/bin"
 
 OPTIMIZE ?= yes
 USEREADLINE ?= yes
-USEBOOST ?= no
-USEHDF5 ?= no
-USEGSL ?= no
+USEBOOST ?= yes
+USEHDF5 ?= yes
+USEGSL ?= yes
 LOGFILE_NAME ?= no
 
-BOOST_LIBRARY_DIR ?= "\"Enter_your_boost_library_path_here\""
-BOOST_INCLUDE_DIR ?= "\"Enter_your_boost_include_path_here\""
-HDF5_LIBRARY_DIR ?= "\"Enter_your_HDF5_library_path_here\""
-HDF5_INCLUDE_DIR ?= "\"Enter_your_HDF5_include_path_here\""
-GSL_LIBRARY_DIR ?= "\"Enter_your_GSL_library_path_here\""
-GSL_INCLUDE_DIR ?= "\"Enter_your_GSL_include_path_here\""
+BOOST_LIBRARY_DIR ?= "$(PREFIX)/lib"
+BOOST_INCLUDE_DIR ?= "$(PREFIX)/include/boost"
+HDF5_LIBRARY_DIR ?= "$(PREFIX)/lib"
+HDF5_INCLUDE_DIR ?= "$(PREFIX)/include/hdf5"
+GSL_LIBRARY_DIR ?= "$(PREFIX)/lib"
+GSL_INCLUDE_DIR ?= "$(PREFIX)/include/gsl"
 MOTHUR_FILES="\"Enter_your_default_path_here\""
 MOTHUR_TOOLS="\"Enter_your_mothur_tools_path_here\""
 VERSION = "\"1.48.3\""
@@ -50,8 +50,8 @@ ifeq  ($(strip $(OPTIMIZE)),yes)
     CXXFLAGS += -O3
 endif
 
-CXXFLAGS += -std=c++11 -pthread -DVERSION=${VERSION} -DRELEASE_DATE=${RELEASE_DATE}
-LDFLAGS += -std=c++11 -pthread
+CXXFLAGS += -std=c++14 -pthread -DVERSION=${VERSION} -DRELEASE_DATE=${RELEASE_DATE} -I$(PREFIX)/include
+LDFLAGS += -std=c++14 -pthread -L$(PREFIX)/lib
 
 ifeq  ($(strip $(MOTHUR_FILES)),"\"Enter_your_default_path_here\"")
 else
@@ -109,8 +109,8 @@ endif
     subdirs :=  $(sort $(dir $(filter-out  $(skipUchime), $(wildcard source/*/))))
     subDirIncludes = $(patsubst %, -I %, $(subdirs))
     subDirLinking =  $(patsubst %, -L%, $(subdirs))
-    CXXFLAGS += -I. $(subDirIncludes)
-    LDFLAGS += $(subDirLinking)
+    CXXFLAGS += -I. -Isource $(subDirIncludes)
+    LDFLAGS += -Lsource $(subDirLinking)
 
 
 #
@@ -118,6 +118,8 @@ endif
 #
     OBJECTS=$(patsubst %.cpp,%.o,$(wildcard $(addsuffix *.cpp,$(subdirs))))
     OBJECTS+=$(patsubst %.c,%.o,$(wildcard $(addsuffix *.c,$(subdirs))))
+    OBJECTS+=$(patsubst %.cpp,%.o,$(wildcard source/*.cpp))
+    OBJECTS+=$(patsubst %.cpp,%.o,$(wildcard source/*.c))
     OBJECTS+=$(patsubst %.cpp,%.o,$(wildcard *.cpp))
     OBJECTS+=$(patsubst %.c,%.o,$(wildcard *.c))
 
