diff --git a/Makefile b/Makefile
index 11a777db..993e7d8c 100644
--- a/Makefile
+++ b/Makefile
@@ -108,13 +108,13 @@ endif
   subdirs :=  $(sort $(dir $(filter-out  $(skipUchime), $(wildcard source/*/)))) source/
   subDirIncludes = $(patsubst %, -I %, $(subdirs))
 	subDirLinking =  $(patsubst %, -L%, $(subdirs))
-	CXXFLAGS += -I. -Isource $(subDirIncludes)
-	LDFLAGS += -Lsource $(subDirLinking)
+	CXXFLAGS += -I. -Isource/ $(subDirIncludes)
+	LDFLAGS += -Lsource/ $(subDirLinking)
 
-		OBJECTS=$(patsubst %.cpp,%.o,$(wildcard $(addsuffix *.cpp,$(subdirs))))
-    OBJECTS+=$(patsubst %.c,%.o,$(wildcard $(addsuffix *.c,$(subdirs))))
-    OBJECTS+=$(patsubst %.cpp,%.o,$(wildcard *.cpp))
-    OBJECTS+=$(patsubst %.c,%.o,$(wildcard *.c))
+	OBJECTS=$(patsubst %.cpp,%.o,$(wildcard $(addsuffix *.cpp,$(subdirs))))
+	OBJECTS+=$(patsubst %.c,%.o,$(wildcard $(addsuffix *.c,$(subdirs))))
+	OBJECTS+=$(patsubst %.cpp,%.o,$(wildcard *.cpp))
+	OBJECTS+=$(patsubst %.c,%.o,$(wildcard *.c))
 
 mothur : $(OBJECTS) uchime
 	$(CXX) $(LDFLAGS) $(TARGET_ARCH) -o $@ $(OBJECTS) $(LIBS)
