diff --git a/source/uchime_src/makefile b/source/uchime_src/makefile
index 428babe..615936e 100644
--- a/source/uchime_src/makefile
+++ b/source/uchime_src/makefile
@@ -1,5 +1,5 @@
-CXXFLAGS = -O3 -D_FILE_OFFSET_BITS=64 -DNDEBUG=1 -DUCHIMES=1
-LDFLAGS = -g
+CXXFLAGS = -O3 -std=c++14 -D_FILE_OFFSET_BITS=64 -DNDEBUG=1 -DUCHIMES=1 -I$(PREFIX)/include
+LDFLAGS = -L$(PREFIX)/lib -g
 
 #
 # Get the list of all .cpp files, rename to .o files
@@ -14,7 +14,6 @@ uchime : $(OBJECTS)
 
 install : uchime
 
-	
 %.o : %.c %.h
 	$(COMPILE.c) $(OUTPUT_OPTION) $<
 %.o : %.cpp %.h
@@ -26,4 +25,4 @@ install : uchime
 
 clean :
 	@rm -f $(OBJECTS)
-	
\ No newline at end of file
+	
