diff --git a/Makefile b/Makefile
index d20d02c..a805ca0 100644
--- a/Makefile
+++ b/Makefile
@@ -109,8 +109,8 @@ endif
     subdirs :=  $(sort $(dir $(filter-out  $(skipUchime), $(wildcard source/*/))))
     subDirIncludes = $(patsubst %, -I %, $(subdirs))
     subDirLinking =  $(patsubst %, -L%, $(subdirs))
-    CXXFLAGS += -I. -Isource $(subDirIncludes)
-    LDFLAGS += -Lsource $(subDirLinking)
+    CXXFLAGS += -I. -Isource/ $(subDirIncludes)
+    LDFLAGS += -Lsource/ $(subDirLinking)
 
 
 #
@@ -118,6 +118,8 @@ endif
 #
     OBJECTS=$(patsubst %.cpp,%.o,$(wildcard $(addsuffix *.cpp,$(subdirs))))
     OBJECTS+=$(patsubst %.c,%.o,$(wildcard $(addsuffix *.c,$(subdirs))))
+    OBJECTS+=$(patsubst %.cpp,%.o,$(wildcard source/*.cpp))
+    OBJECTS+=$(patsubst %.cpp,%.o,$(wildcard source/*.c))
     OBJECTS+=$(patsubst %.cpp,%.o,$(wildcard *.cpp))
     OBJECTS+=$(patsubst %.c,%.o,$(wildcard *.c))
 
