diff --git a/Makefile b/Makefile
index 073a7bb..0e415e7 100644
--- a/Makefile
+++ b/Makefile
@@ -19,8 +19,8 @@
 OPTIMIZE ?= yes
 USEREADLINE ?= yes
 USEBOOST ?= yes
-BOOST_LIBRARY_DIR="\"Enter_your_boost_library_path_here\""
-BOOST_INCLUDE_DIR="\"Enter_your_boost_include_path_here\""
+BOOST_LIBRARY_DIR=${PREFIX}/lib
+BOOST_INCLUDE_DIR=${PREFIX}/include
 MOTHUR_FILES="\"Enter_your_default_path_here\""
 RELEASE_DATE = "\"3/20/2017\""
 VERSION = "\"1.39.5\""
@@ -49,7 +49,7 @@ endif
 
 ifeq  ($(strip $(USEREADLINE)),yes)
     CXXFLAGS += -DUSE_READLINE
-    LIBS += -lreadline
+    LIBS += -lreadline -lncurses
 endif
 
 
diff --git a/makefile-internal b/makefile-internal
index 8debe9a..e1e1593 100644
--- a/makefile-internal
+++ b/makefile-internal
@@ -18,7 +18,7 @@ ifeq  ($(strip $(64BIT_VERSION)),yes)
     TARGET_ARCH += -m64 -static
 
     #if you are a linux user use the following line - specific processor builds in future??
-    #CXXFLAGS += -mtune=generic
+    CXXFLAGS += -mtune=core2
 
     CXXFLAGS += -DBIT_VERSION
 endif
@@ -37,15 +37,17 @@ endif
 #The boost libraries allow you to read gz files.
 ifeq  ($(strip $(USEBOOST)),yes)
     #statically link so the boost install is not required on users machine
-    BOOST_INCLUDE_DIR="/usr/local/include"
-    BOOST_LIBRARY_DIR="/usr/local/lib"
+    BOOST_INCLUDE_DIR=${PREFIX}/include
+    BOOST_LIBRARY_DIR=${PREFIX}/lib
+    
 
-    CXXFLAGS += -DUSE_BOOST -I ${BOOST_INCLUDE_DIR}
+    CXXFLAGS += -DUSE_BOOST -I${BOOST_INCLUDE_DIR} -L${BOOST_LIBRARY_DIR}
+    LDFLAGS += -L${BOOST_LIBRARY_DIR}
 
-    LIBS += ${BOOST_LIBRARY_DIR}/libboost_iostreams.a
+    LIBS += -lboost_iostreams -lz
 
     #if mac
-    LIBS += -lz
+    #LIBS += -lz
 
     #if linux or windows
     #LIBS += ${BOOST_LIBRARY_DIR}/libz.a
