--- makefile-internal	2016-08-09 16:44:08.000000000 +0000
+++ makefile-internal.new	2016-11-18 23:28:01.909986965 +0000
@@ -10,7 +10,7 @@
 
 ifeq  ($(strip $(64BIT_VERSION)),yes)
     #if you are a mac user use the following line
-    TARGET_ARCH += -arch x86_64
+    #TARGET_ARCH += -arch x86_64
 
     #if you using cygwin to build Windows the following line
     #CXX = x86_64-w64-mingw32-g++
@@ -18,7 +18,7 @@
     #TARGET_ARCH += -m64 -static
 
     #if you are a linux user use the following line - specific processor builds in future??
-    #CXXFLAGS += -mtune=generic
+    CXXFLAGS += -mtune=core2
 
     CXXFLAGS += -DBIT_VERSION
 endif
@@ -37,15 +37,16 @@
 #The boost libraries allow you to read gz files.
 ifeq  ($(strip $(USEBOOST)),yes)
     #statically link so the boost install is not required on users machine
-    BOOST_INCLUDE_DIR="/usr/local/include"
-    BOOST_LIBRARY_DIR="/usr/local/lib"
+    BOOST_INCLUDE_DIR=${PREFIX}/include
+    BOOST_LIBRARY_DIR=${PREFIX}/lib
 
-    CXXFLAGS += -DUSE_BOOST -I ${BOOST_INCLUDE_DIR}
+    CXXFLAGS += -DUSE_BOOST -I${BOOST_INCLUDE_DIR} -L${BOOST_LIBRARY_DIR}
+    LDFLAGS += -L${BOOST_LIBRARY_DIR}
 
-    LIBS += ${BOOST_LIBRARY_DIR}/libboost_iostreams.a
+    LIBS += -lboost_iostreams -lz
 
     #if mac
-    LIBS += -lz
+    #LIBS += -lz
 
     #if linux or windows
     #LIBS += ${BOOST_LIBRARY_DIR}/libz.a
--- Makefile	2016-11-18 23:26:30.218928920 +0000
+++ Makefile.new	2016-11-19 01:16:23.950651573 +0000
@@ -12,8 +12,8 @@
 OPTIMIZE ?= yes
 USEREADLINE ?= yes
 USEBOOST ?= yes
-BOOST_LIBRARY_DIR="\"Enter_your_boost_library_path_here\""
-BOOST_INCLUDE_DIR="\"Enter_your_boost_include_path_here\""
+BOOST_LIBRARY_DIR=${PREFIX}/lib
+BOOST_INCLUDE_DIR=${PREFIX}/include
 MOTHUR_FILES="\"Enter_your_default_path_here\""
 RELEASE_DATE = "\"8/9/2016\""
 VERSION = "\"1.38.1\""
@@ -41,7 +41,7 @@
 
 ifeq  ($(strip $(USEREADLINE)),yes)
     CXXFLAGS += -DUSE_READLINE
-    LIBS += -lreadline
+    LIBS += -lreadline -lncurses
 endif
 
 
