--- a/configure	2022-03-29 16:45:50.000000000 +0200
+++ b/configure	2022-03-31 12:00:48.000000000 +0200
@@ -20143,6 +20143,13 @@
 	VRNALIB_OK=true
 fi
 
+    { $as_echo "$as_me:${as_lineno-$LINENO}: Get Vienna RNA configuration from pkg-config and remove -fno-lto." >&5
+$as_echo "$as_me: Get Vienna RNA configuration from pkg-config and remove -fno-lto." >&6;}
+
+    ## strip -fno-lto from VRNA_LIBS
+    VRNA_LIBS="`$as_echo "$VRNA_LIBS" | sed 's/-Wl,-fno-lto//g'`"
+    VRNA_LIBS="`$as_echo "$VRNA_LIBS" | sed 's/-fno-lto//g'`"
+
 fi
 if test "$VRNALIB_OK" != "true" -a "$PKG_CONFIG_AVAILABLE" != "true" ; then :
 
