--- graphlan_annotate.py	2018-06-05 14:52:32.000000000 -0500
+++ graphlan_annotate.py.new	2019-03-19 20:18:46.752259969 -0500
@@ -23,7 +23,7 @@
 
 from sys import argv
 from argparse import ArgumentParser
-from src.graphlan_lib import CircTree as CTree
+from graphlan_lib import CircTree as CTree
 
 
 def read_params(args):
--- graphlan.py	2018-06-05 14:52:32.000000000 -0500
+++ graphlan.py.new	2019-03-19 20:19:08.046264020 -0500
@@ -26,7 +26,7 @@
 
 from sys import argv
 from argparse import ArgumentParser
-from src.graphlan_lib import CircTree as CTree
+from graphlan_lib import CircTree as CTree
 
 
 def read_params(args):
