--- graphlan_annotate.py	2018-06-05 14:52:32.000000000 -0500
+++ graphlan_annotate.py.new	2019-04-23 18:35:08.124974817 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/opt/anaconda1anaconda2anaconda3/bin/python
 
 #------------------------------------------------------------------------------
 # NAME: graphlan_annotate.py
@@ -23,7 +23,7 @@
 
 from sys import argv
 from argparse import ArgumentParser
-from src.graphlan_lib import CircTree as CTree
+from graphlan_lib import CircTree as CTree
 
 
 def read_params(args):
--- graphlan.py	2018-06-05 14:52:32.000000000 -0500
+++ graphlan.py.new	2019-04-23 18:35:44.023976505 -0500
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/opt/anaconda1anaconda2anaconda3/bin/python
 
 #------------------------------------------------------------------------------
 # NAME: graphlan.py
@@ -26,7 +26,7 @@
 
 from sys import argv
 from argparse import ArgumentParser
-from src.graphlan_lib import CircTree as CTree
+from graphlan_lib import CircTree as CTree
 
 
 def read_params(args):
