diff -Naur a/src/fasta_formatter/fasta_formatter.cpp b/src/fasta_formatter/fasta_formatter.cpp
--- a/src/fasta_formatter/fasta_formatter.cpp	2014-01-05 21:10:49.038563296 +0000
+++ b/src/fasta_formatter/fasta_formatter.cpp	2017-07-28 10:52:17.568172280 +0100
@@ -87,12 +87,6 @@
 "\n" \
 "\n";
 
-void usage()
-{
-	printf("%s",usage_string);
-	exit(0);
-}
-
 void parse_command_line(int argc, char* argv[])
 {
 	int opt;
@@ -102,8 +96,9 @@
 		//Parse the default options
 		switch(opt) {
 		case 'h':
-			usage();
-		
+			printf("%s",usage_string);
+			exit(0);
+					
 		case 'i':
 			input_filename = optarg;
 			break;
