diff -Naur scripts/fasta_clipping_histogram.pl scripts/fasta_clipping_histogram.pl
--- scripts/fasta_clipping_histogram.pl	2014-01-05 16:10:49.000000000 -0500
+++ scripts/fasta_clipping_histogram.pl	2018-04-24 14:55:17.509357447 -0400
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 #    FASTX-toolkit - FASTA/FASTQ preprocessing tools.
 #    Copyright (C) 2009-2013  A. Gordon (assafgordon@gmail.com)
diff -Naur scripts/fastx_barcode_splitter.pl scripts/fastx_barcode_splitter.pl
--- scripts/fastx_barcode_splitter.pl	2014-01-05 16:10:49.000000000 -0500
+++ scripts/fastx_barcode_splitter.pl	2018-04-24 14:55:07.909304624 -0400
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 #    FASTX-toolkit - FASTA/FASTQ preprocessing tools.
 #    Copyright (C) 2009-2013  A. Gordon (assafgordon@gmail.com)
diff -Naur src/libfastx/fastx.c src/libfastx/fastx.c
--- src/libfastx/fastx.c	2014-01-05 16:10:49.000000000 -0500
+++ src/libfastx/fastx.c	2018-04-24 14:53:55.188903629 -0400
@@ -126,7 +126,7 @@
 
 	for (i=0; i<strlen(ascii_quality_scores); i++) {
 		pFASTX->quality[i] = (int) (ascii_quality_scores[i] - pFASTX->fastq_ascii_quality_offset ) ;
-		if (pFASTX->quality[i] < -15 || pFASTX->quality[i] > 93) 
+		if (pFASTX->quality[i] < MIN_QUALITY_VALUE || pFASTX->quality[i] > MAX_QUALITY_VALUE) 
 			errx(1, "Invalid quality score value (char '%c' ord %d quality value %d) on line %lld",
 				ascii_quality_scores[i], ascii_quality_scores[i],
 				pFASTX->quality[i], pFASTX->input_line_number );
diff -Naur src/libfastx/fastx.h src/libfastx/fastx.h
--- src/libfastx/fastx.h	2014-01-05 16:10:49.000000000 -0500
+++ src/libfastx/fastx.h	2018-04-24 14:53:55.188903629 -0400
@@ -25,8 +25,8 @@
 /* for PATH_MAX */
 #include <limits.h>
 
-#define MIN_QUALITY_VALUE (-50)
-#define MAX_QUALITY_VALUE 50
+#define MIN_QUALITY_VALUE (-15)
+#define MAX_QUALITY_VALUE 93
 #define QUALITY_VALUES_RANGE (MAX_QUALITY_VALUE-MIN_QUALITY_VALUE)
 
 
