--- configure	2015-12-04 13:14:07.000000000 -0800
+++ oldconfigure	2015-12-04 13:13:34.000000000 -0800
@@ -2865,8 +2865,8 @@
 
 AC_PROG_LIBTOOL
 AX_C_LONG_LONG
-#AX_CXX_HEADER_STDCXX_TR1
-#AX_CXX_COMPILE_STDCXX_11(noext,optional)
+AX_CXX_HEADER_STDCXX_TR1
+AX_CXX_COMPILE_STDCXX_11(noext,optional)
 
 
 PKG_CHECK_MODULES(GTEXTUTILS,gtextutils)
@@ -4111,3 +4111,4 @@
   { $as_echo "$as_me:${as_lineno-$LINENO}: WARNING: unrecognized options: $ac_unrecognized_opts" >&5
 $as_echo "$as_me: WARNING: unrecognized options: $ac_unrecognized_opts" >&2;}
 fi
+
