--- CMakeLists.txt
+++ CMakeLists.txt
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.12)  # required by hunter ZLIB installation
+cmake_minimum_required(VERSION 3.5)  # required by hunter ZLIB installation
 
 # include hunter
 option(HUNTER_STATUS_DEBUG "Hunter debug" OFF)  # comment if does not want hunter debug on
@@ -37,6 +37,9 @@
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fopenmp")
 endif()
 
+# find jsoncpp
+find_package(jsoncpp REQUIRED)
+
 # always link rt
 set(RT_LIBRARY "rt")
 
@@ -184,4 +187,4 @@
 # enable_testing()
 # add_subdirectory(test)
 
-install(TARGETS ${PROJECT_NAME} RUNTIME DESTINATION bin)
\ No newline at end of file
+install(TARGETS ${PROJECT_NAME} RUNTIME DESTINATION bin)
