diff --git a/meson.build b/meson.build
index fd10255..378f784 100644
--- a/meson.build
+++ b/meson.build
@@ -79,24 +79,14 @@ dl_dep  = cc.find_library('dl', required : false)
 # Enable threads in a portable way. Let meson do the hard work
 thread_dep = dependency('threads')
 
-# The following dependencies, if not found in the host, will be
-# downloaded and statically linked
-
 # Zlib
-zlib_dep = dependency('zlib', fallback : ['zlib', 'zlib_dep'])
+zlib_dep = dependency('zlib')
 
 # SQLite3
-sqlite_dep = dependency('sqlite3', version : '>= 3.28.0', fallback : ['sqlite', 'sqlite_dep'])
+sqlite_dep = dependency('sqlite3')
 
 # HTSLib
-htslib_dep = dependency('htslib', version : '>= 1.9', fallback : ['htslib', 'htslib_dep'])
-
-# Libcheck framework is used for testing, so it is not
-# required for building and installing
-check_dep = dependency('check', version : '>= 0.15.0', required : false)
-
-# Sphinx documentation generator
-sphinx_build = find_program('sphinx-build', required : false)
+htslib_dep = dependency('htslib')
 
 # All required deps
 deps = [
@@ -114,9 +104,3 @@ inc = include_directories('.')
 
 # Build static library and executable
 subdir('src')
-
-# Testing
-subdir('tests')
-
-# Documentation
-subdir('docs')
