diff --git a/meson.build b/meson.build
index fd10255..06808d7 100644
--- a/meson.build
+++ b/meson.build
@@ -91,13 +91,6 @@ sqlite_dep = dependency('sqlite3', version : '>= 3.28.0', fallback : ['sqlite',
 # HTSLib
 htslib_dep = dependency('htslib', version : '>= 1.9', fallback : ['htslib', 'htslib_dep'])
 
-# Libcheck framework is used for testing, so it is not
-# required for building and installing
-check_dep = dependency('check', version : '>= 0.15.0', required : false)
-
-# Sphinx documentation generator
-sphinx_build = find_program('sphinx-build', required : false)
-
 # All required deps
 deps = [
   config_dep,
@@ -114,9 +107,3 @@ inc = include_directories('.')
 
 # Build static library and executable
 subdir('src')
-
-# Testing
-subdir('tests')
-
-# Documentation
-subdir('docs')
