From f512a5eaab8b12ef4c78414ffdf813c7ebc872c5 Mon Sep 17 00:00:00 2001
From: Mervin Fansler <mervin.fansler@bric.ku.dk>
Date: Fri, 11 Oct 2024 12:50:25 +0200
Subject: [PATCH] bugfix threads not capped to threads input

---
 R/CreateArrow.R      | 2 +-
 R/MatrixGeneScores.R | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/R/CreateArrow.R b/R/CreateArrow.R
index ac0be626..08243139 100644
--- a/R/CreateArrow.R
+++ b/R/CreateArrow.R
@@ -205,7 +205,7 @@ createArrowFiles <- function(
   if(subThreading){
     h5disableFileLocking()
   }else{
-    args$threads <- length(inputFiles)
+    args$threads <- min(length(inputFiles), threads)
   }
 
   args$minTSS <- NULL
diff --git a/R/MatrixGeneScores.R b/R/MatrixGeneScores.R
index 1d1d10b3..fd342a1a 100644
--- a/R/MatrixGeneScores.R
+++ b/R/MatrixGeneScores.R
@@ -125,7 +125,7 @@ addGeneScoreMatrix <- function(
   if(subThreading){
     h5disableFileLocking()
   }else{
-    args$threads <- length(ArrowFiles)
+    args$threads <- min(length(ArrowFiles), threads)
   }
 
   #Remove Input from args
