--- trinity-plugins/Makefile.old	2017-11-29 16:39:42.851726034 +0200
+++ trinity-plugins/Makefile	2017-12-07 21:44:09.660947521 +0200
@@ -7,11 +7,7 @@
 
 
 
-trinity_essentials: seqtk_target parafly_target trimmomatic_target
-
-trimmomatic_target:
-	ln -sf ${TRIMMOMATIC_CODE} Trimmomatic
-
+trinity_essentials: seqtk_target parafly_target
 
 seqtk_target:
 	tar xvf seqtk-trinity.v0.0.2.tar.gz
@@ -20,7 +16,7 @@
 
 parafly_target:
 	tar -zxvf ${PARAFLY_CODE}.tar.gz && \
-	cd ${PARAFLY_CODE} && sh ./configure --prefix=`pwd` && $(MAKE) install && \
+	cd ${PARAFLY_CODE} && sh ./configure --prefix=`pwd` CC=${PREFIX}/bin/gcc $(INCHWORM_CONFIGURE_FLAGS) CXX=${PREFIX}/bin/g++ CFLAGS="-fopenmp" CXXFLAGS="-fopenmp" && $(MAKE) install && \
 	cp bin/ParaFly ../BIN/
 
 
@@ -40,7 +36,7 @@
 
 clean:
 	rm -rf ./seqtk-trinity-0.0.2
-	cd scaffold_iworm_contigs && $(MAKE) clean
+	#cd scaffold_iworm_contigs && $(MAKE) clean
 	cd parafly-code && $(MAKE) clean
 	rm -f ./parafly # rm symlink
 	rm -f ./Trimmomatic # rm symlink

