diff --git a/configure.ac b/configure.ac
index 25fe884..7de113f 100644
--- a/configure.ac
+++ b/configure.ac
@@ -48,9 +48,6 @@ AC_CHECK_LIB([gsl], [gsl_ran_dirichlet],
 AC_CHECK_LIB([quadmath], [quadmath_snprintf],
              [], [AC_MSG_ERROR([could not find the library quadmath])])
 
-AC_CHECK_LIB([gfortran], [main],
-             [], [AC_MSG_ERROR([could not find the library gfortran])])
-
 AX_BLAS([], [AC_MSG_ERROR([could not find the library gfortran])])
 
 # Set libraries as specified by AX_BLAS documentation
