diff --git a/3rd-party/bamtools/src/bamtools/api/CMakeLists.txt b/3rd-party/bamtools/src/bamtools/api/CMakeLists.txt
index ae4902f..7147865 100644
--- a/3rd-party/bamtools/src/bamtools/api/CMakeLists.txt
+++ b/3rd-party/bamtools/src/bamtools/api/CMakeLists.txt
@@ -48,7 +48,7 @@ else( _WIN32 )
     set( APILibs z )
 endif( _WIN32 )
 
-target_link_libraries( BamTools ${APILibs} )
+target_link_libraries( BamTools ${APILibs} ${ZLIB_LIBRARY})
 
 # set library install destinations
 install( TARGETS BamTools LIBRARY DESTINATION "lib" ARCHIVE DESTINATION "lib")
