--- makefile
+++ makefile
@@ -1,17 +1,17 @@
 abundancebin: main.o profileManager.o myHash.o myMalloc.o myStack.o myStringHash.o
-	g++ -o abundancebin main.o profileManager.o myHash.o myMalloc.o myStack.o myStringHash.o
+	$(CXX) $(LDFLAGS) -o abundancebin main.o profileManager.o myHash.o myMalloc.o myStack.o myStringHash.o
 main.o: main.cpp
-	g++ -c main.cpp
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) -c main.cpp
 profileManager.o: profileManager.cpp
-	g++ -c profileManager.cpp
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) -c profileManager.cpp
 myHash.o: myHash.cpp
-	g++ -c myHash.cpp
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) -c myHash.cpp
 myMalloc.o: myMalloc.cpp
-	g++ -c myMalloc.cpp
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) -c myMalloc.cpp
 myStack.o: myStack.cpp
-	g++ -c myStack.cpp
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) -c myStack.cpp
 myStringHash.o: myStringHash.cpp
-	g++ -c myStringHash.cpp
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) -c myStringHash.cpp
 
 clean:
 	-rm *.o abundancebin
