diff --git a/src/Makefile.am b/src/Makefile.am
index 4bab877..2683af8 100755
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -6,10 +6,10 @@ gadem_SOURCES = main.c mutation.c initial_population.c selection.c \
 		construct_pwm.c	extend_alignment.c align_sites.c \
 		top_kmers.c normalization.c score_subsequence.c transform_pwm.c \
 		sample_wo_replacement.c output.c copy_pwm.c check_pwm_dist.c \
-		consensus.c effect_seq_len.c check_convergence.c read_pwm0.c alloc.h \
+		consensus.c effect_seq_len.c check_convergence.c read_pwm0.c weights.c alloc.h \
 		defines.h evalue_meme.h gadem.h random.h
-                   
-if LARGERINT 
+
+if LARGERINT
    gadem_SOURCES += random_64bits.c
 else
    gadem_SOURCES += random_32bits.c
