diff --git a/src/MinimizerSeeder.cpp b/src/MinimizerSeeder.cpp
index 562cd02..1df4ce1 100644
--- a/src/MinimizerSeeder.cpp
+++ b/src/MinimizerSeeder.cpp
@@ -540,7 +540,7 @@ SeedHit MinimizerSeeder::matchToSeedHit(int nodeId, size_t nodeOffset, size_t se
 {
 	assert(nodeId >= 0);
 	assert((size_t)nodeId < graph.NodeSize());
-	SeedHit result { graph.BigraphNodeID((size_t)nodeId)/2, nodeOffset + graph.NodeOffset((size_t)nodeId), seqPos, minimizerLength, maxCount - count, graph.Reverse((size_t)nodeId) };
+	SeedHit result { static_cast<int>(graph.BigraphNodeID((size_t)nodeId)/2), nodeOffset + graph.NodeOffset((size_t)nodeId), seqPos, minimizerLength, maxCount - count, graph.Reverse((size_t)nodeId) };
 	return result;
 }
 
