diff --git a/makefile b/makefile
index 82aaf50..3dec394 100755
--- a/makefile
+++ b/makefile
@@ -16,7 +16,7 @@ OBJ = $(patsubst %, $(ODIR)/%, $(_OBJ))
 
 LINKFLAGS = $(CPPFLAGS) -Wl,-Bstatic $(LIBS) -Wl,-Bdynamic -Wl,--as-needed -lpthread -pthread -static-libstdc++ $(JEMALLOCFLAGS) `pkg-config --libs libdivsufsort` `pkg-config --libs libdivsufsort64`
 
-VERSION := Branch $(shell git rev-parse --abbrev-ref HEAD) commit $(shell git rev-parse HEAD) $(shell git show -s --format=%ci)
+VERSION := bioconda $(PKG_VERSION)-$(PKG_BUILD_NUMBER)
 
 $(shell mkdir -p bin)
 $(shell mkdir -p obj)
