--- configure	2016-06-08 16:17:37.000000000 -0500
+++ configure.new	2017-01-27 16:20:00.000000000 -0600
@@ -2155,7 +2155,7 @@
     CPPFLAGS=`"${R_HOME}/bin/R" CMD config CPPFLAGS`
     CXX=`"${R_HOME}/bin/R" CMD config CXX`
     CXXFLAGS=`"${R_HOME}/bin/R" CMD config CXXFLAGS`
-    LDFLAGS=`"${R_HOME}/bin/R" CMD config LDFLAGS`
+    LDFLAGS=`"${R_HOME}/bin/R" CMD config LDFLAGS`" -Wl,-rpath ${PREFIX}/lib"
     MAKE=`"${R_HOME}/bin/R" CMD config MAKE`
     ac_ext=c
 ac_cpp='$CPP $CPPFLAGS'
