--- configure	2017-04-24 16:44:10.000000000 -0400
+++ configure.new	2017-10-24 21:24:35.806186099 -0400
@@ -2155,7 +2155,7 @@
     CPPFLAGS=`"${R_HOME}/bin/R" CMD config CPPFLAGS`
     CXX=`"${R_HOME}/bin/R" CMD config CXX`
     CXXFLAGS=`"${R_HOME}/bin/R" CMD config CXXFLAGS`
-    LDFLAGS=`"${R_HOME}/bin/R" CMD config LDFLAGS`
+    LDFLAGS=`"${R_HOME}/bin/R" CMD config LDFLAGS`" -Wl,-rpath ${PREFIX}/lib"
     MAKE=`"${R_HOME}/bin/R" CMD config MAKE`
     ac_ext=c
 ac_cpp='$CPP $CPPFLAGS'
