diff --git a/Makefile b/Makefile
index 92d3be2..8c97ef7 100644
--- a/Makefile
+++ b/Makefile
@@ -19,14 +19,14 @@ ifneq ($(asan),)
 endif
 
 ifeq ($(omp),1)
-	CPPFLAGS+=-DLIBSAIS_OPENMP
-	CFLAGS+=-fopenmp
-	LIBS+=-fopenmp
+	override CPPFLAGS+=-DLIBSAIS_OPENMP
+	override CFLAGS+=-fopenmp
+	override LIBS+=-fopenmp
 endif
 
 ifneq ($(gpl),0)
 	AOBJS+=QSufSort.o bwtgen.o
-	CPPFLAGS+=-DUSE_GPL
+	override CPPFLAGS+=-DUSE_GPL
 endif
 
 ifeq ($(mimalloc),0)
@@ -35,7 +35,7 @@ ifeq ($(mimalloc),0)
 endif
 
 ifeq ($(ARCH), x86_64)
-	CFLAGS+=-msse4.2 -mpopcnt
+	override CFLAGS+=-msse4.2 -mpopcnt
 endif
 
 .SUFFIXES:.c .o
