diff --git i/include/vcf++/VcfFile.hpp w/include/vcf++/VcfFile.hpp
index 4a2f69e..aaf49be 100644
--- i/include/vcf++/VcfFile.hpp
+++ w/include/vcf++/VcfFile.hpp
@@ -35,8 +35,8 @@ THE SOFTWARE.
 #include <string>
 #include <unordered_set>
 
-#include "boost/iostreams/filtering_stream.hpp"
-#include "boost/iostreams/filter/gzip.hpp"
+#include <boost/iostreams/filtering_stream.hpp>
+#include <boost/iostreams/filter/gzip.hpp>
 
 #include "Variant.hpp"
 #include "VcfMetaData.hpp"
diff --git i/src/vcf++/CMakeLists.txt w/src/vcf++/CMakeLists.txt
index c8a6142..beddd90 100755
--- i/src/vcf++/CMakeLists.txt
+++ w/src/vcf++/CMakeLists.txt
@@ -3,6 +3,8 @@ project(vcf++)
 SET(LIBRARY_OUTPUT_PATH ${CMAKE_SOURCE_DIR}/lib)
 
 include_directories(${CMAKE_SOURCE_DIR}/include/vcf++ ${Boost_INCLUDE_DIRS})
+message(STATUS "======= DEBUG Boost_INCLUDE_DIRS: " ${Boost_INCLUDE_DIRS})
+message(STATUS "======= DEBUG Boost_LIB_DIRS: " ${Boost_LIB_DIRS})
 link_directories(${Boost_LIB_DIRS})
 
 add_library(${PROJECT_NAME} VcfFile.cpp Variant.cpp Allele.cpp VcfMetaData.cpp AttributeSet.cpp Sample.cpp Utils.cpp Attribute.cpp JoiningString.cpp Stats.cpp AttributeFilter.cpp SampleAlleleAttributeFilter.cpp FastaReader.cpp FastaRecord.cpp Auxiliaries.cpp Trio.cpp Contig.cpp Regions.cpp)
