diff --git a/external/kmc_api/kmer_defs.h b/external/kmc_api/kmer_defs.h
index 8328822..89ae5e0 100644
--- a/external/kmc_api/kmer_defs.h
+++ b/external/kmc_api/kmer_defs.h
@@ -33,7 +33,12 @@
 
 
 	#include <stdio.h>
-	#include <ext/algorithm>
+    #if defined(__clang__)
+        #include <algorithm>
+    #elif defined(__GNUC__)
+        #include <ext/algorithm>
+        using __gnu_cxx::copy_n;
+    #endif
 	#include <iostream>
 	using namespace std;
 
