--- configure.ac
+++ configure.ac
@@ -283,10 +283,10 @@
 # check for Vienna RNA headers
 
 RNANOTFOUND=0
-AS_IF([test x"$HAVE_PKG_CONFIG" = x"yes"], [
-	AC_MSG_NOTICE([check if VRNA version >= $VRNA_REQUIRED_VERSION])
-	PKG_CHECK_MODULES([VRNA], [RNAlib2 >= $VRNA_REQUIRED_VERSION], [RNANOTFOUND=0], [RNANOTFOUND=1])
-])
+#AS_IF([test x"$HAVE_PKG_CONFIG" = x"yes"], [
+#	AC_MSG_NOTICE([check if VRNA version >= $VRNA_REQUIRED_VERSION])
+#	PKG_CHECK_MODULES([VRNA], [RNAlib2 >= $VRNA_REQUIRED_VERSION], [RNANOTFOUND=0], [RNANOTFOUND=1])
+#])
 
 AS_IF([test "$RNANOTFOUND" = "0"], [
 	AC_MSG_CHECKING([for compilation with Vienna RNA package headers])
