From d1878dbc60b3e4db625f2b3b320c031fe8bfabb6 Mon Sep 17 00:00:00 2001
From: tpob <tpob@tpob.xyz>
Date: Mon, 8 Apr 2024 19:39:30 +0800
Subject: [PATCH] directly write requirements in setup.py

---
 setup.py | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/setup.py b/setup.py
index 2750b76..94ac0bf 100644
--- a/setup.py
+++ b/setup.py
@@ -7,8 +7,7 @@ from setuptools import setup, find_packages
 with open("README.md", "r") as fh:
     long_description = fh.read()
 
-with open("requirements.txt") as f:
-    requirements = f.read().splitlines()
+requirements = ["seqfold>=0.7.3"]
 
 try:
     require("setuptools>=38.3")
-- 
2.31.1

