From aaba304a96a08b459331979042221f6bbae623e1 Mon Sep 17 00:00:00 2001
From: Mervin M Fansler <mmfansler@gmail.com>
Date: Sat, 6 Sep 2025 13:54:26 +0200
Subject: [PATCH] skip unused asm code

---
 src/reader.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/reader.c b/src/reader.c
index 8eafb59..15c8768 100644
--- a/src/reader.c
+++ b/src/reader.c
@@ -962,6 +962,7 @@ int read_fastq_block (memBlock * block, FILE * fp)
   return (1);
 }
 
+#if 0
 static INLINE void
 do_cpuid(uint32_t selector, uint32_t *data)
 {
@@ -972,4 +973,4 @@ do_cpuid(uint32_t selector, uint32_t *data)
       "=d" (data[3])
       : "a"(selector));
 }
-
+#endif
-- 
2.46.0
