--- setup.py	2017-02-08 12:33:39.643035360 +0100
+++ setup.py	2017-02-08 12:33:52.939321021 +0100
@@ -13,9 +13,8 @@
   author = 'Ahmed Mohamed',
   author_email = 'mohamed@kuicr.kyoto-u.ac.jp',
   install_requires=['numpy', 'scipy', 'nmrglue>=0.5'],
-  dependency_links=['https://github.com/jjhelmus/nmrglue/archive/master.zip#egg=nmrglue-0.6'],
   url = 'https://github.com/ahmohamed/nmrpro', # use the URL to the github repo
   license='MIT',
   keywords = ['nmr', 'spectra', 'multi-dimensional'],
   classifiers = [],
-)
\ No newline at end of file
+)
