--- sam_ordered_writer.cpp
+++ sam_ordered_writer.cpp
@@ -135,7 +135,7 @@
 
                 if (i==buffer.end())
                 {
-                    int32_t cutoff_pos1 =  std::max(bam_get_pos1(buffer.front())-window,1);
+                    int32_t cutoff_pos1 =  std::max<int32_t>(bam_get_pos1(buffer.front())-window,1);
                     if (bam_get_pos1(v)<cutoff_pos1)
                     {
 		      notice("[%s:%d %s] Might not be sorted for window size %d at current record %s:%d < %d (%d [last record] - %d), please increase window size to at least %d.\n", __FILE__,__LINE__,__FUNCTION__, window, bam_get_chrom(hdr, v), bam_get_pos1(v), cutoff_pos1, bam_get_pos1(buffer.front()), window, bam_get_pos1(buffer.front())-bam_get_pos1(v)+1);
@@ -218,7 +218,7 @@
   }
   else {
     if (buffer.size()>1) {
-      int32_t cutoff_pos1 =  std::max(bam_get_pos1(buffer.front())-window,1);
+      int32_t cutoff_pos1 =  std::max<int32_t>(bam_get_pos1(buffer.front())-window,1);
       
       while (buffer.size()>1) {
 	if (bam_get_pos1(buffer.back())<=cutoff_pos1) {
