--- defiant.c.old	2019-02-19 20:42:32.000000000 -0600
+++ defiant.c	2019-05-01 00:06:53.616255318 -0500
@@ -33,7 +33,7 @@
 	bool make_heatmap;
 };
 
-int main (int argc, char *restrict argv[]) {
+int main (int argc, char *argv[]) {
 	puts("\n    ____  ___________________    _   ________");
    puts("   / __ \\/ ____/ ____/  _/   |  / | / /_  __/");
    puts("  / / / / __/ / /_   / // /| | /  |/ / / /   ");
