diff --git a/src/Makevars b/src/Makevars
index 404ba20..d47b943 100644
--- a/src/Makevars
+++ b/src/Makevars
@@ -7,7 +7,8 @@ This makefile requires GNU Make.
 endif
 	
 CXX_STD = CXX17
-PKG_CPPFLAGS =-DROUT -I../inst/include -DBOOST_NO_AUTO_PTR -Wno-pedantic  -w -Wfatal-errors
+PKG_CPPFLAGS =-DROUT -I../inst/include -DBOOST_NO_AUTO_PTR -Wno-pedantic  -w -Wfatal-errors \
+              -D_LIBCPP_ENABLE_CXX17_REMOVED_UNARY_BINARY_FUNCTION 
 
 
 
-- 
2.46.0

