diff --git a/setup.py b/setup.py
index 5c1f7cd..a0950ab 100755
--- a/setup.py
+++ b/setup.py
@@ -1,13 +1,16 @@
 import setuptools
 
+# Define a namespace dictionary to hold the variables from the file
+namespace = {}
 with open("mitylib/_version.py", "r") as fh:
-    version = fh.read().replace("__version__ = ", "").strip('""\n')
+    exec(fh.read(), namespace)
+version = namespace['__version__']
 
 with open("README.md", "r") as fh:
     long_description = fh.read()
 
 setuptools.setup(
-    name="{NAME}",
+    name="mity",
     version=version,
     description="A sensitive Mitochondrial variant detection pipeline from WGS data",
     long_description=long_description,
