diff --git a/R/basilisk.R b/R/basilisk.R
index 474c433..ab126a8 100644
--- a/R/basilisk.R
+++ b/R/basilisk.R
@@ -1,12 +1,12 @@
 #' @importFrom basilisk BasiliskEnvironment
 flames_env <- BasiliskEnvironment(
     envname = "flames_env", pkgname = "FLAMES",
-    pip = c("fast-edit-distance==1.2.2", "blaze2==2.5.1", "matplotlib==3.9.1"),
+    pip = c("fast-edit-distance==1.2.2", "blaze2==2.5.1", "matplotlib==3.9.2"), # switch to non-yanked matplotlib version, see https://github.com/matplotlib/matplotlib/issues/28551
     packages = c(
         "python==3.11.9",
         "numpy==2.1.1",
         "scipy==1.14.1",
-        "pysam==0.22.1",
+        "pysam==0.23.2", # switch to working version of pysam, see: https://github.com/pysam-developers/pysam/issues/1339 and here: https://pypi.org/project/pysam/0.23.1/
         "cutadapt==4.9",
         "tqdm==4.66.5",
         "pandas==2.2.3",
