--- a/bin/mango-submit
+++ b/bin/mango-submit
@@ -20,7 +20,15 @@
 # usage: mango-submit [<spark-args> --] <mango-args>
 set -e

-SOURCE_DIR=$(dirname ${BASH_SOURCE[0]})
+# Find original directory of this script, resolving symlinks
+# https://stackoverflow.com/questions/59895/get-the-source-directory-of-a-bash-script-from-within-the-script-itself/246128#246128
+SOURCE="${BASH_SOURCE[0]}"
+while [ -L "$SOURCE" ]; do # resolve $SOURCE until the file is no longer a symlink
+  DIR="$( cd -P "$( dirname "$SOURCE" )" >/dev/null 2>&1 && pwd )"
+  SOURCE="$(readlink "$SOURCE")"
+  [[ $SOURCE != /* ]] && SOURCE="$DIR/$SOURCE" # if $SOURCE was a relative symlink, we need to resolve it relative to the path where the symlink file was located
+done
+SCRIPT_DIR="$( cd -P "$( dirname "$SOURCE" )" >/dev/null 2>&1 && pwd )"

 # Split args into Spark and ADAM args
 DD=False  # DD is "double dash"
@@ -52,10 +60,11 @@


 # get Mango jar
-MANGO_CLI_JAR=$(${SOURCE_DIR}/find-mango-assembly.sh)
+MANGO_CLI_JAR=$(${SCRIPT_DIR}/find-mango-assembly.sh)

-SPARK_SUBMIT=$(${SOURCE_DIR}/find-spark.sh)
+SPARK_SUBMIT=$(${SCRIPT_DIR}/find-spark.sh)
 echo "Using spark-submit=$SPARK_SUBMIT" 1>&2
+ls "$(dirname "$( dirname "$SPARK_SUBMIT" )" )"

 # submit the job to Spark
 "$SPARK_SUBMIT" \
