--- a/bin/mango-notebook
+++ b/bin/mango-notebook
@@ -62,6 +62,16 @@
   echo ${path}
 }

+# Find original directory of this script, resolving symlinks
+# https://stackoverflow.com/questions/59895/get-the-source-directory-of-a-bash-script-from-within-the-script-itself/246128#246128
+SOURCE="${BASH_SOURCE[0]}"
+while [ -h "$SOURCE" ]; do # resolve $SOURCE until the file is no longer a symlink
+  DIR="$( cd -P "$( dirname "$SOURCE" )" >/dev/null 2>&1 && pwd )"
+  SOURCE="$(readlink "$SOURCE")"
+  [[ $SOURCE != /* ]] && SOURCE="$DIR/$SOURCE" # if $SOURCE was a relative symlink, we need to resolve it relative to the path where the symlink file was located
+done
+SCRIPT_DIR="$( cd -P "$( dirname "$SOURCE" )" >/dev/null 2>&1 && pwd )"
+
 # Retrieve the path of py4j zip archive used by current spark
 get_py4jzip(){
   local -a py4j_zip_all_versions
@@ -112,14 +122,13 @@
 # variable declarations
 trap 'err_report ${LINENO} $?' ERR
 declare -r PY4J_ZIP=$(get_py4jzip)
-declare -r SOURCE_DIR=$(dirname ${BASH_SOURCE[0]})
 declare mango_cli_jar=""
 declare wrapper_path
 declare pyspark=""
 parse_args "$@"

 # get Mango jar
-mango_cli_jar=$(${SOURCE_DIR}/find-mango-assembly.sh)
+mango_cli_jar=$(${SCRIPT_DIR}/find-mango-assembly.sh)

 # get pyspark path
 if [[ -z "${SPARK_HOME}" ]]; then
