diff --git a/trimadap.c b/trimadap.c
index 40d42e3..73a278c 100644
--- a/trimadap.c
+++ b/trimadap.c
@@ -37,7 +37,7 @@ typedef struct {
 int main(int argc, char *argv[])
 {
 	int n_adaps, m_adaps;
-	int c, i, j, k, from_stdin;
+	int c, i, j, k;
 	int sa = 1, sb = 2, go = 1, ge = 3, type = 1;
 	int min_sc = 15, min_len = 10;
 	double max_diff = 0.15;
@@ -84,8 +84,7 @@ int main(int argc, char *argv[])
 			p->seq[i] = seq_nt4_table[(uint8_t)p->seq[i]];
 	}
 
-	from_stdin = !isatty(fileno(stdin));
-	if (optind == argc && !from_stdin) {
+	if (optind == argc && isatty(fileno(stdin))) {
 		fprintf(stderr, "\n");
 		fprintf(stderr, "Usage:   trimadap [options] <in.fq>\n\n");
 		fprintf(stderr, "Options: -5 STR     5'-end adapter\n");
@@ -104,7 +103,7 @@ int main(int argc, char *argv[])
 	}
 	for (j = 0; j < 5; ++j) mat[k++] = 0;
 
-	fp = !from_stdin && strcmp(argv[optind], "-")? gzopen(argv[optind], "rb") : gzdopen(fileno(stdin), "rb");
+	fp = optind < argc && strcmp(argv[optind], "-")? gzopen(argv[optind], "rb") : gzdopen(fileno(stdin), "rb");
 	ks = kseq_init(fp);
 	while (kseq_read(ks) >= 0) {
 		if (str.m < ks->seq.m) {
@@ -174,8 +173,8 @@ int main(int argc, char *argv[])
 	for (j = 0; j < n_adaps; ++j) {
 		ta_adap_t *p = &adaps[j];
 		fprintf(stderr, "%-15ld ", (long)p->cnt);
-		for (i = 0; i < p->len; ++i) putchar("ACGTN"[(int)p->seq[i]]);
-		putchar('\n');
+		for (i = 0; i < p->len; ++i) fputc("ACGTN"[(int)p->seq[i]], stderr);
+		fputc('\n', stderr);
 		free(p->seq);
 		free(p->qp);
 	}
