diff --git a/src/SpliceMap/Makefile b/src/SpliceMap/Makefile
index 47b853d..3047d14 100644
--- a/src/SpliceMap/Makefile
+++ b/src/SpliceMap/Makefile
@@ -8,6 +8,7 @@
 #CXXFLAGS=`"${R_HOME}/bin/R" CMD config CXXFLAGS`
 CXX=`"${R_HOME}/bin/R" CMD config CXX`
 CXXFLAGS=`"${R_HOME}/bin/R" CMD config CXXFLAGS`
+LDFLAGS=`"${R_HOME}/bin/R" CMD config LDFLAGS`
 
 all: SpliceMap
 
@@ -15,22 +16,22 @@ clean:
 	rm *.o SpliceMap runSpliceMap_QuasR sortsam amalgamateSAM getSpliceMapUnmapped fuseReorder
 
 SpliceMap: runSpliceMap_QuasR sortsam amalgamateSAM getSpliceMapUnmapped fuseReorder params.o cfgfile.o SpliceMap_utils_QuasR.o main.cpp main.h
-	$(CXX) $(CXXFLAGS) -o SpliceMap main.cpp SpliceMap_utils_QuasR.o params.o cfgfile.o
+	$(CXX) $(LDFLAGS) $(CXXFLAGS) -o SpliceMap main.cpp SpliceMap_utils_QuasR.o params.o cfgfile.o
 
 runSpliceMap_QuasR: SpliceMap_utils_QuasR.o params.o cfgfile.o runSpliceMap_QuasR.cpp runSpliceMap_QuasR.h 
-	$(CXX) $(CXXFLAGS) -o runSpliceMap_QuasR SpliceMap_utils_QuasR.o cfgfile.o runSpliceMap_QuasR.cpp params.o
+	$(CXX) $(LDFLAGS) $(CXXFLAGS) -o runSpliceMap_QuasR SpliceMap_utils_QuasR.o cfgfile.o runSpliceMap_QuasR.cpp params.o
 
 sortsam: sortsam.cpp sortsam.h SpliceMap_utils_QuasR.o params.o
-	$(CXX) $(CXXFLAGS) -o sortsam sortsam.cpp SpliceMap_utils_QuasR.o params.o
+	$(CXX) $(LDFLAGS) $(CXXFLAGS) -o sortsam sortsam.cpp SpliceMap_utils_QuasR.o params.o
 
 amalgamateSAM: amalgamateSAM.cpp amalgamateSAM.h SpliceMap_utils_QuasR.o params.o
-	$(CXX) $(CXXFLAGS) -o amalgamateSAM amalgamateSAM.cpp SpliceMap_utils_QuasR.o params.o
+	$(CXX) $(LDFLAGS) $(CXXFLAGS) -o amalgamateSAM amalgamateSAM.cpp SpliceMap_utils_QuasR.o params.o
 
 getSpliceMapUnmapped: getSpliceMapUnmapped.cpp
-	$(CXX) $(CXXFLAGS) -o getSpliceMapUnmapped getSpliceMapUnmapped.cpp
+	$(CXX) $(LDFLAGS) $(CXXFLAGS) -o getSpliceMapUnmapped getSpliceMapUnmapped.cpp
 
 fuseReorder: fuseReorder.cpp
-	$(CXX) $(CXXFLAGS) -o fuseReorder fuseReorder.cpp
+	$(CXX) $(LDFLAGS) $(CXXFLAGS) -o fuseReorder fuseReorder.cpp
 
 SpliceMap_utils_QuasR.o: SpliceMap_utils_QuasR.cpp SpliceMap_utils_QuasR.h
 	$(CXX) $(CXXFLAGS) -c SpliceMap_utils_QuasR.cpp
