--- auxprogs/compileSpliceCands/Makefile	2017-01-17 15:42:46.129483546 +0400
+++ auxprogs/compileSpliceCands/Makefile	2017-01-17 15:49:00.932632627 +0400
@@ -1,8 +1,8 @@
 compileSpliceCands : compileSpliceCands.o list.h list.o
-	gcc $(LDFLAGS) -o compileSpliceCands compileSpliceCands.o list.o;
+	$(CC) -o compileSpliceCands compileSpliceCands.o list.o;
 #	cp compileSpliceCands ../../bin
 compileSpliceCands.o : compileSpliceCands.c 
-	gcc -Wall -pedantic -ansi $(CPPFLAGS) -c compileSpliceCands.c 
+	$(CC) -Wall -pedantic -ansi $(CPPFLAGS) -c compileSpliceCands.c 
 
 all : compileSpliceCands
 
