--- auxprogs/filterBam/src/Makefile	2016-03-30 18:35:01.000000000 +0400
+++ auxprogs/filterBam/src/Makefile	2016-05-25 17:07:13.875931702 +0400
@@ -8,10 +8,13 @@
 			printElapsedTime.cc sumMandIOperations.cc sumDandIOperations.cc PairednessCoverage.cc
 PROGRAM = filterBam
 OBJECTS = $(SOURCES:.cc=.o)
-BAMTOOLS = /usr/include/bamtools
-INCLUDES = -I$(BAMTOOLS) -Iheaders -I./bamtools
-LIBS = -lbamtools -lz
+BAMTOOLS = ${PREFIX}/include/bamtools
+#INCLUDES = -I$(BAMTOOLS) -Iheaders -I./bamtools
+INCLUDES =  -I${PREFIX}/include/bamtools/ -Iheaders -I./bamtools
+#LIBS = -lbamtools -lz
+LIBS = ${PREFIX}/lib/libbamtools.a -lz
 CFLAGS = -std=c++0x
+CXXFLAGS += -Wall -O2 # -g -p -g -ggdb 
 VPATH = functions
 
 all : $(PROGRAM) $(OBJECTS) CHECKBAM BIN
