--- auxprogs/filterBam/Makefile	2016-03-30 17:40:13.000000000 +0400
+++ auxprogs/filterBam/Makefile	2016-05-25 16:36:14.585884204 +0400
@@ -2,8 +2,14 @@
 #
 # 	Created: 28-February-2012
 #	Last modified: 15-October-2015
+INCLUDES =  ${PREFIX}/include/bamtools/
+LIBS = ${PREFIX}/lib/libbamtools.a -lz
+SOURCES = bam2hints.cc
+OBJECTS = $(SOURCES:.cc=.o)
+CFLAGS = -Wall -g -p -O0
+CXXFLAGS += -Wall -O2 # -g -p -g -ggdb 
 
-all: 
+all:
 	(cd src;make)
 
 clean:
