--- src/calcassoc.cc	2015-08-26 17:42:24.000000000 +0200
+++ src/calcassoc.cc	2017-10-13 23:05:55.325540119 +0200
@@ -21,7 +21,7 @@
 }
 
 void Calcassoc::set_up(Family *first, IV /*maxnumiv*/) {
-  cerr << "calcassoc::set_up() began" << cout;
+  cerr << "calcassoc::set_up() began" << endl;
 
   if (estimatefreq) {
     deletepi();
@@ -100,7 +100,7 @@
 ///////////////////////////////////////////////////////////////////////////
 // Calcassocfrac
 void Calcassocfrac::set_up(Family *first, IV maxnumiv) {
-  cerr << "calcassocFrac::set_up() began" << cout;
+  cerr << "calcassocFrac::set_up() began" << endl;
 
   weights.clear();
   weights.resize(control->map.loci.size());
