--- src/calcviterbi.cc	2015-08-26 17:42:24.000000000 +0200
+++ src/calcviterbi.cc	2017-10-13 23:09:39.635270380 +0200
@@ -14,32 +14,32 @@
 
 void Calcviterbi::set_up(Family */*first*/, IV maxnumiv) {
   cerr << "Ants are fast. Are we 32-bit?" << sizeof(std::size_t) << endl;
-  cerr << "calcviterbi::set_up() began\n" << cout;
+  cerr << "calcviterbi::set_up() began\n" << endl;
 
-  cerr << "calcviterbi::set_up() psibufborrowed=" << psibufborrowed << "\n" << cout;
+  cerr << "calcviterbi::set_up() psibufborrowed=" << psibufborrowed << "\n" << endl;
   if (!psibufborrowed) delete [] psibuf;
   delete psi;
-  cerr << "calcviterbi::set_up() psi deleted" << "\n" << cout;
+  cerr << "calcviterbi::set_up() psi deleted" << "\n" << endl;
 
-  cerr << "calcviterbi::set_up() options_swap?" << options->swap << "\n" << cout;
-  cerr << "calcviterbi::set_up() map.loci.size=" << control->map.loci.size() << ", maxnumiv=" << maxnumiv << "\n" << cout;
+  cerr << "calcviterbi::set_up() options_swap?" << options->swap << "\n" << endl;
+  cerr << "calcviterbi::set_up() map.loci.size=" << control->map.loci.size() << ", maxnumiv=" << maxnumiv << "\n" << endl;
 
   Uint n = Uint(options->swap ? Double(maxnumiv) :
                 (Double(control->map.loci.size())*Double(maxnumiv)));
 
-  cerr << "calcviterbi::set_up() n value=" << n << "\n" << cout;
+  cerr << "calcviterbi::set_up() n value=" << n << "\n" << endl;
 
   if (n == 0) n = 1;
   psibuf = getbuf(n);
-  cerr << "calcviterbi::set_up() getbuf(" << n << ")\n" << cout;
+  cerr << "calcviterbi::set_up() getbuf(" << n << ")\n" << endl;
 
-  cerr << "calcviterbi::set_up() psibuf=" << psibuf << "\n" << cout;
+  cerr << "calcviterbi::set_up() psibuf=" << psibuf << "\n" << endl;
   psibufborrowed = psibuf != 0;
-  cerr << "calcviterbi::set_up() psibufborrowed=" << psibufborrowed << "\n" << cout;
+  cerr << "calcviterbi::set_up() psibufborrowed=" << psibufborrowed << "\n" << endl;
 
   if (!psibufborrowed){
       //PROBLEMS OCCUR HERE
-      cerr << "calcviterbi::set_up() new IV \n" << cout;
+      cerr << "calcviterbi::set_up() new IV \n" << endl;
       cerr << "size of IV:" << sizeof(IV) <<  endl;
       cerr << "total requested: ("
 //           << (sizeof(IV)*n) << "bytes) = ("
@@ -48,12 +48,12 @@
            << (sizeof(IV)*n)/(1024*1024*1024) << "GB)" << endl;
 
       psibuf = new IV[n];
-      cerr << "calcviterbi::set_up() got new IV \n" << cout;
+      cerr << "calcviterbi::set_up() got new IV \n" << endl;
   }
 
-  cerr << "calcviterbi::set_up() new IVMatrix \n" << cout;
+  cerr << "calcviterbi::set_up() new IVMatrix \n" << endl;
   psi = new IVmatrix(psibuf, n, options->swapdirname, "psi." + id + ".");
-  cerr << "calcviterbi::set_up() got new IVMatrix \n" << cout;
+  cerr << "calcviterbi::set_up() got new IVMatrix \n" << endl;
 
 }
 
