--- src/genotypes.cc	2015-08-26 17:42:24.000000000 +0200
+++ src/genotypes.cc	2017-10-13 23:15:03.274988524 +0200
@@ -231,10 +231,10 @@
       dist_female[0] = firstmarker.location_female;
   }
   for (vector<string>::const_iterator mn = ch->second.markers.begin();
-       mn != ch->second.markers.end() && next(mn) != ch->second.markers.end();
+       mn != ch->second.markers.end() && std::next(mn) != ch->second.markers.end();
        mn++, midx++) {
     const Marker &marker = string2marker[*mn];
-    const Marker &nextmarker = string2marker[*next(mn)];
+    const Marker &nextmarker = string2marker[*std::next(mn)];
     dist[midx] = nextmarker.location - marker.location;
     if (options->sexspecific)
       dist_female[midx] = nextmarker.location_female - marker.location_female;
