--- src/inherdist.cc	2015-08-26 17:42:24.000000000 +0200
+++ src/inherdist.cc	2017-10-13 23:04:24.519008503 +0200
@@ -34,7 +34,7 @@
                         DoubleVec p) {
   const string fn = getfilename(famid, gam);
   ifstream f(fn.c_str(), ios::in | ios::binary);
-  assertcond(f != 0, "Unable to read inheritance vector file " + fn);
+  assertcond(!f.is_open(), "Unable to read inheritance vector file " + fn);
   f.seekg(0, ios::beg);
   f.read((char *)(p), numiv*sizeof(Float));
   assertcond(!f.fail(), "Error reading inheritance vector file " + fn);
@@ -57,7 +57,7 @@
 
   const string fn = getfilename(families.back()->id, pos);
   ofstream f(fn.c_str(), ios::out | ios::binary);
-  assertcond(f != 0, "Unable to write inheritance vector file " + fn);
+  assertcond(!f.is_open(), "Unable to write inheritance vector file " + fn);
   f.seekp(ios::beg);
   f.write((char *)(pv), families.back()->numiv()*sizeof(Float));
   assertcond(!f.fail(), "Error write inheritance vector file " + fn);  
