--- src/options.cc	2015-08-26 17:42:24.000000000 +0200
+++ src/options.cc	2017-10-13 22:47:36.689515517 +0200
@@ -620,7 +620,7 @@
   seed[2] = 0x330E;
 
   ifstream optfil(optionsfile.c_str());
-  if (optfil == 0) fatal("Cannot open optionsfile " + optionsfile);
+  if (!optfil.is_open()) fatal("Cannot open optionsfile " + optionsfile);
   message("Using options file " + optionsfile);
 
   Table<File> ftbl;
