--- src/peel.cc	2015-08-26 17:42:24.000000000 +0200
+++ src/peel.cc	2017-10-13 23:12:41.228433765 +0200
@@ -257,7 +257,7 @@
   assertinternal(!fam.empty());
   first = new ESpeople(new ESperson(*fam.begin()), &first, 0);
 
-  for (set<Person *>::const_iterator pi = next(fam.begin());
+  for (set<Person *>::const_iterator pi = std::next(fam.begin());
        pi != fam.end();
        pi++){
     first->insertafter(new ESperson(*pi));
