diff --git a/Makefile b/Makefile
index a8a89d4..5bc078a 100644
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,6 @@
 #Set these variables if needed
-C = gcc
-CC = g++
+C ?= $(CC)
+CC ?= $(CXX)
 FLAGS = -O3 -static -D_NOSQLITE -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC
 
 #Path to MSToolkit
@@ -8,7 +8,7 @@ FLAGS = -O3 -static -D_NOSQLITE -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC
 LIBPATH = ../MSToolkit
 LIBS = -lmstoolkitlite
 
-INCLUDE = -I$(LIBPATH)/include -I$(LIBPATH)/mzParser
+INCLUDE = -I$(PREFIX)/include -I$(LIBPATH)/include -I$(LIBPATH)/mzParser
 
 
 #Do not touch these variables
@@ -18,8 +18,8 @@ HARDKLOR = CHardklor.o CAveragine.o CPeriodicTable.o CHardklorVariant.o CHardklo
 
 #Make statements
 hardklor : HardklorApp.cpp $(HARDKLOR) $(SUPPORT)
-	$(CC) $(FLAGS) $(INCLUDE) $(SUPPORT) $(HARDKLOR) HardklorApp.cpp -L$(LIBPATH) $(LIBS) -o hardklor
-	ar rcs libhardklor.a $(HARDKLOR) $(SUPPORT)
+	$(CC) $(FLAGS) $(INCLUDE) $(SUPPORT) $(HARDKLOR) HardklorApp.cpp -L$(PREFIX)/lib -L$(LIBPATH) $(LIBS) -o hardklor
+	$(AR) rcs libhardklor.a $(HARDKLOR) $(SUPPORT)
 
 clean:
 	rm -f *.o hardklor libhardklor.a
