diff --git a/makefile b/makefile
index 104e33c..0d0bf3d 100755
--- a/makefile
+++ b/makefile
@@ -2,63 +2,63 @@ all: dir astral astral-pro wastral caster-site caster-site_branchlength caster-p
 	echo "*** Installation complete! ***"
 
 astral: dir
-	g++ -std=gnu++11 -march=native -Ofast -pthread src/astral.cpp -o bin/astral4 || g++ -std=c++17 -O2 -pthread src/astral.cpp -o bin/astral4
+	$(CXX) -std=gnu++11 -march=x86-64-v3 -Ofast -pthread src/astral.cpp -o bin/astral4 || $(CXX) -std=c++17 -O3 $(LDFLAGS) -pthread src/astral.cpp -o bin/astral4
 	cp bin/astral4 bin/astral
 
 astral_int128: dir
-	g++ -std=gnu++11 -march=native -D LARGE_DATA -Ofast -pthread src/astral.cpp -o bin/astral4_int128
+	$(CXX) -std=gnu++11 -march=x86-64-v3 -D LARGE_DATA -Ofast $(LDFLAGS) -pthread src/astral.cpp -o bin/astral4_int128
 
 astral-pro: dir
-	g++ -std=gnu++11 -march=native -Ofast -pthread src/astral-pro.cpp -o bin/astral-pro3 || g++ -std=c++17 -O2 -pthread src/astral-pro.cpp -o bin/astral-pro3
+	$(CXX) -std=gnu++11 -march=x86-64-v3 -Ofast -pthread src/astral-pro.cpp -o bin/astral-pro3 || $(CXX) -std=c++17 -O3 $(LDFLAGS) -pthread src/astral-pro.cpp -o bin/astral-pro3
 	cp bin/astral-pro3 bin/astral-pro
 	
 astral-pro_int128: dir
-	g++ -std=gnu++11 -march=native -D LARGE_DATA -Ofast -pthread src/astral-pro.cpp -o bin/astral-pro3_int128
+	$(CXX) -std=gnu++11 -march=x86-64-v3 -D LARGE_DATA -Ofast $(LDFLAGS) -pthread src/astral-pro.cpp -o bin/astral-pro3_int128
 
 wastral: dir
-	g++ -std=gnu++11 -march=native -Ofast -pthread src/astral-hybrid.cpp -o bin/wastral || g++ -std=c++17 -O2 -pthread src/astral-hybrid.cpp -o bin/wastral
+	$(CXX) -std=gnu++11 -march=x86-64-v3 -Ofast -pthread src/astral-hybrid.cpp -o bin/wastral || $(CXX) -std=c++17 -O3 $(LDFLAGS) -pthread src/astral-hybrid.cpp -o bin/wastral
 	cp bin/wastral bin/astral-hybrid
 	
 wastral_precise: dir
-	g++ -std=gnu++11 -march=native -D LARGE_DATA -Ofast -pthread src/astral-hybrid.cpp -o bin/wastral_precise
+	$(CXX) -std=gnu++11 -march=x86-64-v3 -D LARGE_DATA -Ofast $(LDFLAGS) -pthread src/astral-hybrid.cpp -o bin/wastral_precise
 	
 caster-site: dir
-	g++ -std=gnu++11 -march=native -Ofast -pthread src/caster-site.cpp -o bin/caster-site || g++ -std=gnu++17 -O2 -pthread src/caster-site.cpp -o bin/caster-site
+	$(CXX) -std=gnu++11 -march=x86-64-v3 -Ofast -pthread src/caster-site.cpp -o bin/caster-site || $(CXX) -std=gnu++17 -O3 $(LDFLAGS) -pthread src/caster-site.cpp -o bin/caster-site
 
 caster-site_branchlength: dir
-	g++ -std=gnu++11 -march=native -D CUSTOMIZED_ANNOTATION_TERMINAL_LENGTH -Ofast -pthread src/caster-site.cpp -o bin/caster-site_branchlength || g++ -std=gnu++17 -D CUSTOMIZED_ANNOTATION_TERMINAL_LENGTH -O2 -pthread src/caster-site.cpp -o bin/caster-site_branchlength
+	$(CXX) -std=gnu++11 -march=x86-64-v3 -D CUSTOMIZED_ANNOTATION_TERMINAL_LENGTH -Ofast $(LDFLAGS) -pthread src/caster-site.cpp -o bin/caster-site_branchlength || $(CXX) -std=gnu++17 -D CUSTOMIZED_ANNOTATION_TERMINAL_LENGTH -O3 -pthread src/caster-site.cpp -o bin/caster-site_branchlength
 	
 caster-pair: dir
-	g++ -std=gnu++11 -march=native -Ofast -pthread src/caster-pair.cpp -o bin/caster-pair || g++ -std=gnu++17 -O2 -pthread src/caster-pair.cpp -o bin/caster-pair
+	$(CXX) -std=gnu++11 -march=x86-64-v3 -Ofast -pthread src/caster-pair.cpp -o bin/caster-pair || $(CXX) -std=gnu++17 -O3 $(LDFLAGS) -pthread src/caster-pair.cpp -o bin/caster-pair
 
 waster-old_branchlength: dir
-	g++ -std=gnu++11 -march=native -D CUSTOMIZED_ANNOTATION_TERMINAL_LENGTH -Ofast -pthread src/waster-old.cpp -o bin/waster-old_branchlength || g++ -std=gnu++17 -D CUSTOMIZED_ANNOTATION_TERMINAL_LENGTH -O2 -pthread src/waster-old.cpp -o bin/waster-old_branchlength_experimental
+	$(CXX) -std=gnu++11 -march=x86-64-v3 -D CUSTOMIZED_ANNOTATION_TERMINAL_LENGTH -Ofast $(LDFLAGS) -pthread src/waster-old.cpp -o bin/waster-old_branchlength || $(CXX) -std=gnu++17 -D CUSTOMIZED_ANNOTATION_TERMINAL_LENGTH -O3 -pthread src/waster-old.cpp -o bin/waster-old_branchlength_experimental
 	
 waster-old: dir
-	g++ -std=gnu++11 -march=native -Ofast -pthread src/waster-old.cpp -o bin/waster-old || g++ -std=gnu++17 -O2 -pthread src/waster-old.cpp -o bin/waster-old
+	$(CXX) -std=gnu++11 -march=x86-64-v3 -Ofast -pthread src/waster-old.cpp -o bin/waster-old || $(CXX) -std=gnu++17 -O3 $(LDFLAGS) -pthread src/waster-old.cpp -o bin/waster-old
 
 waster_branchlength: dir
-	g++ -std=gnu++11 -march=native -D CUSTOMIZED_ANNOTATION_TERMINAL_LENGTH -Ofast -pthread src/waster.cpp -o bin/waster_branchlength || g++ -std=gnu++17 -D CUSTOMIZED_ANNOTATION_TERMINAL_LENGTH -O2 -pthread src/waster.cpp -o bin/waster_branchlength
+	$(CXX) -std=gnu++11 -march=x86-64-v3 -D CUSTOMIZED_ANNOTATION_TERMINAL_LENGTH -Ofast $(LDFLAGS) -pthread src/waster.cpp -o bin/waster_branchlength || $(CXX) -std=gnu++17 -D CUSTOMIZED_ANNOTATION_TERMINAL_LENGTH -O3 -pthread src/waster.cpp -o bin/waster_branchlength
 	
 waster: dir
-	g++ -std=gnu++11 -march=native -Ofast -pthread src/waster.cpp -o bin/waster || g++ -std=gnu++17 -O2 -pthread src/waster.cpp -o bin/waster
+	$(CXX) -std=gnu++11 -march=x86-64-v3 -Ofast -pthread src/waster.cpp -o bin/waster || $(CXX) -std=gnu++17 -O3 $(LDFLAGS) -pthread src/waster.cpp -o bin/waster
 
 waster-test: waster
 	bash example/sancheck/wasterExample/sancheck-local.sh
 
 dstar: dir
-	g++ -std=gnu++11 -march=native -Ofast src/dstar.cpp -o bin/dstar || g++ -std=gnu++17 -O2 src/dstar.cpp -o bin/dstar
+	$(CXX) -std=gnu++11 -march=x86-64-v3 -Ofast src/dstar.cpp -o bin/dstar || $(CXX) -std=gnu++17 -O3 $(LDFLAGS) -pthread src/dstar.cpp -o bin/dstar
 
 dstar-branch: dir
-	g++ -std=gnu++11 -march=native -Ofast src/dstar-branch.cpp -o bin/dstar-branch || g++ -std=gnu++17 -O2 src/dstar-branch.cpp -o bin/dstar-branch
-	g++ -std=gnu++11 -march=native -Ofast src/dstar-branch-merge-diag.cpp -o bin/dstar-branch-merge-diag || g++ -std=gnu++17 -O2 src/dstar-branch-merge-diag.cpp -o bin/dstar-branch-merge-diag
+	$(CXX) -std=gnu++11 -march=x86-64-v3 -Ofast src/dstar-branch.cpp -o bin/dstar-branch || $(CXX) -std=gnu++17 -O3 $(LDFLAGS) -pthread src/dstar-branch.cpp -o bin/dstar-branch
+	$(CXX) -std=gnu++11 -march=x86-64-v3 -Ofast src/dstar-branch-merge-diag.cpp -o bin/dstar-branch-merge-diag || $(CXX) -std=gnu++17 -O3 $(LDFLAGS) -pthread src/dstar-branch-merge-diag.cpp -o bin/dstar-branch-merge-diag
 
 sister: dir
-	g++ -std=gnu++11 -march=native -Ofast -pthread src/sister.cpp -o bin/sister || g++ -std=gnu++17 -O2 -pthread src/sister.cpp -o bin/sister
+	$(CXX) -std=gnu++11 -march=x86-64-v3 -Ofast -pthread src/sister.cpp -o bin/sister || $(CXX) -std=gnu++17 -O3 $(LDFLAGS) -pthread src/sister.cpp -o bin/sister
 	
 dir:
-	g++ -v 2>&1 | tail -n 1
-	echo 'If installation failed, please ensure g++ version >= 7.5.0'
+	$(CXX) -v 2>&1 | tail -n 1
+	echo 'If installation failed, please ensure $(CXX) version >= 7.5.0'
 	mkdir -p bin
 	
 clean:
