--- gtdbtk/main.py	2019-01-14 11:10:50.000000000 +1000
+++ gtdbtk/main_new.py	2020-03-26 11:55:04.983031963 +1000
@@ -316,6 +316,7 @@

     def parse_options(self, options):
         """Parse user options and call the correct pipeline(s)"""
+        self.logger.warning('You are not running the most recent version of GTDB-Tk!')

         if(options.subparser_name == 'de_novo_wf'):
             check_dependencies(['prodigal', 'hmmalign'])
