--- trinity-plugins/Makefile	2016-07-01 23:32:24.478934797 +0400
+++ trinity-plugins/Makefile	2016-07-02 00:32:49.650063388 +0400
@@ -29,7 +29,7 @@
 htslib_target:
 #	tar xjvf htslib-1.2.1.tar.bz2
 #	cd htslib-1.2.1 && ./configure && $(MAKE)
-	tar xvf ${HTSLIB_CODE} && cd htslib && $(MAKE)
+	tar xvf ${HTSLIB_CODE} && sed -i 's/CPPFLAGS =/#CPPFLAGS =/' htslib/Makefile && cd htslib && $(MAKE)
 
 scaffold_iworm_contigs_target: htslib_target
 	cd scaffold_iworm_contigs && $(MAKE)
