diff --git a/kent/src/inc/bamFile.h b/kent/src/inc/bamFile.h
index 4c517c9..48e4e38 100644
--- a/kent/src/inc/bamFile.h
+++ b/kent/src/inc/bamFile.h
@@ -11,7 +11,7 @@
 #ifndef _IOLIB
 #define _IOLIB 2
 #endif
-#include "htslib/sam.h"
+#include "../htslib/htslib/sam.h"
 typedef samFile samfile_t;
 typedef hts_idx_t bam_index_t;
 typedef bam_hdr_t bam_header_t;
diff --git a/bamFile.c b/bamFile.c
index 9f4f9ab..0b8b972 100644
--- a/kent/src/lib/bamFile.c
+++ b/kent/src/lib/bamFile.c
@@ -7,10 +7,10 @@
 #include "portable.h"
 #include "bamFile.h"
 #include "htmshell.h"
-#include "cram/cram_samtools.h"
-#include "cram/sam_header.h"
-#include "cram/cram_structs.h"
-#include "htslib/cram.h"
+#include "../htslib/cram/cram_samtools.h"
+#include "../htslib/cram/sam_header.h"
+#include "../htslib/cram/cram_structs.h"
+#include "../htslib/htslib/cram.h"
 #include "udc.h"
 #include "psl.h"
 
