diff --git a/src/nicksrc/Makefile b/src/nicksrc/Makefile
index e93d10c..23d5daf 100644
--- a/src/nicksrc/Makefile
+++ b/src/nicksrc/Makefile
@@ -1,10 +1,10 @@
-override CFLAGS += -c -g -p -Wimplicit -fno-common -I./
+override CFLAGS += -g -O3 -Wimplicit -fno-common -I$(PREFIX)/include -I./
 
 all: libnick.a
 
 libnick.a: strsubs.o sortit.o vsubs.o statsubs.o linsubs.o getpars.o xsearch.o gauss.o gds.o
-	ar -r libnick.a $^
-	ranlib libnick.a
+	$(AR) -rcs libnick.a $^
+	$(RANLIB) libnick.a
 
 clean:
 	rm -f *.o
