From 3fd1db48c11837bea0225037ec4e4fec8ae667ee Mon Sep 17 00:00:00 2001
From: "Robert A. Petit III" <robbie.petit@gmail.com>
Date: Mon, 7 Dec 2020 10:06:37 -0500
Subject: [PATCH] Update Makefile

---
 src/Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/Makefile b/src/Makefile
index 813804e..aad8b21 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -29,7 +29,7 @@ PERLP = jackdiff mkpretty
 
 PROGS2 = rexpfit.r  wtjack.pl 
 
-CC = cc
+CC ?= cc
 
 all: $(NLIB) $(PROGS) $(PROGS3) 
 
