--- setup.py.orig	2025-10-03 22:45:48.271123162 -0400
+++ setup.py	2025-10-03 22:46:05.870260704 -0400
@@ -21,7 +21,7 @@
   author_email = 'cgphelp@sanger.ac.uk',
   packages = ['parabam','parabam.command'],
   package_dir = {'parabam':'parabam','parabam.command':'parabam/command'},
-  install_requires = ['numpy','argparse','pysam >= 0.10.0'],
+  install_requires = ['numpy','pysam >= 0.10.0'],
   scripts = ['parabam/bin/parabam'],
   cmdclass = {'sdist': sdist},
   ext_modules=[
