--- bin/bwa.orig	2018-01-06 15:38:56.000000000 -0500
+++ bin/bwa	2018-01-12 15:20:32.878461984 -0500
@@ -1,12 +1,15 @@
-#!/bin/sh
+#!/bin/bash
 OS=$(uname)
-if [ ! ${0%%/*} ]; then
-    bin_dir=$(dirname "$0")
-else
-    bin_dir=$(dirname "$(pwd)/$0")
-fi
-# remove trailing '/.'
-bin_dir=${bin_dir%%/.*}
+
+# Find original directory of bash script, resovling symlinks
+# http://stackoverflow.com/questions/59895/can-a-bash-script-tell-what-directory-its-stored-in/246128#246128
+SOURCE="${BASH_SOURCE[0]}"
+while [ -h "$SOURCE" ]; do # resolve $SOURCE until the file is no longer a symlink
+    DIR="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
+    SOURCE="$(readlink "$SOURCE")"
+    [[ $SOURCE != /* ]] && SOURCE="$DIR/$SOURCE" # if $SOURCE was a relative symlink, we need to resolve it relative to the path where the symlink file was located
+done
+bin_dir="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
 install_dir=$(dirname "$bin_dir")
 lib_dir=$install_dir/lib
 libexec_dir=$install_dir/libexec
