--- bin/sentieon.orig	2016-04-28 20:01:54.000000000 -0400
+++ bin/sentieon	2016-06-17 14:26:15.623750528 -0400
@@ -1,10 +1,16 @@
 #!/bin/sh
 OS=$(uname)
-if [ ! ${0%%/*} ]; then
-    bin_dir=$(dirname "$0")
-else
-    bin_dir=$(dirname "$(pwd)/$0")
-fi
+
+# Find original directory of bash script, resovling symlinks
+# http://stackoverflow.com/questions/59895/can-a-bash-script-tell-what-directory-its-stored-in/246128#246128
+SOURCE="${BASH_SOURCE[0]}"
+while [ -h "$SOURCE" ]; do # resolve $SOURCE until the file is no longer a symlink
+    DIR="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
+    SOURCE="$(readlink "$SOURCE")"
+    [[ $SOURCE != /* ]] && SOURCE="$DIR/$SOURCE" # if $SOURCE was a relative symlink, we need to resolve it relative to the path where the symlink file was located
+done
+bin_dir="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
+
 install_dir=$(dirname "$bin_dir")
 lib_dir=$install_dir/lib
 libexec_dir=$install_dir/libexec
