diff -ruN MITObim-1.9.1/misc_scripts/circules.py MITObim-1.9.1/misc_scripts/circules.py
--- MITObim-1.9.1/misc_scripts/circules.py	2018-09-03 14:01:32.000000000 +0200
+++ MITObim-1.9.1/misc_scripts/circules.py	2020-12-29 14:58:03.224337731 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/usr/bin/env python
 
 import sys,warnings
 import argparse
diff -ruN MITObim-1.9.1/misc_scripts/downsample.py MITObim-1.9.1/misc_scripts/downsample.py
--- MITObim-1.9.1/misc_scripts/downsample.py	2018-09-03 14:01:32.000000000 +0200
+++ MITObim-1.9.1/misc_scripts/downsample.py	2020-12-29 14:58:13.520129158 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/usr/bin/env python
 
 
 """downsample
diff -ruN MITObim-1.9.1/misc_scripts/get_wiggle.pl MITObim-1.9.1/misc_scripts/get_wiggle.pl
--- MITObim-1.9.1/misc_scripts/get_wiggle.pl	2018-09-03 14:01:32.000000000 +0200
+++ MITObim-1.9.1/misc_scripts/get_wiggle.pl	2020-12-29 14:58:24.723903740 +0100
@@ -1,4 +1,4 @@
-#! /usr/bin/perl
+#!/usr/bin/env perl
 #
 # script to extract per base coverage info for unpadded
 # assembly result from MIRA wiggle file
diff -ruN MITObim-1.9.1/misc_scripts/interleave-fastqgz-MITOBIM.py MITObim-1.9.1/misc_scripts/interleave-fastqgz-MITOBIM.py
--- MITObim-1.9.1/misc_scripts/interleave-fastqgz-MITOBIM.py	2018-09-03 14:01:32.000000000 +0200
+++ MITObim-1.9.1/misc_scripts/interleave-fastqgz-MITOBIM.py	2020-12-29 14:58:34.339711535 +0100
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/usr/bin/env python
 # encoding:utf8
 # authors: Erik Garrison, Sébastien Boisvert
 # modified by github@cypridina on 20151104 to work with MITObim
diff -ruN MITObim-1.9.1/MITObim.pl MITObim-1.9.1/MITObim.pl
--- MITObim-1.9.1/MITObim.pl	2018-09-03 14:01:32.000000000 +0200
+++ MITObim-1.9.1/MITObim.pl	2020-12-29 19:30:17.955602013 +0100
@@ -1,4 +1,4 @@
-#! /usr/bin/perl
+#!/usr/bin/env perl
 #
 # MITObim - mitochondrial baiting and iterative mapping
 # wrapper script version 1.9.1
@@ -71,9 +71,7 @@
 #		--insert <int>		insert size of illumina library, default=300, relevant only for proofreading
 
 
-if (scalar @ARGV < 2){
-	die "$PROGRAM\n$VERSION\n$USAGE";
-}
+
 my $command = $0;
 for (@ARGV){
        	$command .= " $_";
