--- mirge/rScripts/difxp_DESeq.R.old	2023-05-11 10:18:53.810745435 -0500
+++ mirge/rScripts/difxp_DESeq.R	2023-05-11 10:23:53.204956327 -0500
@@ -1,14 +1,3 @@
-# Check if ggplot2 is installed, if not installed create a personal library and install there, first time installation may take time!.
-dir.create(Sys.getenv("R_LIBS_USER"), recursive = TRUE, showWarnings = FALSE)  # create personal library
-.libPaths(Sys.getenv("R_LIBS_USER"))  # add to the path
-packages <- c("ggplot2")
-install.packages(setdiff(packages, rownames(installed.packages())))
-
-if (!require("BiocManager", quietly = TRUE))
-    install.packages("BiocManager")
-if (!require("DESeq2", quietly = TRUE))
-	BiocManager::install("DESeq2")
-
 library(ggplot2)
 #suppressMessages(library(DESeq2, quietly = T))
 suppressMessages(library(DESeq2))
