--- TIDDIT.py	2020-09-08 13:52:18.886911002 +0200
+++ TIDDIT.py	2020-09-08 13:52:49.649439603 +0200
@@ -60,14 +60,14 @@
 		print ("error,  could not find the bam file")
 		quit()
 
-	if not os.path.isfile("{}/bin/TIDDIT".format(args.wd)):
+	if not os.path.isfile("{}/TIDDIT".format(args.wd)):
 		print ("error,  could not find the TIDDIT executable file, try rerun the INSTALL.sh script")
 		quit()
 
 	if not args.bam.endswith(".cram"):
-		command_str="{}/bin/TIDDIT --sv -b {} -o {} -p {} -r {} -q {} -n {} -s {}".format(args.wd,args.bam,args.o,args.p,args.r,args.q,args.n,args.s)
+		command_str="{}/TIDDIT --sv -b {} -o {} -p {} -r {} -q {} -n {} -s {}".format(args.wd,args.bam,args.o,args.p,args.r,args.q,args.n,args.s)
 	else:
-		command_str="samtools view -hbu {} -T {} | {}/bin/TIDDIT --sv -b /dev/stdin -o {} -p {} -r {} -q {} -n {} -s {}".format(args.bam,args.ref,args.wd,args.o,args.p,args.r,args.q,args.n,args.s)
+		command_str="samtools view -hbu {} -T {} | {}/TIDDIT --sv -b /dev/stdin -o {} -p {} -r {} -q {} -n {} -s {}".format(args.bam,args.ref,args.wd,args.o,args.p,args.r,args.q,args.n,args.s)
 	
 
 	if args.i:
@@ -82,9 +82,9 @@
 			t=time.time()
 			print ("Generating GC wig file")
 			if args.ref.endswith(".gz"):
-				os.system("zcat {} | {}/bin/TIDDIT --gc -z 50 -o {}".format(args.ref,args.wd,args.o))
+				os.system("zcat {} | {}/TIDDIT --gc -z 50 -o {}".format(args.ref,args.wd,args.o))
 			else:
-				os.system("cat {} | {}/bin/TIDDIT --gc -z 50 -o {}".format(args.ref,args.wd,args.o))
+				os.system("cat {} | {}/TIDDIT --gc -z 50 -o {}".format(args.ref,args.wd,args.o))
 			print ("Constructed GC wig in {} sec".format(time.time()-t))
 
 	if not args.no_cluster:
@@ -108,12 +108,12 @@
 			quit()
 
 	if not args.bam.endswith(".cram"):
-		command="{}/bin/TIDDIT --cov -b {} -o {} -z {}".format(args.wd,args.bam,args.o,args.z)
+		command="{}/TIDDIT --cov -b {} -o {} -z {}".format(args.wd,args.bam,args.o,args.z)
 	else:
 		if not args.ref:
 			print("error, missing reference sequence!")
 			quit()
-		command="samtools view -hbu {} -T {} | {}/bin/TIDDIT --cov -b /dev/stdin -o {} -z {}".format(args.bam,args.ref,args.wd,args.o,args.z)
+		command="samtools view -hbu {} -T {} | {}/TIDDIT --cov -b /dev/stdin -o {} -z {}".format(args.bam,args.ref,args.wd,args.o,args.z)
 
 	if args.w:
 		command += " -w"
@@ -124,7 +124,7 @@
 	if not os.path.isfile(args.bam):
 		print ("error,  could not find the bam file")
 		quit()
-	if not os.path.isfile("{}/bin/TIDDIT".format(args.wd)):
+	if not os.path.isfile("{}/TIDDIT".format(args.wd)):
 		print ("error,  could not find the TIDDIT executable file, try rerun the INSTALL.sh script")
 		quit()
 	os.system(command)
