diff -u -r a/edittag/__init__.py b/edittag/__init__.py
--- a/edittag/__init__.py	2012-07-13 07:42:31.000000000 +0200
+++ b/edittag/__init__.py	2016-11-25 19:12:21.694330464 +0100
@@ -1 +1 @@
-from tests import test
+from edittag.tests import test_c_levenshtein, test_levenshtein
\ No newline at end of file
diff -u -r a/edittag/tests/test_c_levenshtein.py b/edittag/tests/test_c_levenshtein.py
--- a/edittag/tests/test_c_levenshtein.py	2012-07-13 07:42:31.000000000 +0200
+++ b/edittag/tests/test_c_levenshtein.py	2016-12-04 23:47:44.477127611 +0100
@@ -9,8 +9,8 @@
 """
 
 import unittest
-from Levenshtein import distance
-from Levenshtein import hamming
+from edittag.levenshtein import distance
+from edittag.levenshtein import hamming
 
 import pdb
 
diff -u -r a/edittag/tests/test_levenshtein.py b/edittag/tests/test_levenshtein.py
--- a/edittag/tests/test_levenshtein.py	2012-07-13 07:42:31.000000000 +0200
+++ b/edittag/tests/test_levenshtein.py	2016-12-04 23:49:06.364915487 +0100
@@ -9,7 +9,8 @@
 """
 
 import unittest
-from edittag.levenshtein import *
+from edittag.levenshtein import distance
+from edittag.levenshtein import hamming
 
 import pdb
 
diff -u -r a/setup.py b/setup.py
--- a/setup.py	2012-07-13 07:42:31.000000000 +0200
+++ b/setup.py	2016-11-25 17:22:00.016779580 +0100
@@ -1,6 +1,4 @@
-import distribute_setup
-distribute_setup.use_setuptools()
-from setuptools import setup, find_packages
+from setuptools import setup
 
 
 if __name__ == '__main__':
