From 683c9eb1c2bd6b6b3df5853daf597e9e26e9e579 Mon Sep 17 00:00:00 2001
From: Ilya Shlyakhter <ilya@broadinstitute.org>
Date: Tue, 17 Jul 2018 13:30:21 -0400
Subject: [PATCH] fixes for conda

---
 makefile | 24 ++++++++++--------------
 1 file changed, 10 insertions(+), 14 deletions(-)

diff --git a/makefile b/makefile
index bce2834..d36e48d 100644
--- a/makefile
+++ b/makefile
@@ -1,5 +1,5 @@
 all: kmc
-	
+
 KMC_BIN_DIR = bin
 KMC_MAIN_DIR = kmer_counter
 KMC_API_DIR = kmc_api
@@ -7,11 +7,11 @@ KMC_DUMP_DIR = kmc_dump
 KMC_TOOLS_DIR = kmc_tools
 
 CC 	= g++
-CFLAGS	= -Wall -O3 -m64 -static -Wl,--whole-archive -lpthread -Wl,--no-whole-archive -std=c++11 
-CLINK	= -lm -static -O3 -Wl,--whole-archive -lpthread -Wl,--no-whole-archive -std=c++11 
+CFLAGS	= -Wall -O3 -m64 -Wl,--whole-archive -pthread -Wl,--no-whole-archive -std=c++11 
+CLINK	= -lm -O3 -Wl,--whole-archive -pthread -Wl,--no-whole-archive -std=c++11 
 
-KMC_TOOLS_CFLAGS	= -Wall -O3 -m64 -static -Wl,--whole-archive -lpthread -Wl,--no-whole-archive -std=c++14
-KMC_TOOLS_CLINK	= -lm -static -O3 -Wl,--whole-archive -lpthread -Wl,--no-whole-archive -std=c++14
+KMC_TOOLS_CFLAGS	= -Wall -O3 -m64 -Wl,--whole-archive -pthread -Wl,--no-whole-archive -std=c++14
+KMC_TOOLS_CLINK	= -lm -O3 -Wl,--whole-archive -pthread -Wl,--no-whole-archive -std=c++14
 
 DISABLE_ASMLIB = false
 
@@ -36,9 +36,7 @@ $(KMC_MAIN_DIR)/raduls_sse41.o \
 $(KMC_MAIN_DIR)/raduls_avx2.o \
 $(KMC_MAIN_DIR)/raduls_avx.o 
 
-KMC_LIBS = \
-$(KMC_MAIN_DIR)/libs/libz.a \
-$(KMC_MAIN_DIR)/libs/libbz2.a
+KMC_LIBS = -lz -lbz2
 
 KMC_DUMP_OBJS = \
 $(KMC_DUMP_DIR)/nc_utils.o \
@@ -61,9 +59,7 @@ $(KMC_TOOLS_DIR)/fastq_reader.o \
 $(KMC_TOOLS_DIR)/fastq_writer.o \
 $(KMC_TOOLS_DIR)/percent_progress.o
 
-KMC_TOOLS_LIBS = \
-$(KMC_TOOLS_DIR)/libs/libz.a \
-$(KMC_TOOLS_DIR)/libs/libbz2.a 
+KMC_TOOLS_LIBS = -lz -lbz2
 
 ifeq ($(DISABLE_ASMLIB),true)
 	CFLAGS += -DDISABLE_ASMLIB
@@ -91,18 +87,18 @@ $(KMC_MAIN_DIR)/raduls_avx2.o: $(KMC_MAIN_DIR)/raduls_avx2.cpp
 	$(CC) $(CFLAGS) -mavx2 -mfma -fabi-version=0 -c $< -o $@
 $(KMC_MAIN_DIR)/instrset_detect.o: $(KMC_MAIN_DIR)/libs/vectorclass/instrset_detect.cpp
 	$(CC) $(CFLAGS) -c $< -o $@
-	
+
 kmc: $(KMC_OBJS) $(RADULS_OBJS) $(KMC_MAIN_DIR)/instrset_detect.o 
 	-mkdir -p $(KMC_BIN_DIR)
 	$(CC) $(CLINK) -o $(KMC_BIN_DIR)/$@ $^ $(KMC_LIBS)
 kmc_dump: $(KMC_DUMP_OBJS) $(KMC_API_OBJS)
 	-mkdir -p $(KMC_BIN_DIR)
 	$(CC) $(CLINK) -o $(KMC_BIN_DIR)/$@ $^
-	
+
 kmc_tools: $(KMC_TOOLS_OBJS) $(KMC_API_OBJS)
 	-mkdir -p $(KMC_BIN_DIR)
 	$(CC) $(KMC_TOOLS_CLINK) -o $(KMC_BIN_DIR)/$@ $^ $(KMC_TOOLS_LIBS)
-	
+
 clean:
 	-rm $(KMC_MAIN_DIR)/*.o
 	-rm $(KMC_API_DIR)/*.o
-- 
2.14.4

