diff -ru phyml-3.3.20170530.orig/configure.ac phyml-3.3.20170530/configure.ac
--- phyml-3.3.20170530.orig/configure.ac	2017-05-23 00:38:26.000000000 -0600
+++ phyml-3.3.20170530/configure.ac	2017-06-22 13:30:43.000000000 -0600
@@ -1,7 +1,8 @@
 #                                               -*- Autoconf -*-
 # Process this file with autoconf to produce a configure script.
 
-AC_INIT([PhyML],esyscmd([sh -c "date \"+%Y%m%d\" | tr -d '\n'"]),[s.guindon@auckland.ac.nz])
+dnl AC_INIT([PhyML],esyscmd([sh -c "date \"+%Y%m%d\" | tr -d '\n'"]),[s.guindon@auckland.ac.nz])
+AC_INIT([PhyML],[3.2.0],[s.guindon@auckland.ac.nz])
 AM_INIT_AUTOMAKE([foreign])
 m4_ifdef([AM_SILENT_RULES], [AM_SILENT_RULES])
 AC_CONFIG_MACRO_DIR([m4])
