diff --git a/src/bwa_sam_converter.pl b/src/bwa_sam_converter.pl
index 1689596..defb973 100755
--- a/src/bwa_sam_converter.pl
+++ b/src/bwa_sam_converter.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 bwa-SAM-converter perl script
 # Copyright (C) 2009 - 2012  Sebastian Mackowiak
diff --git a/src/clip_adapters.pl b/src/clip_adapters.pl
index 4a99098..29cd73d 100755
--- a/src/clip_adapters.pl
+++ b/src/clip_adapters.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 clip-adapters perl script
 # Copyright (C) 2008 - 2012  Marc Friedländer
diff --git a/src/collapse_reads_md.pl b/src/collapse_reads_md.pl
index cb49c89..8c0991a 100755
--- a/src/collapse_reads_md.pl
+++ b/src/collapse_reads_md.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 collapse-reads-md perl script
 # Copyright (C) 2008 - 2011  Marc Friedländer
diff --git a/src/convert_bowtie_output.pl b/src/convert_bowtie_output.pl
index 4e8af3c..0ac1f5d 100755
--- a/src/convert_bowtie_output.pl
+++ b/src/convert_bowtie_output.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 convert-bowtie-output perl script
 # Copyright (C) 2009 - 2011  Sebastian Mackowiak
diff --git a/src/excise_precursors.pl b/src/excise_precursors.pl
index 0da2e39..0bed63d 100755
--- a/src/excise_precursors.pl
+++ b/src/excise_precursors.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 excise-precursors perl script
 # Copyright (C) 2008 - 2011  Marc Friedländer
diff --git a/src/excise_precursors_iterative_final.pl b/src/excise_precursors_iterative_final.pl
index 49c6504..5a6c66f 100755
--- a/src/excise_precursors_iterative_final.pl
+++ b/src/excise_precursors_iterative_final.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 excise-precursors-iterative-final perl script
 # Copyright (C) 2008 - 2012  Marc Friedländer
diff --git a/src/extract_miRNAs.pl b/src/extract_miRNAs.pl
index 43b5694..7eee060 100755
--- a/src/extract_miRNAs.pl
+++ b/src/extract_miRNAs.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 extract-miRNAs perl script
 # Copyright (C) 2009 - 2011  Sebastian Mackowiak
diff --git a/src/fastaparse.pl b/src/fastaparse.pl
index f6a0cc2..2f68ddf 100755
--- a/src/fastaparse.pl
+++ b/src/fastaparse.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 FASTA-parse perl script
 # Copyright (C) 2008 - 2012  Marc Friedländer
diff --git a/src/fastaselect.pl b/src/fastaselect.pl
index af92c9d..16c096a 100755
--- a/src/fastaselect.pl
+++ b/src/fastaselect.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 FASTA-select perl script
 # Copyright (C) 2008 - 2011  Marc Friedländer
diff --git a/src/fastq2fasta.pl b/src/fastq2fasta.pl
index dea86c8..bee2331 100755
--- a/src/fastq2fasta.pl
+++ b/src/fastq2fasta.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 FASTQ-to-FASTA perl script
 # Copyright (C) 2008 - 2011  Marc Friedländer
diff --git a/src/find_read_count.pl b/src/find_read_count.pl
index 630c2ec..173c29e 100755
--- a/src/find_read_count.pl
+++ b/src/find_read_count.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 find-read-count perl script
 # Copyright (C) 2008 - 2011  Marc Friedländer
diff --git a/src/geo2fasta.pl b/src/geo2fasta.pl
index 8908a3b..e473333 100755
--- a/src/geo2fasta.pl
+++ b/src/geo2fasta.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 GEO-to-FASTA perl script
 # Copyright (C) 2008 - 2011  Marc Friedländer
diff --git a/src/get_mirdeep2_precursors.pl b/src/get_mirdeep2_precursors.pl
index 9962a45..cdb646f 100755
--- a/src/get_mirdeep2_precursors.pl
+++ b/src/get_mirdeep2_precursors.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 get-precursors perl script
 # Copyright (C) 2012 - 2014  Sebastian Mackowiak
diff --git a/src/illumina_to_fasta.pl b/src/illumina_to_fasta.pl
index 7979e59..7d1c4f2 100755
--- a/src/illumina_to_fasta.pl
+++ b/src/illumina_to_fasta.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 Illumina-to-FASTA perl script
 # Copyright (C) 2008 - 2011  Marc Friedländer
diff --git a/src/make_html.pl b/src/make_html.pl
index 02e40ca..51f9b0e 100755
--- a/src/make_html.pl
+++ b/src/make_html.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 make-HTML script
 # Copyright (C) 2009 - 2012, 2014 - 2015, 2018 - 2019 Sebastian Mackowiak
diff --git a/src/make_html2.pl b/src/make_html2.pl
index b97e86e..24c30db 100755
--- a/src/make_html2.pl
+++ b/src/make_html2.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 make-HTML2 perl script
 # Copyright (C) 2009 - 2012, 2019  Sebastian Mackowiak
diff --git a/src/mapper.pl b/src/mapper.pl
index b53be55..6a5b419 100755
--- a/src/mapper.pl
+++ b/src/mapper.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 mapper perl script
 # Copyright (C) 2008 - 2012  Marc Friedländer
diff --git a/src/miRDeep2.pl b/src/miRDeep2.pl
index af39ea9..8f5f500 100755
--- a/src/miRDeep2.pl
+++ b/src/miRDeep2.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 perl script
 # Copyright (C) 2008 - 2012  Marc Friedländer
@@ -29,9 +29,6 @@ use Term::ANSIColor;
 use Cwd qw(abs_path);
 
 
-##
-check_install();
-##
 ## generate log file for run, all output will be printed to it
 
 my $version="2.0.1.3";
@@ -910,15 +907,15 @@ sub output_results{
     if($file_mature_ref_this_species !~ /none/i){
 
         if($options{'q'}){
-            $line="make_html.pl -f $dir/output.mrd -k $dir_tmp/$file_mature_ref_this_species -p $dir_tmp/precursors.coords -s $dir/survey.csv -c -e -q $options{'q'} -x $xopt -r ${scripts}/../Rfam_for_miRDeep.fa -v $sc -y $time $sort_by_sample $OE";
+            $line="make_html.pl -f $dir/output.mrd -k $dir_tmp/$file_mature_ref_this_species -p $dir_tmp/precursors.coords -s $dir/survey.csv -c -e -q $options{'q'} -x $xopt -r ${scripts}/Rfam_for_miRDeep.fa -v $sc -y $time $sort_by_sample $OE";
         }else{
-            $line="make_html.pl -f $dir/output.mrd -k $dir_tmp/$file_mature_ref_this_species -p $dir_tmp/precursors.coords -s $dir/survey.csv -c -e -r ${scripts}/../Rfam_for_miRDeep.fa -v $sc -y $time  $sort_by_sample $OE";
+            $line="make_html.pl -f $dir/output.mrd -k $dir_tmp/$file_mature_ref_this_species -p $dir_tmp/precursors.coords -s $dir/survey.csv -c -e -r ${scripts}/Rfam_for_miRDeep.fa -v $sc -y $time  $sort_by_sample $OE";
         }
     }else{
         if($options{'q'}){
-            $line="make_html.pl -f $dir/output.mrd -p $dir_tmp/precursors.coords -s $dir/survey.csv -c -e -q $options{'q'}  -x $xopt -r ${scripts}/../Rfam_for_miRDeep.fa -v $sc -y $time $sort_by_sample $OE";
+            $line="make_html.pl -f $dir/output.mrd -p $dir_tmp/precursors.coords -s $dir/survey.csv -c -e -q $options{'q'}  -x $xopt -r ${scripts}/Rfam_for_miRDeep.fa -v $sc -y $time $sort_by_sample $OE";
         }else{
-            $line="make_html.pl -f $dir/output.mrd -p $dir_tmp/precursors.coords -v $sc -s $dir/survey.csv -c -e -r ${scripts}/../Rfam_for_miRDeep.fa -y $time $sort_by_sample $OE";
+            $line="make_html.pl -f $dir/output.mrd -p $dir_tmp/precursors.coords -v $sc -s $dir/survey.csv -c -e -r ${scripts}/Rfam_for_miRDeep.fa -y $time $sort_by_sample $OE";
         }
     }
 
@@ -958,7 +955,7 @@ sub test_first_argument{
 
     if($ARGV[0] eq '-u'){system("make_html.pl -u -y 1"); exit;}
 
-    if($ARGV[0] eq '-h' or $ARGV[0] eq '--help'){die $usage;}
+    if($ARGV[0] eq '-h' or $ARGV[0] eq '--help'){print $usage; exit;}
 
     return;
 }
@@ -1046,7 +1043,7 @@ If this did not help please restart youer workstation.\n\n";
 	$ret = checkBIN("perl -e \'use PDF::API2; print \"installed\";\'","installed");
 	die "Error: \tPerl PDF::API2 package not found\nCheck if the perl PDF::API2 package is correctly installed and all Pathes were set correctly.\n$stdm" if($ret);
 
-	if(not -f "$scripts/../Rfam_for_miRDeep.fa"){
+	if(not -f "$scripts/Rfam_for_miRDeep.fa"){
 		die "Error:\t Rfam_for_miRDeep.fa not found in your miRDeep2 scripts directory\nPlease copy this file from the miRDeep2 archive to your miRDeep2 directory\n\n";
 	}
 
@@ -1118,19 +1115,3 @@ sub get_longest_id{
 	return $l;
 }
 
-
-sub check_install{
-	my $a=`which miRDeep2.pl`;
-	chomp $a;
-	$a=abs_path($a);
-	my ( $name0, $bn, $extension0 ) = fileparse ( $a, '\..*' );
-	if(not -f "$bn/../install_successful"){
-		die "Please run the install.pl script first before using the miRDeep2 package
-		The install script is located in ",substr($bn,0,length($bn)-4)," so just do
-
-		cd ",substr($bn,0,length($bn)-4),
-		"\nperl install.pl
-
-		";
-	}
-}
diff --git a/src/miRDeep2_core_algorithm.pl b/src/miRDeep2_core_algorithm.pl
index ff0f3a6..b242b0f 100755
--- a/src/miRDeep2_core_algorithm.pl
+++ b/src/miRDeep2_core_algorithm.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 core-algorithm perl script
 # Copyright (C) 2008 - 2012  Marc Friedländer
diff --git a/src/mirdeep2bed.pl b/src/mirdeep2bed.pl
index f6696a4..7485e96 100755
--- a/src/mirdeep2bed.pl
+++ b/src/mirdeep2bed.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 mirdeep-to-BED perl script
 # Copyright (C) 2009 - 2012  Sebastian Mackowiak
diff --git a/src/parse_mappings.pl b/src/parse_mappings.pl
index 15be453..de70fc1 100755
--- a/src/parse_mappings.pl
+++ b/src/parse_mappings.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 parse-mappings perl script
 # Copyright (C) 2008 - 2011  Marc Friedländer
diff --git a/src/perform_controls.pl b/src/perform_controls.pl
index 0de442d..a3cfce4 100755
--- a/src/perform_controls.pl
+++ b/src/perform_controls.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 perform-controls perl script
 # Copyright (C) 2008 - 2011  Marc Friedländer
diff --git a/src/permute_structure.pl b/src/permute_structure.pl
index a0bc151..f19e1af 100755
--- a/src/permute_structure.pl
+++ b/src/permute_structure.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 permute-structure perl script
 # Copyright (C) 2008 - 2011  Marc Friedländer
diff --git a/src/prepare_signature.pl b/src/prepare_signature.pl
index 7dc52ad..16b8eeb 100755
--- a/src/prepare_signature.pl
+++ b/src/prepare_signature.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 prepare-signature perl script
 # Copyright (C) 2008 - 2012  Marc Friedländer
diff --git a/src/quantifier.pl b/src/quantifier.pl
index bac73c5..058aed1 100755
--- a/src/quantifier.pl
+++ b/src/quantifier.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # quantifier module from the miRDeep2 software package
 # Copyright (C) 2009 - 2012, 2014 - 2017, 2019  Sebastian Mackowiak
diff --git a/src/remove_white_space_in_id.pl b/src/remove_white_space_in_id.pl
index e663a36..c0e1de0 100755
--- a/src/remove_white_space_in_id.pl
+++ b/src/remove_white_space_in_id.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 remove-white-space-in-ID perl script
 # Copyright (C) 2009 - 2011  Sebastian Mackowiak
diff --git a/src/rna2dna.pl b/src/rna2dna.pl
index 909da5c..77f06f5 100755
--- a/src/rna2dna.pl
+++ b/src/rna2dna.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 RNA-to-DNA perl script
 # Copyright (C) 2008 - 2011  Marc Friedländer
diff --git a/src/samFLAGinfo.pl b/src/samFLAGinfo.pl
index d3a5b48..7f9d963 100755
--- a/src/samFLAGinfo.pl
+++ b/src/samFLAGinfo.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 SAM-flag-info perl script
 # Copyright (C) 2009 - 2011  Sebastian Mackowiak
diff --git a/src/sam_reads_collapse.pl b/src/sam_reads_collapse.pl
index c909b74..1c50d1e 100755
--- a/src/sam_reads_collapse.pl
+++ b/src/sam_reads_collapse.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 SAM-reads-collapse perl script
 # Copyright (C) 2009 - 2011  Sebastian Mackowiak
diff --git a/src/sanity_check_genome.pl b/src/sanity_check_genome.pl
index da78321..02ba333 100755
--- a/src/sanity_check_genome.pl
+++ b/src/sanity_check_genome.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl -W
+#!/usr/bin/env perl
 
 # miRDeep2 sanity-check-genome perl script
 # Copyright (C) 2008 - 2011  Marc Friedländer
diff --git a/src/sanity_check_mapping_file.pl b/src/sanity_check_mapping_file.pl
index 4a8009e..f6996c3 100755
--- a/src/sanity_check_mapping_file.pl
+++ b/src/sanity_check_mapping_file.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl -W
+#!/usr/bin/env perl
 
 # miRDeep2 sanity-check-mapping-file perl script
 # Copyright (C) 2008 - 2011  Marc Friedländer
diff --git a/src/sanity_check_mature_ref.pl b/src/sanity_check_mature_ref.pl
index dcea89f..d5c3f8b 100755
--- a/src/sanity_check_mature_ref.pl
+++ b/src/sanity_check_mature_ref.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl -W
+#!/usr/bin/env perl
 
 # miRDeep2 sanity-check-mature-ref perl script
 # Copyright (C) 2008 - 2011  Marc Friedländer
diff --git a/src/sanity_check_reads_ready_file.pl b/src/sanity_check_reads_ready_file.pl
index 99d62c2..bb3598c 100755
--- a/src/sanity_check_reads_ready_file.pl
+++ b/src/sanity_check_reads_ready_file.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl -W
+#!/usr/bin/env perl
 
 # miRDeep2 sanity-check-reads-ready-file perl script
 # Copyright (C) 2008 - 2011  Marc Friedländer
diff --git a/src/select_for_randfold.pl b/src/select_for_randfold.pl
index 2568dcf..1d635a5 100755
--- a/src/select_for_randfold.pl
+++ b/src/select_for_randfold.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 select-for-randfold perl script
 # Copyright (C) 2008 - 2011  Marc Friedländer
diff --git a/src/survey.pl b/src/survey.pl
index 0f7a8d8..33fd695 100755
--- a/src/survey.pl
+++ b/src/survey.pl
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 survey perl script
 # Copyright (C) 2008 - 2011  Marc Friedländer
-- 
2.17.1

