diff -u -r a/src/bwa_sam_converter.pl b/src/bwa_sam_converter.pl
--- a/src/bwa_sam_converter.pl	2017-05-12 16:48:56.000000000 +0200
+++ b/src/bwa_sam_converter.pl	2017-05-12 16:49:06.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 # miRDeep2 bwa-SAM-converter perl script
 # Copyright (C) 2009 - 2012  Sebastian Mackowiak
diff -u -r a/src/clip_adapters.pl b/src/clip_adapters.pl
--- a/src/clip_adapters.pl	2017-05-12 16:48:52.000000000 +0200
+++ b/src/clip_adapters.pl	2017-05-12 16:49:00.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/collapse_reads_md.pl b/src/collapse_reads_md.pl
--- a/src/collapse_reads_md.pl	2017-05-12 16:48:52.000000000 +0200
+++ b/src/collapse_reads_md.pl	2017-05-12 16:49:02.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/convert_bowtie_output.pl b/src/convert_bowtie_output.pl
--- a/src/convert_bowtie_output.pl	2017-05-12 16:48:52.000000000 +0200
+++ b/src/convert_bowtie_output.pl	2017-05-12 16:49:00.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use strict;
 
diff -u -r a/src/excise_precursors_iterative_final.pl b/src/excise_precursors_iterative_final.pl
--- a/src/excise_precursors_iterative_final.pl	2017-05-12 16:48:54.000000000 +0200
+++ b/src/excise_precursors_iterative_final.pl	2017-05-12 16:49:02.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/excise_precursors.pl b/src/excise_precursors.pl
--- a/src/excise_precursors.pl	2017-05-12 16:48:54.000000000 +0200
+++ b/src/excise_precursors.pl	2017-05-12 16:49:04.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/extract_miRNAs.pl b/src/extract_miRNAs.pl
--- a/src/extract_miRNAs.pl	2017-05-12 16:48:56.000000000 +0200
+++ b/src/extract_miRNAs.pl	2017-05-12 16:49:06.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use strict;
 
diff -u -r a/src/fastaparse.pl b/src/fastaparse.pl
--- a/src/fastaparse.pl	2017-05-12 16:48:54.000000000 +0200
+++ b/src/fastaparse.pl	2017-05-12 16:49:06.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/fastaselect.pl b/src/fastaselect.pl
--- a/src/fastaselect.pl	2017-05-12 16:48:58.000000000 +0200
+++ b/src/fastaselect.pl	2017-05-12 16:49:06.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/fastq2fasta.pl b/src/fastq2fasta.pl
--- a/src/fastq2fasta.pl	2017-05-12 16:48:52.000000000 +0200
+++ b/src/fastq2fasta.pl	2017-05-12 16:49:00.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use strict;
 
diff -u -r a/src/find_read_count.pl b/src/find_read_count.pl
--- a/src/find_read_count.pl	2017-05-12 16:48:52.000000000 +0200
+++ b/src/find_read_count.pl	2017-05-12 16:49:02.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/geo2fasta.pl b/src/geo2fasta.pl
--- a/src/geo2fasta.pl	2017-05-12 16:48:54.000000000 +0200
+++ b/src/geo2fasta.pl	2017-05-12 16:49:02.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/get_mirdeep2_precursors.pl b/src/get_mirdeep2_precursors.pl
--- a/src/get_mirdeep2_precursors.pl	2017-05-12 16:48:56.000000000 +0200
+++ b/src/get_mirdeep2_precursors.pl	2017-05-12 16:49:08.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 #######################################################################
 #
diff -u -r a/src/illumina_to_fasta.pl b/src/illumina_to_fasta.pl
--- a/src/illumina_to_fasta.pl	2017-05-12 16:48:58.000000000 +0200
+++ b/src/illumina_to_fasta.pl	2017-05-12 16:49:08.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/make_html2.pl b/src/make_html2.pl
--- a/src/make_html2.pl	2017-05-12 16:48:54.000000000 +0200
+++ b/src/make_html2.pl	2017-05-12 16:49:06.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 
 use strict;
diff -u -r a/src/make_html.pl b/src/make_html.pl
--- a/src/make_html.pl	2017-05-12 16:48:52.000000000 +0200
+++ b/src/make_html.pl	2017-05-12 16:49:00.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 
 use strict;
diff -u -r a/src/mapper.pl b/src/mapper.pl
--- a/src/mapper.pl	2017-05-12 16:48:54.000000000 +0200
+++ b/src/mapper.pl	2017-05-12 16:49:02.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 ## last update 21/04/2016
 
diff -u -r a/src/mirdeep2bed.pl b/src/mirdeep2bed.pl
--- a/src/mirdeep2bed.pl	2017-05-12 16:48:50.000000000 +0200
+++ b/src/mirdeep2bed.pl	2017-05-12 16:49:00.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl 
+#!/usr/bin/env perl
 
 use strict;
 
diff -u -r a/src/miRDeep2_core_algorithm.pl b/src/miRDeep2_core_algorithm.pl
--- a/src/miRDeep2_core_algorithm.pl	2017-05-12 16:48:54.000000000 +0200
+++ b/src/miRDeep2_core_algorithm.pl	2017-05-12 16:49:04.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/miRDeep2.pl b/src/miRDeep2.pl
--- a/src/miRDeep2.pl	2017-05-12 16:48:52.000000000 +0200
+++ b/src/miRDeep2.pl	2017-06-30 04:58:48.878320247 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
@@ -10,9 +10,6 @@
 use File::Basename;
 use Term::ANSIColor;
 
-##
-check_install();
-##
 ## generate log file for run, all output will be printed to it
 
 my $version="2.0.0.8";
@@ -891,15 +888,15 @@
     if($file_mature_ref_this_species !~ /none/i){
 
         if($options{'q'}){
-            $line="make_html.pl -f $dir/output.mrd -k $dir_tmp/$file_mature_ref_this_species -p $dir_tmp/precursors.coords -s $dir/survey.csv -c -e -q $options{'q'} -x $xopt -r ${scripts}/../Rfam_for_miRDeep.fa -v $sc -y $time $sort_by_sample $OE";
+            $line="make_html.pl -f $dir/output.mrd -k $dir_tmp/$file_mature_ref_this_species -p $dir_tmp/precursors.coords -s $dir/survey.csv -c -e -q $options{'q'} -x $xopt -r ${scripts}/Rfam_for_miRDeep.fa -v $sc -y $time $sort_by_sample $OE";
         }else{
-            $line="make_html.pl -f $dir/output.mrd -k $dir_tmp/$file_mature_ref_this_species -p $dir_tmp/precursors.coords -s $dir/survey.csv -c -e -r ${scripts}/../Rfam_for_miRDeep.fa -v $sc -y $time  $sort_by_sample $OE";
+            $line="make_html.pl -f $dir/output.mrd -k $dir_tmp/$file_mature_ref_this_species -p $dir_tmp/precursors.coords -s $dir/survey.csv -c -e -r ${scripts}/Rfam_for_miRDeep.fa -v $sc -y $time  $sort_by_sample $OE";
         }
     }else{
         if($options{'q'}){
-            $line="make_html.pl -f $dir/output.mrd -p $dir_tmp/precursors.coords -s $dir/survey.csv -c -e -q $options{'q'}  -x $xopt -r ${scripts}/../Rfam_for_miRDeep.fa -v $sc -y $time $sort_by_sample $OE";
+            $line="make_html.pl -f $dir/output.mrd -p $dir_tmp/precursors.coords -s $dir/survey.csv -c -e -q $options{'q'}  -x $xopt -r ${scripts}/Rfam_for_miRDeep.fa -v $sc -y $time $sort_by_sample $OE";
         }else{
-            $line="make_html.pl -f $dir/output.mrd -p $dir_tmp/precursors.coords -v $sc -s $dir/survey.csv -c -e -r ${scripts}/../Rfam_for_miRDeep.fa -y $time $sort_by_sample $OE";
+            $line="make_html.pl -f $dir/output.mrd -p $dir_tmp/precursors.coords -v $sc -s $dir/survey.csv -c -e -r ${scripts}/Rfam_for_miRDeep.fa -y $time $sort_by_sample $OE";
         }
     }
 
@@ -939,7 +936,7 @@
 
     if($ARGV[0] eq '-u'){system("make_html.pl -u -y 1"); exit;}
 
-    if($ARGV[0] eq '-h' or $ARGV[0] eq '--help'){die $usage;}
+    if($ARGV[0] eq '-h' or $ARGV[0] eq '--help'){print $usage; exit;}
 
     return;
 }
@@ -1027,7 +1024,7 @@
 	$ret = checkBIN("perl -e \'use PDF::API2; print \"installed\";\'","installed");
 	die "Error: \tPerl PDF::API2 package not found\nCheck if the perl PDF::API2 package is correctly installed and all Pathes were set correctly.\n$stdm" if($ret);
 
-	if(not -f "$scripts/../Rfam_for_miRDeep.fa"){
+	if(not -f "$scripts/Rfam_for_miRDeep.fa"){
 		die "Error:\t Rfam_for_miRDeep.fa not found in your miRDeep2 scripts directory\nPlease copy this file from the miRDeep2 archive to your miRDeep2 directory\n\n";
 	}
 
@@ -1118,19 +1115,3 @@
 	return $l;
 }
 
-
-sub check_install{
-	my $a=`which miRDeep2.pl`;
-	chomp $a;
-	$a=abs_path($a);
-	my ( $name0, $bn, $extension0 ) = fileparse ( $a, '\..*' );
-	if(not -f "$bn/../install_successful"){
-		die "Please run the install.pl script first before using the miRDeep2 package
-		The install script is located in ",substr($bn,0,length($bn)-4)," so just do
-
-		cd ",substr($bn,0,length($bn)-4),
-		"\nperl install.pl
-
-		";
-	}
-}
diff -u -r a/src/parse_mappings.pl b/src/parse_mappings.pl
--- a/src/parse_mappings.pl	2017-05-12 16:48:58.000000000 +0200
+++ b/src/parse_mappings.pl	2017-05-12 16:49:04.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/perform_controls.pl b/src/perform_controls.pl
--- a/src/perform_controls.pl	2017-05-12 16:48:52.000000000 +0200
+++ b/src/perform_controls.pl	2017-05-12 16:49:02.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/permute_structure.pl b/src/permute_structure.pl
--- a/src/permute_structure.pl	2017-05-12 16:48:58.000000000 +0200
+++ b/src/permute_structure.pl	2017-05-12 16:49:08.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/prepare_signature.pl b/src/prepare_signature.pl
--- a/src/prepare_signature.pl	2017-05-12 16:48:56.000000000 +0200
+++ b/src/prepare_signature.pl	2017-05-12 16:49:06.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/quantifier.pl b/src/quantifier.pl
--- a/src/quantifier.pl	2017-05-12 16:48:56.000000000 +0200
+++ b/src/quantifier.pl	2017-05-12 16:49:08.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 ######################
 
diff -u -r a/src/remove_white_space_in_id.pl b/src/remove_white_space_in_id.pl
--- a/src/remove_white_space_in_id.pl	2017-05-12 16:48:52.000000000 +0200
+++ b/src/remove_white_space_in_id.pl	2017-05-12 16:49:02.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/rna2dna.pl b/src/rna2dna.pl
--- a/src/rna2dna.pl	2017-05-12 16:48:58.000000000 +0200
+++ b/src/rna2dna.pl	2017-05-12 16:49:04.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/samFLAGinfo.pl b/src/samFLAGinfo.pl
--- a/src/samFLAGinfo.pl	2017-05-12 16:48:52.000000000 +0200
+++ b/src/samFLAGinfo.pl	2017-05-12 16:49:02.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 
 ## convert decimal number to binary and then to hexadecimal to get read FLAG identification 
diff -u -r a/src/sam_reads_collapse.pl b/src/sam_reads_collapse.pl
--- a/src/sam_reads_collapse.pl	2017-05-12 16:48:58.000000000 +0200
+++ b/src/sam_reads_collapse.pl	2017-05-12 16:49:06.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 
 if(not $ARGV[1]){
diff -u -r a/src/sanity_check_genome.pl b/src/sanity_check_genome.pl
--- a/src/sanity_check_genome.pl	2017-05-12 16:48:54.000000000 +0200
+++ b/src/sanity_check_genome.pl	2017-07-05 16:51:38.380285114 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl -W
+#!/usr/bin/env perl
 
 use strict;
 
diff -u -r a/src/sanity_check_mapping_file.pl b/src/sanity_check_mapping_file.pl
--- a/src/sanity_check_mapping_file.pl	2017-05-12 16:48:52.000000000 +0200
+++ b/src/sanity_check_mapping_file.pl	2017-07-05 16:51:47.220467153 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl -W
+#!/usr/bin/env perl
 
 use strict;
 my $rid;
diff -u -r a/src/sanity_check_mature_ref.pl b/src/sanity_check_mature_ref.pl
--- a/src/sanity_check_mature_ref.pl	2017-05-12 16:48:54.000000000 +0200
+++ b/src/sanity_check_mature_ref.pl	2017-07-05 16:51:58.228692524 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl -W
+#!/usr/bin/env perl
 
 use strict;
 
diff -u -r a/src/sanity_check_reads_ready_file.pl b/src/sanity_check_reads_ready_file.pl
--- a/src/sanity_check_reads_ready_file.pl	2017-05-12 16:48:52.000000000 +0200
+++ b/src/sanity_check_reads_ready_file.pl	2017-07-05 16:52:04.624822815 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl -W
+#!/usr/bin/env perl
 
 use strict;
 
diff -u -r a/src/select_for_randfold.pl b/src/select_for_randfold.pl
--- a/src/select_for_randfold.pl	2017-05-12 16:48:58.000000000 +0200
+++ b/src/select_for_randfold.pl	2017-05-12 16:49:06.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
diff -u -r a/src/survey.pl b/src/survey.pl
--- a/src/survey.pl	2017-05-12 16:48:54.000000000 +0200
+++ b/src/survey.pl	2017-05-12 16:49:04.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 use warnings;
 use strict;
