diff --git a/CMakeLists.txt b/CMakeLists.txt
index a3eebd4..c5ca35e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -28,12 +28,13 @@ ExternalProject_Add(htslib
 
 # Setup boost user config so that it uses the same c++ compiler as other components:
 if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
-	set (BOOST_COMPILER_TAG "gcc")
+	set(BOOST_COMPILER_TAG "gcc")
+	set(B2_OPTIONS ${B2_OPTIONS} "toolset=gcc")
 elseif(CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
-	set (BOOST_COMPILER_TAG "clang")
+	set(BOOST_COMPILER_TAG "clang")
 	set(B2_OPTIONS ${B2_OPTIONS} "toolset=clang")
 elseif(CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang")
-	set (BOOST_COMPILER_TAG "darwin")
+	set(BOOST_COMPILER_TAG "darwin")
 	set(B2_OPTIONS ${B2_OPTIONS} "toolset=clang")
 endif()
 
@@ -45,10 +46,10 @@ endif()
 
 ExternalProject_Add(Boost
 	BUILD_IN_SOURCE YES
-	URL https://boostorg.jfrog.io/artifactory/main/release/1.73.0/source/boost_1_73_0.tar.bz2
+	URL https://boostorg.jfrog.io/artifactory/main/release/1.80.0/source/boost_1_80_0.tar.bz2
-	PATCH_COMMAND ${BOOST_PATCH_COMMAND}
-	CONFIGURE_COMMAND ./bootstrap.sh --prefix=${installDir}
-	BUILD_COMMAND ./b2 install -j8  --with-filesystem --with-system --with-program_options link=static ${B2_OPTIONS}
+	CONFIGURE_COMMAND ./bootstrap.sh --prefix=${installDir}
+	BUILD_COMMAND ./b2 install --prefix=${installDir} --with-filesystem --with-system --with-program_options link=static ${B2_OPTIONS}
 	INSTALL_COMMAND ""
 )
 
@@ -62,7 +63,7 @@ ExternalProject_Add(spdlog
 
 
 ExternalProject_Add(googletest
-	URL https://github.com/google/googletest/archive/refs/tags/release-1.10.0.tar.gz
+	URL https://github.com/google/googletest/archive/refs/tags/release-1.12.1.tar.gz
 	CMAKE_ARGS -DCMAKE_INSTALL_PREFIX:PATH=${installDir}
 		-DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}
 		-DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
