diff --git a/CMakeLists.txt b/CMakeLists.txt
index 9500e8c..e59a19c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -28,12 +28,13 @@ ExternalProject_Add(htslib
 
 # Setup boost user config so that it uses the same c++ compiler as other components:
 if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
-	set (BOOST_COMPILER_TAG "gcc")
+	set(BOOST_COMPILER_TAG "gcc")
+	set(B2_OPTIONS ${B2_OPTIONS} "toolset=gcc")
 elseif(CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
-	set (BOOST_COMPILER_TAG "clang")
+	set(BOOST_COMPILER_TAG "clang")
 	set(B2_OPTIONS ${B2_OPTIONS} "toolset=clang")
 elseif(CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang")
-	set (BOOST_COMPILER_TAG "darwin")
+	set(BOOST_COMPILER_TAG "darwin")
 	set(B2_OPTIONS ${B2_OPTIONS} "toolset=clang")
 endif()
 
@@ -45,10 +46,9 @@ endif()
 
 ExternalProject_Add(Boost
 	BUILD_IN_SOURCE YES
-	URL https://boostorg.jfrog.io/artifactory/main/release/1.73.0/source/boost_1_73_0.tar.bz2
-	PATCH_COMMAND ${BOOST_PATCH_COMMAND}
-	CONFIGURE_COMMAND ./bootstrap.sh --prefix=${installDir}
-	BUILD_COMMAND ./b2 install -j8  --with-filesystem --with-system --with-program_options link=static ${B2_OPTIONS}
+	URL https://boostorg.jfrog.io/artifactory/main/release/1.80.0/source/boost_1_80_0.tar.bz2
+	CONFIGURE_COMMAND ./bootstrap.sh --prefix=${installDir} --with-toolset=${CPP}
+	BUILD_COMMAND ./b2 install --prefix=${installDir} --with-filesystem --with-system --with-program_options link=static ${B2_OPTIONS}
 	INSTALL_COMMAND ""
 )
 
