From 0b636c4ed21211f3f3650eaa8a8a56eb93bd5ce7 Mon Sep 17 00:00:00 2001
From: Lightning Auriga <lightning.auriga@gmail.com>
Date: Mon, 11 Sep 2023 10:01:16 -0400
Subject: [PATCH 3/4] add missing limits import

---
 .../include/graphcore/GraphCoordinates.hh                        | 1 +
 1 file changed, 1 insertion(+)

diff --git a/ehunter/thirdparty/graph-tools-master-f421f4c/include/graphcore/GraphCoordinates.hh b/ehunter/thirdparty/graph-tools-master-f421f4c/include/graphcore/GraphCoordinates.hh
index 7768312..36a4c9b 100644
--- a/ehunter/thirdparty/graph-tools-master-f421f4c/include/graphcore/GraphCoordinates.hh
+++ b/ehunter/thirdparty/graph-tools-master-f421f4c/include/graphcore/GraphCoordinates.hh
@@ -23,6 +23,7 @@
 #include "graphalign/GraphAlignment.hh"
 
 #include <memory>
+#include <limits>
 
 namespace graphtools
 {
-- 
2.39.1

