--- CMakeLists.txt.old	2022-11-04 12:04:37.119028200 +0000
+++ CMakeLists.txt	2022-11-04 13:24:39.332491500 +0000
@@ -10,6 +10,7 @@
 find_package(LibLZMA REQUIRED)
 find_package(CURL REQUIRED)
 find_package(Git REQUIRED)
+find_package(Boost REQUIRED)
 
 include(ExternalProject)
 set(installDir ${CMAKE_CURRENT_BINARY_DIR}/install)
@@ -27,29 +28,6 @@
 )
 
 
-# Setup user config to force same c++ compiler in boost
-if((CMAKE_CXX_COMPILER_ID STREQUAL "GNU") OR (CMAKE_CXX_COMPILER_ID STREQUAL "Clang"))
-    set(BOOST_UCONFIG "${CMAKE_BINARY_DIR}/user-config.jam")
-    if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
-        file(WRITE "${BOOST_UCONFIG}" "using gcc : : \"${CMAKE_CXX_COMPILER}\" ;\n")
-    elseif(CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
-        file(WRITE "${BOOST_UCONFIG}" "using clang : : \"${CMAKE_CXX_COMPILER}\" ;\n")
-        set(B2_OPTIONS ${B2_OPTIONS} "toolset=clang")
-    endif()
-    set(BOOST_PATCH_COMMAND ${CMAKE_COMMAND} -E copy ${BOOST_UCONFIG} tools/build/src/user-config.jam)
-endif()
-
-ExternalProject_Add(Boost
-	BUILD_IN_SOURCE YES
-	URL https://boostorg.jfrog.io/artifactory/main/release/1.73.0/source/boost_1_73_0.tar.bz2
-	UPDATE_COMMAND ""
-	PATCH_COMMAND ${BOOST_PATCH_COMMAND}
-	CONFIGURE_COMMAND ./bootstrap.sh --prefix=${installDir}/lib
-	BUILD_COMMAND ./b2 install -j8   --prefix=${installDir} --with-filesystem --with-system --with-program_options link=static ${B2_OPTIONS}
-	INSTALL_COMMAND ""
-)
-
-
 ExternalProject_Add(spdlog
 	GIT_REPOSITORY "https://github.com/gabime/spdlog.git"
 	GIT_TAG "v1.6.1"
@@ -95,4 +73,4 @@
 
 
 ExternalProject_Add_StepDependencies(ehunter configure
-	Boost spdlog htslib googletest abseil)
+	spdlog htslib googletest abseil)
