--- Cargo.toml.orig	2024-10-07 11:42:31.332000000 +0000
+++ Cargo.toml	2024-10-07 11:40:19.716000000 +0000
@@ -12,7 +12,7 @@
 clap = { version = "4.0.32", features = ["derive"] }
 env_logger = "0.10"
 log = "0.4"
-rust-htslib = { version = "0.37.0", default-features = false, features = ["static"] }
+rust-htslib = { version = "0.47.0", default-features = false, features = ["static"] }
 serde = { version = "1.0", features = ["derive"] }
 csv = "1.1"
 protein-translate = "0.2.0" 
