From 33d12908513f927d8e8edf0fd0523009b558b4a9 Mon Sep 17 00:00:00 2001
From: rob-p <rob-p@users.noreply.github.com>
Date: Thu, 23 Nov 2017 22:15:28 -0500
Subject: [PATCH 1/1] pre AppleClang 8 patch

Signed-off-by: rob-p <rob-p@users.noreply.github.com>
---
 include/FastxParserThreadUtils.hpp | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/include/FastxParserThreadUtils.hpp b/include/FastxParserThreadUtils.hpp
index 7c1273a..234fd55 100644
--- a/include/FastxParserThreadUtils.hpp
+++ b/include/FastxParserThreadUtils.hpp
@@ -27,8 +27,8 @@ ALWAYS_INLINE void yieldSleep() {
 }
 
 ALWAYS_INLINE void backoffExp(size_t& curMaxIters) {
-  thread_local std::uniform_int_distribution<size_t> dist;
-  thread_local std::minstd_rand gen(std::random_device{}());
+  std::uniform_int_distribution<size_t> dist;
+  std::minstd_rand gen(std::random_device{}());
   const size_t spinIters =
       dist(gen, decltype(dist)::param_type{0, curMaxIters});
   curMaxIters = std::min(2 * curMaxIters, MAX_BACKOFF_ITERS);
-- 
2.13.6 (Apple Git-96)

