diff -ru phyloFlash-2.0-beta6.orig/phyloFlash_makedb.pl phyloFlash-2.0-beta6/phyloFlash_makedb.pl
--- phyloFlash-2.0-beta6.orig/phyloFlash_makedb.pl	2017-08-05 14:02:31.000000000 -0600
+++ phyloFlash-2.0-beta6/phyloFlash_makedb.pl	2017-08-05 14:01:46.000000000 -0600
@@ -76,7 +76,7 @@
 
 use strict;
 use FindBin;
-use lib $FindBin::Bin;
+use lib $FindBin::RealBin;
 use PhyloFlash;
 use Pod::Usage;
 use Getopt::Long;
