--- BICseqMulSample.pl.old	2019-04-24 18:20:27.392166365 -0500
+++ BICseqMulSample.pl	2019-04-24 18:20:27.392166365 -0500
@@ -1,16 +1,16 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 use strict;
 
 use FindBin qw($Bin);
 my $path = $Bin;
 
-my $bicseq = "$path/MBIC-seq_v0.1.5/MBICseq";
-my $combinefile = "$path/combineFile/combineFile";
-my $rreport = "$path/R/reportOneSample.R";
-my $estimatLambdaFct = "$path/EstimateLambdaFactor/EstLambdaFct";
-my $bootstrap = "$path/bootstrap/bootstrapTest";
-my $rcombineBoot = "$path/R/combineSegBoostrap.R";
-my $plotSeg = "$path/R/plotProfile.R";
+my $bicseq = "$path/MBICseq";
+my $combinefile = "$path/combineFile";
+my $rreport = "$path/reportOneSample.R";
+my $estimatLambdaFct = "$path/EstLambdaFct";
+my $bootstrap = "$path/bootstrapTest";
+my $rcombineBoot = "$path/combineSegBoostrap.R";
+my $plotSeg = "$path/plotProfile.R";
 
 if(!(-e $plotSeg)) {
 	die("Cannot find $plotSeg\n");
@@ -20,7 +20,6 @@
         die("Cannot find $rcombineBoot\n");
         }
 
-
 if(!(-e $bicseq)){
 	die("Cannot find $bicseq\n");
 	}
--- BICseqOneSample.pl.old	2019-04-24 18:20:27.392166365 -0500
+++ BICseqOneSample.pl	2019-04-24 18:20:27.379166362 -0500
@@ -1,22 +1,20 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 use strict;
 
 use FindBin qw($Bin);
 my $path = $Bin;
 
-my $bicseq = "$path/MBIC-seq_v0.1.5/MBICseq";
-my $combinefile = "$path/combineFile/combineFile";
-my $report = "$path/R/reportOneSample.R";
-my $bootstrap = "$path/bootstrap/bootstrapTest";
-my $estimatLambdaFct = "$path/EstimateLambdaFactor/EstLambdaFct";
-
-my $plotSeg = "$path/R/plotProfile.R";
+my $bicseq = "$path/MBICseq";
+my $combinefile = "$path/combineFile";
+my $report = "$path/reportOneSample.R";
+my $bootstrap = "$path/bootstrap";
+my $estimatLambdaFct = "$path/EstLambdaFct";
+my $plotSeg = "$path/plotProfile.R";
 
 if(!(-e $plotSeg)) {
         die("Cannot find $plotSeg\n");
         }
 
-
 if(!(-e $estimatLambdaFct)){
 	die("Cannot find $estimatLambdaFct\n");
 	}
@@ -34,7 +32,7 @@
         die("Cannot find $bootstrap\n");
         }
 
-my $rcombineBoot = "$path/R/combineSegBoostrap.R";
+my $rcombineBoot = "$path/combineSegBoostrap.R";
 if(!(-e $rcombineBoot)){
         die("Cannot find $rcombineBoot\n");
         }
--- BICseqTwoSample.pl.old	2019-04-24 18:20:27.392166365 -0500
+++ BICseqTwoSample.pl	2019-04-24 18:20:27.392166365 -0500
@@ -1,16 +1,16 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 use strict;
 
 use FindBin qw($Bin);
 my $path = $Bin;
 
-my $bicseq = "$path/MBIC-seq_v0.1.5/MBICseq";
-my $combinefile = "$path/combineFile/combineFile";
-my $rreport = "$path/R/report.R";
-my $estimatLambdaFct = "$path/EstimateLambdaFactor/EstLambdaFct";
-my $bootstrap = "$path/bootstrap/bootstrapTest";
-my $rcombineBoot = "$path/R/combineSegBoostrap.R";
-my $plotSeg = "$path/R/plotProfile.R";
+my $bicseq = "$path/MBICseq";
+my $combinefile = "$path/combineFile";
+my $rreport = "$path/report.R";
+my $estimatLambdaFct = "$path/EstLambdaFct";
+my $bootstrap = "$path/bootstrapTest";
+my $rcombineBoot = "$path/combineSegBoostrap.R";
+my $plotSeg = "$path/plotProfile.R";
 
 if(!(-e $plotSeg)) {
 	die("Cannot find $plotSeg\n");
@@ -20,7 +20,6 @@
         die("Cannot find $rcombineBoot\n");
         }
 
-
 if(!(-e $bicseq)){
 	die("Cannot find $bicseq\n");
 	}
--- genotype.pl.old	2019-04-24 18:20:27.392166365 -0500
+++ genotype.pl	2019-04-24 18:20:27.392166365 -0500
@@ -1,12 +1,12 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 use strict;
 
 use FindBin qw($Bin);
 my $path = $Bin;
 
-my $combinefile = "$path/combineFile/combineFile";
-my $countRead = "$path/genotype/countRead";
-my $genotype = "$path/genotype/genotype";
+my $combinefile = "$path/combineFile";
+my $countRead = "$path/countRead";
+my $genotype = "$path/genotype";
 
 if(!(-e $countRead)) {
 	die("Cannot find $countRead\n");
--- NBICseq-seg.pl.old	2019-04-24 18:20:27.392166365 -0500
+++ NBICseq-seg.pl	2019-04-24 18:20:27.392166365 -0500
@@ -1,23 +1,22 @@
-#!/usr/bin/env perl
+#!/opt/anaconda1anaconda2anaconda3/bin/perl
 use strict;
 
 use FindBin qw($Bin);
 my $path = $Bin;
 
-my $bicseq = "$path/MBIC-seq_v0.1.5/MBICseq";
-my $combinefile = "$path/combineFile/combineFile";
-my $rreport = "$path/R/report.R";
+my $bicseq = "$path/MBICseq";
+my $combinefile = "$path/combineFile";
+my $rreport = "$path/report.R";
 my $bicseqOneSample = "$path/BICseqOneSample.pl";
 my $bidsegTwoSample = "$path/BICseqTwoSample.pl";
 my $bidsegMulSample = "$path/BICseqMulSample.pl";
-my $bootstrap = "$path/bootstrap/bootstrapTest";
-my $plotSeg = "$path/R/plotProfile.R";
+my $bootstrap = "$path/bootstrap";
+my $plotSeg = "$path/plotProfile.R";
 
 if(!(-e $plotSeg)) {
         die("Cannot find $plotSeg\n");
         }
 
-
 if(!(-e $bicseq)){
         die("Cannot find $bicseq\n");
         }
@@ -32,8 +31,8 @@
 	}
 
 
-my $reportOnesample = "$path/R/reportOneSample.R";
-my $rcombineBoot = "$path/R/combineSegBoostrap.R";
+my $reportOnesample = "$path/reportOneSample.R";
+my $rcombineBoot = "$path/combineSegBoostrap.R";
 if(!(-e "$reportOnesample")){
         die("Cannot find $reportOnesample\n");
         }
