--- Makefile.backup	2021-10-20 13:38:18.000000000 +0200
+++ Makefile	2021-10-20 13:42:03.000000000 +0200
@@ -21,7 +21,7 @@
 ## 
 ## If you have AmberTools installed (version >=13), please set
 ## NETCDF to the location of your AmberTools installation; e.g.:
-## NETCDF=/software/amber14/AmberTools
+NETCDF=$(AMBERHOME)
 ##
 ## If you installed netcdf-fortran using macports on MACOSX,
 ## set NETCDF to the location of your macports tree; e.g.:
@@ -35,12 +35,12 @@
 
 ###------------------ Gromacs XTC input support -------------------###
 ## To enable XTC support, uncomment the following line:
-## XTC= yes
+XTC= yes
 
 ## Please do not edit below this line.
 ###----------------------------------------------------------------###
 ifdef NETCDF
-FFLAG_NC= -I$(NETCDF)/include/ -DNETCDF
+FFLAG_NC= -I$(NETCDF)/include/ -DNETCDF -Wno-argument-mismatch
 ifneq ("$(wildcard $(NETCDF)/lib/libnetcdff.a)","")
 libnetcdf= netcdff
 else
@@ -60,8 +60,14 @@
 locate.o lsfit.o manta.o ncerror.o nml.o \
 params.o pdbout.o curpar.o screw.o setup.o smooth.o torp.o xtcerror.o $(XTC_OBJ)
 
+all: Cur+ Canal Canion
+
 Cur+ : $(OBJ) 
 	$(FC) $(OBJ) $(LDFLAG_NC) $(LDFLAG_XTC) -o Cur+
+Canal : canal.o
+	$(FC) $(FFLAG) canal.o -o Canal
+Canion : canion.o
+	$(FC) $(FFLAG) canion.o -o Canion
 
 .SUFFIXES : .o .c .f
 .c.o :
@@ -71,7 +77,7 @@
 	$(FC) -cpp $(FFLAG) $(FFLAG_NC) $(FFLAG_XTC) -c $<
 
 clean: 
-	rm -f Cur+ *.o *.a
+	rm -f Cur+ Canal Canion *.o *.a
 
 f77_molfile.o: f77_molfile.c plugins
 	$(CC) $(CFLAG) -Iinclude -c $<
