From b328b8de0d43e0750c8eb81a5d1727d942648f25 Mon Sep 17 00:00:00 2001
From: Camille Scott <camille.scott.w@gmail.com>
Date: Fri, 9 Mar 2018 18:13:39 -0800
Subject: [PATCH 1/2] patch bam2hints Makefile

---
 auxprogs/bam2hints/Makefile | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/auxprogs/bam2hints/Makefile b/auxprogs/bam2hints/Makefile
index 9eff1ac..ef2431c 100644
--- a/auxprogs/bam2hints/Makefile
+++ b/auxprogs/bam2hints/Makefile
@@ -9,11 +9,11 @@
 
 # Variable definition
 BAMTOOLS=/usr/local/bamtools
-INCLUDES = $(BAMTOOLS)/include
-LIBS = $(BAMTOOLS)/lib/libbamtools.a -lz
+INCLUDES =  ${PREFIX}/include/bamtools/
+LIBS = ${PREFIX}/lib/libbamtools.a -lz
 SOURCES = bam2hints.cc 
 OBJECTS = $(SOURCES:.cc=.o)
-CFLAGS = -Wall -O0 -g -ggdb # -g -p
+CFLAGS = -Wall -g -p -O0
 
 # Recipe(s)
 # $@: full target name of current target. 
@@ -25,7 +25,7 @@ bam2hints : $(OBJECTS)
 all:$(OBJECTS) # Compiles each foo.cc into foo.o
 
 bam2hints.o : $(SOURCES)
-	$(CXX) $(CFLAGS) -c $< -o $@ -I$(INCLUDES) 
+	$(CXX) $(CFLAGS) $(CXXFLAGS) $(CPPFLAGS) $(LDFLAGS) -c $< -o $@ -I$(INCLUDES) 
 
 clean: 
 	rm -f bam2hints.o bam2hints ../../bin/bam2hints
-- 
2.14.3 (Apple Git-98)


From a2a60e12fcbc2dd4cfc310c730eaf93511df5669 Mon Sep 17 00:00:00 2001
From: Camille Scott <camille.scott.w@gmail.com>
Date: Fri, 9 Mar 2018 18:45:13 -0800
Subject: [PATCH 2/2] Patch out another def

---
 auxprogs/bam2hints/Makefile | 1 -
 1 file changed, 1 deletion(-)

diff --git a/auxprogs/bam2hints/Makefile b/auxprogs/bam2hints/Makefile
index ef2431c..2340db8 100644
--- a/auxprogs/bam2hints/Makefile
+++ b/auxprogs/bam2hints/Makefile
@@ -8,7 +8,6 @@
 #	Last modified:	23-October-2017 by Jon Palmer
 
 # Variable definition
-BAMTOOLS=/usr/local/bamtools
 INCLUDES =  ${PREFIX}/include/bamtools/
 LIBS = ${PREFIX}/lib/libbamtools.a -lz
 SOURCES = bam2hints.cc 
-- 
2.14.3 (Apple Git-98)

