From 8db5d7a51c93c9b757550891de5c6768ad4d05d4 Mon Sep 17 00:00:00 2001
From: Camille Scott <camille.scott.w@gmail.com>
Date: Fri, 9 Mar 2018 18:16:46 -0800
Subject: [PATCH] patch compilesplicecands

---
 auxprogs/compileSpliceCands/Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/auxprogs/compileSpliceCands/Makefile b/auxprogs/compileSpliceCands/Makefile
index d2bbb54..73a44f1 100644
--- a/auxprogs/compileSpliceCands/Makefile
+++ b/auxprogs/compileSpliceCands/Makefile
@@ -2,7 +2,7 @@ compileSpliceCands : compileSpliceCands.o list.h list.o
 	$(CC) -o compileSpliceCands compileSpliceCands.o list.o;
 #	cp compileSpliceCands ../../bin
 compileSpliceCands.o : compileSpliceCands.c 
-	$(CC) -Wall -pedantic -ansi -c compileSpliceCands.c 
+	$(CC) -Wall -pedantic -ansi $(CPPFLAGS) -c compileSpliceCands.c 
 
 all : compileSpliceCands
 
-- 
2.14.3 (Apple Git-98)

