From 9cccd5bbffddadd9be9413f7a790ec7f4cbdca46 Mon Sep 17 00:00:00 2001
From: Camille Scott <camille.scott.w@gmail.com>
Date: Fri, 9 Mar 2018 18:22:47 -0800
Subject: [PATCH] patch filterbam src Makefile

---
 auxprogs/filterBam/src/Makefile | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/auxprogs/filterBam/src/Makefile b/auxprogs/filterBam/src/Makefile
index f96e240..a166157 100644
--- a/auxprogs/filterBam/src/Makefile
+++ b/auxprogs/filterBam/src/Makefile
@@ -7,11 +7,12 @@
 SOURCES = filterBam.cc MatePairs.cc getReferenceName.cc initOptions.cc SingleAlignment.cc \
 			printElapsedTime.cc sumMandIOperations.cc sumDandIOperations.cc PairednessCoverage.cc
 PROGRAM = filterBam
+BAMTOOLS = ${PREFIX}/include/bamtools
 OBJECTS = $(SOURCES:.cc=.o)
-BAMTOOLS = /usr/local/bamtools
-INCLUDES = -I$(BAMTOOLS)/include -Iheaders -I./bamtools
-LIBS = $(BAMTOOLS)/lib/libbamtools.a -lz
-CFLAGS = -std=c++0x
+INCLUDES =  -I${PREFIX}/include/bamtools/ -Iheaders -I./bamtools
+
+LIBS = ${PREFIX}/lib/libbamtools.a -lz
+CFLAGS += -std=c++11 -Wall -O2 # -g -p -g -ggdb
 VPATH = functions 
 INSTDIR = ../bin
 
-- 
2.14.3 (Apple Git-98)

