From 2af9c2187abb15e5b1d4903e55352d6a21641fdd Mon Sep 17 00:00:00 2001
From: Camille Scott <camille.scott.w@gmail.com>
Date: Fri, 9 Mar 2018 18:20:28 -0800
Subject: [PATCH] Patch filterBam

---
 auxprogs/filterBam/Makefile | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/auxprogs/filterBam/Makefile b/auxprogs/filterBam/Makefile
index ac69dd4..dbb6ad8 100644
--- a/auxprogs/filterBam/Makefile
+++ b/auxprogs/filterBam/Makefile
@@ -3,6 +3,13 @@
 # 	Created: 28-February-2012
 #	Last modified: 15-October-2015
 
+INCLUDES =  ${PREFIX}/include/bamtools/
+LIBS = ${PREFIX}/lib/libbamtools.a -lz
+SOURCES = bam2hints.cc
+OBJECTS = $(SOURCES:.cc=.o)
+CFLAGS = -Wall -g -p -O0
+CXXFLAGS += -Wall -O2 # -g -p -g -ggdb 
+
 all: 
 	(cd src;make CXX=$(BAMTOOLS_CXX) CC=$(BAMTOOLS_CC) BAMTOOLS=$(BAMTOOLS))
 
-- 
2.14.3 (Apple Git-98)

