From c1723f20c0601e5600ecf41ae89488a41db8ffea Mon Sep 17 00:00:00 2001
From: Camille Scott <camille.scott.w@gmail.com>
Date: Mon, 12 Mar 2018 18:08:10 -0700
Subject: [PATCH] Add mkdir -p call to homegene src

---
 auxprogs/homGeneMapping/src/Makefile | 1 +
 1 file changed, 1 insertion(+)

diff --git a/auxprogs/homGeneMapping/src/Makefile b/auxprogs/homGeneMapping/src/Makefile
index c3eaae1..1a2810e 100644
--- a/auxprogs/homGeneMapping/src/Makefile
+++ b/auxprogs/homGeneMapping/src/Makefile
@@ -28,6 +28,7 @@ all: homGeneMapping
 
 homGeneMapping: main.cc $(OBJS)
 	 $(CXX) $(CFLAGS) -o $@ $^ $(INCLS)
+	mkdir -p ../bin
 	mv homGeneMapping ../bin/
 	cp ../bin/homGeneMapping ../../../bin/homGeneMapping
 clean:	
-- 
2.14.3 (Apple Git-98)

