From 9a6945e866a492befd6830ddd73724b5a8d96273 Mon Sep 17 00:00:00 2001
From: Ilya Shlyakhter <ilya_shl@alum.mit.edu>
Date: Mon, 14 Aug 2017 19:44:45 -0400
Subject: [PATCH] changed tools to shiver_tools

---
 shiver_funcs.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/shiver_funcs.sh b/shiver_funcs.sh
index fd4b7e2..0d24781 100755
--- a/shiver_funcs.sh
+++ b/shiver_funcs.sh
@@ -4,7 +4,7 @@ set -u
 set -o pipefail
 
 ThisDir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
-ToolsDir="$ThisDir"/tools
+ToolsDir="$ThisDir"/shiver_tools
 Code_AlignToConsensus="$ToolsDir/AlignMoreSeqsToPairWithMissingCoverage.py"
 Code_AnalysePileup="$ToolsDir/AnalysePileup.py"
 Code_CallConsensus="$ToolsDir/CallConsensus.py"
-- 
2.12.0

