From b343e299a11607cc19140bb665d166ac1e0f7adb Mon Sep 17 00:00:00 2001
From: Ilya Shlyakhter <ilya_shl@alum.mit.edu>
Date: Wed, 16 Aug 2017 10:28:14 -0400
Subject: [PATCH 4/4] fixed pyfastaq installation

---
 shiver_fastaq | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/shiver_fastaq b/shiver_fastaq
index 8a1c079..4e9f7e9 100755
--- a/shiver_fastaq
+++ b/shiver_fastaq
@@ -5,7 +5,7 @@ set -e -o pipefail
 if conda env list | grep -w "shiver_pyfastaq_env" > /dev/null ; then
     echo already have pyfastaq env
 else
-		conda create -y -n shiver_pyfastaq_env 'python>3 pyfastaq'
+		conda create -y -n shiver_pyfastaq_env pyfastaq
 fi
 
 source activate shiver_pyfastaq_env
-- 
2.12.0

