diff --git a/src/pyOpenMS/pyopenms/__init__.py b/src/pyOpenMS/pyopenms/__init__.py
index 6420a989cb..4bc81ce59f 100755
--- a/src/pyOpenMS/pyopenms/__init__.py
+++ b/src/pyOpenMS/pyopenms/__init__.py
@@ -37,9 +37,9 @@ if sys.platform.startswith("linux"):
     # those are not found. setting LD_LIBRARY_PATH does not work,
     # see: http://stackoverflow.com/questions/1178094
     import ctypes
-    ctypes.cdll.LoadLibrary(os.path.join(here, "libOpenSwathAlgo.so"))
-    ctypes.cdll.LoadLibrary(os.path.join(here, "libOpenMS.so"))
-    ctypes.cdll.LoadLibrary(os.path.join(here, "libSuperHirn.so"))
+    ctypes.cdll.LoadLibrary("libOpenSwathAlgo.so")
+    ctypes.cdll.LoadLibrary("libOpenMS.so")
+    ctypes.cdll.LoadLibrary("libSuperHirn.so")
 
 try:
     from .all_modules import *
