From 4caf370d403002fe9c38fd56b8c46f1718d364ab Mon Sep 17 00:00:00 2001
From: Matus Kosut <matuskosut@gmail.com>
Date: Fri, 9 Apr 2021 17:15:54 +0200
Subject: [PATCH] use shared libs in makevars

---
 src/Makevars | 19 +++----------------
 1 file changed, 3 insertions(+), 16 deletions(-)

diff --git a/src/Makevars b/src/Makevars
index 1ea24566b..3e6c70c62 100644
--- a/src/Makevars
+++ b/src/Makevars
@@ -1,26 +1,13 @@
 PKG_CPPFLAGS = \
--I ../thirdParty/bgen/db/include \
--I ../thirdParty/bgen/genfile/include \
--I ../thirdParty/bgen/genfile/include/genfile \
--I ../thirdParty/bgen/genfile/include/genfile/bgen \
--I ../thirdParty/bgen/db/include/db \
--I ../thirdParty/bgen/3rd_party/boost_1_55_0 \
--I ../thirdParty/bgen/3rd_party/zstd-1.1.0 \
--I ../thirdParty/bgen/3rd_party/zstd-1.1.0/lib \
--I ../thirdParty/bgen/3rd_party/zstd-1.1.0/lib/compress \
--I ../thirdParty/bgen/3rd_party/zstd-1.1.0/lib/decompress \
--I ../thirdParty/bgen/3rd_party/sqlite3 \
--I ../thirdParty/bgen/3rd_party/boost_1_55_0/boost/iostreams \
--I ../thirdParty/cget/include \
--I ../thirdParty/cget/lib \
+-I ${PREFIX}/include \
 -D SQLITE_ENABLE_COLUMN_METADATA -O3 -fpic -Wall -Wextra -pedantic  -DARMA_64BIT_WORD=1
 
 CXX_STD = CXX11 
 
-PKG_LIBS = $(ZLIB_LIB) $(LAPACK_LIBS) $(BLAS_LIBS) $(FLIBS) -L../bgen/3rd_party/boost_1_55_0/boost/iostreams -L../thirdParty/cget/lib/ -L../thirdParty/cget/lib64/ -lsuperlu
+PKG_LIBS = $(ZLIB_LIB) $(LAPACK_LIBS) $(BLAS_LIBS) $(FLIBS) -L${PREFIX}/lib -lsuperlu -lbgen -ldb -lzstd -lsqlite3 -lboost_iostreams -lboost_system -lboost_filesystem -lboost_thread -lboost_timer -lboost_chrono -lboost_date_time
 PKG_LIBS += $(shell ${R_HOME}/bin/Rscript -e "RcppParallel::RcppParallelLibs()")
 
-OBJECTS = RcppExports.o test.o matrixMultiEigen.o SAIGE_fitGLMM_fast.o SAIGE_readDosage_bgen.o SAIGE_readDosage_vcf.o SAIGE_GENE_readDosageMatrix_vcf.o  $(wildcard ../thirdParty/bgen/build/src/*.o) ../thirdParty/bgen/build/db/libdb.a ../thirdParty/bgen/build/3rd_party/zstd-1.1.0/libzstd.a ../thirdParty/bgen/build/3rd_party/sqlite3/libsqlite3.a ../thirdParty/bgen/build/3rd_party/boost_1_55_0/libboost.a $(wildcard ../thirdParty/cget/lib64/*.a)
+OBJECTS = RcppExports.o test.o matrixMultiEigen.o SAIGE_fitGLMM_fast.o SAIGE_readDosage_bgen.o SAIGE_readDosage_vcf.o SAIGE_GENE_readDosageMatrix_vcf.o
 
 all: $(SHLIB)
 $(SHLIB): ${OBJECTS} Makevars
-- 
2.31.0

