From 21ebc0d50201b21d57decae8180d5d1dd6e18a8f Mon Sep 17 00:00:00 2001
From: Lea Waller <lea.waller@charite.de>
Date: Wed, 11 May 2022 10:39:57 +0200
Subject: [PATCH] use shared libs in makevars

---
 src/Makevars | 5 +----
 1 file changed, 1 insertion(+), 4 deletions(-)

diff --git a/src/Makevars b/src/Makevars
index f3841779..e3da7f8b 100644
--- a/src/Makevars
+++ b/src/Makevars
@@ -1,12 +1,9 @@
 PKG_CPPFLAGS = \
--I ../thirdParty/cget/include \
--I ../thirdParty/cget/lib \
--I ../thirdParty/cget/lib64 \
 -D SQLITE_ENABLE_COLUMN_METADATA -O3 -fpic -Wall -Wextra -pedantic  -DARMA_64BIT_WORD=1 -D__STDC_FORMAT_MACROS

 CXX_STD = CXX11

-PKG_LIBS = $(ZLIB_LIB) $(LAPACK_LIBS) $(BLAS_LIBS) $(FLIBS) -L../thirdParty/cget/lib/ -L../thirdParty/cget/lib64/ -lsuperlu -lzstd -L../lib64 -llapack
+PKG_LIBS = $(ZLIB_LIB) $(LAPACK_LIBS) $(BLAS_LIBS) $(FLIBS)  -lsuperlu -lzstd -llapack
 PKG_LIBS += $(shell ${R_HOME}/bin/Rscript -e "RcppParallel::RcppParallelLibs()")


--
2.36.1