diff --git a/src/Makevars b/src/Makevars
index f1e54541..5345d7b5 100644
--- a/src/Makevars
+++ b/src/Makevars
@@ -1,12 +1,9 @@
 PKG_CPPFLAGS = \
--I ../thirdParty/cget/include \
--I ../thirdParty/cget/lib \
--I ../thirdParty/cget/lib64 \
 -D SQLITE_ENABLE_COLUMN_METADATA -O3 -fpic -Wall -Wextra -pedantic  -DARMA_64BIT_WORD=1 -D__STDC_FORMAT_MACROS
 
 CXX_STD = CXX11 
 
-PKG_LIBS = $(ZLIB_LIB) $(LAPACK_LIBS) $(BLAS_LIBS) $(FLIBS) -L../thirdParty/cget/lib/ -L../thirdParty/cget/lib64/ -lsuperlu -lzstd -L../lib64 -llapack
+PKG_LIBS = $(ZLIB_LIB) $(LAPACK_LIBS) $(BLAS_LIBS) $(FLIBS)  -lsuperlu -lzstd -llapack
 PKG_LIBS += $(shell ${R_HOME}/bin/Rscript -e "RcppParallel::RcppParallelLibs()")
