From 661ca2a3fc2e7d3ce1cfb11148e004f523a192e5 Mon Sep 17 00:00:00 2001
From: Matus Kosut <matus.kosut@ntnu.no>
Date: Thu, 5 Nov 2020 20:03:36 +0100
Subject: [PATCH] fix rcpp returns

---
 src/SAIGE_readDosage_bgen.cpp | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/SAIGE_readDosage_bgen.cpp b/src/SAIGE_readDosage_bgen.cpp
index 23c5d1c13..3703625fd 100644
--- a/src/SAIGE_readDosage_bgen.cpp
+++ b/src/SAIGE_readDosage_bgen.cpp
@@ -63,6 +63,7 @@ int numSamples_bgen;
 // [[Rcpp::export]]
 void setIsDropMissingDosages_bgen (bool isdropmissingdosages){
   isDropMissingDosages_bgen = isdropmissingdosages;
+  return;
 }
 
 // // [[Rcpp::export]]
@@ -1058,6 +1059,7 @@ void SetSampleIdx(Rcpp::IntegerVector sample_idx, int Ntest){
 	gmtest_samplesize = Ntest;
 	gm_sample_idx = sample_idx;
 	//cc_idx = cc_index;
+	return;
 }
 
 
@@ -1068,7 +1070,7 @@ void closetestGenoFile_bgenDosage() //needs further check
 //  gm_sample_idx.erase();
 
   printf("closed the genofile!\n");
-
+  return;
 }
 
 
-- 
2.28.0

