From 24735ad0cec9bea2690cad913f48f8b31ae5d951 Mon Sep 17 00:00:00 2001
From: Matus Kosut <matus.kosut@ntnu.no>
Date: Wed, 28 Oct 2020 18:35:42 +0100
Subject: [PATCH 1/2] fix gettimeofday import

---
 src/SAIGE_fitGLMM_fast.cpp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/SAIGE_fitGLMM_fast.cpp b/src/SAIGE_fitGLMM_fast.cpp
index e0f9deded..446e830c0 100644
--- a/src/SAIGE_fitGLMM_fast.cpp
+++ b/src/SAIGE_fitGLMM_fast.cpp
@@ -11,6 +11,7 @@
 #include <cmath>
 #include <ctime>// include this header for calculating execution time 
 #include <cassert>
+#include <boost/date_time.hpp> // for gettimeofday and timeval
 using namespace Rcpp;
 using namespace std;
 using namespace RcppParallel;
-- 
2.23.0

