diff --git a/BMEAN/makefile b/BMEAN/makefile
index f1c4053..251e3be 100644
--- a/BMEAN/makefile
+++ b/BMEAN/makefile
@@ -1,7 +1,7 @@
 # CC=/usr/bin/g++
-CXX= g++
-CFLAGS = -O3 -std=c++11 -lpthread -Ispoa/include
-LFLAGS = -Ispoa/include/ -Lspoa/build/lib/ -lspoa
+CXX ?= $(CXX)
+CFLAGS = -O3 -std=c++14 -lpthread -Ispoa/include -IComplete-Striped-Smith-Waterman-Library/src
+LFLAGS = -Ispoa/include -Lspoa/build/lib -IComplete-Striped-Smith-Waterman-Library/src -LComplete-Striped-Smith-Waterman-Library/src -lspoa
 EXEC=testLR
 OBJS := *.o Complete-Striped-Smith-Waterman-Library/src/*.o
 
@@ -28,7 +28,7 @@ all: $(EXEC)
 
 
 testLR:  testLR.o bmean.o utils.o
-	$(CXX) $(CFLAGS)  $(OBJS) -o $@ $(LFLAGS)
+	$(CXX) $(CFLAGS) $(OBJS) -o $@ $(LFLAGS)
 
 %.o: %.cpp
 	$(CXX) -o $@ -c $< $(CFLAGS) $(LFLAGS)
