diff --git a/BMEAN/makefile b/BMEAN/makefile
index f1c4053..944cd3e 100644
--- a/BMEAN/makefile
+++ b/BMEAN/makefile
@@ -1,6 +1,6 @@
 # CC=/usr/bin/g++
-CXX= g++
-CFLAGS = -O3 -std=c++11 -lpthread -Ispoa/include
+CXX ?= $(CXX)
+CFLAGS = -O3 -std=c++14 -lpthread -Ispoa/include -I$(PREFIX)/include
 LFLAGS = -Ispoa/include/ -Lspoa/build/lib/ -lspoa
 EXEC=testLR
 OBJS := *.o Complete-Striped-Smith-Waterman-Library/src/*.o
