diff --git a/minimap2/Makefile b/minimap2/Makefile
index 18622f5..69a79e6 100644
--- a/minimap2/Makefile
+++ b/minimap2/Makefile
@@ -1,5 +1,5 @@
-CFLAGS=		-g -Wall -O2 -Wc++-compat #-Wextra
-CPPFLAGS=	-DHAVE_KALLOC
+CFLAGS=		-g -Wall -O3 -Wc++-compat #-Wextra
+CPPFLAGS=	-DHAVE_KALLOC -I$(PREFIX)/include
 INCLUDES=
 OBJS=		kthread.o kalloc.o misc.o bseq.o sketch.o sdust.o options.o index.o chain.o align.o hit.o map.o format.o pe.o esterr.o splitidx.o ksw2_ll_sse.o
 PROG=		minimap2
@@ -43,7 +43,7 @@ all:$(PROG)
 extra:all $(PROG_EXTRA)
 
 minimap2:main.o libminimap2.a
-		$(CC) $(CFLAGS) main.o -o $@ -L. -lminimap2 $(LIBS)
+		$(CC) $(CFLAGS) main.o -o $@ -L. -L$(PREFIX)/lib -lminimap2 $(LIBS)
 
 minimap2-lite:example.o libminimap2.a
 		$(CC) $(CFLAGS) $< -o $@ -L. -lminimap2 $(LIBS)
