diff --git a/minimap2/Makefile b/minimap2/Makefile
index 18622f5..b5a96d9 100644
--- a/minimap2/Makefile
+++ b/minimap2/Makefile
@@ -1,4 +1,4 @@
-CFLAGS=		-g -Wall -O2 -Wc++-compat #-Wextra
+CFLAGS=		-g -Wall -O3 -Wc++-compat #-Wextra
 CPPFLAGS=	-DHAVE_KALLOC
 INCLUDES=
 OBJS=		kthread.o kalloc.o misc.o bseq.o sketch.o sdust.o options.o index.o chain.o align.o hit.o map.o format.o pe.o esterr.o splitidx.o ksw2_ll_sse.o
