diff --git a/CONSENT-polish b/CONSENT-polish
index a7bc92e..a74971e 100755
--- a/CONSENT-polish
+++ b/CONSENT-polish
@@ -190,11 +190,11 @@ minimap2 --dual=yes -PD --no-long-join -w5 -g1000 -m30 -n1 -t"$nproc" -I"$minima
 
 echo "["$(date)"] Processing the overlaps"
 LC_COLLATE=C sort -k6,6 -T $tmpdir --parallel="$nproc" $tmpdir/"$alignments" > $tmpdir/"sorted_$alignments"
-$LRSCf/bin/reformatPAF $tmpdir/"sorted_$alignments" $tmpdir/"$alignments"
+reformatPAF $tmpdir/"sorted_$alignments" $tmpdir/"$alignments"
 rm $tmpdir/"sorted_$alignments"
 
 echo "["$(date)"] Polishing the contigs"
-$LRSCf/bin/CONSENT-polishing -a $tmpdir/"$alignments" -s "$minSupport" -S "$maxSupport" -l "$windowSize" -k "$merSize" -c "$commonKMers" -A "$minAnchors" -f "$solid" -m "$windowOverlap" -j "$nproc" -r "$contigs" -R "$reads" -M "$maxMSA" -p "$LRSCf" >> "$out"
+CONSENT-polishing -a $tmpdir/"$alignments" -s "$minSupport" -S "$maxSupport" -l "$windowSize" -k "$merSize" -c "$commonKMers" -A "$minAnchors" -f "$solid" -m "$windowOverlap" -j "$nproc" -r "$contigs" -R "$reads" -M "$maxMSA" -p "$LRSCf" >> "$out"
 echo "["$(date)"] Removing the temporary files"
 rm $tmpdir/"$minimapErrlog"
 rm $tmpdir/"$alignments"
