diff --git a/CONSENT-correct b/CONSENT-correct
index 5d06de0..a9b66be 100755
--- a/CONSENT-correct
+++ b/CONSENT-correct
@@ -186,17 +186,17 @@ else
 fi
 
 echo "["$(date)"] Sorting the overlaps"
-$LRSCf/bin/explode $tmpdir/"$alignments" $tmpdir/$exploded
+CONSENT-explode $tmpdir/"$alignments" $tmpdir/$exploded
 for f in "$tmpdir/$exploded"_*;
 do
 	sort -k1,1 -k 10,10nr -T . --parallel="$nproc" $f > $tmpdir/sorted_$(basename $f)
 	rm $f
 done
-$LRSCf/bin/merge $tmpdir/"$alignments" correction "$tmpdir/sorted_"$exploded*
+CONSENT-merge $tmpdir/"$alignments" correction "$tmpdir/sorted_"$exploded*
 rm "$tmpdir/sorted_"$exploded*
 
 echo "["$(date)"] Correcting the long reads"
-$LRSCf/bin/CONSENT-correction -a $tmpdir/"$alignments" -s "$minSupport" -S "$maxSupport" -l "$windowSize" -k "$merSize" -c "$commonKMers" -A "$minAnchors" -f "$solid" -m "$windowOverlap" -j "$nproc" -r "$reads" -M "$maxMSA" -p "$LRSCf" >> "$out"
+CONSENT-correction -a $tmpdir/"$alignments" -s "$minSupport" -S "$maxSupport" -l "$windowSize" -k "$merSize" -c "$commonKMers" -A "$minAnchors" -f "$solid" -m "$windowOverlap" -j "$nproc" -r "$reads" -M "$maxMSA" -p "$LRSCf" >> "$out"
 
 echo "["$(date)"] Removing the temporary files"
 rm $tmpdir/"$alignments"
