diff --git a/CONSENT-polish b/CONSENT-polish
index 31cc9a1..36371da 100755
--- a/CONSENT-polish
+++ b/CONSENT-polish
@@ -187,20 +187,20 @@ echo "["$(date)"] Aligning the long reads to the contigs (minimap2)"
 minimap2 --dual=yes -PD --no-long-join -w5 -g1000 -m30 -n1 -t"$nproc" -I"$minimapMemory" "$contigs" "$reads" > $tmpdir/"$alignments"
 
 echo "["$(date)"] Sorting the overlaps"
-$LRSCf/bin/explode $tmpdir/"$alignments" $tmpdir/$exploded
+CONSENT-explode $tmpdir/"$alignments" $tmpdir/$exploded
 for f in "$tmpdir/$exploded"_*;
 do
 	sort -k6,6 -k 10,10nr -T . --parallel="$nproc" $f > $tmpdir/sorted_$(basename $f)
 	rm $f
 done
-$LRSCf/bin/merge $tmpdir/"tmp_$alignments" polishing "$tmpdir/sorted_"$exploded*
+CONSENT-merge $tmpdir/"tmp_$alignments" polishing "$tmpdir/sorted_"$exploded*
 rm "$tmpdir/sorted_"$exploded*
-$LRSCf/bin/reformatPAF.py $tmpdir/"tmp_$alignments" > $tmpdir/"$alignments"
+CONSENT-reformatPAF.py $tmpdir/"tmp_$alignments" > $tmpdir/"$alignments"
 rm "$tmpdir/tmp_$alignments"
 
 
 echo "["$(date)"] Polishing the contigs"
-$LRSCf/bin/CONSENT-polishing -a $tmpdir/"$alignments" -s "$minSupport" -S "$maxSupport" -l "$windowSize" -k "$merSize" -c "$commonKMers" -A "$minAnchors" -f "$solid" -m "$windowOverlap" -j "$nproc" -r "$contigs" -R "$reads" -M "$maxMSA" -p "$LRSCf" >> "$out"
+CONSENT-polishing -a $tmpdir/"$alignments" -s "$minSupport" -S "$maxSupport" -l "$windowSize" -k "$merSize" -c "$commonKMers" -A "$minAnchors" -f "$solid" -m "$windowOverlap" -j "$nproc" -r "$contigs" -R "$reads" -M "$maxMSA" -p "$LRSCf" >> "$out"
 echo "["$(date)"] Removing the temporary files"
 rm $tmpdir/"$alignments"
 echo "["$(date)"] Exiting"
