--- Misc/Applications/lib/foldGrammars/Settings.pm	2020-05-30 21:03:20.698373775 +0200
+++ Misc/Applications/lib/foldGrammars/Settings.pm.conda	2020-05-30 21:03:15.414346574 +0200
@@ -34,7 +34,7 @@
 	'centroid_fold', 'centroid_fold',
 	'ct2b.pl', 'ct2b.pl',
 	'gapc', 'gapc',
-	'gcc','gcc',
+	'gcc',$ENV{'CXX'},
 	'ghc','/vol/ghc-7.6/bin/ghc',
 	'hybrid-ss-min', 'hybrid-ss-min-3.8',
 	'readlink', 'readlink',
