--- Build.PL
+++ Build.PL
@@ -76,6 +76,10 @@
 	    if -e "$samtools/$HeaderFile";
 	$sam_include = "$samtools/include"
 	    if -e "$samtools/include/$HeaderFile";
+	$sam_include = "$samtools/include/samtools"
+	    if -e "$samtools/include/samtools/$HeaderFile";
+	$sam_include = "$samtools/include/bam"
+	    if -e "$samtools/include/bam/$HeaderFile";
 	$sam_lib     = $samtools
 	    if -e "$samtools/$LibFile";
 	$sam_lib     = "$samtools/lib"
