From 3b029e472574ebc0a8b68a80d63154d9adb1f23b Mon Sep 17 00:00:00 2001
From: Nils Homer <nilshomer@gmail.com>
Date: Thu, 20 Mar 2025 19:04:11 -0700
Subject: [PATCH] feat: support arm64

---
 installer/rtg | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/installer/rtg b/installer/rtg
index 740e6681..1ee51125 100755
--- a/installer/rtg
+++ b/installer/rtg
@@ -5,7 +5,7 @@ if [[ "$(uname -s)" != "Linux" ]] && [[ "$(uname -s)" != "Darwin" ]]; then
     # If you comment this check out you are on your own :-)
     echo "Sorry, only Linux and MacOS are supported."
     exit 1
-elif [[ "$(uname -m)" != "x86_64" ]]; then
+elif [[ "$(uname -m)" != "x86_64" ]] && [[ "$(uname -m)" != "arm64" ]]; then
     # If you comment this check out you are on your own :-)
     echo "Sorry, you must be running a 64bit operating system."
     exit 1
-- 
2.39.5 (Apple Git-154)

