--- shannon.py.bak	2018-08-30 11:47:27.138347197 -0700
+++ shannon.py	2018-08-30 12:01:56.090925724 -0700
@@ -1,3 +1,4 @@
+#! /usr/bin/env python
 import time
 import sys
 import re
@@ -144,7 +145,7 @@
 # Read input from terminal
 n_inp = sys.argv[1:]
 if '--help' in n_inp:
-	with open('manual.md','r') as fin:
+	with open(os.path.join(shannon_dir, 'manual.md'),'r') as fin:
 		print fin.read()
 	exit_now = True
 	sys.exit()
