diff --git a/src/raccess/run_raccess.mk b/src/raccess/run_raccess.mk
index 6398562..979c2d3 100644
--- a/src/raccess/run_raccess.mk
+++ b/src/raccess/run_raccess.mk
@@ -2,8 +2,8 @@ SRCS_RACCESS__RACCESS   = ${SRCS0} run_raccess.cpp
 OBJS_RACCESS__RACCESS   = ${SRCS_RACCESS__RACCESS:.cpp=.o}
 OBJS_RACCESS__RACCESS_D = ${SRCS_RACCESS__RACCESS:.cpp=.o_D}
 run_raccess: $(OBJS_RACCESS__RACCESS)
-	$(CC) $(CFLAGS) $(INCLUDES) -o $@ $^ $(LIBS)
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) $(LDFLAGS) -o $@ $^ $(LIBS) -lz
 run_raccess_D: $(OBJS_RACCESS__RACCESS_D)
-	$(CC) $(CFLAGS_D) $(INCLUDES) -o $@ $^ $(LIBS_D)
+	$(CC) $(CFLAGS_D) $(INCLUDES) -o $@ $^ $(LIBS_D) -lz
 EXES_ALL += run_raccess
 SRCS_ALL += $(SRCS_RACCESS__RACCESS)
diff --git a/src/raccess/run_raccess_contrafold.mk b/src/raccess/run_raccess_contrafold.mk
index cc89d59..d9696da 100644
--- a/src/raccess/run_raccess_contrafold.mk
+++ b/src/raccess/run_raccess_contrafold.mk
@@ -2,7 +2,7 @@ SRCS_RACCESS__RACCESS_CONTRAFOLD   = ${SRCS0} run_raccess_contrafold.cpp
 OBJS_RACCESS__RACCESS_CONTRAFOLD   = ${SRCS_RACCESS__RACCESS_CONTRAFOLD:.cpp=.o}
 OBJS_RACCESS__RACCESS_CONTRAFOLD_D = ${SRCS_RACCESS__RACCESS_CONTRAFOLD:.cpp=.o_D}
 run_raccess_contrafold: $(OBJS_RACCESS__RACCESS_CONTRAFOLD)
-	$(CC) $(CFLAGS) $(INCLUDES) -o $@ $^ $(LIBS)
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) $(LDFLAGS) -o $@ $^ $(LIBS)
 run_raccess_contrafold_D: $(OBJS_RACCESS__RACCESS_CONTRAFOLD_D)
 	$(CC) $(CFLAGS_D) $(INCLUDES) -o $@ $^ $(LIBS_D)
 EXES_ALL += run_raccess_contrafold
