diff --git a/src/util/def_opt_main.hpp b/src/util/def_opt_main.hpp
index ea38654..d7fc12a 100644
--- a/src/util/def_opt_main.hpp
+++ b/src/util/def_opt_main.hpp
@@ -11,7 +11,7 @@
   static int main(int argc, char** argv, Self& obj) {
 #define opt_item_(_name, _type, _value, _desc)
 #define opt_item(_name, _type, _value, _desc) obj.set_##_name(stot(_type(), _value)); \
-    { os << ("-"#_name"=<"#_type":"_value">\t:\t"_desc"\n");};
+    { os << ("-" #_name "=<" #_type ":" _value ">\t:\t" _desc "\n");};
     ostringstream os; os << "options: -name=<type:default>\t:\tdescription\n\n"; opt_items();
 #undef opt_item
 #undef opt_item_
