diff --git a/src/Makefile b/src/Makefile
index 3e22cb3..cc606e4 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -1,5 +1,6 @@
 CFLAGS1      = 
-CC           = g++
+CXX          ?= g++
+CC           ?= $(CXX)
 # CC = icpc
 
 SRC_DIR      = ./
@@ -11,6 +12,9 @@ LIBS         = # -lz
 CFLAGS_D     = -O0 -g3 -pg -Wall -DDEBUG -fkeep-inline-functions -fkeep-static-consts  \
 	       -finline-limit=0 -fno-default-inline $(CFLAGS1)
 LIBS_D       = # -lz
+CXXFLAGS += -O3 -Wall $(CFLAGS1)
+CPPFLAGS += $(INCLUDES)
+
 
 .PHONY: clean depend all all_mpi depend_all util util_mpi raccess
 
@@ -37,7 +41,7 @@ util_mpi:
 	(cd util/; make all_mpi)
 
 .cpp.o:
-	$(CC) $(CFLAGS) $(INCLUDES) -o $@ -c $<
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) -o $@ -c $<
 .cpp.o_D:
 	$(CC) $(CFLAGS_D) $(INCLUDES) -o $@ -c $<
 clean:
diff --git a/src/Makefile~ b/src/Makefile~
deleted file mode 100644
index f73630a..0000000
--- a/src/Makefile~
+++ /dev/null
@@ -1,63 +0,0 @@
-CFLAGS1      = 
-CC           = g++
-# CC = icpc
-
-SRC_DIR      = ./
-SOFTWARE_DIR = $(SRC_DIR)/../software/
-SUB_DIRS     = util rchange
-
-INCLUDES     = -I$(SRC_DIR)
-CFLAGS       = -O3 -g3 -pg -Wall $(CFLAGS1)
-LIBS         = # -lz
-CFLAGS_D     = -O0 -g3 -pg -Wall -DDEBUG -fkeep-inline-functions -fkeep-static-consts  \
-	       -finline-limit=0 -fno-default-inline $(CFLAGS1)
-LIBS_D       = # -lz
-
-.PHONY: clean depend all all_mpi depend_all util util_mpi raccess
-
-.SUFFIXES: .cpp .hpp .o .o_D
-all: all_
-
-SRCS0    = 
-EXES_ALL =
-SRCS_ALL =
-
-MPICC = mpic++
-LIBS += -lpthread
-EXES_ALL_MPI =
-SRCS_ALL_MPI =
-
-all_:  $(EXES_ALL) util rchange rprofile
-all_mpi: util_mpi $(EXES_ALL_MPI)
-util:
-	(cd util/; make)
-raccess:
-	(cd raccess/; make)
-
-util_mpi:
-	(cd util/; make all_mpi)
-
-.cpp.o:
-	$(CC) $(CFLAGS) $(INCLUDES) -o $@ -c $<
-.cpp.o_D:
-	$(CC) $(CFLAGS_D) $(INCLUDES) -o $@ -c $<
-clean:
-	rm -rf core {.,*,*/*}/{*.exe,*.o,*.o_*,*.da,*.bb,*.bbg,*.gcda,*.gcno,*.gcov,*.stackdump,*.exe}
-
-depend_all:
-	for dir in $(SUB_DIRS); \
-	do (cd $$dir; touch make_deps_o.mk make_deps_o_D.mk; make depend) done;
-	touch make_deps_o.mk make_deps_o_D.mk; make depend
-
-SUFLIST = o o_D 
-depend:
-	for opt in $(SUFLIST);                      \
-	do touch make_deps_$$opt.mk;                \
-	   makedepend -o.$$opt -fmake_deps_$$opt.mk \
-	   -Y -- $(CFLAGS_D) $(INCLUDES) --         \
-	   $(SRCS_ALL) $(SRCS_ALL_MPI);             \
-	done
-include make_deps_o.mk
-include make_deps_o_D.mk
-
-
diff --git a/src/raccess/Makefile b/src/raccess/Makefile
index aa44517..23f5152 100644
--- a/src/raccess/Makefile
+++ b/src/raccess/Makefile
@@ -1,7 +1,8 @@
 BIG_INLINE_LIMIT = 20000
 CFLAGS1 = 
 
-CC = g++
+CXX ?= g++
+CC  ?= $(CXX)
 # CC = icpc
 
 SRC_DIR   = ../
@@ -9,6 +10,8 @@ SRC_DIR   = ../
 INCLUDES = -I$(SRC_DIR)
 
  # CFLAGS = -O3 -Wall -finline-limit=$(BIG_INLINE_LIMIT) $(CFLAGS1)
+CXXFLAGS += -O3 -Wall $(CFLAGS1)
+CPPFLAGS += $(INCLUDES)
 CFLAGS = -O3 -g -Wall $(CFLAGS1)
 # CFLAGS = -O2 -Wall $(CFLAGS1)
 
@@ -50,9 +53,9 @@ util:
 util_mpi:
 	(cd $(UTIL_DIR); make all_mpi)
 .cpp.o:
-	$(CC) $(CFLAGS) $(INCLUDES) -o $@ -c $<
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) -o $@ -c $<
 .c.o:
-	$(CC) $(CFLAGS) $(INCLUDES) -o $@ -c $<
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) -o $@ -c $<
 .c.o_D:
 	$(CC) $(CFLAGS) $(INCLUDES) -o $@ -c $<
 .cpp.o_D:
diff --git a/src/util/Makefile b/src/util/Makefile
index c99065d..5e2f12b 100644
--- a/src/util/Makefile
+++ b/src/util/Makefile
@@ -1,7 +1,9 @@
 BIG_INLINE_LIMIT = 20000
 CFLAGS1 = 
 
-CC = g++
+CXX ?= g++
+CC  ?= $(CXX)
+
 # CC = icpc
 
 SRC_DIR  = ../
@@ -11,6 +13,8 @@ INCLUDES  = -I$(SRC_DIR) -I$(UTIL_DIR)
 
 CFLAGS = -O3 -g3 -pg -Wall $(INCLUDES) $(CFLAGS1)
 LIBS = -lz
+CXXFLAGS += -O3 -Wall $(CFLAGS1)
+CPPFLAGS += $(INCLUDES)
 CFLAGS_D = -O0 -g3 -pg -Wall -DDEBUG                         \
 	   -fkeep-inline-functions -fkeep-static-consts      \
 	   -finline-limit=0 -fno-default-inline $(CFLAGS1)
@@ -33,11 +37,11 @@ all_: $(EXES_ALL)
 all_mpi: util_mpi $(EXES_ALL_MPI)
 
 .cpp.o:
-	$(CC) $(CFLAGS) $(INCLUDES) -o $@ -c $<
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) -o $@ -c $<
 .cpp.o_D:
 	$(CC) $(CFLAGS_D) $(INCLUDES) -o $@ -c $<
 .c.o:
-	$(CC) $(CFLAGS) $(INCLUDES) -o $@ -c $<
+	$(CXX) $(CXXFLAGS) $(CPPFLAGS) -o $@ -c $<
 .c.o_D:
 	$(CC) $(CFLAGS_D) $(INCLUDES) -o $@ -c $<
 clean:
diff --git a/src/util/Makefile~ b/src/util/Makefile~
deleted file mode 100644
index 7277dc2..0000000
--- a/src/util/Makefile~
+++ /dev/null
@@ -1,92 +0,0 @@
-BIG_INLINE_LIMIT = 20000
-CFLAGS1 = 
-
-CC = g++
-
-SRC_DIR  = ../
-UTIL_DIR = $(SRC_DIR)/util/
-BED_DIR  = $(SRC_DIR)/bed/
-MAF_DIR  = $(SRC_DIR)/maf/
-PHAST_DIR = $(SRC_DIR)/phast/
-
-INCLUDES  = -I$(SRC_DIR)
-
-CFLAGS = -O3 -g3 -pg -Wall $(INCLUDES)                       \
-	 -finline-limit=$(BIG_INLINE_LIMIT) $(CFLAGS1)
-LIBS = -lz
-CFLAGS_D = -O0 -g3 -pg -Wall -DDEBUG                         \
-	   -fkeep-inline-functions -fkeep-static-consts      \
-	   -finline-limit=0 -fno-default-inline $(CFLAGS1)
-LIBS_D = -lz
-
-.PHONY: clean depend all util bbls all_mpi util_mpi
-.SUFFIXES: .cpp .hpp .o .o_D
-all: all_
-
-SRCS0    = $(UTIL_DIR)/alpha.cpp $(UTIL_DIR)/opt.cpp
-SRCS1    = $(PHAST_DIR)/vector.c $(PHAST_DIR)/matrix.c $(PHAST_DIR)/numerical_opt.c
-EXES_ALL =
-SRCS_ALL =
-include run_tn93_model.mk
-include run_score_model_gamma.mk
-include run_tree_prob.mk
-include run_phybls_em_train_qsub.mk
-include run_phybls_em_train_qsub_worker.mk
-include run_phybls_em_train.mk
-include run_rev_model.mk
-include run_rev_break_dist.mk
-include run_rev_break_genome.mk
-include run_tdur_nmut.mk
-include run_tdur_nmut_genome.mk
-include run_tdur_pwm_match.mk
-include run_compare_em_grad.mk
-include run_tdur_pwm_match_hist.mk
-include run_tdur_pwm_match_hist_mask.mk
-include run_collect_minmax_tdur.mk
-include run_sample_fdur_nmut.mk
-include run_print_fdur_nmut.mk
-
-ifneq ($(strip $(SCBDSERV)),)
-  MAGI=true
-endif
-ifeq ($(strip $(MAGI)),)
-  MPICC = mpicxx
-else
-  MPICC = mpic++
-  LIBS += -lpthread
-endif
-MPICC = mpic++
-EXES_ALL_MPI =
-SRCS_ALL_MPI =
-
-all_: util $(EXES_ALL)
-
-all_mpi: util_mpi $(EXES_ALL_MPI)
-
-util:
-	(cd ../util/; make)
-util_mpi:
-	(cd ../util/; make all_mpi)
-
-.cpp.o:
-	$(CC) $(CFLAGS) $(INCLUDES) -o $@ -c $<
-.cpp.o_D:
-	$(CC) $(CFLAGS_D) $(INCLUDES) -o $@ -c $<
-.c.o:
-	$(CC) $(CFLAGS) $(INCLUDES) -o $@ -c $<
-.c.o_D:
-	$(CC) $(CFLAGS_D) $(INCLUDES) -o $@ -c $<
-clean:
-	rm -rf core {.,*}/{*.o,*.o_*,*.da,*.bb,*.bbg,*.gcda,*.gcno,*.gcov,*.stackdump,*.exe}
-SUFLIST = o o_D 
-depend:
-	for opt in $(SUFLIST);                      \
-	do touch make_deps_$$opt.mk;                \
-	   makedepend -o.$$opt -fmake_deps_$$opt.mk \
-	   -Y -- $(CFLAGS_D) $(INCLUDES) --         \
-	   $(SRCS_ALL) $(SRCS_ALL_MPI);             \
-	done
-include make_deps_o.mk
-include make_deps_o_D.mk
-
-
