--- clipper/fastq-stats.cpp
+++ clipper/fastq-stats.cpp
@@ -180,14 +180,14 @@
 	while ( (c = getopt (argc, argv, "?DdL:g:x:b:c:w:s:h")) != -1) {
 		switch (c) {
 			case 'c': cyclemax = atoi(optarg); break;
-			case 'D': ++nodup; break;
-			case 'd': ++debug; break;
+			case 'D': nodup = 1; break;
+			case 'd': debug = 1; break;
 			case 'w': window = atoi(optarg); break;
 			case 's': show_max = atoi(optarg); break;
-			case 'x': fastx_outfile = optarg; ++fastx; break;
-			case 'b': brkdown_outfile = optarg; ++brkdown; break;
-			case 'L': ++len_hist; lenhist_outfile = optarg; break;
-			case 'g': gc_outfile = optarg; ++gc; break;
+			case 'x': fastx_outfile = optarg; fastx = 1; break;
+			case 'b': brkdown_outfile = optarg; brkdown = 1; break;
+			case 'L': len_hist = 1; lenhist_outfile = optarg; break;
+			case 'g': gc_outfile = optarg; gc = 1; break;
 			case 'h': usage(stdout); return 0;
 			case '?':
 					  if (!optopt) {
--- clipper/sam-stats.cpp
+++ clipper/sam-stats.cpp
@@ -224,8 +224,8 @@
 
     while ( (c = getopt_long(argc, argv, "?BzArR:Ddx:MhS:", long_options, &long_index)) != -1) {
                 switch (c) {
                 case 'd': ++debug; break;                                       // increment debug level
-                case 'D': ++trackdup; break;
+                case 'D': trackdup = 1; break;
                 case 'B': inbam=1; break;
                 case 'A': max_chr=1000000; break;                               // max chrom
                 case 'R': rnafile=optarg;                                       // pass through
