diff --git a/setup.py b/setup.py
index b597f1c..16ce5d1 100755
--- a/setup.py
+++ b/setup.py
@@ -58,7 +58,7 @@ else:
     install_requires = ['numpy', "numexpr", 'matplotlib>=1.4.3',
         'pandas>=0.19', 'easydev>=0.9.34', 'scipy', "colormap>=1.0",
         'jinja2', 'statsmodels', "scikit-learn", "reports>=0.3.0",
-        "biokit>=0.4", "colorlog", "xlrd"],
+        ],
 
 
 packages = find_packages()

