diff --git a/Makefile b/Makefile
index b8d6a83..a9e1ffb 100644
--- a/Makefile
+++ b/Makefile
@@ -11,8 +11,8 @@ WARNINGS=-Wall -Wextra -Wno-char-subscripts \
         -Wno-missing-braces -Wno-unknown-pragmas
 		# -Wduplicated-branches -Wdangling-else  # -Wsuggest-attribute=malloc   # -Wconversion
 EXTRA?=
-INCPLUS?=
-EXTRA_LD?=
+INCPLUS?=-I$(PREFIX)/include
+EXTRA_LD?=-L$(PREFIX)/lib
 DBG:=
 OS:=$(shell uname)
 FLAGS=
