diff --git a/bonsai/clhash/src/clhash.c b/bonsai/clhash/src/clhash.c
index 63e0fb0..885c82c 100644
--- a/bonsai/clhash/src/clhash.c
+++ b/bonsai/clhash/src/clhash.c
@@ -2,7 +2,11 @@
 
 #include <assert.h>
 #include <string.h>
+#ifdef __ARM_NEON
+#include "sse2neon.h"
+#else
 #include <x86intrin.h>
+#endif
 
 #ifdef __WIN32
 #define posix_memalign(p, a, s) (((*(p)) = _aligned_malloc((s), (a))), *(p) ?0 :errno)
