From 0903111de19d7a0345cd42cc945666a3d8afa258 Mon Sep 17 00:00:00 2001
From: Marcel Bargull <marcel.bargull@udo.edu>
Date: Thu, 4 Mar 2021 15:54:49 +0100
Subject: [PATCH] Consider LDFLAGS while linking

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index a82bd22..cc230aa 100644
--- a/Makefile
+++ b/Makefile
@@ -40,7 +40,7 @@ endif
 all:$(PROG)
 
 unimap:main.o libunimap.a
-		$(CC) $(CFLAGS) main.o -o $@ -L. -lunimap $(LIBS)
+		$(CC) $(CFLAGS) main.o -o $@ $(LDFLAGS) -L. -lunimap $(LIBS)
 
 libunimap.a:$(OBJS)
 		$(AR) -csru $@ $(OBJS)
