diff --git a/Makefile b/Makefile
index b5da88a..86a3778 100644
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,6 @@
-CC=			gcc
-CFLAGS=		-g -Wall -Wc++-compat -std=c99 -O2
-CPPFLAGS=
+CC=			$(CC)
+CFLAGS=		-g -Wall -Wc++-compat -std=c99 -O3
+CPPFLAGS=	-I$(PREFIX)/include
 INCLUDES=	-I.
 OBJS=		kalloc.o gfa-base.o gfa-io.o gfa-aug.o gfa-sub.o gfa-asm.o gfa-util.o \
 			gfa-bbl.o gfa-ed.o gfa-sql.o gfa-gt.o
@@ -27,13 +27,13 @@ libgfa1.a:$(OBJS)
 		$(AR) -csr $@ $(OBJS)
 
 gfa-chk:gfa-chk.l
-		lex $< && $(CC) -O2 lex.yy.c -o $@
+		lex $< && $(CC) -O3 lex.yy.c -o $@
 
 clean:
 		rm -fr gmon.out *.o a.out $(EXE) *~ *.a *.dSYM session* gfa-chk
 
 depend:
-		(LC_ALL=C; export LC_ALL; makedepend -Y -- $(CFLAGS) $(DFLAGS) -- *.c)
+		(LC_ALL=en_US.UTF-8; export LC_ALL; makedepend -Y -- $(CFLAGS) $(DFLAGS) -- *.c)
 
 # DO NOT DELETE
 
