commit b94f4866cc0c5597acdb0e34e17a3165da10f3ce
Author: Nicola Soranzo <nicola.soranzo@earlham.ac.uk>
Date:   Mon Mar 1 01:28:36 2021 +0000

    Remove runtime dependency on `setuptools`

diff --git a/lib/bx/misc/seeklzop.py b/lib/bx/misc/seeklzop.py
index 39fa538..7124f5b 100644
--- a/lib/bx/misc/seeklzop.py
+++ b/lib/bx/misc/seeklzop.py
@@ -6,8 +6,6 @@ import struct
 from io import BytesIO
 
 try:
-    import pkg_resources
-    pkg_resources.require('python_lzo')
     import lzo
 except Exception:
     pass
diff --git a/lib/bx/motif/logo/__init__.py b/lib/bx/motif/logo/__init__.py
index 0cf55ed..ceae03b 100644
--- a/lib/bx/motif/logo/__init__.py
+++ b/lib/bx/motif/logo/__init__.py
@@ -1,7 +1,7 @@
+import os.path
 from io import StringIO
 from string import Template
 
-import pkg_resources
 from numpy import (
     ceil,
     log2,
@@ -52,7 +52,10 @@ def eps_logo(matrix, base_width, height, colors=DNA_DEFAULT_COLORS):
     alphabet = matrix.sorted_alphabet
     rval = StringIO()
     # Read header ans substitute in width / height
-    header = Template(pkg_resources.resource_string(__name__, "template.ps"))
+    template_path = os.path.join(os.path.dirname(__file__), 'template.ps')
+    with open(template_path) as fh:
+        template_str = fh.read()
+    header = Template(template_str)
     rval.write(header.substitute(
         bounding_box_width=ceil(base_width * matrix.width) + PAD,
         bounding_box_height=ceil(height) + PAD))
diff --git a/setup.py b/setup.py
index c7f5231..4471d57 100644
--- a/setup.py
+++ b/setup.py
@@ -113,8 +113,6 @@ command_classes['build_py'] = build_py
 
 # Use epydoc if found
 try:
-    import pkg_resources
-    pkg_resources.require("epydoc")
     import epydoc.cli
     import os
     import os.path
