From 0541565833c3b93215aa0b3b8e5c864eb2dd944a Mon Sep 17 00:00:00 2001
From: Mervin Fansler <mervin.fansler@bric.ku.dk>
Date: Mon, 3 Feb 2025 16:09:39 +0100
Subject: [PATCH] constrain vcf dep

---
 souporcell/Cargo.toml | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/souporcell/Cargo.toml b/souporcell/Cargo.toml
index 3d9b3ef..8f1e40c 100644
--- a/souporcell/Cargo.toml
+++ b/souporcell/Cargo.toml
@@ -11,5 +11,5 @@ rand = "0.7.0"
 itertools = "*"
 statrs = "*"
 rayon = "*"
-vcf = "*"
+vcf = "0.1.0"
 flate2 = "*"
