diff --git a/src/mcmc.c b/src/mcmc.c
index e0412b0..b843d0c 100644
--- a/src/mcmc.c
+++ b/src/mcmc.c
@@ -1618,7 +1618,7 @@ void MCMC_Root_Time(t_tree *tree, int print)
       /* if(tree->aux_tree != NULL) */
         /* { */
           /* PhyML_Printf("\n."); */
-          /* PhyML_Printf("\n. %c [%4d] \n. t: %1f->%1f t1:%f t2: %f \n. r1: %10f->%10f r2: %10f->%10f \n. R: %12f alnL:%10f->%10f[%d] \n. tlnL: %10f->%10f[%d] \n. glnL: %10f->%10f[%d] \n. rlnL: %10f->%10f[%d] \n. tune: %10f hr: %f ratio: %10f K: %10f rate_failed: %d", */
+          /* PhyML_Printf("\n. %c [%4d] \n. t: %1f->%1f t1:%f t2: %f \n. r1: %10f->%10f r2: %10f->%10f \n. R: %12f alnL:%10f->%10f[%d] \n. tlnL: %10f->%10f[%d] ",
           /*              (tree->aux_tree != NULL) ? '*' : 'X', */
           /*              tree->mcmc->run, */
           /*              t0_cur,t0_new, */
@@ -1628,7 +1628,11 @@ void MCMC_Root_Time(t_tree *tree, int print)
           /*              tree->rates->norm_fact, */
           /*              cur_lnL_seq,new_lnL_seq,tree->eval_alnL, */
           /*              cur_lnL_time,new_lnL_time,tree->eval_glnL, */
-          /*              cur_lnL_loc,new_lnL_loc,tree->eval_glnL, */
+#ifdef PHYREX
+	  /* PhyML_Printf("\n. glnL: %10f->%10f[%d] ",
+          /*              cur_lnL_loc,new_lnL_loc,tree->eval_glnL);
+#endif
+	  /* PhyML_Printf("\n. rlnL: %10f->%10f[%d] \n. tune: %10f hr: %f [%f,%f] [%f,%f,%f] ratio: %10f K: %10f failed: %d",
           /*              cur_lnL_rate,new_lnL_rate,tree->eval_rlnL, */
           /*              tune, */
           /*              hr, */
