diff --git a/configure.ac b/configure.ac
index cd08930..b320505 100644
--- a/configure.ac
+++ b/configure.ac
@@ -119,52 +119,28 @@ AM_CONDITIONAL([WANT_WIN], [test "$enable_win" = yes])
 
 AC_ARG_ENABLE([phytime],[AS_HELP_STRING([--enable-phytime],[Compile PhyTime])],[phytime=yes],[phytime=no])
 AM_CONDITIONAL([WANT_PHYTIME], [test "$phytime" = yes])
-if test "$phytime" = yes; then
-   AC_DEFINE([PHYTIME],[1],[PHYTIME tag on])
-fi
 
 AC_ARG_ENABLE([phyml],[AS_HELP_STRING([--enable-phyml],[Compile PhyML])],[phyml=yes],[phyml=no])
 AM_CONDITIONAL([WANT_PHYML], [test "$phyml" = yes])
-if test "$phyml" = yes; then
-   AC_DEFINE([PHYML],[1],[PHYML tag on])
-fi
 
 
 AC_ARG_ENABLE([rwrap],[AS_HELP_STRING([--enable-rwrap],[Compile Rwrap])],[rwrap=yes],[rwrap=no])
 AM_CONDITIONAL([WANT_RWRAP], [test "$rwrap" = yes])
-if test "$rwrap" = yes; then
-   AC_DEFINE([RWRAP],[1],[RWRAP tag on])
-fi
 
 AC_ARG_ENABLE([rf],[AS_HELP_STRING([--enable-rf],[Compile RF])],[rf=yes],[rf=no])
 AM_CONDITIONAL([WANT_RF], [test "$rf" = yes])
-if test "$rf" = yes; then
-   AC_DEFINE([RF],[1],[RF tag on])
-fi
 
 AC_ARG_ENABLE([test],[AS_HELP_STRING([--enable-test],[Compile test])],[test=yes],[test=no])
 AM_CONDITIONAL([WANT_TEST], [test "$test" = yes])
-if test "$test" = yes; then
-   AC_DEFINE([TEST],[1],[TEST tag on])
-fi
 
 AC_ARG_ENABLE([evolve],[AS_HELP_STRING([--enable-evolve],[Compile evolve])],[evolve=yes],[evolve=no])
 AM_CONDITIONAL([WANT_EVOLVE], [test "$evolve" = yes])
-if test "$evolve" = yes; then
-   AC_DEFINE([EVOLVE],[1],[EVOLVE tag on])
-fi
 
 AC_ARG_ENABLE([phyrex],[AS_HELP_STRING([--enable-phyrex],[Compile phyrex])],[phyrex=yes],[phyrex=no])
 AM_CONDITIONAL([WANT_PHYREX], [test "$phyrex" = yes])
-if test "$phyrex" = yes; then
-   AC_DEFINE([PHYREX],[1],[PHYREX tag on])
-fi
 
 AC_ARG_ENABLE([phyrexsim],[AS_HELP_STRING([--enable-phyrexsim],[Compile phyrexsim])],[phyrexsim=yes],[phyrexsim=no])
 AM_CONDITIONAL([WANT_PHYREXSIM], [test "$phyrexsim" = yes])
-if test "$phyrexsim" = yes; then
-   AC_DEFINE([PHYREXSIM],[1],[PHYREXSIM tag on])
-fi
 
 dnl AS_IF([test "x$enable_phyrex" = "xyes"],[PKG_CHECK_MODULES([GTK], [gtk+-3.0])])
 dnl AS_IF([test "x$enable_phyrex" = "xyes"],[PKG_CHECK_MODULES([CAIRO], [cairo])])
@@ -175,27 +151,6 @@ dnl AM_PATH_GTK_3_0(,,AC_MSG_ERROR(windoe-default needs GTK+-3))
 
 AC_ARG_ENABLE([date],[AS_HELP_STRING([--enable-date],[Compile date])],[date=yes],[date=no])
 AM_CONDITIONAL([WANT_DATE], [test "$date" = yes])
-if test "$date" = yes; then
-   AC_DEFINE([DATE],[1],[DATE tag on])
-fi
-
-if test "$phytime" = no; then
-if test "$rwrap" = no; then
-if test "$test" = no; then
-if test "$rf" = no; then
-if test "$evolve" = no; then
-if test "$phyrex" = no; then
-if test "$phyrexsim" = no; then
-if test "$date" = no; then
-   AC_DEFINE([PHYML],[1],[PHYML tag on])
-fi
-fi
-fi
-fi
-fi
-fi
-fi
-fi
 
 AC_CONFIG_FILES([Makefile src/Makefile doc/Makefile])
 AC_OUTPUT
