diff -ur phyml-3.3.20170530.orig/src/utilities.h phyml-3.3.20170530/src/utilities.h
--- phyml-3.3.20170530.orig/src/utilities.h	2017-05-23 00:38:26.000000000 -0600
+++ phyml-3.3.20170530/src/utilities.h	2017-06-21 17:49:32.000000000 -0600
@@ -28,8 +28,6 @@
 #include <float.h>
 #include <assert.h>
 #include <stdbool.h>
-/* #include <malloc/malloc.h> */
-#include <malloc.h>
 
 #if (defined(__AVX__))
 #include <xmmintrin.h>
