--- a/src/mcmc.c
+++ b/src/mcmc.c
@@ -1565,6 +1565,7 @@
 
           RATES_Update_One_Edge_Length(tree->e_root,tree);
           
+#ifdef PHYREX
           if(tree->eval_glnL == YES)
             {
               if(VELOC_Is_Integrated_Velocity(tree->mmod) == YES)
@@ -1577,6 +1578,7 @@
               else
                 new_lnL_loc = (tree->mmod->use_locations == YES) ? LOCATION_Lk(NULL,tree) : 0.0;
             }
+#endif
 
           if(tree->eval_tlnL == YES) new_lnL_time = TIMES_Lk(tree);              
           if(tree->eval_rlnL == YES) new_lnL_rate = RATES_Lk(tree);
