--- a/src/mcmc.c	2022-05-18 13:29:35.000000000 -0800
+++ b/src/mcmc.c	2022-05-18 13:42:37.000000000 -0800
@@ -1469,7 +1469,7 @@
       /* if(tree->aux_tree != NULL) */
         {
           PhyML_Printf("\n.");
-          PhyML_Printf("\n. %c [%4d] \n. t: %1f->%1f t1:%f t2: %f \n. r1: %10f->%10f r2: %10f->%10f \n. R: %12f alnL:%10f->%10f[%d] \n. tlnL: %10f->%10f[%d] \n. glnL: %10f->%10f[%d] \n. rlnL: %10f->%10f[%d] \n. tune: %10f hr: %f [%f,%f] [%f,%f,%f] ratio: %10f K: %10f failed: %d",
+          PhyML_Printf("\n. %c [%4d] \n. t: %1f->%1f t1:%f t2: %f \n. r1: %10f->%10f r2: %10f->%10f \n. R: %12f alnL:%10f->%10f[%d] \n. tlnL: %10f->%10f[%d] ",
                        (tree->aux_tree != NULL) ? '*' : 'X',
                        tree->mcmc->run,
                        t0_cur,t0_new,
@@ -1478,8 +1478,12 @@
                        r2_cur,r2_new,
                        tree->rates->norm_fact,
                        cur_lnL_seq,new_lnL_seq,tree->eval_alnL,
-                       cur_lnL_time,new_lnL_time,tree->eval_glnL,
-                       cur_lnL_loc,new_lnL_loc,tree->eval_glnL,
+                       cur_lnL_time,new_lnL_time,tree->eval_glnL);
+#ifdef PHYREX
+	  PhyML_Printf("\n. glnL: %10f->%10f[%d] ",
+			cur_lnL_loc,new_lnL_loc,tree->eval_glnL);
+#endif
+	  PhyML_Printf("\n. rlnL: %10f->%10f[%d] \n. tune: %10f hr: %f [%f,%f] [%f,%f,%f] ratio: %10f K: %10f failed: %d",
                        cur_lnL_rate,new_lnL_rate,tree->eval_rlnL,
                        tune,
                        hr,
