--- /home/sjanssen/Desktop/sepp/STEFAN/sepp-0.4.0/setup.py	2019-03-31 07:43:25.000000000 +0200
+++ setup.py	2019-04-03 12:12:30.701516881 +0200
@@ -227,7 +227,7 @@
       author_email="smirarab@gmail.com, namphuon@cs.utah.edu",
 
       license="General Public License (GPL)",
-      install_requires=["dendropy >= 4.0.0"],
+      install_requires=[],
       provides=["sepp"],
       scripts=["run_sepp.py", 'run_tipp.py', 'run_upp.py', 'run_abundance.py',
                "split_sequences.py", "run_tipp_tool.py"],
