--- sepp-package/sepp/default.main.config	2019-03-18 10:14:42.363446570 +0100
+++ sepp-package/sepp/default.main.config.conda	2019-04-04 10:10:22.171485197 +0200
@@ -1,20 +1,20 @@
 [pplacer]
-path=~/pplacer
+path=/opt/anaconda1anaconda2anaconda3/bin/pplacer
 
 [hmmalign]
-path=~/hmmalign
+path=/opt/anaconda1anaconda2anaconda3/bin/hmmalign
 
 [hmmsearch]
-path=~/hmmsearch
+path=/opt/anaconda1anaconda2anaconda3/bin/hmmsearch
 piped=False
 elim=10000
 filters=True
 
 [hmmbuild]
-path=~/hmmbuild
+path=/opt/anaconda1anaconda2anaconda3/bin/hmmbuild
 
 [jsonmerger]
-path=~/seppJsonMerger.jar
+path=/opt/anaconda1anaconda2anaconda3/bin/seppJsonMerger.jar
 
 [exhaustive]
 strategy = centroid
