--- setup.py    2016-01-19 17:32:01.000000000 +0100
+++ setup.new.py        2016-02-03 15:48:54.279205749 +0100
@@ -20,10 +20,10 @@
 # Control which 3rd party packages should be installed
 # depending on the python version
 install_requires = [
-    'PyVCF',
-    'setuptools', # needed during execution to load pkg_resources
-    'psutil',
-    'Biopython',
+#    'PyVCF',
+#    'setuptools', # needed during execution to load pkg_resources
+#    'psutil',
+#    'Biopython',
 ]

 # Below needed for Python 2.6

