diff -urN edirect.orig/cmd/build.sh edirect/cmd/build.sh
--- edirect.orig/cmd/build.sh	2022-11-12 07:25:55
+++ edirect/cmd/build.sh	2024-03-08 22:13:47
@@ -13,6 +13,7 @@
   Darwin-*arm* )           platform=Silicon ;;
   CYGWIN_NT-* | MINGW*-* ) platform=CYGWIN_NT ;;
   Linux-*arm* )            platform=ARM ;;
+  Linux-*aarch64* )        platform=ARM64 ;;
   * )                      platform=UNSUPPORTED ;;
 esac
 
@@ -94,7 +95,7 @@
 fi
 
 # erase any existing executables in current directory
-for plt in Darwin Silicon Linux CYGWIN_NT ARM
+for plt in Darwin Silicon Linux CYGWIN_NT ARM ARM64
 do
   rm -f *.$plt
 done
@@ -103,6 +104,7 @@
 mods="darwin amd64 Darwin \
       darwin arm64 Silicon \
       linux amd64 Linux \
+      linux arm64 ARM64 \
       windows 386 CYGWIN_NT \
       linux arm ARM"
 
@@ -126,13 +128,13 @@
   if [ "$cleanup" = true ]
   then
     # remove old executables from target
-    for plt in Darwin Silicon Linux CYGWIN_NT ARM
+    for plt in Darwin Silicon Linux CYGWIN_NT ARM ARM64
     do
       rm -f $target/*.$plt
     done
   fi
   # copy new executables to target
-  for plt in Darwin Silicon Linux CYGWIN_NT ARM
+  for plt in Darwin Silicon Linux CYGWIN_NT ARM ARM64
   do
     for exc in xtract rchive transmute
     do
@@ -145,7 +147,7 @@
 fi
 
 # erase any remaining executables after compiling
-for plt in Darwin Silicon Linux CYGWIN_NT ARM
+for plt in Darwin Silicon Linux CYGWIN_NT ARM ARM64
 do
   rm -f *.$plt
 done
diff -urN edirect.orig/extern/build.sh edirect/extern/build.sh
--- edirect.orig/extern/build.sh	2024-03-04 11:00:38
+++ edirect/extern/build.sh	2024-03-08 22:14:43
@@ -13,6 +13,7 @@
   Darwin-*arm* )           platform=Silicon ;;
   CYGWIN_NT-* | MINGW*-* ) platform=CYGWIN_NT ;;
   Linux-*arm* )            platform=ARM ;;
+  Linux-*aarch64* )        platform=ARM64 ;;
   * )                      platform=UNSUPPORTED ;;
 esac
 
@@ -51,7 +52,7 @@
 fi
 
 # erase any existing executables in current directory
-for plt in Darwin Silicon Linux CYGWIN_NT ARM
+for plt in Darwin Silicon Linux CYGWIN_NT ARM ARM64
 do
   rm -f *.$plt
 done
@@ -64,7 +65,7 @@
 done
 
 # will be using "go run", erase executables after test complication
-for plt in Darwin Silicon Linux CYGWIN_NT ARM
+for plt in Darwin Silicon Linux CYGWIN_NT ARM ARM64
 do
   rm -f *.$plt
 done
diff -urN edirect.orig/rchive edirect/rchive
--- edirect.orig/rchive	2022-01-29 07:25:44
+++ edirect/rchive	2024-03-08 22:16:43
@@ -62,6 +62,7 @@
   Darwin-*arm* )           platform=Silicon ;;
   CYGWIN_NT-* | MINGW*-* ) platform=CYGWIN_NT ;;
   Linux-*arm* )            platform=ARM ;;
+  Linux-*aarch64* )        platform=ARM64 ;;
   * )                      platform=UNSUPPORTED ;;
 esac
 compiled=$0."$platform"
diff -urN edirect.orig/transmute edirect/transmute
--- edirect.orig/transmute	2023-02-22 07:25:33
+++ edirect/transmute	2024-03-08 22:16:43
@@ -402,6 +402,7 @@
   Darwin-*arm* )           platform=Silicon ;;
   CYGWIN_NT-* | MINGW*-* ) platform=CYGWIN_NT ;;
   Linux-*arm* )            platform=ARM ;;
+  Linux-*aarch64* )        platform=ARM64 ;;
   * )                      platform=UNSUPPORTED ;;
 esac
 compiled=$0."$platform"
diff -urN edirect.orig/xtract edirect/xtract
--- edirect.orig/xtract	2024-02-28 09:25:34
+++ edirect/xtract	2024-03-08 22:16:43
@@ -110,6 +110,7 @@
   Darwin-*arm* )           platform=Silicon ;;
   CYGWIN_NT-* | MINGW*-* ) platform=CYGWIN_NT ;;
   Linux-*arm* )            platform=ARM ;;
+  Linux-*aarch64* )        platform=ARM64 ;;
   * )                      platform=UNSUPPORTED ;;
 esac
 compiled=$0."$platform"
