--- a/transmute
+++ b/transmute
@@ -349,7 +349,7 @@ case "$osname-$cputype" in
   Darwin-x86_64 )          platform=Darwin ;;
   Darwin-*arm* )           platform=Silicon ;;
   CYGWIN_NT-* | MINGW*-* ) platform=CYGWIN_NT ;;
-  Linux-*arm* )            platform=ARM ;;
+  Linux-*arm* | Linux-*aarch64* ) platform=ARM ;;
   * )                      platform=UNSUPPORTED ;;
 esac
 compiled=$0."$platform"
