--- setup.sh	2021-06-30 00:25:36.000000000 +0200
+++ setup.sh.new	2021-08-27 10:50:47.884354249 +0200
@@ -26,7 +26,8 @@

 if [ -f xtract."$platform" ]
 then
-  chmod +x xtract."$platform"
+  mv xtract."$osname" xtract
+  chmod +x xtract
 else
   echo "Unable to download xtract executable."
 fi
@@ -108,22 +109,4 @@
   fi
 fi

-if [ -z "$prfx" ]
-then
-  echo ""
-  if [ "$osname" = "Darwin" ]
-  then
-    echo "or manually edit the PATH variable assignments in your .$target and .$zarget files."
-  else
-    echo "or manually edit the PATH variable assignment in your .$target file."
-  fi
-  echo ""
-  echo "Would you like to do that automatically now? [y/N]"
-  read response
-  case "$response" in
-    [Yy]*      ) . $advice; echo "OK, done." ;;
-    [Nn]* | '' ) echo "Holding off, then." ;;
-    *          ) echo "Conservatively taking that as a no." ;;
-  esac
-fi
 rm $advice
