--- setup.sh	2019-10-18 19:50:47.000000000 +0200
+++ setup.sh.new	2020-02-03 08:32:53.591369832 +0100
@@ -62,7 +62,8 @@
 
 if [ -f xtract."$platform" ]
 then
-  chmod +x xtract."$platform"
+  mv xtract."$osname" xtract
+  chmod +x xtract
 else
   echo "Unable to download xtract executable."
 fi
@@ -113,17 +114,4 @@
     | tee $advice
 fi
 
-if [ -z "$prfx" ]
-then
-echo ""
-echo "or manually edit the PATH variable assignment in your .bash_profile file."
-echo ""
-echo "Would you like to do that automatically now? [y/N]"
-read response
-case "$response" in
-  [Yy]*      ) . $advice; echo "OK, done." ;;
-  [Nn]* | '' ) echo "Holding off, then." ;;
-  *          ) echo "Conservatively taking that as a no." ;;
-esac
-fi
 rm $advice
