diff --git i/src/Makefile w/src/Makefile
index 8844e0a..3248657 100644
--- i/src/Makefile
+++ w/src/Makefile
@@ -1,9 +1,9 @@
 LAST_CC=g++
-CC=gcc
-CXX=g++
+CC?=gcc
+CXX?=g++
 COMMON_FLAGS=-Wall -Wextra -pedantic -fopenmp -O3
 INCLUDES = -I ../ext/boost
-CC_FLAGS=${COMMON_FLAGS} 
+CC_FLAGS?=${COMMON_FLAGS}
 LDLIBS	=-lm -fopenmp
 
 ifeq ($(PROFILE), Yes)
@@ -69,18 +69,18 @@ OBJS += $(CPP_SRC:%.cpp=%.o)
 all:ghostx
 
 ghostx: $(OBJS)
-	$(LAST_CC) -o $@ $(OBJS) $(LDFLAGS) $(LDLIBS)
+	$(CXX) -std=c++14 -o $@ $(OBJS) $(LDFLAGS) $(LDLIBS)
 
 .c.o:
 	$(CC) -c $(CC_FLAGS) $< -o $@  $(INCLUDES)
 
 .cpp.o:
-	$(CXX) -c $(CC_FLAGS) $< -o $@  $(INCLUDES)
+	$(CXX) -std=c++14 -c $(CC_FLAGS) $< -o $@  $(INCLUDES)
 
 
 
 .PHONY: clean
 clean:
-	rm ghostx
 	rm -f *.o
 	rm -f *~
+	rm ghostx
