diff --git a/CBlastParser.h b/CBlastParser.h
index 28eaa6f..5d082ac 100644
--- a/CBlastParser.h
+++ b/CBlastParser.h
@@ -539,7 +539,7 @@ class CBlast_parser
   void add(CBlastHit* p)
   {
     lbh.push_back(p);
-    bh_map.insert(bh_map.end(), std::make_pair<faststring, CBlastHit*>(p->get_query_seq_name(), p));
+    bh_map.insert(bh_map.end(), std::make_pair(p->get_query_seq_name(), p));
   }
 
  public:
