--- lib/GI.pm	2018-06-29 20:53:50.113085044 +0200
+++ lib/GI.pm	2018-06-29 21:04:05.041345063 +0200
@@ -4355,7 +4355,7 @@
         if(! -d $CTL_OPT{out_base});
 
    #--make sure repbase is installed
-   if($CTL_OPT{model_org}){
+   if($CTL_OPT{model_org} and !defined($ENV{'REPEATMASKER_LIB_DIR'})){
        my $exe = Cwd::abs_path($CTL_OPT{RepeatMasker});
        my ($lib) = $exe =~ /(.*\/)RepeatMasker$/;
        die "ERROR: Could not determine if RepBase is installed\n" if(! $lib);
