--- gtdbtk/main.py	2019-12-05 13:18:30.000000000 +1000
+++ gtdbtk/main_new.py	2020-03-26 14:58:29.381043987 +1000
@@ -529,6 +529,7 @@
         options : argparse.Namespace
             The CLI arguments input by the user.
         """
+        self.logger.warning('You are not running the most recent version of GTDB-Tk!')

         # Stop processing if python 2 is being used.
         if sys.version_info.major < 3:
