--- setup.py    2016-11-26 23:23:14.000000000 +0100
+++ setup.py    2016-11-26 23:44:54.000000000 +0100
@@ -34,11 +34,10 @@

     scripts=['koeken/koeken.py', 'koeken/lefse_src/format_input.py', 'koeken/lefse_src/run_lefse.py', 'koeken/lefse_src/lefse.py', 'koeken/lefse_src/plot_cladogram.py', 'koeken/lefse_src/hclust2/hclust2.py', 'koeken/pretty_lefse.py'],

-    cmdclass={'install': CustomInstallPackages},

     packages=setuptools.find_packages(),

-    install_requires=['rpy2', 'argparse', 'pandas', 'biopython'],
+    install_requires=['rpy2', 'pandas', 'biopython'],

     classifiers=[
         'Development Status :: 2 - Pre-Alpha',
