diff --git a/reviewer/app/Workflow.cpp b/reviewer/app/Workflow.cpp
index 3954aca71..badd1b2e8 100644
--- a/reviewer/app/Workflow.cpp
+++ b/reviewer/app/Workflow.cpp
@@ -21,6 +21,7 @@
 #include "Workflow.hh"
 
 #include <set>
+#include <sstream>
 
 #include <boost/algorithm/string.hpp>
 
diff --git a/reviewer/thirdparty/graph-tools/src/graphcore/GraphCoordinates.cpp b/reviewer/thirdparty/graph-tools/src/graphcore/GraphCoordinates.cpp
index 2550eda8b..622243c4e 100644
--- a/reviewer/thirdparty/graph-tools/src/graphcore/GraphCoordinates.cpp
+++ b/reviewer/thirdparty/graph-tools/src/graphcore/GraphCoordinates.cpp
@@ -22,6 +22,8 @@
 #include "graphutils/PairHashing.hh"
 
 #include <cassert>
+#include <algorithm>
+#include <limits>
 
 namespace graphtools
 {
