diff --git a/CMakeLists.txt b/CMakeLists.txt
index 022aa4d90..d63cae33e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -14,12 +14,6 @@ include(ExternalProject)
 set(installDir ${CMAKE_CURRENT_BINARY_DIR}/install)
 
 
-ExternalProject_Add(htslib
-	BUILD_IN_SOURCE YES
-	URL https://github.com/samtools/htslib/releases/download/1.10.2/htslib-1.10.2.tar.bz2
-	CONFIGURE_COMMAND ./configure --prefix=${installDir} --disable-libcurl
-	BUILD_COMMAND $(MAKE) lib-static
-)
 
 
 # Setup boost user config so that it uses the same c++ compiler as other components:
