diff --git a/setup.py b/setup.py
index e996012..ed7ffe8 100644
--- a/setup.py
+++ b/setup.py
@@ -1,8 +1,9 @@
-from setuptools import setup
+from setuptools import setup, find_namespace_packages
 
 setup(
     name="m-party",
-    version="1.0.0",
+    version="0.3.0",
+    packages=find_namespace_packages(where="."),
     py_modules=["m-party"],
     install_requires=["upimapi", 
                     "pyyaml", 
@@ -21,7 +22,7 @@ setup(
                     "tqdm"],
     entry_points={
         "console_scripts": [
-            "m-party=m-party:main",
+            "m-party = m-party:main",
         ],
     },
-)
\ No newline at end of file
+)
