--- cpp/variant_file_diff.cpp.old	2019-04-26 22:42:10.461876593 -0500
+++ cpp/variant_file_diff.cpp	2019-04-26 23:14:15.532072986 -0500
@@ -36,7 +36,7 @@
 	for (unsigned int ui=0; ui<meta_data.N_indv; ui++)
 		if (include_indv[ui] == true)
 		{
-			combined_individuals[meta_data.indv[ui]] = make_pair<int,int>((int)ui, -1);
+			combined_individuals[meta_data.indv[ui]] = make_pair((int)ui, -1);
 		}
 
 	for (unsigned int ui=0; ui<file2.meta_data.N_indv; ui++)
@@ -53,7 +53,7 @@
 				combined_individuals[indv_id].second = ui;
 			}
 			else
-				combined_individuals[indv_id] = make_pair<int,int>(-1, (int)ui);
+				combined_individuals[indv_id] = make_pair(-1, (int)ui);
 		}
 }
 
@@ -1491,9 +1491,9 @@
 							if ((phase1 == '|') && (phase2 == '|'))
 							{	// Calculate Phasing error (switch error)
 								N_phased_het_sites[indv_count]++;
-								file1_hap1 = make_pair<string,string>((string)prev_geno_file1[indv_count].first, (string)genotype1.first);
-								file1_hap2 = make_pair<string,string>((string)prev_geno_file1[indv_count].second, (string)genotype1.second);
-								file2_hap1 = make_pair<string,string>((string)prev_geno_file2[indv_count].first, (string)genotype2.first);
+								file1_hap1 = make_pair((string)prev_geno_file1[indv_count].first, (string)genotype1.first);
+								file1_hap2 = make_pair((string)prev_geno_file1[indv_count].second, (string)genotype1.second);
+								file2_hap1 = make_pair((string)prev_geno_file2[indv_count].first, (string)genotype2.first);
 
 								if ((file2_hap1 != file1_hap1) && (file2_hap1 != file1_hap2))
 								{	// Must be a switch error
@@ -1514,8 +1514,8 @@
 								}
 								prev_geno_file1[indv_count] = genotype1;
 								prev_geno_file2[indv_count] = genotype2;
-								prev_pos_file1[indv_count] = make_pair<string,int>(CHROM1,POS1);
-								prev_pos_file2[indv_count] = make_pair<string,int>(CHROM2,POS2);
+								prev_pos_file1[indv_count] = make_pair(CHROM1,POS1);
+								prev_pos_file2[indv_count] = make_pair(CHROM2,POS2);
 							}
 						}
 					}
