diff -ruN Disco-1.2/src/BuildGraph/src/Common.h Disco-1.2/src/BuildGraph/src/Common.h
--- Disco-1.2/src/BuildGraph/src/Common.h	2021-04-07 17:23:03.432443041 +0200
+++ Disco-1.2/src/BuildGraph/src/Common.h	2021-04-07 17:25:14.907339001 +0200
@@ -65,7 +65,7 @@
 // Print which function is currently executing. Only for functions that take long time
 
 
-#define SSTR( x ) dynamic_cast< std::ostringstream & >(( std::ostringstream() << std::dec << x ) ).str()
+#define SSTR( x ) std::to_string(x)
 
 // To keep time information of functions.
 #define CLOCKSTART INT64 mem_start = checkMemoryUsage(); double begin = omp_get_wtime(); cout<<"Currently in file: " << __FILE__ << " Function: "<< __FUNCTION__ << "()" << endl;
diff -ruN Disco-1.2/src/BuildGraphMPI/src/Common.h Disco-1.2/src/BuildGraphMPI/src/Common.h
--- Disco-1.2/src/BuildGraphMPI/src/Common.h	2021-04-07 17:23:03.436443008 +0200
+++ Disco-1.2/src/BuildGraphMPI/src/Common.h	2021-04-07 17:26:01.794949592 +0200
@@ -83,7 +83,7 @@
 // Print which function is currently executing. Only for functions that take long time
 
 
-#define SSTR( x ) dynamic_cast< std::ostringstream & >(( std::ostringstream() << std::dec << x ) ).str()
+#define SSTR( x ) std::to_string(x) 
 
 // To keep time information of functions.
 #define CLOCKSTART INT64 mem_start = checkMemoryUsage(); double begin = omp_get_wtime(); cout<<"Currently in file: " << __FILE__ << " Function: "<< __FUNCTION__ << "()" << endl;
diff -ruN Disco-1.2/src/BuildGraphMPIRMA/src/Common.h Disco-1.2/src/BuildGraphMPIRMA/src/Common.h
--- Disco-1.2/src/BuildGraphMPIRMA/src/Common.h	2021-04-07 17:23:03.436443008 +0200
+++ Disco-1.2/src/BuildGraphMPIRMA/src/Common.h	2021-04-07 17:26:24.738759765 +0200
@@ -69,7 +69,7 @@
 // Print which function is currently executing. Only for functions that take long time
 
 
-#define SSTR( x ) dynamic_cast< std::ostringstream & >(( std::ostringstream() << std::dec << x ) ).str()
+#define SSTR( x ) std::to_string(x)
 
 // To keep time information of functions.
 #define CLOCKSTART INT64 mem_start = checkMemoryUsage(); double begin = omp_get_wtime(); cout<<"Currently in file: " << __FILE__ << " Function: "<< __FUNCTION__ << "()" << endl;
diff -ruN Disco-1.2/src/SimplifyGraph/src/Config.h Disco-1.2/src/SimplifyGraph/src/Config.h
--- Disco-1.2/src/SimplifyGraph/src/Config.h	2021-04-07 17:23:03.436443008 +0200
+++ Disco-1.2/src/SimplifyGraph/src/Config.h	2021-04-07 19:59:14.117036640 +0200
@@ -44,7 +44,7 @@
 #include <sys/wait.h>
 #include <zlib.h>
 
-#define SSTR( x ) dynamic_cast< std::ostringstream & >(( std::ostringstream() << std::dec << x ) ).str()
+#define SSTR( x ) std::to_string(x)
 
 // Local headers
 #include "logcpp/log.h"
diff -ruN Disco-1.2/src/SimplifyGraph/src/OverlapGraphSimple.h Disco-1.2/src/SimplifyGraph/src/OverlapGraphSimple.h
--- Disco-1.2/src/SimplifyGraph/src/OverlapGraphSimple.h	2021-04-07 17:23:03.436443008 +0200
+++ Disco-1.2/src/SimplifyGraph/src/OverlapGraphSimple.h	2021-04-07 19:58:55.341146175 +0200
@@ -14,8 +14,7 @@
 #include "Config.h"
 #include "EdgeSimple.h"
 
-#define SSTR( x ) dynamic_cast< std::ostringstream & >(( std::ostringstream() << std::dec << x ) ).str()
-
+#define SSTR( x ) std::to_string(x)
 
 typedef vector<EdgeSimple *> t_edge_vec;	// vector of pointers to EdgeSimple
 class OverlapGraphSimple
