diff --git i/epik/include/epik/place.h w/epik/include/epik/place.h
index d348755..e78beea 100644
--- i/epik/include/epik/place.h
+++ w/epik/include/epik/place.h
@@ -7,7 +7,17 @@
 #include <i2l/phylo_kmer_db.h>
 #include <i2l/phylo_tree.h>
 
-#ifdef __clang__
+#ifdef __ARM_NEON
+
+#include <cmath>
+
+namespace epik::impl
+{
+    using lwr_type = long double;
+    extern long double (*powl)(long double, long double);
+}
+
+#elif defined(__clang__)
 /// Clang still does not fully support boost::multiprecion.
 /// We use floats instead despite of known precion-related
 /// issues in rare cases
