diff --git a/src/reference_mapping/rm_one_vol_main.mk b/src/reference_mapping/rm_one_vol_main.mk
index 414b4b0..a625625 100644
--- a/src/reference_mapping/rm_one_vol_main.mk
+++ b/src/reference_mapping/rm_one_vol_main.mk
@@ -11,7 +11,7 @@ SOURCES  := rm_one_vol_main.c rm_worker.c
 SRC_INCDIRS  := . 
 
 TGT_LDFLAGS := -L${TARGET_DIR}
-TGT_LDLIBS  := -lontcns
+TGT_LDLIBS  := -lontcns -lz -lpthread  -lstdc++
 TGT_PREREQS := libontcns.a
 
 SUBMAKEFILES :=
