diff --git a/src/trim_bases/largest_cover_range.mk b/src/trim_bases/largest_cover_range.mk
index 17134bf..9e89659 100644
--- a/src/trim_bases/largest_cover_range.mk
+++ b/src/trim_bases/largest_cover_range.mk
@@ -22,7 +22,7 @@ SOURCES  := pm4_aux.c \
 SRC_INCDIRS  := . 
 
 TGT_LDFLAGS := -L${TARGET_DIR}
-TGT_LDLIBS  := -lontcns
+TGT_LDLIBS  := -lontcns -lpthread
 TGT_PREREQS := libontcns.a
 
 SUBMAKEFILES :=
