diff --git a/src/fsa/filter.mk b/src/fsa/filter.mk
index cb13912..5b8c1cd 100644
--- a/src/fsa/filter.mk
+++ b/src/fsa/filter.mk
@@ -9,10 +9,10 @@ TARGET   := fsa_ol_filter
 SOURCES  := ./prog/fsa_ol_filter.cpp
 
 SRC_INCDIRS  := . 
-TGT_CXXFLAGS := -U_GLIBCXX_PARALLEL -std=c++11 -Wall -O3 -D_FILE_OFFSET_BITS=64
+TGT_CXXFLAGS := -U_GLIBCXX_PARALLEL -std=c++11 -Wall -O3 -D_FILE_OFFSET_BITS=64 ${CXXFLAGS}
 
 TGT_LDFLAGS := -L${TARGET_DIR}
-TGT_LDLIBS  := -lfsa
+TGT_LDLIBS  := -lfsa -lz -lpthread
 TGT_PREREQS := libfsa.a
 
 SUBMAKEFILES :=
