diff --git a/src/ctg_cns/filter_m4.mk b/src/ctg_cns/filter_m4.mk
index 822129e..77b441c 100644
--- a/src/ctg_cns/filter_m4.mk
+++ b/src/ctg_cns/filter_m4.mk
@@ -11,7 +11,7 @@ SOURCES  := filter_m4.c pm4.c load_ctg_read_ids.c
 SRC_INCDIRS  := .
 
 TGT_LDFLAGS := -L${TARGET_DIR}
-TGT_LDLIBS  := -lontcns
+TGT_LDLIBS  := -lontcns -lz -lpthread -lstdc++ -lm 
 TGT_PREREQS := libontcns.a
 
 SUBMAKEFILES :=