diff --git a/src/fsa/fsa.mk b/src/fsa/fsa.mk
index 267ccad..9d6166f 100644
--- a/src/fsa/fsa.mk
+++ b/src/fsa/fsa.mk
@@ -10,11 +10,11 @@ SOURCES  := argument_parser.cpp logger.cpp overlap.cpp read_store.cpp sequence.c
              utility.cpp file_io.cpp seq_io.cpp overlap_store.cpp \
    			 ./simple_align.cpp overlap_filter.cpp overlap_stat.cpp overlap_trim.cpp
 
-TGT_CXXFLAGS := -U_GLIBCXX_PARALLEL -std=c++11 -Wall -O3 -D_FILE_OFFSET_BITS=64 
-SRC_INCDIRS  := . 
+TGT_CXXFLAGS := ${TGT_CXXFLAGS} -U_GLIBCXX_PARALLEL -std=c++11 -Wall -O3 -D_FILE_OFFSET_BITS=64 ${CXXFLAGS}
+SRC_INCDIRS  := . -I${PREFIX}/include
 
 TGT_LDFLAGS := -L${TARGET_DIR}
-TGT_LDLIBS  := -lontcns
+TGT_LDLIBS  := -lontcns -lz
 TGT_PREREQS := libontcns.a
 
 SUBMAKEFILES :=
