diff --git a/src/consensus/main.mk b/src/consensus/main.mk
index a4bbb05..df1b004 100644
--- a/src/consensus/main.mk
+++ b/src/consensus/main.mk
@@ -11,7 +11,7 @@ SOURCES  := main.c read_id_pool.c cns_options.c error_estimate.c overlaps_pool.c
 SRC_INCDIRS  := . 
 
 TGT_LDFLAGS := -L${TARGET_DIR}
-TGT_LDLIBS  := -lontcns
+TGT_LDLIBS  := -lontcns -lz -lpthread -lstdc++ -lm
 TGT_PREREQS := libontcns.a
 
 SUBMAKEFILES :=
