diff --git a/src/fsa/assemble.mk b/src/fsa/assemble.mk
index 4713432..f1b8cf0 100644
--- a/src/fsa/assemble.mk
+++ b/src/fsa/assemble.mk
@@ -9,10 +9,10 @@ TARGET   := fsa_assemble
 SOURCES  := ./prog/fsa_assemble.cpp assembly.cpp graph.cpp string_graph.cpp path_graph.cpp
 
 SRC_INCDIRS  := . 
-TGT_CXXFLAGS := -U_GLIBCXX_PARALLEL -std=c++11 -Wall -O3 -D_FILE_OFFSET_BITS=64 
+TGT_CXXFLAGS := -U_GLIBCXX_PARALLEL -std=c++11 -Wall -O3 -D_FILE_OFFSET_BITS=64 ${CXXFLAGS}
 
 TGT_LDFLAGS := -L${TARGET_DIR}
-TGT_LDLIBS  := -lfsa
+TGT_LDLIBS  := -lfsa -lz -lpthread
 TGT_PREREQS := libfsa.a
 
 SUBMAKEFILES :=
