diff --git a/src/extract_sequences/extract_sequences.mk b/src/extract_sequences/extract_sequences.mk
index 266ace6..545ce49 100644
--- a/src/extract_sequences/extract_sequences.mk
+++ b/src/extract_sequences/extract_sequences.mk
@@ -11,7 +11,7 @@ SOURCES  := extract_sequences.c
 SRC_INCDIRS  := . 
 
 TGT_LDFLAGS := -L${TARGET_DIR}
-TGT_LDLIBS  := -lontcns
+TGT_LDLIBS  := -lontcns -lz
 TGT_PREREQS := libontcns.a
 
 SUBMAKEFILES :=
