diff --git a/src/reference_mapping/main.mk b/src/reference_mapping/main.mk
index 96530b3..391ee85 100644
--- a/src/reference_mapping/main.mk
+++ b/src/reference_mapping/main.mk
@@ -11,7 +11,7 @@ SOURCES  := main.c rm_worker.c
 SRC_INCDIRS  := . 
 
 TGT_LDFLAGS := -L${TARGET_DIR}
-TGT_LDLIBS  := -lontcns
+TGT_LDLIBS  := -lontcns -lz -lpthread -lstdc++
 TGT_PREREQS := libontcns.a
 
 SUBMAKEFILES :=
