diff --git a/src/trim_bases_accurate/pm4.mk b/src/trim_bases_accurate/pm4.mk
index 0d9d277..dcba7cf 100644
--- a/src/trim_bases_accurate/pm4.mk
+++ b/src/trim_bases_accurate/pm4.mk
@@ -11,7 +11,7 @@ SOURCES  := pm4_aux.c pm4_main.c
 SRC_INCDIRS  := . 
 
 TGT_LDFLAGS := -L${TARGET_DIR}
-TGT_LDLIBS  := -lontcns
+TGT_LDLIBS  := -lontcns -lpthread
 TGT_PREREQS := libontcns.a
 
 SUBMAKEFILES :=
