diff --git a/src/fsa/rd_extract.mk b/src/fsa/rd_extract.mk
index d23c3fe..5fb8bc6 100644
--- a/src/fsa/rd_extract.mk
+++ b/src/fsa/rd_extract.mk
@@ -9,10 +9,10 @@ TARGET   := fsa_rd_extract
 SOURCES  := ./prog/fsa_rd_extract.cpp read_extract.cpp
 
 SRC_INCDIRS  := . 
-TGT_CXXFLAGS := -U_GLIBCXX_PARALLEL -std=c++11 -Wall -O3 -D_FILE_OFFSET_BITS=64 
+TGT_CXXFLAGS := -U_GLIBCXX_PARALLEL -std=c++11 -Wall -O3 -D_FILE_OFFSET_BITS=64 ${CXXFLAGS}
 
 TGT_LDFLAGS := -L${TARGET_DIR}
-TGT_LDLIBS  := -lfsa
+TGT_LDLIBS  := -lfsa -lz 
 TGT_PREREQS := libfsa.a
 
 SUBMAKEFILES :=
