diff --git a/src/trim_bases/extract_trimmed_reads.mk b/src/trim_bases/extract_trimmed_reads.mk
index 8f493f1..4d06513 100644
--- a/src/trim_bases/extract_trimmed_reads.mk
+++ b/src/trim_bases/extract_trimmed_reads.mk
@@ -10,8 +10,8 @@ SOURCES  := extract_trimmed_reads.c largest_cover_range.c range_list.c detect_ch
 
 SRC_INCDIRS  := . 
 
+TGT_LDLIBS  := -lontcns -lz -lpthread
 TGT_LDFLAGS := -L${TARGET_DIR}
-TGT_LDLIBS  := -lontcns
 TGT_PREREQS := libontcns.a
 
 SUBMAKEFILES :=
