diff --git a/src/trim_bases_accurate/trim_bases.mk b/src/trim_bases_accurate/trim_bases.mk
index 5c2334c..225ff00 100644
--- a/src/trim_bases_accurate/trim_bases.mk
+++ b/src/trim_bases_accurate/trim_bases.mk
@@ -11,7 +11,7 @@ SOURCES  := trim_bases.c
 SRC_INCDIRS  := . 
 
 TGT_LDFLAGS := -L${TARGET_DIR}
-TGT_LDLIBS  := -lontcns
+TGT_LDLIBS  := -lontcns -lz
 TGT_PREREQS := libontcns.a
 
 SUBMAKEFILES :=
