diff --git a/src/trim_bases_accurate0/largest_cover_range.mk b/src/trim_bases_accurate0/largest_cover_range.mk
index f76048f..65d8d92 100644
--- a/src/trim_bases_accurate0/largest_cover_range.mk
+++ b/src/trim_bases_accurate0/largest_cover_range.mk
@@ -21,7 +21,7 @@ SOURCES  := pm4_aux.c \
 SRC_INCDIRS  := . 
 
 TGT_LDFLAGS := -L${TARGET_DIR}
-TGT_LDLIBS  := -lontcns
+TGT_LDLIBS  := -lontcns -lz -lpthread -lstdc++ -lm
 TGT_PREREQS := libontcns.a
 
 SUBMAKEFILES :=
