diff --git a/src/fsa/bridge.mk b/src/fsa/bridge.mk
index dbc50dc..004be32 100644
--- a/src/fsa/bridge.mk
+++ b/src/fsa/bridge.mk
@@ -9,10 +9,10 @@ TARGET   := fsa_ctg_bridge
 SOURCES  := ./prog/fsa_ctg_bridge.cpp contig_bridge.cpp contig_graph.cpp contig_link.cpp contig_link_store.cpp
 
 SRC_INCDIRS  := . 
-TGT_CXXFLAGS := -U_GLIBCXX_PARALLEL -std=c++11 -Wall -O3 -D_FILE_OFFSET_BITS=64 
+TGT_CXXFLAGS := -U_GLIBCXX_PARALLEL -std=c++11 -Wall -O3 -D_FILE_OFFSET_BITS=64 ${CXXFLAGS}
 
 TGT_LDFLAGS := -L${TARGET_DIR}
-TGT_LDLIBS  := -lfsa
+TGT_LDLIBS  := -lfsa -lz -lpthread
 TGT_PREREQS := libfsa.a
 
 SUBMAKEFILES :=
