diff --git a/tool/main.mk b/tool/main.mk
index 543b50e..7c3a20b 100644
--- a/tool/main.mk
+++ b/tool/main.mk
@@ -9,7 +9,7 @@ ${TARGET_DIR}/pigz: ${TARGET_DIR}/../pigz-2.4/pigz
 
 ${TARGET_DIR}/../pigz-2.4/pigz: ../tool/pigz-2.4.tar.gz
 	tar -xzf ../tool/pigz-2.4.tar.gz -C ${TARGET_DIR}/.. || exit 255
-	cd ${TARGET_DIR}/../pigz-2.4 && make
+	cd ${TARGET_DIR}/../pigz-2.4 && make CC=${CC} LDFLAGS="${LDFLAGS}"
 
 clean:
 	rm -f ${TARGET_DIR}/pigz
