diff --git a/gecode/third-party/boost/numeric/interval/hw_rounding.hpp b/gecode/third-party/boost/numeric/interval/hw_rounding.hpp
index 18ad513..b947235 100644
--- a/gecode/third-party/boost/numeric/interval/hw_rounding.hpp
+++ b/gecode/third-party/boost/numeric/interval/hw_rounding.hpp
@@ -33,7 +33,7 @@
 #  include <gecode/third-party/boost/numeric/interval/detail/ia64_rounding_control.hpp>
 #endif
 
-#if defined(GECODE_BOOST_NUMERIC_INTERVAL_NO_HARDWARE) && (defined(__USE_ISOC99) || defined(__MSL__))
+#if defined(GECODE_BOOST_NUMERIC_INTERVAL_NO_HARDWARE) && (defined(__USE_ISOC99) || defined(__MSL__) || (defined __cplusplus && __cplusplus >= 201103L))
 #  include <gecode/third-party/boost/numeric/interval/detail/c99_rounding_control.hpp>
 #endif
 
