diff --git a/makefile b/makefile
index 1a04794..2a9417c 100644
--- a/makefile
+++ b/makefile
@@ -7,11 +7,11 @@ KMC_DUMP_DIR = kmc_dump
 KMC_TOOLS_DIR = kmc_tools
 
 CXX 	?= g++
-CFLAGS	= -Wall -O3 -m64 -static -Wl,--whole-archive -lpthread -Wl,--no-whole-archive -std=c++11
-CLINK	= -lm -static -O3 -Wl,--whole-archive -lpthread -Wl,--no-whole-archive -std=c++11
+CFLAGS	= -Wall -O3 -m64 -Wl,--whole-archive -lpthread -Wl,--no-whole-archive -std=c++11
+CLINK	= -lm -O3 -Wl,--whole-archive -lpthread -Wl,--no-whole-archive -std=c++11
 
-KMC_TOOLS_CFLAGS	= -Wall -O3 -m64 -static -Wl,--whole-archive -lpthread -Wl,--no-whole-archive -std=c++14
-KMC_TOOLS_CLINK	= -lm -static -O3 -Wl,--whole-archive -lpthread -Wl,--no-whole-archive -std=c++14
+KMC_TOOLS_CFLAGS	= -Wall -O3 -m64 -Wl,--whole-archive -lpthread -Wl,--no-whole-archive -std=c++14
+KMC_TOOLS_CLINK	= -lm -O3 -Wl,--whole-archive -lpthread -Wl,--no-whole-archive -std=c++14
 
 DISABLE_ASMLIB ?= false
 
@@ -95,14 +95,14 @@ $(KMC_MAIN_DIR)/instrset_detect.o: $(KMC_MAIN_DIR)/libs/vectorclass/instrset_det
 
 kmc: $(KMC_OBJS) $(RADULS_OBJS) $(KMC_MAIN_DIR)/instrset_detect.o
 	-mkdir -p $(KMC_BIN_DIR)
-	$(CXX) -o $(KMC_BIN_DIR)/$@ $^ $(KMC_LIBS) $(LDFLAGS) $(CLINK)
+	$(CXX) -o $(KMC_BIN_DIR)/$@ $^ -Wl,-Bstatic $(KMC_LIBS) -Wl,-Bdynamic $(LDFLAGS) $(CLINK)
 kmc_dump: $(KMC_DUMP_OBJS) $(KMC_API_OBJS)
 	-mkdir -p $(KMC_BIN_DIR)
 	$(CXX) -o $(KMC_BIN_DIR)/$@ $^ $(LDFLAGS) $(CLINK)
 
 kmc_tools: $(KMC_TOOLS_OBJS) $(KMC_API_OBJS)
 	-mkdir -p $(KMC_BIN_DIR)
-	$(CXX) -o $(KMC_BIN_DIR)/$@ $^ $(KMC_TOOLS_LIBS) $(LDFLAGS) $(KMC_TOOLS_CLINK)
+	$(CXX) -o $(KMC_BIN_DIR)/$@ $^ -Wl,-Bstatic $(KMC_TOOLS_LIBS) -Wl,-Bdynamic $(LDFLAGS) $(KMC_TOOLS_CLINK)
 
 clean:
 	-rm $(KMC_MAIN_DIR)/*.o
