diff --git a/install_pkg.sh b/install_pkg.sh
index 93cfe86..a5cde4a 100755
--- a/install_pkg.sh
+++ b/install_pkg.sh
@@ -30,7 +30,7 @@ export SKIP_DIAMOND=${SKIP_DIAMOND=0}
 export CPU_COUNT=${CPU_COUNT=$(nproc --ignore 1)}
 
 # destination prefic directory
-export PREFIX=${PREFIX="/usr/local"}
+export PREFIX="${PREFIX}"
 
 # get orthologer and brhclus
 export DIR_ORTHOLOGER_PKG=$(find ./ -maxdepth 1 -name "ORTHOLOGER-*" -type d)
@@ -47,8 +47,8 @@ fi
 DIR_ORTHOLOGER_PKG=$( cd -P $DIR_ORTHOLOGER_PKG && pwd)
 DIR_BRHCLUS_PKG=$( cd -P $DIR_BRHCLUS_PKG && pwd)

-export DIR_BRHCLUS_INSTALL=""
-export DIR_ORTHOLOGER_INSTALL=""
+export DIR_BRHCLUS_INSTALL="${PREFIX}"
+export DIR_ORTHOLOGER_INSTALL="${PREFIX}"
 
 print_config() {
     echo "---------------------------------------------------------------"
