--- src/libfaketime.c.orig	2025-02-11 11:34:55.053576612 +0000
+++ src/libfaketime.c	2025-02-11 11:35:38.052192655 +0000
@@ -191,7 +191,7 @@
 static int          (*real_lxstat64)        (int, const char *, struct stat64 *);
 static time_t       (*real_time)            (time_t *);
 static int          (*real_ftime)           (struct timeb *);
-static int          (*real_gettimeofday)    (struct timeval *, void *);
+static int          (*real_gettimeofday)    (struct timeval *, struct timezone *);
 static int          (*real_clock_gettime)   (clockid_t clk_id, struct timespec *tp);
 static int          (*real_timespec_get)    (struct timespec *ts, int base);
 #ifdef FAKE_INTERNAL_CALLS
