--- Makefile
+++ Makefile
@@ -1,7 +1,7 @@
 # Comment with icc
-CC=gcc
-CFLAGS= -O3 -Wall -Wshadow -pedantic -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -DVER32 \
-	-I/opt/local/include/   #-g  #-p  #-m64
+CC?=gcc
+CFLAGS+= -O3 -Wall -Wshadow -pedantic -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -DVER32
+
 
 # Comment with gcc
 #CC=icc
@@ -23,7 +23,7 @@
 all : $(EXECFILE)
 
 $(EXECFILE) : $(OBJFILES)
-	$(CC) $(CFLAGS) -o $(EXECFILE) $(OBJFILES) $(LIBS)	
+	$(CC) $(CFLAGS) $(CPPFLAGS) $(LDFLAGS) -o $(EXECFILE) $(OBJFILES) $(LIBS)	
 
 doc:
 	cd ../Doc; make clean; make pdf; cd ../$(DIRECTORY)_$(VERSION)
