--- SOAPdenovo-Trans/Makefile	2014-04-15 04:37:16.000000000 +0000
+++ SOAPdenovo-Trans/Makefile.new	2017-03-02 22:43:31.155687817 +0000
@@ -1,5 +1,5 @@
 CC=             gcc
-CFLAGS=         -O3 -fomit-frame-pointer -static
+CFLAGS+=         -O3 -fomit-frame-pointer
 #CFLAGS=         -O3 -g -D_DEBUG
 DFLAGS= 
 
@@ -33,5 +33,5 @@
 SUBDIRS=        . 
 LIBPATH=	
-LIBS=      	-pthread -lm -lrt -lbam -lz -L./inc
+LIBS=      	-pthread -lm -lrt -lbam -lz
 EXTRA_FLAGS=
 
