--- MANIFEST.in
+++ MANIFEST.in
@@ -1 +1 @@
-include README* LICENSE requirements.txt goatools/test_data/nbt_3102/* requirements.txt setup_helper.py
+include README* LICENSE requirements.txt requirements.txt setup_helper.py
--- setup.py
+++ setup.py
@@ -35,5 +35,4 @@
     long_description=h.long_description,
-    packages=[name, name + ".test_data", name + ".anno"],
+    packages=[name, name + ".anno"],
     include_package_data=True,
-    package_data={"goatools.test_data.nbt_3102": ["*.*"]},
     scripts=glob('scripts/*.py'),
