diff --git a/setup.py b/setup.py
index 22d1ec6..668d4fc 100755
--- a/setup.py
+++ b/setup.py
@@ -43,8 +43,6 @@ setup(
     author_email='stef@tbi.univie.ac.at',
     url='https://github.com/bad-ants-fleet/ribolands',
     install_requires=['matplotlib', 'networkx', 'future', 'pandas', 'crnsimulator>=0.4'],
-    dependency_links=[
-        'https://github.com/bad-ants-fleet/crnsimulator/tarball/master#egg=crnsimulator-0.4'],
     test_suite='tests',
     packages=['ribolands'],
     scripts=['scripts/BarMap.py',
