diff --git a/makefile b/makefile
index 21492fe..009c44f 100644
--- a/makefile
+++ b/makefile
@@ -73,7 +73,7 @@ ifeq ($(DYNAMIC_LINK), true)
 else
 ifeq ($(uname_S),Darwin)
     CFLAGS  = -Wall -O3 $(ARCH_FLAGS) -std=c++17 $(DEFINE_FLAGS) $(INCLUDES)
-    CLINK	= -lm -O3 -std=c++17 $(ABI_FLAGS) -static-libgcc
+    CLINK	= -lm -O3 -std=c++17 $(ABI_FLAGS)
 else
     CFLAGS  = -Wall -O3 $(ARCH_FLAGS) -std=c++17 $(DEFINE_FLAGS) $(INCLUDES) -static -Wl,--whole-archive -lpthread -Wl,--no-whole-archive
     CLINK	= -lm -static -O3 -std=c++17 $(ABI_FLAGS) -Wl,--whole-archive -lpthread -Wl,--no-whole-archive
