diff --git a/Makefile b/Makefile
index 55b0557..93f1d54 100644
--- a/Makefile
+++ b/Makefile
@@ -62,7 +62,7 @@ clean:
 	cd Zoe; make clean
 
 depend: $(OBJECTS:.o=.c)
-	gcc $(INC) -MM $^ > $@
+	$(CC) $(INC) -MM $^ > $@
 
 tar:
 	rm -rf /tmp/$(APP)
@@ -79,9 +79,9 @@ tar:
 
 gcc:
 	cd Zoe; make;
-	make $(APP)  CC="gcc" CFLAGS="-O2 -Wall -Werror"
-	make $(APP2) CC="gcc" CFLAGS="-O2 -Wall -Werror"
-	make $(APP3) CC="gcc" CFLAGS="-O2 -Wall -Werror"
+	make $(APP)  CC="$(CC)" CFLAGS="-O3 -Wall -Werror"
+	make $(APP2) CC="$(CC)" CFLAGS="-O3 -Wall -Werror"
+	make $(APP3) CC="$(CC)" CFLAGS="-O3 -Wall -Werror"
 
 ###################
 # Inference Rules #
diff --git a/Zoe/Makefile b/Zoe/Makefile
index 2152de4..823b108 100644
--- a/Zoe/Makefile
+++ b/Zoe/Makefile
@@ -63,7 +63,7 @@ tar:
 #################
 
 gcc:
-	make $(APP)  CC="gcc" CFLAGS="-O2 -Wall -Werror"
+	make $(APP)  CC="$(CC)" CFLAGS="-O3 -Wall -Werror"
 
 ###################
 # Inference Rules #
