diff --git a/Makefile b/Makefile
index 55b0557..699bc77 100644
--- a/Makefile
+++ b/Makefile
@@ -62,7 +62,7 @@ clean:
 	cd Zoe; make clean
 
 depend: $(OBJECTS:.o=.c)
-	gcc $(INC) -MM $^ > $@
+	$(CC) $(INC) -MM $^ > $@
 
 tar:
 	rm -rf /tmp/$(APP)
@@ -78,10 +78,10 @@ tar:
 #################
 
 gcc:
-	cd Zoe; make;
-	make $(APP)  CC="gcc" CFLAGS="-O2 -Wall -Werror"
-	make $(APP2) CC="gcc" CFLAGS="-O2 -Wall -Werror"
-	make $(APP3) CC="gcc" CFLAGS="-O2 -Wall -Werror"
+	cd Zoe; make CC=$(CC);
+	make $(APP)  CC=$(CC) CFLAGS="-O3 -Wall -Werror"
+	make $(APP2) CC=$(CC) CFLAGS="-O3 -Wall -Werror"
+	make $(APP3) CC=$(CC) CFLAGS="-O3 -Wall -Werror"
 
 ###################
 # Inference Rules #
diff --git a/Zoe/Makefile b/Zoe/Makefile
index 2152de4..d7b24f3 100644
--- a/Zoe/Makefile
+++ b/Zoe/Makefile
@@ -47,7 +47,7 @@ clean:
 	rm -f *.o $(APP) $(APP2) $(APP3) $(APP4) $(APP5)
 
 depend: $(OBJECTS:.o=.c)
-	gcc -MM $^ > $@
+	$(CC) -MM $^ > $@
 
 tar:
 	rm -rf /tmp/$(APP)
@@ -63,7 +63,7 @@ tar:
 #################
 
 gcc:
-	make $(APP)  CC="gcc" CFLAGS="-O2 -Wall -Werror"
+	make $(APP) CC=$(CC) CFLAGS="-O3 -Wall -Werror"
 
 ###################
 # Inference Rules #
