--- Zoe/Makefile.bak	2019-05-16 14:46:08.804230543 +0200
+++ Zoe/Makefile	2019-05-17 09:58:46.950537640 +0200
@@ -92,14 +92,14 @@
 #################
 
 gcc:
-	make $(APP)  CC="gcc" CFLAGS="-O2 -Wall -Werror"
+	make $(APP)  CC=$(CC) CFLAGS="-O2 -Wall"
 
 all:
-	make $(APP)  CC="gcc" CFLAGS="-O2 -Wall -Werror"
-	make $(APP2) CC="gcc" CFLAGS="-O2 -Wall -Werror"
-	make $(APP3) CC="gcc" CFLAGS="-O2 -Wall -Werror"
-	make $(APP4) CC="gcc" CFLAGS="-O2 -Wall -Werror"
-	make $(APP5) CC="gcc" CFLAGS="-O2 -Wall -Werror"
+	make $(APP)  CC=$(CC) CFLAGS="-O2 -Wall"
+	make $(APP2) CC=$(CC) CFLAGS="-O2 -Wall"
+	make $(APP3) CC=$(CC) CFLAGS="-O2 -Wall"
+	make $(APP4) CC=$(CC) CFLAGS="-O2 -Wall"
+	make $(APP5) CC=$(CC) CFLAGS="-O2 -Wall"
 
 
 ###################
