diff --git a/gofigure.py b/gofigure.py
index 0517691..651e494 100755
--- a/gofigure.py
+++ b/gofigure.py
@@ -1,3 +1,4 @@
+#!/usr/bin/env python3
 ##
 ## Summary Visualizations of Gene Ontology Enrichment with GO-Figure!
 ##
@@ -379,7 +380,7 @@ def process_input(input_file_path,input_type):
 def read_IC(scriptPath):
 	ic_dict = defaultdict(float)
 	frequency_dict = defaultdict(float)
-	for line in open(scriptPath+'/data/ic.tsv'):
+	for line in open(scriptPath+'/ic.tsv'):
 		if not line.startswith('#'):
 			go,ic,frequency = line.strip().split('\t')
 			ic_dict[go] = float(ic)
@@ -394,7 +395,7 @@ def read_OBO(scriptPath):
 	namespace_dict = defaultdict(str)
 	obsolete_dict = defaultdict(set)
 	alt_dict = defaultdict(str)
-	for line in open(scriptPath+'/data/go.obo'):
+	for line in open(scriptPath+'/go.obo'):
 		if line.startswith('id: GO:'):
 			go = line.strip().split('id: ')[1]
 			obsolete_bool = False
@@ -423,7 +424,7 @@ def read_OBO(scriptPath):
 def read_parents_children(scriptPath):
 	parents_dict = defaultdict(list)
 	children_dict = defaultdict(list)
-	for line in open(scriptPath+'/data/relations_full.tsv'):
+	for line in open(scriptPath+'/relations_full.tsv'):
 		if not line.startswith('#'):
 			go,parent = line.strip().split('\t')
 			parents_dict[go].append(parent)
