diff --git a/CMakeLists.txt b/CMakeLists.txt
index 00fc467..6ca437f 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -189,7 +189,7 @@ if(CUDA)
     # -----------------------------------------------------------------------------CUDA--
     # DOC: http://www.cmake.org/cmake/help/v3.0/module/FindCUDA.html
     include(${CMAKE_SOURCE_DIR}/cmake/BuildTypes.cmake)
-    FIND_PACKAGE(CUDA)
+    find_package(CUDAToolkit)
 
     if(CUDA_FOUND)
       message(STATUS "Using cuda wrapper to compile....")
